library(shiny)
library(shinyjs)
library(shinyBS)
library(shinyalert)
library(shinydashboard)
library(dashboardthemes)
library(shinythemes)
library(shinybusy)
library(shinycssloaders)
library(shinydashboardPlus)
library(RColorBrewer)
library(DT)
library(plotly)
library(shinyWidgets)
library(HGNChelper)
library(openxlsx)
library(stringi)
library(cowplot)

####GLOBAL ENVIROMENT####
#Colours
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

#Spectrals colours
Spectrals <- rev(brewer.pal(n = 11, name = "Spectral"))

# Options for Spinner
options(spinner.color="#0275D8", spinner.color.background="#ffffff", spinner.size=3)

#Java code
options(java.parameters = "-Xss2560k")

#Shinyjs code
jsboxcollapsecode <- "shinyjs.collapse = function(boxid) {
$('#' + boxid).closest('.box').find('[data-widget=collapse]').click();
}"

collapseInput <- function(inputId, boxId) {
  tags$script(
    sprintf(
      "$('#%s').closest('.box').on('hidden.bs.collapse', function () {Shiny.onInputChange('%s', true);})",
      boxId, inputId
    ),
    sprintf(
      "$('#%s').closest('.box').on('shown.bs.collapse', function () {Shiny.onInputChange('%s', false);})",
      boxId, inputId
    )
  )
}

#Inactivity timer

inactivity <- sprintf("function idleTimer() {
var t = setTimeout(logout, %s);
window.onmousemove = resetTimer; // catches mouse movements
window.onmousedown = resetTimer; // catches mouse movements
window.onclick = resetTimer;     // catches mouse clicks
window.onscroll = resetTimer;    // catches scrolling
window.onkeypress = resetTimer;  //catches keyboard actions

function logout() {
Shiny.setInputValue('timeOut', '%ss')
}

function resetTimer() {
clearTimeout(t);
t = setTimeout(logout, %s);  // time is in milliseconds (1000 is 1 second)
}
}
idleTimer();", 10800*1000, 10800, 10800*1000)

#Upload size restriction
options(shiny.maxRequestSize = 100000*1024^2)


###Dashboard UI####
sidebar <- dashboardSidebar(
  width = 250,
  
  #Use latest font-awesome icons
  tags$style("@import url(https://use.fontawesome.com/releases/v6.1.1/css/all.css);"),
  
  tags$style(HTML(
    ".pretty > div > label {font-size: 14px !important; line-height: 1px !important;}"
  )),
  
  tags$head(
    tags$style(HTML("
    
  .prevent_click{
  position:fixed; 
  z-index:999;
  width:200%;
  height:200%;
  background-color: transparent;   
  }"))
  ),
  
  conditionalPanel(
    condition = "$(\'html\').hasClass(\'shiny-busy\')",
    tags$div(class = "prevent_click")
  ),
  
  sidebarMenu(id = "orig_tabs",
              menuItem("Introduction", icon = icon("info"), tabName = "intro"),
              sidebarMenuOutput("dynamic_content"),
              menuItem("Help", icon = icon("question-circle"), tabName = "Help")
  )
)

body <- dashboardBody(
  useShinyjs(),
  extendShinyjs(text = jsboxcollapsecode, functions = c("collapse")),
  add_busy_spinner(spin = "fading-circle", position = "bottom-left", color = "#337ab7", height = "100px", width = "100px", margins = c(10,60)),
  tags$head(tags$style(HTML(".small-box {height: 400px; background-color: #e8ebe9 !important; color: #000000 !important; }"))),
  
  tags$head(tags$style(HTML('
  .navbar-custom-menu>.navbar-nav>li>.dropdown-menu {
  width:550px;
  }
  '))),
  
  tags$head(tags$style(
    "html {height: 100%; overflow-y: scroll;}"
  )),
  
  tags$head(
    tags$style(HTML("
    
  .prevent_click{
  position:fixed; 
  z-index:999;
  width:200%;
  height:200%;
  background-color: transparent;   
  }"))
  ),
  
  tags$script(inactivity),
  
  # display load spinner when shiny is busy
  conditionalPanel(
    condition = "$(\'html\').hasClass(\'shiny-busy\')",
    tags$div(class = "prevent_click")
  ),
  
  tabItems(
    
    ####TAB 0####
    tabItem(tabName = "intro",
            
            box(id = "IntroBox", title = h2(HTML('<h2 style = "text-align:justify; margin-top:-50px; margin-bottom:-10px; ">Welcome to ICARUS (<b>I</b>nteractive single <b>C</b>ell RNA-seq <b>A</b>nalysis with <b>R</b> shiny <b>U</b>sing <b>S</b>eurat)<br>')), status = "warning", solidHeader = F, 
                collapsible = T, width = 12,
                
                h4(HTML('<h4 style = "text-align:justify;"> This web server was designed to guide the user through single cell RNA-seq analysis using the
              <a href="https://www.satijalab.org/seurat/" target="_blank">Seurat scRNA-seq analysis toolkit</a>
              via a tutorial style interface. It offers user control over each of the steps to personalise analysis based on the dataset of interest.
              Graphical outputs at each analysis step ensures easy and logical interpretation.
              <br> <br>
              The purpose of this application is to allow the user to interactively visualize single cell RNA-seq data without previous R programming knowledge.
              <br> <br>')),
                img(id="ICARUS",src="ICARUS.png", width="30%", style = "float:right;"),
                
                h4(HTML('<h4 style = "text-align:justify;">
                    Features include: <br>
                    <ol><li>Tutorial inspired user interface!</li>
                    <li>Support for 11 common species!</li>
                    <li>Apply your own cell quality control thresholds and remove cells with low quality (i.e. high mitochondrial percentage)!</li>
                    <li>Removal of cell doublets (multiplets) with <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">DoubletFinder</a>! </li>
                    <li>Adjust your own dimensionality reduction and clustering parameters!</li>
                    <li>3D UMAP and t-SNE plots!</li>
                    <li>Data correction for cell cycle effects!</li>
                    <li>Labelling of cell clusters with <a href="https://doi.org/10.1038/s41467-022-28803-w" target="_blank">sctype</a> and
                    <a href="https://doi.org/10.1038/s41590-018-0276-y" target="_blank">SingleR</a>! </li>
                    <li>Gene expression and gene pathway visualisation!</li>
                    <li>Gene co-expression analysis with <a href="https://doi.org/10.1371/journal.pcbi.1004574" target="_blank">MEGENA</a>! </li>
                    <li>Gene regulatory network identification with <a href="https://doi.org/10.1038/nmeth.4463" target="_blank">SCENIC</a>! </li>
                    <li>Examine cell cluster expression association with GWAS traits using <a href="https://doi.org/10.1371/journal.pcbi.1004219" target="_blank">MAGMA</a>! </li>
                    <li>Trajectory analysis with <a href="https://doi.org/10.1038/nbt.2859" target="_blank">Monocle3</a>!</li>
                    <li>Visualize and analyse cell-cell communication networks with <a href="https://doi.org/10.1038/s41467-021-21246-9" target="_blank">CellChat</a>!</li>
                    <li>Examine drug-gene interactions of differentially expressed genes against the <a href="https://www.dgidb.org" target="_blank">DGIdb 4.0 database</a>!</li>
                    <li>Differential expression analysis and gene set enrichment analysis with
                    <a href="https://doi.org/10.1016/j.xinn.2021.100141" target="_blank">ClusterProfiler</a> and
                    <a href="https://doi.org/10.1039/C5MB00663E" target="_blank">ReactomePA</a>!</li>
                    <li>Custom differential expression analysis with user selected cell groups to compare! </li>
                    <li>Integration with second dataset and adjustment for batch effects!</li>
                    <li>Support for multimodal analysis (i.e. CITE-seq, 10X multiome kit)! </li>
                    <li>Save and continue functionality!</li>
                    <li>Downloadable tables and plots!</li></ol>')),
                
                h4(HTML('<h4 style = "text-align:justify;color:#FF0000">Please refer to the "Help" tab on the sidebar menu for troubleshooting.')),
                br(),
                fluidRow(column(3,
                                valueBox("NEW", p(pickerInput("species1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose from the following species')), choices = c("Homo sapiens", "Bos taurus", "Caenorhabditis elegans", "Canis lupus familiaris", "Danio rerio",
                                                                                                                                                                                                       "Drosophila melanogaster", "Gallus gallus", "Mus musculus", "Rattus norvegicus", "Saccharomyces cerevisiae", "Sus scrofa"), 
                                                              selected = "Homo sapiens"),
                                                  actionBttn("startbuttonintro", "START", 
                                                             style = "jelly",
                                                             color = "primary"),
                                                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"><b>PRESS START TO BEGIN</b><br>'))), icon = icon("paper-plane"), width = 12, color = "aqua")),
                         column(3,       
                                valueBox("CONTINUE",
                                         fileInput("fileinputLOAD", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load previously saved enviroment')), accept = c(".Rdata")),
                                         icon = icon("upload"), width = 12, color = "aqua"))
                ))
    ),
    
    ####TAB 0.5 - LOGS####
    tabItem(tabName = "Logs",
            box(id = "LogsBox", title = strong("Logs"), status = "warning", solidHeader = F, 
                collapsible = T, width = 12,
                htmlOutput("textlogsUI")),
            
            box(id = "LoadedPackagesBox", title = strong("Loaded Packages"), status = "warning", solidHeader = F, 
                collapsible = T, width = 12,
                verbatimTextOutput("sessionInfo")
            )
    ), 
    
    
    ####HELP TAB####
    tabItem(tabName = "Help",
            
            box(id = "HelpBox", title = strong("Help"), status = "warning", solidHeader = F, 
                collapsible = T, width = 12,
                
                h2(HTML('<h2 style = "text-align:justify;color:#000000; margin-top:-20px;"><b>Contact Us</b>')),
                
                h4(HTML('<h4 style = "text-align:justify;"> <b>Version of the application: </b> ICARUS v2.4')),
                actionBttn("updates2", "See what's new in ICARUS v2.4", size = "sm"),
                h4(HTML('<h4 style = "text-align:justify;">ICARUS was designed and maintained by Andrew Jiang based at the Applied Translational Genetics Group, The University of Auckland, Auckland, New Zealand. <br>
                    Please forward any queries to <a href = "mailto: ajia169@aucklanduni.ac.nz">ajia169@aucklanduni.ac.nz</a> or visit the <a href="https://github.com/Enjewl/ICARUS" target="_blank">Github page</a>.')), br(), 
                
                h2(HTML('<h2 style = "text-align:justify;color:#000000; margin-top:-20px;"><b>Citation</b>')),
                
                h4(HTML('<h4 style = "text-align:justify;">If you used ICARUS for your research, please cite the following publication: <br><br>
                Jiang, A., Lehnert, K., You, L. and Snell, R.G. (2022) ICARUS, an interactive web server for single cell RNA-seq analysis. Nucleic Acids Research. https://doi.org/10.1093/nar/gkac322 <br><br>
                        Where applicable, please also cite the relevant R package(s) that ICARUS draws functionality from.')), br(),
                
                h2(strong("Troubleshooting")),
                
                h4(HTML('<h4 style = "text-align:justify;">Please refer to the overview diagram for an outline of each of the analysis steps (Fig.1). 
            For general navigation and usage, please refer to Fig. 2. The save and continue function is detailed in Fig. 3.')),
                
                fluidRow(column(4,
                                h4(HTML('<h4 style = "text-align:justify;"><b>Figure 1 </b>Overview diagram of each analysis step (Click to enlarge).')),
                                img(id="Flow_chart",src="Flow_chart.png",width="100%",style="cursor:pointer;")
                ),
                column(4,
                       h4(HTML('<h4 style = "text-align:justify;"><b>Figure 2 </b>General Navigation (Click to enlarge).')),
                       img(id="Navigation",src="Navigation.png",width="100%",style="cursor:pointer;")
                ),
                column(4,
                       h4(HTML('<h4 style = "text-align:justify;"><b>Figure 3 </b>Save and Continue interface (Click to enlarge).')),
                       img(id="SaveLoad",src="SaveLoad.png",width="100%",style="cursor:pointer;")
                ))
            )
    ), 
    
    ####TAB #1 - Load 1st Dataset####
    tabItem(tabName = "Load",
            fluidRow(
              box(id = "LoadBoxIntro", title = strong("Load your data"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify">A table of gene counts (matrix of UMI counts for each gene per cell) is required for ICARUS. Datasets with single or multiple samples/replicates may be loaded.
                  <br><br>Select between 3 methods of data input:<br>
                  <ol><li><b>Tab delimited table </b>with cells as columns and gene features as rows (see example A). For datasets with multiple samples, each sample can be denoted by an identifier separated by a underscore (Please do not include underscores for single sample datasets).</li>
                  <li><b>10X data</b> files (barcodes.tsv, features.tsv and matrix.mtx, refer to C).</li>
                  <li>Saved <b>Seurat object </b> (RDS file, refer to D)</li></ol><br>')),
                  
                  img(id="Load_ICARUS",src="Load_ICARUS.png",width="60%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonLoad", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua"))
                  ))),
            
            fluidRow(
              column(3,
                     box(
                       id = "LoadBox1", title = "Select single or multiple samples", status = "warning", solidHeader = F, 
                       collapsible = T, width = NULL,
                       
                       pickerInput("singlemultiple", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Does your dataset have single or multiple samples?')), 
                                   choices = c("Please select from the following" = "", "Single", "Multiple"))
                     ),
                     
                     box(
                       id = "LoadBox2", title = "Load your data", status = "warning", solidHeader = F, 
                       collapsible = T, width = NULL,
                       
                       prettyRadioButtons("radioLoad", label = h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose your load method (Single sample)')),
                                          choices = list("10X data", "Seurat object (RDS file)",
                                                         "Count Matrix (txt)", "Sample datasets"), 
                                          selected = "Sample datasets",
                                          icon = icon("check"),
                                          animation = "jelly"),      
                       
                       fileInput("textdirectory", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load barcodes.tsv, features.tsv and matrix.mtx files (can be gzipped .gz)')), 
                                 multiple = T),
                       
                       fileInput("fileinput1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load seurat object (RDS file format)')), 
                                 accept = c(".rds", ".RDS")),
                       
                       fileInput("fileinput2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load count matrix (tab delimited table of gene counts)')),
                                 accept = c("text",
                                            ".txt")),
                       
                       pickerInput("existingdata", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose from the following datasets')),
                                   choices = c("Please select from the following" = "", "3k PBMCs (Seurat - Guided Clustering Tutorial, 10X data)",
                                               "500 PBMCs (10X data)",
                                               "5k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone (10X data)",
                                               "8k CBMCs (Seurat CITE-seq multimodal tutorial)")),
                       actionBttn("buttonexistingdata", "Submit",
                                  style = "jelly",
                                  color = "primary")
                       
                     ),
                     
                     box(
                       id = "LoadBox5", title = "Load your data", status = "warning", solidHeader = F, 
                       collapsible = T, width = NULL,
                       
                       prettyRadioButtons("radioLoadM", label = h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose your load method (Multiple Samples)')),
                                          choices = list("Seurat object (RDS file)", "Count Matrix (txt)", "Sample datasets"), 
                                          selected = "Sample datasets",
                                          icon = icon("check"),
                                          animation = "jelly"),      
                       
                       fileInput("fileinput1M", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load seurat object (RDS file format)')), 
                                 accept = c(".rds", ".RDS")),
                       
                       fileInput("fileinput2M", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load count matrix (tab delimited table of gene counts)')),
                                 accept = c("text",
                                            ".txt")),
                       
                       pickerInput("existingdataM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose from the following datasets')), choices = c("Please select from the following" = "",
                                                                                                                                                                                  "10k Human PBMCs Multiplexed (CellPlex, 10X data)")),
                       actionBttn("buttonexistingdataM", "Submit",
                                  style = "jelly",
                                  color = "primary")
                       
                     ),
                     
                     box(id = "LoadBox3", title = "Label your experiment", status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         textInput("textlabel", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Sample Name:')), 
                                   value = "Sample A"),
                         numericInput("mincells", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Include genes detected in at least this many cells:')),
                                      min = 0, max = 1000000,
                                      value = "3"),
                         numericInput("minfeatures", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Include cells where at least this many genes are detected:')),
                                      min = 0, max = 1000000,
                                      value = "200")),
                     p(style="text-align: center;", actionBttn("NextButtonLoad", "CONTINUE TO NEXT STEP...",
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "LoadBox4", title = "Count Matrix (preview first 500 rows and columns)", status = "warning", 
                         collapsible = T, width = NULL,
                         
                         withSpinner(DT::dataTableOutput("MatrixTable", height = "600"))
                     ))
            ), collapseInput(inputId = "iscollapseboxintro", boxId = "LoadBoxIntro")
    ),
    
    ####TAB 1.5 - Load 2nd Dataset###########
    tabItem(tabName = "Combine",
            fluidRow(
              box(id = "CombineBoxIntro", title = strong("Load second dataset"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify">This step allows the user to load a second dataset that can be integrated with the first dataset.
                          The identical load options are available as previously (refer to Load your data tab). Datasets with single or multiple samples/replicates may be loaded.
                  <br><br>Select between 3 methods of data input:<br>
                  <ol><li><b>Tab delimited table </b>with cells as columns and gene features as rows (see example A). For datasets with multiple samples, each sample can be denoted by an identifier separated by a underscore (Please do not include underscores for single sample datasets).</li>
                  <li><b>10X data</b> files (barcodes.tsv, features.tsv and matrix.mtx, refer to C).</li>
                  <li>Saved <b>Seurat object </b> (RDS file, refer to D)</li></ol><br>')),
                  
                  img(id="Load2_ICARUS",src="Load_ICARUS.png",width="60%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonCombine", "START",
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")),
                           column(3,       
                                  valueBox("SKIP",
                                           p(actionBttn("skipbuttonCombine", "SKIP", 
                                                        style = "jelly",
                                                        color = "primary")),
                                           icon = icon("forward"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(
                       id = "CombineBox1", title = "Select single or multiple samples", status = "warning", solidHeader = F, 
                       collapsible = T, width = NULL,
                       pickerInput("singlemultipleCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Does your dataset have single or multiple samples?')), 
                                   choices = c("Please select from the following" = "", "Single", "Multiple"))
                     ),
                     
                     box(
                       id = "CombineBox2", title = "Load your data", status = "warning", solidHeader = F, 
                       collapsible = T, width = NULL,
                       
                       prettyRadioButtons("radioCombine", label = h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose your load method (Single Sample)')),
                                          choices = list("10X data", "Seurat object (RDS file)",
                                                         "Count Matrix (txt)", "Sample datasets"), 
                                          selected = "Sample datasets",
                                          icon = icon("check"),
                                          animation = "jelly"),  
                       
                       fileInput("textdirectoryCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load barcodes.tsv, features.tsv and matrix.mtx files (can be gzipped .gz)')), 
                                 multiple = T),
                       
                       fileInput("fileinput1Combine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load seurat object (RDS file format)')), 
                                 accept = c(".rds", ".RDS")),
                       
                       fileInput("fileinput2Combine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load count matrix (tab delimited table of gene counts)')),
                                 accept = c("text",
                                            ".txt")),
                       
                       pickerInput("existingdataCombine",h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose from the following datasets')), choices = c("Please select from the following" = "", "3k PBMCs (Seurat - Guided Clustering Tutorial, 10X data)",
                                                                                                                                                                                       "500 PBMCs (10X data)",
                                                                                                                                                                                       "5k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone (10X data)",
                                                                                                                                                                                       "8k CBMCs (Seurat CITE-seq multimodal tutorial)")),
                       actionBttn("buttonexistingdataCombine", "Submit",
                                  style = "jelly",
                                  color = "primary")
                     ),
                     
                     box(
                       id = "CombineBox5", title = "Load your data", status = "warning", solidHeader = F, 
                       collapsible = T, width = NULL,
                       
                       prettyRadioButtons("radioMCombine", label = h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose your load method (Multiple Samples)')),
                                          choices = list("Seurat object (RDS file)", "Count Matrix (txt)", "Sample datasets"), 
                                          selected = "Sample datasets",
                                          icon = icon("check"),
                                          animation = "jelly"),      
                       
                       fileInput("fileinput1MCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load seurat object (RDS file format)')), 
                                 accept = c(".rds", ".RDS")),
                       
                       fileInput("fileinput2MCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load count matrix (tab delimited table of gene counts)')),
                                 accept = c("text",
                                            ".txt")),
                       
                       pickerInput("existingdataMCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose from the following datasets')), choices = c("Please select from the following" = "",
                                                                                                                                                                                         "10k Human PBMCs Multiplexed (CellPlex, 10X data)")),
                       actionBttn("buttonexistingdataMCombine", "Submit",
                                  style = "jelly",
                                  color = "primary")
                       
                     ),
                     
                     box(id = "CombineBox3", title = "Label your experiment", status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         textInput("textlabelCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Sample Name:')), 
                                   value = "Sample B"),
                         numericInput("mincellsCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Include genes detected in at least this many cells:')),
                                      min = 0, max = 1000000,
                                      value = "3"),
                         numericInput("minfeaturesCombine", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Include cells where at least this many genes are detected:')),
                                      min = 0, max = 1000000,
                                      value = "200")),
                     p(style="text-align: center;", actionBttn("NextButtonCombine", "CONTINUE TO NEXT STEP...",
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "CombineBox4", title = "Count Matrix (preview first 500 rows and columns)", status = "warning", 
                         collapsible = T, width = NULL,
                         
                         withSpinner(DT::dataTableOutput("MatrixTableCombine", height = "600"))
                     ))
            ), collapseInput(inputId = "iscollapseboxCombine", boxId = "CombineBoxIntro")
    ),
    
    
    ####TAB #2 - Quality Control (1st Dataset)####    
    tabItem(tabName = "QC",
            fluidRow(
              box(id = "QCBoxIntro", title = strong("Quality Control"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify">A quality control (QC) step is crucial to remove low quality cells that may arise from cell 
                             damage or systematic errors in library preparation (PCR amplification and reverse transcription 
                             errors). If not removed, these low-quality cells can distort results in downstream analyses including
                             formation of spurious clusters due to similarities in the damaged-induced expression profiles 
                             between otherwise distinct subpopulations.<br><br>
                          In this step the user may apply thresholds to commonly used quality control metrics including:<br>
                          <ol><li>Number of unique genes per cell - Low quality cells will have very few genes, cell multiplets will have a high gene count.</li>
                          <li>Number of molecules detected per cell - Low quality cells will have very few molecules, cell multiplets will have a high molecule count.</li>
                          <li>Mitochondrial percentage - Low quality cells will contain high mitochondrial contamination.</li>
                          <li>Ribosomal percentage - Ribosomal percentage will differ depending on the starting cell types, this metric can be useful to removal of cells that do not fall within the expected range.</li></ol><br>')),
                  
                  img(id="QC_ICARUS",src="QC_ICARUS.png",width="60%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonQC", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "QCBox1", title = p("Quality Control", actionBttn("QCInfo1", "Info")), status = "warning", solidHeader = F,
                         collapsible = TRUE,  width = NULL,
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Number of unique genes (nFeature_RNA)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("nFeature1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum number:')),
                                                        min = 0, max = 100000,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("nFeature2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum number:')),
                                                        min = 0, max = 100000,
                                                        value = "100000"))))),
                         
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Number of molecules detected per cell (nCount_RNA)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("nCount1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum number:')),
                                                        min = 0, max = 1000000,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("nCount2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum number:')),
                                                        min = 0, max = 1000000,
                                                        value = "1000000"))))),
                         
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Mitochondrial Percentage (percent.mt)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("percent.mt1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("percent.mt2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "100"))))),
                         
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Ribosomal Percentage (percent.ribo)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("percent.ribo1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("percent.ribo2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "100"))))),
                         
                         
                         p(style="text-align: center;", actionBttn("SQC", "APPLY", 
                                                                   style = "jelly",
                                                                   color = "success",
                                                                   icon = icon("play")),
                           actionBttn("rewindQC", "RESTART", 
                                      style = "jelly",
                                      color = "danger",
                                      icon = icon("history")))
                     ),
                     p(style="text-align: center;", actionBttn("NextButtonQC", "CONTINUE TO NEXT STEP...",
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              column(9,
                     box(id = "QCBox2", title = p("Violin Plots", actionBttn("downloadQCViolin", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                         collapsible = TRUE,  width = NULL,
                         
                         fluidRow(
                           column(6,
                                  sliderInput("sizeQC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 2, step = 1)),
                           column(6,
                                  sliderInput("QCplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("plotUI"))))
            ),
            fluidRow(
              column(3),
              column(9,
                     column(6,
                            box(id = "QCBox3", title = p("nFeature_RNA vs nCount_RNA", actionBttn("downloadQCUMI", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                                collapsible = TRUE,  width = NULL,
                                
                                withSpinner(plotOutput("plot3", height = "600")))
                     ),
                     column(6,
                            box(id = "QCBox4", title = p("nCount_RNA vs percent.mt", actionBttn("downloadQCFeature", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                                collapsible = TRUE,  width = NULL,
                                
                                withSpinner(plotOutput("plot2", height = "600"))
                            ))
              )
            ),
            
            collapseInput(inputId = "iscollapseboxQC", boxId = "QCBoxIntro")
    ),
    
    ####TAB 2.25 - Doublet Removal (1st Dataset)####        
    tabItem(tabName = "Doublet",
            fluidRow(
              box(id = "DBoxIntro", title = strong("Doublet Removal"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify">
                          Cell doublets (multiplets) may arise during scRNA-seq library preparation (i.e., partitioning of multiple cells into a single droplet for droplet-based microfluidics methods). 
                          These doublets will exhibit a loss of single cell status and can compromise downstream analysis by creating spurious intermediate populations or transition states. <br> <br>
                          ICARUS incorporates the <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">DoubletFinder</a> R package to detect and remove doublets. 
                          DoubletFinder performs the following: <br>
                          <ol><li>Simulation of "artificial doublets" from existing scRNA-seq data.</li>
                          <li>Merge simulated artificial doublets with real data and perform dimensionality reduction. </li>
                          <li>A k-nearest neighbour graph is developed, and each cell is scored based on its proximity to artificial doublets. The highest scoring cells are assigned as real doublets. </li>
                          <li>The user has the option to visualise and remove these doublets.</li></ol><br>')),
                  
                  img(id="Doublet_ICARUS",src="Doublet_ICARUS.png",width="80%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please note doublet removal may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.')),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonD", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")),
                           column(3,       
                                  valueBox("SKIP",
                                           p(actionBttn("skipbuttonD", "SKIP", 
                                                        style = "jelly",
                                                        color = "primary")),
                                           icon = icon("forward"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "DBox1", title = p("Detect Doublets", actionBttn("DBoxInfo", "Info")), status = "warning", solidHeader = F,
                         width = NULL, collapsible = T,
                         p(style="text-align: justify;color:#FF0000", strong("Please note doublet removal may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.")),
                         numericInput("doublesim", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Estimated percentage of doublets in dataset (tailor for your dataset)')),
                                      min = 0, max = 100,
                                      value = 3, step = 1),
                         
                         p(style="text-align: center;", 
                           actionBttn("doublesimGO", "SIMULATE DOUBLETS",
                                      style = "jelly",
                                      color = "success",
                                      icon = icon("play"),
                                      size = "md"),
                           actionBttn("doubleremove", "REMOVE DOUBLETS",
                                      style = "jelly",
                                      color = "warning",
                                      icon = icon("play"),
                                      size = "md"),
                         )
                     ),
                     p(style="text-align: center;", actionBttn("NextButtonD", "CONTINUE TO NEXT STEP...", 
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "DBox2", title = p("UMAP/t-SNE Plot"), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(
                           column(6,
                                  prettyRadioButtons("radioD", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(6,
                                  pickerInput("plotclustersD", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                              choices = list("Detected Doublets", "Sample"),
                                              selected = "Detected Doublets", width = "98%"))
                         ),
                         
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeD", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           column(3,
                                  sliderInput("sizeD", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;"> Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityD", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           
                           column(3,
                                  sliderInput("Dplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         
                         withSpinner(htmlOutput("plotdoublesimUI")))
              )),
            
            collapseInput(inputId = "iscollapseboxD", boxId = "DBoxIntro")
    ),
    
    ####TAB 2.5 - Quality Control (2nd dataset)####
    tabItem(tabName = "QC2",
            fluidRow(
              box(id = "QCBoxIntro2", title = strong("Quality Control (2nd Dataset)"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify">A quality control (QC) step is crucial to remove low quality cells that may arise from cell 
                             damage or systematic errors in library preparation (PCR amplification and reverse transcription 
                             errors). If not removed, these low-quality cells can distort results in downstream analyses including
                             formation of spurious clusters due to similarities in the damaged-induced expression profiles 
                             between otherwise distinct subpopulations.<br><br>
                          In this step the user may apply thresholds to commonly used quality control metrics including:<br>
                          <ol><li>Number of unique genes per cell - Low quality cells will have very few genes, cell multiplets will have a high gene count.</li>
                          <li>Number of molecules detected per cell - Low quality cells will have very few molecules, cell multiplets will have a high molecule count.</li>
                          <li>Mitochondrial percentage - Low quality cells will contain high mitochondrial contamination.</li>
                          <li>Ribosomal percentage - Ribosomal percentage will differ depending on the starting cell types, this metric can be useful to removal of cells that do not fall within the expected range.</li></ol><br>')),
                  
                  img(id="QC2_ICARUS",src="QC_ICARUS.png",width="60%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonQC2", "START",
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "QCBox12", title = p("Quality Control", actionBttn("QC2Info1", "Info")), status = "warning", solidHeader = F,
                         collapsible = TRUE,  width = NULL,
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Number of unique genes (nFeature_RNA)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("nFeature12", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum number:')),
                                                        min = 0, max = 100000,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("nFeature22", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum number:')),
                                                        min = 0, max = 100000,
                                                        value = "100000"))))),
                         
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Number of molecules detected per cell (nCount_RNA)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("nCount12", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum number:')),
                                                        min = 0, max = 1000000,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("nCount22", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum number:')),
                                                        min = 0, max = 1000000,
                                                        value = "1000000"))))),
                         
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Mitochondrial Percentage (percent.mt)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("percent.mt12", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("percent.mt22", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "100"))))),
                         
                         fluidRow(
                           column(12,
                                  h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Ribosomal Percentage (percent.ribo)</b>')),
                                  fluidRow(
                                    column(6,  
                                           numericInput("percent.ribo12", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "0")),
                                    
                                    column(6,
                                           numericInput("percent.ribo22", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Maximum percentage:')),
                                                        min = 0, max = 100,
                                                        value = "100"))))),
                         
                         
                         p(style="text-align: center;", actionBttn("SQC2", "APPLY", 
                                                                   style = "jelly",
                                                                   color = "success",
                                                                   icon = icon("play")),
                           actionBttn("rewindQC2", "RESTART",
                                      style = "jelly",
                                      color = "danger",
                                      icon = icon("history")))),
                     p(style="text-align: center;", actionBttn("NextButtonQC2", "CONTINUE TO NEXT STEP...",
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              column(9,
                     box(id = "QCBox22", title = p("Violin Plots", actionBttn("downloadQCViolin2", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                         collapsible = TRUE,  width = NULL,
                         
                         fluidRow(
                           column(6,
                                  sliderInput("sizeQC2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 2, step = 1)),
                           column(6,
                                  sliderInput("QC2plotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("plotCombineUI"))))
            ),
            fluidRow(
              column(3),
              column(9,
                     column(6,
                            box(id = "QCBox32", title = p("nFeature_RNA vs nCount_RNA", actionBttn("downloadQCUMI2", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                                collapsible = TRUE,  width = NULL,
                                
                                withSpinner(plotOutput("plot3Combine", height = "600")))
                     ),
                     column(6,
                            box(id = "QCBox42", title = p("nCount_RNA vs percent.mt", actionBttn("downloadQCFeature2", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                                collapsible = TRUE,  width = NULL,
                                
                                withSpinner(plotOutput("plot2Combine", height = "600"))
                            ))
              )
              
            ),
            
            collapseInput(inputId = "iscollapseboxQC2", boxId = "QCBoxIntro2")
    ),
    
    ####TAB 2.75 - Doublet Removal (2nd Dataset)####        
    tabItem(tabName = "Doublet2",
            fluidRow(
              box(id = "DBoxIntro2", title = strong("Doublet Removal (2nd Dataset)"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  h4(HTML('<h4 style = "text-align:justify">
                          Cell doublets (multiplets) may arise during scRNA-seq library preparation (i.e., partitioning of multiple cells into a single droplet for droplet-based microfluidics methods). 
                          These doublets will exhibit a loss of single cell status and can compromise downstream analysis by creating spurious intermediate populations or transition states. <br> <br>
                          ICARUS incorporates the <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">DoubletFinder</a> R package to detect and remove doublets. 
                          DoubletFinder performs the following: <br>
                          <ol><li>Simulation of "artificial doublets" from existing scRNA-seq data.</li>
                          <li>Merge simulated artificial doublets with real data and perform dimensionality reduction. </li>
                          <li>A k-nearest neighbour graph is developed, and each cell is scored based on its proximity to artificial doublets. The highest scoring cells are assigned as real doublets. </li>
                          <li>The user has the option to visualise and remove these doublets.</li></ol>')),
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please note doublet removal may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.<br>')),
                  
                  img(id="Doublet2_ICARUS",src="Doublet_ICARUS.png",width="80%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonD2", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")),
                           column(3,       
                                  valueBox("SKIP",
                                           p(actionBttn("skipbuttonD2", "SKIP", 
                                                        style = "jelly",
                                                        color = "primary")),
                                           icon = icon("forward"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "DBox12", title = p("Detect Doublets", actionBttn("DBoxInfo2", "Info")), status = "warning", solidHeader = F,
                         width = NULL, collapsible = T,
                         p(style="text-align: justify;color:#FF0000", strong("Please note doublet removal may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.")),
                         numericInput("doublesim2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Estimated percentage of doublets in dataset (tailor for your dataset)')),
                                      min = 0, max = 100,
                                      value = 3, step = 1),
                         
                         p(style="text-align: center;", 
                           actionBttn("doublesimGO2", "SIMULATE DOUBLETS",
                                      style = "jelly",
                                      color = "success",
                                      icon = icon("play"),
                                      size = "md"),
                           actionBttn("doubleremove2", "REMOVE DOUBLETS",
                                      style = "jelly",
                                      color = "warning",
                                      icon = icon("play"),
                                      size = "md"),
                         )
                     ),
                     p(style="text-align: center;", actionBttn("NextButtonD2", "CONTINUE TO NEXT STEP...", 
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "DBox22", title = p("UMAP/t-SNE Plot"), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(
                           column(6,
                                  prettyRadioButtons("radioD2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(6,
                                  pickerInput("plotclustersD2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                              choices = list("Detected Doublets", "Sample"),
                                              selected = "Detected Doublets", width = "98%"))
                         ),
                         
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeD2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           column(3,
                                  sliderInput("sizeD2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityD2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           
                           column(3,
                                  sliderInput("D2plotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         
                         withSpinner(htmlOutput("plotdoublesim2UI")))
              )),
            
            collapseInput(inputId = "iscollapseboxD2", boxId = "DBoxIntro2")
    ),
    
    ####TAB 3 - Dimensionality Reduction####        
    tabItem(tabName = "DR",
            fluidRow(
              box(id = "DRBoxIntro", title = strong("Dimensionality Reduction"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify">In single cell RNA-seq analyses, individual cells may be characterised by its transcriptomic expression profile comprised of thousands of genes. 
                    If the expression level of each gene represents a dimension of data, each cell can be allocated a location in the high-dimensional expression space. 
                    The purpose of dimensionality reduction is to enable the analysis and visualisation of high-dimensional data in a low dimensional plane whilst retaining 
                    the major properties of the original data. This is possible as biological processes encompass multiple genes (i.e., genes within a pathway) and therefore, 
                    multiple features can be compressed into a single dimension. <br><br>
                          The following are performed at this step:<br>
                          <ol><li>Either log Normalization of the data by a scale factor of 10,000 or Normalization by <a href="https://doi.org/10.1186/s13059-019-1874-1" target="_blank">SCTransform</a>.
                          A Normalization step ensures technical biases or noise present from sequencing are kept to a minimum. </li>
                          <li>A linear transformation is applied to scale expression levels so that the mean expression of genes across cells equates to zero and 
                          the variance across cells equates to one. This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate.</li>
                          <li>Dimensionality reduction using Principal Component Analysis (<b>PCA</b>) with the user defined number of variable features to incorporate.
                          Seurat prioritises a set of highly variable genes when performing its dimensionality reduction, this number may be altered in the <b>NEW</b> box below.</li>
                          <li>An option to impute dropouts (false zeros in the dataset due to low amounts of mRNA in individual cells resulting in insufficient mRNA capture) is available. 
                          The <a href="https://www.nature.com/articles/s41467-021-27729-z" target="_blank">Adaptively-thresholded low rank approximation (ALRA) method</a> to impute dropouts is used.</li></ol><br>')),
                  
                  img(id="DR_ICARUS",src="DR_ICARUS.png",width="60%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startDR", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            collapseInput(inputId = "iscollapsebox1", boxId = "DRBoxIntro"),
            
            fluidRow(
              box(id = "DRBox1", title = p("Variable Features Plot", actionBttn("DRInfo1", "Info"), actionBttn("downloadDRVF", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, 
                  
                  withSpinner(plotOutput("plot4", height = 600))),
              
              box(id = "DRBox2", title = p("Dimension Reduction Heatmap", actionBttn("DRInfo2", "Info"), actionBttn("downloadDRPC", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, 
                  sidebar = boxSidebar(id = "boxsidebar1", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("integer", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">View Dimension:')),
                                                    min = 1, max = 100,
                                                    value = 1, step = 1, width = "98%"),
                                       numericInput("PCnfeatures", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of genes to plot:')),
                                                    min = 2, max = 1000,
                                                    value = 30, step = 1, width = "98%"),
                                       numericInput("PCncells", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of cells to plot:')),
                                                    min = 2, 
                                                    value = 500, step = 1, width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  withSpinner(plotOutput("plot5", height = 600)))
            ),
            
            fluidRow(
              box(id = "DRBox3", title = p("Elbow Plot", actionBttn("DRInfo3", "Info"), actionBttn("downloadDRElbow", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, 
                  withSpinner(plotOutput("plot6", height = 600))),
              box(id = "DRBox4", title = p("Loadings Plot", actionBttn("DRInfo4", "Info"), actionBttn("downloadDRLoadings", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, 
                  sidebar = boxSidebar(id = "boxsidebar_loadings", startOpen = TRUE, width =25, background = "#e8ebe9",
                                       numericInput("integer2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">View Dimension:')),
                                                    min = 1, max = 100,
                                                    value = 1, step = 1, width = "98%"),
                                       numericInput("Loadingsnfeatures", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of genes to plot:')),
                                                    min = 0, max = 1000,
                                                    value = 30, step = 1, width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  withSpinner(plotOutput("plot7", height = 600)))
            ),
            p(style="text-align: center;", actionBttn("NextButtonDR", "CONTINUE TO NEXT STEP...", 
                                                      style = "jelly",
                                                      color = "primary",
                                                      icon = icon("play"), block = T))
    ),
    
    ####TAB 3.5 - Dimensionality reduction (Integrated datasets)####
    tabItem(tabName = "DR2",
            fluidRow(
              box(id = "DRBoxIntro2", title = strong("Integrate and Perform Dimensionality Reduction"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify">In single cell RNA-seq analyses, individual cells may be characterised by its transcriptomic expression profile comprised of thousands of genes. 
                    If the expression level of each gene represents a dimension of data, each cell can be allocated a location in the high-dimensional expression space. 
                    The purpose of dimensionality reduction is to enable the analysis and visualisation of high-dimensional data in a low dimensional plane whilst retaining 
                    the major properties of the original data. This is possible as biological processes encompass multiple genes (i.e., genes within a pathway) and therefore, 
                    multiple features can be compressed into a single dimension.<br><br>
                          The following are performed at this step:<br>
                          <ol><li>Either log Normalization of the data by a scale factor of 10,000 or Normalization by <a href="https://doi.org/10.1186/s13059-019-1874-1" target="_blank">SCTransform</a>. 
                          A Normalization step ensures technical biases or noise present from sequencing are kept to a minimum.</li>
                          <li>A linear transformation is applied to scale expression levels so that the mean expression of genes across cells equates to zero and 
                          the variance across cells equates to one. This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate.</li>
                          <li>Dimensionality reduction is performed using Principal Component Analysis (<b>PCA</b>) with the user defined number of variable features.
                          Seurat prioritises a set of highly variable genes when performing its dimensionality reduction, this number may be altered in the <b>NEW</b> box below.</li>
                          <li>Data integration with either 
                          <a href="https://doi.org/10.1016/j.cell.2019.05.031" target="_blank">Canonical Correlation Analysis (CCA)</a>,
                          <a href="https://satijalab.org/seurat/articles/integration_rpca.html" target="_blank">Reciprocal PCA (RPCA)</a> or 
                          <a href="https://doi.org/10.1038/s41592-019-0619-0" target="_blank">Harmony</a> methodologies. 
                          For CCA and RPCA methods, the number of <b> k-anchors </b> (strength of integration) 
                          can be adjusted, please increase the k-anchors parameter (default value = 5) for samples where integration of certain cell types are not aligned. </li></ol><br>')),
                  
                  img(id="DR2_ICARUS",src="DR2_ICARUS.png",width="100%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startDR2", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            collapseInput(inputId = "iscollapseboxDR2", boxId = "DRBoxIntro2"),
            
            p(style = "text-align: right;", 
              actionBttn("downloadMerged", "Download Integrated Gene Count Matrix", size = "lg"),
              
              fluidRow(
                box(id = "DRBox1DR2", title = p("Variable Features Plot", actionBttn("DR2Info1", "Info"), actionBttn("downloadDR2VF1", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                    collapsible = TRUE, 
                    sidebar = boxSidebar(id = "boxsidebarXDR2", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                         numericInput("integerVFPlot", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">View variable features plot for sample:')),
                                                      min = 1, max = 50,
                                                      value = 1, step = 1, width = "98%"),
                                         p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                    
                    withSpinner(plotOutput("DR2VF1", height = 600))),
                
                box(id = "DRBox2DR2", title = p("Dimension Reduction Heatmap (Integrated Dataset)", actionBttn("DR2Info3", "Info"), actionBttn("downloadDR2PC", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                    collapsible = TRUE, 
                    sidebar = boxSidebar(id = "boxsidebar1DR2", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                         numericInput("integerDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">View Dimension:')),
                                                      min = 1, max = 100,
                                                      value = 1, step = 1, width = "98%"),
                                         numericInput("PCnfeaturesDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of genes to plot:')),
                                                      min = 2, max = 1000,
                                                      value = 30, step = 1, width = "98%"),
                                         numericInput("PCncellsDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of cells to plot:')),
                                                      min = 2, 
                                                      value = 500, step = 1, width = "98%"),
                                         p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                    ),
                    withSpinner(plotOutput("plot5DR2", height = 600)))
              ),
              
              
              fluidRow(
                box(id = "DRBox3DR2", title = p("Elbow Plot (Integrated Dataset)", actionBttn("DR2Info4", "Info"), actionBttn("downloadDR2Elbow", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                    collapsible = TRUE, 
                    withSpinner(plotOutput("plot6DR2", height = 600))),
                box(id = "DRBox4DR2", title = p("Loadings Plot (Integrated Dataset)", actionBttn("DR2Info5", "Info"), actionBttn("downloadDR2Loadings", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                    collapsible = TRUE, 
                    sidebar = boxSidebar(id = "boxsidebar_loadingsDR2", startOpen = TRUE, width =25, background = "#e8ebe9",
                                         numericInput("integer2DR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">View Dimension:')),
                                                      min = 1, max = 100,
                                                      value = 1, step = 1, width = "98%"),
                                         numericInput("LoadingsnfeaturesDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of genes to plot:')),
                                                      min = 1, max = 1000,
                                                      value = 30, step = 1, width = "98%"),
                                         p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                    ),
                    withSpinner(plotOutput("plot7DR2", height = 600)))
              ),
              
              p(style="text-align: center;", actionBttn("NextButtonDR2", "CONTINUE TO NEXT STEP...", 
                                                        style = "jelly",
                                                        color = "primary",
                                                        icon = icon("play"), block = T))
            )
    ),
    
    ####TAB 4 - Clustering####        
    tabItem(tabName = "Cluster",
            fluidRow(
              box(id = "CBoxIntro", title = strong("Clustering"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify">The outcome of clustering is to group together cells with similar expression profiles for human interpretation. <br><br>
                          ICARUS employs Seurat\'s graph-based community detection clustering approach. In brief, a graph of k-nearest neighbours (k-NN) is formed between cells in high dimensional space 
                          where each cell is a node that is connected to its nearest cells. 
                          The edges of each connection are then weighted based on its similarities to neighbouring cells and a community detection algorithm is utilised to define clusters. <br><br>
                          Clustering parameters including number of dimensions, number of k-nearest neighbours for graph construction and choice of clustering algorithm (either Louvain, SLM or Leiden) may be adjusted in this step.
                          UMAP parameters (nearest neighbours and minimum distance) and t-SNE parameters (perplexity and number of iterations) can also be adjusted.<br><br>')),
                  
                  img(id="Clustering_ICARUS",src="Clustering_ICARUS.png",width="80%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please note clustering may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.')),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonC", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "CBox2", title = p("Select Parameters", actionBttn("CBoxInfo", "Info")), status = "warning", solidHeader = F,
                         width = NULL, collapsible = T,
                         p(style="text-align: justify;color:#FF0000", strong("Please note data clustering may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.")),
                         h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Nearest-neighbour graph construction</b>')),
                         numericInput("integera", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of dimensions')),
                                      min = 0, max = 100,
                                      value = 15, step = 1),
                         
                         numericInput("integerb", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">k.param')),
                                      value = 20, step = 1),
                         
                         numericInput("integerc", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">n.trees')),
                                      value = 50, step = 1),
                         
                         h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>Clustering Parameters</b>')),
                         numericInput("integerd", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Resolution')),
                                      min = 0, max = 10,
                                      value = 0.3, step = 0.1),
                         
                         pickerInput("select1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Clustering Algorithm')), 
                                     choices = list("Louvain", "SLM",
                                                    "Leiden"), selected = "Louvain"),
                         
                         h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>UMAP Parameters</b>')), 
                         
                         numericInput("integere", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of dimensions')),
                                      min = 0, max = 100,
                                      value = 15, step = 1),
                         
                         numericInput("integerf", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">k-nearest-neighbours')),
                                      min = 0, max = 1000,
                                      value = 20, step = 1),
                         
                         numericInput("integerg", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">min.dist')),
                                      min = 0, max = 10,
                                      value = 0.3, step = 0.01),
                         
                         h4(HTML('<h4 style = "text-align:justify;color:#000000"><b>t-SNE Parameters</b>')),
                         numericInput("tsne1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of dimensions')),
                                      min = 0, max = 100,
                                      value = 15, step = 1),
                         
                         numericInput("tsne2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Perplexity')),
                                      value = 30, step = 1),
                         
                         numericInput("tsne3", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Max iterations')),
                                      value = 1000, step = 100),
                         
                         p(style="text-align: center;", actionBttn("button5", "APPLY",
                                                                   style = "jelly",
                                                                   color = "success",
                                                                   icon = icon("play"),
                                                                   size = "lg"),
                         )
                     ),
                     p(style="text-align: center;", actionBttn("NextButtonC", "CONTINUE TO NEXT STEP...", 
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "CBox1", title = p("UMAP/t-SNE Plot"), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(
                           column(6,
                                  prettyRadioButtons("radio", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(6,
                                  pickerInput("plotclustersC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                              choices = list("Seurat Clusters", "Sample"),
                                              selected = "Seurat Clusters", width = "98%"))
                         ),
                         
                         fluidRow(
                           column(3,
                                  sliderInput("labelsize", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           column(3,
                                  sliderInput("size", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacity", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           column(3,
                                  sliderInput("Cplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         
                         withSpinner(htmlOutput("plot8UI")))
              )),
            
            fluidRow(
              box(id = "CBox3", title = p("Cell Composition Summary", actionBttn("CBoxInfo3", "Info"), actionBttn("downloadComposition", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "boxsidebar_samplesC", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("Sample_selectionC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select group to compare:')),
                                                   choices = list("Seurat Clusters", "Labelled Cells"), selected = "Seurat Clusters", width = "98%"),
                                       materialSwitch("switchPercent", label = h4(HTML('<h4 style = "text-align:justify;color:#000000">Show composition as %')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  fluidRow(
                    column(6,
                           sliderInput("labelsizeC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    
                    column(6,
                           sliderInput("HeightC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("sampledistributionUI"))
              )
            ),
            
            collapseInput(inputId = "iscollapsebox2", boxId = "CBoxIntro")
    ),
    
    
    ####TAB 5 - Cell cycle removal ####
    tabItem(tabName = "CC",
            fluidRow(
              box(id = "CCBoxIntro", title = strong("Data Correction"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Data correction aims to remove further technical and biological confounders. 
                          For example, cell clusters may be formed largely due to cell cycle state rather than differences in underlying biology.
                  The choice whether to correct for these effects will depend on the intention of the downstream analysis. <br><br>
                          This step allows the regression of the following:<br>
                          <ol><li><b>Cell cycle genes:</b> Expression of cell cycle phase canonical markers (taken from <a href="https://doi.org/10.1126/science.aad0501" target="_blank">Triosh et al., 2016</a>).</li>
                          <li><b>User gene(s) of interest:</b> Average normalised gene count for each selected gene are computed and regressed from the data (using the vars.to.regress function in Seurat::ScaleData).</li>
                          <li><b>User gene pathway of interest:</b> Average normalised gene count for all genes in the gene pathway <a href="https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp" target="_blank">(GSEA/Msigdb database)</a>
                          are computed for each cell and regressed from the data (using the vars.to.regress function in Seurat::ScaleData).</li></ol><br>')),
                  
                  img(id="DataCorrection_ICARUS",src="DataCorrection_ICARUS.png",width="80%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please note data correction may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.')),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonCC", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "CCBox1", title = p("Effect Regression", actionBttn("CCBoxInfo", "Info")), status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         p(style="text-align: justify;color:#FF0000", strong("Please note data correction may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.")),
                         pickerInput("CC1a", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Regress the following:')),
                                     choices = list("Please select from the following" = "", "Confounding Effects", "Gene(s) of interest", "Gene pathway"), selected = "")
                     ),
                     
                     box(id = "CCBox1a", title = "Confounding Effects", status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         virtualSelectInput("CC1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Regress the following:')),
                                            choices = NULL, multiple = T,
                                            hideClearButton = F,
                                            disableSelectAll = F),
                         
                         p(style="text-align: center;", actionBttn("regress1", "APPLY", icon = icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"),
                           actionBttn("rewindCC", "RESTART", icon = icon("history"),
                                      style = "jelly",
                                      color = "danger"))
                     ),
                     
                     box(id = "CCBox1b", title = "Regress gene(s) of interest", status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         virtualSelectInput("CC2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Regress gene or genes:')), choices = NULL, multiple = T,
                                            hideClearButton = F,
                                            disableSelectAll = T,
                                            search = T,
                                            placeholder = "Please type your gene(s) of interest"),
                         p(style="text-align: center;", actionBttn("regress2", "APPLY", icon = icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"),
                           actionBttn("rewindCC2", "RESTART", icon = icon("history"),
                                      style = "jelly",
                                      color = "danger"))
                     ),
                     box(id = "CCBox1c", title = "Regress gene pathway", status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         virtualSelectInput("CC3", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Regress a gene pathway:')), choices = NULL,
                                            hideClearButton = F,
                                            disableSelectAll = T,
                                            search = T,
                                            placeholder = "Please type your pathway of interest"),
                         
                         p(style="text-align: center;", actionBttn("regress3", "APPLY", icon = icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"),
                           actionBttn("rewindCC3", "RESTART", icon = icon("history"),
                                      style = "jelly",
                                      color = "danger"))
                     ),
                     p(style="text-align: center;", actionBttn("NextButtonCC", "CONTINUE TO NEXT STEP...",
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "CCBox2", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         prettyRadioButtons("radioCC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  
                                  column(6,
                                         pickerInput("plotclustersCC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                                                     selected = "Cell Cycle Phase",
                                                     width = "98%")
                                  )),
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeCC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("sizeCC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityCC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           column(3,
                                  sliderInput("CCplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                           
                         ),
                         
                         withSpinner(htmlOutput("plotCCUI")),
                         htmlOutput("GenepathwaytextCC")
                     ))),
            
            collapseInput(inputId = "iscollapsebox3", boxId = "CCBoxIntro")
            
    ),
    
    
    ####TAB 6 - Label####        
    tabItem(tabName = "Label",
            fluidRow(
              box(id = "LBoxIntro", title = strong("Cluster Labelling"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style="text-align:justify"> Marker genes for each cluster may be analysed and compared against known cell markers to assign a cluster cell type. 
                  ICARUS enables users to add custom cell cluster labels or annotate clusters against either the sctype reference set or the singleR reference set. <br><br>
                  
                  The <a href="https://doi.org/10.1038/s41467-022-28803-w" target="_blank">sctype reference dataset</a> compiled cell markers from 
                  <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/" target="_blank">CellMarker</a> and 
                  <a href="https://panglaodb.se/" target="_blank">PanglaoDB</a> databases. These include markers for the brain, pancreas, immune system, liver,
                  eye, kidney, lung, embryo, gastrointestinal tract, muscle and skin. <br> <br>
                  
                  The <a href="https://doi.org/10.1038/s41590-018-0276-y" target="_blank">singleR</a> R package incorporates cell markers from the following databases: <br>
                  <b>Brain</b>
                  <ol><li>Darmanis Brain Data: scRNA-seq of 8 adult and 4 fetal human brain temporal lobe tissue (<a href="https://doi.org/10.1073/pnas.1507125112" target="_blank">Darmanis et al., 2015</a>). </li>
                  <li>Zhong Prefrontal Cortex Data: scRNA-seq of developing human embryonic prefrontal cortex from gestational weeks 8 to 26 (<a href="https://doi.org/10.1038/nature25980" target="_blank">Zhong et al., 2018</a>). </li></ol>
                  <b>Immune system</b>
                          <ol><li>Blueprint Encode: scRNA-seq of 259 bulk RNA-seq samples generated by Blueprint and ENCODE from pure populations of stroma and immune cells
                          (<a href="https://doi.org/10.3324/haematol.2013.094243" target="_blank">Martens and Stunnenberg, 2013;</a>
                          <a href="https://doi.org/10.1038/nature11247" target="_blank">The ENCODE Consortium, 2012</a>).</li>
                          <li>Database of Immune Cell Expression (DICE): scRNA-seq of 1561 bulk RNA-seq samples generated by DICE from pure populations of human immune cells.
                          (<a href="https://doi.org/10.1016/j.cell.2018.10.022" target="_blank">Schmiedel et al., 2018</a>).</li>
                          <li>Immunologic Genome Project (ImmGen): scRNA-seq of 830 microarray samples generated by ImmGen from pure populations of murine immune cells 
                          (<a href="http://www.immgen.org/" target="_blank">http://www.immgen.org/</a>).</li>
                          <li>Monaco Immune Data: scRNA-seq of 114 bulk RNA-seq samples of sorted immune cell populations that can be found in GSE107011 
                          (<a href="https://doi.org/10.1016/j.celrep.2019.01.041" target="_blank">Monaco et al., 2019</a>).</li>
                          <li>Novershtern Hematopoietic Data: scRNA-seq of 211 bulk human microarray samples of sorted hematopoietic cell populations that can be found in GSE24759 
                          (<a href="https://doi.org/10.1016/j.cell.2011.01.004" target="_blank">Novershtern et al., 2011</a>).</li></ol>
                  <b>Pancreas</b><br>
                          <ol><li>Baron Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors and 2 mouse strains (<a href="https://doi.org/10.1016/j.cels.2016.08.011" target="_blank">Baron et al., 2016</a>). </li>
                          <li>Lawlor Pancreas Data: scRNA-seq of 5 non-diabetic and 8 type 2 diabetic human islet samples (<a href="https://doi.org/10.1101/gr.212720.116" target="_blank">Lawlor et al., 2017</a>). </li>
                          <li>Muraro Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors (<a href="https://doi.org/10.1016/j.cels.2016.09.002" target="_blank">Muraro et al., 2016</a>). </li>
                          <li>Segerstolpe Pancreas Data: scRNA-seq of human islet cells from 6 healthy and 4 type 2 diabetic donors (<a href="https://doi.org/10.1016/j.cmet.2016.08.020" target="_blank">Segerstolpe et al., 2016</a>). </li></ol>
                  <b>Multiple organs</b>        
                          <ol><li>Mouse RNA-seq: 358 bulk RNA-seq samples of sorted cell populations 
                          (<a href="https://doi.org/10.1101/gr.240093.118" target="_blank">Benayoun et al., 2019</a>).</li>
                          <li>Human Primary Cell Atlas: scRNA-seq of 713 microarray samples from the Human Primary Cell Atlas 
                          (<a href="https://doi.org/10.1186/1471-2164-14-632" target="_blank">Mabbott et al., 2013</a>).</li>
                          <li>He Organ Atlas: scRNA-seq of 15 tissue organs of one adult donor. Tissue organs include bladder, blood, common bile duct, oesophagus, heart, liver,
                          lymph node, bone marrow, muscle, rectum, skin, small intestine, spleen, stomach and trachea 
                          (<a href="https://doi.org/10.1186/s13059-020-02210-0" target="_blank">He et al., 2020</a>).</li></ol><br>')),
                  
                  img(id="Labelling_ICARUS",src="Labelling_ICARUS.png",width="80%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please note cluster labelling using reference marker databases may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.')),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonL", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "LBox3", title = p("Label Clusters", actionBttn("LBoxInfo1", "Info")), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         p(style="text-align: justify;color:#FF0000", strong("Please note data cluster labelling using reference marker databases may take a long time depending on the size of the input dataset(s). Please save the output and use the load function for repeat analysis.")),
                         
                         
                         pickerInput("selectLabelInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Assign cell labels with')),
                                     choices = list("Please select from the following" = "", "sctype reference cell marker dataset", "SingleR reference cell marker dataset", "Own cell labels"), selected = "",
                                     width = "98%")
                     ),
                     box(id = "LBox4", title = "SingleR reference dataset", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         pickerInput("selecttissueSingleR", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select starting tissue')), 
                                     choices = list("Please select from the following" = "", "Immune system", "Brain",
                                                    "Pancreas", "Multiple"),
                                     selected = "",
                                     width = "98%"),
                         
                         pickerInput("select2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select singleR reference dataset (Immune system)')), 
                                     choices = list("Blueprint Encode", "DICE Immune Cell Expression Data", "Immunologic Genome Project",
                                                    "Monaco Mouse Immune Data", "Novershtern Hematopoietic Data"),
                                     selected = "Blueprint Encode",
                                     width = "98%"),
                         
                         pickerInput("selectBrain", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select singleR reference dataset (Brain)')), 
                                     choices = list("Darmanis Brain Data", "Zhong Prefrontal Cortex Data"),
                                     selected = "Darmanis Brain Data",
                                     width = "98%"),
                         
                         pickerInput("selectPancreas", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select singleR reference dataset (Pancreas)')), 
                                     choices = list("Baron Mouse Pancreas Data", "Baron Human Pancreas Data", "Lawlor Pancreas Data", "Muraro Pancreas Data",
                                                    "Segerstolpe Pancreas Data"),
                                     selected = "Baron Human Pancreas Data",
                                     width = "98%"),
                         
                         pickerInput("selectMultiple", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select singleR reference dataset (Multiple)')), 
                                     choices = list("Human Primary Cell Atlas", "Mouse RNA-seq", "He Organ Atlas"),
                                     selected = "Human Primary Cell Atlas",
                                     width = "98%"),
                         
                         pickerInput("selectHe", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select from the following tissue types for He Organ Atlas')), 
                                     choices = list("Bladder", "Blood", "Common.bile.duct", "Esophagus", "Heart", "Liver",
                                                    "Lymph.node", "Marrow", "Muscle", "Rectum", "Skin", "Small.intestine", "Spleen",
                                                    "Stomach", "Trachea"),
                                     selected = "Blood",
                                     width = "98%"),
                         
                         p(style = "text-align:center;", actionBttn("ReferenceButtonImmune", "APPLY", icon("play"),
                                                                    style = "jelly",
                                                                    color = "success")),
                         p(style = "text-align:center;", actionBttn("ReferenceButtonBrain", "APPLY", icon("play"),
                                                                    style = "jelly",
                                                                    color = "success")),
                         p(style = "text-align:center;", actionBttn("ReferenceButtonPancreas", "APPLY", icon("play"),
                                                                    style = "jelly",
                                                                    color = "success")),
                         p(style = "text-align:center;", actionBttn("ReferenceButtonMultiple", "APPLY", icon("play"),
                                                                    style = "jelly",
                                                                    color = "success"))
                     ),
                     box(id = "LBox5", title = "Own labels", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         h4(HTML('<h4 style = "text-align:justify;color:#000000;"> Use own labels')),
                         
                         uiOutput("LabelClustersTextInput"),
                         
                         p(style = "text-align:center;", actionBttn("LabelClustersTextButton", "APPLY", icon("play"),
                                                                    style = "jelly",
                                                                    color = "success"))
                     ),
                     
                     box(id = "LBox6", title = "sctype reference dataset", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         pickerInput("selecttissue", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select starting tissue type')), 
                                     choices = list("Immune system", "Pancreas",
                                                    "Liver", "Eye", "Muscle", "Skin",
                                                    "Kidney", "Brain", "Lung", "Embryo", "Gastrointestinal tract"),
                                     width = "98%"),
                         
                         materialSwitch("switchUnknown", label = h4(HTML('<h4 style = "text-align:justify;color:#000000">Set low-confident clusters to "unknown"')), value = TRUE, status = "primary"),
                         
                         p(style = "text-align:center;", actionBttn("ReferenceButtonsctype", "APPLY", icon("play"),
                                                                    style = "jelly",
                                                                    color = "success"))
                     )
              ),
              
              column(9,
                     box(id = "LBox1", title = "UMAP/t-SNE plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         fluidRow(column(6,
                                         prettyRadioButtons("radioL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  
                                  column(6,
                                         pickerInput("plotclustersL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                                                     selected = "Seurat Clusters", width = "98%")
                                  )),
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("sizeL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           
                           column(3,
                                  sliderInput("Lplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("plot9UI"))
                     ))),
            
            fluidRow( 
              box(id = "LBox2", title = p("Marker Genes", actionBttn("LBoxInfo2", "Info"), downloadBttn("downloadmarkersL", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "boxsidebar_labelcells", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("LC_clusters", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by:')),
                                                   choices = list("Seurat Clusters", "Labelled Cells"),
                                                   width = "98%"),
                                       
                                       pickerInput("testuseL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select test:')),
                                                   choices = list("wilcox", "bimod", "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2"),
                                                   width = "98%"),
                                       
                                       numericInput("log2FCL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Log2 fold change threshold')), 
                                                    min = 0, max = 10,
                                                    value = 1, step = 0.01, width = "98%"),
                                       
                                       numericInput("minL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">min.pct')),
                                                    min = 0, max = 1,
                                                    value = 0.3, step = 0.01, width = "98%"),
                                       
                                       pickerInput("padjustL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value adjustment method')),
                                                   choices = list("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                                                   selected = "fdr",
                                                   width = "98%"),
                                       numericInput("pvalL", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                                                    min = 0.1, max = 1,
                                                    value = 0.05, step = 0.01, width = "98%"),
                                       p(style="text-align: center;", actionBttn("CMG", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  fluidRow(column(9),
                           column(3,
                                  virtualSelectInput("refInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Query cluster')), choices = NULL, width = "98%")
                           )
                  ),
                  
                  withSpinner(DT::dataTableOutput("markers", height = "500"))
              ),
              
              box(id = "LBox7", title = p("Marker Gene Expression Summary", actionBttn("LBoxInfo3", "Info"), actionBttn("downloadmarkersUporDown", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(
                    column(6,
                           sliderInput("labelsizeUporDown", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    
                    column(6,
                           sliderInput("HeightUporDown", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("UporDownUI"))
              ),
              
              box(id = "LBox8", title = p("Marker Gene Map", actionBttn("LBoxInfo4", "Info"), actionBttn("downloadmarkersCluster", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "MGCluster", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("MGNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of genes per cluster to show (ordered by fold change)')),
                                                    min = 1, max = 100,
                                                    value = 20, step = 1, width = "98%"),
                                       materialSwitch("checkboxgene_markers", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Gene Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxcluster_markers", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Cluster Labels')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(4,
                           sliderInput("genefontmarkers", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Gene Label Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1)),
                    column(4,
                           sliderInput("categoryfontmarkers", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Cluster Label Size')),
                                       min = 1, max = 20,
                                       value = 8, step = 1)),
                    column(4,
                           sliderInput("nodemarkers", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1))
                  ),
                  
                  withSpinner(plotOutput("MarkersMap", height = 1500)))
            ),
            collapseInput(inputId = "iscollapsebox4", boxId = "LBoxIntro")
    ),
    
    ####TAB 7 - Gene expression#######
    
    tabItem(tabName = "GE",
            fluidRow(
              box(id = "GEBoxIntro", title = strong("Gene Expression"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> This step allows the user to visualize normalised counts for gene(s) or gene pathway 
                  (<a href="https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp" target="_blank">GSEA/Msigdb gene pathway</a>
                  including KEGG, Reactome, WikiPathways and Gene Ontology) across cell clusters. <br><br>')),
                  
                  img(id="GeneExpression_ICARUS",src="GeneExpression_ICARUS.png",width="50%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonGE", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "GEBox1", title = p("Gene Expression", actionBttn("GEBoxInfo", "Info")), status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         
                         virtualSelectInput("GEInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Enter single or multiple genes:')), choices = NULL, multiple = T, 
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = T,
                                            placeholder = "Please type your gene(s) of interest"
                         ),
                         p(style="text-align: center;", actionBttn("gene1", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success")),
                         
                         p(style="text-align: center;", h4(HTML('<h4 style = "text-align: center; color:#000000; "><b>OR</b>'))),
                         
                         virtualSelectInput("GSEInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize gene pathway:')), choices = c("Please select from the following" = ""), selected = "",
                                            search = T,
                                            hideClearButton = T,
                                            placeholder = "Please type your pathway of interest"),
                         
                         p(style="text-align: center;", actionBttn("gene2", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     )
              ),
              
              column(9,
                     box(id = "GEBox2", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         prettyRadioButtons("radioGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  column(6,
                                         pickerInput("plotclustersGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Gene Expression"),
                                                     selected = "Gene Expression",
                                                     width = "98%"))
                         ),
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("sizeGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           
                           column(3,
                                  sliderInput("GEplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("plotGEUI")),
                         htmlOutput("Genepathwaytext")
                     ))),
            
            fluidRow(
              box(id = "GEBox3", title = p("Violin Plots", actionBttn("downloadGEViolin", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "boxsidebarviolinGE", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("VlnInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by:')),
                                                   choices = list("Please select from the following" = "", "Labelled Cells", "Seurat Clusters", "Cell Cycle Phase"),
                                                   selected = "",
                                                   width = "98%"),
                                       virtualSelectInput("refineGEVLN", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Filter Clusters:')),
                                                          choices = NULL, 
                                                          multiple = T,
                                                          hideClearButton = F,
                                                          search = T,
                                                          disableSelectAll = F,
                                                          width = "98%"),
                                       materialSwitch("checkboxvln", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Split by sample')), value = FALSE, status = "primary"),
                                       virtualSelectInput("refineGEVLN_SAMPLES", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Filter Samples:')),
                                                          choices = NULL, 
                                                          multiple = T,
                                                          hideClearButton = F,
                                                          search = T,
                                                          disableSelectAll = F,
                                                          width = "98%"),
                                       materialSwitch("mergedGE", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Plot average expression of all genes')), value = FALSE, status = "primary"),
                                       p(style="text-align: center;", actionBttn("GEDOTGO2", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(
                    column(4,
                           sliderInput("VlnsizeGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    column(4,
                           sliderInput("Vlnpt", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Point Size')),
                                       min = 0, max = 5,
                                       value = 2, step = 0.5)),
                    
                    column(4,
                           sliderInput("HeightGEVln", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 1000,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("VlnUI"))
              ),
              box(id = "GEBox4", title = p("Dot Plot", actionBttn("downloadGEDot", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "boxsidebarGEDot", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("GEDotInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by:')),
                                                   choices = list("Please select from the following" = "", "Labelled Cells", "Seurat Clusters", "Cell Cycle Phase"),
                                                   selected = "", 
                                                   width = "98%"),
                                       virtualSelectInput("refineGEDOT", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Filter Clusters:')),
                                                          choices = NULL, 
                                                          multiple = T,
                                                          hideClearButton = F,
                                                          search = T,
                                                          disableSelectAll = F,
                                                          width = "98%"),
                                       materialSwitch("checkboxGEDot", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Split by sample')), value = FALSE, status = "primary"),
                                       virtualSelectInput("refineGEDOT_SAMPLES", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Filter Samples:')),
                                                          choices = NULL, 
                                                          multiple = T,
                                                          hideClearButton = F,
                                                          search = T,
                                                          disableSelectAll = F,
                                                          width = "98%"),
                                       p(style="text-align: center;", actionBttn("GEDOTGO", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(
                    column(4,
                           sliderInput("GEDotSize", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Font Size')),
                                       min = 1, max = 50,
                                       value = 10, step = 1)),
                    column(4,
                           sliderInput("dotscale", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Dot Scale')),
                                       min = 1, max = 100,
                                       value = 6, step = 1)),
                    column(4,
                           sliderInput("HeightGEDot", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("GEDotUI"))
              )),
            
            collapseInput(inputId = "iscollapsebox5", boxId = "GEBoxIntro")
    ),
    
    ####TAB 8 - MEGENA #####
    tabItem(tabName = "MEGENA",
            fluidRow(
              box(id = "MEGENABoxIntro", title = strong("Gene Co-Expression Networks"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Many biological networks contain patterns of gene expression that are evolutionary conserved. 
                  These patterns can be characterised and grouped into hierarchical co-expression modules which can infer biological functionality. 
                  A method of detecting these gene Co-Expression modules is <a href="https://doi.org/10.1371/journal.pcbi.1004574" target="_blank">MEGENA</a>
                  (Multiscale Embedded Gene Co-Expression Network Analysis) which employs a 
                  Planar Maximally filtered graph to extract significant gene interactions and networks of co-expression modules. <br> <br>
                  MEGENA identifies gene co-expression networks through:
                  <ol><li>Identification of significant gene interactions (Planar filtered network (PFN) construction through embedding on a topological sphere). </li>
                  <li>Multiscale clustering analysis to group similar and dissimilar gene modules. </li>
                  <li>Multiscale hub analysis to detect connected hubs of individual clusters. </li></ol>
                  ICARUS performs MEGENA on a set of Seurat highly variable genes (see Dimensionality Reduction tab) or differentially expressed genes to generate a set of hierarchically ordered co-expression 
                  modules, where larger modules progressively branch into smaller submodules.<br>')), br(), br(),
                  
                  img(id="MEGENA_ICARUS",src="MEGENA_ICARUS.png",width="70%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startMEGENA", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )
            ),
            
            collapseInput(inputId = "iscollapseboxMEGENA", boxId = "MEGENABoxIntro"),
            
            fluidRow(
              box(id = "MEGENABox1", title = p("Co-Expression Modules", actionBttn("MEGENABoxInfo1", "Info"), actionBttn("downloadmodulesMEGENA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F, 
                  collapsible = T, width = 12,
                  sidebar = boxSidebar(id = "sidebarMEGENA1", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       virtualSelectInput("MEGENA_modulesInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize sub-module')), choices = NULL,
                                                          search = T,
                                                          hideClearButton = T,
                                                          placeholder = "Please type your sub-module of interest"),
                                       materialSwitch("checkboxMEGENA", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Label module hubs only')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  fluidRow(
                    column(4,
                           prettyRadioButtons("radioMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                              list("View All Modules","View Sub-module (refer to sidebar)"), inline = TRUE, selected = "View All Modules",
                                              icon = icon("check"),
                                              animation = "jelly")),
                    column(4,
                           sliderInput("genefontMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 100,
                                       value = 40, step = 1)),
                    column(4,
                           sliderInput("TranspraentMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Transparency')),
                                       min = 0, max = 1,
                                       value = 0.5, step = 0.1))
                  ),
                  
                  withSpinner(plotOutput("CoExMEGENA", height = "1000")))
            ),
            
            fluidRow(
              box(id = "MEGENABox2", title = p("Co-Expression Modules Info", actionBttn("MEGENABoxInfo2", "Info"), downloadBttn("downloadtable_MEGENA", "Download .csv")), status = "warning", solidHeader = F, 
                  collapsible = T, width = 12,
                  
                  withSpinner(DT::dataTableOutput("MEGENA_table", height = "500")))
            ),
            
            fluidRow(
              box(id = "MEGENABox3", title = p("Sunburst Plot", actionBttn("MEGENABoxInfo3", "Info"), actionBttn("downloadSunburstMEGENA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "sidebarMEGENA2", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       h4(HTML('<h4 style = "text-align:justify;color:#000000;"><b>Find Marker Genes Settings</b>')),
                                       pickerInput("testuseMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select test:')),
                                                   choices = list("wilcox", "bimod", "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2"),
                                                   width = "98%"),
                                       
                                       numericInput("log2FCMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Log2 fold change threshold')), 
                                                    min = 0, max = 10,
                                                    value = 0.5, step = 0.01, width = "98%"),
                                       
                                       numericInput("minMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">min.pct')),
                                                    min = 0, max = 1,
                                                    value = 0.3, step = 0.01, width = "98%"),
                                       pickerInput("padjustMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value adjustment method')),
                                                   choices = list("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                                                   selected = "fdr",
                                                   width = "98%"),
                                       numericInput("pvalMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                                                    min = 0.1, max = 1,
                                                    value = 0.05, step = 0.01, width = "98%"),
                                       p(style="text-align: center;", actionBttn("SunMEGENA", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(12,
                           prettyRadioButtons("radioMEGENAselect", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                              list("Cell Type Module Activity","Differentially Expressed Genes (Number of DEGs)", "Differentially Expressed Genes (Percentage of DEGs)", "Custom Differentially Expressed Genes (Number of DEGs)",
                                                   "Custom Differentially Expressed Genes (Percentage of DEGs)"), inline = TRUE, selected = "Cell Type Module Activity",
                                              icon = icon("check"),
                                              animation = "jelly"))
                  ),
                  
                  withSpinner(plotOutput("Sunburst_MEGENA", height = "1000")), br(), br(),
                  h4(htmlOutput("textMEGENA")),
                  withSpinner(DT::dataTableOutput("MEGENA_markertable")))
            ),
            
            
            fluidRow(
              box(id = "MEGENABox4", title = p("Co-Expression Module GO Heatmap", actionBttn("MEGENABoxInfo4", "Info"), actionBttn("downloadGOMEGENA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "sidebarMEGENA3", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       p(style="text-align: center;", actionBttn("GO_MODGO", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  fluidRow(
                    column(4,
                           sliderInput("GOtermsMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of top GO terms shown (per module)')), 
                                       min = 1, max = 100,
                                       value = 4, step = 1, width = "98%")),
                    column(4,
                           sliderInput("GOtextMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 1, max = 30,
                                       value = 12, step = 1)),
                    column(4,
                           sliderInput("GOheightMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 25, step = 1))),
                  
                  withSpinner(htmlOutput("GOHeatmap_MEGENAUI")))
            )
    ),
    
    ####TAB 9 - SCENIC #####
    tabItem(tabName = "SCENIC",
            fluidRow(
              box(id = "SCENICBoxIntro", title = strong("Gene Regulatory Networks"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> A cell&#8217s transcriptional state may be characterised by gene regulatory networks (GRNs) that are formed from transcription factors 
                          and cofactors that regulate each other and their downstream gene targets. ICARUS employs the <a href="https://doi.org/10.1038/nmeth.4463" target="_blank">SCENIC</a>
                          R package to characterise cell cluster/cell type specific GRNs 
                          using a set of variable genes (seurat variable features, please refer to the Dimensionality Reduction tab) or user computed differentially expressed genes (refer to the differential expression and custom differential expression tabs). <br> <br>
                          The SCENIC workflow consists of: <br>
                          <ol><li>Identification of coexpression modules with GENIE3 (between transcription factors and variable/DE genes). </li>
                          <li>Cis-regulatory motif analysis. Motifs in the promoter of genes (up to 500bp upstream the TSS) and in the 20kb around the TSS (+/- 10kb) are scored. 
                          Modules with significant motif enrichment are retained and indirect targets are removed. These modules are termed <b>regulons</b>. </li>
                          <li>Regulon activity across cell clusters are scored using the AUCell algorithm. </li></ol><br>')),
                  
                  img(id="Regulon_ICARUS",src="Regulon_ICARUS.png",width="70%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> <b>WARNING: ICARUS by default limits regulon construction to 100 cells per cluster/cell type to improve computational speed. 
                          The user may choose to increase the number of cells analysed. However, these analyses could take hour to days depending on the size of the analysis. </b>')),
                  
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startSCENIC", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )
            ),
            
            collapseInput(inputId = "iscollapseboxSCENIC", boxId = "SCENICBoxIntro"),
            
            fluidRow(
              column(3,
                     box(id = "SCENICBox1a", title = p("Regulon Expression", actionBttn("RegulonBoxInfo", "Info")), status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         
                         virtualSelectInput("SCENICInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select regulon to visualise its expression:')), choices = c("Please select from the following" = ""), selected = "",
                                            search = T,
                                            hideClearButton = F,
                                            placeholder = "Please type your regulon of interest"),
                         p(style="text-align: center;", actionBttn("regulon1", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     )
              ),
              
              column(9,
                     box(id = "SCENICBox1b", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         prettyRadioButtons("radioSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  column(6,
                                         pickerInput("plotclustersSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Regulon"),
                                                     selected = "Regulon",
                                                     width = "98%"))
                         ),
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("sizeSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacitySCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           
                           column(3,
                                  sliderInput("SCENICplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("plotSCENICUI"))
                     ))),
            
            fluidRow(
              box(id = "SCENICBox1", title = p("Gene Regulatory Network Heatmap", actionBttn("SCENICBoxInfo1", "Info"), actionBttn("downloadheatmapSCENIC", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(
                    column(6,
                           sliderInput("fontSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Font Size')),
                                       min = 0, max = 50,
                                       value = 14, step = 1)),
                    column(6,
                           sliderInput("HeightSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Height')),
                                       min = 0, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("Heatmap_SCENICUI")))
            ),
            
            fluidRow(
              box(id = "SCENICBox2", title = p("Gene Regulatory Network Dotplot", actionBttn("SCENICBoxInfo2", "Info"), actionBttn("downloaddotplotSCENIC", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  
                  fluidRow(
                    column(2,
                           prettyRadioButtons("radioSCENICrss", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                              list("All","Individual"), inline = TRUE, selected = "All",
                                              icon = icon("check"),
                                              animation = "jelly")),
                    column(5,
                           sliderInput("fontSCENIC2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Font Size')),
                                       min = 0, max = 50,
                                       value = 14, step = 1)),
                    column(5,
                           sliderInput("HeightSCENIC2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Height')),
                                       min = 0, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  fluidRow(
                    column(6,
                           virtualSelectInput("Regulon_Select2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize Individual Cell Population:')), choices = NULL,
                                              search = T,
                                              hideClearButton = F,
                                              placeholder = "Please type your cell population of interest",
                                              width = "98%"),
                    )),
                  
                  
                  withSpinner(htmlOutput("Dotplot_SCENICUI")))
            ),
            
            
            fluidRow(
              box(id = "SCENICBox3", title = p("Regulon Transcription Factor Map", actionBttn("SCENICBoxInfo3", "Info"), actionBttn("downloadRegulonSCENIC", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "SCENIC_TF", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("SCENICNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of TFs to show per regulon')),
                                                    min = 1, max = 100,
                                                    value = 20, step = 1, width = "98%"),
                                       materialSwitch("checkboxgene_SCENIC", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show TF Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxregulon_SCENIC", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Regulon Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxshowextended", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Extended Regulons')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(4,
                           sliderInput("genefontSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">TF Label Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1)),
                    column(4,
                           sliderInput("categoryfontSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Regulon Label Size')),
                                       min = 1, max = 20,
                                       value = 8, step = 1)),
                    column(4,
                           sliderInput("nodeSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1))
                  ),
                  
                  withSpinner(plotOutput("TF_Regulon_Map", height = 1500)))
            )
            
    ),
    
    ####TAB 10 - MAGMA#######
    
    tabItem(tabName = "MAGMA",
            fluidRow(
              box(id = "MAGMABoxIntro", title = strong("GWAS Associations"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Genome wide association studies have identified several loci in various genes that are associated with a trait of interest. 
                  The expression profiles of cell clusters can be compared against these GWAS loci to identify potential causal cell types underlying complex traits. <br><br>
                  ICARUS employs the multi-marker analysis of genomic annotation (MAGMA) methodology to identify increased linear association between cluster derived gene sets and GWAS traits.
                  The user may either upload their own GWAS summary statistics or select biological traits from hundreds of GWAS datasets curated by <a href="https://github.com/neurogenomics/MAGMA_Files_Public" target="_blank">Neurogenomics/MAGMA_Files_Public</a> 
                  which includes gene level association statistics for 10b upstream and 1.5kb downstream of each gene. <br><br>')),
                  
                  img(id="MAGMA_ICARUS",src="MAGMA_ICARUS.png",width="65%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonMAGMA", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "MAGMABox1", title = "MAGMA Data Input", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         pickerInput("selectMAGMAInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select GWAS Data Input')),
                                     choices = list("Please select from the following" = "", "Own GWAS Data (GWAS Summary Statistics)", "Public GWAS Datasets"), selected = "",
                                     width = "98%")
                     ),
                     
                     box(id = "MAGMABox1a", title = p("GWAS Summary Statistics", actionBttn("MAGMABoxInfo1", "Info")), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         h4(style="text-align: justify;color:#FF0000", HTML('Please make sure uploaded GWAS summary statistics file contains <q>SNP</q>, <q>CHR</q>, <q>BP</q> as first three columns AND at least one of these columns: <q>Z</q>,<q>OR</q>,<q>BETA</q>,<q>LOG_ODDS</q>,<q>SIGNED_SUMSTAT</q>.
                                                                             see below for an example dataset <br>
                                                                            NOTE: ONLY HUMAN DATASETS CAN BE PROCESSED.</a>')),
                         
                         downloadBttn("downloadtable_MAGMAExample", "Download Example GWAS Summary Statistics"),
                         
                         textInput("textlabelMAGMA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">GWAS Summary Statistics Trait:')), 
                                   placeholder = "e.g., Alzheimer's Disease"),
                         
                         numericInput("MAGMA_N", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-40px;">Number of GWAS Individuals')), 
                                      value = NULL, min = 1),
                         
                         pickerInput("MAGMA_Pop", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">GWAS Population')),
                                     choices = list("European", "African", "Ad Mixed American", "East Asian", "South Asian"),
                                     width = "98%"),
                         
                         pickerInput("GRCh", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Genome Build of GWAS Summary Statistics')),
                                     choices = list("GRCh37", "GRCh38"),
                                     width = "98%"),
                         
                         fileInput("fileinputMAGMA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load GWAS Summary Statistics')),
                                   accept = c("text",
                                              ".txt")),
                         
                         virtualSelectInput("MAGMA_samples2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sample(s)')), 
                                            choices = NULL, 
                                            multiple = T,
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = F,
                                            width = "98%"),
                         pickerInput("MAGMA_selectcluster2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by')),
                                     choices = list("Seurat Clusters", "Labelled Cells"),
                                     width = "98%"),
                         numericInput("n_MAGMA", h5(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Downsample MAGMA analysis to X cells per cluster/cell type (randomly selected)')),
                                      min = 1,
                                      value = 2000, step = 50, 
                                      width = "100%"),
                         materialSwitch("Linearmode1", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Perform Linear and Top 10% enrichment (Turn off for linear enrichment only)')), value = TRUE, status = "primary"),
                         
                         p(style="text-align: center;", actionBttn("MAGMAGO2", "START", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     ),
                     
                     box(id = "MAGMABox1b", title = p("Public GWAS Datasets", actionBttn("MAGMABoxInfo2", "Info")), status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         
                         virtualSelectInput("MAGMA_samples", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sample(s)')), 
                                            choices = NULL, 
                                            multiple = T,
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = F,
                                            width = "98%"),
                         pickerInput("MAGMA_selectcluster", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by')),
                                     choices = list("Seurat Clusters", "Labelled Cells"),
                                     width = "98%"),
                         
                         virtualSelectInput("MAGMA_GWAS", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select GWAS dataset(s)')), 
                                            choices = NULL, 
                                            multiple = T,
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = T,
                                            placeholder = "Please type your GWAS/Trait of interest",
                                            width = "98%"),
                         numericInput("n_MAGMA2", h5(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Downsample MAGMA analysis to X cells per cluster/cell type (randomly selected)')),
                                      min = 1,
                                      value = 2000, step = 50, 
                                      width = "100%"),
                         materialSwitch("Linearmode2", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Perform Linear and Top 10% enrichment (Turn off for linear enrichment only)')), value = TRUE, status = "primary"),
                         
                         p(style="text-align: center;", actionBttn("MAGMAGO", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     )
              ),
              
              column(9,
                     box(id = "MAGMABox2a", title = "Uploaded GWAS Dataset", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         withSpinner(DT::dataTableOutput("MAGMA_GWAS_table_Uploaded", height = "500"))),
                     
                     box(id = "MAGMABox2b", title = p("Public GWAS Datasets", downloadBttn("downloadtable_MAGMAGWAS", "Download .csv")), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         withSpinner(DT::dataTableOutput("MAGMA_GWAS_table", height = "500")))
              )),
            
            fluidRow(
              box(id = "MAGMABox3", title = p("GWAS Cell Cluster/Cell Type Association", actionBttn("MAGMABoxInfo3", "Info"), actionBttn("downloadMAGMAHeat", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(
                    column(6,
                           sliderInput("fontMAGMA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 1, max = 40,
                                       value = 12, step = 1)),
                    column(6,
                           sliderInput("HeightMAGMA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("MAGMA_HeatUI"))
              ),
              box(id = "MAGMABox4", title = p("GWAS Cell Cluster/Cell Type Association Info", actionBttn("MAGMABoxInfo4", "Info"), downloadBttn("downloadtable_MAGMA", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  withSpinner(DT::dataTableOutput("MAGMA_table", height = "500"))
              )),
            
            collapseInput(inputId = "iscollapseboxMAGMA", boxId = "MAGMABoxIntro")
    ),
    
    ####TAB 11 - Trajectory Analysis ####
    tabItem(tabName = "TA",
            fluidRow(
              box(id = "TABoxIntro", title = strong("Trajectory Analysis"), status = "warning", solidHeader = F,
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> The expression profiles of cells change during development, in response to stimuli and throughout life. 
                          Trajectory analysis aims to determine the sequence of gene expression changes in your dataset. ICARUS employs the 
                          <a href="https://doi.org/10.1038/nbt.2859" target="_blank">Monocle3</a> algorithm to graph cells 
                          according to their progress in <b>pseudotime</b> (a measure of how much progress an individual cell has made, i.e., cell differentiation). 
                          Pseudotime estimates the amount of transcriptional change a cell undergoes from its beginning to end states.<br><br>')),
                  
                  img(id="TrajectoryAnalysis_ICAURS",src="TrajectoryAnalysis_ICARUS.png",width="40%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color: #FF0000;"> Please use the lasso tool that can be found at the top right hand corner of the interactive plot to select your groups of cells.')),
                  img(id="lasso3",src="lasso_FINAL.png",width="35%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonTA", "START",
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "TABox1", title = p("Trajectory Analysis", actionBttn("TABoxInfo", "Info")), status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         pickerInput("TAselect", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select root cells as cell cluster/cell type or using the lasso select:')),
                                     choices = c("Please select from the following" = "", "Cell cluster/Cell type", "Lasso select"))
                     ),
                     
                     box(id = "TABox2", title = "Root cells selection (Cell cluster/Cell type)", status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         pickerInput("TA_picker", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select root node:')),
                                     choices = list("Seurat Clusters", "Labelled Cells")),
                         virtualSelectInput("TA_pickercluster", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select root cluster')), choices = NULL, multiple = T,
                                            search = T,
                                            disableSelectAll = T,
                                            hideClearButton = F),
                         p(style="text-align: center;", actionBttn("TAGO", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     ),
                     
                     box(id = "TABox3", title = "Root cells selection (lasso selection)", status = "warning", solidHeader = F,
                         collapsible = T, width = NULL,
                         h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please use the lasso select function to select your root cells, check to see if you have selected the correct cells in the table below and press APPLY to begin trajectory analysis.')),
                         p(style="text-align: center;", actionBttn("TAGO2", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     )
              ),
              
              column(9,
                     box(id = "TABox4", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         prettyRadioButtons("radioTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  column(6,
                                         pickerInput("plotclustersTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                                                     width = "98%"))
                         ),
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           column(3,
                                  sliderInput("sizeTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           column(3,
                                  sliderInput("TAplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("plotTAUI"))
                     ),
                     box(id = "TABox5", title = "Selected Cells (lasso)", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         withSpinner(DT::dataTableOutput("selectedplotTA", height = "500"))
                     ))),
            
            fluidRow(
              column(12,
                     box(id = "TABox6", title = p("Monocle3 Trajectory Visualisation", actionBttn("TABoxInfo2", "Info"), actionBttn("downloadUMAPmonocle", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(
                           column(2,
                           column(12,
                                  prettyRadioButtons("reductionmethodTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("2D","3D"), inline = TRUE, selected = "2D",
                                                     icon = icon("check"),
                                                     animation = "jelly"))),
                           column(10,
                           column(6,
                                  sliderInput("labelsizemonocleTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           column(6,
                                  sliderInput("sizemonocleTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 10,
                                              value = 5, step = 0.5))
                           )),
                         fluidRow(
                           column(2),
                           column(10,                           
                             column(6,
                                    sliderInput("opacitymonocleTA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                                min = 0, max = 1,
                                                value = 0.8, step = 0.1)),
                             column(6,
                                    sliderInput("TA2plotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                                min = 0, max = 100,
                                                value = 20, step = 1))
                             )),
                         
                         withSpinner(htmlOutput("plotmonocleTAUI"))
                     )
              )),
            
            fluidRow(
              column(12,
                     box(id = "TABox7", title = p("Pseudotime Visualisation", actionBttn("TABoxInfo3", "Info"), actionBttn("downloadmonocle3viz", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(
                           column(2,
                                  prettyRadioButtons("monocle3pseudo", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("Seurat Clusters","Labelled Cells"), inline = TRUE, selected = "Seurat Clusters",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(5,
                                  sliderInput("fontsizepseudotimeviz", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           
                           column(5,
                                  sliderInput("Heightpseudotimeviz", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("monocle3vizUI"))
                     ))
            ),
            
            fluidRow(
              column(12,
                     box(id = "TABox8", title = p("Genes Changes Across Pseudotime", actionBttn("TABoxInfo4", "Info"), downloadBttn("downloadmonocle3DEG", "Download .csv", icon = icon("download"))), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         sidebar = boxSidebar(id = "sidebarTA", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                              p(style="text-align: center;", actionBttn("GO_TA", "START", icon("play"),
                                                                                        style = "jelly",
                                                                                        color = "success")),
                                              p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                         
                         withSpinner(DT::dataTableOutput("monocle3DEG", height = "500"))
                     ))
            ),
            
            fluidRow(
              column(3,
                     box(id = "TABox9", title = p("Gene Input", actionBttn("TABoxInfo5", "Info")), status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         
                         virtualSelectInput("MonocleInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Enter single or multiple genes:')), choices = NULL, multiple = T, 
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = T,
                                            placeholder = "Please type your gene(s) of interest"
                         )
                     )
              ),
              
              column(9,
                     box(id = "TABox10", title = p("Gene Expression Across Pseudotime", actionBttn("TABoxInfo6", "Info"), actionBttn("downloadmonocleGAP", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         sliderInput("fontsizemonocle3select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                                     min = 1, max = 50,
                                                     value = 12, step = 1)),
                                  
                                  column(6,
                                         sliderInput("HeightDEGviz", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                                     min = 1, max = 100,
                                                     value = 20, step = 1))
                         ),
                         
                         withSpinner(htmlOutput("monocle3DEGvizUI"))
                     ))),
            
            collapseInput(inputId = "iscollapseboxTA", boxId = "TABoxIntro")
    ),
    
    ####TAB 12 - CellChat #####
    tabItem(tabName = "Chat",
            fluidRow(
              box(id = "ChatBoxIntro", title = strong("Cell-Cell Communication"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Cells interact and communicate with each other. This interaction may be between cells of similar cell types, cells of different cell types 
                          and/or cells at different states (i.e., diseased vs phenotypically normal). The coordination of cell-cell communication is vital for many biological processes including cell cycle 
                          activation, initiation of apoptosis and activation of cellular development and differentiation. <br> <br>
                          ICARUS employs the <a href="https://doi.org/10.1038/s41467-021-21246-9" target="_blank">CellChat</a> tool to quantitatively infer and analyse intercellular communication networks from 
                          single cell RNA-seq data. CellChat incorporates a manually curated comprehensive database of ligand-receptor pairs, soluble agonists/antagonists and stimulatory/inhibitory 
                          membrane bound co-receptors to infer cell-cell communication interactions based on social network analysis tools, pattern recognition methods and manifold learning approaches. <br><br>')),
                  
                  img(id="CellChat_ICARUS",src="CellChat_ICARUS.png",width="70%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startChat", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            collapseInput(inputId = "iscollapseboxChat", boxId = "ChatBoxIntro"),
            
            fluidRow(
              box(id = "ChatBox1", title = p("Cell-Cell Communication Network (Single Dataset)", actionBttn("ChatBoxInfo1", "Info"), actionBttn("downloadCommunicationChat", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "Cell-Cellside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("Communication_Select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Cell Cluster/Cell Type')),
                                                   choices = NULL,
                                                   width = "98%"),
                                       materialSwitch("checkboxCommunication", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show ALL')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  withSpinner(plotOutput("Interactions_CellChat", height = "800")))
            ),
            
            fluidRow(
              box(id = "ChatBox2", title = p("Signalling Pathways", actionBttn("ChatBoxInfo2", "Info"), actionBttn("downloadSignalChat", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(3,
                                  prettyRadioButtons("radioChat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("Signalling Roles", "Hierarchy Plot", "Circle Plot","Chord Plot", "Heatmap"), inline = TRUE, selected = "Hierarchy Plot",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(3,
                                  sliderInput("sizeSignalling", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 10, step = 1)),
                           column(3,
                                  sliderInput("WidthSignalling_Chat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Width')),
                                              min = 1, max = 100,
                                              value = 50, step = 1)),
                           column(3,
                                  sliderInput("HeightSignalling_Chat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  br(),
                  
                  fluidRow(
                    column(4,
                           virtualSelectInput("Signalling_Select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Signalling Pathway')),
                                              choices = NULL,
                                              search = T,
                                              placeholder = "Please type your signalling pathway of interest")),
                  ),
                  
                  withSpinner(htmlOutput("Signalling_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox3", title = p("Ligand-Receptor Interaction For Signalling Pathways", actionBttn("ChatBoxInfo3", "Info"), actionBttn("downloadLigandChat", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(4,
                                  prettyRadioButtons("radioLigandChat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("Bar Graph","Dot Plot","Chord Plot", "Gene Expression"), inline = TRUE, selected = "Bar Graph",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(4,
                                  sliderInput("sizeLigand", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(4,
                                  sliderInput("HeightLigand", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  br(),
                  
                  fluidRow(
                    column(4,
                           virtualSelectInput("Ligand_Select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Signalling Pathway')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your signalling pathway of interest")),
                    column(4,
                           virtualSelectInput("source_select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Source Cell Population')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your source cell population/cell type of interest")),
                    column(4,
                           virtualSelectInput("target_select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Target Cell Population')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your target cell population/cell type of interest"))
                  ),
                  
                  withSpinner(htmlOutput("Ligand_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox4", title = p("Signalling Roles", actionBttn("ChatBoxInfo4", "Info"), actionBttn("downloadRolesChat", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(3,
                                  prettyRadioButtons("radioRolesChat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("Heatmap","Scatter Plot"), inline = TRUE, selected = "Heatmap",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(3,
                                  sliderInput("sizeRoles", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(3,
                                  sliderInput("WidthRoles", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Width')),
                                              min = 1, max = 100,
                                              value = 25, step = 1)),
                           column(3,
                                  sliderInput("HeightRoles", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("Roles_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox5", title = p("Signalling Patterns", actionBttn("ChatBoxInfo5", "Info"), actionBttn("downloadPatternsChat", "Download Plot (Incoming)", icon = icon("download")), actionBttn("downloadPatternsChat2", "Download Plot (Outgoing)", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "Patternsside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       sliderInput("Patterns_Select", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Number of Patterns')),
                                                   min = 1, max = 5,
                                                   value = 2, step = 1,
                                                   width = "98%"),
                                       p(style="text-align: center;", actionBttn("GOPATTERNS", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(column(3,
                                  prettyRadioButtons("radioPatternsChat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("River Plot", "Dot Plot"), inline = TRUE, selected = "River Plot",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(3,
                                  sliderInput("sizePatterns", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(3,
                                  sliderInput("WidthPatterns", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Width')),
                                              min = 1, max = 100,
                                              value = 18, step = 1)),
                           column(3,
                                  sliderInput("HeightPatterns", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  withSpinner(htmlOutput("Patterns_CellChatUI")),
                  br(),
                  withSpinner(htmlOutput("Patterns_CellChatUI2")))
            ),
            
            fluidRow(
              box(id = "ChatBox6", title = p("Manifold and Classification Learning Analysis of Signaling Networks", actionBttn("ChatBoxInfo6", "Info"), actionBttn("downloadManifoldChat", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(4,
                                  prettyRadioButtons("radioManifoldChat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("Functional Classification","Structural Classification"), inline = TRUE, selected = "Structural Classification",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(4,
                                  sliderInput("sizeManifold", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 10,
                                              value = 6, step = 1)),
                           column(4,
                                  sliderInput("HeightManifold", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("Manifold_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox7", title = p("Cell-Cell Communication Network (Two Datasets)", actionBttn("ChatBoxInfo7", "Info"), actionBttn("downloadCommunicationChatM", "Download Plot (Number of Interactions)", icon = icon("download")), actionBttn("downloadCommunicationChatM2", "Download Plot (Strength of Interactions)", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "Cell-CellsideM", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("Communication_SelectM1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Cell Cluster/Cell Type (Dataset #1)')),
                                                   choices = NULL,
                                                   width = "98%"),
                                       pickerInput("Communication_SelectM2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Cell Cluster/Cell Type (Dataset #2)')),
                                                   choices = NULL,
                                                   width = "98%"),
                                       materialSwitch("checkboxCommunicationM", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show ALL')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  withSpinner(plotOutput("Interactions_CellChatM", height = "800")),
                  withSpinner(plotOutput("Interactions_CellChatM2", height = "800"))
              )
            ),
            
            fluidRow(
              box(id = "ChatBox8", title = p("Differential Interaction (Comparison Between Datasets)", actionBttn("ChatBoxInfo8", "Info"), actionBttn("downloadCellCell", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(4,
                                  prettyRadioButtons("radioCellCell", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("Bar Graph", "Communication Network", "Heatmap"), inline = TRUE, selected = "Communication Network",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(4,
                                  sliderInput("sizeCellCell", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(4,
                                  sliderInput("HeightCellCell", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("CellCell_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox9", title = p("Signalling Pathways", actionBttn("ChatBoxInfo9", "Info"), actionBttn("downloadSignalChatM", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(3,
                                  prettyRadioButtons("radioChatM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("Signalling Roles", "Hierarchy Plot", "Circle Plot","Chord Plot", "Heatmap"), inline = TRUE, selected = "Hierarchy Plot",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(3,
                                  sliderInput("sizeSignallingM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 10, step = 1)),
                           column(3,
                                  sliderInput("WidthSignalling_ChatM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Width')),
                                              min = 1, max = 100,
                                              value = 50, step = 1)),
                           column(3,
                                  sliderInput("HeightSignalling_ChatM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  br(),
                  
                  fluidRow(
                    column(4,
                           virtualSelectInput("Signalling_SelectM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Signalling Pathway')),
                                              choices = NULL,
                                              search = T,
                                              placeholder = "Please type your signalling pathway of interest")),
                    column(4,
                           virtualSelectInput("SignallingDataset_SelectM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Dataset')),
                                              choices = c("Dataset 1", "Dataset 2"))),
                  ),
                  br(), br(),
                  h4(strong(htmlOutput("SignallingM_CellChattext"))),
                  withSpinner(htmlOutput("SignallingM_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox10", title = p("Ligand-Receptor Interaction For Signalling Pathways", actionBttn("ChatBoxInfo10", "Info"), actionBttn("downloadLigandChatM", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(4,
                                  prettyRadioButtons("radioLigandChatM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')), 
                                                     list("Dot Plot", "Gene Expression"), inline = TRUE, selected = "Dot Plot",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(4,
                                  sliderInput("sizeLigandM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(4,
                                  sliderInput("HeightLigandM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  br(),
                  
                  fluidRow(
                    column(4,
                           virtualSelectInput("Ligand_SelectM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Signalling Pathway')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your signalling pathway of interest")),
                    column(4,
                           virtualSelectInput("source_selectM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Source Cell Population')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your source cell population/cell type of interest")),
                    column(4,
                           virtualSelectInput("target_selectM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Target Cell Population')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your target cell population/cell type of interest"))
                  ),
                  
                  withSpinner(htmlOutput("LigandM_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox11", title = p("Identification of Upregulated and Downregulated Signalling Ligand-Receptor Pairs", actionBttn("ChatBoxInfo11", "Info"), actionBttn("downloadLigandChatMM", "Download Plot (Upregulated)", icon = icon("download")), actionBttn("downloadLigandChatMM2", "Download Plot (Downregulated)", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "Cell-CellsideMM", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("Dataset_SelectMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Query Dataset')),
                                                   choices = NULL),
                                       pickerInput("Dataset_SelectMM2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Comparison Dataset')),
                                                   choices = NULL),
                                       p(style="text-align: center;", actionBttn("GOCHAT", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(column(6,
                                  sliderInput("sizeLigandMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(6,
                                  sliderInput("HeightLigandMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  fluidRow(
                    column(4,
                           virtualSelectInput("source_selectMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Source Cell Population')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your source cell population/cell type of interest")),
                    column(4,
                           virtualSelectInput("target_selectMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Target Cell Population')),
                                              choices = NULL,
                                              multiple = T,
                                              search = T,
                                              hideClearButton = F,
                                              disableSelectAll = F,
                                              placeholder = "Please type your target cell population/cell type of interest"))
                  ),
                  
                  withSpinner(htmlOutput("LigandMM1_CellChatUI")),
                  withSpinner(htmlOutput("LigandMM2_CellChatUI"))
              )
            ),
            
            fluidRow(
              box(id = "ChatBox12", title = p("Signalling Roles", actionBttn("ChatBoxInfo12", "Info"), actionBttn("downloadRolesChatM", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(3,
                                  prettyRadioButtons("radioRolesChatM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("Bar Graph","Scatter Plot", "Heatmap"), inline = TRUE, selected = "Bar Graph",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(3,
                                  sliderInput("sizeRolesM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 50,
                                              value = 12, step = 1)),
                           column(3,
                                  sliderInput("WidthRolesM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Width')),
                                              min = 1, max = 100,
                                              value = 25, step = 1)),
                           column(3,
                                  sliderInput("HeightRolesM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  br(),
                  
                  fluidRow(
                    column(4,
                           virtualSelectInput("signallingroles_MM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Signalling Role')),
                                              choices = list("Incoming", "Outgoing", "All"),
                                              selected = "All")),
                  ),
                  
                  withSpinner(htmlOutput("RolesM_CellChatUI")))
            ),
            
            fluidRow(
              box(id = "ChatBox13", title = p("Manifold and Classification Learning Analysis of Signaling Networks", actionBttn("ChatBoxInfo13", "Info"), actionBttn("downloadManifoldChatM", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(4,
                                  prettyRadioButtons("radioManifoldChatM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("Functional Classification","Structural Classification"), inline = TRUE, selected = "Structural Classification",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(4,
                                  sliderInput("sizeManifoldM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 1, max = 10,
                                              value = 6, step = 1)),
                           column(4,
                                  sliderInput("HeightManifoldM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 1, max = 100,
                                              value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("ManifoldM_CellChatUI")))
            )
            
    ),
    
    ####TAB 13 - Multimodal#######
    
    tabItem(tabName = "MM",
            fluidRow(
              box(id = "MMBoxIntro", title = strong("Multimodal Analysis"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Multimodal analysis refers to the simultaneous measurements of several data types from the same cell.
                  For example, methodologies including <a href="https://doi.org/10.1038/nmeth.4380" target="_blank">CITE-seq</a>, 
                  <a href="https://doi.org/10.1186/s13059-018-1603-1" target="_blank">cell hasing oligos</a> and 
                  <a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression" target="_blank">10X multiome kit</a> 
                  allow measurements of single cell transcriptomes and cell-surface proteins of the same cell.
                  ICARUS enables a side by side visualisation of multi-modal data to enable easy comparisons and analysis. <br> <br>
                  
                  ICARUS requires a matrix of multimodal data gene counts. This may be uploaded as a: 
                  <ol><li><b>Tab delimited</b> table with cells as columns and gene features as rows (see example A). Please make sure the cell names (column names) match the loaded scRNA-seq cell names. <br> </li>
                  <li><b>10X data</b> files (barcodes.tsv, features.tsv and matrix.mtx, refer to B).</li>
                  <li>Saved <b>Seurat object </b> (RDS file, refer to C).</li></ol><br>')),
                  
                  img(id="Multimodal_ICARUS",src="Multimodal_ICARUS.png",width="60%"), br(), br(),
                  
                  fluidRow(
                    column(3,       
                           valueBox("NEW",
                                    actionBttn("startbuttonMM", "START",
                                               style = "jelly",
                                               color = "primary"),
                                    icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(
                       id = "MMBox4", title = "Load your multimodal data", status = "warning", solidHeader = F, 
                       collapsible = T, collapsed = T, width = NULL,
                       
                       prettyRadioButtons("radioLoadMM", label = h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose your load method')),
                                          choices = list("10X data", "Count Matrix (txt)", "Seurat object (RDS file)", "Sample datasets"), 
                                          selected = "Sample datasets",
                                          icon = icon("check"),
                                          animation = "jelly"), 
                       
                       fileInput("fileinputMMload1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load barcodes.tsv, features.tsv and matrix.mtx files (can be gzipped .gz)')), 
                                 multiple = T),
                       
                       actionBttn("buttonfileinputMMload1", "Submit",
                                  style = "jelly",
                                  color = "primary"),
                       
                       fileInput("fileinputMMload2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load count matrix (tab delimited table of gene counts)')),
                                 accept = c("text",
                                            ".txt")),
                       actionBttn("buttonfileinputMMload2", "Submit",
                                  style = "jelly",
                                  color = "primary"),
                       
                       fileInput("fileinputMMload3", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Load seurat object (RDS file format)')),
                                 accept = c(".rds", ".RDS")),
                       actionBttn("buttonfileinputMMload3", "Submit",
                                  style = "jelly",
                                  color = "primary"),
                       
                       pickerInput("existingdataMM",h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Choose from the following datasets')), choices = c("Please select from the following" = "",
                                                                                                                                                                                  "8k CBMCs (Seurat CITE-seq multimodal tutorial)")),
                       actionBttn("buttonexistingdataMM", "Submit",
                                  style = "jelly",
                                  color = "primary")
                       
                     )
              ),
              column(9,
                     box(id = "MMBox5", title = "Multimodal count matrix (preview first 500 rows and columns)", status = "warning", 
                         collapsible = T, collapsed = T, width = NULL,
                         
                         withSpinner(DT::dataTableOutput("MatrixTableMM", height = "600"))
                     ))
            ),
            
            fluidRow(
              column(3,
                     box(id = "MMBox1", title = p("Gene Expression", actionBttn("MMBoxInfo", "Info")), status = "warning", solidHeader = F, 
                         collapsible = T, width = NULL,
                         
                         virtualSelectInput("MMInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Enter gene:')), choices = NULL, multiple = F, 
                                            search = T,
                                            placeholder = "Please select gene(s) to view its expression"),
                         p(style="text-align: center;", actionBttn("geneMM", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                     )
              ),
              
              column(9,
                     box(id = "MMBox2", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         prettyRadioButtons("radioMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  column(6,
                                         pickerInput("plotclustersMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Gene expression"),
                                                     selected = "Gene expression"))
                         ),
                         fluidRow(
                           column(3,
                                  sliderInput("labelsizeMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("sizeMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(3,
                                  sliderInput("opacityMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1)),
                           column(3,
                                  sliderInput("MMplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                              min = 0, max = 100,
                                              value = 20, step = 1))
                         ),
                         fluidRow( 
                           column(6,
                                  strong("scRNA-seq"),        
                                  withSpinner(htmlOutput("plotMM1UI"))
                           ),
                           column(6,
                                  strong("Multimodal analysis"),
                                  withSpinner(htmlOutput("plotMM2UI"))
                           )
                         )
                     )
              )
            ),
            
            fluidRow(
              box(id = "MMBox3", title = p("Violin Plots", actionBttn("downloadMMViolin", "Download Plot (scRNA-seq)", icon = icon("download")),
                                           actionBttn("downloadMMViolin2", "Download Plot (Multimodal data)", icon = icon("download"))
              ), status = "warning", solidHeader = F,
              collapsible = TRUE, width = 12,
              sidebar = boxSidebar(id = "boxsidebarviolinMM", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                   pickerInput("VlnInputMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by:')),
                                               choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase"),
                                               selected = "Labelled Cells", width = "98%"),
                                   materialSwitch("checkboxvlnMM", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Split by sample')), value = FALSE),
                                   p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
              ),
              
              fluidRow(
                column(4,
                       sliderInput("VlnsizeMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                   min = 0, max = 50,
                                   value = 10, step = 1)),
                column(4,
                       sliderInput("VlnptMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Point Size')),
                                   min = 0, max = 5,
                                   value = 2, step = 0.5)),
                
                column(4,
                       sliderInput("HeightMMVln", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                   min = 1, max = 1000,
                                   value = 15, step = 1))
              ),
              
              strong("scRNA-seq"),
              withSpinner(htmlOutput("VlnMMUI")),
              strong("Multimodal analysis"),
              withSpinner(htmlOutput("VlnMM2UI"))
              )),
            
            
            collapseInput(inputId = "iscollapseboxMM", boxId = "MMBoxIntro"),
            collapseInput(inputId = "iscollapseboxMM2", boxId = "MMBox4"),
            collapseInput(inputId = "iscollapseboxMM3", boxId = "MMBox5")
            
            
    ),
    
    
    ####TAB 14 - Differential expression######
    tabItem(tabName = "DE",
            fluidRow(
              box(id = "DEBoxIntro", title = strong("Differential Expression"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> ICARUS enables users to perform pairwise differential expression tests between clusters of interest or between samples within cell cluster(s) and output a list of differentially expressed genes. 
                  Statistical significance testing between pairwise observations is conducted with the Wilcoxon rank sum test (Wilcoxon-Mann-Whitney test) with Bonferroni multiple testing correction. 
                  The list of differentially expressed genes are then probed for enriched gene pathways (gene set enrichment analysis) to reveal potential affected biological pathways.
                  Gene set enrichment and visualisation performed using 
                  <a href="https://doi.org/10.1016/j.xinn.2021.100141" target="_blank">ClusterProfiler</a> and
                  <a href="https://doi.org/10.1039/C5MB00663E" target="_blank">ReactomePA</a> R packages. <br> <br>
                  Several outputs are produced in this step including: <br>
                  <ol><li>Table of differentially expressed genes. </li>
                  <li>Table of enriched pathways (Gene Ontology/WikiPathways/Reactome/KEGG pathways).</li>
                  <li>Volcano plot of differentially expressed genes.</li>
                  <li>Visualisation of log2 fold changes for specific gene pathway. </ol></li><br>')),
                  
                  img(id="DifferentialExpression_ICARUS",src="DifferentialExpression_ICARUS.png",width="80%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonDE", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              column(3,
                     box(id = "DEBox1", title = "Differential Expression", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         pickerInput("SelectComparison", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Compare between clusters or between samples')), 
                                     choices = list("Please select from the following" = "", "Between Clusters", "Between Samples"), selected = "")
                     ),
                     box(id = "DEBox1a", title = "DE Between Clusters", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         pickerInput("DE_clusters", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select clusters by:')),
                                     choices = list("Seurat Clusters", "Labelled Cells")),
                         
                         virtualSelectInput("DEInput1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cluster(s) 1')), choices = NULL, multiple = T,
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = F),
                         
                         p(style="text-align: center;", h4(HTML('<h4 style = "text-align: center; color:#000000;"><b>VS</b>'))),
                         
                         virtualSelectInput("DEInput2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cluster(s) 2')), choices = NULL, multiple = T,
                                            search = T,
                                            hideClearButton = F,
                                            disableSelectAll = F),
                         
                         p(style="text-align: center;", actionBttn("DE1", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                         
                     ),
                     box(id = "DEBox1b", title = "DE Between Samples", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         pickerInput("DE_clusters_samples", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select clusters by:')),
                                     choices = list("Seurat Clusters", "Labelled Cells")),
                         
                         virtualSelectInput("DECluster", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cluster(s)')), choices = NULL, multiple = T,
                                            search = T,
                                            disableSelectAll = F,
                                            hideClearButton = F),
                         
                         virtualSelectInput("DEInput1_samples", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sample(s) 1')), choices = NULL, multiple = T,
                                            search = T,
                                            disableSelectAll = F,
                                            hideClearButton = F),
                         
                         p(style="text-align: center;", h4(HTML('<h4 style = "text-align: center; color:#000000;"><b>VS</b>'))),
                         
                         virtualSelectInput("DEInput2_samples", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sample(s) 2')), choices = NULL, multiple = T,
                                            search = T,
                                            disableSelectAll = F,
                                            hideClearButton = F),
                         
                         p(style="text-align: center;", actionBttn("DE2", "APPLY", icon("play"),
                                                                   style = "jelly",
                                                                   color = "success"))
                         
                     ),
                     p(style="text-align: center;", actionBttn("NextButtonDE", "CONTINUE TO PATHWAY ANALYSIS...", 
                                                               style = "jelly",
                                                               color = "primary",
                                                               icon = icon("play"), block = T))
              ),
              
              column(9,
                     box(id = "DEBox2", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                         collapsible = TRUE, width = NULL,
                         
                         fluidRow(column(6,
                                         prettyRadioButtons("radioDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                            list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                            icon = icon("check"),
                                                            animation = "jelly")),
                                  column(6,
                                         pickerInput("plotclustersDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                                     choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Differential Expression"),
                                                     selected = "Sample"))
                         ),
                         fluidRow(
                           column(4,
                                  sliderInput("labelsizeDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                              min = 0, max = 50,
                                              value = 10, step = 1)),
                           
                           column(4,
                                  sliderInput("sizeDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                              min = 0, max = 20,
                                              value = 10, step = 1)),
                           
                           column(4,
                                  sliderInput("opacityDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                              min = 0, max = 1,
                                              value = 0.8, step = 0.1))
                         ),
                         
                         withSpinner(plotlyOutput("plotDE", height = "1200"))
                     )
              )),
            
            fluidRow( 
              box(id = "DEBox3", title = p("Differentially Expressed Genes", actionBttn("DEBoxInfo1", "Info"), downloadBttn("downloadmarkersDE", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "boxsidebar_DE", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       
                                       pickerInput("testuseDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select test:')),
                                                   choices = list("wilcox", "bimod", "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2"),
                                                   width = "98%"),
                                       
                                       numericInput("logDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Log2 fold change threshold')), 
                                                    min = 0, max = 1,
                                                    value = 0, step = 0.01, width = "98%"),
                                       
                                       numericInput("minDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">min.pct')),
                                                    min = 0, max = 1,
                                                    value = 0.25, step = 0.01, width = "98%"),
                                       
                                       materialSwitch("onlyposDE", label = h4(HTML('<h4 style = "text-align:justify;color:#000000">Return only upregulated genes')), value = FALSE, status = "primary"),
                                       
                                       pickerInput("padjustDE1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value adjustment method')),
                                                   choices = list("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                                                   selected = "fdr",
                                                   width = "98%"),
                                       
                                       numericInput("pvalDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                                                    min = 0.1, max = 1,
                                                    value = 0.05, step = 0.01, width = "98%"),
                                       p(style="text-align: center;", actionBttn("buttonDEGDE", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  withSpinner(DT::dataTableOutput("markersDE", height = "600"))
              ),
              
              box(id = "DEBox4", title = p("Gene Set Enrichment", actionBttn("DEBoxInfo2", "Info"), downloadBttn("downloadYuDE", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "boxsidebarDE", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("enrichInputDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select enrichment analysis')), 
                                                   choices = list("Gene Ontology", "WikiPathways",
                                                                  "KEGG Pathways", "REACTOME"),
                                                   selected = "Gene Ontology", width = "98%"),
                                       
                                       pickerInput("GOlabelsDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sub-ontology')), 
                                                   choices = list("BP", "MF",
                                                                  "CC"),
                                                   selected = "BP",
                                                   width = "98%"),
                                       
                                       pickerInput("padjustDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value adjustment method')),
                                                   choices = list("none", "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                                                   selected = "fdr",
                                                   width = "98%"),
                                       
                                       numericInput("pvalueDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')), 
                                                    value = 0.05, min = 0, max = 1, step = 0.001,
                                                    width = "98%"),
                                       
                                       p(style="text-align: center;", actionBttn("buttonEnrichDE", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  
                  withSpinner(DT::dataTableOutput("GO_DE", height = "600"))
              )
            ),
            
            fluidRow( 
              box(id = "DEBox5", title = p("Volcano Plot", actionBttn("DEBoxInfo3", "Info"), actionBttn("downloadvolcanoDE", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "volcanoDE", startOpen = TRUE, width = 50, background = "#e8ebe9",
                                       
                                       sliderInput("logvolcano", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Log2 fold change threshold')), 
                                                   min = -2.5, max = 2.5,
                                                   value = c(-0.5, 0.5), step = 0.01, width = "98%"),
                                       
                                       numericInput("pvolcano", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                                                    min = 0, max = 1,
                                                    value = 0.05, step = 0.001, width = "98%"),
                                       p(style="text-align: center;", actionBttn("buttonvolcano", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  fluidRow(
                    column(4,
                           sliderInput("labelsizevolcanoDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    column(4,
                           sliderInput("sizevolcanoDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                       min = 0, max = 5,
                                       value = 2, step = 0.1)),
                    
                    column(4,
                           sliderInput("opacityvolcanoDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                       min = 0, max = 1,
                                       value = 0.8, step = 0.1))
                  ),
                  
                  withSpinner(plotlyOutput("volcanoplot", height = "600"))
              ),
              
              box(id = "DEBox6", title = p("Gene Pathways", actionBttn("DEBoxInfo4", "Info"), actionBttn("downloadPathwayPlotDE", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "GenePathwaysDE", startOpen = TRUE, width = 50, background = "#e8ebe9",
                                       
                                       virtualSelectInput("GPInput", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize gene pathways:')), choices = NULL,
                                                          search = T,
                                                          placeholder = "Please type your pathway of interest",
                                                          width = "98%"),
                                       p(style="text-align: center;", actionBttn("GP_actions", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  
                  
                  fluidRow(
                    column(4,
                           sliderInput("labelsizeGPDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    
                    column(4,
                           sliderInput("sizeGPDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                       min = 0, max = 5,
                                       value = 2, step = 0.1)),
                    
                    column(4,
                           sliderInput("opacityGPDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                       min = 0, max = 1,
                                       value = 0.8, step = 0.1))
                  ),
                  
                  
                  withSpinner(plotlyOutput("GP_plot", height = "600"))
              )
            ),
            
            collapseInput(inputId = "iscollapsebox6", boxId = "DEBoxIntro")
            
            
    ),
    
    ####TAB 15 - Custom Differential expression######
    tabItem(tabName = "SDE",
            fluidRow(
              box(id = "SDEBoxIntro", title = strong("Custom Differential Expression"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> This step allows the user to investigate genes that are differentially
                  expressed between user selected cells. Pairwise differential expression tests are computed between selected cell groups and a list of differentially expressed genes is produced. 
                  Statistical significance testing between pairwise observations is conducted with the Wilcoxon rank sum test (Wilcoxon-Mann-Whitney test) with Bonferroni multiple testing correction. 
                  Gene set enrichment analysis of significantly differentially expressed genes may reveal affected gene pathways. Gene set enrichment and visualisation performed using 
                  <a href="https://doi.org/10.1016/j.xinn.2021.100141" target="_blank">ClusterProfiler</a> and
                  <a href="https://doi.org/10.1039/C5MB00663E" target="_blank">ReactomePA</a> R packages. <br> <br>
                  Several outputs are produced in this step including: <br>
                  <ol><li>Table of differentially expressed genes. </li>
                  <li>Table of enriched pathways (Gene Ontology/WikiPathways/Reactome/KEGG pathways).</li>
                  <li>Volcano plot of differentially expressed genes.</li>
                  <li>Visualisation of log2 fold changes for specific gene pathway. </ol></li> <br>')),
                  
                  img(id="DifferentialExpression2_ICARUS",src="DifferentialExpression_ICARUS.png",width="80%"), br(), br(),
                  
                  h4(HTML('<h4 style = "text-align:justify;color:#FF0000"> Please use the lasso tool that can be found at the top right hand corner of the interactive plot to select your groups of cells. 
                          Press COMPARE to compare the two groups of lasso selected cells.')),
                  
                  img(id="lasso",src="lasso_FINAL.png",width="35%"), br(), br(),
                  
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonSDE", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            fluidRow(
              box(id = "SDEBox1", title = "UMAP/t-SNE Plot", status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  fluidRow(column(6,
                                  prettyRadioButtons("radioSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                                     list("UMAP","3D UMAP","t-SNE", "3D t-SNE"), inline = TRUE, selected = "UMAP",
                                                     icon = icon("check"),
                                                     animation = "jelly")),
                           column(6,
                                  pickerInput("plotclustersSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label cells by:')),
                                              choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                             "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"))),
                  ),
                  fluidRow(
                    column(3,
                           sliderInput("labelsizeSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    
                    column(3,
                           sliderInput("sizeSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                       min = 0, max = 20,
                                       value = 10, step = 1)),
                    
                    column(3,
                           sliderInput("opacitySDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                       min = 0, max = 1,
                                       value = 0.8, step = 0.1)),
                    column(3,
                           sliderInput("customplotheight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  fluidRow( 
                    column(6,
                           h4(HTML('<h4 style = "text-align:justify;color:#000000;"><b>Use lasso tool to select cells for group 1</b>')),        
                           withSpinner(htmlOutput("plotSDE1UI"))
                    ),
                    column(6,
                           h4(HTML('<h4 style = "text-align:justify;color:#000000;"><b>Use lasso tool to select cells for group 2</b>')),
                           withSpinner(htmlOutput("plotSDE2UI"))
                    )
                  ),
                  p(style="text-align: center;", actionBttn("SelectedCellsGO", "COMPARE", icon = icon("dna"),
                                                            style = "jelly",
                                                            color = "success",
                                                            size = "lg"))             
              ),
              p(style="text-align: center;", actionBttn("NextButtonSDE", "CONTINUE TO PATHWAY ANALYSIS...", 
                                                        style = "jelly",
                                                        color = "primary",
                                                        icon = icon("play"), block = T))
            ),
            
            fluidRow( 
              box(id = "SDEBox2", title = p("Selected Cells (Group 1)", downloadBttn("downloadlasso1", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  
                  withSpinner(DT::dataTableOutput("selectedplot1", height = "700"))
              ),
              
              box(id = "SDEBox3", title = p("Selected Cells (Group 2)", downloadBttn("downloadlasso2", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  
                  withSpinner(DT::dataTableOutput("selectedplot2", height = "700"))
              )
            ),
            
            fluidRow( 
              box(id = "SDEBox4", title = p("Differentially Expressed Genes", actionBttn("SDEBoxInfo1", "Info"), downloadBttn("downloadmarkersSDE", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "boxsidebar_SDE", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       
                                       pickerInput("testuseSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select test:')),
                                                   choices = list("wilcox", "bimod", "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2"),
                                                   width = "98%"),
                                       
                                       numericInput("logSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Log2 fold change threshold')), 
                                                    min = 0, max = 1,
                                                    value = 0, step = 0.01, width = "98%"),
                                       
                                       numericInput("minSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">min.pct')),
                                                    min = 0, max = 1,
                                                    value = 0.25, step = 0.01, width = "98%"),
                                       materialSwitch("onlyposSDE", label = h4(HTML('<h4 style = "text-align:justify;color:#000000">Return only upregulated genes')), value = FALSE, status = "primary"),
                                       
                                       pickerInput("padjustSDE1", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value adjustment method')),
                                                   choices = list("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                                                   selected = "fdr",
                                                   width = "98%"),
                                       numericInput("pvalSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                                                    min = 0.1, max = 1,
                                                    value = 0.05, step = 0.01, width = "98%"),
                                       p(style="text-align: center;", actionBttn("buttonSDEGDE", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  withSpinner(DT::dataTableOutput("markersSDEs", height = "600"))
              ),
              
              box(id = "SDEBox5", title = p("Gene Set Enrichment", actionBttn("SDEBoxInfo2", "Info"), downloadBttn("downloadYuSDE", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "boxsidebarSDE", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       pickerInput("enrichInputSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select enrichment analysis')), 
                                                   choices = list("Gene Ontology", "WikiPathways",
                                                                  "KEGG Pathways", "REACTOME"),
                                                   selected = "Gene Ontology",
                                                   width = "98%"),
                                       
                                       pickerInput("GOlabelsSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sub-ontology')), 
                                                   choices = list("BP", "MF",
                                                                  "CC"),
                                                   selected = "BP",
                                                   width = "98%"),
                                       
                                       pickerInput("padjustSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value adjustment method')),
                                                   choices = list("none", "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                                                   selected = "fdr",
                                                   width = "98%"),
                                       
                                       numericInput("pvalueSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')), 
                                                    value = 0.05, min = 0, max = 1, step = 0.001,
                                                    width = "98%"),
                                       
                                       p(style="text-align: center;", actionBttn("buttonEnrichSDE", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  
                  withSpinner(DT::dataTableOutput("GO_SDEs", height = "600"))
              )
            ),
            
            fluidRow( 
              box(id = "SDEBox6", title = p("Volcano Plot", actionBttn("SDEBoxInfo3", "Info"), actionBttn("downloadvolcanoSDE", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "volcanoSDE", startOpen = TRUE, width = 50, background = "#e8ebe9",
                                       
                                       sliderInput("logvolcanoSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Log2 fold change threshold')), 
                                                   min = -2.5, max = 2.5,
                                                   value = c(-0.5, 0.5), step = 0.01, width = "98%"),
                                       
                                       numericInput("pvolcanoSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                                                    min = 0, max = 1,
                                                    value = 0.05, step = 0.001, width = "98%"),
                                       p(style="text-align: center;", actionBttn("buttonvolcanoSDE", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  fluidRow(
                    column(4,
                           sliderInput("labelsizevolcanoSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    
                    column(4,
                           sliderInput("sizevolcanoSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                       min = 0, max = 5,
                                       value = 2, step = 0.1)),
                    
                    column(4,
                           sliderInput("opacityvolcanoSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                       min = 0, max = 1,
                                       value = 0.8, step = 0.1))
                  ),
                  
                  withSpinner(plotlyOutput("volcanoplotSDE", height = "600"))
              ),
              
              box(id = "SDEBox7", title = p("Gene Pathways", actionBttn("SDEBoxInfo4", "Info"), actionBttn("downloadPathwayPlotSDE", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE,
                  sidebar = boxSidebar(id = "GenePathwaysSDE", startOpen = TRUE, width = 50, background = "#e8ebe9",
                                       
                                       virtualSelectInput("GPInputSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize gene pathways:')), choices = NULL,
                                                          search = T,
                                                          placeholder = "Please type your pathway of interest",
                                                          width = "98%"),
                                       p(style="text-align: center;", actionBttn("GP_actionsSDE", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  
                  
                  fluidRow(
                    column(4,
                           sliderInput("labelsizeGPSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0, max = 50,
                                       value = 10, step = 1)),
                    column(4,
                           sliderInput("sizeGPSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Point Size')),
                                       min = 0, max = 5,
                                       value = 2, step = 0.1)),
                    
                    column(4,
                           sliderInput("opacityGPSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Opacity')),
                                       min = 0, max = 1,
                                       value = 0.8, step = 0.1))
                  ),
                  
                  
                  withSpinner(plotlyOutput("GP_plotSDE", height = "600"))
              )
            ),
            collapseInput(inputId = "iscollapseboxSDE", boxId = "SDEBoxIntro")
            
            
    ),
    
    ####TAB 14.5 - Pathway Analysis (Differential Expression Analysis) #####
    tabItem(tabName = "PA",
            fluidRow(
              box(id = "PABoxIntro", title = strong("Pathway Analysis"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Additional graphical outputs for gene set enrichment analysis. The outputs of this tab use the list of differentially expressed genes computed
                  in the "Differential Expression" tab.
                  <br><br>Visualisations of enriched terms in the form of:
                  <ol><li><b>Dot plot:</b> Enriched terms are plotted in order of gene ratio (number of enriched genes in gene pathway/total number of genes in gene pathway). </li>
                  <li><b>Gene Concept Network:</b> Visualisation of complex association and interactions between genes and enriched terms as a network. </li>
                  <li><b>Enrichment map: </b>Network of enriched terms with edges connecting overlapping gene pathways. Mutually overlapping gene pathways are clustered together allowing easy identification of related pathways. </li>
                  <li><b>View specific pathways: </b>View selected reactome pathway. </li></ol><br>
                          ')),
                  
                  img(id="DifferentialExpression3_ICARUS",src="DifferentialExpression_ICARUS.png",width="80%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startPA", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            collapseInput(inputId = "iscollapsebox7", boxId = "PABoxIntro"),
            
            fluidRow(
              box(id = "PABox1", title = p("Visualisation of Enriched Terms", actionBttn("PABoxInfo1", "Info"), actionBttn("downloaddotplotPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "DPside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("DotPlotNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    min = 0, max = 100,
                                                    value = 20, step = 1, width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(
                    column(4,
                           prettyRadioButtons("radioPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                              list("Dot Plot","Ridge Plot","Heatmap"), inline = TRUE, selected = "Dot Plot",
                                              icon = icon("check"),
                                              animation = "jelly")),
                    column(4,
                           sliderInput("fontPA_Dot", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 1, max = 40,
                                       value = 12, step = 1)),
                    column(4,
                           sliderInput("HeightPA_Dot", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("dotplotUI")))
            ),
            
            fluidRow(
              box(id = "PABox3", title = p("Gene Concept Network", actionBttn("PABoxInfo2", "Info"), actionBttn("downloadGCplotPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "UPside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("GCNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    min = 1, max = 100,
                                                    value = 20, step = 1, width = "98%"),
                                       materialSwitch("checkboxgene_GC", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Gene Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxcategory_GC", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Category Labels')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(3,
                           sliderInput("genefontPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Gene Label Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("categoryfontPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Category Label Size')),
                                       min = 0.1, max = 3,
                                       value = 1.5, step = 0.1)),
                    column(3,
                           sliderInput("nodePA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("HeightPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("GCplotUI")))
            ),
            
            
            fluidRow(
              box(id = "PABox2", title = p("Enrichment Map", actionBttn("PABoxInfo3", "Info"), actionBttn("downloademapplotPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "EMside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("EMNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    value = 20, min = 1, max = 100,
                                                    width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(3,
                           sliderInput("fontPA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("linePA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Line Size')),
                                       min = 0.1, max = 3,
                                       value = 0.5, step = 0.1)),
                    column(3,
                           sliderInput("nodePA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("HeightPA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("emapplotUI")))
            ),
            
            fluidRow(
              box(id = "PABox5", title = p("Tree Plot", actionBttn("PABoxInfo5", "Info"), actionBttn("downloadtreeplotPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "Treeside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("TreeNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    value = 20, min = 1, max = 100,
                                                    width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(4,
                           sliderInput("ClustersPA_T", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of Clusters to show')),
                                       min = 1, max = 20,
                                       value = 5, step = 1)),
                    column(4,
                           sliderInput("fontPA_T", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 1, max = 20,
                                       value = 5, step = 1)),
                    column(4,
                           sliderInput("HeightPA_T", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("treeplotUI")))
            ),
            
            fluidRow(
              box(id = "PABox4", title = p("View Specific Pathways", actionBttn("PABoxInfo4", "Info"), actionBttn("downloadpathwayPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12, 
                  sidebar = boxSidebar(id = "ViewSpecificPathways", startOpen = TRUE, width = 50, background = "#e8ebe9",
                                       virtualSelectInput("PAPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize Reactome pathway:')), choices = NULL, 
                                                          search = T,
                                                          placeholder = "Please type your pathway of interest",
                                                          width = "98%"),
                                       p(style="text-align: center;", actionBttn("PAPAInput", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(
                    column(12,
                           sliderInput("HeightPA_PP", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  withSpinner(htmlOutput("pathwayplotUI")))
            )
            
    ),
    
    ####TAB 14.75 - Drug Gene Database (Differential Expression Analysis) #####
    tabItem(tabName = "DRUG",
            fluidRow(
              box(id = "DRUGBoxIntro", title = strong("Drug Gene Interactions"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify">
                  The <a href="https://www.dgidb.org" target="_blank">Drug Gene Interaction Database (DGIdb)</a> curates information on druggable genes from publications, known affected pathways 
                  (Gene Ontology, Human Protein Atlas, IDG) and publicly available databases (DrugBank, PharmGKB, Chembl, Drug Target Commons and TTD).
                  ICARUS offers the user the option to query differentially expressed genes (refer to differential expression tab) against DGIdb v4.0 and returns a list of potential drug targets with 
                  information of interaction type (i.e., inhibitor, agonist, blocker), interaction claim source (i.e., PharmGKB, ChemblInteractions, etc), interaction group score 
                  (score takes into account number of drug and gene partners and number of supporting publications) and the relevant pubmed reference if available. <br><br>
                  Identification of drug-gene targets can facilitate targeted perturbations of key molecular pathways. 
                  In the context of dysregulation or disease, this could provide an avenue towards a therapeutic opportunity through repurposed drugs. <br><br>')),
                  
                  img(id="DGI_ICARUS",src="DGI_ICARUS.png",width="30%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW", actionBttn("startbuttonDRUG", "START", 
                                                             style = "jelly",
                                                             color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )
            ),
            
            collapseInput(inputId = "iscollapseboxDRUG", boxId = "DRUGBoxIntro"),
            
            fluidRow( 
              box(id = "DRUGBox1", title = p("Drug Gene Interactions", actionBttn("DRUGBoxInfo1", "Info"), downloadBttn("downloadmarkersDRUG", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  withSpinner(DT::dataTableOutput("markersDRUG", height = "500"))
              ),
              
              box(id = "DRUGBox2", title = p("Drug Gene Interactions Map", actionBttn("DRUGBoxInfo2", "Info"), actionBttn("downloadGeneDrugMap", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "DRUGCluster", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("DRUGNumber", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of interactions to show')),
                                                    value = 500, 
                                                    width = "98%"),
                                       materialSwitch("checkboxgene_DRUG", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Gene Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxcluster_DRUG", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Drug Labels')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(4,
                           sliderInput("genefontDRUG", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Gene Label Size')),
                                       min = 1, max = 20,
                                       value = 8, step = 1)),
                    column(4,
                           sliderInput("categoryfontDRUG", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Drug Label Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1)),
                    column(4,
                           sliderInput("nodeDRUG", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1))
                  ),
                  
                  withSpinner(plotOutput("DRUGMap", height = 1500))
              )
            )
    ),
    
    ####TAB 15.5 - Pathway Analysis (Custom Differential Expression Analysis) #####
    tabItem(tabName = "SPA",
            fluidRow(
              box(id = "SPABoxIntro", title = strong("Custom Pathway Analysis"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> Additional graphical outputs for gene set enrichment analysis. The outputs of this tab use the set of differentially expressed genes computed
                  in the "Custom Differential Expression" tab.
                  <br><br>Visualisations of enriched terms in the form of:
                  <ol><li><b>Dot plot:</b> Enriched terms are plotted in order of gene ratio (number of enriched genes in gene pathway/total number of genes in gene pathway). </li>
                  <li><b>Gene Concept Network:</b> Visualisation of complex association and interactions between genes and enriched terms as a network. </li>
                  <li><b>Enrichment map: </b>Network of enriched terms with edges connecting overlapping gene pathways. Mutually overlapping gene pathways are clustered together allowing easy identification of related pathways. </li>
                  <li><b>View specific pathways: </b>View selected reactome pathway. </li></ol><br>
                          ')),
                  
                  img(id="DifferentialExpression4_ICARUS",src="DifferentialExpression_ICARUS.png",width="80%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startSPA", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )),
            
            collapseInput(inputId = "iscollapseboxSPA", boxId = "SPABoxIntro"),
            
            fluidRow(
              box(id = "SPABox1", title = p("Visualisation of Enriched Terms", actionBttn("SPABoxInfo1", "Info"), actionBttn("downloaddotplotSPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "SDPside", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("DotPlotNumberSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    min = 0, max = 100,
                                                    value = 20, step = 1, width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  
                  fluidRow(
                    column(4,
                           prettyRadioButtons("radioSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Plot Options')),
                                              list("Dot Plot","Ridge Plot","Heatmap"), inline = TRUE, selected = "Dot Plot",
                                              icon = icon("check"),
                                              animation = "jelly")),
                    column(4,
                           sliderInput("fontSPA_Dot", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 1, max = 40,
                                       value = 12, step = 1)),
                    column(4,
                           sliderInput("HeightSPA_Dot", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("dotplotUISPA")))
            ),
            
            fluidRow(
              box(id = "SPABox3", title = p("Gene Concept Network", actionBttn("SPABoxInfo2", "Info"), actionBttn("downloadGCplotSPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "UPsideSPA", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("GCNumberSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    min = 1, max = 100,
                                                    value = 20, step = 1, width = "98%"),
                                       materialSwitch("checkboxgene_GCSPA", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;;">Show Gene Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxcategory_GCSPA", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Category Labels')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(3,
                           sliderInput("genefontSPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Gene Label Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("categoryfontSPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Category Label Size')),
                                       min = 0.1, max = 3,
                                       value = 1.5, step = 0.1)),
                    column(3,
                           sliderInput("nodeSPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("HeightSPA_GC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("GCplotUISPA")))
            ),
            
            fluidRow(
              box(id = "SPABox2", title = p("Enrichment Map", actionBttn("SPABoxInfo3", "Info"), actionBttn("downloademapplotSPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "EMsideSPA", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("EMNumberSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    value = 20, min = 1, max = 100, width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(3,
                           sliderInput("fontSPA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("lineSPA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Line Size')),
                                       min = 0.1, max = 3,
                                       value = 0.5, step = 0.1)),
                    column(3,
                           sliderInput("nodeSPA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 0.1, max = 3,
                                       value = 1, step = 0.1)),
                    column(3,
                           sliderInput("HeightSPA_EM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("emapplotUISPA")))
            ),
            
            fluidRow(
              box(id = "SPABox5", title = p("Tree Plot", actionBttn("SPABoxInfo5", "Info"), actionBttn("downloadtreeplotSPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "TreesideSPA", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("TreeNumberSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of categories to show')),
                                                    value = 20, min = 1, max = 100,
                                                    width = "98%"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(4,
                           sliderInput("ClustersSPA_T", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of Clusters to show')),
                                       min = 1, max = 20,
                                       value = 5, step = 1)),
                    column(4,
                           sliderInput("fontSPA_T", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Label Size')),
                                       min = 1, max = 20,
                                       value = 5, step = 1)),
                    column(4,
                           sliderInput("HeightSPA_T", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  
                  withSpinner(htmlOutput("treeplotSPAUI")))
            ),
            
            fluidRow(
              box(id = "SPABox4", title = p("View Specific Pathways", actionBttn("SPABoxInfo4", "Info"), actionBttn("downloadpathwaySPA", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12, 
                  sidebar = boxSidebar(id = "ViewSpecificPathwaysSPA", startOpen = TRUE, width = 50, background = "#e8ebe9",
                                       virtualSelectInput("PAPASPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Visualize Reactome pathway:')), choices = NULL, 
                                                          search = T,
                                                          placeholder = "Please type your pathway of interest",
                                                          width = "98%"),
                                       p(style="text-align: center;", actionBttn("PAPAInputSPA", "APPLY", icon("play"),
                                                                                 style = "jelly",
                                                                                 color = "success")),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")
                  ),
                  fluidRow(
                    column(12,
                           sliderInput("HeightSPA_PP", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Plot Height')),
                                       min = 1, max = 100,
                                       value = 20, step = 1))
                  ),
                  withSpinner(htmlOutput("pathwayplotSPAUI")))
            )
            
    ),
    
    ####TAB 15.75 - Drug Gene Database (Custom Differential Expression Analysis)#####
    tabItem(tabName = "DRUG2",
            fluidRow(
              box(id = "DRUGBoxIntro2", title = strong("Custom Drug Gene Interactions"), status = "warning", solidHeader = F, 
                  collapsible = T, collapsed = T, width = 12,
                  
                  h4(HTML('<h4 style = "text-align:justify"> 
                  The <a href="https://www.dgidb.org" target="_blank">Drug Gene Interaction Database (DGIdb)</a> curates information on druggable genes from publications, known affected pathways 
                  (Gene Ontology, Human Protein Atlas, IDG) and publicly available databases (DrugBank, PharmGKB, Chembl, Drug Target Commons and TTD).
                  ICARUS offers the user the option to query differentially expressed genes (refer to custom differential expression tab) against DGIdb v4.0 and returns a list of potential drug targets with 
                  information of interaction type (i.e., inhibitor, agonist, blocker), interaction claim source (i.e., PharmGKB, ChemblInteractions, etc), interaction group score 
                  (score takes into account number of drug and gene partners and number of supporting publications) and the relevant pubmed reference if available. <br><br>
                  Identification of drug-gene targets can facilitate targeted perturbations of key molecular pathways. 
                  In the context of dysregulation or disease, this could provide an avenue towards a therapeutic opportunity through repurposed drugs. <br><br>')),
                  
                  img(id="DGI_ICARUS2",src="DGI_ICARUS.png",width="30%"), br(), br(),
                  
                  fluidRow(column(3,
                                  valueBox("NEW",
                                           actionBttn("startbuttonDRUG2", "START", 
                                                      style = "jelly",
                                                      color = "primary"), icon = icon("paper-plane"), width = 12, color = "aqua")))
              )
            ),
            
            collapseInput(inputId = "iscollapseboxDRUG2", boxId = "DRUGBoxIntro2"),
            
            fluidRow( 
              box(id = "DRUGBox12", title = p("Drug Gene Interactions", actionBttn("DRUGBoxInfo12", "Info"), downloadBttn("downloadmarkersDRUG2", "Download .csv")), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  
                  withSpinner(DT::dataTableOutput("markersDRUG2", height = "500"))
              ),
              
              box(id = "DRUGBox22", title = p("Drug Gene Interactions Map", actionBttn("DRUGBoxInfo22", "Info"), actionBttn("downloadGeneDrugMap2", "Download Plot", icon = icon("download"))), status = "warning", solidHeader = F,
                  collapsible = TRUE, width = 12,
                  sidebar = boxSidebar(id = "DRUGCluster2", startOpen = TRUE, width = 25, background = "#e8ebe9",
                                       numericInput("DRUGNumber2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of interactions to show')),
                                                    value = 500, 
                                                    width = "98%"),
                                       materialSwitch("checkboxgene_DRUG2", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Gene Labels')), value = TRUE, status = "primary"),
                                       materialSwitch("checkboxcluster_DRUG2", label = h4(HTML('<h4 style = "text-align:justify;color:#000000;">Show Drug Labels')), value = TRUE, status = "primary"),
                                       p(style = "color: #FF0000;", "*This sidebar is minimizable by clicking the cog icon on the top right of the box")),
                  
                  fluidRow(
                    column(4,
                           sliderInput("genefontDRUG2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Gene Label Size')),
                                       min = 1, max = 20,
                                       value = 8, step = 1)),
                    column(4,
                           sliderInput("categoryfontDRUG2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Drug Label Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1)),
                    column(4,
                           sliderInput("nodeDRUG2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Node Size')),
                                       min = 1, max = 20,
                                       value = 4, step = 1))
                  ),
                  
                  withSpinner(plotOutput("DRUGMap2", height = 1500)))
            )
    )
    
  )
)



# Put them together into a dashboardPage
ui <- dashboardPage(
  dashboardHeader(
    
    
    title = img(id="ICARUS",src="ICARUS.png",width="70%"),
    dropdownMenu(type = "notifications",
                 notificationItem(
                   text = p("ICARUS has been updated to v2.4",
                            actionBttn("updates", "See what's new in ICARUS v2.4", size = "sm")),
                   icon = icon("gear")
                 )
    ),
    tags$li(class = "dropdown", p(style="padding-left: 20px; padding-top: 4px; margin-bottom:-10px", 
                                  actionBttn("downloadDataSaveButton", "Save", icon = icon("save"),
                                             size = "lg"))),
    titleWidth = 250),
  sidebar,
  body,
  md = TRUE,
  skin = "black",
  title = "ICARUS",
  footer = dashboardFooter(right = h6("Created by: Andrew Jiang", br(), "Github: github.com/Enjewl/ICARUS", br(), "Applied Translational Genetics Group", br(), "Centre for Brain Research", br(), "The University of Auckland"),
                           left = p(img(id="ATGlogo", src="ATGlogo.png", width="10%", style = "margin-right: 30px"), img(id="CBR_logo", src="CBR_logo.png", width="20%", style = "margin-right: 30px"), img(id="Nectar_logo", src="Nectar_logo.png", width="15%")))
)



server <- function(input, output, session) { 
  
  ####Menu manipulation####
  #Timeout
  observeEvent(input$timeOut, { 
    shinyalert("SESSION TIMEOUT", paste("Session timeout due to", input$timeOut, "inactivity -", Sys.time()), type = "error", showConfirmButton = F)
    session$close()
  })
  
  observeEvent(input$startbuttonintro, {
    withProgress(message = "Getting everything ready, won't take long...", 
                 value = 0, {
                   incProgress(0.1)
                   
                   incProgress(0.2)
                   library(SeuratWrappers)
                   library(MAST)
                   library(DESeq2)
                   library(BiocParallel)
                   library(harmony)
                   library(Seurat)
                   library(uwot)
                   library(glmGamPoi)
                   library(sctransform)
                   library(svglite)
                   library(ggrepel)
                   library(tidyverse)
                   library(data.table)
                   library(tools)   
                   library(clusterProfiler)
                   library(ReactomePA)
                   library(msigdbr)
                   library(enrichplot)
                   library(graphite)
                   library(limma)
                   library(msigdbr)
                   library(org.Hs.eg.db)
                   library(org.Bt.eg.db)
                   library(org.Ce.eg.db)
                   library(org.Cf.eg.db)
                   library(org.Dr.eg.db)
                   library(org.Dm.eg.db)
                   library(org.Gg.eg.db)
                   library(org.Mm.eg.db)
                   library(org.Rn.eg.db)
                   library(org.Sc.sgd.db)
                   library(org.Ss.eg.db)
                   
                   updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                   
                   incProgress(0.3)
                   
                   seuratreactive$msig <- NULL
                   seuratreactive$orgDb <- NULL
                   seuratreactive$EKEGG <- NULL
                   seuratreactive$REACT <- NULL
                   seuratreactive$PAREACTOME <- NULL
                   
                   #Load
                   seuratreactive$obj_original <- NULL
                   seuratreactive$matrixtable <- NULL
                   seuratreactive$mincells <- NULL
                   seuratreactive$minfeatures <- NULL
                   
                   #Doublet
                   seuratreactive$plot.data.doublets <- NULL
                   seuratreactive$Remove <- NULL
                   seuratreactive$Removed <- NULL
                   seuratreactive$doublesim <- NULL
                   
                   #Load2
                   seuratreactive$obj_original2 <- NULL
                   seuratreactive$matrixtable2 <- NULL
                   seuratreactive$mincellsCombine <- NULL
                   seuratreactive$minfeaturesCombine <- NULL
                   
                   #QC
                   seuratreactive$objQC <- NULL
                   seuratreactive$type <- NULL
                   seuratreactive$nFeature1 <- NULL
                   seuratreactive$nFeature2 <- NULL
                   seuratreactive$nCount1 <- NULL
                   seuratreactive$nCount2 <- NULL
                   seuratreactive$percent.mt1 <- NULL
                   seuratreactive$percent.mt2 <- NULL
                   seuratreactive$percent.ribo1 <- NULL
                   seuratreactive$percent.ribo2 <- NULL
                   
                   #QC2
                   seuratreactive$type2 <- NULL
                   seuratreactive$objQC2 <- NULL
                   seuratreactive$nFeature12 <- NULL
                   seuratreactive$nFeature22 <- NULL
                   seuratreactive$nCount12 <- NULL
                   seuratreactive$nCount22 <- NULL
                   seuratreactive$percent.mt12 <- NULL
                   seuratreactive$percent.mt22 <- NULL
                   seuratreactive$percent.ribo12 <- NULL
                   seuratreactive$percent.ribo22 <- NULL
                   
                   #D2
                   seuratreactive$Remove2 <- NULL
                   seuratreactive$plot.data.doublets2 <- NULL
                   seuratreactive$Removed2 <- NULL
                   seuratreactive$doublesim2 <- NULL
                   
                   #DR
                   seuratreactive$obj <- NULL
                   seuratreactive$variablefeatures <- NULL
                   seuratreactive$SCTransformDR <- NULL
                   seuratreactive$ALRAL <- NULL
                   seuratreactive$npcDR <- NULL
                   
                   #DR2
                   seuratreactive$obj <- NULL
                   seuratreactive$SCTransformDR2 <- NULL
                   seuratreactive$npcDR2 <- NULL
                   seuratreactive$Combined.list <- NULL
                   seuratreactive$VFintegrated <- NULL
                   seuratreactive$integrationreduction <- NULL
                   seuratreactive$kanchor <- NULL
                   
                   #C
                   seuratreactive$integera <- NULL
                   seuratreactive$integerb <- NULL
                   seuratreactive$integerc <- NULL
                   seuratreactive$integerd <- NULL
                   seuratreactive$integere <- NULL
                   seuratreactive$integerf <- NULL
                   seuratreactive$integerg <- NULL
                   seuratreactive$select1 <- NULL
                   seuratreactive$tsne1 <- NULL
                   seuratreactive$tsne2 <- NULL
                   seuratreactive$tsne3 <- NULL
                   seuratreactive$plot.data <- NULL
                   seuratreactive$plot.data.original <- NULL
                   seuratreactive$Identifiers <- NULL
                   
                   #CC
                   seuratreactive$reductionCC1 <- NULL
                   seuratreactive$reductionCC2 <- NULL
                   seuratreactive$reductionCC3 <- NULL
                   seuratreactive$GeneR <- NULL
                   
                   #L
                   seuratreactive$reference_annotation <- NULL
                   seuratreactive$select2 <- NULL
                   seuratreactive$selecttissue <- NULL
                   seuratreactive$new.cell.ids <- NULL
                   seuratreactive$markers <- NULL
                   
                   #MEGENA
                   seuratreactive$MEGENA_samples <- NULL
                   seuratreactive$variablegenesMEGENA <- NULL
                   seuratreactive$MEGENA_selectmethod <- NULL
                   seuratreactive$MEGENAmin.size <- NULL
                   seuratreactive$MEGENApval <- NULL
                   seuratreactive$MEGENA_selectcluster <- NULL
                   seuratreactive$n_MEGENA <- NULL
                   seuratreactive$log2FCMEGENA <- NULL
                   seuratreactive$padjustMEGENA <- NULL
                   seuratreactive$testuseMEGENA <- NULL
                   seuratreactive$minMEGENA <- NULL
                   seuratreactive$pvalMEGENA <- NULL
                   seuratreactive$summary.output <- NULL
                   seuratreactive$markersMEGENA <- NULL
                   seuratreactive$moduleGO_df <- NULL
                   seuratreactive$VF <- NULL
                   seuratreactive$g <- NULL
                   
                   #SCENIC
                   seuratreactive$variablefeaturesSCENIC <- NULL
                   seuratreactive$variablegenesSCENIC <- NULL
                   seuratreactive$regulonAUC <- NULL
                   seuratreactive$regulons <- NULL
                   seuratreactive$cellInfo <- NULL
                   seuratreactive$n_SCENIC <- NULL
                   seuratreactive$SCENICdb <- NULL
                   seuratreactive$SCENIC_samples <- NULL
                   seuratreactive$regulonnames <- NULL
                   seuratreactive$titleSCENIC <- NULL
                   seuratreactive$SCENICInput <- NULL
                   seuratreactive$SCENIC_selectcluster <- NULL
                   seuratreactive$SCENIC_selectmethod <- NULL
                   seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                   seuratreactive$rss <- NULL
                   
                   #MAGMA
                   seuratreactive$MAGMA_samples <- NULL
                   seuratreactive$MAGMA_GWAS <- NULL
                   seuratreactive$MAGMA_Final_Results <- NULL
                   seuratreactive$meta <- NULL
                   seuratreactive$MAGMA_selectcluster <- NULL
                   seuratreactive$MAGMA_samples2 <- NULL
                   seuratreactive$MAGMA_selectcluster2 <- NULL
                   seuratreactive$textlabelMAGMA <- NULL
                   seuratreactive$MAGMA_N <- NULL
                   seuratreactive$GRCh <- NULL
                   seuratreactive$MAGMA_Pop <- NULL
                   seuratreactive$MAGMA_dirs <- NULL
                   seuratreactive$genesOutPath2 <- NULL
                   seuratreactive$n_MAGMA <- NULL
                   seuratreactive$n_MAGMA2 <- NULL
                   
                   #CellChat
                   seuratreactive$cellchat <- NULL
                   seuratreactive$cellchatM <- NULL
                   seuratreactive$object.list <- NULL
                   seuratreactive$ChatChoose <- NULL
                   seuratreactive$ChatSample <- NULL
                   seuratreactive$Chatdb <- NULL
                   seuratreactive$ChatMultiple1Sample <- NULL
                   seuratreactive$ChatMultiple2Sample <- NULL
                   seuratreactive$Chat_selectcluster <- NULL
                   seuratreactive$Chat_number <- NULL
                   seuratreactive$Chatpval <- NULL
                   
                   #DRUG
                   seuratreactive$markersDRUG <- NULL
                   seuratreactive$DRUGdf <- NULL
                   
                   #DRUG2
                   seuratreactive$markersDRUG2 <- NULL
                   seuratreactive$DRUGdf2 <- NULL
                   
                   #GE
                   seuratreactive$titleGE <- NULL
                   seuratreactive$GEInput <- NULL
                   seuratreactive$GSEInput <- NULL
                   seuratreactive$p <- NULL
                   seuratreactive$q <- NULL
                   
                   seuratreactive$objGEyes <- NULL
                   seuratreactive$Gene_vector_msig <- NULL
                   
                   #TA
                   seuratreactive$TAselectedcells <- NULL
                   seuratreactive$TA_pickercluster <- NULL
                   seuratreactive$seurat_monocle <- NULL
                   seuratreactive$Monocle3_genes <- NULL
                   
                   #MM
                   seuratreactive$MM <- NULL
                   seuratreactive$plot.dataM <- NULL
                   seuratreactive$plot.dataMM <- NULL
                   
                   #DE
                   seuratreactive$onlyposDE <- NULL
                   seuratreactive$Subset_GENES <- NULL
                   seuratreactive$DE_clusters_samples <- NULL
                   seuratreactive$DECluster <- NULL
                   seuratreactive$DEInput1_samples <- NULL
                   seuratreactive$DEInput2_samples <- NULL
                   seuratreactive$SelectComparison <- NULL
                   seuratreactive$DE_clusters <- NULL
                   seuratreactive$GOlabelsDE <- NULL
                   seuratreactive$DEInput1 <- NULL
                   seuratreactive$DEInput2 <- NULL
                   seuratreactive$logDE <- NULL
                   seuratreactive$minDE <- NULL
                   seuratreactive$padjustDE1 <- NULL
                   seuratreactive$testuseDE <- NULL
                   seuratreactive$enrichInputDE <- NULL
                   seuratreactive$pvalueDE <- NULL
                   seuratreactive$pvalDE <- NULL
                   seuratreactive$padjustDE <- NULL
                   seuratreactive$volcanovalue1 <- NULL
                   seuratreactive$volcanovalue2 <- NULL
                   seuratreactive$pvaluevolcano <- NULL
                   seuratreactive$markersDE <- NULL
                   seuratreactive$YuDE <- NULL
                   seuratreactive$markersDEvolcano <- NULL
                   seuratreactive$PATHWAY <- NULL
                   
                   #SDE
                   seuratreactive$onlyposSDE <- NULL
                   seuratreactive$Subset_GENESSDE <- NULL
                   seuratreactive$selectedkey1 <- NULL
                   seuratreactive$selectedkey2 <- NULL
                   seuratreactive$GOlabelsSDE <- NULL
                   seuratreactive$logSDE <- NULL
                   seuratreactive$minSDE <- NULL
                   seuratreactive$padjustSDE1 <- NULL
                   seuratreactive$testuseSDE <- NULL
                   seuratreactive$enrichInputSDE <- NULL
                   seuratreactive$pvalueSDE <- NULL
                   seuratreactive$pvalSDE <- NULL
                   seuratreactive$padjustSDE <- NULL
                   seuratreactive$volcanoSDEvalue1 <- NULL
                   seuratreactive$volcanoSDEvalue2 <- NULL
                   seuratreactive$pvaluevolcanoSDE <- NULL
                   seuratreactive$PATHWAYSDE <- NULL
                   
                   seuratreactive$markersSDE <- NULL
                   seuratreactive$YuSDE <- NULL
                   seuratreactive$markersSDEvolcano <- NULL
                   
                   #PA
                   seuratreactive$PAchoose <- NULL
                   seuratreactive$padjustPA <- NULL
                   seuratreactive$pvaluePA <- NULL
                   seuratreactive$GOlabelsPA <- NULL
                   Yu$result <- NULL
                   Yu$ema <- NULL
                   seuratreactive$geneListPA <- NULL
                   
                   #SPA
                   seuratreactive$SPAchoose <- NULL
                   seuratreactive$padjustSPA <- NULL
                   seuratreactive$pvalueSPA <- NULL
                   seuratreactive$GOlabelsSPA <- NULL
                   YuSPA$result <- NULL
                   YuSPA$ema <- NULL
                   seuratreactive$geneListSPA <- NULL
                   
                   shinyjs::hide("NextButtonLoad")
                   shinyjs::hide("NextButtonQC")
                   shinyjs::hide("NextButtonD")
                   shinyjs::hide("NextButtonCombine")
                   shinyjs::hide("NextButtonQC2")
                   shinyjs::hide("NextButtonD2")
                   shinyjs::hide("NextButtonDR")
                   shinyjs::hide("NextButtonDR2")
                   shinyjs::hide("NextButtonC")
                   shinyjs::hide("NextButtonCC")
                   shinyjs::hide("NextButtonDE")
                   shinyjs::hide("NextButtonSDE")
                   
                   #Load
                   if (input$iscollapseboxintro == TRUE) {
                     js$collapse("LoadBoxIntro")
                   }
                   shinyjs::hide("LoadBox1")
                   shinyjs::hide("LoadBox2")
                   shinyjs::hide("LoadBox3")
                   shinyjs::hide("LoadBox5")
                   shinyjs::hide("LoadBox4")
                   
                   
                   #QC
                   if(input$iscollapseboxQC == TRUE) {
                     js$collapse("QCBoxIntro")
                   }
                   shinyjs::hide("QCBox1")
                   shinyjs::hide("QCBox2")
                   shinyjs::hide("QCBox3")
                   shinyjs::hide("QCBox4")
                   shinyjs::enable("startbuttonQC")
                   
                   #D
                   if(input$iscollapseboxD == TRUE) {
                     js$collapse("DBoxIntro")
                   }
                   shinyjs::hide("DBox1")
                   shinyjs::hide("DBox2")
                   
                   
                   #Load2
                   if(input$iscollapseboxCombine == TRUE) {
                     js$collapse("CombineBoxIntro")
                   }
                   shinyjs::hide("CombineBox1")
                   shinyjs::hide("CombineBox2")
                   shinyjs::hide("CombineBox3")
                   shinyjs::hide("CombineBox4")
                   shinyjs::hide("CombineBox5")
                   
                   
                   #QC2
                   if(input$iscollapseboxQC2 == TRUE) {
                     js$collapse("QCBoxIntro2")
                   }
                   shinyjs::hide("QCBox12")
                   shinyjs::hide("QCBox22")
                   shinyjs::hide("QCBox32")
                   shinyjs::hide("QCBox42")
                   shinyjs::enable("startbuttonQC2")
                   
                   #D2
                   if(input$iscollapseboxD2 == TRUE) {
                     js$collapse("DBoxIntro2")
                   }
                   shinyjs::hide("DBox12")
                   shinyjs::hide("DBox22")
                   
                   
                   #DR
                   if(input$iscollapsebox1 == TRUE) {
                     js$collapse("DRBoxIntro")
                   }
                   shinyjs::hide("DRBox1")
                   shinyjs::hide("DRBox2")
                   shinyjs::hide("DRBox3")
                   shinyjs::hide("DRBox4")
                   shinyjs::enable("startbuttonDR")
                   
                   #DR2
                   if(input$iscollapseboxDR2 == TRUE) {
                     js$collapse("DRBoxIntro2")
                   }
                   shinyjs::hide("DRBox1DR2")
                   shinyjs::hide("DRBox2DR2")
                   shinyjs::hide("DRBox3DR2")
                   shinyjs::hide("DRBox4DR2")
                   shinyjs::hide("downloadMerged")
                   shinyjs::enable("startbuttonDR2")
                   
                   #C
                   if (input$iscollapsebox2 == TRUE) {
                     js$collapse("CBoxIntro")
                   }
                   shinyjs::hide("CBox1")
                   shinyjs::hide("CBox2")
                   shinyjs::hide("CBox3")
                   shinyjs::enable("startbuttonC")
                   
                   #CC
                   if (input$iscollapsebox3 == TRUE) {
                     js$collapse("CCBoxIntro")
                   }
                   shinyjs::hide("CCBox1")
                   shinyjs::hide("CCBox1a")
                   shinyjs::hide("CCBox1b")
                   shinyjs::hide("CCBox1c")
                   shinyjs::hide("CCBox2")
                   shinyjs::enable("startbuttonCC")
                   
                   #L
                   if (input$iscollapsebox4 == TRUE) {
                     js$collapse("LBoxIntro")
                   }
                   shinyjs::hide("LBox1")
                   shinyjs::hide("LBox2")
                   shinyjs::hide("LBox3")
                   shinyjs::hide("LBox4")
                   shinyjs::hide("LBox5")
                   shinyjs::hide("LBox6")
                   shinyjs::hide("LBox7")
                   shinyjs::hide("LBox8")
                   
                   #MEGENA
                   if (input$iscollapseboxMEGENA == TRUE) {
                     js$collapse("MEGENABoxIntro")
                   }
                   shinyjs::hide("MEGENABox1")
                   shinyjs::hide("MEGENABox2")
                   shinyjs::hide("MEGENABox3")
                   shinyjs::hide("MEGENABox4")
                   
                   #SCENIC
                   if (input$iscollapseboxSCENIC == TRUE) {
                     js$collapse("SCENICBoxIntro")
                   }
                   shinyjs::hide("SCENICBox1a")
                   shinyjs::hide("SCENICBox1b")
                   shinyjs::hide("SCENICBox1")
                   shinyjs::hide("SCENICBox2")
                   shinyjs::hide("SCENICBox3")
                   
                   #MAGMA
                   if (input$iscollapseboxMAGMA == TRUE) {
                     js$collapse("MAGMABoxIntro")
                   }
                   shinyjs::hide("MAGMABox1")
                   shinyjs::hide("MAGMABox1a")
                   shinyjs::hide("MAGMABox1b")
                   shinyjs::hide("MAGMABox2a")
                   shinyjs::hide("MAGMABox2b")
                   shinyjs::hide("MAGMABox3")
                   shinyjs::hide("MAGMABox4")
                   
                   #CellChat
                   if (input$iscollapseboxChat == TRUE) {
                     js$collapse("ChatBoxIntro")
                   }
                   shinyjs::hide("ChatBox1")
                   shinyjs::hide("ChatBox2")
                   shinyjs::hide("ChatBox3")
                   shinyjs::hide("ChatBox4")
                   shinyjs::hide("ChatBox5")
                   shinyjs::hide("ChatBox6")
                   shinyjs::hide("ChatBox7")
                   shinyjs::hide("ChatBox8")
                   shinyjs::hide("ChatBox9")
                   shinyjs::hide("ChatBox10")
                   shinyjs::hide("ChatBox11")
                   shinyjs::hide("ChatBox12")
                   shinyjs::hide("ChatBox13")
                   
                   #DRUG
                   if (input$iscollapseboxDRUG == TRUE) {
                     js$collapse("DRUGBoxIntro")
                   }
                   shinyjs::hide("DRUGBox1")
                   shinyjs::hide("DRUGBox2")
                   
                   #DRUG2
                   if (input$iscollapseboxDRUG2 == TRUE) {
                     js$collapse("DRUGBoxIntro2")
                   }
                   shinyjs::hide("DRUGBox12")
                   shinyjs::hide("DRUGBox22")
                   
                   #GE
                   if (input$iscollapsebox5 == TRUE) {
                     js$collapse("GEBoxIntro")
                   }
                   shinyjs::hide("GEBox1")
                   shinyjs::hide("GEBox2")
                   shinyjs::hide("GEBox3")
                   shinyjs::hide("GEBox4")
                   
                   #TA
                   if (input$iscollapseboxTA == TRUE) {
                     js$collapse("TABoxIntro")
                   }
                   shinyjs::hide("TABox1")
                   shinyjs::hide("TABox2")
                   shinyjs::hide("TABox3")
                   shinyjs::hide("TABox4")
                   shinyjs::hide("TABox5")
                   shinyjs::hide("TABox6")
                   shinyjs::hide("TABox7")
                   shinyjs::hide("TABox8")
                   shinyjs::hide("TABox9")
                   shinyjs::hide("TABox10")
                   
                   #MM
                   if (input$iscollapseboxMM == TRUE) {
                     js$collapse("MMBoxIntro")
                   }
                   if (input$iscollapseboxMM2 == TRUE) {
                     js$collapse("MMBox4")
                   }
                   
                   if (input$iscollapseboxMM3 == TRUE) {
                     js$collapse("MMBox5")
                   }
                   shinyjs::hide("MMBox1")
                   shinyjs::hide("MMBox2")
                   shinyjs::hide("MMBox3")
                   shinyjs::hide("MMBox4")
                   shinyjs::hide("MMBox5")
                   
                   #DE
                   if (input$iscollapsebox6 == TRUE) {
                     js$collapse("DEBoxIntro")
                   }
                   shinyjs::hide("DEBox1")
                   shinyjs::hide("DEBox1a")
                   shinyjs::hide("DEBox1b")
                   shinyjs::hide("DEBox2")
                   shinyjs::hide("DEBox3")
                   shinyjs::hide("DEBox4")
                   shinyjs::hide("DEBox5")
                   shinyjs::hide("DEBox6")
                   shinyjs::enable("startbuttonDE")
                   
                   #SDE
                   if (input$iscollapseboxSDE == TRUE) {
                     js$collapse("SDEBoxIntro")
                   }
                   shinyjs::hide("SDEBox1")
                   shinyjs::hide("SDEBox2")
                   shinyjs::hide("SDEBox3")
                   shinyjs::hide("SDEBox4")
                   shinyjs::hide("SDEBox5")
                   shinyjs::hide("SDEBox6")
                   shinyjs::hide("SDEBox7")
                   shinyjs::enable("startbuttonSDE")
                   
                   #PA  
                   if (input$iscollapsebox7 == TRUE) {
                     js$collapse("PABoxIntro")
                   }
                   shinyjs::hide("PABox1")
                   shinyjs::hide("PABox2")
                   shinyjs::hide("PABox3")
                   shinyjs::hide("PABox4")
                   shinyjs::hide("PABox5")
                   
                   #SPA
                   if (input$iscollapseboxSPA == TRUE) {
                     js$collapse("SPABoxIntro")
                   }
                   shinyjs::hide("SPABox1")
                   shinyjs::hide("SPABox2")
                   shinyjs::hide("SPABox3")
                   shinyjs::hide("SPABox4")
                   shinyjs::hide("SPABox5")
                   
                   incProgress(0.5)
                   
                   if (input$species1 == "Homo sapiens"){
                     seuratreactive$msig <- readRDS("Human.msig")
                     seuratreactive$orgDb <- "org.Hs.eg.db"
                     seuratreactive$EKEGG <- "hsa"
                     seuratreactive$REACT <- "human"
                     seuratreactive$PAREACTOME <- "hsapiens"
                   }
                   else if (input$species1 == "Bos taurus") {
                     seuratreactive$msig <- readRDS("Cattle.msig")
                     seuratreactive$orgDb <- "org.Bt.eg.db"
                     seuratreactive$EKEGG <- "bta"
                     seuratreactive$REACT <- "bovine"
                     seuratreactive$PAREACTOME <- "btaurus"
                   }
                   else if (input$species1 == "Caenorhabditis elegans") {
                     seuratreactive$msig <- readRDS("Celegans.msig")
                     seuratreactive$orgDb <- "org.Ce.eg.db"
                     seuratreactive$EKEGG <- "tst"
                     seuratreactive$REACT <- "celegans"
                     seuratreactive$PAREACTOME <- "celegans"
                   }
                   else if (input$species1 == "Canis lupus familiaris") {
                     seuratreactive$msig <- readRDS("Dog.msig")
                     seuratreactive$orgDb <- "org.Cf.eg.db"
                     seuratreactive$EKEGG <- "cfa"
                     seuratreactive$REACT <- "canine"
                     seuratreactive$PAREACTOME <- "cfamiliaris"
                   }
                   else if (input$species1 == "Danio rerio") {
                     seuratreactive$msig <- readRDS("Zebrafish.msig")
                     seuratreactive$orgDb <- "org.Dr.eg.db"
                     seuratreactive$EKEGG <- "dre"
                     seuratreactive$REACT <- "zebrafish"
                     seuratreactive$PAREACTOME <- "celegans"
                   }
                   else if (input$species1 == "Drosophila melanogaster") {
                     seuratreactive$msig <- readRDS("Fly.msig")
                     seuratreactive$orgDb <- "org.Dm.eg.db"
                     seuratreactive$EKEGG <- "dme"
                     seuratreactive$REACT <- "fly"
                     seuratreactive$PAREACTOME <- "dmelanogaster"
                   }
                   else if (input$species1 == "Gallus gallus") {
                     seuratreactive$msig <- readRDS("Chicken.msig")
                     seuratreactive$orgDb <- "org.Gg.eg.db"
                     seuratreactive$EKEGG <- "gga"
                     seuratreactive$REACT <- "chicken"
                     seuratreactive$PAREACTOME <- "ggallus"
                   }
                   else if (input$species1 == "Mus musculus") {
                     seuratreactive$msig <- readRDS("Mouse.msig")
                     seuratreactive$orgDb <- "org.Mm.eg.db"
                     seuratreactive$EKEGG <- "mmu"
                     seuratreactive$REACT <- "mouse"
                     seuratreactive$PAREACTOME <- "mmusculus"
                   }
                   else if (input$species1 == "Rattus norvegicus") {
                     seuratreactive$msig <- readRDS("Rat.msig")
                     seuratreactive$orgDb <- "org.Rn.eg.db"
                     seuratreactive$EKEGG <- "rno"
                     seuratreactive$REACT <- "rat"
                     seuratreactive$PAREACTOME <- "rnorvegicus"
                   }
                   else if (input$species1 == "Saccharomyces cerevisiae") {
                     seuratreactive$msig <- readRDS("Yeast.msig")
                     seuratreactive$orgDb <- "org.Sc.sgd.db"
                     seuratreactive$EKEGG <- "sce"
                     seuratreactive$REACT <- "yeast"
                     seuratreactive$PAREACTOME <- "scerevisiae"
                   }
                   else if (input$species1 == "Sus scrofa") {
                     seuratreactive$msig <- readRDS("Pig.msig")
                     seuratreactive$orgDb <- "org.Ss.eg.db"
                     seuratreactive$EKEGG <- "ssc"
                     seuratreactive$REACT <- "pig"
                     seuratreactive$PAREACTOME <- "sscrofa"
                   }
                   
                   incProgress(0.8)
                   
                   shinyjs::hide("NextButtonLoad")
                   
                   
                   if(input$iscollapseboxintro == TRUE) {
                     js$collapse("LoadBoxIntro")
                   }
                   shinyjs::show("startbuttonLoad")
                   shinyjs::hide("LoadBox1")
                   shinyjs::hide("LoadBox2")
                   shinyjs::hide("LoadBox3")
                   shinyjs::hide("LoadBox4")
                   shinyjs::hide("LoadBox5")
                   
                   output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                    menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T,
                                                                             badgeColor = "green", badgeLabel = "new")
                   ))
                 })
  })
  
  
  ##INFO BUTTONS####
  ####Updates######
  observeEvent(input$updates, {
    showModal(modalDialog(
      title = strong("ICARUS CHANGE LOG"),
      HTML('<h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.4</span> - 1st January, 2023</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW An option to impute dropouts using <a href="https://www.nature.com/articles/s41467-021-27729-z" target="_blank">Adaptively-thresholded low rank approximation (ALRA) method.</a></li>
             </ul>
           </ul>
      
      
      <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.3</span> - 18th October, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Improved search functionality for genes and gene pathways.</li>
                          <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW The height of ICARUS windows are now adjustable for many of the plots.</li>
                                       <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW A document describing the objects in the ICARUS enviroment is now <a href="https://github.com/Enjewl/ICARUS/blob/master/ICARUS_enviroment_descriptions.md" target="_blank">available.</a> </li>
             </ul>
           </ul>
      
      
      <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.2</span> - 20th August, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Analyze gene Co-expression with MEGENA, read more about it <a href="https://doi.org/10.1371/journal.pcbi.1004574" target="_blank">here</a>.</li>
                          <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Analyze gene regulatory networks from transcription factors and target genes with SCENIC, read more about it <a href="https://doi.org/10.1038/nmeth.4463" target="_blank">here</a>.</li>
                                       <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Examine cell cluster expression association with various Genome Wide Association Studies (GWAS) traits using MAGMA, read more about it <a href="https://doi.org/10.1371/journal.pcbi.1004219" target="_blank">MAGMA</a>.</li>
                                                    <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Visualize and analyze cell-cell communication networks (ligand-receptor interactions between cells) with CellChat, read more about it <a href="https://doi.org/10.1038/s41467-021-21246-9" target="_blank">here</a>.</li>
                                                                                                                                                            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Examine drug-gene interactions of differentially expressed genes against the <a href="https://www.dgidb.org" target="_blank">DGIdb 4.0 database</a>!</li>
             </ul>
           </ul>
      
      <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.1</span> - 10th May, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>ICARUS is now published in <a href="https://doi.org/10.1093/nar/gkac322" target="_blank">Nucleic Acids Research!</a></li>
             </ul>
           </ul>
           
           <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.0 </span> - 21st April, 2022</h4>
            <ul class="list-unstyled mt-3"> 
            <ul>
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Perform trajectory analysis with <b>Monocle3</b>, read more about it <a href="https://doi.org/10.1038/nbt.2859" target="_blank">here</a>.</li>
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Detection and removal of doublets in data with <b>DoubletFinder</b>, read more about it <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">here</a>.</li>
             
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Cell cluster labelling with <b>sctype</b>. The <a href="https://doi.org/10.1038/s41467-022-28803-w" target="_blank">sctype reference dataset</a> compiled cell markers from 
                  <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/" target="_blank">CellMarker</a> and 
                  <a href="https://panglaodb.se/" target="_blank">PanglaoDB</a> databases and includes markers for the brain, pancreas, immune system, liver, eye, kidney, lung, embryo, gastrointestinal tract, muscle and skin.</li>
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Reference datasets for cell labelling with <b>SingleR</b>. These include: <br>
                  <b>Brain</b>
                  <ul><li>Darmanis Brain Data: scRNA-seq of 8 adult and 4 fetal human brain temporal lobe tissue (<a href="https://doi.org/10.1073/pnas.1507125112" target="_blank">Darmanis et al., 2015</a>). </li>
                  <li>Zhong Prefrontal Cortex Data: scRNA-seq of developing human embryonic prefrontal cortex from gestational weeks 8 to 26 (<a href="https://doi.org/10.1038/nature25980" target="_blank">Zhong et al., 2018</a>). </li></ul>
                  <b>Pancreas</b><br>
                          <ul><li>Baron Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors and 2 mouse strains (<a href="https://doi.org/10.1016/j.cels.2016.08.011" target="_blank">Baron et al., 2016</a>). </li>
                          <li>Lawlor Pancreas Data: scRNA-seq of 5 non-diabetic and 8 type 2 diabetic human islet samples (<a href="https://doi.org/10.1101/gr.212720.116" target="_blank">Lawlor et al., 2017</a>). </li>
                          <li>Muraro Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors (<a href="https://doi.org/10.1016/j.cels.2016.09.002" target="_blank">Muraro et al., 2016</a>). </li>
                          <li>Segerstolpe Pancreas Data: scRNA-seq of human islet cells from 6 healthy and 4 type 2 diabetic donors (<a href="https://doi.org/10.1016/j.cmet.2016.08.020" target="_blank">Segerstolpe et al., 2016</a>). </li></ul>
                  <b>Multiple organs</b>        
                          <ul><li>He Organ Atlas: scRNA-seq of 15 tissue organs of one adult donor. Tissue organs include bladder, blood, common bile duct, oesophagus, heart, liver,
                          lymph node, bone marrow, muscle, rectum, skin, small intestine, spleen, stomach and trachea 
                          (<a href="https://doi.org/10.1186/s13059-020-02210-0" target="_blank">He et al., 2020</a>).</li></ul>
             </li>
             </ul>
             </ul>            

             <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 1.0</span> - 15th March, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>Initial Release</li>
             </ul>
           </ul>'),
      easyClose = TRUE,
      footer = NULL,
      size = "xl",
    ))
  })
  
  observeEvent(input$updates2, {
    showModal(modalDialog(
      title = strong("ICARUS CHANGE LOG"),
      HTML('<h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.4</span> - 1st January, 2023</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW An option to impute dropouts using <a href="https://www.nature.com/articles/s41467-021-27729-z" target="_blank">Adaptively-thresholded low rank approximation (ALRA) method</a>.</li>
             </ul>
           </ul>
           
      <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.3</span> - 18th October, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Improved search functionality for genes and gene pathways.</li>
                          <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW The height of ICARUS windows are now adjustable for many of the plots.</li>
                                       <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW A document describing the objects in the ICARUS enviroment is now <a href="https://github.com/Enjewl/ICARUS/blob/master/ICARUS_enviroment_descriptions.md" target="_blank">available.</a> </li>
             </ul>
           </ul>
      
      
      <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.2</span> - 20th August, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Analyze gene Co-expression with MEGENA, read more about it <a href="https://doi.org/10.1371/journal.pcbi.1004574" target="_blank">here</a>.</li>
                          <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Analyze gene regulatory networks from transcription factors and target genes with SCENIC, read more about it <a href="https://doi.org/10.1038/nmeth.4463" target="_blank">here</a>.</li>
                                       <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Examine cell cluster expression association with various Genome Wide Association Studies (GWAS) traits using MAGMA, read more about it <a href="https://doi.org/10.1371/journal.pcbi.1004219" target="_blank">MAGMA</a>.</li>
                                                    <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Visualize and analyze cell-cell communication networks (ligand-receptor interactions between cells) with CellChat, read more about it <a href="https://doi.org/10.1038/s41467-021-21246-9" target="_blank">here</a>.</li>
                                                                                                                                                            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Examine drug-gene interactions of differentially expressed genes against the <a href="https://www.dgidb.org" target="_blank">DGIdb 4.0 database</a>!</li>
             </ul>
           </ul>
      
      <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.1</span> - 10th May, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>ICARUS is now published in <a href="https://doi.org/10.1093/nar/gkac322" target="_blank">Nucleic Acids Research!</a></li>
             </ul>
           </ul>
           
           <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 2.0 </span> - 21st April, 2022</h4>
            <ul class="list-unstyled mt-3"> 
            <ul>
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Perform trajectory analysis with <b>Monocle3</b>, read more about it <a href="https://doi.org/10.1038/nbt.2859" target="_blank">here</a>.</li>
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Detection and removal of doublets in data with <b>DoubletFinder</b>, read more about it <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">here</a>.</li>
             
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Cell cluster labelling with <b>sctype</b>. The <a href="https://doi.org/10.1038/s41467-022-28803-w" target="_blank">sctype reference dataset</a> compiled cell markers from 
                  <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/" target="_blank">CellMarker</a> and 
                  <a href="https://panglaodb.se/" target="_blank">PanglaoDB</a> databases and includes markers for the brain, pancreas, immune system, liver, eye, kidney, lung, embryo, gastrointestinal tract, muscle and skin.</li>
            <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>NEW Reference datasets for cell labelling with <b>SingleR</b>. These include: <br>
                  <b>Brain</b>
                  <ul><li>Darmanis Brain Data: scRNA-seq of 8 adult and 4 fetal human brain temporal lobe tissue (<a href="https://doi.org/10.1073/pnas.1507125112" target="_blank">Darmanis et al., 2015</a>). </li>
                  <li>Zhong Prefrontal Cortex Data: scRNA-seq of developing human embryonic prefrontal cortex from gestational weeks 8 to 26 (<a href="https://doi.org/10.1038/nature25980" target="_blank">Zhong et al., 2018</a>). </li></ul>
                  <b>Pancreas</b><br>
                          <ul><li>Baron Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors and 2 mouse strains (<a href="https://doi.org/10.1016/j.cels.2016.08.011" target="_blank">Baron et al., 2016</a>). </li>
                          <li>Lawlor Pancreas Data: scRNA-seq of 5 non-diabetic and 8 type 2 diabetic human islet samples (<a href="https://doi.org/10.1101/gr.212720.116" target="_blank">Lawlor et al., 2017</a>). </li>
                          <li>Muraro Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors (<a href="https://doi.org/10.1016/j.cels.2016.09.002" target="_blank">Muraro et al., 2016</a>). </li>
                          <li>Segerstolpe Pancreas Data: scRNA-seq of human islet cells from 6 healthy and 4 type 2 diabetic donors (<a href="https://doi.org/10.1016/j.cmet.2016.08.020" target="_blank">Segerstolpe et al., 2016</a>). </li></ul>
                  <b>Multiple organs</b>        
                          <ul><li>He Organ Atlas: scRNA-seq of 15 tissue organs of one adult donor. Tissue organs include bladder, blood, common bile duct, oesophagus, heart, liver,
                          lymph node, bone marrow, muscle, rectum, skin, small intestine, spleen, stomach and trachea 
                          (<a href="https://doi.org/10.1186/s13059-020-02210-0" target="_blank">He et al., 2020</a>).</li></ul>
             </li>
             </ul>
             </ul>            

             <h4 class="mt-4"> <span class="p-2 bg-light shadow rounded text-success"> ICARUS Version 1.0</span> - 15th March, 2022</h4>
             <ul class="list-unstyled mt-3">
             <ul>
             <li class="text-muted ml-3"><i class="mdi mdi-circle-medium mr-2"></i>Initial Release</li>
             </ul>
           </ul>'),
      easyClose = TRUE,
      footer = NULL,
      size = "xl",
    ))
  })
  
  ####QC1####
  observeEvent(input$QCInfo1, {
    showModal(modalDialog(
      title = strong("Quality Control"),
      h4(HTML('<h4 style="text-align: justify;">
         Allows the user to apply quality control thresholds to the cells within the loaded dataset. These include specifying thresholds for: <br>
        <ol><li>Number of unique genes per cell - Low quality cells will have very few genes, cell multiplets will have a high gene count.</li>
        <li>Number of molecules detected per cell - Low quality cells will have very few molecules, cell multiplets will have a high molecule count.</li>
        <li>Mitochondrial percentage - Low quality cells will contain high mitochondrial contamination.</li>
        <li>Ribosomal percentage - Ribosomal percentage will differ depending on the starting cell types, this metric can be useful to removal of cells that do not fall within the expected range.</li></ol><br>
         Violin plots visualise the distribution of each quality control metric.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####QC2####
  observeEvent(input$QC2Info1, {
    showModal(modalDialog(
      title = strong("Quality Control"),
      h4(HTML('<h4 style="text-align: justify;">
         Allows the user to apply quality control thresholds to the cells within the loaded dataset. These include specifying thresholds for: <br>
        <ol><li>Number of unique genes per cell - Low quality cells will have very few genes, cell multiplets will have a high gene count.</li>
        <li>Number of molecules detected per cell - Low quality cells will have very few molecules, cell multiplets will have a high molecule count.</li>
        <li>Mitochondrial percentage - Low quality cells will contain high mitochondrial contamination.</li>
        <li>Ribosomal percentage - Ribosomal percentage will differ depending on the starting cell types, this metric can be useful to removal of cells that do not fall within the expected range.</li></ol><br>
         Violin plots visualise the distribution of each quality control metric.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####D####
  observeEvent(input$DBoxInfo, {
    showModal(modalDialog(
      title = strong("Detect Doublets"),
      h4(HTML('<h4 style = "text-align:justify">
                          Cell doublets (multiplets) may arise during scRNA-seq library preparation (i.e., partitioning of multiple cells into a single droplet for droplet-based microfluidics methods). 
                          These doublets will exhibit a loss of single cell status and can compromise downstream analysis by creating spurious intermediate populations or transition states. <br> <br>
                          ICARUS incorporates the <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">DoubletFinder</a> R package to detect and remove doublets. 
                          DoubletFinder performs the following: <br>
                          <ol><li>Simulation of "artificial doublets" from existing scRNA-seq data.</li>
                          <li>Merge simulated artificial doublets with real data and perform dimensionality reduction. </li>
                          <li>A k-nearest neighbour graph is developed, and each cell is scored based on its proximity to artificial doublets. The highest scoring cells are assigned as real doublets. </li>
                          <li>The user has the option to visualise and remove these doublets.</li></ol>')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####D2####
  observeEvent(input$DBoxInfo2, {
    showModal(modalDialog(
      title = strong("Detect Doublets"),
      h4(HTML('<h4 style = "text-align:justify">
                          Cell doublets (multiplets) may arise during scRNA-seq library preparation (i.e., partitioning of multiple cells into a single droplet for droplet-based microfluidics methods). 
                          These doublets will exhibit a loss of single cell status and can compromise downstream analysis by creating spurious intermediate populations or transition states. <br> <br>
                          ICARUS incorporates the <a href="https://doi.org/10.1016/j.cels.2019.03.003" target="_blank">DoubletFinder</a> R package to detect and remove doublets. 
                          DoubletFinder performs the following: <br>
                          <ol><li>Simulation of "artificial doublets" from existing scRNA-seq data.</li>
                          <li>Merge simulated artificial doublets with real data and perform dimensionality reduction. </li>
                          <li>A k-nearest neighbour graph is developed, and each cell is scored based on its proximity to artificial doublets. The highest scoring cells are assigned as real doublets. </li>
                          <li>The user has the option to visualise and remove these doublets.</li></ol>')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  
  ####DR1####
  observeEvent(input$DRInfo1, {
    showModal(modalDialog(
      title = strong("Variable Features Plot"),
      h4(style="text-align: justify;",
         "The variable features plot showcases the top variable genes in the dataset for which the algorithm
         prioritises when performing dimensionality reduction. You may adjust the number of variable genes that are prioritised in the 'NEW' box."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DRInfo2, {
    showModal(modalDialog(
      title = strong("Dimension Reduction Heatmap"),
      h4(style="text-align: justify;",
         "Dimension reduction heatmap sorts cells and genes by their dimensionality reduction scores. This allows for nice visualisation
         of sources of heterogeneity within the dataset. You can visualize different dimensions using the sidebar navigation."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DRInfo3, {
    showModal(modalDialog(
      title = strong("Elbow Plot"),
      h4(style="text-align: justify;",
         "The Variance explained by each successive dimension plot (Elbow Plot) can be helpful
        in determining the number of dimensions that captures the majority of the variance (often referred to
        as the 'elbow point'). This elbow point is the recommended number of dimensions to use for clustering."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DRInfo4, {
    showModal(modalDialog(
      title = strong("Loadings Plot"),
      h4(style="text-align: justify;",
         "The loading plot visualizes top genes associated with reduction components. This allows for nice visualisation
         of sources of heterogeneity within the dataset. You can visualize different dimensions using the sidebar navigation."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####DR2####
  observeEvent(input$DR2Info1, {
    showModal(modalDialog(
      title = strong("Variable Features Plot"),
      h4(style="text-align: justify;",
         "The variable features plot showcases the top variable genes in the sample for which the algorithm
         prioritises when performing dimensionality reduction. Please use the sidebar menu to scroll through variable features plots for each sample.
         You may adjust the number of variable genes that are prioritised in the 'NEW' box."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DR2Info3, {
    showModal(modalDialog(
      title = strong("Dimension Reduction Heatmap (Integrated Dataset)"),
      h4(style="text-align: justify;",
         "Dimensionality reduction heatmap sorts cells and genes by their dimensionality reduction scores. This allows for nice visualisation
         of sources of heterogeneity within the integrated dataset. You can visualize different dimensions using the sidebar navigation."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DR2Info4, {
    showModal(modalDialog(
      title = strong("Elbow Plot (Integrated Dataset)"),
      h4(style="text-align: justify;",
         "The Variance explained by each successive dimension plot (Elbow Plot) can be helpful
        in determining the number of dimensions that captures the majority of the variance (often referred to
        as the 'elbow point'). This elbow point is the recommended number of dimensions to use for clustering."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DR2Info5, {
    showModal(modalDialog(
      title = strong("Loadings Plot (Integrated Dataset)"),
      h4(style="text-align: justify;",
         "The loading plot visualizes top genes associated with reduction components. This allows for nice visualisation
         of sources of heterogeneity within the integrated dataset. You can visualize different dimensions using the sidebar navigation."),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####C#####
  observeEvent(input$CBoxInfo, {
    showModal(modalDialog(
      title = strong("Select Clustering Parameters"),
      h4(HTML('<h4 style="text-align: justify;">The following clustering parameters may be adjusted in this step:<br>
                     <b>Nearest-neighbour graph construction </b> <br>
                     <ol><li><b>Number of Dimensions:</b> Dimensions of reduction to use as input. </li>
                     <li><b>k.param:</b> Defines k for the k-nearest neighbor algorithm. Larger k-values will allow more preservation of global structure.</li>
                     <li><b>n.trees:</b> More trees gives higher precision when using annoy approximate nearest neighbor search.</li></ol>
                     <b>Clustering Parameters</b><br>
                     <ol><li><b>Resolution:</b> Defines the number of clusters (use a larger number if you want to obtain a larger number of clusters).</li>
                     <li><b>Clustering Algorithm:</b> Select between Louvain, SLM and Leiden community detection algorithms.</li></ol>
                     <b>UMAP Parameters</b><br>
                     <ol><li><b>Number of Dimensions:</b> Dimensions of reduction to use as input.</li>
                     <li><b>k-nearest-neighbours:</b> This determines the number of neighboring points used in local approximations of manifold structure. Larger values will result in more global structure being preserved at the loss of detailed local structure. In general this parameter should be in the range 5 to 50.</li>
                     <li><b>min.dist:</b> Controls the distance between each point. Larger values ensure embedded points are more evenly distributed, while smaller values allow the algorithm to optimise more accurately with regard to local structure. Sensible values are in the range 0.001 to 0.5.</li></ol>
                     <b>t-SNE Parameters</b><br>
                     <ol><li><b>Number of Dimensions:</b> Dimensions of reduction to use as input.</li>
                     <li><b>Perplexity:</b> The approximation of the number of close neighbours when calculating similarity scores. A low perplexity will focus more on local structure, whereas high perplexity will preserve global structure.</li>
                     <li><b>Max iterations:</b> Number of iterations to be performed.</li></ol>')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$CBoxInfo3, {
    showModal(modalDialog(
      title = strong("View cell composition by sample"),
      h4(HTML('<h4 style="text-align: justify;">View composition of each cluster/cell type for each sample. Composition maybe shown as total cell number or as a % of all cells.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####CC#####
  observeEvent(input$CCBoxInfo, {
    showModal(modalDialog(
      title = strong("Effect Regression"),
      h4(HTML('<h4 style = "text-align:justify">This step allows the regression of the following:<br>
                          <ol><li><b>Cell cycle effect:</b> Expression of cell cycle phase canonical markers (taken from <a href="https://doi.org/10.1126/science.aad0501" target="_blank">Triosh et al., 2016</a>) are regressed from the data.</li>
                          <li><b>Number of unique genes (nFeature_RNA)</b></li>
                          <li><b>Number of molecules detected per cell (nCount_RNA)</b></li>
                          <li><b>Mitochondrial percentage</b></li>
                          <li><b>Ribosomal percentage</b></li>
                          <li><b>User input gene(s) of interest:</b> Average normalised gene(s) count are computed and regressed from the data (using the vars.to.regress function in Seurat::ScaleData).</li>
                          <li><b>User input gene pathway of interest:</b> Average normalised gene count for all genes in the gene pathway <a href="https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp" target="_blank">(GSEA/Msigdb database)</a> are computed and regressed from the data (using the vars.to.regress function in Seurat::ScaleData).</li></ol>
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####L#####
  observeEvent(input$LBoxInfo1, {
    showModal(modalDialog(
      title = strong("Label Clusters"),
      h4(HTML('<h4 style="text-align:justify"> Marker genes for each cluster may be analysed and compared against known cell markers to assign a cluster cell type. 
                  ICARUS enables users to add custom cell cluster labels or annotate clusters against either the sctype reference set or the singleR reference set. <br><br>
                  
                  The <a href="https://doi.org/10.1038/s41467-022-28803-w" target="_blank">sctype reference dataset</a> compiled cell markers from 
                  <a href="http://biocc.hrbmu.edu.cn/CellMarker/" target="_blank">CellMarker</a> and 
                  <a href="https://panglaodb.se/" target="_blank">PanglaoDB</a> databases. These include markers for the brain, pancreas, immune system, liver,
                  eye, kidney, lung, embryo, gastrointestinal tract, muscle and skin. <br> <br>
                  
                  The <a href="https://doi.org/10.1038/s41590-018-0276-y" target="_blank">singleR</a> R package incorporates cell markers from the following databases: <br>
                  <b>Brain</b>
                  <ol><li>Darmanis Brain Data: scRNA-seq of 8 adult and 4 fetal human brain temporal lobe tissue (<a href="https://doi.org/10.1073/pnas.1507125112" target="_blank">Darmanis et al., 2015</a>). </li>
                  <li>Zhong Prefrontal Cortex Data: scRNA-seq of developing human embryonic prefrontal cortex from gestational weeks 8 to 26 (<a href="https://doi.org/10.1038/nature25980" target="_blank">Zhong et al., 2018</a>). </li></ol>
                  <b>Immune system</b>
                          <ol><li>Blueprint Encode: scRNA-seq of 259 bulk RNA-seq samples generated by Blueprint and ENCODE from pure populations of stroma and immune cells
                          (<a href="https://doi.org/10.3324/haematol.2013.094243" target="_blank">Martens and Stunnenberg, 2013;</a>
                          <a href="https://doi.org/10.1038/nature11247" target="_blank">The ENCODE Consortium, 2012</a>).</li>
                          <li>Database of Immune Cell Expression (DICE): scRNA-seq of 1561 bulk RNA-seq samples generated by DICE from pure populations of human immune cells.
                          (<a href="https://doi.org/10.1016/j.cell.2018.10.022" target="_blank">Schmiedel et al., 2018</a>).</li>
                          <li>Immunologic Genome Project (ImmGen): scRNA-seq of 830 microarray samples generated by ImmGen from pure populations of murine immune cells 
                          (<a href="http://www.immgen.org/" target="_blank">http://www.immgen.org/</a>).</li>
                          <li>Monaco Immune Data: scRNA-seq of 114 bulk RNA-seq samples of sorted immune cell populations that can be found in GSE107011 
                          (<a href="https://doi.org/10.1016/j.celrep.2019.01.041" target="_blank">Monaco et al., 2019</a>).</li>
                          <li>Novershtern Hematopoietic Data: scRNA-seq of 211 bulk human microarray samples of sorted hematopoietic cell populations that can be found in GSE24759 
                          (<a href="https://doi.org/10.1016/j.cell.2011.01.004" target="_blank">Novershtern et al., 2011</a>).</li></ol>
                  <b>Pancreas</b><br>
                          <ol><li>Baron Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors and 2 mouse strains (<a href="https://doi.org/10.1016/j.cels.2016.08.011" target="_blank">Baron et al., 2016</a>). </li>
                          <li>Lawlor Pancreas Data: scRNA-seq of 5 non-diabetic and 8 type 2 diabetic human islet samples (<a href="https://doi.org/10.1101/gr.212720.116" target="_blank">Lawlor et al., 2017</a>). </li>
                          <li>Muraro Pancreas Data: scRNA-seq of pancreatic cells from 4 human donors (<a href="https://doi.org/10.1016/j.cels.2016.09.002" target="_blank">Muraro et al., 2016</a>). </li>
                          <li>Segerstolpe Pancreas Data: scRNA-seq of human islet cells from 6 healthy and 4 type 2 diabetic donors (<a href="https://doi.org/10.1016/j.cmet.2016.08.020" target="_blank">Segerstolpe et al., 2016</a>). </li></ol>
                  <b>Multiple organs</b>        
                          <ol><li>Mouse RNA-seq: 358 bulk RNA-seq samples of sorted cell populations 
                          (<a href="https://doi.org/10.1101/gr.240093.118" target="_blank">Benayoun et al., 2019</a>).</li>
                          <li>Human Primary Cell Atlas: scRNA-seq of 713 microarray samples from the Human Primary Cell Atlas 
                          (<a href="https://doi.org/10.1186/1471-2164-14-632" target="_blank">Mabbott et al., 2013</a>).</li>
                          <li>He Organ Atlas: scRNA-seq of 15 tissue organs of one adult donor. Tissue organs include bladder, blood, common bile duct, oesophagus, heart, liver,
                          lymph node, bone marrow, muscle, rectum, skin, small intestine, spleen, stomach and trachea 
                          (<a href="https://doi.org/10.1186/s13059-020-02210-0" target="_blank">He et al., 2020</a>).</li></ol>
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$LBoxInfo2, {
    showModal(modalDialog(
      title = strong("Marker Genes"),
      h4(HTML('<h4 style = "text-align:justify"> Table of marker genes for query cluster calculated using Seurat::FindAllMarkers function which compares each cluster to all other clusters. <br> <br>
              <b>Table columns:</b> <br>
              <b>Gene name:</b> Marker gene. <br>
              <b>Log2 fold change:</b> Log2 fold-change of the average expression between the query cluster against all other cluters. Positive values indicate that the gene is more highly expressed in the query cluster. <br>
              <b>Percentage of cell expressed: </b> Percentage of cells that the gene is expressed in. <br>
              <b>P-value: </b> Unadjusted p-value. <br>
              <b>Adjusted p-value:</b> Adjusted p-value. <br><br>
              User can apply thresholds to the type of test used, p-value adjustment method, p-value threshold, log2 fold change and minimum percentage of cells where a gene is detected (min.pct) in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$LBoxInfo3, {
    showModal(modalDialog(
      title = strong("Marker Gene Expression Summary"),
      h4(HTML('<h4 style = "text-align:justify"> Summary of marker genes expression for each cluster. 
      The plot shows the fraction of upregulated marker genes against all marker genes for each cluster.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  
  observeEvent(input$LBoxInfo4, {
    showModal(modalDialog(
      title = strong("Marker Genes Map"),
      h4(HTML('<h4 style = "text-align:justify"> Net map of marker genes expressed in each cluster. Colour of the lines represent the log2 fold change.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####MEGENA#####
  observeEvent(input$MEGENABoxInfo1, {
    showModal(modalDialog(
      title = strong("Co-expression Modules"),
      h4(HTML('<h4 style = "text-align:justify"> MEGENA computed co-expression modules. Triangles represent major modules hubs. Colours represent individual module clusters.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$MEGENABoxInfo2, {
    showModal(modalDialog(
      title = strong("Co-Expression Modules Info"),
      h4(HTML('<h4 style = "text-align:justify"> Table of MEGENA computed Co-expression modules. <br> <br>
              <b>Table columns:</b> <br>
              <b>Module ID:</b> Coexpression module. <br>
              <b>Module Size:</b> Number of co-expressed genes in module. <br>
              <b>Module Parent:</b> Hierarchical connection to another module. <br>
              <b>Module Hub:</b> Key defining gene of the module. <br>
              <b>Module p-value:</b> p-value significance of the formed co-expression module. <br><br>
              User can visualize various modules in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  
  observeEvent(input$MEGENABoxInfo3, {
    showModal(modalDialog(
      title = strong("Sunburst Plot"),
      h4(HTML('<h4 style = "text-align:justify"> Sunburst plot of module hierarchy. The colour of the plot represents either 
      <ol><li>The cell cluster/cell type with the highest activity of the corresponding coexpression module. </li>
              <li> The number or percentage of differentially expressed genes </li>
              <li> The number or percentage of custom differentially expressed genes </li></ol>
              Module activity is computed as a percentage of cluster marker gene overlap (Seurat::FindAllMarkers) with co-expression modules. <br> 
              Marker gene detection parameters, test type, log2 fold change threshold, p-value adjustment method, adjusted p-value threshold and 
              minimum percentage of marker gene expression (i.e. only include marker genes expressed in X% minimum percentage of cells in the cluster) can be adjusted in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$MEGENABoxInfo4, {
    showModal(modalDialog(
      title = strong("Co-Expression Modules Gene Ontology Heatmap"),
      h4(HTML('<h4 style = "text-align:justify"> Heatmap of significant GO terms associated with each co-expression module. <br> <br>
      The user may further select from 3 sub-ontologies including Biological Processes (BP), Cellular Components (CC) and Molecular Function (MF). <br><br>
              Please note that this step is computationally intensive and can take upwards of 30 minutes to complete.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####SCENIC#####
  observeEvent(input$RegulonBoxInfo, {
    showModal(modalDialog(
      title = strong("Regulon Expression"),
      h4(HTML('<h4 style = "text-align:justify"> Regulon activity across UMAP/t-SNE plots. <br>
              Please select a regulon from the list to view its activity across UMAP and t-SNE plots.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SCENICBoxInfo1, {
    showModal(modalDialog(
      title = strong("Gene Regulatory Network Heatmap"),
      h4(HTML('<h4 style = "text-align:justify"> Heatmap showing regulon activity across cell clusters.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SCENICBoxInfo2, {
    showModal(modalDialog(
      title = strong("Gene Regulatory Network Dotplot"),
      h4(HTML('<h4 style = "text-align:justify"> Dotplot of regulon specificity score (rss) across cell clusters. 
      The rss is calculated based on regulon activity and exclusivity to a certain cell cluster. 
              A higher z value indicates a higher rss score. <br>
              
              *Only regulons and cell types with any RSS > 0.01 are shown')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  
  observeEvent(input$SCENICBoxInfo3, {
    showModal(modalDialog(
      title = strong("Regulon Transcription Factor Map"),
      h4(HTML('<h4 style = "text-align:justify"> Map of regulons and shared transcription factors. The number of transcription factors shown can be altered in the sidebar menu.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  
  ####MAGMA#####
  
  observeEvent(input$MAGMABoxInfo1, {
    showModal(modalDialog(
      title = strong("Upload GWAS summary statistics for MAGMA cell type association analysis"),
      h4(HTML('<h4 style = "text-align:justify"> ICARUS performs MAGMA cell type association analysis on a user uploaded GWAS summary statistics.
      Please make sure uploaded GWAS summary statistics file contains SNP, CHR, BP as first three columns AND at least one of these 
      columns: Z,OR,BETA,LOG_ODDS,SIGNED_SUMSTAT. <br><br>
      NOTE: ONLY HUMAN DATASETS CAN BE PROCESSED.
      Please input a dataset name, number of GWAS individuals in the dataset, GWAS ethnic population and the sample of interest from single cell data to perform the analysis on.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$MAGMABoxInfo2, {
    showModal(modalDialog(
      title = strong("Select public GWAS dataset for MAGMA cell type association analysis"),
      h4(HTML('<h4 style = "text-align:justify"> ICARUS performs MAGMA cell type association analysis on public GWAS datasets. 
              Please select the GWAS dataset(s) and the sample of interest from single cell data to perform the analysis on. <br><br>
              Available GWAS datasets can be visualised in the table.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$MAGMABoxInfo3, {
    showModal(modalDialog(
      title = strong("GWAS Cell cluster/cell type association"),
      h4(HTML('<h4 style = "text-align:justify"> Heatmap of cell cluster/cell type association with GWAS datasets. -log10 FDR values are coloured in the heatmap. <br><br>
              <b>Linear mode: </b>Perform linear association for all significant MAGMA computed gene sets. <br>
              <b>Top 10% mode: </b>Perform linear association for top 10% significant MAGMA computed gene sets.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  
  observeEvent(input$MAGMABoxInfo4, {
    showModal(modalDialog(
      title = strong("GWAS Cell Cluster/Cell Type Association Info Table"),
      h4(HTML('<h4 style = "text-align:justify"> Summary table for GWAS cell cluster/cell type association. <br><br>
              <b>Celltype_id: </b>Cell cluster/cell type of interest. <br>
              <b>GWAS: </b>GWAS dataset ID. <br>
              <b>Trait: </b>GWAS dataset associated trait.<br>
              <b>Enrichment mode: </b>Linear mode performs association analysis with all significant MAGMA computed gene sets. Top10% mode performs
              association analysis with top 10% significant MAGMA computed gene sets.<br>
              <b>P: </b>P-value.<br>
              <b>log10p: </b>Log10 P-value.<br>
              <b>FDR: </b>FDR corrected P-value.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####GE#####
  observeEvent(input$GEBoxInfo, {
    showModal(modalDialog(
      title = strong("Gene Expression"),
      h4(HTML('<h4 style = "text-align:justify"> Select gene(s) or gene pathway to visualize normalised counts across cell clusters. List of gene pathways curated from 
              <a href="https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp" target="_blank">GSEA/Msigdb database</a> including 
      KEGG, Reactome, WikiPathways and Gene Ontology gene pathways.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####Chat#####
  observeEvent(input$ChatBoxInfo1, {
    showModal(modalDialog(
      title = strong("Cell-Cell Communication Network (Single Dataset)"),
      h4(HTML('<h4 style = "text-align:justify"> Aggregated cell-cell communication network shown as the number of interactions or the 
              total interaction strengths (weights) between any two cell populations. <br> <br>
              The option to view individual cell populations can be toggled in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo2, {
    showModal(modalDialog(
      title = strong("Cell-Cell Signalling Pathways"),
      h4(HTML('<h4 style = "text-align:justify"> Further visualisations of the cell-cell communication network for major signalling pathways. 
              Visualisations in the form of a hierarchy plot, chord diagram, circle plot and heatmap. <br> <br>
              <b>Hierarchy Plot: </b>Circle sizes are proportional to number of cells in each cell populations. Solid and open circles represent source and
              target, respectively. Edge colours are consistent with the sources as senders, and edge weights are proportional
              to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Chord Diagram: </b>The inner thinner bar colours represent the targets that recieve signal from the outer bar. The inner bar size is proportional
              to the signal strength received by the targets. Edge colours are consistent with the sources as senders, and edge weights are proportional
              to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Circle Plot: </b>Circle sizes are proportional to the number of cells in each cell populations. Edge colours are consistent with the sources as senders, 
              and edge weights are proportional to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Heatmap: </b>Heatmap of the communication probability between any two cell populations. The top coloured bar 
      plot represents the sum of the columns of values displayed in the heatmap. The right colored bar plot represents the sum of row of values. <br> <br>
              <b>Signalling roles </b> in the form of dominant senders, receivers, mediators and influencers for each cell cluster/cell type are inferred from network centrality
              measures. Please refer to the <a href="https://doi.org/10.1038/s41467-021-21246-9" target="_blank">CellChat paper</a> for more details.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo3, {
    showModal(modalDialog(
      title = strong("Ligand-Receptor Interaction For Signalling Pathways"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of the contribution of each ligand-receptor pair to signalling pathway(s).<br> <br>
               <b>Bar Graph: </b>Visualisation of cell-cell communications mediated by each ligand receptor pairs for the signalling pathway(s). <br>
               <b>Dot Plot: </b>Dot plot of the communication probabilities of signalling pathway(s) between any two cell populations. <br>
               Note* ONLY significant (p <0.05) interactions are shown. <br>
               <b>Chord Plot: </b>The inner thinner bar colours represent the targets that recieve signal from the outer bar. The inner bar size is proportional
              to the signal strength received by the targets. Edge colours are consistent with the sources as senders, and edge weights are proportional
              to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Gene Expression: </b>Gene expression of ligand-receptor pairs across cell populations.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo4, {
    showModal(modalDialog(
      title = strong("Cell-Cell Signalling Roles"),
      h4(HTML('<h4 style = "text-align:justify"> Summary of incoming and outgoing signalling pathways across all cell populations. <br> <br>
            <b>Heatmap: </b>summarising the information from outgoing and incoming signalling for each signalling pathway across cell populations. The top coloured bar 
      plot represents the sum of the columns of values displayed in the heatmap. The right colored bar plot represents the sum of row of values. <br>
      <b>Scatter Plot: </b>Scatter plot of the outgoing and incoming interaction strengths and allows identification of cell populations with significant
      changes in sending and receiving signals between different datasets.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo5, {
    showModal(modalDialog(
      title = strong("Cell-Cell Signalling Patterns"),
      h4(HTML('<h4 style = "text-align:justify"> Examination of global communication patterns to explore how multiple cell populations and signalling
      pathways coordinate together. <br> <br>
      <b>Incoming Patterns </b>show how the target cells coordinate with each other as well as how they coordinate
      with signalling pathways to respond to incoming signals. <br>
      <b>Outgoing Patterns </b>show how sender cells coordinate with each other as well as how they coordinate with 
      signalling pathways to drive communication. <br>
      <b>River Plot: </b>River plot shows associations of communication patterns with cell populations and signalling pathways. <br>
      <b>Dot Plot: </b>Dot plot shows the contribution score of each signalling pathways with cell populations.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo6, {
    showModal(modalDialog(
      title = strong("Manifold and Classification Learning Analysis of Signaling Networks"),
      h4(HTML('<h4 style = "text-align:justify"> Quantification of the cell-cell communication similarity between all significant signalling pathways.
              Similarity comparison may be done based on functional or structural similarity. <br> <br>
              <b>Function similarity: </b>High functional similarity indicates that senders and recievers are similar and therefore the two signalling pathways
              or ligand receptor pairs exhibit similar and/or redundant roles. <br>
              <b>Strucutral similarity: </b>Compares signalling network structure without considering the similarity of senders and recievers.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo7, {
    showModal(modalDialog(
      title = strong("Cell-Cell Communication Network (Two Datasets)"),
      h4(HTML('<h4 style = "text-align:justify"> Aggregated cell-cell communication network shown as the number of interactions or the 
              total interaction strengths (weights) between any two cell populations. <br> <br>
              The option to view individual cell populations can be toggled in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo8, {
    showModal(modalDialog(
      title = strong("Differential Interaction (Comparison Between Datasets)"),
      h4(HTML('<h4 style = "text-align:justify"> Comparison of cell-cell communication between datasets. <br> <br>
      <b>Bar Graph: </b>Total number of interactions and interaction strength of the inferred cell-cell communication networks from the two datasets. <br>
      <b>Communication Network: </b>Circle plot showing the differential number of interaction or interaction strength of cell-cell communication
      network between the two datasets. Red coloured edges represent increased signalling in the second dataset compared to the first. Blue coloured
      edges represent decreased signalling in the second dataset compared to the first dataset. <br>
      <b>Heatmap: </b>Heatmap of the differential number of interactions or interaction strength of cell-cell communication
      network between the two datasets. The top coloured bar plot represents the sum of the columns of values displayed in the heatmap (incoming signalling). 
      The right colored bar plot represents the sum of row of values (outgoing signaling). 
      In the colorbar, red (or blue) represents increased (or decreased) signaling in the second dataset compared to the first dataset.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo9, {
    showModal(modalDialog(
      title = strong("Cell-Cell Signalling Pathways"),
      h4(HTML('<h4 style = "text-align:justify"> Further visualisations of the cell-cell communication network for major signalling pathways. 
              Visualisations in the form of a hierarchy plot, chord diagram, circle plot and heatmap. <br> <br>
              <b>Hierarchy Plot: </b>Circle sizes are proportional to number of cells in each cell populations. Solid and open circles represent source and
              target, respectively. Edge colours are consistent with the sources as senders, and edge weights are proportional
              to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Chord Diagram: </b>The inner thinner bar colours represent the targets that recieve signal from the outer bar. The inner bar size is proportional
              to the signal strength received by the targets. Edge colours are consistent with the sources as senders, and edge weights are proportional
              to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Circle Plot: </b>Circle sizes are proportional to the number of cells in each cell populations. Edge colours are consistent with the sources as senders, 
              and edge weights are proportional to the interaction strength. Thicker edge line indicates a stronger signal. <br>
              <b>Heatmap: </b>Heatmap of the communication probability between any two cell populations. The top coloured bar 
      plot represents the sum of the columns of values displayed in the heatmap. The right colored bar plot represents the sum of row of values. <br> <br>
              <b>Signalling roles </b> in the form of dominant senders, receivers, mediators and influencers for each cell cluster/cell type are inferred from network centrality
              measures. Please refer to the <a href="https://doi.org/10.1038/s41467-021-21246-9" target="_blank">CellChat paper</a> for more details.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo10, {
    showModal(modalDialog(
      title = strong("Ligand-Receptor Interaction For Signalling Pathways"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of the contribution of each ligand-receptor pair to signalling pathway(s).<br> <br>
               <b>Dot Plot: </b>Dot plot of the communication probabilities of signalling pathway(s) between any two clusters/cell types of cells. <br>
               Note* ONLY significant (p <0.05) interactions are shown. <br>
              <b>Gene Expression: </b>Gene expression of ligand-receptor pairs across cell populations.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo11, {
    showModal(modalDialog(
      title = strong("Identification of Upregulated and Downregulated Signalling Ligand-Receptor Pairs"),
      h4(HTML('<h4 style = "text-align:justify"> Upregulated and downregulated signalling ligand-receptor pairs identified based on differential gene expression
      between the two datasets for each cell cluster/cell type. Higher communication probability indicates increased signalling in the one dataset compared to the other dataset. <br>
              Note* ONLY significant (p <0.05) interactions are shown.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo12, {
    showModal(modalDialog(
      title = strong("Cell-Cell Signalling Roles"),
      h4(HTML('<h4 style = "text-align:justify"> Summary of incoming and outgoing signalling pathways across all cell populations. <br> <br>
      <b>Bar Graph: </b>Comparison of the overall information flow (sum of communication probability among all pairs of cell populations in the network)
      of each signalling pathway in the two datasets. <br>
      <b>Heatmap: </b>Heatmap summarising the information from outgoing and incoming signalling for each signalling pathway across cell cluster/cell type. The top coloured bar 
      plot represents the sum of the columns of values displayed in the heatmap. The right colored bar plot represents the sum of row of values. <br>
      <b>Scatter Plot: </b>Scatter plot compares the outgoing and incoming interaction strengths and allows identification of cell populations with significant
      changes in sending and receiving signals between different datasets.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$ChatBoxInfo13, {
    showModal(modalDialog(
      title = strong("Manifold and Classification Learning Analysis of Signaling Networks"),
      h4(HTML('<h4 style = "text-align:justify"> Quantification of the cell-cell communication similarity between all significant signalling pathways.
              Similarity comparison may be done based on functional or structural similarity. <br> <br>
              <b>Function similarity: </b>High functional similarity indicates that senders and recievers are similar and therefore the two signalling pathways
              or ligand receptor pairs exhibit similar and/or redundant roles. <br>
              <b>Strucutral similarity: </b>Compares signalling network structure without considering the similarity of senders and recievers.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####TA#####
  observeEvent(input$TABoxInfo, {
    showModal(modalDialog(
      title = strong("Trajectory Analysis"),
      h4(HTML('<h4 style = "text-align:justify"> ICARUS employs the 
                          <a href="https://doi.org/10.1038/nbt.2859" target="_blank">Monocle3</a> algorithm to graph cells
                          according to their progress in <b>pseudotime</b> (a measure of how much progress an individual cell has made, i.e., cell differentiation). 
                          Pseudotime estimates the amount of transcriptional change a cell undergoes from its beginning to end states. <br><br>
                          
                          Calculation of pseudotime requires the user to select where the beginning of the biological process is (i.e., root cells). Root cells may be selected from
                          a cell cluster/cell type or manually selected using the lasso select function.')),
      img(id="lasso2",src="lasso_FINAL.png",width="100%"),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$TABoxInfo2, {
    showModal(modalDialog(
      title = strong("Monocle3 Trajectory Visualisation"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of trajectories computed by Monocle3 measured by pseudotime. Pseudotime estimates the amount of transcriptional change a cell undergoes from its beginning to end states. <br><br>
              The black lines show the structure of the graph. Note that the graph is not fully connected: cells in different partitions are in distinct components of the graph. 
              The circles with numbers in them denote special points within the graph. White circles denote the points at which pseudotime begins (beginning states). Each leaf, denoted by light gray circles, corresponds to a different outcome (i.e. cell fate) of the trajectory. 
              Black circles indicate branch nodes, in which cells can travel to one of several outcomes.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$TABoxInfo3, {
    showModal(modalDialog(
      title = strong("Box Plot Pseudotime visualisation"),
      h4(HTML('<h4 style = "text-align:justify"> Box plot visualisation of pseudotime across cell clusters/cell types.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$TABoxInfo4, {
    showModal(modalDialog(
      title = strong("Gene Changes Across Pseudotime"),
      h4(HTML('<h4 style = "text-align:justify"> Table of genes that change across pseudotime. <br><br>
              <b>Table columns:</b> <br>
              <b>Gene name:</b> Genes that change across pseudotime. <br>
              <b>p-value:</b> p-value. <br>
              <b>Morans I:</b> Morans I tells you whether cells at nearby positions on a trajectory will have similar (or dissimilar) expression levels for the gene being tested. Morans I ranges from -1 to +1. A +1 means that nearby cells will have
              perfectly similar expression; 0 represents no correlation and -1 means that neighbouring cells will be anti-correlated.<br>
              <b>q-value:</b> q-value.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$TABoxInfo5, {
    showModal(modalDialog(
      title = strong("Gene Input"),
      h4(HTML('<h4 style = "text-align:justify"> Gene input for plots on the right (gene expression across pseudotime).')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$TABoxInfo6, {
    showModal(modalDialog(
      title = strong("Gene Expression Across Pseudotime"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of gene expression across pseudotime as a line graph.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####DRUG#####
  observeEvent(input$DRUGBoxInfo1, {
    showModal(modalDialog(
      title = strong("Drug Gene Interactions"),
      h4(HTML('<h4 style = "text-align:justify"> Table of drug gene interactions for differentially expressed genes from the Differential Expression tab. <br> <br>
              <b>Table columns:</b> <br>
              <b>Gene name:</b> Differentially expressed genes. <br>
              <b>Log2 fold change:</b> Log2 fold-change of the average expression between the query cluster against all other cluters. Positive values indicate that the gene is more highly expressed in the query cluster. <br>
              <b>Adjusted p-value:</b> Adjusted p-value of significance. <br>
              <b>Drug Claim Name:</b> Name of drug with known interaction with the gene of interest. <br>
              <b>Drug Concept ID:</b> Chembl ID - includes additional information about the drug. <br>
              <b>Interaction Claim Source:</b> Relevant source where drug interaction is listed. <br>
              <b>Interaction types:</b> Type of drug-gene interaction. <br>
              <b>Interaction Group Score:</b> Amount of supporting evidence for the drug-gene interaction (higher scores = more evidence). 
              The interaction group score takes into account the number of confirmed drug-gene partners and number of supporting publications). <br>
              <b>PMIDs:</b> Pubmed reference (if available).')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DRUGBoxInfo2, {
    showModal(modalDialog(
      title = strong("Drug Gene Interactions Map"),
      h4(HTML('<h4 style = "text-align:justify"> Net map of drug-gene interactions for cluster specific differentially expressed genes. Colour of the lines represent the log2 fold change.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####DRUG2#####
  observeEvent(input$DRUGBoxInfo12, {
    showModal(modalDialog(
      title = strong("Drug Gene Interactions"),
      h4(HTML('<h4 style = "text-align:justify"> Table of drug gene interactions for differentially expressed genes from the Custom Differential Expression tab. <br> <br>
              <b>Table columns:</b> <br>
              <b>Gene name:</b> Differentially expressed genes. <br>
              <b>Log2 fold change:</b> Log2 fold-change of the average expression between the query cluster against all other cluters. Positive values indicate that the gene is more highly expressed in the query cluster. <br>
              <b>Adjusted p-value:</b> Adjusted p-value of significance. <br>
              <b>Drug Claim Name:</b> Name of drug with known interaction with the gene of interest. <br>
              <b>Drug Concept ID:</b> Chembl ID - includes additional information about the drug. <br>
              <b>Interaction Claim Source:</b> Relevant source where drug interaction is listed. <br>
              <b>Interaction types:</b> Type of drug-gene interaction. <br>
              <b>Interaction Group Score:</b> Amount of supporting evidence for the drug-gene interaction (higher scores = more evidence). 
              The interaction group score takes into account the number of confirmed drug-gene partners and number of supporting publications). <br>
              <b>PMIDs:</b> Pubmed reference (if available).')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DRUGBoxInfo22, {
    showModal(modalDialog(
      title = strong("Drug Gene Interactions Map"),
      h4(HTML('<h4 style = "text-align:justify"> Net map of drug-gene interactions for cluster specific differentially expressed genes. Colour of the lines represent the log2 fold change.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####MM#####
  observeEvent(input$MMBoxInfo, {
    showModal(modalDialog(
      title = strong("Gene Expression"),
      h4(HTML('<h4 style = "text-align:justify"> Select gene(s) to visualize normalised counts across cell clusters.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####DE#####
  observeEvent(input$DEBoxInfo1, {
    showModal(modalDialog(
      title = strong("Differentially Expressed Genes"),
      h4(HTML('<h4 style = "text-align:justify"> Table of differentially expressed genes calculated using Seurat::FindMarkers command which performs differential expression based on the non-parametric Wilcoxon rank sum test. <br> <br>
              <b>Table columns:</b> <br>
              <b>Gene Name:</b> Differentially expressed gene. <br>
              <b>p-value:</b> Unadjusted p-value. <br>
              <b>Log2 fold change:</b> Log2 fold-change of the average expression between the two clusters. Positive values indicate that the gene is more highly expressed in the first cluster. <br>
              <b>pct.1:</b> The percentage of cells where the gene is detected in the first cluster/sample. <br>
              <b>pct.2:</b> The percentage of cells where the gene is detected in the second cluster/sample. <br>
              <b>Adjusted p-value:</b> Adjusted p-value. <br><br>
              User can apply thresholds to the type of test used, p-value adjustment method, p-value threshold, log2 fold change and minimum percentage of cells where a gene is detected in either group (min.pct) in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DEBoxInfo2, {
    showModal(modalDialog(
      title = strong("Gene set Enrichment Analysis"),
      h4(HTML('<h4 style = "text-align:justify"> Table of enriched pathways. Gene set enrichment analysis performed with                   
                    <a href="https://doi.org/10.1016/j.xinn.2021.100141" target="_blank">ClusterProfiler</a> and
                    <a href="https://doi.org/10.1039/C5MB00663E" target="_blank">ReactomePA</a> R packages. <br> <br>
                    <b>Table columns:</b><br>
              <b>ID:</b> Pathway identifier. <br>
              <b>Description:</b> Enriched pathway. <br>
              <b>Adjusted p-value:</b> User input multiple testing correction adjusted p-value. <br><br>
              User can apply thresholds to the p-value and p-value adjustment method in the minimizable sidebar.<br><br>
              If Gene Ontology enrichment analysis is selected, the user may further select from 3 sub-ontologies including Biological Processes (BP), Cellular Components (CC) and Molecular Function (MF).
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DEBoxInfo3, {
    showModal(modalDialog(
      title = strong("Volcano Plot"),
      h4(HTML('<h4 style = "text-align:justify"> Volcano plot of differentially expressed genes. <br><br>
              User can apply thresholds to the p-value and log2 fold change in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$DEBoxInfo4, {
    showModal(modalDialog(
      title = strong("Gene Pathways"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of log2 fold changes for specific gene pathway (GSEA/Msigdb gene pathways including KEGG, Reactome, WikiPathways and Gene Ontology). <br> <br>
              Select gene pathway to visualize using the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####SDE#####
  observeEvent(input$SDEBoxInfo1, {
    showModal(modalDialog(
      title = strong("Differentially Expressed Genes"),
      h4(HTML('<h4 style = "text-align:justify"> Table of differentially expressed genes calculated using Seurat::FindMarkers command which performs differential expression based on the non-parametric Wilcoxon rank sum test. <br> <br>
                            <b>Table columns:</b> <br>
              <b>Gene Name:</b> Differentially expressed gene. <br>
              <b>p-value:</b> Unadjusted p-value. <br>
              <b>Log2 fold change:</b> Log2 fold-change of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group (Group 1). <br>
              <b>pct.1:</b> The percentage of cells where the gene is detected in the first group (Group 1). <br>
              <b>pct.2:</b> The percentage of cells where the gene is detected in the second group (Group 2). <br>
              <b>Adjusted p-value:</b> Adjusted p-value. <br><br>
              User can apply thresholds to the type of test used, p-value adjustment method, p-value threshold, log2 fold change and minimum percentage of cells where a gene is detected in either group (min.pct) in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SDEBoxInfo2, {
    showModal(modalDialog(
      title = strong("Gene Set Enrichment Analysis"),
      h4(HTML('<h4 style = "text-align:justify"> Table of enriched pathways. Gene set enrichment analysis performed with                   
                    <a href="https://doi.org/10.1016/j.xinn.2021.100141" target="_blank">ClusterProfiler</a> and
                    <a href="https://doi.org/10.1039/C5MB00663E" target="_blank">ReactomePA</a> R packages. <br> <br>
                                        <b>Table columns:</b><br>
              <b>ID:</b> Pathway identifier. <br>
              <b>Description:</b> Enriched pathway. <br>
              <b>Adjusted p-value:</b> User input multiple testing correction adjusted p-value. <br><br>
              User can apply thresholds to the p-value and p-value adjustment method in the minimizable sidebar.<br><br>
              If Gene Ontology enrichment analysis is selected, the user may further select from 3 sub-ontologies including Biological Processes (BP), Cellular Components (CC) and Molecular Function (MF).
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SDEBoxInfo3, {
    showModal(modalDialog(
      title = strong("Volcano Plot"),
      h4(HTML('<h4 style = "text-align:justify"> Volcano plot of differentially expressed genes. <br><br>
              User can apply thresholds to the p-value and log2 fold change in the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SDEBoxInfo4, {
    showModal(modalDialog(
      title = strong("Gene Pathways"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of log2 fold changes for specific gene pathway (GSEA/Msigdb gene pathways including KEGG, Reactome, WikiPathways and Gene Ontology). <br> <br>
              Select gene pathway to visualize using the minimizable sidebar.')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####PA#####
  observeEvent(input$PABoxInfo1, {
    showModal(modalDialog(
      title = strong("Visualisation of Enriched Terms"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of enriched terms as either a dot plot, ridge plot or heatmap. <br>
      <b>Dot Plot</b><br>
              Enriched terms are plotted in order of gene ratio (number of enriched genes in gene pathway/total number of genes in gene pathway). 
              Size of points represent the number of significantly enriched genes in the gene pathway Colour of the points represent the p-value. <br>
              <b>Ridge Plot </b><br>
              The ridge plot will visualise expression distribution of core enriched genes for enriched categories. It helps users to interpret up/down-regulated pathways. <br>
              <b>Heatmap</b><br>
              Heatmap of enriched terms to identify expression patterns. <br> <br>
              The number of categories shown may be adjusted in the minimizable sidebar.
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$PABoxInfo2, {
    showModal(modalDialog(
      title = strong("Gene Concept Network"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of complex association and interactions between genes and enriched terms as a network. 
      Colour of gene points represent the log2 fold change. <br> <br>
              The number of categories shown may be adjusted in the minimizable sidebar. Gene and category labels may be toggled on and off.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$PABoxInfo3, {
    showModal(modalDialog(
      title = strong("Enrichment Map"),
      h4(HTML('<h4 style = "text-align:justify"> 
              Network of enriched terms with edges connecting overlapping gene pathways. 
              Mutually overlapping gene pathways are clustered together allowing easy identification of affected pathways.
              Size of points represent the number of significantly enriched genes in the gene pathway. Colour of the points represent the p-value. <br> <br>
              The number of categories shown may be adjusted in the minimizable sidebar.
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$PABoxInfo4, {
    showModal(modalDialog(
      title = strong("View Specific Pathways"),
      h4(HTML('<h4 style = "text-align:justify">  
              View selected reactome pathway. Colour of the points represent the log2 fold change.       
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$PABoxInfo5, {
    showModal(modalDialog(
      title = strong("Tree Plot"),
      h4(HTML('<h4 style = "text-align:justify">  
              Hierarchical clustering of enriched terms. This reduces the complexity of enriched terms and improve user interpretation ability.       
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####SPA#####
  observeEvent(input$SPABoxInfo1, {
    showModal(modalDialog(
      title = strong("Visualisation of Enriched Terms"),
      h4(HTML('<h4 style = "text-align:justify"> Visualisation of enriched terms as either a dot plot, ridge plot or heatmap. <br>
      <b>Dot Plot</b><br>
              Enriched terms are plotted in order of gene ratio (number of enriched genes in gene pathway/total number of genes in gene pathway). 
              Size of points represent the number of significantly enriched genes in the gene pathway Colour of the points represent the p-value. <br>
              <b>Ridge Plot </b><br>
              The ridge plot will visualise expression distribution of core enriched genes for enriched categories. It helps users to interpret up/down-regulated pathways. <br>
              <b>Heatmap</b><br>
              Heatmap of enriched terms to identify expression patterns. <br> <br>
              The number of categories shown may be adjusted in the minimizable sidebar.
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SPABoxInfo2, {
    showModal(modalDialog(
      title = strong("Gene Concept Network"),
      h4(HTML('<h4 style = "text-align:justify"> Gene Concept Network:</b> Visualisation of complex association and interactions between genes and enriched terms as a network.
      Colour of gene points represent the log2 fold change. <br> <br>
              The number of categories shown may be adjusted in the minimizable sidebar. Gene and category labels may be toggled on and off.
              ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SPABoxInfo3, {
    showModal(modalDialog(
      title = strong("Enrichment Map"),
      h4(HTML('<h4 style = "text-align:justify"> 
              Network of enriched terms with edges connecting overlapping gene pathways. 
              Mutually overlapping gene pathways are clustered together allowing easy identification of affected pathways.
              Size of points represent the number of significantly enriched genes in the gene pathway. Colour of the points represent the p-value. <br> <br>
              The number of categories shown may be adjusted in the minimizable sidebar.
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SPABoxInfo4, {
    showModal(modalDialog(
      title = strong("View Specific Pathways"),
      h4(HTML('<h4 style = "text-align:justify">  
              View selected reactome pathway. Colour of the points represent the log2 fold change.       
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$SPABoxInfo5, {
    showModal(modalDialog(
      title = strong("Tree Plot"),
      h4(HTML('<h4 style = "text-align:justify">  
              Hierarchical clustering of enriched terms. This reduces the complexity of enriched terms and improve user interpretation ability.       
                          ')),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  ####Download Data#####
  
  observe({
    if (!is.null(seuratreactive$objQC)) {
      shinyjs::show("downloadDataSaveButton")
    }
    else {
      shinyjs::hide("downloadDataSaveButton")
    }
  })
  
  observeEvent(input$downloadDataSaveButton, {
    showModal(modalDialog(
      title = strong("Save Options"),
      h4(HTML('<h4 style = "text-align:justify">  
              Select between saving the current working enviroment for reloading into ICARUS or saving the Seurat object ONLY for analysis in R.       
                          ')),
      
      h4(downloadBttn("downloadDataSave", "ICARUS Enviroment File", size = "lg"),
         downloadBttn("downloadDataSaveRDS", "Seurat Object ONLY", size = "lg")),
      size = "l",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadDataSaveRDS <- downloadHandler(
    filename = function() {
      
      if (!is.null(seuratreactive$integera)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")
      }
      
      else if (!is.null(seuratreactive$VFintegrated)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")
      }
      
      else if (!is.null(seuratreactive$variablefeatures)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")
      }
      
      else if (!is.null(seuratreactive$Removed2)) {
        paste(paste(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        paste(paste(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")
      }
      
      else if (!is.null(seuratreactive$Removed)) {
        paste(paste(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")
      }
      
      else if (!is.null(seuratreactive$objQC)) {
        paste(paste(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_SeuratObject", ".RDS", sep = "")       
      }
    },
    content = function(file) {
      if (!is.null(seuratreactive$obj)) {
        Final_Seurat_Object <- seuratreactive$obj
        
        saveRDS(Final_Seurat_Object, file = file)
      }
      else {
        Seurat_Object <- seuratreactive$obj_original
        saveRDS(Seurat_Object, file = file)
      }
    }
  )
  
  output$downloadDataSave <- downloadHandler(
    filename = function() {
      
      if (!is.null(seuratreactive$markersDRUG2)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_CustomDrugGeneInteractions_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(YuSPA$result)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_CustomPathwayAnalysis_Completed", ".Rdata", sep = "")
      }  
      
      else if (!is.null(seuratreactive$markersSDE)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_CustomDifferentialExpression_Completed", ".Rdata", sep = "")
      }  
      
      else if (!is.null(seuratreactive$markersDRUG)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_DrugGeneInteractions_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(Yu$result)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_PathwayAnalysis_Completed", ".Rdata", sep = "")
      }  
      
      else if (!is.null(seuratreactive$markersDE)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_DifferentialExpression_Completed", ".Rdata", sep = "")
      }  
      
      else if (!is.null(seuratreactive$MM)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_MultimodalAnalysis_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$Chatdb)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_Cell-CellCommunication_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$TAselectedcells)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_TrajectoryAnalysis_Completed", ".Rdata", sep = "")
      } 
      
      else if (!is.null(seuratreactive$MAGMA_Final_Results)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_GWAS_Associations_Completed", ".Rdata", sep = "")
      } 
      
      else if (!is.null(seuratreactive$regulons)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_GeneRegulatoryNetwork_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$summary.output)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_GeneCoexpressionModules_Completed", ".Rdata", sep = "")
      } 
      
      else if (!is.null(seuratreactive$new.cell.ids)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_Labelling_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$reductionCC1)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_DataCorrection_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$integera)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_Clustering_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$VFintegrated)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_Integrated_DimensionalityReduction_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$variablefeatures)) {
        paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Seurat_DimensionalityReduction_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$Removed2)) {
        paste(paste(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_Seurat_Integrated_DoubletRemoval_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        paste(paste(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_SeuratIntegrated_QualityControl_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$Removed)) {
        paste(paste(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_Seurat_DoubletRemoval_Completed", ".Rdata", sep = "")
      }
      
      else if (!is.null(seuratreactive$objQC)) {
        paste(paste(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_Seurat_QualityControl_Completed", ".Rdata", sep = "")       
      }
    },
    content = function(file) {
      
      #Load
      mincells <- seuratreactive$mincells 
      minfeatures <- seuratreactive$minfeatures 
      
      #Load2
      mincellsCombine <- seuratreactive$mincellsCombine 
      minfeaturesCombine <- seuratreactive$minfeaturesCombine 
      
      #QC
      PostQC_dataset1 <- seuratreactive$objQC
      type <- seuratreactive$type
      original_dataset1 <- seuratreactive$obj_original
      QC_nFeature_threshold1_dataset1 <- seuratreactive$nFeature1
      QC_nFeature_threshold2_dataset1 <- seuratreactive$nFeature2
      QC_nCount_threshold1_dataset1 <- seuratreactive$nCount1 
      QC_nCount_threshold2_dataset1 <- seuratreactive$nCount2
      QC_percent.mt_threshold1_dataset1 <- seuratreactive$percent.mt1
      QC_percent.mt_threshold2_dataset1 <- seuratreactive$percent.mt2 
      QC_percent.ribo_threshold1_dataset1 <- seuratreactive$percent.ribo1
      QC_percent.ribo_threshold2_dataset1 <- seuratreactive$percent.ribo2
      
      #D
      Remove <- seuratreactive$Remove
      Removed <- seuratreactive$Removed
      doublesim <- seuratreactive$doublesim
      plot.data.doublets <- seuratreactive$plot.data.doublets
      
      #QC2
      PostQC_dataset2 <- seuratreactive$objQC2
      type2 <- seuratreactive$type2
      original_dataset2 <- seuratreactive$obj_original2
      QC_nFeature_threshold1_dataset2 <- seuratreactive$nFeature12
      QC_nFeature_threshold2_dataset2 <- seuratreactive$nFeature22
      QC_nCount_threshold1_dataset2 <- seuratreactive$nCount12
      QC_nCount_threshold2_dataset2 <- seuratreactive$nCount22
      QC_percent.mt_threshold1_dataset2 <- seuratreactive$percent.mt12
      QC_percent.mt_threshold2_dataset2<- seuratreactive$percent.mt22
      QC_percent.ribo_threshold1_dataset2 <- seuratreactive$percent.ribo12
      QC_percent.ribo_threshold2_dataset2 <- seuratreactive$percent.ribo22
      
      #D2
      Remove2 <- seuratreactive$Remove2
      Removed2 <- seuratreactive$Removed2
      doublesim2 <- seuratreactive$doublesim2
      plot.data.doublets2 <- seuratreactive$plot.data.doublets2
      
      #DR
      Final_Seurat_Object <- seuratreactive$obj
      variablefeatures <- seuratreactive$variablefeatures
      SCTransformDR <- seuratreactive$SCTransformDR
      ALRAL <- seuratreactive$ALRAL
      npcDR <- seuratreactive$npcDR
      
      #DR2
      Combined.list <- seuratreactive$Combined.list
      VFintegrated <- seuratreactive$VFintegrated
      integrationreduction <- seuratreactive$integrationreduction
      kanchor <- seuratreactive$kanchor
      SCTransformDR2 <- seuratreactive$SCTransformDR2
      npcDR2 <- seuratreactive$npcDR2
      
      #C
      NumberofPCs_GraphConstruction <- seuratreactive$integera
      k.param <- seuratreactive$integerb
      n.trees <- seuratreactive$integerc
      Resolution <- seuratreactive$integerd
      NumberofPCs_UMAP <- seuratreactive$integere
      knn_UMAP <- seuratreactive$integerf
      min.dist_UMAP <- seuratreactive$integerg
      ClusteringAlgorithm <- seuratreactive$select1
      NumberofPCs_tsne <- seuratreactive$tsne1
      Perplexity_tsne <- seuratreactive$tsne2
      max.iterations_tsne <- seuratreactive$tsne3
      CellComposition <- seuratreactive$Identifiers
      plot.data <- seuratreactive$plot.data
      plot.data.original <- seuratreactive$plot.data.original
      
      #CC
      Regress_Confounder <- seuratreactive$reductionCC1
      Regress_genes <- seuratreactive$reductionCC2 
      Regress_pathway <- seuratreactive$reductionCC3 
      GeneR <- seuratreactive$GeneR
      msigdb <- seuratreactive$msig
      orgDb <- seuratreactive$orgDb 
      KEGG_species<- seuratreactive$EKEGG
      REACTOME_species <- seuratreactive$REACT
      REACTOME_species2 <- seuratreactive$PAREACTOME
      
      #L
      new.cell.ids <- seuratreactive$new.cell.ids
      reference_annotation <- seuratreactive$reference_annotation
      select2 <- seuratreactive$select2
      selecttissue <- seuratreactive$selecttissue
      
      #MEGENA
      MEGENA_samples <- seuratreactive$MEGENA_samples
      variablegenesMEGENA <- seuratreactive$variablegenesMEGENA
      MEGENA_selectmethod <- seuratreactive$MEGENA_selectmethod
      MEGENAmin.size <- seuratreactive$MEGENAmin.size
      MEGENApval <- seuratreactive$MEGENApval
      MEGENA_selectcluster <- seuratreactive$MEGENA_selectcluster
      n_MEGENA <- seuratreactive$n_MEGENA
      log2FCMEGENA <- seuratreactive$log2FCMEGENA
      padjustMEGENA <- seuratreactive$padjustMEGENA
      testuseMEGENA <- seuratreactive$testuseMEGENA
      minMEGENA <- seuratreactive$minMEGENA
      pvalMEGENA <- seuratreactive$pvalMEGENA
      summary.output <- seuratreactive$summary.output
      markersMEGENA <- seuratreactive$markersMEGENA
      moduleGO_df <- seuratreactive$moduleGO_df
      VF <- seuratreactive$VF
      g <- seuratreactive$g
      
      #SCENIC
      variablefeaturesSCENIC <- seuratreactive$variablefeaturesSCENIC
      variablegenesSCENIC <- seuratreactive$variablegenesSCENIC
      regulonAUC <- seuratreactive$regulonAUC
      regulons <- seuratreactive$regulons
      cellInfo <- seuratreactive$cellInfo
      n_SCENIC <- seuratreactive$n_SCENIC
      SCENICdb <- seuratreactive$SCENICdb
      SCENIC_samples <- seuratreactive$SCENIC_samples
      regulonnames <- seuratreactive$regulonnames
      titleSCENIC <- seuratreactive$titleSCENIC
      SCENICInput <- seuratreactive$SCENICInput
      SCENIC_selectcluster <- seuratreactive$SCENIC_selectcluster
      SCENIC_selectmethod <- seuratreactive$SCENIC_selectmethod
      regulonActivity_byCellType_Scaled <- seuratreactive$regulonActivity_byCellType_Scaled
      rss <- seuratreactive$rss
      
      #MAGMA
      MAGMA_samples <- seuratreactive$MAGMA_samples
      MAGMA_GWAS <- seuratreactive$MAGMA_GWAS
      MAGMA_Final_Results <- seuratreactive$MAGMA_Final_Results
      meta <- seuratreactive$meta
      MAGMA_selectcluster <- seuratreactive$MAGMA_selectcluster
      MAGMA_samples2 <- seuratreactive$MAGMA_samples2
      MAGMA_selectcluster2 <- seuratreactive$MAGMA_selectcluster2
      textlabelMAGMA <- seuratreactive$textlabelMAGMA
      MAGMA_N <- seuratreactive$MAGMA_N
      GRCh <- seuratreactive$GRCh
      MAGMA_Pop <- seuratreactive$MAGMA_Pop
      n_MAGMA <- seuratreactive$n_MAGMA
      n_MAGMA2 <- seuratreactive$n_MAGMA2
      
      #TA
      TAselectedcells <- seuratreactive$TAselectedcells
      TA_pickercluster <- seuratreactive$TA_pickercluster
      seurat_monocle <- seuratreactive$seurat_monocle
      Monocle3_genes <- seuratreactive$Monocle3_genes
      
      #CellChat
      cellchat <- seuratreactive$cellchat
      cellchatM <- seuratreactive$cellchatM
      object.list <- seuratreactive$object.list
      ChatChoose <- seuratreactive$ChatChoose
      ChatSample <- seuratreactive$ChatSample
      Chatdb <- seuratreactive$Chatdb
      ChatMultiple1Sample <- seuratreactive$ChatMultiple1Sample
      ChatMultiple2Sample <- seuratreactive$ChatMultiple2Sample
      Chat_selectcluster <- seuratreactive$Chat_selectcluster 
      Chat_number <- seuratreactive$Chat_number
      Chatpval <- seuratreactive$Chatpval
      
      #DRUG
      markersDRUG <- seuratreactive$markersDRUG
      DRUGdf <- seuratreactive$DRUGdf
      
      #DRUG2
      markersDRUG2 <- seuratreactive$markersDRUG2
      DRUGdf2 <- seuratreactive$DRUGdf2
      
      #MM
      MM <- seuratreactive$MM
      
      #GE
      titleGE <- seuratreactive$titleGE
      GeneExpression_GeneInput <- seuratreactive$GEInput
      GeneExpression_PathwayInput <- seuratreactive$GSEInput
      GenePathwayNames <- seuratreactive$Gene_vector_msig
      
      #SDE
      onlyposSDE <- seuratreactive$onlyposSDE
      selectedcells_lasso1 <- seuratreactive$selectedkey1
      selectedcells_lasso2 <- seuratreactive$selectedkey2
      GOlabelsSDE <- seuratreactive$GOlabelsSDE
      CustomDifferentialExpression_log_threshold <- seuratreactive$logSDE
      CustomDifferentialExpression_min.pct_threshold <- seuratreactive$minSDE
      padjustSDE1 <- seuratreactive$padjustSDE1
      testuseSDE <- seuratreactive$testuseSDE
      CustomDifferentialExpression_GeneSetEnrichmentInput <- seuratreactive$enrichInputSDE
      CustomDifferentialExpression_GeneSetEnrichment_pvalue_threshold <- seuratreactive$pvalueSDE
      pvalSDE <- seuratreactive$pvalSDE
      CustomDifferentialExpression_GeneSetEnrichment_padjust_method <- seuratreactive$padjustSDE
      CustomDifferentialExpression_volcano_Foldchange_threshold1 <- seuratreactive$volcanoSDEvalue1
      CustomDifferentialExpression_volcano_Foldchange_threshold2 <- seuratreactive$volcanoSDEvalue2
      CustomDifferentialExpression_volcano_pvalue <- seuratreactive$pvaluevolcanoSDE
      CustomDifferentialExpression_Table <- seuratreactive$markersSDE
      CustomDifferentialExpression_GeneSetEnrichment <- seuratreactive$YuSDE
      CustomDifferentialExpression_volcano_markers <- seuratreactive$markersSDEvolcano
      CustomDifferentialExpression_GeneSetEnrichment_PATHWAY <- seuratreactive$PATHWAYSDE
      
      #SPA
      SelectCustomPathwayAnalysis <- seuratreactive$SPAchoose
      CustomPathwayAnalysis_pvalue_adjustment_method <- seuratreactive$padjustSPA
      GOlabelsSPA  <- seuratreactive$GOlabelsSPA 
      CustomPathwayAnalysis_pvalue <- seuratreactive$pvalueSPA
      CustomPathwayAnalysis_GeneSetEnrichment <- YuSPA$result
      CustomPathwayAnalysis_GeneSetEnrichment2 <- YuSPA$ema
      geneList_CustomPathway <- seuratreactive$geneListSPA
      
      #DE
      onlyposDE <- seuratreactive$onlyposDE
      DifferentialExpression_clusters_BetweenSampleComparison <- seuratreactive$DE_clusters_samples
      DifferentialExpression_ClusterOfInterest_BetweenSampleComparison <- seuratreactive$DECluster
      DifferentialExpression_SampleOfInterest1_BetweenSampleComparison <- seuratreactive$DEInput1_samples
      DifferentialExpression_SampleOfInterest2_BetweenSampleComparison <- seuratreactive$DEInput2_samples
      DifferentialExpression_comparison <- seuratreactive$SelectComparison 
      DifferentialExpression_clusters_BetweenClustersComparison <- seuratreactive$DE_clusters
      DifferentialExpression_SampleOfInterest1_BetweenClustersComparison <- seuratreactive$DEInput1
      DifferentialExpression_SampleOfInterest2_BetweenClustersComparison <- seuratreactive$DEInput2
      DifferentialExpression_log_threshold <- seuratreactive$logDE
      DifferentialExpression_min.pct_threshold <- seuratreactive$minDE
      padjustDE1 <- seuratreactive$padjustDE1
      testuseDE <- seuratreactive$testuseDE
      DifferentialExpression_GeneSetEnrichmentInput <- seuratreactive$enrichInputDE
      GOlabelsDE <- seuratreactive$GOlabelsDE
      DifferentialExpression_GeneSetEnrichment_pvalue_threshold <- seuratreactive$pvalueDE
      pvalDE <- seuratreactive$pvalDE
      DifferentialExpression_GeneSetEnrichment_padjust_method <- seuratreactive$padjustDE
      DifferentialExpression_volcano_Foldchange_threshold1 <- seuratreactive$volcanovalue1
      DifferentialExpression_volcano_Foldchange_threshold2 <- seuratreactive$volcanovalue2
      DifferentialExpression_volcano_pvalue <- seuratreactive$pvaluevolcano
      DifferentialExpression_Table <- seuratreactive$markersDE
      DifferentialExpression_GeneSetEnrichment <- seuratreactive$YuDE
      DifferentialExpression_volcano_markers <- seuratreactive$markersDEvolcano
      DifferentialExpression_GeneSetEnrichment_PATHWAY <- seuratreactive$PATHWAY
      
      #PA
      SelectPathwayAnalysis <- seuratreactive$PAchoose
      PathwayAnalysis_pvalue_adjustment_method <- seuratreactive$padjustPA
      PathwayAnalysis_pvalue <- seuratreactive$pvaluePA
      GOlabelsPA <- seuratreactive$GOlabelsPA
      PathwayAnalysis_GeneSetEnrichment <- Yu$result
      PathwayAnalysis_GeneSetEnrichment2 <- Yu$ema
      geneListPathway <- seuratreactive$geneListPA
      
      save("PostQC_dataset1", "type", "original_dataset1", "QC_nFeature_threshold1_dataset1", "QC_nFeature_threshold2_dataset1", "QC_nCount_threshold1_dataset1", "QC_nCount_threshold2_dataset1",
           "QC_percent.mt_threshold1_dataset1", "QC_percent.mt_threshold2_dataset1", "QC_percent.ribo_threshold1_dataset1", "QC_percent.ribo_threshold2_dataset1", "mincells", "minfeatures",
           "Remove", "Removed", "doublesim", "plot.data.doublets",
           "PostQC_dataset2", "type2", "original_dataset2", "QC_nFeature_threshold1_dataset2", "QC_nFeature_threshold2_dataset2", "QC_nCount_threshold1_dataset2", "QC_nCount_threshold2_dataset2",
           "QC_percent.mt_threshold1_dataset2", "QC_percent.mt_threshold2_dataset2", "QC_percent.ribo_threshold1_dataset2", "QC_percent.ribo_threshold2_dataset2", "mincellsCombine", "minfeaturesCombine",
           "Remove2", "Removed2", "doublesim2", "plot.data.doublets2", 
           "Final_Seurat_Object", "variablefeatures", "SCTransformDR", "ALRAL", "npcDR",
           "Combined.list", "VFintegrated", "integrationreduction", "kanchor", "SCTransformDR2", "npcDR2",
           "NumberofPCs_GraphConstruction", "k.param", "n.trees", "Resolution", "NumberofPCs_UMAP", "knn_UMAP", "min.dist_UMAP", "ClusteringAlgorithm", "NumberofPCs_tsne", "Perplexity_tsne",
           "max.iterations_tsne", "CellComposition", "plot.data", "plot.data.original",
           "Regress_Confounder", "Regress_genes", "Regress_pathway", "GeneR", "msigdb", "orgDb", "KEGG_species", "REACTOME_species", "REACTOME_species2",
           "new.cell.ids", "reference_annotation", "select2", "selecttissue", "TAselectedcells", "TA_pickercluster", "Monocle3_genes", "seurat_monocle", "MM",
           "MEGENA_samples", "variablegenesMEGENA", "MEGENA_selectmethod", "MEGENApval", "MEGENAmin.size", "MEGENA_selectcluster", "n_MEGENA", "log2FCMEGENA", "testuseMEGENA", "padjustMEGENA", "minMEGENA", "pvalMEGENA", "summary.output", "markersMEGENA", "moduleGO_df", "VF", "g",
           "variablefeaturesSCENIC", "variablegenesSCENIC", "regulonAUC", "rss", "regulonActivity_byCellType_Scaled", "regulons", "cellInfo", "n_SCENIC", "SCENICdb", "SCENIC_samples", "regulonnames", "titleSCENIC", "SCENICInput", "SCENIC_selectcluster", "SCENIC_selectmethod",
           "MAGMA_samples", "MAGMA_GWAS", "MAGMA_Final_Results", "meta", "MAGMA_selectcluster", "n_MAGMA", "n_MAGMA2", "MAGMA_selectcluster2", "MAGMA_samples2", "textlabelMAGMA", "MAGMA_N", "GRCh", "MAGMA_Pop",
           "titleGE", "GeneExpression_GeneInput", "GeneExpression_PathwayInput", "GenePathwayNames",
           "cellchat", "cellchatM", "object.list", "ChatChoose", "ChatSample", "Chatdb", "ChatMultiple1Sample", "ChatMultiple2Sample", "Chat_selectcluster", "Chat_number", "Chatpval",
           "markersDRUG", "markersDRUG2", "DRUGdf", "DRUGdf2",
           "selectedcells_lasso1", "selectedcells_lasso2", "GOlabelsSDE", "padjustSDE1", "onlyposSDE", "testuseSDE", "CustomDifferentialExpression_log_threshold", "CustomDifferentialExpression_min.pct_threshold", "CustomDifferentialExpression_GeneSetEnrichmentInput",
           "CustomDifferentialExpression_GeneSetEnrichment_pvalue_threshold", "pvalSDE", "CustomDifferentialExpression_GeneSetEnrichment_padjust_method", "CustomDifferentialExpression_volcano_Foldchange_threshold1",
           "CustomDifferentialExpression_volcano_Foldchange_threshold2", "CustomDifferentialExpression_volcano_pvalue", "CustomDifferentialExpression_Table", "CustomDifferentialExpression_GeneSetEnrichment",
           "CustomDifferentialExpression_volcano_markers", "CustomDifferentialExpression_GeneSetEnrichment_PATHWAY",
           "SelectCustomPathwayAnalysis", "CustomPathwayAnalysis_pvalue_adjustment_method", "GOlabelsSPA", "CustomPathwayAnalysis_pvalue", "CustomPathwayAnalysis_GeneSetEnrichment",
           "CustomPathwayAnalysis_GeneSetEnrichment2", "geneList_CustomPathway",
           "DifferentialExpression_clusters_BetweenSampleComparison", "padjustDE1", "onlyposDE", "testuseDE", "DifferentialExpression_ClusterOfInterest_BetweenSampleComparison", "DifferentialExpression_SampleOfInterest1_BetweenSampleComparison",
           "DifferentialExpression_SampleOfInterest2_BetweenSampleComparison", "DifferentialExpression_comparison", "DifferentialExpression_clusters_BetweenClustersComparison", "DifferentialExpression_SampleOfInterest1_BetweenClustersComparison",
           "DifferentialExpression_SampleOfInterest2_BetweenClustersComparison", "DifferentialExpression_log_threshold", "DifferentialExpression_min.pct_threshold", "DifferentialExpression_GeneSetEnrichmentInput", "GOlabelsDE",
           "DifferentialExpression_GeneSetEnrichment_pvalue_threshold", "pvalDE", "DifferentialExpression_GeneSetEnrichment_padjust_method", "DifferentialExpression_volcano_Foldchange_threshold1", "DifferentialExpression_volcano_Foldchange_threshold2",
           "DifferentialExpression_volcano_pvalue", "DifferentialExpression_Table", "DifferentialExpression_GeneSetEnrichment", "DifferentialExpression_volcano_markers", "DifferentialExpression_GeneSetEnrichment_PATHWAY",
           "SelectPathwayAnalysis", "PathwayAnalysis_pvalue_adjustment_method", "PathwayAnalysis_pvalue", "GOlabelsPA", "PathwayAnalysis_GeneSetEnrichment", "PathwayAnalysis_GeneSetEnrichment2", "geneListPathway",
           file = file)
    }
  )
  
  ####Download images#####
  ####QC####
  observeEvent(input$downloadQCViolin, {
    showModal(modalDialog(
      title = strong("Download Quality Control Violin Plot"),
      numericInput("QCViolinHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("QCViolinWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("QCViolinfileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("QCViolindownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$QCViolindownloadoutput <- downloadHandler(
    filename = function(){
      paste0(paste0(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_QualityControl_ViolinPlot", ".", input$QCViolinfileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = VlnPlot(seuratreactive$objQC, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"), group.by = "orig.ident", pt.size = input$sizeQC, ncol = 2),
             width = input$QCViolinWidth,
             height = input$QCViolinHeight,
             units = "mm",
             device = input$QCViolinfileformat)
    }
  )
  
  observeEvent(input$downloadQCUMI, {
    showModal(modalDialog(
      title = strong("Download Quality Control nCount_RNA vs nFeature_RNA Plot"),
      numericInput("QCUMIHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("QCUMIWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("QCUMIfileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("QCUMIdownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$QCUMIdownloadoutput <- downloadHandler(
    filename = function(){
      paste0(paste0(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_QualityControl_nfeatureRNA_Vs_nCountRNA", ".", input$QCUMIfileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot =  FeatureScatter(seuratreactive$objQC, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident", pt.size = 2) + NoLegend(),
             width = input$QCUMIWidth,
             height = input$QCUMIHeight,
             units = "mm",
             device = input$QCUMIfileformat)
    }
  )
  
  observeEvent(input$downloadQCFeature, {
    showModal(modalDialog(
      title = strong("Download Quality Control nCount_RNA vs percent.mt Plot"),
      numericInput("QCFeatureHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("QCFeatureWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("QCFeaturefileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("QCFeaturedownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$QCFeaturedownloadoutput <- downloadHandler(
    filename = function(){
      paste0(paste0(levels(as.factor(seuratreactive$objQC@meta.data$orig.ident)), collapse = "_"), "_QualityControl_nCountRNA_Vs_percent.mt", ".", input$QCFeaturefileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = FeatureScatter(seuratreactive$objQC, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident", pt.size = 2) + NoLegend(),
             width = input$QCFeatureWidth,
             height = input$QCFeatureHeight,
             units = "mm",
             device = input$QCFeaturefileformat)
    }
  )
  
  ####QC2####
  observeEvent(input$downloadQCViolin2, {
    showModal(modalDialog(
      title = strong("Download Quality Control Violin Plot"),
      numericInput("QCViolinHeight2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("QCViolinWidth2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("QCViolinfileformat2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("QCViolindownloadoutput2", "Download"),
      easyClose = TRUE,
      size = "s",
      footer = NULL
    ))
  })
  
  output$QCViolindownloadoutput2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_QualityControl_ViolinPlot_SecondDataset.", input$QCViolinfileformat2, sep = "") 
    },
    
    content = function(file) {
      ggsave(file, plot = VlnPlot(seuratreactive$objQC2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"), group.by = "orig.ident", pt.size = input$sizeQC2, ncol = 2),
             width = input$QCViolinWidth2,
             height = input$QCViolinHeight2,
             units = "mm",
             device = input$QCViolinfileformat2)
    }
  )
  
  observeEvent(input$downloadQCUMI2, {
    showModal(modalDialog(
      title = strong("Download Quality Control nCount_RNA vs nFeature_RNA Plot"),
      numericInput("QCUMI2Height", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("QCUMI2Width", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("QCUMI2fileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("QCUMI2downloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$QCUMI2downloadoutput <- downloadHandler(
    filename = function(){
      paste0(paste0(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_QualityControl_nfeatureRNA_Vs_nCountRNA_SecondDataset", ".", input$QCUMI2fileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot =  FeatureScatter(seuratreactive$objQC2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident", pt.size = 2) + NoLegend(),
             width = input$QCUMI2Width,
             height = input$QCUMI2Height,
             units = "mm",
             device = input$QCUMI2fileformat)
    }
  )
  
  observeEvent(input$downloadQCFeature2, {
    showModal(modalDialog(
      title = strong("Download Quality Control nCount_RNA vs percent.mt Plot"),
      numericInput("QCFeature2Height", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("QCFeature2Width", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("QCFeature2fileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("QCFeature2downloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$QCFeature2downloadoutput <- downloadHandler(
    filename = function(){
      paste0(paste0(levels(as.factor(seuratreactive$objQC2@meta.data$orig.ident)), collapse = "_"), "_QualityControl_nCountRNA_Vs_percent.mt_SecondDataset", ".", input$QCFeature2fileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = FeatureScatter(seuratreactive$objQC2, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 2, group.by = "orig.ident",) + NoLegend(),
             width = input$QCFeature2Width,
             height = input$QCFeature2Height,
             units = "mm",
             device = input$QCFeature2fileformat)
    }
  )
  
  ####DR#####
  observeEvent(input$downloadDRVF, {
    showModal(modalDialog(
      title = strong("Download Variable Features Plot"),
      numericInput("VFHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("VFWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("VFfileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("VFdownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$VFdownloadoutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DimensionalityReduction_VariableFeaturesPlot.", input$VFfileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = LabelPoints(VariableFeaturePlot(seuratreactive$obj, pt.size = 2), head(VariableFeatures(seuratreactive$obj), 10), repel = TRUE),
             width = input$VFWidth,
             height = input$VFHeight,
             units = "mm",
             device = input$VFfileformat)
    }
  )
  
  observeEvent(input$downloadDRPC, {
    showModal(modalDialog(
      title = strong("Download Dimensionality Reduction Heatmap"),
      numericInput("PCHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("PCWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("PCfileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("PCdownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$PCdownloadoutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DimensionalityReduction_Heatmap.", input$PCfileformat, sep = "")
    }, 
    
    content = function(file) {
      p <- DimHeatmap(seuratreactive$obj, dims = input$integer, nfeatures = input$PCnfeatures, cells = input$PCncells, balanced = TRUE, ncol = 1, fast = F)
      p[[1]][["labels"]][["title"]] <- paste("PC_", input$integer, collapse = "")
      ggsave(file, plot = p,
             width = input$PCWidth,
             height = input$PCHeight,
             units = "mm",
             device = input$PCfileformat)
    }
  )
  
  observeEvent(input$downloadDRElbow, {
    showModal(modalDialog(
      title = strong("Download Elbow Plot"),
      numericInput("ElbowHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("ElbowWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("Elbowfileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("Elbowdownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$Elbowdownloadoutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DimensionalityReduction_ElbowPlot.", input$Elbowfileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = ElbowPlot(seuratreactive$obj, ndims = seuratreactive$npcDR),
             width = input$ElbowWidth,
             height = input$ElbowHeight,
             units = "mm",
             device = input$Elbowfileformat)
    }
  )
  
  observeEvent(input$downloadDRLoadings, {
    showModal(modalDialog(
      title = strong("Download Loadings Plot"),
      numericInput("LoadingsHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("LoadingsWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("Loadingsfileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("Loadingsdownloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$Loadingsdownloadoutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DimensionalityReduction_LoadingsPlot.", input$Loadingsfileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = VizDimLoadings(seuratreactive$obj, dims = input$integer2, nfeatures = input$Loadingsnfeatures),
             width = input$LoadingsWidth,
             height = input$LoadingsHeight,
             units = "mm",
             device = input$Loadingsfileformat)
    }
  )
  
  #####DR2####  
  observeEvent(input$downloadDR2VF1, {
    showModal(modalDialog(
      title = strong("Download Variable Features Plot (Integrated Dataset)"),
      numericInput("DR2VF1Height", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("DR2VF1Width", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("DR2VF1fileformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("DR2VF1downloadoutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$DR2VF1downloadoutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_IntegratedDataset_DimensionalityReduction_VariableFeaturesPlot.", input$DR2VF1fileformat, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = LabelPoints(VariableFeaturePlot(seuratreactive$Combined.list[[input$integerVFPlot]], pt.size = 2), head(VariableFeatures(seuratreactive$Combined.list[[input$integerVFPlot]]), 10), repel = TRUE),
             width = input$DR2VF1Width,
             height = input$DR2VF1Height,
             units = "mm",
             device = input$DR2VF1fileformat)
    }
  )
  
  observeEvent(input$downloadDR2PC, {
    showModal(modalDialog(
      title = strong("Download Dimensionality Reduction Heatmap"),
      numericInput("PCHeightDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("PCWidthDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("PCfileformatDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("PCdownloadoutputDR2", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$PCdownloadoutputDR2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_IntegratedDataset_DimensionalityReduction_Heatmap.", input$PCfileformatDR2, sep = "")
    }, 
    
    content = function(file) {
      if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
        p <- DimHeatmap(seuratreactive$obj, dims = input$integerDR2, nfeatures = input$PCnfeaturesDR2, cells = input$PCncellsDR2, balanced = TRUE, ncol = 1, fast = F)
      }
      else if (seuratreactive$integrationreduction == "harmony") {
        p <- DimHeatmap(seuratreactive$obj, dims = input$integerDR2, reduction = "harmony", nfeatures = input$PCnfeaturesDR2, cells = input$PCncellsDR2, balanced = TRUE, ncol = 1, fast = F)
      }
      
      p[[1]][["labels"]][["title"]] <- paste("Dimension_", input$integerDR2, collapse = "")
      
      ggsave(file, plot = p,
             width = input$PCWidthDR2,
             height = input$PCHeightDR2,
             units = "mm",
             device = input$PCfileformatDR2)
    }
  )
  
  observeEvent(input$downloadDR2Elbow, {
    showModal(modalDialog(
      title = strong("Download Elbow Plot"),
      numericInput("ElbowHeightDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("ElbowWidthDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ElbowfileformatDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("ElbowdownloadoutputDR2", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$ElbowdownloadoutputDR2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_IntegratedDataset_DimensionalityReduction_ElbowPlot.", input$ElbowfileformatDR2, sep = "")
    }, 
    
    content = function(file) {
      if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
        q <- ElbowPlot(seuratreactive$obj, ndims = seuratreactive$npcDR2)
      }
      else if (seuratreactive$integrationreduction == "harmony") {
        q <- ElbowPlot(seuratreactive$obj, reduction = "harmony", ndims = seuratreactive$npcDR2)
      }
      
      ggsave(file, plot = q,
             width = input$ElbowWidthDR2,
             height = input$ElbowHeightDR2,
             units = "mm",
             device = input$ElbowfileformatDR2)
    }
  )
  
  observeEvent(input$downloadDR2Loadings, {
    showModal(modalDialog(
      title = strong("Download Loadings Plot"),
      numericInput("LoadingsHeightDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("LoadingsWidthDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("LoadingsfileformatDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("LoadingsdownloadoutputDR2", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$LoadingsdownloadoutputDR2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_IntegratedDataset_DimensionalityReduction_LoadingsPlot.", input$LoadingsfileformatDR2, sep = "")
    }, 
    
    content = function(file) {
      if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
        q <- VizDimLoadings(seuratreactive$obj, dims = input$integer2DR2, nfeatures = input$LoadingsnfeaturesDR2)
      }
      else if (seuratreactive$integrationreduction == "harmony") {
        q <- VizDimLoadings(seuratreactive$obj, dims = input$integer2DR2, reduction = "harmony", nfeatures = input$LoadingsnfeaturesDR2)
      }
      
      ggsave(file, plot = q,
             width = input$LoadingsWidthDR2,
             height = input$LoadingsHeightDR2,
             units = "mm",
             device = input$LoadingsfileformatDR2)
    }
  )
  
  ####C######
  observeEvent(input$downloadComposition, {
    showModal(modalDialog(
      title = strong("Download Cell Composition plot"),
      numericInput("ViolinHeightSamplesC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("ViolinWidthSamplesC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ViolinFileInputSamplesC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("ViolindownloadoutputSamplesC", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$ViolindownloadoutputSamplesC <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_CellCompositionPlot.", input$ViolinFileInputSamplesC, sep = "")
    }, 
    
    content = function(file) {
      if (input$Sample_selectionC == "Seurat Clusters") {
        if (input$switchPercent == FALSE) {
          plot <- ggplot(seuratreactive$Graph, aes(fill=`Seurat Clusters`, y=`Value`, x=`Sample`, label = `Value`)) + 
            geom_bar(position="stack", stat="identity") + 
            geom_text(size = input$labelsizeC/2, position = position_stack(vjust = 0.5)) +
            ylab("") + 
            theme_bw() +
            theme(text = element_text(size=input$labelsizeC),
                  axis.text.x= element_text(size = input$labelsizeC),
                  axis.text.y= element_text(size = input$labelsizeC))
        }
        else {
          plot <- ggplot(seuratreactive$Graph, aes(x=Sample, y=Percentage, fill = `Seurat Clusters`)) +
            geom_bar(position = "fill", stat = "identity") +
            scale_y_continuous(labels = scales::percent) +
            geom_text(aes(label = paste0(Percentage*100,"%")), 
                      position = position_stack(vjust = 0.5), size = input$labelsizeC/2) +
            ylab("") +
            theme_bw() +
            theme(text = element_text(size=input$labelsizeC),
                  axis.text.x= element_text(size = input$labelsizeC),
                  axis.text.y= element_text(size = input$labelsizeC))
        }
      }
      else if (input$Sample_selectionC == "Labelled Cells") {
        if (input$switchPercent == FALSE) {
          plot <- ggplot(seuratreactive$Graph, aes(fill=`Labelled Cells`, y=`Value`, x=`Sample`, label = `Value`)) + 
            geom_bar(position="stack", stat="identity") + 
            geom_text(size = input$labelsizeC/2, position = position_stack(vjust = 0.5)) +
            ylab("") + 
            theme_bw() +
            theme(text = element_text(size=input$labelsizeC),
                  axis.text.x= element_text(size = input$labelsizeC),
                  axis.text.y= element_text(size = input$labelsizeC))
        }
        else {
          plot <- ggplot(seuratreactive$Graph, aes(x=Sample, y=Percentage, fill = `Labelled Cells`)) +
            geom_bar(position = "fill", stat = "identity") +
            scale_y_continuous(labels = scales::percent) +
            geom_text(aes(label = paste0(Percentage*100,"%")), 
                      position = position_stack(vjust = 0.5), size = input$labelsizeC/2) +
            ylab("") +
            theme_bw() +
            theme(text = element_text(size=input$labelsizeC),
                  axis.text.x= element_text(size = input$labelsizeC),
                  axis.text.y= element_text(size = input$labelsizeC))
        }
      }
      ggsave(file, plot = plot,
             width = input$ViolinWidthSamplesC,
             height = input$ViolinHeightSamplesC,
             units = "mm",
             device = input$ViolinFileInputSamplesC)
      
    }
  )
  
  ####Labelling####
  output$downloadmarkersL <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Labelling_MarkerGenesTable.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$markers, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadmarkersUporDown, {
    showModal(modalDialog(
      title = strong("Download Marker Gene Expression Summary Plot"),
      numericInput("downloadmarkersUporDownHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadmarkersUporDownWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadmarkersUporDownFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadmarkersUporDownOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadmarkersUporDownOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_MarkerGeneExpressionSummaryPlot.", input$downloadmarkersUporDownFile, sep = "")
    }, 
    
    content = function(file) {
      
      summary <- seuratreactive$markers %>%
        group_by(Cluster) %>%
        summarise(sum = sum(`Log2 fold change` > 0), n = n()) %>% 
        mutate(Ratio = signif((sum/n) - 0.5, 4)) %>% 
        mutate(`Expression Change`= ifelse(Ratio>0,"Upregulated","Downregulated"))
      
      plot <- ggplot(summary, aes(fill=`Expression Change`, y=Ratio, x=Cluster)) + 
        geom_bar(position="stack", stat="identity") + 
        ylab("Upregulated markers/total markers") + 
        scale_y_continuous(limits = c(-0.5, 0.5), labels = function(y) y + 0.5) +
        scale_fill_manual(values=c("blue","red"), name="") +
        theme_bw() + 
        theme(axis.text.x = element_text(angle = 90, vjust=0.5),
              text = element_text(size=input$labelsizeUporDown))
      
      ggsave(file, plot = plot,
             width = input$downloadmarkersUporDownWidth,
             height = input$downloadmarkersUporDownHeight,
             units = "mm",
             device = input$downloadmarkersUporDownFile)
    }
  )
  
  observeEvent(input$downloadmarkersCluster, {
    showModal(modalDialog(
      title = strong("Download Marker Gene Map"),
      numericInput("downloadmarkersClusterHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadmarkersClusterWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadmarkersClusterFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadmarkersClusterOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadmarkersClusterOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_MarkerGeneMap.", input$downloadmarkersClusterFile, sep = "")
    }, 
    
    content = function(file) {
      markersmap <- seuratreactive$markers[seuratreactive$markers$`Adjusted p-value` <= 0.05,] %>% 
        group_by(`Cluster`) %>% 
        top_n(n = input$MGNumber, wt = `Log2 fold change`)
      
      edges <- markersmap[,c("Cluster", "Gene Name", "Log2 fold change")]
      
      nodes1 <- markersmap %>%
        group_by(`Gene Name`) %>%
        summarise(node_color = "gene", n = 2) %>% 
        drop_na()
      
      nodes2 <- markersmap %>%
        group_by(Cluster) %>%
        summarise(node_color = "cluster", n = 2.1) %>% 
        drop_na()
      
      colnames(nodes2)[1] <- c("Gene Name")
      
      nodes <- full_join(nodes1, nodes2)
      
      graph = graph_from_data_frame(edges, directed = FALSE, nodes)
      
      if (input$checkboxgene_markers == TRUE && input$checkboxcluster_markers == TRUE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodemarkers, input$nodemarkers/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontmarkers, repel=TRUE) +
          geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontmarkers, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      else if (input$checkboxgene_markers == FALSE && input$checkboxcluster_markers == TRUE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodemarkers, input$nodemarkers/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontmarkers, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      else if (input$checkboxgene_markers == TRUE && input$checkboxcluster_markers == FALSE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodemarkers, input$nodemarkers/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontmarkers, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      ggsave(file, plot = plot,
             width = input$downloadmarkersClusterWidth,
             height = input$downloadmarkersClusterHeight,
             units = "mm",
             device = input$downloadmarkersClusterFile)
    }
  )
  
  ####MEGENA####
  observeEvent(input$downloadmodulesMEGENA, {
    showModal(modalDialog(
      title = strong("Download Coexpression Module Plot"),
      numericInput("downloadmodulesMEGENAHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadmodulesMEGENAWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadmodulesMEGENAFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadmodulesMEGENAOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadmodulesMEGENAOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_CoexpressionModulesPlot.", input$downloadmodulesMEGENAFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$radioMEGENA == "View All Modules") {
        module.table <- seuratreactive$summary.output$module.table
        colnames(module.table)[1] <- "id" # first column of module table must be labelled as "id".
        
        plot <- plot_module_hierarchy(module.table = module.table,label.scaleFactor = input$genefontMEGENA,
                                      arrow.size = 0.015, node.label.color = "blue", edge.color = "red")
      }
      else {
        plot <- plot_module(output.summary = seuratreactive$summary.output,PFN = seuratreactive$g,subset.module = input$MEGENA_modulesInput,
                            layout = "kamada.kawai",label.hubs.only = input$checkboxMEGENA,
                            gene.set = NULL,color.code =  "grey",
                            output.plot = FALSE, out.dir = "modulePlot",col.names = palette(rainbow(6)),label.scaleFactor = input$genefontMEGENA,
                            hubLabel.col = "black", label.alpha = input$TranspraentMEGENA, show.topn.hubs = Inf,show.legend = TRUE)
      }
      
      ggsave(file, plot = plot[[1]],
             width = input$downloadmodulesMEGENAWidth,
             height = input$downloadmodulesMEGENAHeight,
             units = "mm",
             device = input$downloadmodulesMEGENAFile)
    }
  )
  
  output$downloadtable_MEGENA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_CoExpressionModules.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$summary.output$module.table[,c(1,2,3,4,6)], file, row.names = F)
    }
  )
  
  observeEvent(input$downloadSunburstMEGENA, {
    showModal(modalDialog(
      title = strong("Download Coexpression Sunburst Plot"),
      numericInput("downloadSunburstMEGENAHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadSunburstMEGENAWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadSunburstMEGENAFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadSunburstMEGENAOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadSunburstMEGENAOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_CoexpressionSunburstPlot.", input$downloadSunburstMEGENAFile, sep = "")
    }, 
    
    content = function(file) {
      
      ggsave(file, plot = seuratreactive$sunburst,
             width = input$downloadSunburstMEGENAWidth,
             height = input$downloadSunburstMEGENAHeight,
             units = "mm",
             device = input$downloadSunburstMEGENAFile)
    }
  )
  
  observeEvent(input$downloadGOMEGENA, {
    showModal(modalDialog(
      title = strong("Download Coexpression GO Heatmap"),
      numericInput("downloadGOMEGENAHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadGOMEGENAWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadGOMEGENAFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadGOMEGENAOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadGOMEGENAOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_CoexpressionGOHeatmap.", input$downloadGOMEGENAFile, sep = "")
    }, 
    
    content = function(file) {
      
      plot <- plotModuleGO(seuratreactive$moduleGO_df, nTerms = input$GOtermsMEGENA, text_size = input$GOtextMEGENA, coord_flip = TRUE, axis_x_text_angle = 90) +
        theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust = 1))
      
      ggsave(file, plot = plot,
             width = input$downloadGOMEGENAWidth,
             height = input$downloadGOMEGENAHeight,
             units = "mm",
             device = input$downloadGOMEGENAFile)
    }
  )
  
  ####SCENIC####
  observeEvent(input$downloadheatmapSCENIC, {
    showModal(modalDialog(
      title = strong("Download Gene Regulatory Network Heatmap"),
      numericInput("downloadheatmapSCENICHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadheatmapSCENICWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadheatmapSCENICFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadheatmapSCENICOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadheatmapSCENICOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_RegulonHeatmap.", input$downloadheatmapSCENICFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadheatmapSCENICFile == "png") {
        
        png(filename = file, width = input$downloadheatmapSCENICWidth, height = input$downloadheatmapSCENICHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(Heatmap(seuratreactive$regulonActivity_byCellType_Scaled, name="Regulon activity",
                      column_names_gp = grid::gpar(fontsize = input$fontSCENIC),
                      row_names_gp = grid::gpar(fontsize = input$fontSCENIC)))
        
        dev.off()
      }
      
      else if (input$downloadheatmapSCENICFile == "jpeg") {
        jpeg(filename = file, width = input$downloadheatmapSCENICWidth, height = input$downloadheatmapSCENICHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(Heatmap(seuratreactive$regulonActivity_byCellType_Scaled, name="Regulon activity",
                      column_names_gp = grid::gpar(fontsize = input$fontSCENIC),
                      row_names_gp = grid::gpar(fontsize = input$fontSCENIC)))
        
        dev.off()
      }
      
      else if (input$downloadheatmapSCENICFile == "tiff") {
        tiff(filename = file, width = input$downloadheatmapSCENICWidth, height = input$downloadheatmapSCENICHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(Heatmap(seuratreactive$regulonActivity_byCellType_Scaled, name="Regulon activity",
                      column_names_gp = grid::gpar(fontsize = input$fontSCENIC),
                      row_names_gp = grid::gpar(fontsize = input$fontSCENIC)))
        
        dev.off()
      }
      
      else if (input$downloadheatmapSCENICFile == "bmp") {
        bmp(filename = file, width = input$downloadheatmapSCENICWidth, height = input$downloadheatmapSCENICHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(Heatmap(seuratreactive$regulonActivity_byCellType_Scaled, name="Regulon activity",
                      column_names_gp = grid::gpar(fontsize = input$fontSCENIC),
                      row_names_gp = grid::gpar(fontsize = input$fontSCENIC)))
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloaddotplotSCENIC, {
    showModal(modalDialog(
      title = strong("Download Gene Regulatory Network Dotplot"),
      numericInput("downloaddotplotSCENICHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloaddotplotSCENICWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloaddotplotSCENICFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloaddotplotSCENICOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloaddotplotSCENICOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_RegulonDotplot.", input$downloaddotplotSCENICFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$radioSCENICrss == "All") {
        
        rssPlot <- plotRSS(seuratreactive$rss)
        
        plot <- rssPlot$plot + 
          theme(axis.text.x = element_text(angle=90, hjust=1, vjust = 0.5),
                text = element_text(size=input$fontSCENIC2))
      }
      else {
        plot <- plotRSS_oneSet(seuratreactive$rss, setName = input$Regulon_Select2, n = length(rownames(seuratreactive$rss))-1) +
          theme(text = element_text(size=input$fontSCENIC2),
              axis.text.x= element_text(size = input$fontSCENIC2),
              axis.text.y= element_text(size = input$fontSCENIC2))
      }
      
      ggsave(file, plot = plot,
             width = input$downloaddotplotSCENICWidth,
             height = input$downloaddotplotSCENICHeight,
             units = "mm",
             device = input$downloaddotplotSCENICFile)
    }
  )
  
  observeEvent(input$downloadRegulonSCENIC, {
    showModal(modalDialog(
      title = strong("Download Regulon Transcription Factor Map"),
      numericInput("downloadRegulonSCENICHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadRegulonSCENICWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadRegulonSCENICFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadRegulonSCENICOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadRegulonSCENICOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_RegulonTFMap.", input$downloadRegulonSCENICFile, sep = "")
    }, 
    
    content = function(file) {
      
      onlyDuplicatedExtended <- function(regulonNames)
      {
        regulonNames <- unname(regulonNames)
        tfs <- getTF(regulonNames)
        tfs <- gsub("_extended", "", tfs)
        splitRegulons <- split(regulonNames, tfs)[unique(tfs)]
        
        ret <- sapply(splitRegulons, function(x) 
        {
          split(x, !grepl("_extended", x))[[1]] # False (direct) will be first
        })
        
        return(ret)
      }
      
      if (input$checkboxshowextended == TRUE) {
        regulons <- seuratreactive$regulons[seuratreactive$regulons$Regulon %in% onlyDuplicatedExtended(seuratreactive$regulons$Regulon),]
      }
      else if (input$checkboxshowextended == FALSE) {
        regulons <- seuratreactive$regulons[-grep("_extended", seuratreactive$regulons$Regulon), ]
      }
      
      regulons <- regulons[,c(1:input$SCENICNumber)]
      
      regulons <- reshape2::melt(regulons, id.vars = "Regulon") %>% na.omit()
      
      regulons$Regulon <- paste0(regulons$Regulon, "_Regulon")
      
      
      edges <- regulons[,c("Regulon", "value")]
      
      nodes1 <- regulons %>%
        group_by(`Regulon`) %>%
        summarise(node_color = "Regulon") %>%
        drop_na()
      
      nodes2 <- regulons %>%
        group_by(`value`) %>%
        summarise(node_color = "TFs") %>%
        drop_na()
      
      colnames(nodes2)[1] <- "Regulon"
      
      nodes <- full_join(nodes1, nodes2)
      
      graph = graph_from_data_frame(edges, directed = FALSE, nodes)
      
      if (input$checkboxgene_SCENIC == TRUE && input$checkboxregulon_SCENIC == TRUE) {
        plot <- ggraph(graph, layout = "fr") +
          geom_edge_link(aes(width = 1), color = "light blue") +
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "Regulon", input$nodeSCENIC, input$nodeSCENIC/2)) +
          geom_node_text(aes(label = ifelse(node_color == "Regulon", name, NA_character_)), size = input$categoryfontSCENIC, repel=TRUE) +
          geom_node_text(aes(label = ifelse(node_color == "TFs", name, NA_character_)), size = input$genefontSCENIC, repel=TRUE) +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() +
          guides(color = FALSE, edge_width = FALSE)
      }
      
      else if (input$checkboxgene_SCENIC == FALSE && input$checkboxregulon_SCENIC == TRUE) {
        plot <- ggraph(graph, layout = "fr") +
          geom_edge_link(aes(width = 1), color = "light blue") +
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "Regulon", input$nodeSCENIC, input$nodeSCENIC/2)) +
          geom_node_text(aes(label = ifelse(node_color == "Regulon", name, NA_character_)), size = input$categoryfontSCENIC, repel=TRUE) +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() +
          guides(color = FALSE, edge_width = FALSE)
      }
      
      else if (input$checkboxgene_SCENIC == TRUE && input$checkboxregulon_SCENIC == FALSE) {
        plot <- ggraph(graph, layout = "fr") +
          geom_edge_link(aes(width = 1), color = "light blue") +
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "Regulon", input$nodeSCENIC, input$nodeSCENIC/2)) +
          geom_node_text(aes(label = ifelse(node_color == "TFs", name, NA_character_)), size = input$genefontSCENIC, repel=TRUE) +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() +
          guides(color = FALSE, edge_width = FALSE)
      }
      
      ggsave(file, plot = plot,
             width = input$downloadRegulonSCENICWidth,
             height = input$downloadRegulonSCENICHeight,
             units = "mm",
             device = input$downloadRegulonSCENICFile)
    }
  )
  
  ####MAGMA####
  
  output$downloadtable_MAGMAExample <- downloadHandler(
    filename = function() {
      paste("MAGMA_GWAS_Example_DataInput.txt")
    }, 
    
    content = function(file) {
      GWAS_Example <- read.table("educational_attainment.tsv")
      write.table(GWAS_Example, file, row.names = F, col.names = F, quote = F, sep = "\t")
    }
  )
  
  output$downloadtable_MAGMAGWAS <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_MAGMAGWASTable.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$meta, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadMAGMAHeat, {
    showModal(modalDialog(
      title = strong("Download GWAS Cell Type Association Plot"),
      numericInput("downloadMAGMAHeatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadMAGMAHeatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadMAGMAHeatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadMAGMAHeatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadMAGMAHeatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_MAGMA_CellTypeAssociation.", input$downloadMAGMAHeatFile, sep = "")
    }, 
    
    content = function(file) {
      
      plot <- MAGMA.Celltyping::results_heatmap(
        merged_results = seuratreactive$MAGMA_Final_Results, 
        fdr_thresh = 1,
        y_var = "Trait",
        x_var = "Celltype_id",
        fill_var = "-log10(FDR)") + theme(text = element_text(size = input$fontMAGMA),
                                          axis.text.x = element_text(size = input$fontMAGMA, angle=90, vjust=0.5))
      
      ggsave(file, plot = plot,
             width = input$downloadMAGMAHeatWidth,
             height = input$downloadMAGMAHeatHeight,
             units = "mm",
             device = input$downloadMAGMAHeatFile)
    }
  )
  
  output$downloadtable_MAGMA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_MAGMA_CellTypeAssociationInfo.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$MAGMA_Final_Results[,c("Celltype_id", "GWAS", "Trait", "EnrichmentMode", "P", "log10p", "FDR")], file, row.names = F)
    }
  )
  
  
  ####GE######
  observeEvent(input$downloadGEViolin, {
    showModal(modalDialog(
      title = strong("Download Violin Plot"),
      numericInput("ViolinHeightGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("ViolinWidthGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ViolinFileInputGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("ViolindownloadoutputGE", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$ViolindownloadoutputGE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_ViolinPlot_GeneExpression.", input$ViolinFileInputGE, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$p,
             width = input$ViolinWidthGE,
             height = input$ViolinHeightGE,
             units = "mm",
             device = input$ViolinFileInputGE)
    }
  )  
  
  observeEvent(input$downloadGEDot, {
    showModal(modalDialog(
      title = strong("Download Dot Plot"),
      numericInput("DotHeightGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("DotWidthGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("DotFileInputGE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("DotdownloadoutputGE", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$DotdownloadoutputGE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DotPlot_GeneExpression.", input$DotFileInputGE, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$q,
             width = input$DotWidthGE,
             height = input$DotHeightGE,
             units = "mm",
             device = input$DotFileInputGE)
    }
  )
  
  ####TA#####
  observeEvent(input$downloadUMAPmonocle, {
    showModal(modalDialog(
      title = strong("Download Monocle3 trajectories"),
      numericInput("downloadUMAPmonocleHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadUMAPmonocleWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadUMAPmonocleformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadUMAPmonoclebutton", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadUMAPmonoclebutton <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Monocle3_trajectories.", input$downloadUMAPmonocleformat, sep = "")
    }, 
    
    content = function(file) {
      
      p <- plot_cells(
        cds = seuratreactive$seurat_monocle,
        color_cells_by = "pseudotime",
        reduction_method = "UMAP",
        cell_size = input$sizemonocleTA/4,
        graph_label_size = input$sizemonocleTA,
        alpha = input$opacitymonocleTA
      ) + theme(text = element_text(size=input$labelsizemonocleTA),
                axis.text.x= element_text(size = input$labelsizemonocleTA),
                axis.text.y= element_text(size = input$labelsizemonocleTA))
      
      ggsave(file, plot = p,
             width = input$downloadUMAPmonocleWidth,
             height = input$downloadUMAPmonocleHeight,
             units = "mm",
             device = input$downloadUMAPmonocleformat)
    }
  )
  
  observeEvent(input$downloadmonocle3viz, {
    showModal(modalDialog(
      title = strong("Download Box Plot"),
      numericInput("downloadmonocle3vizHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadmonocle3vizWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadmonocle3vizformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadmonocle3vizButton", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadmonocle3vizButton <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Monocle3_BoxPlot.", input$downloadmonocle3vizformat, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$monocle3pseudo == "Seurat Clusters") {
        seuratreactive$seurat_monocle$monocle3_pseudotime <- pseudotime(seuratreactive$seurat_monocle)
        data.pseudo <- as.data.frame(colData(seuratreactive$seurat_monocle))
        
        p <- ggplot(data.pseudo, aes(monocle3_pseudotime, reorder(seurat_clusters, monocle3_pseudotime), fill = seurat_clusters)) + 
          geom_boxplot() + 
          ylab("Seurat Clusters") +
          xlab("Monocle3 Pseudotime") +
          theme_bw(base_size = input$fontsizepseudotimeviz)
      }
      else {
        seuratreactive$seurat_monocle$monocle3_pseudotime <- pseudotime(seuratreactive$seurat_monocle)
        data.pseudo <- as.data.frame(colData(seuratreactive$seurat_monocle))
        
        p <- ggplot(data.pseudo, aes(monocle3_pseudotime, reorder(ident, monocle3_pseudotime), fill = ident)) + 
          geom_boxplot() + 
          ylab("Labelled Cells") + 
          xlab("Monocle3 Pseudotime") +
          theme_bw(base_size = input$fontsizepseudotimeviz)
      }
      
      ggsave(file, plot = p,
             width = input$downloadmonocle3vizWidth,
             height = input$downloadmonocle3vizHeight,
             units = "mm",
             device = input$downloadmonocle3vizformat)
    }
  )
  
  output$downloadmonocle3DEG <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Monocle3_trajectory_DEGs.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$Monocle3_genes, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadmonocleGAP, {
    showModal(modalDialog(
      title = strong("Download Gene Expression Across Pseudotime Plot"),
      numericInput("downloadmonocleGAPHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadmonocleGAPWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadmonocleGAPformat", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadmonocleGAPButton", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadmonocleGAPButton <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Monocle3_GenesAcrossPseudotime.", input$downloadmonocleGAPformat, sep = "")
    }, 
    
    content = function(file) {
      
      fData(seuratreactive$seurat_monocle)$gene_short_name <- rownames(fData(seuratreactive$seurat_monocle))
      
      my_genes <- row.names(subset(fData(seuratreactive$seurat_monocle), gene_short_name %in% input$MonocleInput)) 
      cds_subset <- seuratreactive$seurat_monocle[my_genes,]
      p <- plot_genes_in_pseudotime(cds_subset, color_cells_by = "monocle3_pseudotime") + theme_bw(base_size = input$fontsizemonocle3select) + guides(colour = FALSE)
      
      ggsave(file, plot = p,
             width = input$downloadmonocleGAPWidth,
             height = input$downloadmonocleGAPHeight,
             units = "mm",
             device = input$downloadmonocleGAPformat)
    }
  )
  
  ####Chat######
  observeEvent(input$downloadCommunicationChat, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Communication Network Plot"),
      numericInput("downloadCommunicationChatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadCommunicationChatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadCommunicationChatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadCommunicationChatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadCommunicationChatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellCommunicationNetworkPlot.", input$downloadCommunicationChatFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadCommunicationChatFile == "png") {
        
        png(filename = file, width = input$downloadCommunicationChatWidth, height = input$downloadCommunicationChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunication == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_circle(seuratreactive$cellchat@net$count, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Number of interactions")
          netVisual_circle(seuratreactive$cellchat@net$weight, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$count), ncol = ncol(seuratreactive$cellchat@net$count), dimnames = dimnames(seuratreactive$cellchat@net$count))
          mat1[input$Communication_Select, ] <- seuratreactive$cellchat@net$count[input$Communication_Select, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$count), title.name = paste(input$Communication_Select, "Number of interactions"))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$weight), ncol = ncol(seuratreactive$cellchat@net$weight), dimnames = dimnames(seuratreactive$cellchat@net$weight))
          mat2[input$Communication_Select, ] <- seuratreactive$cellchat@net$weight[input$Communication_Select, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$weight), title.name = paste(input$Communication_Select, "Interaction weights/strength"))
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatFile == "jpeg") {
        jpeg(filename = file, width = input$downloadCommunicationChatWidth, height = input$downloadCommunicationChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunication == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_circle(seuratreactive$cellchat@net$count, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Number of interactions")
          netVisual_circle(seuratreactive$cellchat@net$weight, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$count), ncol = ncol(seuratreactive$cellchat@net$count), dimnames = dimnames(seuratreactive$cellchat@net$count))
          mat1[input$Communication_Select, ] <- seuratreactive$cellchat@net$count[input$Communication_Select, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$count), title.name = paste(input$Communication_Select, "Number of interactions"))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$weight), ncol = ncol(seuratreactive$cellchat@net$weight), dimnames = dimnames(seuratreactive$cellchat@net$weight))
          mat2[input$Communication_Select, ] <- seuratreactive$cellchat@net$weight[input$Communication_Select, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$weight), title.name = paste(input$Communication_Select, "Interaction weights/strength"))
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatFile == "tiff") {
        tiff(filename = file, width = input$downloadCommunicationChatWidth, height = input$downloadCommunicationChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunication == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_circle(seuratreactive$cellchat@net$count, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Number of interactions")
          netVisual_circle(seuratreactive$cellchat@net$weight, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$count), ncol = ncol(seuratreactive$cellchat@net$count), dimnames = dimnames(seuratreactive$cellchat@net$count))
          mat1[input$Communication_Select, ] <- seuratreactive$cellchat@net$count[input$Communication_Select, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$count), title.name = paste(input$Communication_Select, "Number of interactions"))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$weight), ncol = ncol(seuratreactive$cellchat@net$weight), dimnames = dimnames(seuratreactive$cellchat@net$weight))
          mat2[input$Communication_Select, ] <- seuratreactive$cellchat@net$weight[input$Communication_Select, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$weight), title.name = paste(input$Communication_Select, "Interaction weights/strength"))
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatFile == "bmp") {
        bmp(filename = file, width = input$downloadCommunicationChatWidth, height = input$downloadCommunicationChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunication == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_circle(seuratreactive$cellchat@net$count, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Number of interactions")
          netVisual_circle(seuratreactive$cellchat@net$weight, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$count), ncol = ncol(seuratreactive$cellchat@net$count), dimnames = dimnames(seuratreactive$cellchat@net$count))
          mat1[input$Communication_Select, ] <- seuratreactive$cellchat@net$count[input$Communication_Select, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$count), title.name = paste(input$Communication_Select, "Number of interactions"))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$weight), ncol = ncol(seuratreactive$cellchat@net$weight), dimnames = dimnames(seuratreactive$cellchat@net$weight))
          mat2[input$Communication_Select, ] <- seuratreactive$cellchat@net$weight[input$Communication_Select, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$weight), title.name = paste(input$Communication_Select, "Interaction weights/strength"))
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadSignalChat, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Signalling Pathways Plot"),
      numericInput("downloadSignalChatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadSignalChatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadSignalChatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadSignalChatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadSignalChatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellSignallingPathwaysPlot.", input$downloadSignalChatFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadSignalChatFile == "png") {
        
        png(filename = file, width = input$downloadSignalChatWidth, height = input$downloadSignalChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioChat == "Hierarchy Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select,  vertex.receiver = seq(1,length(unique(seuratreactive$cellchat@idents))/2), layout = "hierarchy", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Circle Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "circle", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Chord Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "chord", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Heatmap") {
          par(mfrow=c(1,1))
          print(netVisual_heatmap(seuratreactive$cellchat, signaling = input$Signalling_Select, color.heatmap = "Reds", font.size = input$sizeSignalling, font.size.title = input$sizeSignalling + 2))
        }
        else if (input$radioChat == "Signalling Roles") {
          par(mfrow=c(1,1))
          print(netAnalysis_signalingRole_network(seuratreactive$cellchat, signaling = input$Signalling_Select, width = input$WidthSignalling_Chat, height = input$HeightSignalling_Chat, font.size = input$sizeSignalling))
        }
        
        dev.off()
      }
      
      else if (input$downloadSignalChatFile == "jpeg") {
        jpeg(filename = file, width = input$downloadSignalChatWidth, height = input$downloadSignalChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioChat == "Hierarchy Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select,  vertex.receiver = seq(1,length(unique(seuratreactive$cellchat@idents))/2), layout = "hierarchy", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Circle Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "circle", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Chord Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "chord", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Heatmap") {
          par(mfrow=c(1,1))
          print(netVisual_heatmap(seuratreactive$cellchat, signaling = input$Signalling_Select, color.heatmap = "Reds", font.size = input$sizeSignalling, font.size.title = input$sizeSignalling + 2))
        }
        else if (input$radioChat == "Signalling Roles") {
          par(mfrow=c(1,1))
          print(netAnalysis_signalingRole_network(seuratreactive$cellchat, signaling = input$Signalling_Select, width = input$WidthSignalling_Chat, height = input$HeightSignalling_Chat, font.size = input$sizeSignalling))
        }
        
        dev.off()
      }
      
      else if (input$downloadSignalChatFile == "tiff") {
        tiff(filename = file, width = input$downloadSignalChatWidth, height = input$downloadSignalChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioChat == "Hierarchy Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select,  vertex.receiver = seq(1,length(unique(seuratreactive$cellchat@idents))/2), layout = "hierarchy", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Circle Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "circle", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Chord Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "chord", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Heatmap") {
          par(mfrow=c(1,1))
          print(netVisual_heatmap(seuratreactive$cellchat, signaling = input$Signalling_Select, color.heatmap = "Reds", font.size = input$sizeSignalling, font.size.title = input$sizeSignalling + 2))
        }
        else if (input$radioChat == "Signalling Roles") {
          par(mfrow=c(1,1))
          print(netAnalysis_signalingRole_network(seuratreactive$cellchat, signaling = input$Signalling_Select, width = input$WidthSignalling_Chat, height = input$HeightSignalling_Chat, font.size = input$sizeSignalling))
        }
        
        dev.off()
      }
      
      else if (input$downloadSignalChatFile == "bmp") {
        bmp(filename = file, width = input$downloadSignalChatWidth, height = input$downloadSignalChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioChat == "Hierarchy Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select,  vertex.receiver = seq(1,length(unique(seuratreactive$cellchat@idents))/2), layout = "hierarchy", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Circle Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "circle", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Chord Plot") {
          par(mfrow=c(1,1))
          print(netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "chord", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10))
        }
        else if (input$radioChat == "Heatmap") {
          par(mfrow=c(1,1))
          print(netVisual_heatmap(seuratreactive$cellchat, signaling = input$Signalling_Select, color.heatmap = "Reds", font.size = input$sizeSignalling, font.size.title = input$sizeSignalling + 2))
        }
        else if (input$radioChat == "Signalling Roles") {
          par(mfrow=c(1,1))
          print(netAnalysis_signalingRole_network(seuratreactive$cellchat, signaling = input$Signalling_Select, width = input$WidthSignalling_Chat, height = input$HeightSignalling_Chat, font.size = input$sizeSignalling))
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadLigandChat, {
    showModal(modalDialog(
      title = strong("Download Ligand Receptor Interactions Plot"),
      numericInput("downloadLigandChatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadLigandChatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadLigandChatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadLigandChatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadLigandChatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_LigandReceptorInteractionsPlot.", input$downloadLigandChatFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadLigandChatFile == "png") {
        
        png(filename = file, width = input$downloadLigandChatWidth, height = input$downloadLigandChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_Select)) {
          if (input$radioLigandChat == "Bar Graph") {
            par(mfrow=c(1,1))
            print(netAnalysis_contribution(seuratreactive$cellchat, signaling = input$Ligand_Select, font.size = input$sizeLigand, font.size.title = input$sizeLigand))
          }
          else if (input$radioLigandChat == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchat, sources.use = input$source_select, targets.use = input$target_select, signaling = input$Ligand_Select, remove.isolate = FALSE, font.size = input$sizeLigand, font.size.title = input$sizeLigand, angle.x = 90))
          }
          else if (input$radioLigandChat == "Chord Plot") {
            par(mfrow=c(1,1))
            print(netVisual_chord_gene(seuratreactive$cellchat, signaling = input$Ligand_Select, lab.cex = input$sizeLigand/10))
          }
          else if (input$radioLigandChat == "Gene Expression") {
            par(mfrow=c(1,1))
            print(plotGeneExpression(seuratreactive$cellchat, signaling = input$Ligand_Select))
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadLigandChatFile == "jpeg") {
        jpeg(filename = file, width = input$downloadLigandChatWidth, height = input$downloadLigandChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_Select)) {
          if (input$radioLigandChat == "Bar Graph") {
            par(mfrow=c(1,1))
            print(netAnalysis_contribution(seuratreactive$cellchat, signaling = input$Ligand_Select, font.size = input$sizeLigand, font.size.title = input$sizeLigand))
          }
          else if (input$radioLigandChat == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchat, sources.use = input$source_select, targets.use = input$target_select, signaling = input$Ligand_Select, remove.isolate = FALSE, font.size = input$sizeLigand, font.size.title = input$sizeLigand, angle.x = 90))
          }
          else if (input$radioLigandChat == "Chord Plot") {
            par(mfrow=c(1,1))
            print(netVisual_chord_gene(seuratreactive$cellchat, signaling = input$Ligand_Select, lab.cex = input$sizeLigand/10))
          }
          else if (input$radioLigandChat == "Gene Expression") {
            par(mfrow=c(1,1))
            print(plotGeneExpression(seuratreactive$cellchat, signaling = input$Ligand_Select))
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadLigandChatFile == "tiff") {
        tiff(filename = file, width = input$downloadLigandChatWidth, height = input$downloadLigandChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_Select)) {
          if (input$radioLigandChat == "Bar Graph") {
            par(mfrow=c(1,1))
            print(netAnalysis_contribution(seuratreactive$cellchat, signaling = input$Ligand_Select, font.size = input$sizeLigand, font.size.title = input$sizeLigand))
          }
          else if (input$radioLigandChat == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchat, sources.use = input$source_select, targets.use = input$target_select, signaling = input$Ligand_Select, remove.isolate = FALSE, font.size = input$sizeLigand, font.size.title = input$sizeLigand, angle.x = 90))
          }
          else if (input$radioLigandChat == "Chord Plot") {
            par(mfrow=c(1,1))
            print(netVisual_chord_gene(seuratreactive$cellchat, signaling = input$Ligand_Select, lab.cex = input$sizeLigand/10))
          }
          else if (input$radioLigandChat == "Gene Expression") {
            par(mfrow=c(1,1))
            print(plotGeneExpression(seuratreactive$cellchat, signaling = input$Ligand_Select))
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadLigandChatFile == "bmp") {
        bmp(filename = file, width = input$downloadLigandChatWidth, height = input$downloadLigandChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_Select)) {
          if (input$radioLigandChat == "Bar Graph") {
            par(mfrow=c(1,1))
            print(netAnalysis_contribution(seuratreactive$cellchat, signaling = input$Ligand_Select, font.size = input$sizeLigand, font.size.title = input$sizeLigand))
          }
          else if (input$radioLigandChat == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchat, sources.use = input$source_select, targets.use = input$target_select, signaling = input$Ligand_Select, remove.isolate = FALSE, font.size = input$sizeLigand, font.size.title = input$sizeLigand, angle.x = 90))
          }
          else if (input$radioLigandChat == "Chord Plot") {
            par(mfrow=c(1,1))
            print(netVisual_chord_gene(seuratreactive$cellchat, signaling = input$Ligand_Select, lab.cex = input$sizeLigand/10))
          }
          else if (input$radioLigandChat == "Gene Expression") {
            par(mfrow=c(1,1))
            print(plotGeneExpression(seuratreactive$cellchat, signaling = input$Ligand_Select))
          }
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadRolesChat, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Signalling Roles Plot"),
      numericInput("downloadRolesChatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadRolesChatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadRolesChatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadRolesChatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadRolesChatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellSignallingRolesPlot.", input$downloadRolesChatFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadRolesChatFile == "png") {
        
        png(filename = file, width = input$downloadRolesChatWidth, height = input$downloadRolesChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChat == "Heatmap") {
          ht1 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          ht2 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          print(ht1 + ht2)
        }
        else if (input$radioRolesChat == "Scatter Plot") {
          print(netAnalysis_signalingRole_scatter(seuratreactive$cellchat, font.size = input$sizeRoles, font.size.title = input$sizeRoles, label.size	= input$sizeRoles - 8))
        }
        
        dev.off()
      }
      
      else if (input$downloadRolesChatFile == "jpeg") {
        jpeg(filename = file, width = input$downloadRolesChatWidth, height = input$downloadRolesChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChat == "Heatmap") {
          ht1 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          ht2 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          print(ht1 + ht2)
        }
        else if (input$radioRolesChat == "Scatter Plot") {
          print(netAnalysis_signalingRole_scatter(seuratreactive$cellchat, font.size = input$sizeRoles, font.size.title = input$sizeRoles, label.size	= input$sizeRoles - 8))
        }
        
        dev.off()
      }
      
      else if (input$downloadRolesChatFile == "tiff") {
        tiff(filename = file, width = input$downloadRolesChatWidth, height = input$downloadRolesChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChat == "Heatmap") {
          ht1 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          ht2 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          print(ht1 + ht2)
        }
        else if (input$radioRolesChat == "Scatter Plot") {
          print(netAnalysis_signalingRole_scatter(seuratreactive$cellchat, font.size = input$sizeRoles, font.size.title = input$sizeRoles, label.size	= input$sizeRoles - 8))
        }
        
        dev.off()
      }
      
      else if (input$downloadRolesChatFile == "bmp") {
        bmp(filename = file, width = input$downloadRolesChatWidth, height = input$downloadRolesChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChat == "Heatmap") {
          ht1 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          ht2 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
          print(ht1 + ht2)
        }
        else if (input$radioRolesChat == "Scatter Plot") {
          print(netAnalysis_signalingRole_scatter(seuratreactive$cellchat, font.size = input$sizeRoles, font.size.title = input$sizeRoles, label.size	= input$sizeRoles - 8))
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadPatternsChat, {
    showModal(modalDialog(
      title = strong("Download Incoming Cell-Cell Signalling Patterns Plot"),
      numericInput("downloadPatternsChatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadPatternsChatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadPatternsChatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadPatternsChatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadPatternsChatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_IncomingCell-CellSignallingPatternsPlot.", input$downloadPatternsChatFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadPatternsChatFile == "png") {
        
        png(filename = file, width = input$downloadPatternsChatWidth, height = input$downloadPatternsChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
      
      else if (input$downloadPatternsChatFile == "jpeg") {
        jpeg(filename = file, width = input$downloadPatternsChatWidth, height = input$downloadPatternsChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
      
      else if (input$downloadPatternsChatFile == "tiff") {
        tiff(filename = file, width = input$downloadPatternsChatWidth, height = input$downloadPatternsChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
      
      else if (input$downloadPatternsChatFile == "bmp") {
        bmp(filename = file, width = input$downloadPatternsChatWidth, height = input$downloadPatternsChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadPatternsChat2, {
    showModal(modalDialog(
      title = strong("Download Outgoing Cell-Cell Signalling Patterns Plot"),
      numericInput("downloadPatternsChat2Height", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadPatternsChat2Width", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadPatternsChat2File", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadPatternsChat2Output", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadPatternsChat2Output <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_OutgoingCell-CellSignallingPatternsPlot.", input$downloadPatternsChat2File, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadPatternsChat2File == "png") {
        
        png(filename = file, width = input$downloadPatternsChat2Width, height = input$downloadPatternsChat2Height,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
      
      else if (input$downloadPatternsChat2File == "jpeg") {
        jpeg(filename = file, width = input$downloadPatternsChat2Width, height = input$downloadPatternsChat2Height,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
      
      else if (input$downloadPatternsChat2File == "tiff") {
        tiff(filename = file, width = input$downloadPatternsChat2Width, height = input$downloadPatternsChat2Height,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
      
      else if (input$downloadPatternsChat2File == "bmp") {
        bmp(filename = file, width = input$downloadPatternsChat2Width, height = input$downloadPatternsChat2Height,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioPatternsChat == "River Plot") {
          print(netAnalysis_river(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2))
        }
        else if (input$radioPatternsChat == "Dot Plot") {
          print(netAnalysis_dot(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2))
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadManifoldChat, {
    showModal(modalDialog(
      title = strong("Download Manifold and Classification Learning Analysis of Signaling Networks Plot"),
      numericInput("downloadManifoldChatHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadManifoldChatWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadManifoldChatFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadManifoldChatOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadManifoldChatOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_ManifoldLearningPlot.", input$downloadManifoldChatFile, sep = "")
    }, 
    
    content = function(file) {
      
      if (input$downloadManifoldChatFile == "png") {
        
        png(filename = file, width = input$downloadManifoldChatWidth, height = input$downloadManifoldChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChat == "Functional Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "functional", label.size = input$sizeManifold))
        }
        else if (input$radioManifoldChat == "Structural Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "structural", label.size = input$sizeManifold))
        }
        
        dev.off()
      }
      
      else if (input$downloadManifoldChatFile == "jpeg") {
        jpeg(filename = file, width = input$downloadManifoldChatWidth, height = input$downloadManifoldChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChat == "Functional Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "functional", label.size = input$sizeManifold))
        }
        else if (input$radioManifoldChat == "Structural Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "structural", label.size = input$sizeManifold))
        }
        
        dev.off()
      }
      
      else if (input$downloadManifoldChatFile == "tiff") {
        tiff(filename = file, width = input$downloadManifoldChatWidth, height = input$downloadManifoldChatHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChat == "Functional Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "functional", label.size = input$sizeManifold))
        }
        else if (input$radioManifoldChat == "Structural Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "structural", label.size = input$sizeManifold))
        }
        
        dev.off()
      }
      
      else if (input$downloadManifoldChatFile == "bmp") {
        bmp(filename = file, width = input$downloadManifoldChatWidth, height = input$downloadManifoldChatHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChat == "Functional Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "functional", label.size = input$sizeManifold))
        }
        else if (input$radioManifoldChat == "Structural Classification") {
          print(netVisual_embedding(seuratreactive$cellchat, type = "structural", label.size = input$sizeManifold))
        }
        
        dev.off()
      }
    }
  )
  ####CellChat2
  
  observeEvent(input$downloadCommunicationChatM, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Communication Network Plot (Number of Interactions)"),
      numericInput("downloadCommunicationChatMHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadCommunicationChatMWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadCommunicationChatMFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadCommunicationChatMOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadCommunicationChatMOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellCommunicationNetworkPlot_NumberofInteractions.", input$downloadCommunicationChatMFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadCommunicationChatMFile == "png") {
        
        png(filename = file, width = input$downloadCommunicationChatMWidth, height = input$downloadCommunicationChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$count, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$count), edge.width.max = 12, title.name = paste0("Number of interactions - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$count), ncol = ncol(seuratreactive$object.list[[1]]@net$count), dimnames = dimnames(seuratreactive$object.list[[1]]@net$count))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$count[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$count), title.name = paste(input$Communication_SelectM1, "Number of interactions - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$count), ncol = ncol(seuratreactive$object.list[[2]]@net$count), dimnames = dimnames(seuratreactive$object.list[[2]]@net$count))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$count[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$count), title.name = paste(input$Communication_SelectM2, "Number of interactions - ", names(seuratreactive$object.list)[2]))
          
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatMFile == "jpeg") {
        jpeg(filename = file, width = input$downloadCommunicationChatMWidth, height = input$downloadCommunicationChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$count, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$count), edge.width.max = 12, title.name = paste0("Number of interactions - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$count), ncol = ncol(seuratreactive$object.list[[1]]@net$count), dimnames = dimnames(seuratreactive$object.list[[1]]@net$count))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$count[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$count), title.name = paste(input$Communication_SelectM1, "Number of interactions - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$count), ncol = ncol(seuratreactive$object.list[[2]]@net$count), dimnames = dimnames(seuratreactive$object.list[[2]]@net$count))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$count[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$count), title.name = paste(input$Communication_SelectM2, "Number of interactions - ", names(seuratreactive$object.list)[2]))
          
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatMFile == "tiff") {
        tiff(filename = file, width = input$downloadCommunicationChatMWidth, height = input$downloadCommunicationChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$count, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$count), edge.width.max = 12, title.name = paste0("Number of interactions - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$count), ncol = ncol(seuratreactive$object.list[[1]]@net$count), dimnames = dimnames(seuratreactive$object.list[[1]]@net$count))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$count[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$count), title.name = paste(input$Communication_SelectM1, "Number of interactions - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$count), ncol = ncol(seuratreactive$object.list[[2]]@net$count), dimnames = dimnames(seuratreactive$object.list[[2]]@net$count))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$count[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$count), title.name = paste(input$Communication_SelectM2, "Number of interactions - ", names(seuratreactive$object.list)[2]))
          
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatMFile == "bmp") {
        bmp(filename = file, width = input$downloadCommunicationChatMWidth, height = input$downloadCommunicationChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$count, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$count), edge.width.max = 12, title.name = paste0("Number of interactions - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$count), ncol = ncol(seuratreactive$object.list[[1]]@net$count), dimnames = dimnames(seuratreactive$object.list[[1]]@net$count))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$count[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$count), title.name = paste(input$Communication_SelectM1, "Number of interactions - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$count), ncol = ncol(seuratreactive$object.list[[2]]@net$count), dimnames = dimnames(seuratreactive$object.list[[2]]@net$count))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$count[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$count), title.name = paste(input$Communication_SelectM2, "Number of interactions - ", names(seuratreactive$object.list)[2]))
          
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadCommunicationChatM2, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Communication Network Plot (Strength of Interactions)"),
      numericInput("downloadCommunicationChatM2Height", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadCommunicationChatM2Width", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadCommunicationChatM2File", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadCommunicationChatM2Output", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadCommunicationChatM2Output <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellCommunicationNetworkPlot_StrengthofInteractions.", input$downloadCommunicationChatM2File, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadCommunicationChatM2File == "png") {
        
        png(filename = file, width = input$downloadCommunicationChatM2Width, height = input$downloadCommunicationChatM2Height,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$weight, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$weight), edge.width.max = 12, title.name = paste0("Interaction weights/strength - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$weight), ncol = ncol(seuratreactive$object.list[[1]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[1]]@net$weight))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$weight[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$weight), title.name = paste(input$Communication_SelectM1, "Interaction weights/strength - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$weight), ncol = ncol(seuratreactive$object.list[[2]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[2]]@net$weight))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$weight[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$weight), title.name = paste(input$Communication_SelectM2, "Interaction weights/strength - ", names(seuratreactive$object.list)[2]))
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatM2File == "jpeg") {
        jpeg(filename = file, width = input$downloadCommunicationChatM2Width, height = input$downloadCommunicationChatM2Height,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$weight, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$weight), edge.width.max = 12, title.name = paste0("Interaction weights/strength - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$weight), ncol = ncol(seuratreactive$object.list[[1]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[1]]@net$weight))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$weight[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$weight), title.name = paste(input$Communication_SelectM1, "Interaction weights/strength - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$weight), ncol = ncol(seuratreactive$object.list[[2]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[2]]@net$weight))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$weight[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$weight), title.name = paste(input$Communication_SelectM2, "Interaction weights/strength - ", names(seuratreactive$object.list)[2]))
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatM2File == "tiff") {
        tiff(filename = file, width = input$downloadCommunicationChatM2Width, height = input$downloadCommunicationChatM2Height,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$weight, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$weight), edge.width.max = 12, title.name = paste0("Interaction weights/strength - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$weight), ncol = ncol(seuratreactive$object.list[[1]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[1]]@net$weight))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$weight[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$weight), title.name = paste(input$Communication_SelectM1, "Interaction weights/strength - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$weight), ncol = ncol(seuratreactive$object.list[[2]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[2]]@net$weight))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$weight[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$weight), title.name = paste(input$Communication_SelectM2, "Interaction weights/strength - ", names(seuratreactive$object.list)[2]))
        }
        
        dev.off()
      }
      
      else if (input$downloadCommunicationChatM2File == "bmp") {
        bmp(filename = file, width = input$downloadCommunicationChatM2Width, height = input$downloadCommunicationChatM2Height,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$checkboxCommunicationM == TRUE) {
          par(mfrow = c(1,2), xpd=TRUE)
          for (i in 1:length(seuratreactive$object.list)) {
            netVisual_circle(seuratreactive$object.list[[i]]@net$weight, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$weight), edge.width.max = 12, title.name = paste0("Interaction weights/strength - ", names(seuratreactive$object.list)[i]))
          }
        }
        else {
          par(mfrow = c(1,2), xpd=TRUE)
          mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$weight), ncol = ncol(seuratreactive$object.list[[1]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[1]]@net$weight))
          mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$weight[input$Communication_SelectM1, ]
          netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$weight), title.name = paste(input$Communication_SelectM1, "Interaction weights/strength - ", names(seuratreactive$object.list)[1]))
          
          mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$weight), ncol = ncol(seuratreactive$object.list[[2]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[2]]@net$weight))
          mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$weight[input$Communication_SelectM2, ]
          netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$weight), title.name = paste(input$Communication_SelectM2, "Interaction weights/strength - ", names(seuratreactive$object.list)[2]))
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadCellCell, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Differential Interactions Between Datasets"),
      numericInput("downloadCellCellHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadCellCellWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadCellCellFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadCellCellOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadCellCellOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DifferentialInteractionBetweenDatasets.", input$downloadCellCellFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadCellCellFile == "png") {
        
        png(filename = file, width = input$downloadCellCellWidth, height = input$downloadCellCellHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioCellCell == "Communication Network") {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T, measure = "weight")
        }
        else if (input$radioCellCell == "Bar Graph") {
          gg1 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), size.text = input$sizeCellCell)
          gg2 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), measure = "weight", size.text = input$sizeCellCell)
          print(gg1 + gg2)
        }
        else if (input$radioCellCell == "Heatmap") {
          gg1 <- netVisual_heatmap(seuratreactive$cellchatM, font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          gg2 <- netVisual_heatmap(seuratreactive$cellchatM, measure = "weight", font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          print(gg1 + gg2)
        }
        
        dev.off()
      }
      
      else if (input$downloadCellCellFile == "jpeg") {
        jpeg(filename = file, width = input$downloadCellCellWidth, height = input$downloadCellCellHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioCellCell == "Communication Network") {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T, measure = "weight")
        }
        else if (input$radioCellCell == "Bar Graph") {
          gg1 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), size.text = input$sizeCellCell)
          gg2 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), measure = "weight", size.text = input$sizeCellCell)
          print(gg1 + gg2)
        }
        else if (input$radioCellCell == "Heatmap") {
          gg1 <- netVisual_heatmap(seuratreactive$cellchatM, font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          gg2 <- netVisual_heatmap(seuratreactive$cellchatM, measure = "weight", font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          print(gg1 + gg2)
        }
        
        dev.off()
      }
      
      else if (input$downloadCellCellFile == "tiff") {
        tiff(filename = file, width = input$downloadCellCellWidth, height = input$downloadCellCellHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioCellCell == "Communication Network") {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T, measure = "weight")
        }
        else if (input$radioCellCell == "Bar Graph") {
          gg1 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), size.text = input$sizeCellCell)
          gg2 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), measure = "weight", size.text = input$sizeCellCell)
          print(gg1 + gg2)
        }
        else if (input$radioCellCell == "Heatmap") {
          gg1 <- netVisual_heatmap(seuratreactive$cellchatM, font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          gg2 <- netVisual_heatmap(seuratreactive$cellchatM, measure = "weight", font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          print(gg1 + gg2)
        }
        
        dev.off()
      }
      
      else if (input$downloadCellCellFile == "bmp") {
        bmp(filename = file, width = input$downloadCellCellWidth, height = input$downloadCellCellHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioCellCell == "Communication Network") {
          par(mfrow = c(1,2), xpd=TRUE)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T)
          netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T, measure = "weight")
          
        }
        else if (input$radioCellCell == "Bar Graph") {
          gg1 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), size.text = input$sizeCellCell)
          gg2 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), measure = "weight", size.text = input$sizeCellCell)
          gg1 + gg2
        }
        else if (input$radioCellCell == "Heatmap") {
          gg1 <- netVisual_heatmap(seuratreactive$cellchatM, font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          gg2 <- netVisual_heatmap(seuratreactive$cellchatM, measure = "weight", font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
          pgg1 + gg2
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadSignalChatM, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Signalling Pathways Plot"),
      numericInput("downloadSignalChatMHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadSignalChatMWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadSignalChatMFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadSignalChatMOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadSignalChatMOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellSignallingPathwaysPlot.", input$downloadSignalChatMFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadSignalChatMFile == "png") {
        
        png(filename = file, width = input$downloadSignalChatMWidth, height = input$downloadSignalChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Signalling_SelectM)) {
          
          if (input$SignallingDataset_SelectM == "Dataset 1") {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[1]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
          else {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[2]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadSignalChatMFile == "jpeg") {
        jpeg(filename = file, width = input$downloadSignalChatMWidth, height = input$downloadSignalChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Signalling_SelectM)) {
          
          if (input$SignallingDataset_SelectM == "Dataset 1") {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[1]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
          else {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[2]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadSignalChatMFile == "tiff") {
        tiff(filename = file, width = input$downloadSignalChatMWidth, height = input$downloadSignalChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Signalling_SelectM)) {
          
          if (input$SignallingDataset_SelectM == "Dataset 1") {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[1]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
          else {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[2]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadSignalChatMFile == "bmp") {
        bmp(filename = file, width = input$downloadSignalChatMWidth, height = input$downloadSignalChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        if (!is.null(input$Signalling_SelectM)) {
          
          if (input$SignallingDataset_SelectM == "Dataset 1") {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[1]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
          else {
            if (input$radioChatM == "Hierarchy Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[2]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Circle Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Chord Plot") {
              par(mfrow=c(1,1))
              print(netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10))
            }
            else if (input$radioChatM == "Heatmap") {
              par(mfrow=c(1,1))
              print(netVisual_heatmap(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2))
            }
            else if (input$radioChatM == "Signalling Roles") {
              par(mfrow=c(1,1))
              print(netAnalysis_signalingRole_network(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM))
            }
          }
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadLigandChatM, {
    showModal(modalDialog(
      title = strong("Download Ligand Receptor Interactions Plot"),
      numericInput("downloadLigandChatMHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadLigandChatMWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadLigandChatMFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadLigandChatMOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadLigandChatMOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_LigandReceptorInteractionsPlot.", input$downloadLigandChatMFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadLigandChatMFile == "png") {
        
        png(filename = file, width = input$downloadLigandChatMWidth, height = input$downloadLigandChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_SelectM)) {
          
          if (input$radioLigandChatM == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectM, targets.use = input$target_selectM, comparison = c(1, 2), signaling = input$Ligand_SelectM, remove.isolate = F, font.size = input$sizeLigandM, font.size.title = input$sizeLigandM, angle.x = 90))
          }
          else if (input$radioLigandChatM == "Gene Expression") {
            seuratreactive$cellchatM@meta$datasets = factor(seuratreactive$cellchatM@meta$datasets, levels = c(names(seuratreactive$object.list)[1], names(seuratreactive$object.list)[2])) # set factor level
            print(plotGeneExpression(seuratreactive$cellchatM, signaling = input$Ligand_SelectM, split.by = "datasets", colors.ggplot = T))
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMFile == "jpeg") {
        jpeg(filename = file, width = input$downloadLigandChatMWidth, height = input$downloadLigandChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_SelectM)) {
          
          if (input$radioLigandChatM == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectM, targets.use = input$target_selectM, comparison = c(1, 2), signaling = input$Ligand_SelectM, remove.isolate = F, font.size = input$sizeLigandM, font.size.title = input$sizeLigandM, angle.x = 90))
          }
          else if (input$radioLigandChatM == "Gene Expression") {
            seuratreactive$cellchatM@meta$datasets = factor(seuratreactive$cellchatM@meta$datasets, levels = c(names(seuratreactive$object.list)[1], names(seuratreactive$object.list)[2])) # set factor level
            print(plotGeneExpression(seuratreactive$cellchatM, signaling = input$Ligand_SelectM, split.by = "datasets", colors.ggplot = T))
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMFile == "tiff") {
        tiff(filename = file, width = input$downloadLigandChatMWidth, height = input$downloadLigandChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_SelectM)) {
          
          if (input$radioLigandChatM == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectM, targets.use = input$target_selectM, comparison = c(1, 2), signaling = input$Ligand_SelectM, remove.isolate = F, font.size = input$sizeLigandM, font.size.title = input$sizeLigandM, angle.x = 90))
          }
          else if (input$radioLigandChatM == "Gene Expression") {
            seuratreactive$cellchatM@meta$datasets = factor(seuratreactive$cellchatM@meta$datasets, levels = c(names(seuratreactive$object.list)[1], names(seuratreactive$object.list)[2])) # set factor level
            print(plotGeneExpression(seuratreactive$cellchatM, signaling = input$Ligand_SelectM, split.by = "datasets", colors.ggplot = T))
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMFile == "bmp") {
        bmp(filename = file, width = input$downloadLigandChatMWidth, height = input$downloadLigandChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (!is.null(input$Ligand_SelectM)) {
          
          if (input$radioLigandChatM == "Dot Plot") {
            par(mfrow=c(1,1))
            print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectM, targets.use = input$target_selectM, comparison = c(1, 2), signaling = input$Ligand_SelectM, remove.isolate = F, font.size = input$sizeLigandM, font.size.title = input$sizeLigandM, angle.x = 90))
          }
          else if (input$radioLigandChatM == "Gene Expression") {
            seuratreactive$cellchatM@meta$datasets = factor(seuratreactive$cellchatM@meta$datasets, levels = c(names(seuratreactive$object.list)[1], names(seuratreactive$object.list)[2])) # set factor level
            print(plotGeneExpression(seuratreactive$cellchatM, signaling = input$Ligand_SelectM, split.by = "datasets", colors.ggplot = T))
          }
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadLigandChatMM, {
    showModal(modalDialog(
      title = strong("Download Upregulated signalling ligand-receptor pairs"),
      numericInput("downloadLigandChatMMHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadLigandChatMMWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadLigandChatMMFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadLigandChatMMOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadLigandChatMMOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DifferentialExpressionAnalysisofSignallingPathways_Upregulated.", input$downloadLigandChatMMFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadLigandChatMMFile == "png") {
        
        png(filename = file, width = input$downloadLigandChatMMWidth, height = input$downloadLigandChatMMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.up[, "interaction_name", drop = F], comparison = c(1, 2), angle.x = 90, remove.isolate = F,title.name = paste0("Up-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMMFile == "jpeg") {
        jpeg(filename = file, width = input$downloadLigandChatMMWidth, height = input$downloadLigandChatMMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.up[, "interaction_name", drop = F], comparison = c(1, 2), angle.x = 90, remove.isolate = F,title.name = paste0("Up-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMMFile == "tiff") {
        tiff(filename = file, width = input$downloadLigandChatMMWidth, height = input$downloadLigandChatMMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.up[, "interaction_name", drop = F], comparison = c(1, 2), angle.x = 90, remove.isolate = F,title.name = paste0("Up-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMMFile == "bmp") {
        bmp(filename = file, width = input$downloadLigandChatMMWidth, height = input$downloadLigandChatMMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.up[, "interaction_name", drop = F], comparison = c(1, 2), angle.x = 90, remove.isolate = F,title.name = paste0("Up-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadLigandChatMM2, {
    showModal(modalDialog(
      title = strong("Download downregulated signalling ligand-receptor pairs"),
      numericInput("downloadLigandChatMM2Height", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadLigandChatMM2Width", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadLigandChatMM2File", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadLigandChatMM2Output", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadLigandChatMM2Output <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DifferentialExpressionAnalysisofSignallingPathways_Downregulated.", input$downloadLigandChatMM2File, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadLigandChatMM2File == "png") {
        
        png(filename = file, width = input$downloadLigandChatMM2Width, height = input$downloadLigandChatMM2Height,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.down[, "interaction_name", drop = F], comparison = c(1, 2),  angle.x = 90, remove.isolate = F,title.name = paste0("Down-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMM2File == "jpeg") {
        jpeg(filename = file, width = input$downloadLigandChatMM2Width, height = input$downloadLigandChatMM2Height,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.down[, "interaction_name", drop = F], comparison = c(1, 2),  angle.x = 90, remove.isolate = F,title.name = paste0("Down-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMM2File == "tiff") {
        tiff(filename = file, width = input$downloadLigandChatMM2Width, height = input$downloadLigandChatMM2Height,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.down[, "interaction_name", drop = F], comparison = c(1, 2),  angle.x = 90, remove.isolate = F,title.name = paste0("Down-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
      
      else if (input$downloadLigandChatMM2File == "bmp") {
        bmp(filename = file, width = input$downloadLigandChatMM2Width, height = input$downloadLigandChatMM2Height,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        print(netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.down[, "interaction_name", drop = F], comparison = c(1, 2),  angle.x = 90, remove.isolate = F,title.name = paste0("Down-regulated signaling in ", input$Dataset_SelectMM)))
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadRolesChatM, {
    showModal(modalDialog(
      title = strong("Download Cell-Cell Signalling Roles Plot"),
      numericInput("downloadRolesChatMHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadRolesChatMWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadRolesChatMFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadRolesChatMOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadRolesChatMOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_Cell-CellSignallingRoles.", input$downloadRolesChatMFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadRolesChatMFile == "png") {
        
        png(filename = file, width = input$downloadRolesChatMWidth, height = input$downloadRolesChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChatM == "Scatter Plot") {
          num.link <- sapply(seuratreactive$object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
          gg <- list()
          for (i in 1:length(seuratreactive$object.list)) {
            gg[[i]] <- netAnalysis_signalingRole_scatter(seuratreactive$object.list[[i]], title = names(seuratreactive$object.list)[i], weight.MinMax = c(min(num.link[[i]]), max(num.link[[i]])),  font.size = input$sizeRolesM, font.size.title = input$sizeRolesM, label.size	= input$sizeRolesM - 8)
          }
          
          print(patchwork::wrap_plots(plots = gg))
        }
        else if (input$radioRolesChatM == "Bar Graph") {
          gg1 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = T, do.stat = TRUE, font.size = input$sizeRolesM)
          gg2 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = F, do.stat = TRUE, font.size = input$sizeRolesM)
          print(gg1 + gg2)
        }
        else if (input$radioRolesChatM == "Heatmap") {
          if (input$signallingroles_MM == "Outgoing") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "Incoming") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "All") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadRolesChatMFile == "jpeg") {
        jpeg(filename = file, width = input$downloadRolesChatMWidth, height = input$downloadRolesChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChatM == "Scatter Plot") {
          num.link <- sapply(seuratreactive$object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
          gg <- list()
          for (i in 1:length(seuratreactive$object.list)) {
            gg[[i]] <- netAnalysis_signalingRole_scatter(seuratreactive$object.list[[i]], title = names(seuratreactive$object.list)[i], weight.MinMax = c(min(num.link[[i]]), max(num.link[[i]])),  font.size = input$sizeRolesM, font.size.title = input$sizeRolesM, label.size	= input$sizeRolesM - 8)
          }
          
          print(patchwork::wrap_plots(plots = gg))
        }
        else if (input$radioRolesChatM == "Bar Graph") {
          gg1 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = T, do.stat = TRUE, font.size = input$sizeRolesM)
          gg2 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = F, do.stat = TRUE, font.size = input$sizeRolesM)
          print(gg1 + gg2)
        }
        else if (input$radioRolesChatM == "Heatmap") {
          if (input$signallingroles_MM == "Outgoing") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "Incoming") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "All") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadRolesChatMFile == "tiff") {
        tiff(filename = file, width = input$downloadRolesChatMWidth, height = input$downloadRolesChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChatM == "Scatter Plot") {
          num.link <- sapply(seuratreactive$object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
          gg <- list()
          for (i in 1:length(seuratreactive$object.list)) {
            gg[[i]] <- netAnalysis_signalingRole_scatter(seuratreactive$object.list[[i]], title = names(seuratreactive$object.list)[i], weight.MinMax = c(min(num.link[[i]]), max(num.link[[i]])),  font.size = input$sizeRolesM, font.size.title = input$sizeRolesM, label.size	= input$sizeRolesM - 8)
          }
          
          print(patchwork::wrap_plots(plots = gg))
        }
        else if (input$radioRolesChatM == "Bar Graph") {
          gg1 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = T, do.stat = TRUE, font.size = input$sizeRolesM)
          gg2 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = F, do.stat = TRUE, font.size = input$sizeRolesM)
          print(gg1 + gg2)
        }
        else if (input$radioRolesChatM == "Heatmap") {
          if (input$signallingroles_MM == "Outgoing") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "Incoming") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "All") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
        }
        
        dev.off()
      }
      
      else if (input$downloadRolesChatMFile == "bmp") {
        bmp(filename = file, width = input$downloadRolesChatMWidth, height = input$downloadRolesChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioRolesChatM == "Scatter Plot") {
          num.link <- sapply(seuratreactive$object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
          gg <- list()
          for (i in 1:length(seuratreactive$object.list)) {
            gg[[i]] <- netAnalysis_signalingRole_scatter(seuratreactive$object.list[[i]], title = names(seuratreactive$object.list)[i], weight.MinMax = c(min(num.link[[i]]), max(num.link[[i]])),  font.size = input$sizeRolesM, font.size.title = input$sizeRolesM, label.size	= input$sizeRolesM - 8)
          }
          
          print(patchwork::wrap_plots(plots = gg))
        }
        else if (input$radioRolesChatM == "Bar Graph") {
          gg1 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = T, do.stat = TRUE, font.size = input$sizeRolesM)
          gg2 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = F, do.stat = TRUE, font.size = input$sizeRolesM)
          print(gg1 + gg2)
        }
        else if (input$radioRolesChatM == "Heatmap") {
          if (input$signallingroles_MM == "Outgoing") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "Incoming") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
          else if (input$signallingroles_MM == "All") {
            pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
            ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
            print(ht1 + ht2)
          }
        }
        
        dev.off()
      }
    }
  )
  
  observeEvent(input$downloadManifoldChatM, {
    showModal(modalDialog(
      title = strong("Download Manifold and Classification Learning Analysis of Signaling Networks Plot"),
      numericInput("downloadManifoldChatMHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadManifoldChatMWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("downloadManifoldChatMFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("jpeg", "tiff", "png", "bmp"),
                         selected = "png"),
      downloadBttn("downloadManifoldChatMOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadManifoldChatMOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_ManifoldLearningPlot.", input$downloadManifoldChatMFile, sep = "")
    },
    
    content = function(file) {
      
      if (input$downloadManifoldChatMFile == "png") {
        
        png(filename = file, width = input$downloadManifoldChatMWidth, height = input$downloadManifoldChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChatM == "Functional Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "functional", label.size = input$sizeManifoldM))
        }
        else if (input$radioManifoldChatM == "Structural Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "structural", label.size = input$sizeManifoldM))
        }
        
        dev.off()
      }
      
      else if (input$downloadManifoldChatMFile == "jpeg") {
        jpeg(filename = file, width = input$downloadManifoldChatMWidth, height = input$downloadManifoldChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChatM == "Functional Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "functional", label.size = input$sizeManifoldM))
        }
        else if (input$radioManifoldChatM == "Structural Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "structural", label.size = input$sizeManifoldM))
        }
        
        dev.off()
      }
      
      else if (input$downloadManifoldChatMFile == "tiff") {
        tiff(filename = file, width = input$downloadManifoldChatMWidth, height = input$downloadManifoldChatMHeight,
             units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChatM == "Functional Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "functional", label.size = input$sizeManifoldM))
        }
        else if (input$radioManifoldChatM == "Structural Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "structural", label.size = input$sizeManifoldM))
        }
        
        dev.off()
      }
      
      else if (input$downloadManifoldChatMFile == "bmp") {
        bmp(filename = file, width = input$downloadManifoldChatMWidth, height = input$downloadManifoldChatMHeight,
            units = "mm", pointsize = 12, bg = "white", res = 600)
        
        if (input$radioManifoldChatM == "Functional Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "functional", label.size = input$sizeManifoldM))
        }
        else if (input$radioManifoldChatM == "Structural Classification") {
          print(netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "structural", label.size = input$sizeManifoldM))
        }
        
        dev.off()
      }
    }
  )
  
  ####DRUG####
  output$downloadmarkersDRUG <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SignificantGeneDrugInteractions.csv", sep = "")
    }, 
    
    content = function(file) {
      
      write.csv(seuratreactive$DRUGdf, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadGeneDrugMap, {
    showModal(modalDialog(
      title = strong("Download Drug Gene Interaction Map"),
      numericInput("downloadGeneDrugMapHeight", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadGeneDrugMapWidth", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ddownloadGeneDrugMapFile", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadGeneDrugMapOutput", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadGeneDrugMapOutput <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DrugGeneInteractionMap.", input$ddownloadGeneDrugMapFile, sep = "")
    }, 
    
    content = function(file) {
      
      DRUGmap <- seuratreactive$DRUGdf[1:input$DRUGNumber,]
      
      edges <- DRUGmap[,c("Gene Name", "Drug Claim Name", "Log2 fold change")]
      
      nodes1 <- DRUGmap %>%
        group_by(`Gene Name`) %>%
        summarise(node_color = "gene", n = 2) %>% 
        drop_na()
      
      nodes2 <- DRUGmap %>%
        group_by(`Drug Claim Name`) %>%
        summarise(node_color = "cluster", n = 2.1) %>% 
        drop_na()
      
      colnames(nodes2)[1] <- c("Gene Name")
      
      nodes <- full_join(nodes1, nodes2)
      
      graph = graph_from_data_frame(edges, directed = FALSE, nodes)
      
      if (input$checkboxgene_DRUG == TRUE && input$checkboxcluster_DRUG == TRUE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG, input$nodeDRUG/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG, repel=TRUE) +
          geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      else if (input$checkboxgene_DRUG == FALSE && input$checkboxcluster_DRUG == TRUE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG, input$nodeDRUG/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      else if (input$checkboxgene_DRUG == TRUE && input$checkboxcluster_DRUG == FALSE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG, input$nodeDRUG/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      ggsave(file, plot = plot,
             width = input$downloadGeneDrugMapWidth,
             height = input$downloadGeneDrugMapHeight,
             units = "mm",
             device = input$ddownloadGeneDrugMapFile)
    }
  )
  
  ####DRUG2####
  output$downloadmarkersDRUG2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SignificantGeneDrugInteractions.csv", sep = "")
    }, 
    
    content = function(file) {
      
      write.csv(seuratreactive$DRUGdf2, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadGeneDrugMap2, {
    showModal(modalDialog(
      title = strong("Download Drug Gene Interaction Map"),
      numericInput("downloadGeneDrugMapHeight2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("downloadGeneDrugMapWidth2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ddownloadGeneDrugMapFile2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("downloadGeneDrugMapOutput2", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadGeneDrugMapOutput2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DrugGeneInteractionMap.", input$ddownloadGeneDrugMapFile2, sep = "")
    }, 
    
    content = function(file) {
      
      DRUGmap <- seuratreactive$DRUGdf2[1:input$DRUGNumber2,]
      
      edges <- DRUGmap[,c("Gene Name", "Drug Claim Name", "Log2 fold change")]
      
      nodes1 <- DRUGmap %>%
        group_by(`Gene Name`) %>%
        summarise(node_color = "gene", n = 2) %>% 
        drop_na()
      
      nodes2 <- DRUGmap %>%
        group_by(`Drug Claim Name`) %>%
        summarise(node_color = "cluster", n = 2.1) %>% 
        drop_na()
      
      colnames(nodes2)[1] <- c("Gene Name")
      
      nodes <- full_join(nodes1, nodes2)
      
      graph = graph_from_data_frame(edges, directed = FALSE, nodes)
      
      if (input$checkboxgene_DRUG2 == TRUE && input$checkboxcluster_DRUG2 == TRUE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG2, input$nodeDRUG2/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG2, repel=TRUE) +
          geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG2, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      else if (input$checkboxgene_DRUG2 == FALSE && input$checkboxcluster_DRUG2 == TRUE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG2, input$nodeDRUG2/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG2, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      else if (input$checkboxgene_DRUG2 == TRUE && input$checkboxcluster_DRUG2 == FALSE) {
        plot <- ggraph(graph, layout = "fr") + 
          geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
          scale_edge_width(range = c(1, 2))+
          geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG2, input$nodeDRUG2/2)) + 
          geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG2, repel=TRUE) +
          scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
          scale_color_manual(values = c("grey", "green")) +
          theme_graph() + 
          guides(edge_width = FALSE, color = FALSE)
      }
      
      ggsave(file, plot = plot,
             width = input$downloadGeneDrugMapWidth2,
             height = input$downloadGeneDrugMapHeight2,
             units = "mm",
             device = input$ddownloadGeneDrugMapFile2)
    }
  )
  
  ####MM######
  observeEvent(input$downloadMMViolin, {
    showModal(modalDialog(
      title = strong("Download Violin Plot (scRNA-seq)"),
      numericInput("ViolinHeightMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("ViolinWidthMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ViolinFileInputMM", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("ViolindownloadoutputMM", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$ViolindownloadoutputMM <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_ViolinPlot_GeneExpression.", input$ViolinFileInputMM, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$pMM,
             width = input$ViolinWidthMM,
             height = input$ViolinHeightMM,
             units = "mm",
             device = input$ViolinFileInputMM)
    }
  )
  
  observeEvent(input$downloadMMViolin2, {
    showModal(modalDialog(
      title = strong("Download Violin Plot (Multimodal data)"),
      numericInput("ViolinHeightMM2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("ViolinWidthMM2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("ViolinFileInputMM2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("ViolindownloadoutputMM2", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$ViolindownloadoutputMM2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_ViolinPlot_GeneExpression_MultimodalAnalysis.", input$ViolinFileInputMM2, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$pMM2,
             width = input$ViolinWidthMM2,
             height = input$ViolinHeightMM2,
             units = "mm",
             device = input$ViolinFileInputMM2)
    }
  )
  
  ####DE####
  output$downloadmarkersDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DifferentialExpressedGenesTable.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$markersDE, file, row.names = F)
    }
  )
  
  output$downloadYuDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DifferentialExpressedGenesEnrichmentTable.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$YuDE, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadvolcanoDE, {
    showModal(modalDialog(
      title = strong("Download Volcano Plot"),
      numericInput("VolcanoHeightDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("VolcanoWidthDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("VolcanoFileInputDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("VolcanodownloadoutputDE", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$VolcanodownloadoutputDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_VolcanoPlot_DifferentialExpression.", input$VolcanoFileInputDE, sep = "")
    }, 
    
    content = function(file) {
      p <- ggplot(data=seuratreactive$markersDEvolcano, aes(x= `Log2 fold change`, y=-log10(`Adjusted p-value`), col = `Expression`, 
      )) +
        geom_point(size = input$sizevolcanoDE, alpha = input$opacityvolcanoDE) + 
        theme_minimal() +
        theme(legend.position="none", text = element_text(size=input$labelsizevolcanoDE)) +
        ylab("-log10(p-value)") +
        xlab("Log2 fold change") +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_vline(xintercept=c(seuratreactive$volcanovalue1, seuratreactive$volcanovalue2), col="red", linetype = "dotted", size = 1.5) +
        geom_hline(yintercept=-log10(seuratreactive$pvaluevolcano), col="blue", linetype = "dashed", size = 1.5)
      
      ggsave(file, plot = p,
             width = input$VolcanoWidthDE,
             height = input$VolcanoHeightDE,
             units = "mm",
             device = input$VolcanoFileInputDE)
    }
  )
  
  observeEvent(input$downloadPathwayPlotDE, {
    showModal(modalDialog(
      title = strong("Download Pathway Plot"),
      numericInput("PathwayPlotHeightDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("PathwayPlotWidthDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("PathwayPlotFileInputDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("PathwayPlotdownloadoutputDE", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$PathwayPlotdownloadoutputDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_PathwayViewer_DifferentialExpression.", input$PathwayPlotFileInputDE, sep = "")
    }, 
    
    content = function(file) {
      q <- ggplot(seuratreactive$Subset_GENES, aes(x=reorder(`Gene Name`, `Log2 fold change`), y=`Log2 fold change`)) +
        geom_point(size = input$sizeGPDE, alpha = input$opacityGPDE, col = "red") + 
        theme_minimal() +
        theme(legend.position="none", text = element_text(size=input$labelsizeGPDE)) +
        ylab("Log2 fold change") +
        xlab("") +
        ggtitle(seuratreactive$title) +
        theme(plot.title = element_text(vjust=0.5)) +
        geom_hline(yintercept=-log10(1), size = 1)
      
      ggsave(file, plot = q,
             width = input$PathwayPlotWidthDE,
             height = input$PathwayPlotHeightDE,
             units = "mm",
             device = input$PathwayPlotFileInputDE)
    }
  )
  
  ####SDE####
  output$downloadlasso1 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SelectedDifferentialExpression_Lasso1SelectedCells.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$lasso1, file, row.names = F)
    }
  )
  
  output$downloadlasso2 <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SelectedDifferentialExpression_Lasso2SelectedCells.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$lasso2, file, row.names = F)
    }
  )
  
  output$downloadmarkersSDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SelectedDifferentialExpressedGenesTable.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$markersSDE, file, row.names = F)
    }
  )
  
  output$downloadYuSDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_SelectedDifferentialExpressedGenesEnrichmentTable.csv", sep = "")
    }, 
    
    content = function(file) {
      write.csv(seuratreactive$YuSDE, file, row.names = F)
    }
  )
  
  observeEvent(input$downloadvolcanoSDE, {
    showModal(modalDialog(
      title = strong("Download Volcano Plot"),
      numericInput("VolcanoHeightSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("VolcanoWidthSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("VolcanoFileInputSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("VolcanodownloadoutputSDE", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$VolcanodownloadoutputSDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_VolcanoPlot_SelectedDifferentialExpression.", input$VolcanoFileInputSDE, sep = "")
    }, 
    
    content = function(file) {
      p <- ggplot(data=seuratreactive$markersSDEvolcano, aes(x= `Log2 fold change`, y=-log10(`Adjusted p-value`), col = `Expression`)) +
        geom_point(size = input$sizevolcanoSDE, alpha = input$opacityvolcanoSDE) +
        theme_minimal() +
        theme(legend.position="none", text = element_text(size=input$labelsizevolcanoSDE)) +
        ylab("-log10(p-value)") +
        xlab("Log2 fold change") +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_vline(xintercept=c(seuratreactive$volcanoSDEvalue1, seuratreactive$volcanoSDEvalue2), col="red", linetype = "dotted", size = 1.5) +
        geom_hline(yintercept=-log10(seuratreactive$pvaluevolcanoSDE), col="blue", linetype = "dashed", size = 1.5)
      
      ggsave(file, plot = p,
             width = input$VolcanoWidthSDE,
             height = input$VolcanoHeightSDE,
             units = "mm",
             device = input$VolcanoFileInputSDE)
    }
  )
  
  observeEvent(input$downloadPathwayPlotSDE, {
    showModal(modalDialog(
      title = strong("Download Pathway Plot"),
      numericInput("PathwayPlotHeightSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("PathwayPlotWidthSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("PathwayPlotFileInputSDE", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("PathwayPlotdownloadoutputSDE", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$PathwayPlotdownloadoutputSDE <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_PathwayViewer_SelectedDifferentialExpression.", input$PathwayPlotFileInputSDE, sep = "")
    }, 
    
    content = function(file) {
      q <- ggplot(seuratreactive$Subset_GENESSDE, aes(x=reorder(`Gene Name`, `Log2 fold change`), y=`Log2 fold change`)) +
        geom_point(size = input$sizeGPSDE, alpha = input$opacityGPSDE, col = "red") +
        theme_minimal() +
        theme(legend.position="none", text = element_text(size=input$labelsizeGPSDE)) +
        ylab("Log2 fold change") +
        xlab("") +
        ggtitle(seuratreactive$titleSDE) +
        theme(plot.title = element_text(vjust=0.5)) +
        geom_hline(yintercept=-log10(1), size = 1)
      
      ggsave(file, plot = q,
             width = input$PathwayPlotWidthSDE,
             height = input$PathwayPlotHeightSDE,
             units = "mm",
             device = input$PathwayPlotFileInputSDE)
    }
  )
  
  ####PA#####
  observeEvent(input$downloaddotplotPA, {
    showModal(modalDialog(
      title = strong("Download Dot Plot"),
      numericInput("dotplotHeightPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("dotplotWidthPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("dotplotFileInputPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("dotplotdownloadoutputPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$dotplotdownloadoutputPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DotPlot_PathwayAnalysis.", input$dotplotFileInputPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$dotplotPA,
             width = input$dotplotWidthPA,
             height = input$dotplotHeightPA,
             units = "mm",
             device = input$dotplotFileInputPA)
    }
  )
  
  observeEvent(input$downloadGCplotPA, {
    showModal(modalDialog(
      title = strong("Download Gene Concept Network"),
      numericInput("GCplotHeightPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("GCplotWidthPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("GCplotFileInputPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("GCplotdownloadoutputPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$GCplotdownloadoutputPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_GeneConceptNetwork_PathwayAnalysis.", input$GCplotFileInputPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$GCplotPA,
             width = input$GCplotWidthPA,
             height = input$GCplotHeightPA,
             units = "mm",
             device = input$GCplotFileInputPA)
    }
  )
  
  observeEvent(input$downloademapplotPA, {
    showModal(modalDialog(
      title = strong("Download Emap Plot"),
      numericInput("emapplotHeightPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("emapplotWidthPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("emapplotFileInputPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("emapplotdownloadoutputPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$emapplotdownloadoutputPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_EmapPlot_PathwayAnalysis.", input$emapplotFileInputPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$emapplotPA,
             width = input$emapplotWidthPA,
             height = input$emapplotHeightPA,
             units = "mm",
             device = input$emapplotFileInputPA)
    }
  )
  
  observeEvent(input$downloadpathwayPA, {
    showModal(modalDialog(
      title = strong("Download Pathway Plot"),
      numericInput("pathwayHeightPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("pathwayWidthPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("pathwayFileInputPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("pathwaydownloadoutputPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$pathwaydownloadoutputPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_PathwayPlot_PathwayAnalysis.", input$pathwayFileInputPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = Yu$PAPA,
             width = input$pathwayWidthPA,
             height = input$pathwayHeightPA,
             units = "mm",
             device = input$pathwayFileInputPA)
    }
  )
  
  observeEvent(input$downloadtreeplotPA, {
    showModal(modalDialog(
      title = strong("Download Tree Plot"),
      numericInput("TreeHeightPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("TreeWidthPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("TreeFileInputPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("TreedownloadoutputPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$TreedownloadoutputPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_TreePlot_PathwayAnalysis.", input$TreeFileInputPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$treeplotPA,
             width = input$TreeWidthPA,
             height = input$TreeHeightPA,
             units = "mm",
             device = input$TreeFileInputPA)
    }
  )
  
  ####SPA#####
  observeEvent(input$downloaddotplotSPA, {
    showModal(modalDialog(
      title = strong("Download Dot Plot"),
      numericInput("dotplotHeightSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("dotplotWidthSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("dotplotFileInputSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("dotplotdownloadoutputSPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$dotplotdownloadoutputSPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_DotPlot_SelectedPathwayAnalysis.", input$dotplotFileInputSPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$dotplotSPA,
             width = input$dotplotWidthSPA,
             height = input$dotplotHeightSPA,
             units = "mm",
             device = input$dotplotFileInputSPA)
    }
  )
  
  observeEvent(input$downloadGCplotSPA, {
    showModal(modalDialog(
      title = strong("Download Gene Concept Network"),
      numericInput("GCplotHeightSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("GCplotWidthSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("GCplotFileInputSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("GCplotdownloadoutputSPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$GCplotdownloadoutputSPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_GeneConceptNetwork_SelectedPathwayAnalysis.", input$GCplotFileInputSPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$GCplotSPA,
             width = input$GCplotWidthSPA,
             height = input$GCplotHeightSPA,
             units = "mm",
             device = input$GCplotFileInputSPA)
    }
  )
  
  observeEvent(input$downloademapplotSPA, {
    showModal(modalDialog(
      title = strong("Download Emap Plot"),
      numericInput("emapplotHeightSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("emapplotWidthSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("emapplotFileInputSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("emapplotdownloadoutputSPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$emapplotdownloadoutputSPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_EmapPlot_SelectedPathwayAnalysis.", input$emapplotFileInputSPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = seuratreactive$emapplotSPA,
             width = input$emapplotWidthSPA,
             height = input$emapplotHeightSPA,
             units = "mm",
             device = input$emapplotFileInputSPA)
    }
  )
  
  observeEvent(input$downloadpathwaySPA, {
    showModal(modalDialog(
      title = strong("Download Pathway Plot"),
      numericInput("pathwayHeightSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Height of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      numericInput("pathwayWidthSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Width of plot (mm)')),
                   value = "480",
                   min = "0",
                   max = "1000"),
      virtualSelectInput("pathwayFileInputSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select type of file format')),
                         choices = c("pdf", "jpeg", "tiff", "png", "bmp", "svg"),
                         selected = "png"),
      downloadBttn("pathwaydownloadoutputSPA", "Download"),
      size = "s",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$pathwaydownloadoutputSPA <- downloadHandler(
    filename = function() {
      paste(paste(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_PathwayPlot_SelectedPathwayAnalysis.", input$pathwayFileInputSPA, sep = "")
    }, 
    
    content = function(file) {
      ggsave(file, plot = YuSPA$PAPA,
             width = input$pathwayWidthSPA,
             height = input$pathwayHeightSPA,
             units = "mm",
             device = input$pathwayFileInputSPA)
    }
  )
  
  ####TAB 0 ####
  
  shinyjs::onclick("Flow_chart",  
                   showModal(modalDialog(
                     title = strong("Figure 1 Flow chart of single cell RNA-seq analysis"),
                     img(id="Flow_chart",src="Flow_chart.png",width="100%",style="cursor:pointer;"),
                     size = "l",
                     easyClose = TRUE,
                     footer = NULL
                   )))
  
  shinyjs::onclick("SaveLoad",  
                   showModal(modalDialog(
                     title = strong("Figure 3 Save and Continue interface"),
                     img(id="SaveLoad",src="SaveLoad.png",width="100%",style="cursor:pointer;"),
                     size = "l",
                     easyClose = TRUE,
                     footer = NULL
                   )))
  
  shinyjs::onclick("Navigation",  
                   showModal(modalDialog(
                     title = strong("Figure 2 General Navigation"),
                     img(id="Navigation",src="Navigation.png",width="100%",style="cursor:pointer;"),
                     size = "l",
                     easyClose = TRUE,
                     footer = NULL
                   )))
  
  observe({
    js$collapse("LoadBoxIntro")
    js$collapse("CombineBoxIntro")
    js$collapse("QCBoxIntro")
    js$collapse("DBoxIntro")
    js$collapse("DBoxIntro2")
    js$collapse("QCBoxIntro2")
    js$collapse("DRBoxIntro")
    js$collapse("DRBoxIntro2")
    js$collapse("CBoxIntro")
    js$collapse("CCBoxIntro")
    js$collapse("LBoxIntro")
    js$collapse("MEGENABoxIntro")
    js$collapse("SCENICBoxIntro")
    js$collapse("MAGMABoxIntro")
    js$collapse("GEBoxIntro")
    js$collapse("TABoxIntro")
    js$collapse("ChatBoxIntro")
    js$collapse("DRUGBoxIntro")
    js$collapse("DRUGBoxIntro2")
    js$collapse("DEBoxIntro")
    js$collapse("SDEBoxIntro")
    js$collapse("PABoxIntro")
    js$collapse("SPABoxIntro")
    js$collapse("MMBoxIntro")
    js$collapse("MMBox4")
    js$collapse("MMBox5")
  })
  
  ####TAB 0.5 - LOGS#####
  
  output$textlogsUI <- renderUI({
    strspecies <- paste("<h4>Selected species: ", seuratreactive$REACT)
    
    strLoad <- ifelse (seuratreactive$type == "single",
                       paste("<h4>Load Your Data </h4>",
                             "Name of sample: ", paste(unique(seuratreactive$obj_original@meta.data$orig.ident), collapse = " "), "<br>",
                             "Include genes detected in at least this many cells: ", seuratreactive$mincells, "<br>",
                             "Include cells where at least this many genes are detected: ", seuratreactive$minfeatures),
                       paste("<h4>Load Your Data </h4>",
                             "Name of sample: ", paste(unique(seuratreactive$obj_original@meta.data$orig.ident), collapse = " "))
    )
    
    strLoad2 <- ifelse (is.null(seuratreactive$obj_original2),
                        paste("<h4> Second dataset was not loaded </h4>"),
                        ifelse(seuratreactive$type2 == "single",
                               paste("<h4>Load Your Data (Second Dataset) </h4>",
                                     "Name of sample: ", paste(unique(seuratreactive$obj_original2@meta.data$orig.ident), collapse = " "), "<br>",
                                     "Include genes detected in at least this many cells: ", seuratreactive$mincellsCombine, "<br>",
                                     "Include cells where at least this many genes are detected: ", seuratreactive$minfeaturesCombine),
                               ifelse (seuratreactive$type2 == "multiple",
                                       paste("<h4>Load Your Data (Second Dataset) </h4>",
                                             "Name of sample: ", paste(unique(seuratreactive$obj_original2@meta.data$orig.ident), collapse = " ")),
                                       paste("<h4> Second Dataset was not loaded </h4>")
                               )))
    
    strQC <- ifelse (!is.null(seuratreactive$objQC),
                     paste("<h4>Quality Control </h4>",
                           "<b>The following thresholds were implemented: </b> <br>",
                           "Number of unique genes (nFeature_RNA): ", paste(seuratreactive$nFeature1), "to", paste(seuratreactive$nFeature2), "<br>",
                           "Number of molecules detected per cell (nCount_RNA): ", paste(seuratreactive$nCount1), "to", paste(seuratreactive$nCount2), "<br>",
                           "Mitochondrial percentage (percent.mt): ", paste(seuratreactive$percent.mt1), "to", paste(seuratreactive$percent.mt2), "<br>",
                           "Ribosomal percentage (percent.ribo): ", paste(seuratreactive$percent.ribo1), "to", paste(seuratreactive$percent.ribo2)), 
                     paste("<h4> Quality Control was not completed </h4>")
    )
    
    strD <- ifelse (!is.null(seuratreactive$Removed),
                    paste("<h4>Doublet Removal </h4>",
                          "Estimated percentage of doublets in dataset: ", paste(seuratreactive$doublesim), "<br>",
                          "Doublets removed: ", paste(seuratreactive$Remove, collapse = ", ")), 
                    paste("<h4> Doublet Removal was not completed </h4>")
    )
    
    strQC2 <- ifelse (!is.null(seuratreactive$objQC2),
                      paste("<h4>Quality Control (Second Dataset) </h4>",
                            "<b>The following thresholds were implemented: </b> <br>",
                            "Number of unique genes (nFeature_RNA): ", paste(seuratreactive$nFeature12), "to", paste(seuratreactive$nFeature22), "<br>",
                            "Number of molecules detected per cell (nCount_RNA): ", paste(seuratreactive$nCount12), "to", paste(seuratreactive$nCount22), "<br>",
                            "Mitochondrial percentage (percent.mt): ", paste(seuratreactive$percent.mt12), "to", paste(seuratreactive$percent.mt22), "<br>",
                            "Ribosomal percentage (percent.ribo): ", paste(seuratreactive$percent.ribo12), "to", paste(seuratreactive$percent.ribo22)), 
                      paste("<h4> Quality Control (Second Dataset) was not completed (Second dataset was not loaded) </h4>")
    )
    
    strD2 <- ifelse (!is.null(seuratreactive$Removed2),
                     paste("<h4>Doublet Removal (Second dataset) </h4>",
                           "Estimated percentage of doublets in dataset: ", paste(seuratreactive$doublesim2), "<br>",
                           "Doublets removed: ", paste(seuratreactive$Remove2, collapse = ", ")), 
                     paste("<h4> Doublet Removal (second dataset) was not completed (second dataset was not loaded) </h4>")
    )
    
    strDR <- ifelse (!is.null(seuratreactive$variablefeatures),
                     paste("<h4>Dimensionality Reduction </h4>",
                           "Number of variable features incoporated: ", paste(seuratreactive$variablefeatures), "<br>",
                           "Normalization method: ", paste(seuratreactive$SCTransformDR), "<br>",
                           "Number of Dimensions: ", paste(seuratreactive$npcDR), "<br>",
                           "Imputed dropouts: ", paste(seuratreactive$ALRAL)),
                     ifelse(!is.null(seuratreactive$VFintegrated),
                            paste("<h4>Dimensionality Reduction </h4>",
                                  "Number of variable features incoporated: ", paste(seuratreactive$VFintegrated), "<br>",
                                  "Normalization method: ", paste(seuratreactive$SCTransformDR2), "<br>",
                                  "Number of Dimensions: ", paste(seuratreactive$npcDR2), "<br>",
                                  "Integration method: ", paste(seuratreactive$integrationreduction), "<br>",
                                  "Strength of integration: ", paste(seuratreactive$kanchor)),
                            paste("<h4> Dimensionality Reduction was not completed </h4>"))
    )
    
    strC <- ifelse (!is.null(seuratreactive$integera),
                    paste("<h4>Clustering </h4>",
                          "<b>Nearest-neighbour graph construction </b> <br>",
                          "Number of dimensions: ", paste(seuratreactive$integera), "<br>",
                          "k.param: ", paste(seuratreactive$integerb), "<br>",
                          "n.trees: ", paste(seuratreactive$integerc), "<br>",
                          "<b>Clustering Parameters </b> <br>",
                          "Resolution: ", paste(seuratreactive$integerd), "<br>",
                          "Clustering algorithm: ", paste(seuratreactive$select1), "<br>",
                          "<b>UMAP Parameters </b> <br>",
                          "Number of dimenions: ", paste(seuratreactive$integere), "<br>",
                          "k-nearest-neighbours: ", paste(seuratreactive$integerf), "<br>",
                          "min.dist: ", paste(seuratreactive$integerg), "<br>",
                          "<b>t-SNE Parameters </b> <br>",
                          "Number of dimensions: ", paste(seuratreactive$tsne1), "<br>",
                          "Perplexity: ", paste(seuratreactive$tsne2), "<br>",
                          "Max iterations: ", paste(seuratreactive$tsne3)),
                    paste("<h4> Clustering was not completed </h4>")
    )
    
    strCC <- ifelse (!is.null(seuratreactive$reductionCC1) & !is.null(seuratreactive$reductionCC1),
                     paste("<h4>Data Correction </h4>",
                           "The following were regressed out of the dataset: ", paste(seuratreactive$reductionCC1, collapse = ", "), "<br>",
                           "The following genes were regressed out of the dataset: ", paste(seuratreactive$reductionCC2, collapse = ", "), "<br>",
                           "The following gene pathway were regressed out of the dataset: ", paste(seuratreactive$reductionCC3, collapse = ", ")
                     ),
                     paste("<h4> Data Correction was not completed </h4>")
    )
    
    strL <- ifelse (is.null(seuratreactive$new.cell.ids),
                    paste("<h4>Labelling was not completed </h4>"),
                    
                    ifelse(paste(seuratreactive$reference_annotation) == "singleR database",
                           paste("<h4>Labelling </h4>",
                                 "Reference cell marker database: ", paste(seuratreactive$select2), "<br>",
                                 "Cluster labels: ", paste(unique(seuratreactive$obj@active.ident), collapse = ", ")
                           ),
                           ifelse(paste(seuratreactive$reference_annotation) == "sctype database",
                                  paste("<h4>Labelling </h4>",
                                        "Tissue type:", paste(seuratreactive$selecttissue), "<br>",
                                        "Cluster labels: ", paste(unique(seuratreactive$obj@active.ident), collapse = ", ")
                                  ),
                                  ifelse(paste(unique(seuratreactive$reference_annotation)) == "custom",
                                         paste("<h4>Labelling </h4>",
                                               "Cluster labels: ", paste(unique(seuratreactive$obj@active.ident), collapse = ", ")
                                         ),
                                         paste("<h4>Labelling was not completed </h4>"))))
                    
    )
    
    strMEGENA <- ifelse (is.null(seuratreactive$summary.output),
                         paste("<h4>Gene Co-Expression Modules Construction was not completed </h4>"),
                         ifelse (seuratreactive$MEGENA_selectmethod == "Variable Genes",
                                 paste("<h4>Gene Co-Expression Modules Construction </h4>",
                                       "<b>Co-expression Module construction settings:</b>", "<br>",
                                       "Method of input: ", paste(seuratreactive$MEGENA_selectmethod), "<br>",
                                       "Number of variable genes: ", paste(seuratreactive$variablegenesMEGENA), "<br>",
                                       "p-value threshold: ", paste(seuratreactive$MEGENApval), "<br>",
                                       "Minimum module size: ", paste(seuratreactive$MEGENAmin.size), "<br>",
                                       "Dataset analysed: ", paste(seuratreactive$MEGENA_samples, collapse = ","), "<br>",
                                       "Downsample MEGENA module construction to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_MEGENA, "cells"), "<br>",
                                       "<b>Cell cluster marker gene detection settings (for sunburst plot): </b>", "<br>",
                                       "Selected test: ", paste(seuratreactive$testuseMEGENA), "<br>",
                                       "Log2 Fold Change: ", paste(seuratreactive$log2FCMEGENA), "<br>",
                                       "p-value adjustment method: ", paste(seuratreactive$padjustMEGENA), "<br>",
                                       "p-value threshold: ", paste(seuratreactive$pvalMEGENA), "<br>",
                                       "Limit to genes expressed in X% of cells in cluster: ", paste(seuratreactive$minMEGENA)
                                 ),
                                 paste("<h4>Gene Co-Expression Modules Construction </h4>",
                                       "<b>Co-expression Module construction settings:</b>", "<br>",
                                       "Method of input: ", paste(seuratreactive$MEGENA_selectmethod), "<br>",
                                       "p-value threshold: ", paste(seuratreactive$MEGENApval), "<br>",
                                       "Minimum module size: ", paste(seuratreactive$MEGENAmin.size), "<br>",
                                       "Dataset analysed: ", paste(seuratreactive$MEGENA_samples, collapse = ","), "<br>",
                                       "Downsample MEGENA module construction to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_MEGENA, "cells"), "<br>",
                                       "<b>Cell cluster marker gene detection settings (for sunburst plot): </b>", "<br>",
                                       "Selected test: ", paste(seuratreactive$testuseMEGENA), "<br>",
                                       "Log2 Fold Change: ", paste(seuratreactive$log2FCMEGENA), "<br>",
                                       "p-value adjustment method: ", paste(seuratreactive$padjustMEGENA), "<br>",
                                       "p-value threshold: ", paste(seuratreactive$pvalMEGENA), "<br>",
                                       "Limit to genes expressed in X% of cells in cluster: ", paste(seuratreactive$minMEGENA)
                                 ))
    )
    
    strSCENIC <- ifelse (is.null(seuratreactive$regulons),
                         paste("<h4>Gene Regulatory Network Construction was not completed </h4>"),
                         ifelse (seuratreactive$SCENIC_selectmethod == "Variable Genes",
                                 paste("<h4>Gene Regulatory Network Construction </h4>",
                                       "Method of input: ", paste(seuratreactive$SCENIC_selectmethod), "<br>",
                                       "Number of variable genes: ", paste(seuratreactive$variablegenesSCENIC), "<br>",
                                       "Dataset analysed: ", paste(seuratreactive$SCENIC_samples, collapse = ","), "<br>",
                                       "<b>Regulon Construction settings (SCENIC options): </b>", "<br>",
                                       "Database used: ", paste(seuratreactive$SCENICdb), "<br>",
                                       "Downsample SCENIC analysis to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_SCENIC, "cells")
                                 ),
                                 paste("<h4>Gene Regulatory Network Construction </h4>",
                                       "Method of input: ", paste(seuratreactive$SCENIC_selectmethod), "<br>",
                                       "Dataset analysed: ", paste(seuratreactive$SCENIC_samples, collapse = ","), "<br>",
                                       "<b>Regulon Construction settings (SCENIC options): </b>", "<br>",
                                       "Database used: ", paste(seuratreactive$SCENICdb), "<br>",
                                       "Downsample SCENIC analysis to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_SCENIC, "cells")
                                 ))
    )
    
    strMAGMA <- ifelse (is.null(seuratreactive$MAGMA_Final_Results),
                        paste("<h4>MAGMA Cell Cluster/Cell Type Association Analysis was not completed </h4>"),
                        ifelse (!is.null(seuratreactive$MAGMA_samples) && !is.null(seuratreactive$MAGMA_samples2), 
                                paste("<h4>MAGMA Cell Cluster/Cell Type Association Analysis </h4>",
                                      "Public GWAS Dataset selected: ", paste(seuratreactive$MAGMA_GWAS, collapse = ", "), "<br>",
                                      "Samples analysed (Public GWAS Dataset): ", paste(seuratreactive$MAGMA_samples, collapse = ","), "<br>",
                                      "Own GWAS Dataset: ", paste(seuratreactive$textlabelMAGMA), "<br>",
                                      "Samples analysed (Own GWAS Dataset): ", paste(seuratreactive$MAGMA_samples2, collapse = ","), "<br>",
                                      "Number of GWAS individuals: ", paste(seuratreactive$MAGMA_N), "<br>",
                                      "Reference genome build: ", paste(seuratreactive$GRCh), "<br>",
                                      "GWAS population: ", paste(seuratreactive$MAGMA_Pop), "<br>",
                                      "Downsample MAGMA analysis to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_MAGMA2, "cells")),
                                ifelse(!is.null(seuratreactive$MAGMA_samples) && is.null(seuratreactive$MAGMA_samples2),
                                       paste("<h4>MAGMA Cell Cluster/Cell Type Association Analysis </h4>",
                                             "Public GWAS Dataset selected: ", paste(seuratreactive$MAGMA_GWAS, collapse = ", "), "<br>",
                                             "Samples analysed (Public GWAS Dataset): ", paste(seuratreactive$MAGMA_samples, collapse = ","), "<br>",
                                             "Downsample MAGMA analysis to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_MAGMA2, "cells")),
                                       paste("<h4>MAGMA Cell Cluster/Cell Type Association Analysis </h4>",
                                             "Own GWAS Dataset: ", paste(seuratreactive$textlabelMAGMA), "<br>",
                                             "Samples analysed (Own GWAS Dataset): ", paste(seuratreactive$MAGMA_samples2, collapse = ","), "<br>",
                                             "Number of GWAS individuals: ", paste(seuratreactive$MAGMA_N), "<br>",
                                             "Reference genome build: ", paste(seuratreactive$GRCh), "<br>",
                                             "GWAS population: ", paste(seuratreactive$MAGMA_Pop), "<br>",
                                             "Downsample MAGMA analysis to X cells per cluster/cell type (randomly selected): ", paste(seuratreactive$n_MAGMA, "cells")))
                        ))
    
    strTA <- ifelse (is.null(seuratreactive$TAselectedcells),
                     paste("<h4>Trajectory Analysis was not completed </h4>"),
                     paste("<h4>Trajectory Analysis </h4>",
                           "Root cluster: ", paste(seuratreactive$TA_pickercluster, collapse = ", "), "<br>",
                           "Root cell names: ", paste(seuratreactive$TAselectedcells, collapse = ", ")
                     ))
    
    strChat <- ifelse (!is.null(seuratreactive$cellchat),
                       paste("<h4>Cell-Cell Communication Analysis (Single Dataset) </h4>",
                             "Dataset analysed: ", paste(seuratreactive$ChatSample, collapse = ","), "<br>",
                             "Filter out the cell-cell communication in groups with less than ", seuratreactive$Chat_number, "cells", "<br>",
                             "p-value threshold: ", seuratreactive$Chatpval, "<br>",
                             "Ligand-receptor database: ", paste(seuratreactive$Chatdb)),
                       ifelse(!is.null(seuratreactive$cellchatM),
                              paste("<h4>Cell-Cell Communication Analysis (Comparison Between Two Datasets) </h4>",
                                    "Dataset 1: ", paste(seuratreactive$ChatMultiple1Sample), "<br>",
                                    "Dataset 2: ", paste(seuratreactive$ChatMultiple2Sample), "<br>",
                                    "Filter out the cell-cell communication in groups with less than ", seuratreactive$Chat_number, "cells", "<br>",
                                    "p-value threshold: ", seuratreactive$Chatpval, "<br>",
                                    "Ligand-receptor database: ", paste(seuratreactive$Chatdb)),
                              paste("<h4> Cell-Cell Communication Analysis was not completed </h4>"))
    )
    
    strDRUG <- ifelse (is.null(seuratreactive$markersDRUG),
                       paste("<h4>Drug Gene Interaction Analysis was not completed </h4>"),
                       paste("<h4>Drug Gene Interaction Analysis was completed on differentially expressed genes</h4>"
                       ))
    
    strDRUG2 <- ifelse (is.null(seuratreactive$markersDRUG2),
                        paste("<h4>Custom Drug Gene Interaction Analysis was not completed </h4>"),
                        paste("<h4>Custom Drug Gene Interaction Analysis was completed on differentially expressed genes</h4>"
                        ))
    
    strDE <- ifelse (is.null(seuratreactive$markersDE),
                     paste("<h4>Differential Expression Analysis was not completed </h4>"),
                     
                     ifelse (seuratreactive$SelectComparison == "Between Clusters",
                             paste("<h4>Differential Expression Analysis Between Clusters </h4>",
                                   "Clusters being compared: ", paste(seuratreactive$DEInput1, collapse = ","), "vs", paste(seuratreactive$DEInput2, collapse = ","), "<br>",
                                   "<b>Differential Expression test parameters </b> <br>",
                                   "Test selected: ", seuratreactive$testuseDE, "<br>",
                                   "Log2 fold change threshold: ", seuratreactive$logDE, "<br>",
                                   "Report only upregulated genes: ", seuratreactive$onlyposDE, "<br>",
                                   "p-value adjustment method: ", seuratreactive$padjustDE1, "<br>",
                                   "p-value threshold: ", seuratreactive$pvalDE, "<br>",
                                   "Only test genes that are detected in a minimum fraction of cells (min.pct): ", seuratreactive$minDE, "<br>",
                                   "<b>Gene set Enrichment </b> <br>",
                                   "Select enrichment analysis: ", seuratreactive$enrichInputDE, "<br>",
                                   "Select sub-ontology: ", seuratreactive$GOlabelsDE, "<br>",
                                   "p-value cutoff: ", seuratreactive$pvalueDE, "<br>",
                                   "p-value adjust method: ", seuratreactive$padjustDE, "<br>",
                                   "<b>Volcano Plot </b> <br>",
                                   "Log2 fold change thresholds: ", seuratreactive$volcanovalue1, "to", seuratreactive$volcanovalue2, "<br>",
                                   "p-value cutoff: ", seuratreactive$pvaluevolcano),
                             
                             paste("<h4>Differential Expression Analysis Between Samples </h4>",
                                   "Samples being compared: ", paste(seuratreactive$DEInput1_samples, collapse = ","), "vs", paste(seuratreactive$DEInput2_samples, collapse = ","), "<br>",
                                   "Cluster of interest: ", paste(seuratreactive$DECluster, collapse = ","), "<br>",
                                   "<b>Differential Expression test parameters </b> <br>",
                                   "Test selected: ", seuratreactive$testuseDE, "<br>",
                                   "Log2 fold change threshold: ", seuratreactive$logDE, "<br>",
                                   "Report only upregulated genes: ", seuratreactive$onlyposDE, "<br>",
                                   "p-value adjustment method: ", seuratreactive$padjustDE1, "<br>",
                                   "p-value threshold: ", seuratreactive$pvalDE, "<br>",
                                   "Only test genes that are detected in a minimum fraction of cells (min.pct): ", seuratreactive$minDE, "<br>",
                                   "<b>Gene set Enrichment </b> <br>",
                                   "Select enrichment analysis: ", seuratreactive$enrichInputDE, "<br>",
                                   "Select sub-ontology: ", seuratreactive$GOlabelsDE, "<br>",
                                   "p-value cutoff: ", seuratreactive$pvalueDE, "<br>",
                                   "p-value adjust method: ", seuratreactive$padjustDE, "<br>",
                                   "<b>Volcano Plot </b> <br>",
                                   "Log2 fold change thresholds: ", seuratreactive$volcanovalue1, "to", seuratreactive$volcanovalue2, "<br>",
                                   "p-value cutoff: ", seuratreactive$pvaluevolcano))
    )
    
    strSDE <- ifelse (is.null(seuratreactive$markersSDE),
                      paste("<h4>Custom Differential Expression Analysis was not completed </h4>"),
                      
                      paste("<h4>Custom Differential Expression Analysis </h4>",
                            "<b>Cells being compared: </b> <br>",
                            "Cells in group 1: ", paste(seuratreactive$selectedkey1, collapse = ", "), "<br>",
                            "Cells in group 2: ", paste(seuratreactive$selectedkey2, collapse = ", "), "<br>",
                            "<b>Differential Expression test parameters </b> <br>",
                            "Test selected: ", seuratreactive$testuseSDE, "<br>",
                            "Log2 fold change threshold: ", seuratreactive$logSDE, "<br>",
                            "Report only upregulated genes: ", seuratreactive$onlyposSDE, "<br>",
                            "p-value adjustment method: ", seuratreactive$padjustSDE1, "<br>",
                            "p-value threshold: ", seuratreactive$pvalSDE, "<br>",
                            "Only test genes that are detected in a minimum fraction of cells (min.pct): ", seuratreactive$minSDE, "<br>",
                            "<b>Pathway Enrichment </b> <br>",
                            "Select enrichment analysis: ", seuratreactive$enrichInputSDE, "<br>",
                            "Select sub-ontology: ", seuratreactive$GOlabelsSDE, "<br>",
                            "p-value cutoff: ", seuratreactive$pvalueSDE, "<br>",
                            "p-value adjust method: ", seuratreactive$padjustSDE, "<br>",
                            "<b>Volcano Plot </b> <br>",
                            "Log2 fold change thresholds: ", seuratreactive$volcanoSDEvalue1, "to", seuratreactive$volcanoSDEvalue2, "<br>",
                            "p-value cutoff: ", seuratreactive$pvaluevolcanoSDE)
    )
    
    strPA <- ifelse (is.null(Yu$result),
                     paste("<h4>Pathway Analysis was not completed </h4>"),
                     
                     paste("<h4>Pathway Analysis </h4>",
                           "Select enrichment analysis: ", seuratreactive$PAchoose, "<br>",
                           "Select sub-ontology: ", seuratreactive$GOlabelsPA, "<br>",
                           "p-value cutoff: ", seuratreactive$pvaluePA, "<br>",
                           "p-value adjust method: ", seuratreactive$padjustPA)
    )
    
    strSPA <- ifelse (is.null(YuSPA$result),
                      paste("<h4>Custom Pathway Analysis was not completed </h4>"),
                      
                      paste("<h4>Custom Pathway Analysis </h4>",
                            "Select enrichment analysis: ", seuratreactive$SPAchoose, "<br>",
                            "Select sub-ontology: ", seuratreactive$GOlabelsSPA, "<br>",
                            "p-value cutoff: ", seuratreactive$pvalueSPA, "<br>",
                            "p-value adjust method: ", seuratreactive$padjustSPA)
    )
    
    HTML(paste(strspecies, strLoad, strQC, strD, strLoad2, strQC2, strD2, strDR, strC, strCC, strL, strMEGENA, strSCENIC, strMAGMA, strTA, strChat, strDE, strPA, strDRUG, strSDE, strSPA, strDRUG2, sep = '<br>'))
  })
  
  output$sessionInfo <- renderPrint({
    print("Loaded Packages")
    sessionInfo()
  })
  
  #####LOAD#####
  
  observeEvent(input$fileinputLOAD, {
    
    tryCatch({
      withProgress(message = 'Loading previously saved enviroment...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file = input$fileinputLOAD
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     library(MAST)
                     library(DESeq2)
                     library(BiocParallel)
                     library(MEGENA)
                     library(DGCA)
                     library(ggplot2)
                     library(ggraph)
                     library(GOstats)
                     library(HGNChelper)
                     library(reshape)
                     library(MAGMA.Celltyping)
                     library(EWCE) 
                     library(ComplexHeatmap)
                     library(ggplot2)
                     library(ggraph)
                     library(igraph)
                     library(monocle3)
                     library(SeuratWrappers)
                     library(scran)
                     library(harmony)
                     library(PCAtools)
                     library(scRNAseq)
                     library(Seurat)
                     library(uwot)
                     library(DoubletFinder)
                     library(sctransform)
                     library(SingleR)
                     library(celldex)
                     library(clusterProfiler)
                     library(ReactomePA)
                     library(msigdbr)
                     library(enrichplot)
                     library(graphite)
                     library(limma)
                     library(org.Hs.eg.db)
                     library(org.Bt.eg.db)
                     library(org.Ce.eg.db)
                     library(org.Cf.eg.db)
                     library(org.Dr.eg.db)
                     library(org.Dm.eg.db)
                     library(org.Gg.eg.db)
                     library(org.Mm.eg.db)
                     library(org.Rn.eg.db)
                     library(org.Sc.sgd.db)
                     library(org.Ss.eg.db)
                     library(svglite)
                     library(ggrepel)
                     library(tidyverse)
                     library(data.table)
                     library(tools)
                     library(SCENIC)
                     library(AUCell)
                     library(RcisTarget)
                     library(doParallel)
                     library(doRNG)
                     library(CellChat)
                     library(patchwork)
                     library(NMF)
                     library(ggalluvial)
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     seuratreactive$msig <- NULL
                     seuratreactive$orgDb <- NULL
                     seuratreactive$EKEGG <- NULL
                     seuratreactive$REACT <- NULL
                     seuratreactive$PAREACTOME <- NULL
                     
                     #Load
                     seuratreactive$obj_original <- NULL
                     seuratreactive$matrixtable <- NULL
                     seuratreactive$mincells <- NULL
                     seuratreactive$minfeatures <- NULL
                     
                     #Load2
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     seuratreactive$mincellsCombine <- NULL
                     seuratreactive$minfeaturesCombine <- NULL
                     
                     #QC
                     seuratreactive$objQC <- NULL
                     seuratreactive$type <- NULL
                     seuratreactive$nFeature1 <- NULL
                     seuratreactive$nFeature2 <- NULL
                     seuratreactive$nCount1 <- NULL
                     seuratreactive$nCount2 <- NULL
                     seuratreactive$percent.mt1 <- NULL
                     seuratreactive$percent.mt2 <- NULL
                     seuratreactive$percent.ribo1 <- NULL
                     seuratreactive$percent.ribo2 <- NULL
                     
                     #D
                     seuratreactive$plot.data.doublets <- NULL
                     seuratreactive$Remove <- NULL
                     seuratreactive$Removed <- NULL
                     seuratreactive$doublesim <- NULL
                     
                     #QC2
                     seuratreactive$type2 <- NULL
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$ALRAL <- NULL
                     seuratreactive$npcDR <- NULL
                     
                     #DR2
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL  
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #SDE
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     shinyjs::hide("NextButtonLoad")
                     shinyjs::hide("NextButtonQC")
                     shinyjs::hide("NextButtonD")
                     shinyjs::hide("NextButtonCombine")
                     shinyjs::hide("NextButtonQC2")
                     shinyjs::hide("NextButtonD2")
                     shinyjs::hide("NextButtonDR")
                     shinyjs::hide("NextButtonDR2")
                     shinyjs::hide("NextButtonC")
                     shinyjs::hide("NextButtonCC")
                     shinyjs::hide("NextButtonDE")
                     shinyjs::hide("NextButtonSDE")
                     
                     #Load
                     if (input$iscollapseboxintro == TRUE) {
                       js$collapse("LoadBoxIntro")
                     }
                     shinyjs::hide("LoadBox1")
                     shinyjs::hide("LoadBox2")
                     shinyjs::hide("LoadBox3")
                     shinyjs::hide("LoadBox5")
                     shinyjs::hide("LoadBox4")
                     
                     
                     #QC
                     if(input$iscollapseboxQC == TRUE) {
                       js$collapse("QCBoxIntro")
                     }
                     shinyjs::hide("QCBox1")
                     shinyjs::hide("QCBox2")
                     shinyjs::hide("QCBox3")
                     shinyjs::hide("QCBox4")
                     shinyjs::enable("startbuttonQC")
                     
                     #D
                     if(input$iscollapseboxD == TRUE) {
                       js$collapse("DBoxIntro")
                     }
                     shinyjs::hide("DBox1")
                     shinyjs::hide("DBox2")
                     
                     
                     #Load2
                     if(input$iscollapseboxCombine == TRUE) {
                       js$collapse("CombineBoxIntro")
                     }
                     shinyjs::hide("CombineBox1")
                     shinyjs::hide("CombineBox2")
                     shinyjs::hide("CombineBox3")
                     shinyjs::hide("CombineBox4")
                     shinyjs::hide("CombineBox5")
                     
                     
                     #QC2
                     if(input$iscollapseboxQC2 == TRUE) {
                       js$collapse("QCBoxIntro2")
                     }
                     shinyjs::hide("QCBox12")
                     shinyjs::hide("QCBox22")
                     shinyjs::hide("QCBox32")
                     shinyjs::hide("QCBox42")
                     shinyjs::enable("startbuttonQC2")
                     
                     #D2
                     if(input$iscollapseboxD2 == TRUE) {
                       js$collapse("DBoxIntro2")
                     }
                     shinyjs::hide("DBox12")
                     shinyjs::hide("DBox22")
                     
                     
                     #DR
                     if(input$iscollapsebox1 == TRUE) {
                       js$collapse("DRBoxIntro")
                     }
                     shinyjs::hide("DRBox1")
                     shinyjs::hide("DRBox2")
                     shinyjs::hide("DRBox3")
                     shinyjs::hide("DRBox4")
                     shinyjs::enable("startbuttonDR")
                     
                     #DR2
                     if(input$iscollapseboxDR2 == TRUE) {
                       js$collapse("DRBoxIntro2")
                     }
                     shinyjs::hide("DRBox1DR2")
                     shinyjs::hide("DRBox2DR2")
                     shinyjs::hide("DRBox3DR2")
                     shinyjs::hide("DRBox4DR2")
                     shinyjs::hide("downloadMerged")
                     shinyjs::enable("startbuttonDR2")
                     
                     #C
                     if (input$iscollapsebox2 == TRUE) {
                       js$collapse("CBoxIntro")
                     }
                     shinyjs::hide("CBox1")
                     shinyjs::hide("CBox2")
                     shinyjs::hide("CBox3")
                     shinyjs::enable("startbuttonC")
                     
                     #CC
                     if (input$iscollapsebox3 == TRUE) {
                       js$collapse("CCBoxIntro")
                     }
                     shinyjs::hide("CCBox1")
                     shinyjs::hide("CCBox1a")
                     shinyjs::hide("CCBox1b")
                     shinyjs::hide("CCBox1c")
                     shinyjs::hide("CCBox2")
                     shinyjs::enable("startbuttonCC")
                     
                     #L
                     if (input$iscollapsebox4 == TRUE) {
                       js$collapse("LBoxIntro")
                     }
                     shinyjs::hide("LBox1")
                     shinyjs::hide("LBox2")
                     shinyjs::hide("LBox3")
                     shinyjs::hide("LBox4")
                     shinyjs::hide("LBox5")
                     shinyjs::hide("LBox6")
                     shinyjs::hide("LBox7")
                     shinyjs::hide("LBox8")
                     
                     #MEGENA
                     if (input$iscollapseboxMEGENA == TRUE) {
                       js$collapse("MEGENABoxIntro")
                     }
                     shinyjs::hide("MEGENABox1")
                     shinyjs::hide("MEGENABox2")
                     shinyjs::hide("MEGENABox3")
                     shinyjs::hide("MEGENABox4")
                     
                     #SCENIC
                     if (input$iscollapseboxSCENIC == TRUE) {
                       js$collapse("SCENICBoxIntro")
                     }
                     shinyjs::hide("SCENICBox1a")
                     shinyjs::hide("SCENICBox1b")
                     shinyjs::hide("SCENICBox1")
                     shinyjs::hide("SCENICBox2")
                     shinyjs::hide("SCENICBox3")
                     
                     #MAGMA
                     if (input$iscollapseboxMAGMA == TRUE) {
                       js$collapse("MAGMABoxIntro")
                     }
                     shinyjs::hide("MAGMABox1")
                     shinyjs::hide("MAGMABox1a")
                     shinyjs::hide("MAGMABox1b")
                     shinyjs::hide("MAGMABox2a")
                     shinyjs::hide("MAGMABox2b")
                     shinyjs::hide("MAGMABox3")
                     shinyjs::hide("MAGMABox4")
                     
                     #CellChat
                     if (input$iscollapseboxChat == TRUE) {
                       js$collapse("ChatBoxIntro")
                     }
                     shinyjs::hide("ChatBox1")
                     shinyjs::hide("ChatBox2")
                     shinyjs::hide("ChatBox3")
                     shinyjs::hide("ChatBox4")
                     shinyjs::hide("ChatBox5")
                     shinyjs::hide("ChatBox6")
                     shinyjs::hide("ChatBox7")
                     shinyjs::hide("ChatBox8")
                     shinyjs::hide("ChatBox9")
                     shinyjs::hide("ChatBox10")
                     shinyjs::hide("ChatBox11")
                     shinyjs::hide("ChatBox12")
                     shinyjs::hide("ChatBox13")
                     
                     #DRUG
                     if (input$iscollapseboxDRUG == TRUE) {
                       js$collapse("DRUGBoxIntro")
                     }
                     shinyjs::hide("DRUGBox1")
                     shinyjs::hide("DRUGBox2")
                     
                     #DRUG2
                     if (input$iscollapseboxDRUG2 == TRUE) {
                       js$collapse("DRUGBoxIntro2")
                     }
                     shinyjs::hide("DRUGBox12")
                     shinyjs::hide("DRUGBox22")
                     
                     #GE
                     if (input$iscollapsebox5 == TRUE) {
                       js$collapse("GEBoxIntro")
                     }
                     shinyjs::hide("GEBox1")
                     shinyjs::hide("GEBox2")
                     shinyjs::hide("GEBox3")
                     shinyjs::hide("GEBox4")
                     
                     #TA
                     if (input$iscollapseboxTA == TRUE) {
                       js$collapse("TABoxIntro")
                     }
                     shinyjs::hide("TABox1")
                     shinyjs::hide("TABox2")
                     shinyjs::hide("TABox3")
                     shinyjs::hide("TABox4")
                     shinyjs::hide("TABox5")
                     shinyjs::hide("TABox6")
                     shinyjs::hide("TABox7")
                     shinyjs::hide("TABox8")
                     shinyjs::hide("TABox9")
                     shinyjs::hide("TABox10")
                     
                     #MM
                     if (input$iscollapseboxMM == TRUE) {
                       js$collapse("MMBoxIntro")
                     }
                     if (input$iscollapseboxMM2 == TRUE) {
                       js$collapse("MMBox4")
                     }
                     
                     if (input$iscollapseboxMM3 == TRUE) {
                       js$collapse("MMBox5")
                     }
                     shinyjs::hide("MMBox1")
                     shinyjs::hide("MMBox2")
                     shinyjs::hide("MMBox3")
                     shinyjs::hide("MMBox4")
                     shinyjs::hide("MMBox5")
                     
                     #DE
                     if (input$iscollapsebox6 == TRUE) {
                       js$collapse("DEBoxIntro")
                     }
                     shinyjs::hide("DEBox1")
                     shinyjs::hide("DEBox1a")
                     shinyjs::hide("DEBox1b")
                     shinyjs::hide("DEBox2")
                     shinyjs::hide("DEBox3")
                     shinyjs::hide("DEBox4")
                     shinyjs::hide("DEBox5")
                     shinyjs::hide("DEBox6")
                     shinyjs::enable("startbuttonDE")
                     
                     #SDE
                     if (input$iscollapseboxSDE == TRUE) {
                       js$collapse("SDEBoxIntro")
                     }
                     shinyjs::hide("SDEBox1")
                     shinyjs::hide("SDEBox2")
                     shinyjs::hide("SDEBox3")
                     shinyjs::hide("SDEBox4")
                     shinyjs::hide("SDEBox5")
                     shinyjs::hide("SDEBox6")
                     shinyjs::hide("SDEBox7")
                     shinyjs::enable("startbuttonSDE")
                     
                     #PA  
                     if (input$iscollapsebox7 == TRUE) {
                       js$collapse("PABoxIntro")
                     }
                     shinyjs::hide("PABox1")
                     shinyjs::hide("PABox2")
                     shinyjs::hide("PABox3")
                     shinyjs::hide("PABox4")
                     shinyjs::hide("PABox5")
                     
                     #SPA
                     if (input$iscollapseboxSPA == TRUE) {
                       js$collapse("SPABoxIntro")
                     }
                     shinyjs::hide("SPABox1")
                     shinyjs::hide("SPABox2")
                     shinyjs::hide("SPABox3")
                     shinyjs::hide("SPABox4")
                     shinyjs::hide("SPABox5")
                     
                     load(file$datapath)
                     
                     seuratreactive$GeneR <- GeneR
                     seuratreactive$msig <- msigdb
                     seuratreactive$orgDb <- orgDb
                     seuratreactive$EKEGG <- KEGG_species
                     seuratreactive$REACT <- REACTOME_species
                     seuratreactive$PAREACTOME <- REACTOME_species2
                     
                     #Load
                     seuratreactive$mincells <- mincells
                     seuratreactive$minfeatures <- minfeatures
                     
                     #Load2
                     seuratreactive$mincellsCombine <- mincellsCombine
                     seuratreactive$minfeaturesCombine <- minfeaturesCombine
                     
                     #QC
                     if (exists("PostQC_dataset1")) {
                       
                       if (!is.null(PostQC_dataset1)) {
                         
                         if(input$iscollapseboxQC == FALSE) {
                           js$collapse("QCBoxIntro")
                         }
                         shinyjs::show("QCBox1")
                         shinyjs::show("QCBox2")
                         shinyjs::show("QCBox3")
                         shinyjs::show("QCBox4")
                         shinyjs::disable("startbuttonQC")
                         shinyjs::show("NextButtonQC")
                         
                         seuratreactive$objQC <- PostQC_dataset1
                         seuratreactive$type <- type
                         seuratreactive$obj_original <- original_dataset1
                         seuratreactive$nFeature1 <- QC_nFeature_threshold1_dataset1
                         seuratreactive$nFeature2 <- QC_nFeature_threshold2_dataset1
                         seuratreactive$nCount1 <- QC_nCount_threshold1_dataset1
                         seuratreactive$nCount2 <- QC_nCount_threshold2_dataset1
                         seuratreactive$percent.mt1 <- QC_percent.mt_threshold1_dataset1
                         seuratreactive$percent.mt2 <- QC_percent.mt_threshold2_dataset1
                         seuratreactive$percent.ribo1 <- QC_percent.ribo_threshold1_dataset1
                         seuratreactive$percent.ribo2 <- QC_percent.ribo_threshold2_dataset1
                       } 
                     }
                     
                     #D
                     if (exists("Removed")) {
                       
                       if (!is.null(Removed)) {
                         
                         if(input$iscollapseboxD == FALSE) {
                           js$collapse("DBoxIntro")
                         }
                         shinyjs::show("DBox1")
                         shinyjs::show("DBox2")
                         shinyjs::show("DBox3")
                         shinyjs::show("DBox4")
                         shinyjs::show("NextButtonD")
                         shinyjs::disable("startbuttonD")
                         
                         seuratreactive$Remove <- Remove
                         seuratreactive$Removed <- Removed
                         seuratreactive$doublesim <- doublesim
                         seuratreactive$plot.data.doublets <- plot.data.doublets
                       }
                     }
                     
                     #QC2
                     if (exists("PostQC_dataset2")) {
                       
                       if (!is.null(PostQC_dataset2)) {
                         
                         if(input$iscollapseboxQC2 == FALSE) {
                           js$collapse("QCBoxIntro2")
                         }
                         shinyjs::show("QCBox12")
                         shinyjs::show("QCBox22")
                         shinyjs::show("QCBox32")
                         shinyjs::show("QCBox42")
                         shinyjs::disable("startbuttonQC2")
                         shinyjs::show("NextButtonQC2")
                         
                         shinyjs::hide("NextButtonQC")
                         
                         seuratreactive$objQC2 <- PostQC_dataset2
                         seuratreactive$type2 <- type2
                         seuratreactive$obj_original2 <- original_dataset2
                         seuratreactive$nFeature12 <- QC_nFeature_threshold1_dataset2
                         seuratreactive$nFeature22 <- QC_nFeature_threshold2_dataset2
                         seuratreactive$nCount12 <- QC_nCount_threshold1_dataset2
                         seuratreactive$nCount22 <- QC_nCount_threshold2_dataset2
                         seuratreactive$percent.mt12 <- QC_percent.mt_threshold1_dataset2
                         seuratreactive$percent.mt22 <- QC_percent.mt_threshold2_dataset2
                         seuratreactive$percent.ribo12 <- QC_percent.ribo_threshold1_dataset2
                         seuratreactive$percent.ribo22 <- QC_percent.ribo_threshold2_dataset2
                       }
                     }
                     
                     #D2
                     if (exists("Removed2")) {
                       if (!is.null(Removed2)) {
                         
                         if(input$iscollapseboxD2 == FALSE) {
                           js$collapse("DBoxIntro2")
                         }
                         shinyjs::show("DBox12")
                         shinyjs::show("DBox22")
                         shinyjs::disable("startbuttonD2")
                         shinyjs::show("NextButtonD2")
                         
                         shinyjs::hide("NextButtonQC2")
                         
                         seuratreactive$Remove2 <- Remove2
                         seuratreactive$Removed2 <- Removed2
                         seuratreactive$doublesim2 <- doublesim2
                         seuratreactive$plot.data.doublets2 <- plot.data.doublets2
                       }
                     }
                     
                     #DR
                     if (exists("variablefeatures")) {
                       if (!is.null(variablefeatures)) {
                         
                         if(input$iscollapsebox1 == FALSE) {
                           js$collapse("DRBoxIntro")
                         }
                         shinyjs::show("DRBox1")
                         shinyjs::show("DRBox2")
                         shinyjs::show("DRBox3")
                         shinyjs::show("DRBox4")
                         shinyjs::show("NextButtonDR")
                         
                         shinyjs::hide("NextButtonQC")
                         
                         seuratreactive$obj <- Final_Seurat_Object
                         seuratreactive$variablefeatures <- variablefeatures
                         seuratreactive$SCTransformDR <- SCTransformDR
                         seuratreactive$npcDR <- npcDR
                         seuratreactive$ALRAL <- ALRAL
                       }
                     }
                     
                     #DR2
                     if (exists("VFintegrated")) {
                       if (!is.null(VFintegrated)) {
                         
                         if (input$iscollapseboxDR2 == FALSE) {
                           js$collapse("DRBoxIntro2")
                         }
                         shinyjs::show("DRBox1DR2")
                         shinyjs::show("DRBox2DR2")
                         shinyjs::show("DRBox3DR2")
                         shinyjs::show("DRBox4DR2")
                         shinyjs::show("downloadMerged")
                         shinyjs::show("NextButtonDR2")
                         
                         shinyjs::hide("NextButtonD2")
                         
                         shinyjs::hide("NextButtonQC2")
                         
                         seuratreactive$obj <- Final_Seurat_Object
                         seuratreactive$Combined.list <- Combined.list
                         seuratreactive$VFintegrated <- VFintegrated
                         seuratreactive$integrationreduction <- integrationreduction
                         seuratreactive$kanchor <- kanchor
                         seuratreactive$SCTransformDR2 <- SCTransformDR2
                         seuratreactive$npcDR2 <- npcDR2
                       }
                     }
                     
                     #C
                     if (exists("NumberofPCs_GraphConstruction")) {
                       
                       if (!is.null(NumberofPCs_GraphConstruction)) {
                         
                         if (input$iscollapsebox2 == FALSE) {
                           js$collapse("CBoxIntro")
                         }
                         shinyjs::show("CBox1")
                         shinyjs::show("CBox2")
                         shinyjs::show("CBox3")
                         shinyjs::disable("startbuttonC")
                         shinyjs::show("NextButtonC")
                         
                         shinyjs::hide("NextButtonDR")
                         shinyjs::hide("NextButtonDR2")
                         
                         seuratreactive$integera <- NumberofPCs_GraphConstruction
                         seuratreactive$integerb <- k.param
                         seuratreactive$integerc <- n.trees
                         seuratreactive$integerd <- Resolution
                         seuratreactive$integere <- NumberofPCs_UMAP
                         seuratreactive$integerf <- knn_UMAP
                         seuratreactive$integerg <- min.dist_UMAP
                         seuratreactive$select1 <- ClusteringAlgorithm
                         seuratreactive$tsne1 <- NumberofPCs_tsne
                         seuratreactive$tsne2 <- Perplexity_tsne
                         seuratreactive$tsne3 <- max.iterations_tsne
                         seuratreactive$Identifiers <- CellComposition
                         seuratreactive$plot.data <- plot.data
                         seuratreactive$plot.data.original <- plot.data.original
                       }
                     }
                     
                     #CC
                     if (exists("Regress_Confounder")) {
                       if (!is.null(Regress_Confounder)) {
                         
                         if (input$iscollapsebox3 == FALSE) {
                           js$collapse("CCBoxIntro")
                         }
                         shinyjs::show("CCBox1")
                         shinyjs::show("CCBox2")
                         shinyjs::disable("startbuttonCC")
                         shinyjs::show("NextButtonCC")
                         
                         shinyjs::hide("NextButtonC")
                         
                         seuratreactive$reductionCC1 <- Regress_Confounder
                         seuratreactive$reductionCC2 <- Regress_genes
                         seuratreactive$reductionCC3 <- Regress_pathway
                       }
                     }
                     
                     #L
                     if (exists("new.cell.ids")) {
                       if (!is.null(new.cell.ids)) {
                         
                         if (input$iscollapsebox4 == FALSE) {
                           js$collapse("LBoxIntro")
                         }
                         shinyjs::show("LBox1")
                         shinyjs::show("LBox2")
                         shinyjs::show("LBox3")
                         shinyjs::show("LBox7")
                         shinyjs::show("LBox8")
                         
                         shinyjs::hide("NextButtonCC")
                         
                         seuratreactive$new.cell.ids <- new.cell.ids
                         seuratreactive$reference_annotation <- reference_annotation
                         seuratreactive$select2 <- select2
                         seuratreactive$selecttissue <- selecttissue
                       }
                     }
                     
                     #MEGENA
                     if (exists("summary.output")) {
                       
                       if (!is.null(summary.output)) {
                         
                         if (input$iscollapseboxMEGENA == FALSE) {
                           js$collapse("MEGENABoxIntro")
                         }
                         shinyjs::show("MEGENABox1")
                         shinyjs::show("MEGENABox2")
                         shinyjs::show("MEGENABox3")
                         shinyjs::show("MEGENABox4")
                         
                         seuratreactive$MEGENA_samples <- MEGENA_samples
                         seuratreactive$variablegenesMEGENA <- variablegenesMEGENA
                         seuratreactive$MEGENA_selectmethod <- MEGENA_selectmethod
                         seuratreactive$MEGENApval <- MEGENApval
                         seuratreactive$MEGENAmin.size <- MEGENAmin.size
                         seuratreactive$MEGENA_selectcluster <- MEGENA_selectcluster
                         seuratreactive$n_MEGENA <- n_MEGENA
                         seuratreactive$log2FCMEGENA <- log2FCMEGENA
                         seuratreactive$padjustMEGENA <- padjustMEGENA
                         seuratreactive$testuseMEGENA <- testuseMEGENA
                         seuratreactive$minMEGENA <- minMEGENA
                         seuratreactive$pvalMEGENA <- pvalMEGENA
                         seuratreactive$summary.output <- summary.output
                         seuratreactive$markersMEGENA <- markersMEGENA
                         seuratreactive$moduleGO_df <- moduleGO_df
                         seuratreactive$VF <- VF
                         seuratreactive$g <- g
                       }
                     }
                     
                     #SCENIC
                     if (exists("regulons")) {
                       
                       if (!is.null(regulons)) {
                         
                         if (input$iscollapseboxSCENIC == FALSE) {
                           js$collapse("SCENICBoxIntro")
                         }
                         shinyjs::show("SCENICBox1a")
                         shinyjs::show("SCENICBox1b")
                         shinyjs::show("SCENICBox1")
                         shinyjs::show("SCENICBox2")
                         shinyjs::show("SCENICBox3")
                         
                         seuratreactive$variablefeaturesSCENIC <- variablefeaturesSCENIC
                         seuratreactive$variablegenesSCENIC <- variablegenesSCENIC
                         seuratreactive$regulonAUC <- regulonAUC
                         seuratreactive$regulons <- regulons
                         seuratreactive$cellInfo <- cellInfo
                         seuratreactive$n_SCENIC <- n_SCENIC
                         seuratreactive$SCENICdb <- SCENICdb
                         seuratreactive$SCENIC_samples <- SCENIC_samples
                         seuratreactive$regulonnames <- regulonnames
                         seuratreactive$titleSCENIC <- titleSCENIC
                         seuratreactive$SCENICInput <- SCENICInput
                         seuratreactive$SCENIC_selectcluster <- SCENIC_selectcluster
                         seuratreactive$SCENIC_selectmethod <- SCENIC_selectmethod
                         seuratreactive$regulonActivity_byCellType_Scaled <- regulonActivity_byCellType_Scaled
                         seuratreactive$rss <- rss
                       }
                     }
                     
                     #MAGMA
                     if (exists("MAGMA_Final_Results")) {
                       
                       if (!is.null(MAGMA_Final_Results)) {
                         
                         if (input$iscollapseboxMAGMA == FALSE) {
                           js$collapse("MAGMABoxIntro")
                         }
                         shinyjs::show("MAGMABox1")
                         shinyjs::hide("MAGMABox1a")
                         shinyjs::hide("MAGMABox1b")
                         shinyjs::hide("MAGMABox2a")
                         shinyjs::show("MAGMABox2b")
                         shinyjs::show("MAGMABox3")
                         shinyjs::show("MAGMABox4")
                         
                         seuratreactive$MAGMA_samples <- MAGMA_samples
                         seuratreactive$MAGMA_GWAS <- MAGMA_GWAS
                         seuratreactive$MAGMA_Final_Results <- MAGMA_Final_Results
                         seuratreactive$meta <- meta
                         seuratreactive$MAGMA_selectcluster <- MAGMA_selectcluster
                         seuratreactive$MAGMA_selectcluster2 <- MAGMA_selectcluster2
                         seuratreactive$MAGMA_samples2 <- MAGMA_samples2
                         seuratreactive$textlabelMAGMA <- textlabelMAGMA
                         seuratreactive$MAGMA_N <- MAGMA_N
                         seuratreactive$GRCh <- GRCh
                         seuratreactive$MAGMA_Pop <- MAGMA_Pop
                         seuratreactive$n_MAGMA <- n_MAGMA
                         seuratreactive$n_MAGMA2 <- n_MAGMA2
                       }
                     }
                     
                     #TA
                     if (exists("TAselectedcells")) {
                       
                       if (!is.null(TAselectedcells)) {
                         
                         if (input$iscollapseboxTA == FALSE) {
                           js$collapse("TABoxIntro")
                         }
                         shinyjs::show("TABox1")
                         shinyjs::show("TABox4")
                         shinyjs::show("TABox5")
                         shinyjs::show("TABox6")
                         shinyjs::show("TABox7")
                         shinyjs::show("TABox8")
                         shinyjs::show("TABox9")
                         shinyjs::show("TABox10")
                         shinyjs::hide("TABox2")
                         shinyjs::hide("TABox3")
                         
                         seuratreactive$TAselectedcells <- TAselectedcells
                         seuratreactive$TA_pickercluster <- TA_pickercluster
                         seuratreactive$seurat_monocle <- seurat_monocle
                         seuratreactive$Monocle3_genes <- Monocle3_genes
                       }
                     }
                     
                     #CellChat
                     if (exists("cellchat") && exists("cellchatM")) {
                       
                       if (!is.null(cellchat) | !is.null(cellchatM)) {
                         
                         if (!is.null(cellchat)) {
                           if (input$iscollapseboxChat == FALSE) {
                             js$collapse("ChatBoxIntro")
                           }
                           shinyjs::show("ChatBox1")
                           shinyjs::show("ChatBox2")
                           shinyjs::show("ChatBox3")
                           shinyjs::show("ChatBox4")
                           shinyjs::show("ChatBox5")
                           shinyjs::show("ChatBox6")
                           shinyjs::hide("ChatBox7")
                           shinyjs::hide("ChatBox8")
                           shinyjs::hide("ChatBox9")
                           shinyjs::hide("ChatBox10")
                           shinyjs::hide("ChatBox11")
                           shinyjs::hide("ChatBox12")
                           shinyjs::hide("ChatBox13")
                         }
                         else {
                           if (input$iscollapseboxChat == FALSE) {
                             js$collapse("ChatBoxIntro")
                           }
                           shinyjs::hide("ChatBox1")
                           shinyjs::hide("ChatBox2")
                           shinyjs::hide("ChatBox3")
                           shinyjs::hide("ChatBox4")
                           shinyjs::hide("ChatBox5")
                           shinyjs::hide("ChatBox6")
                           shinyjs::show("ChatBox7")
                           shinyjs::show("ChatBox8")
                           shinyjs::show("ChatBox9")
                           shinyjs::show("ChatBox10")
                           shinyjs::show("ChatBox11")
                           shinyjs::show("ChatBox12")
                           shinyjs::show("ChatBox13")
                         }
                         
                         seuratreactive$cellchat <- cellchat
                         seuratreactive$cellchatM <- cellchatM
                         seuratreactive$object.list <- object.list
                         seuratreactive$ChatChoose <- ChatChoose
                         seuratreactive$ChatSample <- ChatSample
                         seuratreactive$Chatdb <- Chatdb
                         seuratreactive$ChatMultiple1Sample <- ChatMultiple1Sample
                         seuratreactive$ChatMultiple2Sample <- ChatMultiple2Sample
                         seuratreactive$Chat_selectcluster <- Chat_selectcluster
                         seuratreactive$Chat_number <- Chat_number
                         seuratreactive$Chatpval <- Chatpval
                         
                       }
                     }
                     
                     #MM
                     if (exists("MM")) {
                       
                       if (!is.null(MM)) {
                         
                         if (input$iscollapseboxMM == FALSE) {
                           js$collapse("MMBoxIntro")
                         }
                         if (input$iscollapseboxMM2 == FALSE) {
                           js$collapse("MMBox4")
                         }
                         
                         if (input$iscollapseboxMM3 == FALSE) {
                           js$collapse("MMBox5")
                         }
                         
                         shinyjs::show("MMBox1")
                         shinyjs::show("MMBox2")
                         shinyjs::show("MMBox3")
                         shinyjs::show("MMBox4")
                         shinyjs::show("MMBox5")
                         
                         seuratreactive$MM <- MM
                       }
                     }
                     
                     #DE
                     if (exists("DifferentialExpression_GeneSetEnrichment")) {
                       
                       if (!is.null(DifferentialExpression_GeneSetEnrichment)) {
                         
                         if (input$iscollapsebox6 == FALSE) {
                           js$collapse("DEBoxIntro")
                         }
                         shinyjs::show("DEBox1")
                         shinyjs::show("DEBox2")
                         shinyjs::show("DEBox3")
                         shinyjs::show("DEBox4")
                         shinyjs::show("DEBox5")
                         shinyjs::show("DEBox6")
                         shinyjs::disable("startbuttonDE")
                         
                         if (exists("markersDE") && exists("markersDRUG")) {
                           if (!is.null(markersDE) && !is.null(markersDRUG)) {
                             shinyjs::show("NextButtonDE")
                           }
                           else {
                             shinyjs::hide("NextButtonDE")
                           }
                         }
                         
                         seuratreactive$onlyposDE <- onlyposDE
                         seuratreactive$DE_clusters_samples <- DifferentialExpression_clusters_BetweenSampleComparison
                         seuratreactive$DECluster <- DifferentialExpression_ClusterOfInterest_BetweenSampleComparison
                         seuratreactive$DEInput1_samples <- DifferentialExpression_SampleOfInterest1_BetweenSampleComparison
                         seuratreactive$DEInput2_samples <- DifferentialExpression_SampleOfInterest2_BetweenSampleComparison
                         seuratreactive$SelectComparison <- DifferentialExpression_comparison
                         seuratreactive$DE_clusters <- DifferentialExpression_clusters_BetweenClustersComparison
                         seuratreactive$DEInput1 <- DifferentialExpression_SampleOfInterest1_BetweenClustersComparison
                         seuratreactive$DEInput2 <- DifferentialExpression_SampleOfInterest2_BetweenClustersComparison
                         seuratreactive$logDE <- DifferentialExpression_log_threshold
                         seuratreactive$minDE <- DifferentialExpression_min.pct_threshold
                         seuratreactive$padjustDE1 <- padjustDE1
                         seuratreactive$testuseDE <- testuseDE
                         seuratreactive$enrichInputDE <- DifferentialExpression_GeneSetEnrichmentInput
                         seuratreactive$GOlabelsDE <- GOlabelsDE 
                         seuratreactive$pvalueDE <- DifferentialExpression_GeneSetEnrichment_pvalue_threshold
                         seuratreactive$pvalDE <- pvalDE
                         seuratreactive$padjustDE <- DifferentialExpression_GeneSetEnrichment_padjust_method
                         seuratreactive$volcanovalue1 <- DifferentialExpression_volcano_Foldchange_threshold1
                         seuratreactive$volcanovalue2 <- DifferentialExpression_volcano_Foldchange_threshold2
                         seuratreactive$pvaluevolcano <- DifferentialExpression_volcano_pvalue
                         seuratreactive$markersDE <- DifferentialExpression_Table
                         seuratreactive$YuDE <- DifferentialExpression_GeneSetEnrichment
                         seuratreactive$markersDEvolcano <- DifferentialExpression_volcano_markers
                         seuratreactive$PATHWAY <- DifferentialExpression_GeneSetEnrichment_PATHWAY
                       }
                     }
                     
                     #PA
                     if (exists("PathwayAnalysis_GeneSetEnrichment")) {
                       
                       if (!is.null(PathwayAnalysis_GeneSetEnrichment)) {
                         
                         if (input$iscollapsebox7 == FALSE) {
                           js$collapse("PABoxIntro")
                         }
                         shinyjs::show("PABox1")
                         shinyjs::show("PABox2")
                         shinyjs::show("PABox3")
                         shinyjs::show("PABox4")
                         shinyjs::show("PABox5")
                         
                         shinyjs::disable("startbuttonDE")
                         shinyjs::hide("NextButtonDE")
                         
                         seuratreactive$PAchoose <- SelectPathwayAnalysis
                         seuratreactive$padjustPA <- PathwayAnalysis_pvalue_adjustment_method
                         seuratreactive$pvaluePA <- PathwayAnalysis_pvalue
                         seuratreactive$GOlabelsPA <- GOlabelsPA
                         Yu$result <- PathwayAnalysis_GeneSetEnrichment
                         Yu$ema <- PathwayAnalysis_GeneSetEnrichment2
                         seuratreactive$geneListPA <- geneListPathway
                       }
                     }
                     
                     #DRUG
                     if (exists("markersDRUG")) {
                       if (!is.null(markersDRUG)) {
                         
                         seuratreactive$markersDRUG <- markersDRUG
                         seuratreactive$DRUGdf <- DRUGdf
                         
                         if (input$iscollapseboxDRUG == FALSE) {
                           js$collapse("DRUGBoxIntro")
                         }
                         shinyjs::show("DRUGBox1")
                         shinyjs::show("DRUGBox2")
                         
                         shinyjs::hide("NextButtonDE")
                       }
                     }
                     
                     #SDE
                     if (exists("CustomDifferentialExpression_GeneSetEnrichment")) {
                       
                       if (!is.null(CustomDifferentialExpression_GeneSetEnrichment)) {
                         
                         if (input$iscollapseboxSDE == FALSE) {
                           js$collapse("SDEBoxIntro")
                         }
                         shinyjs::show("SDEBox1")
                         shinyjs::show("SDEBox2")
                         shinyjs::show("SDEBox3")
                         shinyjs::show("SDEBox4")
                         shinyjs::show("SDEBox5")
                         shinyjs::show("SDEBox6")
                         shinyjs::show("SDEBox7")
                         shinyjs::disable("startbuttonSDE")
                         
                         if (exists("markersSDE") && exists("markersDRUG2")) {
                           if (!is.null(markersSDE) && !is.null(markersDRUG2)) {
                             shinyjs::show("NextButtonSDE")
                           }
                           else {
                             shinyjs::hide("NextButtonSDE")
                           }
                         }
                         
                         seuratreactive$onlyposSDE <- onlyposSDE
                         seuratreactive$selectedkey1 <- selectedcells_lasso1
                         seuratreactive$selectedkey2 <- selectedcells_lasso2
                         seuratreactive$GOlabelsSDE <- GOlabelsSDE
                         seuratreactive$logSDE <- CustomDifferentialExpression_log_threshold
                         seuratreactive$minSDE <- CustomDifferentialExpression_min.pct_threshold
                         seuratreactive$padjustSDE1 <- padjustSDE1
                         seuratreactive$testuseSDE <- testuseSDE
                         seuratreactive$enrichInputSDE <- CustomDifferentialExpression_GeneSetEnrichmentInput
                         seuratreactive$pvalueSDE <- CustomDifferentialExpression_GeneSetEnrichment_pvalue_threshold
                         seuratreactive$pvalSDE <- pvalSDE
                         seuratreactive$padjustSDE <- CustomDifferentialExpression_GeneSetEnrichment_padjust_method
                         seuratreactive$volcanoSDEvalue1 <- CustomDifferentialExpression_volcano_Foldchange_threshold1
                         seuratreactive$volcanoSDEvalue2 <- CustomDifferentialExpression_volcano_Foldchange_threshold2
                         seuratreactive$pvaluevolcanoSDE <- CustomDifferentialExpression_volcano_pvalue
                         seuratreactive$markersSDE <- CustomDifferentialExpression_Table
                         seuratreactive$YuSDE <- CustomDifferentialExpression_GeneSetEnrichment
                         seuratreactive$markersSDEvolcano <- CustomDifferentialExpression_volcano_markers
                         seuratreactive$PATHWAYSDE <- CustomDifferentialExpression_GeneSetEnrichment_PATHWAY
                       }
                     }
                     
                     #SPA
                     if (exists("CustomPathwayAnalysis_GeneSetEnrichment")) {
                       
                       if (!is.null(CustomPathwayAnalysis_GeneSetEnrichment)) {
                         
                         if (input$iscollapseboxSPA == FALSE) {
                           js$collapse("SPABoxIntro")
                         }
                         shinyjs::show("SPABox1")
                         shinyjs::show("SPABox2")
                         shinyjs::show("SPABox3")
                         shinyjs::show("SPABox4")
                         shinyjs::show("SPABox5")
                         
                         shinyjs::disable("startbuttonSDE")
                         shinyjs::hide("NextButtonSDE")
                         
                         seuratreactive$SPAchoose <- SelectCustomPathwayAnalysis
                         seuratreactive$padjustSPA <- CustomPathwayAnalysis_pvalue_adjustment_method
                         seuratreactive$GOlabelsSPA <- GOlabelsSPA
                         seuratreactive$pvalueSPA <- CustomPathwayAnalysis_pvalue
                         YuSPA$result <- CustomPathwayAnalysis_GeneSetEnrichment
                         YuSPA$ema <- CustomPathwayAnalysis_GeneSetEnrichment2
                         seuratreactive$geneListSPA <- geneList_CustomPathway
                       }
                     }
                     
                     #DRUG2
                     if (exists("markersDRUG2") && exists("DRUGdf2")) {
                       if (!is.null(markersDRUG2) && !is.null(DRUGdf2)) {
                         
                         if (input$iscollapseboxDRUG2 == FALSE) {
                           js$collapse("DRUGBoxIntro2")
                         }
                         shinyjs::show("DRUGBox12")
                         shinyjs::show("DRUGBox22")
                         
                         shinyjs::hide("NextButtonSDE")
                         
                         seuratreactive$markersDRUG2 <- markersDRUG2
                         seuratreactive$DRUGdf2 <- DRUGdf2
                       }
                     }
                     
                     
                     if (!is.null(seuratreactive$objQC)) {
                       
                       incProgress(0.8)
                       
                       updateNumericInput(session, 'nFeature1', value = paste(seuratreactive$nFeature1))
                       updateNumericInput(session, 'nFeature2', value = paste(seuratreactive$nFeature2))
                       updateNumericInput(session, 'nCount1', value = paste(seuratreactive$nCount1))
                       updateNumericInput(session, 'nCount2', value = paste(seuratreactive$nCount2))
                       updateNumericInput(session, 'percent.mt1', value = paste(seuratreactive$percent.mt1))
                       updateNumericInput(session, 'percent.mt2', value = paste(seuratreactive$percent.mt2))
                       updateNumericInput(session, 'percent.ribo1', value = paste(seuratreactive$percent.ribo1))
                       updateNumericInput(session, 'percent.ribo2', value = paste(seuratreactive$percent.ribo2))
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = T),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                     if (!is.null(seuratreactive$Removed)) {
                       
                       updateVirtualSelect("doublesim", selected = seuratreactive$doublesim)
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = T),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$objQC2)) {
                       
                       updateNumericInput(session, 'nFeature12', value = paste(seuratreactive$nFeature12))
                       updateNumericInput(session, 'nFeature22', value = paste(seuratreactive$nFeature22))
                       updateNumericInput(session, 'nCount12', value = paste(seuratreactive$nCount12))
                       updateNumericInput(session, 'nCount22', value = paste(seuratreactive$nCount22))
                       updateNumericInput(session, 'percent.mt12', value = paste(seuratreactive$percent.mt12))
                       updateNumericInput(session, 'percent.mt22', value = paste(seuratreactive$percent.mt22))
                       updateNumericInput(session, 'percent.ribo12', value = paste(seuratreactive$percent.ribo12))
                       updateNumericInput(session, 'percent.ribo22', value = paste(seuratreactive$percent.ribo22))
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = T),
                                                                        menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$Removed2)) {
                       
                       updateVirtualSelect("doublesim2", selected = seuratreactive$doublesim2)
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                        menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = T),
                                                                        menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$variablefeatures)) {
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                        menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = T),
                                                                        menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                       incProgress(0.8)
                     }
                     
                     if(!is.null(seuratreactive$VFintegrated)) {
                       
                       if (!is.null(seuratreactive$objQC2)) {
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                          menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       else {
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       incProgress(0.8)
                     }
                     
                     
                     if (!is.null(seuratreactive$integera)) {
                       
                       updateNumericInput(session, 'integera', value = seuratreactive$integera)
                       updateNumericInput(session, 'integerb', value = seuratreactive$integerb)
                       updateNumericInput(session, 'integerc', value = seuratreactive$integerc)
                       updateNumericInput(session, 'integerd', value = seuratreactive$integerd)
                       updatePickerInput(session, 'select1', selected = seuratreactive$select1)
                       updateNumericInput(session, 'integere', value = seuratreactive$integere)
                       updateNumericInput(session, 'integerf', value = seuratreactive$integerf)
                       updateNumericInput(session, 'integerg', value = seuratreactive$integerg)
                       updateNumericInput(session, 'tsne1', value = seuratreactive$tsne1)
                       updateNumericInput(session, 'tsne2', value = seuratreactive$tsne2)
                       updateNumericInput(session, 'tsne3', value = seuratreactive$tsne3)
                       
                       if (is.null(seuratreactive$objQC2)){
                         if (is.null(seuratreactive$VFintegrated)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                           )
                           )
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                           ))
                         }
                       }
                       else{
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                          menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T),
                                                                          menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$reductionCC1)) {
                       
                       updatePickerInput(session, 'CC1a', selected = "")
                       
                       if (is.null(seuratreactive$new.cell.ids) && is.null(seuratreactive$summary.output) && is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && is.null(seuratreactive$objGEyes) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$new.cell.ids)) {
                       
                       updatePickerInput(session, 'selectLabelInput', selected = "")  
                       
                       if (is.null(seuratreactive$summary.output) && is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && is.null(seuratreactive$objGEyes) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$summary.output)) {
                       
                       updateNumericInput(session, 'log2FCMEGENA', value = seuratreactive$log2FCMEGENA)
                       updatePickerInput(session, 'pajustMEGENA', selected = seuratreactive$pajustMEGENA)
                       updatePickerInput(session, 'testuseMEGENA', selected = seuratreactive$testuseMEGENA)
                       updateNumericInput(session, 'minMEGENA', value = seuratreactive$minMEGENA)
                       updateNumericInput(session, 'pvalMEGENA', value = seuratreactive$pvalMEGENA)
                       
                       if (is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = T),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = T),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = T),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$regulons)) {
                       
                       updateVirtualSelect('SCENICInput', selected = seuratreactive$SCENICInput)
                       
                       if (is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = T),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = T),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = T),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$MAGMA_Final_Results)) {
                       
                       updateVirtualSelect('MAGMA_samples', selected = seuratreactive$MAGMA_samples)
                       
                       updateVirtualSelect('MAGMA_samples2', selected = seuratreactive$MAGMA_samples2)
                       
                       updateVirtualSelect('MAGMA_GWAS', selected = seuratreactive$MAGMA_GWAS)
                       
                       updatePickerInput(session, 'MAGMA_selectcluster', selected = seuratreactive$MAGMA_selectcluster)
                       
                       updatePickerInput(session, 'MAGMA_selectcluster2', selected = seuratreactive$MAGMA_selectcluster2)
                       
                       updateTextInput(session, 'textlabelMAGMA', value = seuratreactive$textlabelMAGMA)
                       
                       updateNumericInput(session, 'MAGMA_N', value = seuratreactive$MAGMA_N)
                       
                       updatePickerInput(session, 'GRCh', selected = seuratreactive$GRCh)
                       
                       updatePickerInput(session, 'MAGMA_Pop', selected = seuratreactive$MAGMA_Pop)
                       
                       updateNumericInput(session, 'n_MAGMA', value = seuratreactive$n_MAGMA)
                       
                       updateNumericInput(session, 'n_MAGMA2', value = seuratreactive$n_MAGMA2)
                       
                       if (is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = T),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = T),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = T),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$TAselectedcells)) {
                       
                       updatePickerInput(session, "TAselect", selected = "")
                       
                       updateVirtualSelect('MonocleInput', choices = c(rownames(seuratreactive$obj)))
                       
                       if (is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = T),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = T),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = T),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$Chatdb)) {
                       if (is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = T),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = T),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = T),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$MM)) {
                       
                       if (is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
                         if (is.null(seuratreactive$objQC2)) {
                           if (is.null(seuratreactive$VFintegrated)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = T),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = T),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         
                         else if (!is.null(seuratreactive$objQC2)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = T),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     
                     if (!is.null(seuratreactive$markersDE)) {
                       
                       seuratreactive$title <- NULL
                       
                       updatePickerInput(session, 'SelectComparison', selected = seuratreactive$SelectComparison)
                       updatePickerInput(session, 'DE_clusters', selected = seuratreactive$DE_clusters)
                       updateVirtualSelect('DEInput1', selected = seuratreactive$DEInput1)
                       updateVirtualSelect('DEInput2', selected = seuratreactive$DEInput2)
                       
                       updatePickerInput(session, 'DE_clusters_samples', selected = seuratreactive$DE_clusters_samples)
                       updateVirtualSelect('DECluster', selected = seuratreactive$DECluster)
                       updateVirtualSelect('DEInput1_samples', selected = seuratreactive$DEInput1_samples)
                       updateVirtualSelect('DEInput2_samples', selected = seuratreactive$DEInput2_samples)
                       
                       updatePickerInput(session, 'GOlabelsDE', selected = paste(seuratreactive$GOlabelsDE))
                       updatePickerInput(session, 'padjustDE1', selected = paste(seuratreactive$padjustDE1))
                       updatePickerInput(session, 'testuseDE', selected = paste(seuratreactive$testuseDE))
                       updateNumericInput(session, 'logDE', value = paste(seuratreactive$logDE))
                       updateNumericInput(session, 'minDE', value = paste(seuratreactive$minDE))
                       updatePickerInput(session, 'enrichInputDE', selected = paste(seuratreactive$enrichInputDE))
                       updateNumericInput(session, 'pvalueDE', value = paste(seuratreactive$pvalueDE))
                       updateNumericInput(session, 'pvalDE', value = paste(seuratreactive$pvalDE))
                       updatePickerInput(session, 'padjustDE', selected = paste(seuratreactive$padjustDE))
                       updateSliderInput(session, 'logvolcano', value = c(paste(seuratreactive$volcanovalue1), paste(seuratreactive$volcanovalue2)))
                       updateNumericInput(session, 'pvolcano', value = paste(seuratreactive$pvaluevolcano))
                       updateMaterialSwitch(session, 'onlyposDE', value = seuratreactive$onlyposDE)
                       
                       if (is.null(seuratreactive$objQC2)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           if (!is.null(seuratreactive$markersSDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         else {
                           if (!is.null(seuratreactive$markersSDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                       }
                       
                       else {
                         if (!is.null(seuratreactive$markersSDE)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(Yu$result)) {
                       
                       if (is.null(seuratreactive$objQC2)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           if (!is.null(seuratreactive$markersSDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         else {
                           if (!is.null(seuratreactive$markersSDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                       }
                       
                       else {
                         if (!is.null(seuratreactive$markersSDE)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)
                     }
                     
                     if (!is.null(seuratreactive$markersDRUG)) {
                       
                       if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = T),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = T),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       
                       else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                          menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                          menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                          menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                          menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                          menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                          menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                          menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                          menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                          menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                          menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                          menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                          menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                          menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = T),
                                                                          menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                          menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                          menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                       }
                       
                       else if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = T),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = T),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       
                       else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                          menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                          menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                          menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                          menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                          menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                          menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                          menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                          menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                          menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                          menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                          menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                          menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                          menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = T),
                                                                          menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                       }
                     }
                     
                     if (!is.null(seuratreactive$markersSDE)) {
                       
                       seuratreactive$titleSDE <- NULL
                       
                       updateNumericInput(session, 'logSDE', value = paste(seuratreactive$logSDE))
                       updateNumericInput(session, 'minSDE', value = paste(seuratreactive$minSDE))
                       updatePickerInput(session, 'padjustSDE1', selected = paste(seuratreactive$padjustSDE1))
                       updatePickerInput(session, 'testuseSDE', selected = paste(seuratreactive$testuseSDE))
                       updatePickerInput(session, 'GOlabelsSDE', selected = paste(seuratreactive$GOlabelsSDE))
                       updatePickerInput(session, 'enrichInputSDE', selected = paste(seuratreactive$enrichInputSDE))
                       updateNumericInput(session, 'pvalueSDE', value = paste(seuratreactive$pvalueSDE))
                       updateNumericInput(session, 'pvalSDE', value = paste(seuratreactive$pvalSDE))
                       updatePickerInput(session, 'padjustSDE', selected = paste(seuratreactive$padjustSDE))
                       updateSliderInput(session, 'logvolcanoSDE', value = c(paste(seuratreactive$volcanoSDEvalue1), paste(seuratreactive$volcanoSDEvalue2)))
                       updateNumericInput(session, 'pvolcanoSDE', value = paste(seuratreactive$pvaluevolcanoSDE))
                       updateMaterialSwitch(session, 'onlyposSDE', value = seuratreactive$onlyposSDE)
                       
                       if (is.null(seuratreactive$objQC2)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           if (!is.null(seuratreactive$markersDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         else {
                           if (!is.null(seuratreactive$markersDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                       }
                       
                       else {
                         if (!is.null(seuratreactive$markersDE)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)                
                     } 
                     
                     if (!is.null(YuSPA$result)) {
                       
                       if (is.null(seuratreactive$objQC2)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           if (!is.null(seuratreactive$markersDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                         else {
                           if (!is.null(seuratreactive$markersDE)) {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                              menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                           else {
                             output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                              menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                              menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                              menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                              menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                              menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                              menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                              menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                              menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                              menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                              menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                              menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                              menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                              menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                              menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                              menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                              menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                              menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                              menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T),
                                                                              menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                              menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                           }
                         }
                       }
                       
                       else {
                         if (!is.null(seuratreactive$markersDE)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                            menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       incProgress(0.8)                 
                     }
                     
                     if (!is.null(seuratreactive$markersDRUG2)) {
                       
                       if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = T),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                            menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = T),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       else if (is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
                         if (is.null(seuratreactive$VFintegrated)) {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = T),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                         else {
                           output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                            menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                            menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                            menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                            menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                            menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                            menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                            menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                            menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                            menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                            menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                            menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                            menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                            menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                            menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                            menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                            menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                            menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                            menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                            menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = T),
                                                                            menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                         }
                       }
                       
                       else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                          menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                          menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                          menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                          menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                          menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                          menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                          menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                          menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                          menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                          menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                          menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                          menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                                          menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                                          menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                          menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                          menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = T),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                       }
                       else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                                          menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                                          menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                                          menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                                          menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                                          menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                                          menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                                          menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                                          menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                                          menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                                          menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                                          menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                                          menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                                          menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                                          menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                                          menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                                          menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = T),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
                       }
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  ####TAB1 - Load 1st dataset####
  
  observeEvent(input$startbuttonLoad, {
    if (input$iscollapseboxintro == FALSE) {
      js$collapse("LoadBoxIntro")
    }
    shinyjs::show("LoadBox1")
    shinyjs::hide("LoadBox2")
    shinyjs::hide("LoadBox3")
    shinyjs::hide("LoadBox5")
    shinyjs::show("LoadBox4")
    
    shinyjs::enable("startbuttonQC")
    shinyjs::hide("NextButtonQC")
  })
  
  observe({
    if (input$singlemultiple == "Single") {
      shinyjs::show("LoadBox2")
      shinyjs::hide("LoadBox5")
      shinyjs::reset("radioLoad")
    }
    else if (input$singlemultiple == "Multiple") {
      shinyjs::hide("LoadBox2")
      shinyjs::hide("LoadBox3")
      shinyjs::show("LoadBox5")
      shinyjs::reset("radioLoadM")
    }
  })
  
  observe({
    if (input$radioLoad == "10X data") {
      shinyjs::show("textdirectory")
      shinyjs::hide("fileinput1")
      shinyjs::hide("fileinput2")
      shinyjs::hide("existingdata")
      shinyjs::hide("buttonexistingdata")
      shinyjs::show("LoadBox3")
    }
    else if (input$radioLoad == "Seurat object (RDS file)") {
      shinyjs::hide("textdirectory")
      shinyjs::show("fileinput1")
      shinyjs::hide("fileinput2")
      shinyjs::hide("existingdata")
      shinyjs::hide("buttonexistingdata")
      shinyjs::show("LoadBox3")
    }
    
    else if (input$radioLoad == "Count Matrix (txt)") {
      shinyjs::hide("textdirectory")
      shinyjs::hide("fileinput1")
      shinyjs::show("fileinput2")
      shinyjs::hide("existingdata")
      shinyjs::hide("buttonexistingdata")
      shinyjs::show("LoadBox3")
    }
    
    else if (input$radioLoad == "Sample datasets") {
      shinyjs::hide("textdirectory")
      shinyjs::hide("fileinput1")
      shinyjs::hide("fileinput2")
      shinyjs::show("existingdata")
      shinyjs::show("buttonexistingdata")
      shinyjs::hide("LoadBox3")
    }
  })
  
  observe({
    if (input$radioLoadM == "Seurat object (RDS file)") {
      shinyjs::show("fileinput1M")
      shinyjs::hide("fileinput2M")
      shinyjs::hide("existingdataM")
      shinyjs::hide("buttonexistingdataM")
    }
    
    else if (input$radioLoadM == "Count Matrix (txt)") {
      shinyjs::hide("fileinput1M")
      shinyjs::show("fileinput2M")
      shinyjs::hide("existingdataM")
      shinyjs::hide("buttonexistingdataM")
    }
    
    else if (input$radioLoadM == "Sample datasets") {
      shinyjs::hide("fileinput1M")
      shinyjs::hide("fileinput2M")
      shinyjs::show("existingdataM")
      shinyjs::show("buttonexistingdataM")
    }
  })
  
  observe({
    if(is.null(input$existingdata) | !is.null(input$existingdata) && input$existingdata == "") {
      shinyjs::disable("buttonexistingdata")
    }
    else{
      shinyjs::enable("buttonexistingdata")
    }
  })
  
  observe({
    if(is.null(input$existingdataM) | !is.null(input$existingdataM) && input$existingdataM == "") {
      shinyjs::disable("buttonexistingdataM")
    }
    else{
      shinyjs::enable("buttonexistingdataM")
    }
  })
  
  observeEvent(input$tabs, {
    if (input$tabs == "Load") {
      updatePickerInput(session, 'singlemultiple', selected = "")  
    }
  })
  
  observeEvent(input$startbuttonLoad, {
    updatePickerInput(session, 'singlemultiple', selected = "") 
  })
  
  seuratreactive <- reactiveValues()
  
  observeEvent(input$fileinput1, {
    
    tryCatch({
      withProgress(message = 'Loading Seurat object...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     
                     file = input$fileinput1
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat_file <- readRDS(file$datapath)
                     
                     if (length(unique(seurat_file@meta.data$orig.ident)) > 1) {
                       shinyalert("ERROR!", "File contains multiple samples, please select the multiple samples option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     
                     else {
                       
                       seurat <- seurat_file@assays$RNA@counts
                       
                       if(class(seurat_file) == "list"){
                         seuratreactive$obj_original  <- CreateSeuratObject(counts = seurat$`Gene Expression`, project = input$textlabel, min.cells = input$mincells, min.features = input$minfeatures)}
                       else {
                         seuratreactive$obj_original  <- CreateSeuratObject(counts = seurat, project = input$textlabel, min.cells = input$mincells, min.features = input$minfeatures)}
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       seuratreactive$mincells <- input$mincells
                       seuratreactive$minfeatures <- input$minfeatures
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #Doublet
                       seuratreactive$plot.data.doublets <- NULL
                       seuratreactive$Remove <- NULL
                       seuratreactive$Removed <- NULL
                       seuratreactive$doublesim <- NULL
                       
                       #Load2
                       seuratreactive$obj_original2 <- NULL
                       seuratreactive$matrixtable2 <- NULL
                       seuratreactive$mincellsCombine <- NULL
                       seuratreactive$minfeaturesCombine <- NULL
                       
                       #QC
                       seuratreactive$objQC <- NULL
                       seuratreactive$nFeature1 <- NULL
                       seuratreactive$nFeature2 <- NULL
                       seuratreactive$nCount1 <- NULL
                       seuratreactive$nCount2 <- NULL
                       seuratreactive$percent.mt1 <- NULL
                       seuratreactive$percent.mt2 <- NULL
                       seuratreactive$percent.ribo1 <- NULL
                       seuratreactive$percent.ribo2 <- NULL
                       
                       #QC2
                       seuratreactive$type2 <- NULL
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       seuratreactive$npcDR <- NULL
                       
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       
                       #C
                       
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       if (!is.null(seuratreactive$obj_original)) {
                         
                         if(input$iscollapseboxQC == TRUE) {
                           js$collapse("QCBoxIntro")
                         }
                         shinyjs::hide("QCBox1")
                         shinyjs::hide("QCBox2")
                         shinyjs::hide("QCBox3")
                         shinyjs::hide("QCBox4")
                         shinyjs::enable("startbuttonQC")
                         
                         shinyjs::hide("NextButtonQC")
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       seuratreactive$matrixtable <- "yes"
                       shinyjs::show("NextButtonLoad")
                       
                       incProgress(0.8)
                       
                       seuratreactive$type = "single"
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  
  observeEvent(input$textdirectory, {
    
    tryCatch({
      withProgress(message = 'Loading 10X files...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file1 <- input$textdirectory[[1, 'datapath']]
                     file2 <- input$textdirectory[[2, 'datapath']]
                     file3 <- input$textdirectory[[3, 'datapath']]
                     
                     filesdir = dirname(file1)
                     
                     file.rename(file1, paste0(filesdir,'/',input$textdirectory$name[1]))
                     file.rename(file2, paste0(filesdir,'/',input$textdirectory$name[2]))
                     file.rename(file3, paste0(filesdir,'/',input$textdirectory$name[3]))
                     
                     seurat <- Read10X(data.dir = filesdir)
                     
                     if(class(seurat) == "list"){
                       seurat <- CreateSeuratObject(counts = seurat$`Gene Expression`, project = input$textlabel, min.cells = input$mincells, min.features = input$minfeatures)}
                     else {
                       seurat <- CreateSeuratObject(counts = seurat, project = input$textlabel, min.cells = input$mincells, min.features = input$minfeatures)}
                     
                     if (length(unique(seurat@meta.data$orig.ident)) > 1) {
                       shinyalert("ERROR!", "File contains multiple samples, please select the multiple samples option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     
                     else {
                       
                       seuratreactive$obj_original <- seurat
                       
                       seuratreactive$matrixtable <- "yes"
                       shinyjs::show("NextButtonLoad")
                       
                       seuratreactive$mincells <- input$mincells
                       seuratreactive$minfeatures <- input$minfeatures
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #Doublet
                       seuratreactive$plot.data.doublets <- NULL
                       seuratreactive$Remove <- NULL
                       seuratreactive$Removed <- NULL
                       seuratreactive$doublesim <- NULL
                       
                       #Load2
                       seuratreactive$obj_original2 <- NULL
                       seuratreactive$matrixtable2 <- NULL
                       seuratreactive$mincellsCombine <- NULL
                       seuratreactive$minfeaturesCombine <- NULL
                       
                       #QC
                       seuratreactive$objQC <- NULL
                       seuratreactive$nFeature1 <- NULL
                       seuratreactive$nFeature2 <- NULL
                       seuratreactive$nCount1 <- NULL
                       seuratreactive$nCount2 <- NULL
                       seuratreactive$percent.mt1 <- NULL
                       seuratreactive$percent.mt2 <- NULL
                       seuratreactive$percent.ribo1 <- NULL
                       seuratreactive$percent.ribo2 <- NULL
                       
                       #QC2
                       seuratreactive$type2 <- NULL
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       
                       #C
                       
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       
                       if (!is.null(seuratreactive$obj_original)) {
                         
                         if(input$iscollapseboxQC == TRUE) {
                           js$collapse("QCBoxIntro")
                         }
                         shinyjs::hide("QCBox1")
                         shinyjs::hide("QCBox2")
                         shinyjs::hide("QCBox3")
                         shinyjs::hide("QCBox4")
                         
                         shinyjs::hide("NextButtonQC")
                         shinyjs::enable("startbuttonQC")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       incProgress(0.8)
                       
                       seuratreactive$type = "single"
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if you have loaded 3 files (barcodes.tsv, features.tsv, matrix.mtx).", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$fileinput2, {
    tryCatch({
      withProgress(message = 'Loading count matrix...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file = input$fileinput2
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat <- read.table(file$datapath, sep = "\t", header = T, row.names = 1, check.names = F)
                     
                     if (any(grepl("_", colnames(seurat)))) {
                       shinyalert("ERROR!", "File contains multiple samples, please select the multiple samples option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       
                       seuratreactive$obj_original <- CreateSeuratObject(counts = seurat, project = input$textlabel, min.cells = input$mincells, min.features = input$minfeatures)
                       
                       seuratreactive$matrixtable <- "yes"
                       shinyjs::show("NextButtonLoad")
                       seuratreactive$mincells <- input$mincells
                       seuratreactive$minfeatures <- input$minfeatures
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #Doublet
                       seuratreactive$plot.data.doublets <- NULL
                       seuratreactive$Remove <- NULL
                       seuratreactive$Removed <- NULL
                       seuratreactive$doublesim <- NULL
                       
                       #Load2
                       seuratreactive$obj_original2 <- NULL
                       seuratreactive$matrixtable2 <- NULL
                       seuratreactive$mincellsCombine <- NULL
                       seuratreactive$minfeaturesCombine <- NULL
                       
                       #QC
                       seuratreactive$objQC <- NULL
                       seuratreactive$nFeature1 <- NULL
                       seuratreactive$nFeature2 <- NULL
                       seuratreactive$nCount1 <- NULL
                       seuratreactive$nCount2 <- NULL
                       seuratreactive$percent.mt1 <- NULL
                       seuratreactive$percent.mt2 <- NULL
                       seuratreactive$percent.ribo1 <- NULL
                       seuratreactive$percent.ribo2 <- NULL
                       
                       #QC2
                       seuratreactive$type2 <- NULL
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       
                       if (!is.null(seuratreactive$obj_original)) {
                         
                         if(input$iscollapseboxQC == TRUE) {
                           js$collapse("QCBoxIntro")
                         }
                         shinyjs::hide("QCBox1")
                         shinyjs::hide("QCBox2")
                         shinyjs::hide("QCBox3")
                         shinyjs::hide("QCBox4")
                         
                         shinyjs::hide("NextButtonQC")
                         
                         shinyjs::enable("startbuttonQC")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       incProgress(0.8)
                       
                       seuratreactive$type = "single"
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$buttonexistingdata, {
    if (input$existingdata == "3k PBMCs (Seurat - Guided Clustering Tutorial, 10X data)") {
      
      withProgress(message = 'Loading 3k PBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original <- readRDS("pbmc3k.rds")
                     seuratreactive$matrixtable <- "yes"
                     shinyjs::show("NextButtonLoad")
                     
                     seuratreactive$mincells <- "NA"
                     seuratreactive$minfeatures <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     
                     #Doublet
                     seuratreactive$plot.data.doublets <- NULL
                     seuratreactive$Remove <- NULL
                     seuratreactive$Removed <- NULL
                     seuratreactive$doublesim <- NULL
                     
                     #Load2
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     seuratreactive$mincellsCombine <- NULL
                     seuratreactive$minfeaturesCombine <- NULL
                     
                     #QC
                     seuratreactive$objQC <- NULL
                     seuratreactive$nFeature1 <- NULL
                     seuratreactive$nFeature2 <- NULL
                     seuratreactive$nCount1 <- NULL
                     seuratreactive$nCount2 <- NULL
                     seuratreactive$percent.mt1 <- NULL
                     seuratreactive$percent.mt2 <- NULL
                     seuratreactive$percent.ribo1 <- NULL
                     seuratreactive$percent.ribo2 <- NULL
                     
                     #QC2
                     seuratreactive$type2 <- NULL
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original)) {
                       
                       if(input$iscollapseboxQC == TRUE) {
                         js$collapse("QCBoxIntro")
                       }
                       shinyjs::hide("QCBox1")
                       shinyjs::hide("QCBox2")
                       shinyjs::hide("QCBox3")
                       shinyjs::hide("QCBox4")
                       
                       shinyjs::hide("NextButtonQC")
                       shinyjs::enable("startbuttonQC")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     incProgress(0.8)
                     
                     seuratreactive$type = "single"
                   })
    }
    
    else if (input$existingdata == "500 PBMCs (10X data)") {
      
      withProgress(message = 'Loading 500 PBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original <- readRDS("500_Human_PBMC.rds")
                     seuratreactive$matrixtable <- "yes"
                     shinyjs::show("NextButtonLoad")
                     
                     seuratreactive$mincells <- "NA"
                     seuratreactive$minfeatures <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #Doublet
                     seuratreactive$plot.data.doublets <- NULL
                     seuratreactive$Remove <- NULL
                     seuratreactive$Removed <- NULL
                     seuratreactive$doublesim <- NULL
                     
                     #Load2
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     seuratreactive$mincellsCombine <- NULL
                     seuratreactive$minfeaturesCombine <- NULL
                     
                     #QC
                     seuratreactive$objQC <- NULL
                     seuratreactive$nFeature1 <- NULL
                     seuratreactive$nFeature2 <- NULL
                     seuratreactive$nCount1 <- NULL
                     seuratreactive$nCount2 <- NULL
                     seuratreactive$percent.mt1 <- NULL
                     seuratreactive$percent.mt2 <- NULL
                     seuratreactive$percent.ribo1 <- NULL
                     seuratreactive$percent.ribo2 <- NULL
                     
                     #QC2
                     seuratreactive$type2 <- NULL
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original)) {
                       
                       if(input$iscollapseboxQC == TRUE) {
                         js$collapse("QCBoxIntro")
                       }
                       shinyjs::hide("QCBox1")
                       shinyjs::hide("QCBox2")
                       shinyjs::hide("QCBox3")
                       shinyjs::hide("QCBox4")
                       
                       shinyjs::hide("NextButtonQC")
                       shinyjs::enable("startbuttonQC")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     incProgress(0.8)
                     
                     seuratreactive$type = "single"
                   })
    }
    
    else if (input$existingdata == "5k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone (10X data)") {
      
      withProgress(message = 'Loading 5k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original <- readRDS("5k_Mouse_E18.rds")
                     seuratreactive$matrixtable <- "yes"
                     shinyjs::show("NextButtonLoad")
                     
                     seuratreactive$mincells <- "NA"
                     seuratreactive$minfeatures <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #Doublet
                     seuratreactive$plot.data.doublets <- NULL
                     seuratreactive$Remove <- NULL
                     seuratreactive$Removed <- NULL
                     seuratreactive$doublesim <- NULL
                     
                     #Load2
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     seuratreactive$mincellsCombine <- NULL
                     seuratreactive$minfeaturesCombine <- NULL
                     
                     #QC
                     seuratreactive$objQC <- NULL
                     seuratreactive$nFeature1 <- NULL
                     seuratreactive$nFeature2 <- NULL
                     seuratreactive$nCount1 <- NULL
                     seuratreactive$nCount2 <- NULL
                     seuratreactive$percent.mt1 <- NULL
                     seuratreactive$percent.mt2 <- NULL
                     seuratreactive$percent.ribo1 <- NULL
                     seuratreactive$percent.ribo2 <- NULL
                     
                     #QC2
                     seuratreactive$type2 <- NULL
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original)) {
                       
                       if(input$iscollapseboxQC == TRUE) {
                         js$collapse("QCBoxIntro")
                       }
                       shinyjs::hide("QCBox1")
                       shinyjs::hide("QCBox2")
                       shinyjs::hide("QCBox3")
                       shinyjs::hide("QCBox4")
                       
                       shinyjs::hide("NextButtonQC")
                       shinyjs::enable("startbuttonQC")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     incProgress(0.8)
                     
                     seuratreactive$type = "single"
                   })
    }
    
    else if (input$existingdata == "8k CBMCs (Seurat CITE-seq multimodal tutorial)") {
      
      withProgress(message = 'Loading 8k CBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original <- readRDS("8k_CBMCs.rds")
                     seuratreactive$matrixtable <- "yes"
                     shinyjs::show("NextButtonLoad")
                     
                     seuratreactive$mincells <- "NA"
                     seuratreactive$minfeatures <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #Doublet
                     seuratreactive$plot.data.doublets <- NULL
                     seuratreactive$Remove <- NULL
                     seuratreactive$Removed <- NULL
                     seuratreactive$doublesim <- NULL
                     
                     #Load2
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     seuratreactive$mincellsCombine <- NULL
                     seuratreactive$minfeaturesCombine <- NULL
                     
                     #QC
                     seuratreactive$objQC <- NULL
                     seuratreactive$nFeature1 <- NULL
                     seuratreactive$nFeature2 <- NULL
                     seuratreactive$nCount1 <- NULL
                     seuratreactive$nCount2 <- NULL
                     seuratreactive$percent.mt1 <- NULL
                     seuratreactive$percent.mt2 <- NULL
                     seuratreactive$percent.ribo1 <- NULL
                     seuratreactive$percent.ribo2 <- NULL
                     
                     #QC2
                     seuratreactive$type2 <- NULL
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original)) {
                       
                       if(input$iscollapseboxQC == TRUE) {
                         js$collapse("QCBoxIntro")
                       }
                       shinyjs::hide("QCBox1")
                       shinyjs::hide("QCBox2")
                       shinyjs::hide("QCBox3")
                       shinyjs::hide("QCBox4")
                       
                       shinyjs::hide("NextButtonQC")
                       shinyjs::enable("startbuttonQC")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     incProgress(0.8)
                     
                     seuratreactive$type = "single"
                   })
    }
  })
  
  observeEvent(input$fileinput1M, {
    
    tryCatch({
      withProgress(message = 'Loading seurat object...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     
                     file = input$fileinput1M
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat <- readRDS(file$datapath)
                     
                     if (length(unique(seurat@meta.data$orig.ident)) == 1) {
                       shinyalert("ERROR!", "File contains a single sample, please select the single sample option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     
                     else {
                       
                       seuratreactive$obj_original <- seurat
                       
                       seuratreactive$matrixtable <- "yes"
                       shinyjs::show("NextButtonLoad")
                       
                       incProgress(0.8)
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #Doublet
                       seuratreactive$plot.data.doublets <- NULL
                       seuratreactive$Remove <- NULL
                       seuratreactive$Removed <- NULL
                       seuratreactive$doublesim <- NULL
                       
                       #Load2
                       seuratreactive$obj_original2 <- NULL
                       seuratreactive$matrixtable2 <- NULL
                       seuratreactive$mincellsCombine <- NULL
                       seuratreactive$minfeaturesCombine <- NULL
                       
                       #QC
                       seuratreactive$objQC <- NULL
                       seuratreactive$nFeature1 <- NULL
                       seuratreactive$nFeature2 <- NULL
                       seuratreactive$nCount1 <- NULL
                       seuratreactive$nCount2 <- NULL
                       seuratreactive$percent.mt1 <- NULL
                       seuratreactive$percent.mt2 <- NULL
                       seuratreactive$percent.ribo1 <- NULL
                       seuratreactive$percent.ribo2 <- NULL
                       
                       #QC2
                       seuratreactive$type2 <- NULL
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       if (!is.null(seuratreactive$obj_original)) {
                         
                         if(input$iscollapseboxQC == TRUE) {
                           js$collapse("QCBoxIntro")
                         }
                         shinyjs::hide("QCBox1")
                         shinyjs::hide("QCBox2")
                         shinyjs::hide("QCBox3")
                         shinyjs::hide("QCBox4")
                         
                         shinyjs::hide("NextButtonQC")
                         shinyjs::enable("startbuttonQC")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       
                       seuratreactive$type = "multiple"
                       
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })
    
  })
  
  observeEvent(input$fileinput2M, {
    
    tryCatch({
      withProgress(message = 'Loading count matrix...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file = input$fileinput2M
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat <- read.table(file$datapath, sep = "\t", header = T, row.names = 1, check.names = F)
                     
                     if (any(!grepl("_", colnames(seurat)))) {
                       shinyalert("ERROR!", "File contains a single sample, please select the sample sample option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       
                       seuratreactive$obj_original <- CreateSeuratObject(counts = seurat)
                       seuratreactive$matrixtable <- "yes"
                       shinyjs::show("NextButtonLoad")
                       incProgress(0.8)
                       seuratreactive$type = "multiple"
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #Doublet
                       seuratreactive$plot.data.doublets <- NULL
                       seuratreactive$Remove <- NULL
                       seuratreactive$Removed <- NULL
                       seuratreactive$doublesim <- NULL
                       
                       #Load2
                       seuratreactive$obj_original2 <- NULL
                       seuratreactive$matrixtable2 <- NULL
                       seuratreactive$mincellsCombine <- NULL
                       seuratreactive$minfeaturesCombine <- NULL
                       
                       #QC
                       seuratreactive$objQC <- NULL
                       seuratreactive$nFeature1 <- NULL
                       seuratreactive$nFeature2 <- NULL
                       seuratreactive$nCount1 <- NULL
                       seuratreactive$nCount2 <- NULL
                       seuratreactive$percent.mt1 <- NULL
                       seuratreactive$percent.mt2 <- NULL
                       seuratreactive$percent.ribo1 <- NULL
                       seuratreactive$percent.ribo2 <- NULL
                       
                       #QC2
                       seuratreactive$type2 <- NULL
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       if (!is.null(seuratreactive$obj_original)) {
                         
                         if(input$iscollapseboxQC == TRUE) {
                           js$collapse("QCBoxIntro")
                         }
                         shinyjs::hide("QCBox1")
                         shinyjs::hide("QCBox2")
                         shinyjs::hide("QCBox3")
                         shinyjs::hide("QCBox4")
                         
                         shinyjs::hide("NextButtonQC")
                         shinyjs::enable("startbuttonQC")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })  
  })
  
  observeEvent(input$buttonexistingdataM, {
    if (input$existingdataM == "10k Human PBMCs Multiplexed (CellPlex, 10X data)") {
      
      withProgress(message = 'Loading 10k Human PBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original <- readRDS("10k_Human_PBMCs_CellPlex.rds")
                     seuratreactive$matrixtable <- "yes"
                     shinyjs::show("NextButtonLoad")
                     incProgress(0.8)
                     seuratreactive$type = "multiple"
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #Doublet
                     seuratreactive$plot.data.doublets <- NULL
                     seuratreactive$Remove <- NULL
                     seuratreactive$Removed <- NULL
                     seuratreactive$doublesim <- NULL
                     
                     #Load2
                     seuratreactive$obj_original2 <- NULL
                     seuratreactive$matrixtable2 <- NULL
                     seuratreactive$mincellsCombine <- NULL
                     seuratreactive$minfeaturesCombine <- NULL
                     
                     #QC
                     seuratreactive$objQC <- NULL
                     seuratreactive$nFeature1 <- NULL
                     seuratreactive$nFeature2 <- NULL
                     seuratreactive$nCount1 <- NULL
                     seuratreactive$nCount2 <- NULL
                     seuratreactive$percent.mt1 <- NULL
                     seuratreactive$percent.mt2 <- NULL
                     seuratreactive$percent.ribo1 <- NULL
                     seuratreactive$percent.ribo2 <- NULL
                     
                     #QC2
                     seuratreactive$type2 <- NULL
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original)) {
                       
                       if(input$iscollapseboxQC == TRUE) {
                         js$collapse("QCBoxIntro")
                       }
                       shinyjs::hide("QCBox1")
                       shinyjs::hide("QCBox2")
                       shinyjs::hide("QCBox3")
                       shinyjs::hide("QCBox4")
                       
                       shinyjs::hide("NextButtonQC")
                       shinyjs::enable("startbuttonQC")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = T),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                   })
    }
  })
  
  output$MatrixTable <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$obj_original), 'You must load your data first...'))
    
    head(as.data.frame(seuratreactive$obj_original@assays$RNA@counts), c(500L, 500L))
    
  }, options = list(scrollX = TRUE))
  
  observeEvent(input$NextButtonLoad, {
    if (!is.null(seuratreactive$obj_original)) {
      
      if(input$iscollapseboxQC == TRUE) {
        js$collapse("QCBoxIntro")
      }
      shinyjs::hide("QCBox1")
      shinyjs::hide("QCBox2")
      shinyjs::hide("QCBox3")
      shinyjs::hide("QCBox4")
      
      shinyjs::hide("NextButtonLoad")
      shinyjs::hide("NextButtonQC")
      
      shinyjs::enable("startbuttonQC")
      
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
    else{
      shinyalert("ERROR!", "Please reupload your file.", type = "error", confirmButtonCol = "#337ab7")
    }
  },ignoreInit = T)
  
  ####TAB1.5 - Integrate 2nd dataset####
  
  observeEvent(input$startbuttonCombine, {
    if(input$iscollapseboxCombine == FALSE) {
      js$collapse("CombineBoxIntro")
    }
    shinyjs::show("CombineBox1")
    shinyjs::hide("CombineBox2")
    shinyjs::hide("CombineBox3")
    shinyjs::hide("CombineBox5")
    shinyjs::show("CombineBox4")
    
    shinyjs::enable("startbuttonQC2")
    shinyjs::hide("NextButtonQC2")
    
    shinyjs::hide("NextButtonD")
    
    
  })
  
  observe({
    if (input$singlemultipleCombine == "Single") {
      shinyjs::show("CombineBox2")
      shinyjs::hide("CombineBox5")
      shinyjs::reset("radioCombine")
    }
    else if (input$singlemultipleCombine == "Multiple") {
      shinyjs::hide("CombineBox2")
      shinyjs::hide("CombineBox3")
      shinyjs::show("CombineBox5")
      shinyjs::reset("radioMCombine")
    }
  })
  
  observe({
    if (input$radioCombine == "10X data") {
      shinyjs::show("textdirectoryCombine")
      shinyjs::hide("fileinput1Combine")
      shinyjs::hide("fileinput2Combine")
      shinyjs::hide("existingdataCombine")
      shinyjs::hide("buttonexistingdataCombine")
      shinyjs::show("CombineBox3")
    }
    else if (input$radioCombine == "Seurat object (RDS file)") {
      shinyjs::hide("textdirectoryCombine")
      shinyjs::show("fileinput1Combine")
      shinyjs::hide("fileinput2Combine")
      shinyjs::hide("existingdataCombine")
      shinyjs::hide("buttonexistingdataCombine")
      shinyjs::show("CombineBox3")
    }
    
    else if (input$radioCombine == "Count Matrix (txt)") {
      shinyjs::hide("textdirectoryCombine")
      shinyjs::hide("fileinput1Combine")
      shinyjs::show("fileinput2Combine")
      shinyjs::hide("existingdataCombine")
      shinyjs::hide("buttonexistingdataCombine")
      shinyjs::show("CombineBox3")
    }
    
    else if (input$radioCombine == "Sample datasets") {
      shinyjs::hide("textdirectoryCombine")
      shinyjs::hide("fileinput1Combine")
      shinyjs::hide("fileinput2Combine")
      shinyjs::show("existingdataCombine")
      shinyjs::show("buttonexistingdataCombine")
      shinyjs::hide("CombineBox3")
    }
  })
  
  observe({
    if (input$radioMCombine == "Seurat object (RDS file)") {
      shinyjs::show("fileinput1MCombine")
      shinyjs::hide("fileinput2MCombine")
      shinyjs::hide("existingdataMCombine")
      shinyjs::hide("buttonexistingdataMCombine")
    }
    
    else if (input$radioMCombine == "Count Matrix (txt)") {
      shinyjs::hide("fileinput1MCombine")
      shinyjs::show("fileinput2MCombine")
      shinyjs::hide("existingdataMCombine")
      shinyjs::hide("buttonexistingdataMCombine")
    }
    
    else if (input$radioMCombine == "Sample datasets") {
      shinyjs::hide("fileinput1MCombine")
      shinyjs::hide("fileinput2MCombine")
      shinyjs::show("existingdataMCombine")
      shinyjs::show("buttonexistingdataMCombine")
    }
  })
  
  observe({
    if(is.null(input$existingdataCombine) | !is.null(input$existingdataCombine) && input$existingdataCombine == "") {
      shinyjs::disable("buttonexistingdataCombine")
    }
    else{
      shinyjs::enable("buttonexistingdataCombine")
    }
  })
  
  observe({
    if(is.null(input$existingdataMCombine) | !is.null(input$existingdataMCombine) && input$existingdataMCombine == "") {
      shinyjs::disable("buttonexistingdataMCombine")
    }
    else{
      shinyjs::enable("buttonexistingdataMCombine")
    }
  })
  
  observeEvent(input$tabs, {
    if (input$tabs == "Combine") {
      updatePickerInput(session, 'singlemultipleCombine', selected = "")  
    }
  })
  
  observeEvent(input$startbuttonCombine, {
    updatePickerInput(session, 'singlemultipleCombine', selected = "")  
  })
  
  observeEvent(input$fileinput1Combine, {
    tryCatch({
      withProgress(message = 'Loading Seurat object...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     
                     file = input$fileinput1Combine
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat_file <- readRDS(file$datapath)
                     
                     if (length(unique(seurat_file@meta.data$orig.ident)) > 1) {
                       shinyalert("ERROR!", "File contains multiple samples, please select the multiple samples option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       
                       seurat <- seurat_file@assays$RNA@counts
                       
                       if(class(seurat_file) == "list"){
                         seuratreactive$obj_original2  <- CreateSeuratObject(counts = seurat$`Gene Expression`, project = input$textlabelCombine, min.cells = input$mincellsCombine, min.features = input$minfeaturesCombine)}
                       else {
                         seuratreactive$obj_original2  <- CreateSeuratObject(counts = seurat, project = input$textlabelCombine, min.cells = input$mincellsCombine, min.features = input$minfeaturesCombine)}
                       
                       incProgress(0.8)
                       
                       seuratreactive$matrixtable2 <- "yes"
                       shinyjs::show("NextButtonCombine")
                       seuratreactive$type2 <- "single"
                       seuratreactive$mincellsCombine <- input$mincellsCombine
                       seuratreactive$minfeaturesCombine <- input$minfeaturesCombine
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original2)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #QC2
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       
                       if (!is.null(seuratreactive$obj_original2)) {
                         
                         if(input$iscollapseboxQC2 == TRUE) {
                           js$collapse("QCBoxIntro2")
                         }
                         shinyjs::hide("QCBox12")
                         shinyjs::hide("QCBox22")
                         shinyjs::hide("QCBox32")
                         shinyjs::hide("QCBox42")
                         
                         shinyjs::enable("startbuttonQC2")
                         shinyjs::hide("NextButtonQC2")
                         
                         shinyjs::hide("NextButtonD")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       
                       
                       if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                         shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                       }
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })  
  })
  
  
  observeEvent(input$textdirectoryCombine, {
    tryCatch({
      withProgress(message = 'Loading 10X files...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file1 <- input$textdirectoryCombine[[1, 'datapath']]
                     file2 <- input$textdirectoryCombine[[2, 'datapath']]
                     file3 <- input$textdirectoryCombine[[3, 'datapath']]
                     
                     filesdir = dirname(file1)
                     
                     file.rename(file1, paste0(filesdir,'/',input$textdirectoryCombine$name[1]))
                     file.rename(file2, paste0(filesdir,'/',input$textdirectoryCombine$name[2]))
                     file.rename(file3, paste0(filesdir,'/',input$textdirectoryCombine$name[3]))
                     
                     seurat <- Read10X(data.dir = filesdir)
                     
                     if(class(seurat) == "list"){
                       seurat <- CreateSeuratObject(counts = seurat$`Gene Expression`, project = input$textlabelCombine, min.cells = input$mincellsCombine, min.features = input$minfeaturesCombine)}
                     else {
                       seurat <- CreateSeuratObject(counts = seurat, project = input$textlabelCombine, min.cells = input$mincellsCombine, min.features = input$minfeaturesCombine)}
                     
                     incProgress(0.8)
                     
                     if (length(unique(seurat@meta.data$orig.ident)) > 1) {
                       shinyalert("ERROR!", "File contains multiple samples, please select the multiple samples option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     
                     else {
                       
                       seuratreactive$obj_original2 <- seurat
                       
                       seuratreactive$matrixtable2 <- "yes"
                       seuratreactive$type2 <- "single"
                       shinyjs::show("NextButtonCombine")
                       
                       seuratreactive$mincellsCombine <- input$mincellsCombine
                       seuratreactive$minfeaturesCombine <- input$minfeaturesCombine
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original2)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #QC2
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       
                       if (!is.null(seuratreactive$obj_original2)) {
                         
                         if(input$iscollapseboxQC2 == TRUE) {
                           js$collapse("QCBoxIntro2")
                         }
                         shinyjs::hide("QCBox12")
                         shinyjs::hide("QCBox22")
                         shinyjs::hide("QCBox32")
                         shinyjs::hide("QCBox42")
                         
                         shinyjs::enable("startbuttonQC2")
                         shinyjs::hide("NextButtonQC2")
                         
                         shinyjs::hide("NextButtonD")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       
                       
                       if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                         shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                       }
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if you have loaded 3 files (barcodes.tsv, features.tsv, matrix.mtx).", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$fileinput2Combine, {
    tryCatch({
      withProgress(message = 'Loading count matrix...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file = input$fileinput2Combine
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat <- read.table(file$datapath, sep = "\t", header = T, row.names = 1, check.names = F)
                     
                     if (any(grepl("_", colnames(seurat)))){
                       shinyalert("ERROR!", "File contains multiple samples, please use select the multiple samples option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       
                       seuratreactive$obj_original2 <- CreateSeuratObject(counts = seurat, project = input$textlabelCombine, min.cells = input$mincellsCombine, min.features = input$minfeaturesCombine)
                       
                       incProgress(0.8)
                       seuratreactive$matrixtable2 <- "yes"
                       seuratreactive$type2 <- "single"
                       shinyjs::show("NextButtonCombine")
                       
                       seuratreactive$mincellsCombine <- input$mincellsCombine
                       seuratreactive$minfeaturesCombine <- input$minfeaturesCombine
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original2)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #QC2
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       
                       if (!is.null(seuratreactive$obj_original2)) {
                         
                         if(input$iscollapseboxQC2 == TRUE) {
                           js$collapse("QCBoxIntro2")
                         }
                         shinyjs::hide("QCBox12")
                         shinyjs::hide("QCBox22")
                         shinyjs::hide("QCBox32")
                         shinyjs::hide("QCBox42")
                         
                         shinyjs::enable("startbuttonQC2")
                         shinyjs::hide("NextButtonQC2")
                         
                         shinyjs::hide("NextButtonD")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       
                       if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                         shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                       }
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })  
  })
  
  observeEvent(input$buttonexistingdataCombine, {
    if (input$existingdataCombine == "3k PBMCs (Seurat - Guided Clustering Tutorial, 10X data)") {
      
      withProgress(message = 'Loading 3k PBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original2 <- readRDS("pbmc3k.rds")
                     incProgress(0.8)
                     
                     seuratreactive$type2 <- "single"
                     seuratreactive$matrixtable2 <- "yes"
                     shinyjs::show("NextButtonCombine")
                     seuratreactive$mincellsCombine <- "NA"
                     seuratreactive$minfeaturesCombine <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #QC2
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original2)) {
                       
                       if(input$iscollapseboxQC2 == TRUE) {
                         js$collapse("QCBoxIntro2")
                       }
                       shinyjs::hide("QCBox12")
                       shinyjs::hide("QCBox22")
                       shinyjs::hide("QCBox32")
                       shinyjs::hide("QCBox42")
                       
                       shinyjs::enable("startbuttonQC2")
                       shinyjs::hide("NextButtonQC2")
                       
                       shinyjs::hide("NextButtonD")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                     if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                       shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                     }
                   })
    }
    
    else if (input$existingdataCombine == "500 PBMCs (10X data)") {
      
      withProgress(message = 'Loading 500 PBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original2 <- readRDS("500_Human_PBMC.rds")
                     incProgress(0.8)
                     
                     seuratreactive$type2 <- "single"
                     seuratreactive$matrixtable2 <- "yes"
                     shinyjs::show("NextButtonCombine")
                     
                     seuratreactive$mincellsCombine <- "NA"
                     seuratreactive$minfeaturesCombine <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #QC2
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original2)) {
                       
                       if(input$iscollapseboxQC2 == TRUE) {
                         js$collapse("QCBoxIntro2")
                       }
                       shinyjs::hide("QCBox12")
                       shinyjs::hide("QCBox22")
                       shinyjs::hide("QCBox32")
                       shinyjs::hide("QCBox42")
                       
                       shinyjs::enable("startbuttonQC2")
                       shinyjs::hide("NextButtonQC2")
                       
                       shinyjs::hide("NextButtonD")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                     if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                       shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                     }
                   })
    }
    
    else if (input$existingdataCombine == "5k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone (10X data)") {
      
      withProgress(message = 'Loading 5k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original2 <- readRDS("5k_Mouse_E18.rds")
                     incProgress(0.8)
                     
                     seuratreactive$type2 <- "single"
                     seuratreactive$matrixtable2 <- "yes"
                     shinyjs::show("NextButtonCombine")
                     
                     seuratreactive$mincellsCombine <- "NA"
                     seuratreactive$minfeaturesCombine <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #QC2
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original2)) {
                       
                       if(input$iscollapseboxQC2 == TRUE) {
                         js$collapse("QCBoxIntro2")
                       }
                       shinyjs::hide("QCBox12")
                       shinyjs::hide("QCBox22")
                       shinyjs::hide("QCBox32")
                       shinyjs::hide("QCBox42")
                       
                       shinyjs::enable("startbuttonQC2")
                       shinyjs::hide("NextButtonQC2")
                       
                       shinyjs::hide("NextButtonD")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                     if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                       shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                     }
                   })
    }
    
    else if (input$existingdataCombine == "8k CBMCs (Seurat CITE-seq multimodal tutorial)") {
      
      withProgress(message = 'Loading 8k CBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original2 <- readRDS("8k_CBMCs.rds")
                     incProgress(0.8)
                     
                     seuratreactive$type2 <- "single"
                     seuratreactive$matrixtable2 <- "yes"
                     shinyjs::show("NextButtonCombine")
                     
                     seuratreactive$mincellsCombine <- "NA"
                     seuratreactive$minfeaturesCombine <- "NA"
                     
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #QC2
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original2)) {
                       
                       if(input$iscollapseboxQC2 == TRUE) {
                         js$collapse("QCBoxIntro2")
                       }
                       shinyjs::hide("QCBox12")
                       shinyjs::hide("QCBox22")
                       shinyjs::hide("QCBox32")
                       shinyjs::hide("QCBox42")
                       
                       shinyjs::enable("startbuttonQC2")
                       shinyjs::hide("NextButtonQC2")
                       
                       shinyjs::hide("NextButtonD")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                     if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                       shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                     }
                   })
    }
  })
  
  observeEvent(input$fileinput1MCombine, {
    tryCatch({
      withProgress(message = 'Loading Seurat object...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     
                     file = input$fileinput1MCombine
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat <- readRDS(file$datapath)
                     
                     if (length(unique(seurat@meta.data$orig.ident)) == 1) {
                       shinyalert("ERROR!", "File contains a single sample, please select the single sample option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     
                     else {
                       seuratreactive$obj_original2 <- seurat
                       
                       incProgress(0.8)
                       
                       seuratreactive$type2 <- "multiple"
                       seuratreactive$matrixtable2 <- "yes"
                       shinyjs::show("NextButtonCombine")
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original2)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #QC2
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       if (!is.null(seuratreactive$obj_original2)) {
                         
                         if(input$iscollapseboxQC2 == TRUE) {
                           js$collapse("QCBoxIntro2")
                         }
                         shinyjs::hide("QCBox12")
                         shinyjs::hide("QCBox22")
                         shinyjs::hide("QCBox32")
                         shinyjs::hide("QCBox42")
                         
                         shinyjs::enable("startbuttonQC2")
                         shinyjs::hide("NextButtonQC2")
                         
                         shinyjs::hide("NextButtonD")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       
                       if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                         shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                       }
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })  
  })
  
  observeEvent(input$fileinput2MCombine, {
    tryCatch({
      withProgress(message = 'Loading count matrix...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file = input$fileinput2MCombine
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seurat <- read.table(file$datapath, sep = "\t", header = T, row.names = 1, check.names = F)
                     
                     if (any(!grepl("_", colnames(seurat)))){
                       shinyalert("ERROR!", "File contains a single sample, please select the single sample option.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       
                       seuratreactive$obj_original2 <- CreateSeuratObject(counts = seurat)
                       
                       incProgress(0.8)
                       seuratreactive$type2 <- "multiple"
                       seuratreactive$matrixtable2 <- "yes"
                       shinyjs::show("NextButtonCombine")
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       if(any(grepl("ENS[A-Z]+00", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Ensembl Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("GRCh[[:digit:]]+", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains GRCh prefix, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(any(grepl("^[[:digit:]]+$", rownames(seuratreactive$obj_original2)))){
                         shinyalert("WARNING!", "File contains Entrez Identifiers, Calculation of mitochondrial and ribosomal expressed genes may not be accurate. Additionally, cell type annotation and pathway analysis may not be accurate.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       if(length(colnames(seuratreactive$obj_original2)) > 15000){
                         shinyalert("WARNING!", "File contains more than 15,000 cells. ICARUS is not recommended for large datasets. Some processing steps may take extra long.", type = "warning", confirmButtonCol = "#337ab7")
                       }
                       
                       updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                       
                       #QC2
                       seuratreactive$objQC2 <- NULL
                       seuratreactive$nFeature12 <- NULL
                       seuratreactive$nFeature22 <- NULL
                       seuratreactive$nCount12 <- NULL
                       seuratreactive$nCount22 <- NULL
                       seuratreactive$percent.mt12 <- NULL
                       seuratreactive$percent.mt22 <- NULL
                       seuratreactive$percent.ribo12 <- NULL
                       seuratreactive$percent.ribo22 <- NULL
                       
                       #D2
                       seuratreactive$Remove2 <- NULL
                       seuratreactive$plot.data.doublets2 <- NULL
                       seuratreactive$Removed2 <- NULL
                       seuratreactive$doublesim2 <- NULL
                       
                       #DR
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR <- NULL
                       seuratreactive$npcDR <- NULL
                       seuratreactive$variablefeatures <- NULL
                       seuratreactive$ALRAL <- NULL
                       
                       #DR2
                       seuratreactive$obj <- NULL
                       seuratreactive$SCTransformDR2 <- NULL
                       seuratreactive$npcDR2 <- NULL
                       seuratreactive$Combined.list <- NULL
                       seuratreactive$VFintegrated <- NULL
                       seuratreactive$integrationreduction <- NULL
                       seuratreactive$kanchor <- NULL
                       
                       #C
                       seuratreactive$integera <- NULL
                       seuratreactive$integerb <- NULL
                       seuratreactive$integerc <- NULL
                       seuratreactive$integerd <- NULL
                       seuratreactive$integere <- NULL
                       seuratreactive$integerf <- NULL
                       seuratreactive$integerg <- NULL
                       seuratreactive$select1 <- NULL
                       seuratreactive$tsne1 <- NULL
                       seuratreactive$tsne2 <- NULL
                       seuratreactive$tsne3 <- NULL
                       seuratreactive$plot.data <- NULL
                       seuratreactive$plot.data.original <- NULL
                       
                       seuratreactive$Identifiers <- NULL
                       
                       #CC
                       seuratreactive$reductionCC1 <- NULL
                       seuratreactive$reductionCC2 <- NULL
                       seuratreactive$reductionCC3 <- NULL
                       seuratreactive$GeneR <- NULL
                       
                       
                       #L
                       seuratreactive$reference_annotation <- NULL
                       seuratreactive$select2 <- NULL
                       seuratreactive$selecttissue <- NULL
                       seuratreactive$new.cell.ids <- NULL
                       seuratreactive$markers <- NULL
                       
                       #MEGENA
                       seuratreactive$MEGENA_samples <- NULL
                       seuratreactive$variablegenesMEGENA <- NULL
                       seuratreactive$MEGENA_selectmethod <- NULL
                       seuratreactive$MEGENApval <- NULL
                       seuratreactive$MEGENAmin.size <- NULL
                       seuratreactive$MEGENA_selectcluster <- NULL
                       seuratreactive$n_MEGENA <- NULL
                       seuratreactive$log2FCMEGENA <- NULL
                       seuratreactive$padjustMEGENA <- NULL
                       seuratreactive$testuseMEGENA <- NULL
                       seuratreactive$minMEGENA <- NULL
                       seuratreactive$pvalMEGENA <- NULL
                       seuratreactive$summary.output <- NULL
                       seuratreactive$markersMEGENA <- NULL
                       seuratreactive$moduleGO_df <- NULL
                       seuratreactive$VF <- NULL
                       seuratreactive$g <- NULL
                       
                       #SCENIC
                       seuratreactive$variablefeaturesSCENIC <- NULL
                       seuratreactive$variablegenesSCENIC <- NULL
                       seuratreactive$regulonAUC <- NULL
                       seuratreactive$regulons <- NULL
                       seuratreactive$cellInfo <- NULL
                       seuratreactive$n_SCENIC <- NULL
                       seuratreactive$SCENICdb <- NULL
                       seuratreactive$SCENIC_samples <- NULL
                       seuratreactive$regulonnames <- NULL
                       seuratreactive$titleSCENIC <- NULL
                       seuratreactive$SCENICInput <- NULL
                       seuratreactive$SCENIC_selectcluster <- NULL
                       seuratreactive$SCENIC_selectmethod <- NULL
                       seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                       seuratreactive$rss <- NULL
                       
                       #MAGMA
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_Final_Results <- NULL
                       seuratreactive$meta <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$MAGMA_dirs <- NULL
                       seuratreactive$genesOutPath2 <- NULL
                       seuratreactive$n_MAGMA <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                       
                       #CellChat
                       seuratreactive$cellchat <- NULL
                       seuratreactive$cellchatM <- NULL
                       seuratreactive$object.list <- NULL
                       seuratreactive$ChatChoose <- NULL
                       seuratreactive$ChatSample <- NULL
                       seuratreactive$Chatdb <- NULL
                       seuratreactive$ChatMultiple1Sample <- NULL
                       seuratreactive$ChatMultiple2Sample <- NULL
                       seuratreactive$Chat_selectcluster <- NULL
                       seuratreactive$Chat_number <- NULL
                       seuratreactive$Chatpval <- NULL
                       
                       #DRUG
                       seuratreactive$markersDRUG <- NULL
                       seuratreactive$DRUGdf <- NULL
                       
                       #DRUG2
                       seuratreactive$markersDRUG2 <- NULL
                       seuratreactive$DRUGdf2 <- NULL
                       
                       #GE
                       seuratreactive$titleGE <- NULL
                       seuratreactive$GEInput <- NULL
                       seuratreactive$GSEInput <- NULL
                       seuratreactive$p <- NULL
                       seuratreactive$q <- NULL
                       seuratreactive$objGEyes <- NULL
                       seuratreactive$Gene_vector_msig <- NULL
                       
                       #TA
                       seuratreactive$TAselectedcells <- NULL
                       seuratreactive$TA_pickercluster <- NULL
                       seuratreactive$seurat_monocle <- NULL
                       seuratreactive$Monocle3_genes <- NULL
                       
                       #MM
                       seuratreactive$MM <- NULL
                       seuratreactive$plot.dataM <- NULL
                       seuratreactive$plot.dataMM <- NULL
                       
                       #DE
                       seuratreactive$onlyposDE <- NULL
                       seuratreactive$Subset_GENES <- NULL
                       seuratreactive$DE_clusters_samples <- NULL
                       seuratreactive$DECluster <- NULL
                       seuratreactive$DEInput1_samples <- NULL
                       seuratreactive$DEInput2_samples <- NULL
                       seuratreactive$SelectComparison <- NULL
                       seuratreactive$DE_clusters <- NULL
                       seuratreactive$GOlabelsDE <- NULL
                       seuratreactive$DEInput1 <- NULL
                       seuratreactive$DEInput2 <- NULL
                       seuratreactive$logDE <- NULL
                       seuratreactive$minDE <- NULL
                       seuratreactive$padjustDE1 <- NULL
                       seuratreactive$testuseDE <- NULL
                       seuratreactive$enrichInputDE <- NULL
                       seuratreactive$pvalueDE <- NULL
                       seuratreactive$pvalDE <- NULL
                       seuratreactive$padjustDE <- NULL
                       seuratreactive$volcanovalue1 <- NULL
                       seuratreactive$volcanovalue2 <- NULL
                       seuratreactive$pvaluevolcano <- NULL
                       seuratreactive$markersDE <- NULL
                       seuratreactive$YuDE <- NULL
                       seuratreactive$markersDEvolcano <- NULL
                       seuratreactive$PATHWAY <- NULL
                       
                       #SDE
                       seuratreactive$Subset_GENESSDE <- NULL
                       seuratreactive$selectedkey1 <- NULL
                       seuratreactive$selectedkey2 <- NULL
                       seuratreactive$onlyposSDE <- NULL
                       seuratreactive$GOlabelsSDE <- NULL
                       seuratreactive$logSDE <- NULL
                       seuratreactive$minSDE <- NULL
                       seuratreactive$padjustSDE1 <- NULL
                       seuratreactive$testuseSDE <- NULL
                       seuratreactive$enrichInputSDE <- NULL
                       seuratreactive$pvalueSDE <- NULL
                       seuratreactive$pvalSDE <- NULL
                       seuratreactive$padjustSDE <- NULL
                       seuratreactive$volcanoSDEvalue1 <- NULL
                       seuratreactive$volcanoSDEvalue2 <- NULL
                       seuratreactive$pvaluevolcanoSDE <- NULL
                       seuratreactive$PATHWAYSDE <- NULL
                       
                       seuratreactive$markersSDE <- NULL
                       seuratreactive$YuSDE <- NULL
                       seuratreactive$markersSDEvolcano <- NULL
                       
                       #PA
                       seuratreactive$PAchoose <- NULL
                       seuratreactive$padjustPA <- NULL
                       seuratreactive$pvaluePA <- NULL
                       seuratreactive$GOlabelsPA <- NULL
                       Yu$result <- NULL
                       Yu$ema <- NULL
                       seuratreactive$geneListPA <- NULL
                       
                       #SPA
                       seuratreactive$SPAchoose <- NULL
                       seuratreactive$padjustSPA <- NULL
                       seuratreactive$pvalueSPA <- NULL
                       seuratreactive$GOlabelsSPA <- NULL
                       YuSPA$result <- NULL
                       YuSPA$ema <- NULL
                       seuratreactive$geneListSPA <- NULL
                       
                       if (!is.null(seuratreactive$obj_original2)) {
                         
                         if(input$iscollapseboxQC2 == TRUE) {
                           js$collapse("QCBoxIntro2")
                         }
                         shinyjs::hide("QCBox12")
                         shinyjs::hide("QCBox22")
                         shinyjs::hide("QCBox32")
                         shinyjs::hide("QCBox42")
                         
                         shinyjs::enable("startbuttonQC2")
                         shinyjs::hide("NextButtonQC2")
                         
                         shinyjs::hide("NextButtonD")
                         
                         
                         output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                          menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                          menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                          menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                          menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                          menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                   badgeColor = "green", badgeLabel = "new"),
                                                                          menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                         ))
                       }
                       
                       if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                         shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                       }
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })  
  })
  
  observeEvent(input$buttonexistingdataMCombine, {
    if (input$existingdataMCombine == "10k Human PBMCs Multiplexed (CellPlex, 10X data)") {
      
      withProgress(message = 'Loading 10k Human PBMCs dataset...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     seuratreactive$obj_original2 <- readRDS("10k_Human_PBMCs_CellPlex.rds")
                     incProgress(0.8)
                     seuratreactive$type2 <- "multiple"
                     seuratreactive$matrixtable2 <- "yes"
                     shinyjs::show("NextButtonCombine")
                     shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                     
                     updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
                     
                     #QC2
                     seuratreactive$objQC2 <- NULL
                     seuratreactive$nFeature12 <- NULL
                     seuratreactive$nFeature22 <- NULL
                     seuratreactive$nCount12 <- NULL
                     seuratreactive$nCount22 <- NULL
                     seuratreactive$percent.mt12 <- NULL
                     seuratreactive$percent.mt22 <- NULL
                     seuratreactive$percent.ribo12 <- NULL
                     seuratreactive$percent.ribo22 <- NULL
                     
                     #D2
                     seuratreactive$Remove2 <- NULL
                     seuratreactive$plot.data.doublets2 <- NULL
                     seuratreactive$Removed2 <- NULL
                     seuratreactive$doublesim2 <- NULL
                     
                     #DR
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR <- NULL
                     seuratreactive$npcDR <- NULL
                     seuratreactive$variablefeatures <- NULL
                     seuratreactive$ALRAL <- NULL
                     
                     #DR2
                     seuratreactive$obj <- NULL
                     seuratreactive$SCTransformDR2 <- NULL
                     seuratreactive$npcDR2 <- NULL
                     seuratreactive$Combined.list <- NULL
                     seuratreactive$VFintegrated <- NULL
                     seuratreactive$integrationreduction <- NULL
                     seuratreactive$kanchor <- NULL
                     
                     #C
                     seuratreactive$integera <- NULL
                     seuratreactive$integerb <- NULL
                     seuratreactive$integerc <- NULL
                     seuratreactive$integerd <- NULL
                     seuratreactive$integere <- NULL
                     seuratreactive$integerf <- NULL
                     seuratreactive$integerg <- NULL
                     seuratreactive$select1 <- NULL
                     seuratreactive$tsne1 <- NULL
                     seuratreactive$tsne2 <- NULL
                     seuratreactive$tsne3 <- NULL
                     seuratreactive$plot.data <- NULL
                     seuratreactive$plot.data.original <- NULL
                     
                     seuratreactive$Identifiers <- NULL
                     
                     #CC
                     seuratreactive$reductionCC1 <- NULL
                     seuratreactive$reductionCC2 <- NULL
                     seuratreactive$reductionCC3 <- NULL
                     seuratreactive$GeneR <- NULL
                     
                     
                     #L
                     seuratreactive$reference_annotation <- NULL
                     seuratreactive$select2 <- NULL
                     seuratreactive$selecttissue <- NULL
                     seuratreactive$new.cell.ids <- NULL
                     seuratreactive$markers <- NULL
                     
                     #MEGENA
                     seuratreactive$MEGENA_samples <- NULL
                     seuratreactive$variablegenesMEGENA <- NULL
                     seuratreactive$MEGENA_selectmethod <- NULL
                     seuratreactive$MEGENApval <- NULL
                     seuratreactive$MEGENAmin.size <- NULL
                     seuratreactive$MEGENA_selectcluster <- NULL
                     seuratreactive$n_MEGENA <- NULL
                     seuratreactive$log2FCMEGENA <- NULL
                     seuratreactive$padjustMEGENA <- NULL
                     seuratreactive$testuseMEGENA <- NULL
                     seuratreactive$minMEGENA <- NULL
                     seuratreactive$pvalMEGENA <- NULL
                     seuratreactive$summary.output <- NULL
                     seuratreactive$markersMEGENA <- NULL
                     seuratreactive$moduleGO_df <- NULL
                     seuratreactive$VF <- NULL
                     seuratreactive$g <- NULL
                     
                     #SCENIC
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     #MAGMA
                     seuratreactive$MAGMA_samples <- NULL
                     seuratreactive$MAGMA_GWAS <- NULL
                     seuratreactive$MAGMA_Final_Results <- NULL
                     seuratreactive$meta <- NULL
                     seuratreactive$MAGMA_selectcluster <- NULL
                     seuratreactive$MAGMA_selectcluster2 <- NULL
                     seuratreactive$MAGMA_samples2 <- NULL
                     seuratreactive$textlabelMAGMA <- NULL
                     seuratreactive$MAGMA_N <- NULL
                     seuratreactive$GRCh <- NULL
                     seuratreactive$MAGMA_Pop <- NULL
                     seuratreactive$MAGMA_dirs <- NULL
                     seuratreactive$genesOutPath2 <- NULL
                     seuratreactive$n_MAGMA <- NULL
                     seuratreactive$n_MAGMA2 <- NULL
                     
                     #CellChat
                     seuratreactive$cellchat <- NULL
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- NULL
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- NULL
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- NULL
                     seuratreactive$Chat_number <- NULL
                     seuratreactive$Chatpval <- NULL
                     
                     #DRUG
                     seuratreactive$markersDRUG <- NULL
                     seuratreactive$DRUGdf <- NULL
                     
                     #DRUG2
                     seuratreactive$markersDRUG2 <- NULL
                     seuratreactive$DRUGdf2 <- NULL
                     
                     #GE
                     seuratreactive$titleGE <- NULL
                     seuratreactive$GEInput <- NULL
                     seuratreactive$GSEInput <- NULL
                     seuratreactive$p <- NULL
                     seuratreactive$q <- NULL
                     seuratreactive$objGEyes <- NULL
                     seuratreactive$Gene_vector_msig <- NULL
                     
                     #TA
                     seuratreactive$TAselectedcells <- NULL
                     seuratreactive$TA_pickercluster <- NULL
                     seuratreactive$seurat_monocle <- NULL
                     seuratreactive$Monocle3_genes <- NULL
                     
                     #MM
                     seuratreactive$MM <- NULL
                     seuratreactive$plot.dataM <- NULL
                     seuratreactive$plot.dataMM <- NULL
                     
                     #DE
                     seuratreactive$onlyposDE <- NULL
                     seuratreactive$Subset_GENES <- NULL
                     seuratreactive$DE_clusters_samples <- NULL
                     seuratreactive$DECluster <- NULL
                     seuratreactive$DEInput1_samples <- NULL
                     seuratreactive$DEInput2_samples <- NULL
                     seuratreactive$SelectComparison <- NULL
                     seuratreactive$DE_clusters <- NULL
                     seuratreactive$GOlabelsDE <- NULL
                     seuratreactive$DEInput1 <- NULL
                     seuratreactive$DEInput2 <- NULL
                     seuratreactive$logDE <- NULL
                     seuratreactive$minDE <- NULL
                     seuratreactive$padjustDE1 <- NULL
                     seuratreactive$testuseDE <- NULL
                     seuratreactive$enrichInputDE <- NULL
                     seuratreactive$pvalueDE <- NULL
                     seuratreactive$pvalDE <- NULL
                     seuratreactive$padjustDE <- NULL
                     seuratreactive$volcanovalue1 <- NULL
                     seuratreactive$volcanovalue2 <- NULL
                     seuratreactive$pvaluevolcano <- NULL
                     seuratreactive$markersDE <- NULL
                     seuratreactive$YuDE <- NULL
                     seuratreactive$markersDEvolcano <- NULL
                     seuratreactive$PATHWAY <- NULL
                     
                     #SDE
                     seuratreactive$Subset_GENESSDE <- NULL
                     seuratreactive$selectedkey1 <- NULL
                     seuratreactive$selectedkey2 <- NULL
                     seuratreactive$onlyposSDE <- NULL
                     seuratreactive$GOlabelsSDE <- NULL
                     seuratreactive$logSDE <- NULL
                     seuratreactive$minSDE <- NULL
                     seuratreactive$padjustSDE1 <- NULL
                     seuratreactive$testuseSDE <- NULL
                     seuratreactive$enrichInputSDE <- NULL
                     seuratreactive$pvalueSDE <- NULL
                     seuratreactive$pvalSDE <- NULL
                     seuratreactive$padjustSDE <- NULL
                     seuratreactive$volcanoSDEvalue1 <- NULL
                     seuratreactive$volcanoSDEvalue2 <- NULL
                     seuratreactive$pvaluevolcanoSDE <- NULL
                     seuratreactive$PATHWAYSDE <- NULL
                     
                     seuratreactive$markersSDE <- NULL
                     seuratreactive$YuSDE <- NULL
                     seuratreactive$markersSDEvolcano <- NULL
                     
                     #PA
                     seuratreactive$PAchoose <- NULL
                     seuratreactive$padjustPA <- NULL
                     seuratreactive$pvaluePA <- NULL
                     seuratreactive$GOlabelsPA <- NULL
                     Yu$result <- NULL
                     Yu$ema <- NULL
                     seuratreactive$geneListPA <- NULL
                     
                     #SPA
                     seuratreactive$SPAchoose <- NULL
                     seuratreactive$padjustSPA <- NULL
                     seuratreactive$pvalueSPA <- NULL
                     seuratreactive$GOlabelsSPA <- NULL
                     YuSPA$result <- NULL
                     YuSPA$ema <- NULL
                     seuratreactive$geneListSPA <- NULL
                     
                     if (!is.null(seuratreactive$obj_original2)) {
                       
                       if(input$iscollapseboxQC2 == TRUE) {
                         js$collapse("QCBoxIntro2")
                       }
                       shinyjs::hide("QCBox12")
                       shinyjs::hide("QCBox22")
                       shinyjs::hide("QCBox32")
                       shinyjs::hide("QCBox42")
                       
                       shinyjs::enable("startbuttonQC2")
                       shinyjs::hide("NextButtonQC2")
                       
                       shinyjs::hide("NextButtonD")
                       
                       
                       output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                                        menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                                        menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                                        menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                                        menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T),
                                                                        menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F,
                                                                                 badgeColor = "green", badgeLabel = "new"),
                                                                        menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
                       ))
                     }
                     
                     if (identical(seuratreactive$obj_original@assays$RNA@counts, seuratreactive$obj_original2@assays$RNA@counts)) {
                       shinyalert("Warning!", "You have uploaded two identical datasets, this may affect the integration process...", type = "warning", confirmButtonCol = "#337ab7")      
                     }
                   })
    }
  })
  
  output$MatrixTableCombine <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$obj_original2), 'You must load your data first...'))
    
    head(as.data.frame(seuratreactive$obj_original2@assays$RNA@counts), c(500L, 500L)) %>% na.omit()
    
  }, options = list(scrollX = TRUE))
  
  observeEvent(input$NextButtonCombine, {
    
    if (!is.null(seuratreactive$obj_original2)) {
      
      if(input$iscollapseboxQC2 == TRUE) {
        js$collapse("QCBoxIntro2")
      }
      shinyjs::hide("QCBox12")
      shinyjs::hide("QCBox22")
      shinyjs::hide("QCBox32")
      shinyjs::hide("QCBox42")
      
      shinyjs::enable("startbuttonQC2")
      shinyjs::hide("NextButtonQC2")
      
      shinyjs::hide("NextButtonCombine")
      
      shinyjs::hide("NextButtonD")
      
      
      
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
    else{
      shinyalert("ERROR!", "Please reupload your file.", type = "error", confirmButtonCol = "#337ab7")
    }
  },ignoreInit = T)
  
  ####TAB2 - QC (1st dataset)####
  
  observeEvent(input$tabs, {
    if (input$tabs == "QC" & !is.null(seuratreactive$variablefeatures) | input$tabs == "QC" & !is.null(seuratreactive$Removed) | input$tabs == "QC" & !is.null(seuratreactive$VFintegrated)) {
      shinyalert("Warning!", "If you adjust quality control parameters, all the following tabs will disappear and you will have to 
                 re-do the analysis at each step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (is.na(input$nFeature1) |
        is.na(input$nFeature2) |
        is.na(input$nCount1) |
        is.na(input$nCount2) |
        is.na(input$percent.mt1) |
        is.na(input$percent.mt2) |
        is.na(input$percent.ribo1) |
        is.na(input$percent.ribo2))
    {
      shinyjs::disable("SQC")
    }
    else {
      shinyjs::enable("SQC")
    }
  })
  
  observeEvent(input$startbuttonQC |input$rewindQC, {
    tryCatch({
      seuratreactive$obj_original[["percent.mt"]] <- PercentageFeatureSet(seuratreactive$obj_original, pattern = "^MT-|^mt-|^Mt-")
      seuratreactive$obj_original[["percent.ribo"]] <- PercentageFeatureSet(seuratreactive$obj_original, pattern = "^RP[SL][[:digit:]]|^RPLP[[:digit:]]|^RPSA|
                                                                              ^Rp[sl][[:digit:]]|^Rplp[[:digit:]]|^Rpsa|
                                                                              ^rp[sl][[:digit:]]|^rplp[[:digit:]]|^rpsa")
      
      if(sum(seuratreactive$obj_original[["percent.mt"]]) == 0) {
        shinyalert("WARNING!", "No Mitochondrial genes detected in dataset (no genes that begin with MT detected), please check your gene names.", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      if(sum(seuratreactive$obj_original[["percent.ribo"]]) == 0){
        shinyalert("WARNING!", "No Ribosomal genes detected in dataset (no genes that begin with RPS, RPL, RPLP, RPSA detected), please check your gene names.", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      seuratreactive$objQC <- seuratreactive$obj_original
      
      seuratreactive$nFeature1 <- "0"
      seuratreactive$nFeature2 <- "100000"
      seuratreactive$nCount1 <- "0"
      seuratreactive$nCount2 <- "1000000"
      seuratreactive$percent.mt1 <- "0"
      seuratreactive$percent.mt2 <- "100"
      seuratreactive$percent.ribo1 <- "0"
      seuratreactive$percent.ribo2 <- "100"
      
      if(input$iscollapseboxQC == FALSE) {
        js$collapse("QCBoxIntro")
      }
      shinyjs::show("QCBox1")
      shinyjs::show("QCBox2")
      shinyjs::show("QCBox3")
      shinyjs::show("QCBox4")
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check you have the correct seurat object uploaded.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T
  )
  
  observeEvent(input$SQC, {
    tryCatch({
      seuratreactive$objQC <- subset(seuratreactive$obj_original, subset = nFeature_RNA >= input$nFeature1 & nFeature_RNA <= input$nFeature2
                                     & nCount_RNA >= input$nCount1 & nCount_RNA <= input$nCount2
                                     & percent.mt >= input$percent.mt1 & percent.mt <= input$percent.mt2
                                     & percent.ribo >= input$percent.ribo1 & percent.ribo <= input$percent.ribo2
      )
      
      seuratreactive$nFeature1 <- input$nFeature1
      seuratreactive$nFeature2 <- input$nFeature2
      seuratreactive$nCount1 <- input$nCount1
      seuratreactive$nCount2 <- input$nCount2
      seuratreactive$percent.mt1 <- input$percent.mt1
      seuratreactive$percent.mt2 <- input$percent.mt2
      seuratreactive$percent.ribo1 <- input$percent.ribo1
      seuratreactive$percent.ribo2 <- input$percent.ribo2
    },
    error = function(e) {
      shinyalert("ERROR!", "Selection parameters out of bounds, please re-select parameters.", type = "error", confirmButtonCol = "#337ab7")
    }
    )
  })
  
  output$plotUI <- renderUI({
    plotOutput("plot", height = (400 + input$QCplotheight*40))
  })
  
  output$plot <- renderPlot({
    validate(
      need(!is.null(seuratreactive$objQC), "You must complete the quality control step"))
    VlnPlot(seuratreactive$objQC, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"), group.by = "orig.ident", pt.size = input$sizeQC, ncol = 2) 
  })
  
  output$plot2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$objQC), "You must complete the quality control step"))
    FeatureScatter(seuratreactive$objQC, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident", pt.size = 2) + NoLegend() })
  
  output$plot3 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$objQC), "You must complete the quality control step"))
    FeatureScatter(seuratreactive$objQC, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident", pt.size = 2) + NoLegend() })
  
  observeEvent(input$SQC | input$rewindQC | input$startbuttonQC, {
    
    if (!is.null(seuratreactive$objQC)) {
      if(input$iscollapseboxD == TRUE) {
        js$collapse("DBoxIntro")
      }
      shinyjs::hide("DBox1")
      shinyjs::hide("DBox2")
      
      
      shinyjs::hide("NextButtonD")
      shinyjs::enable("startbuttonD")
      
      shinyjs::disable("startbuttonQC")
      shinyjs::show("NextButtonQC")
      
      shinyjs::hide("NextButtonLoad")
      
      updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      
      
      #Doublet
      seuratreactive$plot.data.doublets <- NULL
      seuratreactive$Remove <- NULL
      seuratreactive$Removed <- NULL
      seuratreactive$doublesim <- NULL
      
      #Load2
      seuratreactive$obj_original2 <- NULL
      seuratreactive$matrixtable2 <- NULL
      seuratreactive$mincellsCombine <- NULL
      seuratreactive$minfeaturesCombine <- NULL
      
      #QC2
      seuratreactive$type2 <- NULL
      seuratreactive$objQC2 <- NULL
      seuratreactive$nFeature12 <- NULL
      seuratreactive$nFeature22 <- NULL
      seuratreactive$nCount12 <- NULL
      seuratreactive$nCount22 <- NULL
      seuratreactive$percent.mt12 <- NULL
      seuratreactive$percent.mt22 <- NULL
      seuratreactive$percent.ribo12 <- NULL
      seuratreactive$percent.ribo22 <- NULL
      
      #D2
      seuratreactive$Remove2 <- NULL
      seuratreactive$plot.data.doublets2 <- NULL
      seuratreactive$Removed2 <- NULL
      seuratreactive$doublesim2 <- NULL
      
      #DR
      seuratreactive$obj <- NULL
      seuratreactive$SCTransformDR <- NULL
      seuratreactive$npcDR <- NULL
      seuratreactive$variablefeatures <- NULL
      seuratreactive$ALRAL <- NULL
      
      #DR2
      seuratreactive$obj <- NULL
      seuratreactive$SCTransformDR2 <- NULL
      seuratreactive$npcDR2 <- NULL
      seuratreactive$Combined.list <- NULL
      seuratreactive$VFintegrated <- NULL
      seuratreactive$integrationreduction <- NULL
      seuratreactive$kanchor <- NULL
      
      #C
      seuratreactive$integera <- NULL
      seuratreactive$integerb <- NULL
      seuratreactive$integerc <- NULL
      seuratreactive$integerd <- NULL
      seuratreactive$integere <- NULL
      seuratreactive$integerf <- NULL
      seuratreactive$integerg <- NULL
      seuratreactive$select1 <- NULL
      seuratreactive$tsne1 <- NULL
      seuratreactive$tsne2 <- NULL
      seuratreactive$tsne3 <- NULL
      seuratreactive$plot.data <- NULL
      seuratreactive$plot.data.original <- NULL
      
      seuratreactive$Identifiers <- NULL
      
      #CC
      seuratreactive$reductionCC1 <- NULL
      seuratreactive$reductionCC2 <- NULL
      seuratreactive$reductionCC3 <- NULL
      seuratreactive$GeneR <- NULL
      
      
      #L
      seuratreactive$reference_annotation <- NULL
      seuratreactive$select2 <- NULL
      seuratreactive$selecttissue <- NULL
      seuratreactive$new.cell.ids <- NULL
      seuratreactive$markers <- NULL
      
      #MEGENA
      seuratreactive$MEGENA_samples <- NULL
      seuratreactive$variablegenesMEGENA <- NULL
      seuratreactive$MEGENA_selectmethod <- NULL
      seuratreactive$MEGENApval <- NULL
      seuratreactive$MEGENAmin.size <- NULL
      seuratreactive$MEGENA_selectcluster <- NULL
      seuratreactive$n_MEGENA <- NULL
      seuratreactive$log2FCMEGENA <- NULL
      seuratreactive$padjustMEGENA <- NULL
      seuratreactive$testuseMEGENA <- NULL
      seuratreactive$minMEGENA <- NULL
      seuratreactive$pvalMEGENA <- NULL
      seuratreactive$summary.output <- NULL
      seuratreactive$markersMEGENA <- NULL
      seuratreactive$moduleGO_df <- NULL
      seuratreactive$VF <- NULL
      seuratreactive$g <- NULL
      
      #SCENIC
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$variablegenesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$n_SCENIC <- NULL
      seuratreactive$SCENICdb <- NULL
      seuratreactive$SCENIC_samples <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$SCENIC_selectcluster <- NULL
      seuratreactive$SCENIC_selectmethod <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
      
      #MAGMA
      seuratreactive$MAGMA_samples <- NULL
      seuratreactive$MAGMA_GWAS <- NULL
      seuratreactive$MAGMA_Final_Results <- NULL
      seuratreactive$meta <- NULL
      seuratreactive$MAGMA_selectcluster <- NULL
      seuratreactive$MAGMA_selectcluster2 <- NULL
      seuratreactive$MAGMA_samples2 <- NULL
      seuratreactive$textlabelMAGMA <- NULL
      seuratreactive$MAGMA_N <- NULL
      seuratreactive$GRCh <- NULL
      seuratreactive$MAGMA_Pop <- NULL
      seuratreactive$MAGMA_dirs <- NULL
      seuratreactive$genesOutPath2 <- NULL
      seuratreactive$n_MAGMA <- NULL
      seuratreactive$n_MAGMA2 <- NULL
      
      #CellChat
      seuratreactive$cellchat <- NULL
      seuratreactive$cellchatM <- NULL
      seuratreactive$object.list <- NULL
      seuratreactive$ChatChoose <- NULL
      seuratreactive$ChatSample <- NULL
      seuratreactive$Chatdb <- NULL
      seuratreactive$ChatMultiple1Sample <- NULL
      seuratreactive$ChatMultiple2Sample <- NULL
      seuratreactive$Chat_selectcluster <- NULL
      seuratreactive$Chat_number <- NULL
      seuratreactive$Chatpval <- NULL
      
      #DRUG
      seuratreactive$markersDRUG <- NULL
      seuratreactive$DRUGdf <- NULL
      
      #DRUG2
      seuratreactive$markersDRUG2 <- NULL
      seuratreactive$DRUGdf2 <- NULL
      
      #GE
      seuratreactive$titleGE <- NULL
      seuratreactive$GEInput <- NULL
      seuratreactive$GSEInput <- NULL
      seuratreactive$p <- NULL
      seuratreactive$q <- NULL
      seuratreactive$objGEyes <- NULL
      seuratreactive$Gene_vector_msig <- NULL
      
      #TA
      seuratreactive$TAselectedcells <- NULL
      seuratreactive$TA_pickercluster <- NULL
      seuratreactive$seurat_monocle <- NULL
      seuratreactive$Monocle3_genes <- NULL
      
      #MM
      seuratreactive$MM <- NULL
      seuratreactive$plot.dataM <- NULL
      seuratreactive$plot.dataMM <- NULL
      
      #DE
      seuratreactive$onlyposDE <- NULL
      seuratreactive$Subset_GENES <- NULL
      seuratreactive$DE_clusters_samples <- NULL
      seuratreactive$DECluster <- NULL
      seuratreactive$DEInput1_samples <- NULL
      seuratreactive$DEInput2_samples <- NULL
      seuratreactive$SelectComparison <- NULL
      seuratreactive$DE_clusters <- NULL
      seuratreactive$GOlabelsDE <- NULL
      seuratreactive$DEInput1 <- NULL
      seuratreactive$DEInput2 <- NULL
      seuratreactive$logDE <- NULL
      seuratreactive$minDE <- NULL
      seuratreactive$padjustDE1 <- NULL
      seuratreactive$testuseDE <- NULL
      seuratreactive$enrichInputDE <- NULL
      seuratreactive$pvalueDE <- NULL
      seuratreactive$pvalDE <- NULL
      seuratreactive$padjustDE <- NULL
      seuratreactive$volcanovalue1 <- NULL
      seuratreactive$volcanovalue2 <- NULL
      seuratreactive$pvaluevolcano <- NULL
      seuratreactive$markersDE <- NULL
      seuratreactive$YuDE <- NULL
      seuratreactive$markersDEvolcano <- NULL
      seuratreactive$PATHWAY <- NULL
      
      #SDE
      seuratreactive$Subset_GENESSDE <- NULL
      seuratreactive$selectedkey1 <- NULL
      seuratreactive$selectedkey2 <- NULL
      seuratreactive$onlyposSDE <- NULL
      seuratreactive$GOlabelsSDE <- NULL
      seuratreactive$logSDE <- NULL
      seuratreactive$minSDE <- NULL
      seuratreactive$padjustSDE1 <- NULL
      seuratreactive$testuseSDE <- NULL
      seuratreactive$enrichInputSDE <- NULL
      seuratreactive$pvalueSDE <- NULL
      seuratreactive$pvalSDE <- NULL
      seuratreactive$padjustSDE <- NULL
      seuratreactive$volcanoSDEvalue1 <- NULL
      seuratreactive$volcanoSDEvalue2 <- NULL
      seuratreactive$pvaluevolcanoSDE <- NULL
      seuratreactive$PATHWAYSDE <- NULL
      
      seuratreactive$markersSDE <- NULL
      seuratreactive$YuSDE <- NULL
      seuratreactive$markersSDEvolcano <- NULL
      
      #PA
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      #SPA
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = T),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F,
                                                                badgeColor = "green", badgeLabel = "optional"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
  },ignoreInit = T)
  
  observeEvent(input$NextButtonQC, {
    
    if(input$iscollapseboxD == TRUE) {
      js$collapse("DBoxIntro")
    }
    shinyjs::hide("DBox1")
    shinyjs::hide("DBox2")
    
    
    shinyjs::hide("NextButtonD")
    shinyjs::enable("startbuttonD")
    
    shinyjs::disable("startbuttonQC")
    shinyjs::hide("NextButtonQC")
    
    shinyjs::hide("NextButtonLoad")
    
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = T,
                                                              badgeColor = "green", badgeLabel = "optional"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T) 
  
  ####TAB2.25 - Doublet Removal (1st dataset) ####
  observeEvent(input$tabs, {
    if (input$tabs == "Doublet" & !is.null(seuratreactive$objQC2) | input$tabs == "Doublet" & !is.null(seuratreactive$variablefeatures)) {
      shinyalert("Warning!", "If you adjust doublet removal, all the following tabs will disappear and you will have to 
                 re-do the analysis at each step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (is.na(input$doublesim)) {
      shinyjs::disable("doublesimGO")
    }
  })
  
  observeEvent(input$startbuttonD, {
    library(PCAtools)
    library(DoubletFinder)
    
    if (!is.null(seuratreactive$obj_original) & length(unique(seuratreactive$obj_original@meta.data$orig.ident)) > 1) {
      shinyalert("Warning!", "File contains multiple samples, doublet detection on aggregated scRNA-seq data representing
                    multiple distinct samples (i.e., control vs treatment, samples sequenced on different lanes) may be inaccurate due to
                   technical differences. Proceed with caution.", type = "warning", confirmButtonCol = "#337ab7")      
    }
    
    if(input$iscollapseboxD == FALSE) {
      js$collapse("DBoxIntro")
    }
    shinyjs::show("DBox1")
    shinyjs::show("DBox2")
    
    
    shinyjs::hide("NextButtonCombine")
    
    shinyjs::disable("startbuttonD")
    shinyjs::show("NextButtonD")
    
    shinyjs::hide("NextButtonQC")
  }, ignoreInit = T)
  
  observe(
    if (is.null(seuratreactive$Remove)) {
      shinyjs::disable("doubleremove")
    }
    else {
      shinyjs::enable("doubleremove")
    }
  )
  
  observeEvent(input$doublesimGO, {
    tryCatch({
      withProgress(message = 'Detecting doublets. Please wait...',
                   value = 0, {
                     incProgress(1/20)
                     
                     seuratreactive$objQC <- subset(seuratreactive$obj_original, subset = nFeature_RNA >= as.numeric(seuratreactive$nFeature1) & nFeature_RNA <= as.numeric(seuratreactive$nFeature2)
                                                    & nCount_RNA >= as.numeric(seuratreactive$nCount1) & nCount_RNA <= as.numeric(seuratreactive$nCount2)
                                                    & percent.mt >= as.numeric(seuratreactive$percent.mt1) & percent.mt <= as.numeric(seuratreactive$percent.mt2)
                                                    & percent.ribo >= as.numeric(seuratreactive$percent.ribo1) & percent.ribo <= as.numeric(seuratreactive$percent.ribo2))
                     
                     #Identification of variable features
                     seuratreactive$objQC <- FindVariableFeatures(seuratreactive$objQC, selection.method = "vst", nfeatures = 2000)
                     
                     #Normalization
                     seuratreactive$objQC <- NormalizeData(seuratreactive$objQC)
                     
                     #Data scaling
                     seuratreactive$objQC <- ScaleData(seuratreactive$objQC)
                     
                     #Dimensionality reduction
                     seuratreactive$objQC <- RunPCA(seuratreactive$objQC, npcs = 50)
                     incProgress(7/20)
                     
                     elbow <- ElbowPlot(seuratreactive$objQC, ndims = 50)
                     
                     Elbowpoint <- min(elbow$data[elbow$data$stdev <= findElbowPoint(elbow$data$stdev),]$dims) + 5
                     
                     seuratreactive$objQC <- FindNeighbors(seuratreactive$objQC, dims = 1:Elbowpoint)
                     seuratreactive$objQC <- FindClusters(seuratreactive$objQC)
                     
                     #2D
                     #RunUMAP
                     seuratreactive$objQC <- RunUMAP(seuratreactive$objQC, dims = 1:Elbowpoint, n.components = 2L)
                     incProgress(9/20)
                     #Run tsne
                     seuratreactive$objQC <- RunTSNE(seuratreactive$objQC, dims = 1:Elbowpoint, dim.embed = 2)
                     
                     #3D
                     seuratreactive$objQC <- RunUMAP(seuratreactive$objQC, dims = 1:Elbowpoint, n.components = 3L)
                     incProgress(11/20)
                     #Run tsne
                     seuratreactive$objQC <- RunTSNE(seuratreactive$objQC, dims = 1:Elbowpoint, dim.embed = 3)
                     incProgress(13/20)
                     
                     # pK Identification (no ground-truth)
                     sweep.res.list <- paramSweep_v3(seuratreactive$objQC, PCs = 1:Elbowpoint, sct = FALSE)
                     sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
                     bcmvn <- find.pK(sweep.stats)
                     
                     incProgress(15/20)
                     
                     # Homotypic Doublet Proportion Estimate
                     homotypic.prop <- modelHomotypic(seuratreactive$objQC@active.ident)           
                     nExp_poi <- round(input$doublesim/100*nrow(seuratreactive$objQC@meta.data))  ## Assuming 4% doublet formation rate - tailor for your dataset
                     nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
                     
                     seuratreactive$doublesim <- input$doublesim
                     
                     seuratreactive$objQC <- doubletFinder_v3(seuratreactive$objQC, PCs = 1:Elbowpoint, pN = 0.25, pK = as.numeric(paste0(bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK)), nExp = nExp_poi, reuse.pANN = F, sct = FALSE)
                     seuratreactive$objQC <- doubletFinder_v3(seuratreactive$objQC, PCs = 1:Elbowpoint, pN = 0.25, pK = as.numeric(paste0(bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK)), nExp = nExp_poi.adj, 
                                                              reuse.pANN = paste0("pANN_", "0.25_", bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK, "_", nExp_poi), sct = FALSE)
                     
                     incProgress(19/20)
                     
                     seuratreactive$Remove <- cbind.data.frame(rownames(seuratreactive$objQC@meta.data), seuratreactive$objQC@meta.data[[paste0("DF.classifications_", "0.25_", bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK, "_", nExp_poi.adj)]])
                     colnames(seuratreactive$Remove) <- c("Cell", "Doublet")
                     
                     seuratreactive$Remove <- seuratreactive$Remove[grep('Doublet', seuratreactive$Remove$Doublet),]
                     seuratreactive$Remove <- as.list(seuratreactive$Remove$Cell)
                     
                     #Plot
                     plot.data2D <- FetchData(object = seuratreactive$objQC, 
                                              vars = c("UMAP_1", "UMAP_2", "tSNE_1", "tSNE_2", "orig.ident", 
                                                       paste0("DF.classifications_", "0.25_", bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK, "_", nExp_poi.adj)))
                     
                     colnames(plot.data2D) <- c("UMAP_1_2D", "UMAP_2_2D", "tSNE_1_2D", "tSNE_2_2D", "Sample", "Detected Doublets")
                     
                     
                     plot.data3D <- FetchData(object = seuratreactive$objQC, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "tSNE_1", "tSNE_2", "tSNE_3"))
                     
                     colnames(plot.data3D) <- c("UMAP_1_3D", "UMAP_2_3D", "UMAP_3_3D", "tSNE_1_3D", "tSNE_2_3D", "tSNE_3_3D")
                     
                     seuratreactive$plot.data.doublets <- merge(plot.data2D, plot.data3D, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data.doublets) <- seuratreactive$plot.data.doublets$Row.names
                     
                     seuratreactive$plot.data.doublets <- seuratreactive$plot.data.doublets[,-1]
                     
                     
                     shinyjs::hide("NextButtonCombine")
                     
                     shinyjs::show("NextButtonD")
                     shinyjs::disable("startbuttonD")
                     
                     shinyjs::hide("NextButtonQC")
                     
                     incProgress(20/20)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Selection parameters out of bounds, please re-select parameters.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$doubleremove, {
    tryCatch({
      withProgress(message = 'Performing Dimensionality Reduction. Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     if (is.null(seuratreactive$Remove)) {
                       shinyalert("ERROR!", "Doublet removal failed. You need to detect your doublets first.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       incProgress(5/10)
                       
                       seuratreactive$objQC <- subset(seuratreactive$obj_original, subset = nFeature_RNA >= as.numeric(seuratreactive$nFeature1) & nFeature_RNA <= as.numeric(seuratreactive$nFeature2)
                                                      & nCount_RNA >= as.numeric(seuratreactive$nCount1) & nCount_RNA <= as.numeric(seuratreactive$nCount2)
                                                      & percent.mt >= as.numeric(seuratreactive$percent.mt1) & percent.mt <= as.numeric(seuratreactive$percent.mt2)
                                                      & percent.ribo >= as.numeric(seuratreactive$percent.ribo1) & percent.ribo <= as.numeric(seuratreactive$percent.ribo2))
                       incProgress(7/10)
                       
                       seuratreactive$objQC <- seuratreactive$objQC[,!colnames(seuratreactive$objQC) %in% seuratreactive$Remove]
                     }
                     
                     incProgress(10/10)
                     
                     seuratreactive$Removed <- "yes"
                     
                     
                     shinyjs::hide("NextButtonCombine")
                     
                     shinyjs::show("NextButtonD")
                     shinyjs::disable("startbuttonD")
                     
                     shinyjs::hide("NextButtonQC")
                     
                     shinyalert("SUCCESS!", "Doublets Removed!", type = "success", confirmButtonCol = "#337ab7")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "You have already removed your doublets!", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$plotdoublesimUI <- renderUI({
    plotlyOutput("plotdoublesim", height = (400 + input$Dplotheight*40))
  })
  
  output$plotdoublesim <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data.doublets), "Please select your estimated percentage of doublets and click SIMULATE DOUBLETS to begin the detection process"))
    if(input$radioD == "UMAP") {
      plot_ly(data = seuratreactive$plot.data.doublets, 
              x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
              color = seuratreactive$plot.data.doublets[, input$plotclustersD],
              colors = col_vector,
              type = "scatter", 
              mode = "markers", 
              marker = list(size = input$sizeD),
              opacity = input$opacityD,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD))
    }
    else if (input$radioD == "3D UMAP") {
      plot_ly(data = seuratreactive$plot.data.doublets, 
              x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D, 
              color = seuratreactive$plot.data.doublets[, input$plotclustersD],
              colors = col_vector,
              type = "scatter3d", 
              mode = "markers", 
              marker = list(size = input$sizeD),
              opacity = input$opacityD,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD))
    }
    else if (input$radioD == "t-SNE") {
      plot_ly(data = seuratreactive$plot.data.doublets, 
              x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
              color = seuratreactive$plot.data.doublets[, input$plotclustersD],
              colors = col_vector,
              type = "scatter", 
              mode = "markers", 
              marker = list(size = input$sizeD), 
              opacity = input$opacityD,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD))
    }
    else if (input$radioD == "3D t-SNE") {
      plot_ly(data = seuratreactive$plot.data.doublets, 
              x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
              color = seuratreactive$plot.data.doublets[, input$plotclustersD], 
              colors = col_vector,
              type = "scatter3d", 
              mode = "markers", 
              marker = list(size = input$sizeD),
              opacity = input$opacityD,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD))
    }
  })
  
  observeEvent(input$skipbuttonCombine, {
    updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
    
    #Load2
    seuratreactive$obj_original2 <- NULL
    seuratreactive$matrixtable2 <- NULL
    seuratreactive$mincellsCombine <- NULL
    seuratreactive$minfeaturesCombine <- NULL
    
    #QC2
    seuratreactive$type2 <- NULL
    seuratreactive$objQC2 <- NULL
    seuratreactive$nFeature12 <- NULL
    seuratreactive$nFeature22 <- NULL
    seuratreactive$nCount12 <- NULL
    seuratreactive$nCount22 <- NULL
    seuratreactive$percent.mt12 <- NULL
    seuratreactive$percent.mt22 <- NULL
    seuratreactive$percent.ribo12 <- NULL
    seuratreactive$percent.ribo22 <- NULL
    
    #D2
    seuratreactive$Remove2 <- NULL
    seuratreactive$plot.data.doublets2 <- NULL
    seuratreactive$Removed2 <- NULL
    seuratreactive$doublesim2 <- NULL
    
    #DR
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR <- NULL
    seuratreactive$npcDR <- NULL
    seuratreactive$variablefeatures <- NULL
    seuratreactive$ALRAL <- NULL
    
    #DR2
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR2 <- NULL
    seuratreactive$npcDR2 <- NULL
    seuratreactive$Combined.list <- NULL
    seuratreactive$VFintegrated <- NULL
    seuratreactive$integrationreduction <- NULL
    seuratreactive$kanchor <- NULL
    
    #C
    seuratreactive$integera <- NULL
    seuratreactive$integerb <- NULL
    seuratreactive$integerc <- NULL
    seuratreactive$integerd <- NULL
    seuratreactive$integere <- NULL
    seuratreactive$integerf <- NULL
    seuratreactive$integerg <- NULL
    seuratreactive$select1 <- NULL
    seuratreactive$tsne1 <- NULL
    seuratreactive$tsne2 <- NULL
    seuratreactive$tsne3 <- NULL
    seuratreactive$plot.data <- NULL
    seuratreactive$plot.data.original <- NULL
    
    seuratreactive$Identifiers <- NULL
    
    #CC
    seuratreactive$reductionCC1 <- NULL
    seuratreactive$reductionCC2 <- NULL
    seuratreactive$reductionCC3 <- NULL
    seuratreactive$GeneR <- NULL
    
    
    #L
    seuratreactive$reference_annotation <- NULL
    seuratreactive$select2 <- NULL
    seuratreactive$selecttissue <- NULL
    seuratreactive$new.cell.ids <- NULL
    seuratreactive$markers <- NULL
    
    #MEGENA
    seuratreactive$MEGENA_samples <- NULL
    seuratreactive$variablegenesMEGENA <- NULL
    seuratreactive$MEGENA_selectmethod <- NULL
    seuratreactive$MEGENApval <- NULL
    seuratreactive$MEGENAmin.size <- NULL
    seuratreactive$MEGENA_selectcluster <- NULL
    seuratreactive$n_MEGENA <- NULL
    seuratreactive$log2FCMEGENA <- NULL
    seuratreactive$padjustMEGENA <- NULL
    seuratreactive$testuseMEGENA <- NULL
    seuratreactive$minMEGENA <- NULL
    seuratreactive$pvalMEGENA <- NULL
    seuratreactive$summary.output <- NULL
    seuratreactive$markersMEGENA <- NULL
    seuratreactive$moduleGO_df <- NULL
    seuratreactive$VF <- NULL
    seuratreactive$g <- NULL
    
    #SCENIC
    seuratreactive$variablefeaturesSCENIC <- NULL
    seuratreactive$variablegenesSCENIC <- NULL
    seuratreactive$regulonAUC <- NULL
    seuratreactive$regulons <- NULL
    seuratreactive$cellInfo <- NULL
    seuratreactive$n_SCENIC <- NULL
    seuratreactive$SCENICdb <- NULL
    seuratreactive$SCENIC_samples <- NULL
    seuratreactive$regulonnames <- NULL
    seuratreactive$titleSCENIC <- NULL
    seuratreactive$SCENICInput <- NULL
    seuratreactive$SCENIC_selectcluster <- NULL
    seuratreactive$SCENIC_selectmethod <- NULL
    seuratreactive$regulonActivity_byCellType_Scaled <- NULL
    seuratreactive$rss <- NULL
    
    #MAGMA
    seuratreactive$MAGMA_samples <- NULL
    seuratreactive$MAGMA_GWAS <- NULL
    seuratreactive$MAGMA_Final_Results <- NULL
    seuratreactive$meta <- NULL
    seuratreactive$MAGMA_selectcluster <- NULL
    seuratreactive$MAGMA_selectcluster2 <- NULL
    seuratreactive$MAGMA_samples2 <- NULL
    seuratreactive$textlabelMAGMA <- NULL
    seuratreactive$MAGMA_N <- NULL
    seuratreactive$GRCh <- NULL
    seuratreactive$MAGMA_Pop <- NULL
    seuratreactive$MAGMA_dirs <- NULL
    seuratreactive$genesOutPath2 <- NULL
    seuratreactive$n_MAGMA <- NULL
    seuratreactive$n_MAGMA2 <- NULL
    
    #CellChat
    seuratreactive$cellchat <- NULL
    seuratreactive$cellchatM <- NULL
    seuratreactive$object.list <- NULL
    seuratreactive$ChatChoose <- NULL
    seuratreactive$ChatSample <- NULL
    seuratreactive$Chatdb <- NULL
    seuratreactive$ChatMultiple1Sample <- NULL
    seuratreactive$ChatMultiple2Sample <- NULL
    seuratreactive$Chat_selectcluster <- NULL
    seuratreactive$Chat_number <- NULL
    seuratreactive$Chatpval <- NULL
    
    #DRUG
    seuratreactive$markersDRUG <- NULL
    seuratreactive$DRUGdf <- NULL
    
    #DRUG2
    seuratreactive$markersDRUG2 <- NULL
    seuratreactive$DRUGdf2 <- NULL
    
    #GE
    seuratreactive$titleGE <- NULL
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    seuratreactive$objGEyes <- NULL
    seuratreactive$Gene_vector_msig <- NULL
    
    #TA
    seuratreactive$TAselectedcells <- NULL
    seuratreactive$TA_pickercluster <- NULL
    seuratreactive$seurat_monocle <- NULL
    seuratreactive$Monocle3_genes <- NULL
    
    #MM
    seuratreactive$MM <- NULL
    seuratreactive$plot.dataM <- NULL
    seuratreactive$plot.dataMM <- NULL
    
    #DE
    seuratreactive$onlyposDE <- NULL
    seuratreactive$Subset_GENES <- NULL
    seuratreactive$DE_clusters_samples <- NULL
    seuratreactive$DECluster <- NULL
    seuratreactive$DEInput1_samples <- NULL
    seuratreactive$DEInput2_samples <- NULL
    seuratreactive$SelectComparison <- NULL
    seuratreactive$DE_clusters <- NULL
    seuratreactive$GOlabelsDE <- NULL
    seuratreactive$DEInput1 <- NULL
    seuratreactive$DEInput2 <- NULL
    seuratreactive$logDE <- NULL
    seuratreactive$minDE <- NULL
    seuratreactive$padjustDE1 <- NULL
    seuratreactive$testuseDE <- NULL
    seuratreactive$enrichInputDE <- NULL
    seuratreactive$pvalueDE <- NULL
    seuratreactive$pvalDE <- NULL
    seuratreactive$padjustDE <- NULL
    seuratreactive$volcanovalue1 <- NULL
    seuratreactive$volcanovalue2 <- NULL
    seuratreactive$pvaluevolcano <- NULL
    seuratreactive$markersDE <- NULL
    seuratreactive$YuDE <- NULL
    seuratreactive$markersDEvolcano <- NULL
    seuratreactive$PATHWAY <- NULL
    
    #SDE
    seuratreactive$Subset_GENESSDE <- NULL
    seuratreactive$selectedkey1 <- NULL
    seuratreactive$selectedkey2 <- NULL
    seuratreactive$onlyposSDE <- NULL
    seuratreactive$GOlabelsSDE <- NULL
    seuratreactive$logSDE <- NULL
    seuratreactive$minSDE <- NULL
    seuratreactive$padjustSDE1 <- NULL
    seuratreactive$testuseSDE <- NULL
    seuratreactive$enrichInputSDE <- NULL
    seuratreactive$pvalueSDE <- NULL
    seuratreactive$pvalSDE <- NULL
    seuratreactive$padjustSDE <- NULL
    seuratreactive$volcanoSDEvalue1 <- NULL
    seuratreactive$volcanoSDEvalue2 <- NULL
    seuratreactive$pvaluevolcanoSDE <- NULL
    seuratreactive$PATHWAYSDE <- NULL
    
    seuratreactive$markersSDE <- NULL
    seuratreactive$YuSDE <- NULL
    seuratreactive$markersSDEvolcano <- NULL
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    shinyjs::hide("CombineBox1")
    shinyjs::hide("CombineBox2")
    shinyjs::hide("CombineBox3")
    shinyjs::hide("CombineBox4")
    shinyjs::hide("CombineBox5")
    shinyjs::hide("NextButtonCombine")
    shinyjs::enable("startbuttonCombine")
    
    shinyjs::enable("startbuttonDR")
    shinyjs::hide("NextButtonDR")
    shinyjs::enable("startbuttonDR2")
    shinyjs::hide("NextButtonDR2")
    shinyjs::enable("startbuttonQC2")
    shinyjs::hide("NextButtonQC2")
    
    
    shinyjs::hide("NextButtonD")
    
    if(input$iscollapsebox1 == TRUE) {
      js$collapse("DRBoxIntro")
    }
    shinyjs::hide("DRBox1")
    shinyjs::hide("DRBox2")
    shinyjs::hide("DRBox3")
    shinyjs::hide("DRBox4")
    
    if(input$iscollapseboxDR2 == TRUE) {
      js$collapse("DRBoxIntro2")
    }
    shinyjs::hide("DRBox1DR2")
    shinyjs::hide("DRBox2DR2")
    shinyjs::hide("DRBox3DR2")
    shinyjs::hide("DRBox4DR2")
    shinyjs::hide("downloadMerged")
    
    if (length(unique(seuratreactive$objQC@meta.data$orig.ident)) == 1) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
    else {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
  }, ignoreInit = T)
  
  observeEvent(input$skipbuttonD, {
    updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
    #Doublet
    seuratreactive$plot.data.doublets <- NULL
    seuratreactive$Remove <- NULL
    seuratreactive$Removed <- NULL
    seuratreactive$doublesim <- NULL
    
    #Load2
    seuratreactive$obj_original2 <- NULL
    seuratreactive$matrixtable2 <- NULL
    seuratreactive$mincellsCombine <- NULL
    seuratreactive$minfeaturesCombine <- NULL
    
    #QC2
    seuratreactive$type2 <- NULL
    seuratreactive$objQC2 <- NULL
    seuratreactive$nFeature12 <- NULL
    seuratreactive$nFeature22 <- NULL
    seuratreactive$nCount12 <- NULL
    seuratreactive$nCount22 <- NULL
    seuratreactive$percent.mt12 <- NULL
    seuratreactive$percent.mt22 <- NULL
    seuratreactive$percent.ribo12 <- NULL
    seuratreactive$percent.ribo22 <- NULL
    
    #D2
    seuratreactive$Remove2 <- NULL
    seuratreactive$plot.data.doublets2 <- NULL
    seuratreactive$Removed2 <- NULL
    seuratreactive$doublesim2 <- NULL
    
    #DR
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR <- NULL
    seuratreactive$npcDR <- NULL
    seuratreactive$variablefeatures <- NULL
    seuratreactive$ALRAL <- NULL
    
    #DR2
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR2 <- NULL
    seuratreactive$npcDR2 <- NULL
    seuratreactive$Combined.list <- NULL
    seuratreactive$VFintegrated <- NULL
    seuratreactive$integrationreduction <- NULL
    seuratreactive$kanchor <- NULL
    
    #C
    seuratreactive$integera <- NULL
    seuratreactive$integerb <- NULL
    seuratreactive$integerc <- NULL
    seuratreactive$integerd <- NULL
    seuratreactive$integere <- NULL
    seuratreactive$integerf <- NULL
    seuratreactive$integerg <- NULL
    seuratreactive$select1 <- NULL
    seuratreactive$tsne1 <- NULL
    seuratreactive$tsne2 <- NULL
    seuratreactive$tsne3 <- NULL
    seuratreactive$plot.data <- NULL
    seuratreactive$plot.data.original <- NULL
    
    seuratreactive$Identifiers <- NULL
    
    #CC
    seuratreactive$reductionCC1 <- NULL
    seuratreactive$reductionCC2 <- NULL
    seuratreactive$reductionCC3 <- NULL
    seuratreactive$GeneR <- NULL
    
    
    #L
    seuratreactive$reference_annotation <- NULL
    seuratreactive$select2 <- NULL
    seuratreactive$selecttissue <- NULL
    seuratreactive$new.cell.ids <- NULL
    seuratreactive$markers <- NULL
    
    #MEGENA
    seuratreactive$MEGENA_samples <- NULL
    seuratreactive$variablegenesMEGENA <- NULL
    seuratreactive$MEGENA_selectmethod <- NULL
    seuratreactive$MEGENApval <- NULL
    seuratreactive$MEGENAmin.size <- NULL
    seuratreactive$MEGENA_selectcluster <- NULL
    seuratreactive$n_MEGENA <- NULL
    seuratreactive$log2FCMEGENA <- NULL
    seuratreactive$padjustMEGENA <- NULL
    seuratreactive$testuseMEGENA <- NULL
    seuratreactive$minMEGENA <- NULL
    seuratreactive$pvalMEGENA <- NULL
    seuratreactive$summary.output <- NULL
    seuratreactive$markersMEGENA <- NULL
    seuratreactive$moduleGO_df <- NULL
    seuratreactive$VF <- NULL
    seuratreactive$g <- NULL
    
    #SCENIC
    seuratreactive$variablefeaturesSCENIC <- NULL
    seuratreactive$variablegenesSCENIC <- NULL
    seuratreactive$regulonAUC <- NULL
    seuratreactive$regulons <- NULL
    seuratreactive$cellInfo <- NULL
    seuratreactive$n_SCENIC <- NULL
    seuratreactive$SCENICdb <- NULL
    seuratreactive$SCENIC_samples <- NULL
    seuratreactive$regulonnames <- NULL
    seuratreactive$titleSCENIC <- NULL
    seuratreactive$SCENICInput <- NULL
    seuratreactive$SCENIC_selectcluster <- NULL
    seuratreactive$SCENIC_selectmethod <- NULL
    seuratreactive$regulonActivity_byCellType_Scaled <- NULL
    seuratreactive$rss <- NULL
    
    #MAGMA
    seuratreactive$MAGMA_samples <- NULL
    seuratreactive$MAGMA_GWAS <- NULL
    seuratreactive$MAGMA_Final_Results <- NULL
    seuratreactive$meta <- NULL
    seuratreactive$MAGMA_selectcluster <- NULL
    seuratreactive$MAGMA_selectcluster2 <- NULL
    seuratreactive$MAGMA_samples2 <- NULL
    seuratreactive$textlabelMAGMA <- NULL
    seuratreactive$MAGMA_N <- NULL
    seuratreactive$GRCh <- NULL
    seuratreactive$MAGMA_Pop <- NULL
    seuratreactive$MAGMA_dirs <- NULL
    seuratreactive$genesOutPath2 <- NULL
    seuratreactive$n_MAGMA <- NULL
    seuratreactive$n_MAGMA2 <- NULL
    
    #CellChat
    seuratreactive$cellchat <- NULL
    seuratreactive$cellchatM <- NULL
    seuratreactive$object.list <- NULL
    seuratreactive$ChatChoose <- NULL
    seuratreactive$ChatSample <- NULL
    seuratreactive$Chatdb <- NULL
    seuratreactive$ChatMultiple1Sample <- NULL
    seuratreactive$ChatMultiple2Sample <- NULL
    seuratreactive$Chat_selectcluster <- NULL
    seuratreactive$Chat_number <- NULL
    seuratreactive$Chatpval <- NULL
    
    #DRUG
    seuratreactive$markersDRUG <- NULL
    seuratreactive$DRUGdf <- NULL
    
    #DRUG2
    seuratreactive$markersDRUG2 <- NULL
    seuratreactive$DRUGdf2 <- NULL
    
    #GE
    seuratreactive$titleGE <- NULL
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    seuratreactive$objGEyes <- NULL
    seuratreactive$Gene_vector_msig <- NULL
    
    #TA
    seuratreactive$TAselectedcells <- NULL
    seuratreactive$TA_pickercluster <- NULL
    seuratreactive$seurat_monocle <- NULL
    seuratreactive$Monocle3_genes <- NULL
    
    #MM
    seuratreactive$MM <- NULL
    seuratreactive$plot.dataM <- NULL
    seuratreactive$plot.dataMM <- NULL
    
    #DE
    seuratreactive$onlyposDE <- NULL
    seuratreactive$Subset_GENES <- NULL
    seuratreactive$DE_clusters_samples <- NULL
    seuratreactive$DECluster <- NULL
    seuratreactive$DEInput1_samples <- NULL
    seuratreactive$DEInput2_samples <- NULL
    seuratreactive$SelectComparison <- NULL
    seuratreactive$DE_clusters <- NULL
    seuratreactive$GOlabelsDE <- NULL
    seuratreactive$DEInput1 <- NULL
    seuratreactive$DEInput2 <- NULL
    seuratreactive$logDE <- NULL
    seuratreactive$minDE <- NULL
    seuratreactive$padjustDE1 <- NULL
    seuratreactive$testuseDE <- NULL
    seuratreactive$enrichInputDE <- NULL
    seuratreactive$pvalueDE <- NULL
    seuratreactive$pvalDE <- NULL
    seuratreactive$padjustDE <- NULL
    seuratreactive$volcanovalue1 <- NULL
    seuratreactive$volcanovalue2 <- NULL
    seuratreactive$pvaluevolcano <- NULL
    seuratreactive$markersDE <- NULL
    seuratreactive$YuDE <- NULL
    seuratreactive$markersDEvolcano <- NULL
    seuratreactive$PATHWAY <- NULL
    
    #SDE
    seuratreactive$Subset_GENESSDE <- NULL
    seuratreactive$selectedkey1 <- NULL
    seuratreactive$selectedkey2 <- NULL
    seuratreactive$onlyposSDE <- NULL
    seuratreactive$GOlabelsSDE <- NULL
    seuratreactive$logSDE <- NULL
    seuratreactive$minSDE <- NULL
    seuratreactive$padjustSDE1 <- NULL
    seuratreactive$testuseSDE <- NULL
    seuratreactive$enrichInputSDE <- NULL
    seuratreactive$pvalueSDE <- NULL
    seuratreactive$pvalSDE <- NULL
    seuratreactive$padjustSDE <- NULL
    seuratreactive$volcanoSDEvalue1 <- NULL
    seuratreactive$volcanoSDEvalue2 <- NULL
    seuratreactive$pvaluevolcanoSDE <- NULL
    seuratreactive$PATHWAYSDE <- NULL
    
    seuratreactive$markersSDE <- NULL
    seuratreactive$YuSDE <- NULL
    seuratreactive$markersSDEvolcano <- NULL
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    shinyjs::hide("DBox1")
    shinyjs::hide("DBox2")
    
    if(input$iscollapseboxCombine == TRUE) {
      js$collapse("CombineBoxIntro")
    }
    shinyjs::hide("CombineBox1")
    shinyjs::hide("CombineBox2")
    shinyjs::hide("CombineBox3")
    shinyjs::hide("CombineBox4")
    shinyjs::hide("CombineBox5")
    
    
    shinyjs::hide("NextButtonCombine")
    
    shinyjs::hide("NextButtonD")
    
    shinyjs::hide("NextButtonQC")
    
    shinyjs::enable("startbuttonD")
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T,
                                                              badgeColor = "green", badgeLabel = "optional"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T)
  
  observeEvent(input$doublesimGO | input$startbuttonD | input$doubleremove, {
    
    if(input$iscollapseboxCombine == TRUE) {
      js$collapse("CombineBoxIntro")
    }
    shinyjs::hide("CombineBox1")
    shinyjs::hide("CombineBox2")
    shinyjs::hide("CombineBox3")
    shinyjs::hide("CombineBox4")
    shinyjs::hide("CombineBox5")
    
    
    shinyjs::hide("NextButtonCombine")
    
    shinyjs::show("NextButtonD")
    
    shinyjs::hide("NextButtonQC")
    updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
    
    
    #Load2
    seuratreactive$obj_original2 <- NULL
    seuratreactive$matrixtable2 <- NULL
    seuratreactive$mincellsCombine <- NULL
    seuratreactive$minfeaturesCombine <- NULL
    
    #QC2
    seuratreactive$type2 <- NULL
    seuratreactive$objQC2 <- NULL
    seuratreactive$nFeature12 <- NULL
    seuratreactive$nFeature22 <- NULL
    seuratreactive$nCount12 <- NULL
    seuratreactive$nCount22 <- NULL
    seuratreactive$percent.mt12 <- NULL
    seuratreactive$percent.mt22 <- NULL
    seuratreactive$percent.ribo12 <- NULL
    seuratreactive$percent.ribo22 <- NULL
    
    #D2
    seuratreactive$Remove2 <- NULL
    seuratreactive$plot.data.doublets2 <- NULL
    seuratreactive$Removed2 <- NULL
    seuratreactive$doublesim2 <- NULL
    
    #DR
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR <- NULL
    seuratreactive$npcDR <- NULL
    seuratreactive$variablefeatures <- NULL
    seuratreactive$ALRAL <- NULL
    
    #DR2
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR2 <- NULL
    seuratreactive$npcDR2 <- NULL
    seuratreactive$Combined.list <- NULL
    seuratreactive$VFintegrated <- NULL
    seuratreactive$integrationreduction <- NULL
    seuratreactive$kanchor <- NULL
    
    #C
    seuratreactive$integera <- NULL
    seuratreactive$integerb <- NULL
    seuratreactive$integerc <- NULL
    seuratreactive$integerd <- NULL
    seuratreactive$integere <- NULL
    seuratreactive$integerf <- NULL
    seuratreactive$integerg <- NULL
    seuratreactive$select1 <- NULL
    seuratreactive$tsne1 <- NULL
    seuratreactive$tsne2 <- NULL
    seuratreactive$tsne3 <- NULL
    seuratreactive$plot.data <- NULL
    seuratreactive$plot.data.original <- NULL
    
    seuratreactive$Identifiers <- NULL
    
    #CC
    seuratreactive$reductionCC1 <- NULL
    seuratreactive$reductionCC2 <- NULL
    seuratreactive$reductionCC3 <- NULL
    seuratreactive$GeneR <- NULL
    
    
    #L
    seuratreactive$reference_annotation <- NULL
    seuratreactive$select2 <- NULL
    seuratreactive$selecttissue <- NULL
    seuratreactive$new.cell.ids <- NULL
    seuratreactive$markers <- NULL
    
    #MEGENA
    seuratreactive$MEGENA_samples <- NULL
    seuratreactive$variablegenesMEGENA <- NULL
    seuratreactive$MEGENA_selectmethod <- NULL
    seuratreactive$MEGENApval <- NULL
    seuratreactive$MEGENAmin.size <- NULL
    seuratreactive$MEGENA_selectcluster <- NULL
    seuratreactive$n_MEGENA <- NULL
    seuratreactive$log2FCMEGENA <- NULL
    seuratreactive$padjustMEGENA <- NULL
    seuratreactive$testuseMEGENA <- NULL
    seuratreactive$minMEGENA <- NULL
    seuratreactive$pvalMEGENA <- NULL
    seuratreactive$summary.output <- NULL
    seuratreactive$markersMEGENA <- NULL
    seuratreactive$moduleGO_df <- NULL
    seuratreactive$VF <- NULL
    seuratreactive$g <- NULL
    
    #SCENIC
    seuratreactive$variablefeaturesSCENIC <- NULL
    seuratreactive$variablegenesSCENIC <- NULL
    seuratreactive$regulonAUC <- NULL
    seuratreactive$regulons <- NULL
    seuratreactive$cellInfo <- NULL
    seuratreactive$n_SCENIC <- NULL
    seuratreactive$SCENICdb <- NULL
    seuratreactive$SCENIC_samples <- NULL
    seuratreactive$regulonnames <- NULL
    seuratreactive$titleSCENIC <- NULL
    seuratreactive$SCENICInput <- NULL
    seuratreactive$SCENIC_selectcluster <- NULL
    seuratreactive$SCENIC_selectmethod <- NULL
    seuratreactive$regulonActivity_byCellType_Scaled <- NULL
    seuratreactive$rss <- NULL
    
    #MAGMA
    seuratreactive$MAGMA_samples <- NULL
    seuratreactive$MAGMA_GWAS <- NULL
    seuratreactive$MAGMA_Final_Results <- NULL
    seuratreactive$meta <- NULL
    seuratreactive$MAGMA_selectcluster <- NULL
    seuratreactive$MAGMA_selectcluster2 <- NULL
    seuratreactive$MAGMA_samples2 <- NULL
    seuratreactive$textlabelMAGMA <- NULL
    seuratreactive$MAGMA_N <- NULL
    seuratreactive$GRCh <- NULL
    seuratreactive$MAGMA_Pop <- NULL
    seuratreactive$MAGMA_dirs <- NULL
    seuratreactive$genesOutPath2 <- NULL
    seuratreactive$n_MAGMA <- NULL
    seuratreactive$n_MAGMA2 <- NULL
    
    #CellChat
    seuratreactive$cellchat <- NULL
    seuratreactive$cellchatM <- NULL
    seuratreactive$object.list <- NULL
    seuratreactive$ChatChoose <- NULL
    seuratreactive$ChatSample <- NULL
    seuratreactive$Chatdb <- NULL
    seuratreactive$ChatMultiple1Sample <- NULL
    seuratreactive$ChatMultiple2Sample <- NULL
    seuratreactive$Chat_selectcluster <- NULL
    seuratreactive$Chat_number <- NULL
    seuratreactive$Chatpval <- NULL
    
    #DRUG
    seuratreactive$markersDRUG <- NULL
    seuratreactive$DRUGdf <- NULL
    
    #DRUG2
    seuratreactive$markersDRUG2 <- NULL
    seuratreactive$DRUGdf2 <- NULL
    
    #GE
    seuratreactive$titleGE <- NULL
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    seuratreactive$objGEyes <- NULL
    seuratreactive$Gene_vector_msig <- NULL
    
    #TA
    seuratreactive$TAselectedcells <- NULL
    seuratreactive$TA_pickercluster <- NULL
    seuratreactive$seurat_monocle <- NULL
    seuratreactive$Monocle3_genes <- NULL
    
    #MM
    seuratreactive$MM <- NULL
    seuratreactive$plot.dataM <- NULL
    seuratreactive$plot.dataMM <- NULL
    
    #DE
    seuratreactive$onlyposDE <- NULL
    seuratreactive$Subset_GENES <- NULL
    seuratreactive$DE_clusters_samples <- NULL
    seuratreactive$DECluster <- NULL
    seuratreactive$DEInput1_samples <- NULL
    seuratreactive$DEInput2_samples <- NULL
    seuratreactive$SelectComparison <- NULL
    seuratreactive$DE_clusters <- NULL
    seuratreactive$GOlabelsDE <- NULL
    seuratreactive$DEInput1 <- NULL
    seuratreactive$DEInput2 <- NULL
    seuratreactive$logDE <- NULL
    seuratreactive$minDE <- NULL
    seuratreactive$padjustDE1 <- NULL
    seuratreactive$testuseDE <- NULL
    seuratreactive$enrichInputDE <- NULL
    seuratreactive$pvalueDE <- NULL
    seuratreactive$pvalDE <- NULL
    seuratreactive$padjustDE <- NULL
    seuratreactive$volcanovalue1 <- NULL
    seuratreactive$volcanovalue2 <- NULL
    seuratreactive$pvaluevolcano <- NULL
    seuratreactive$markersDE <- NULL
    seuratreactive$YuDE <- NULL
    seuratreactive$markersDEvolcano <- NULL
    seuratreactive$PATHWAY <- NULL
    
    #SDE
    seuratreactive$Subset_GENESSDE <- NULL
    seuratreactive$selectedkey1 <- NULL
    seuratreactive$selectedkey2 <- NULL
    seuratreactive$onlyposSDE <- NULL
    seuratreactive$GOlabelsSDE <- NULL
    seuratreactive$logSDE <- NULL
    seuratreactive$minSDE <- NULL
    seuratreactive$padjustSDE1 <- NULL
    seuratreactive$testuseSDE <- NULL
    seuratreactive$enrichInputSDE <- NULL
    seuratreactive$pvalueSDE <- NULL
    seuratreactive$pvalSDE <- NULL
    seuratreactive$padjustSDE <- NULL
    seuratreactive$volcanoSDEvalue1 <- NULL
    seuratreactive$volcanoSDEvalue2 <- NULL
    seuratreactive$pvaluevolcanoSDE <- NULL
    seuratreactive$PATHWAYSDE <- NULL
    
    seuratreactive$markersSDE <- NULL
    seuratreactive$YuSDE <- NULL
    seuratreactive$markersSDEvolcano <- NULL
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = T),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F,
                                                              badgeColor = "green", badgeLabel = "optional"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  },ignoreInit = T)
  
  observeEvent(input$NextButtonD, {
    
    if(input$iscollapseboxCombine == TRUE) {
      js$collapse("CombineBoxIntro")
    }
    shinyjs::hide("CombineBox1")
    shinyjs::hide("CombineBox2")
    shinyjs::hide("CombineBox3")
    shinyjs::hide("CombineBox4")
    shinyjs::hide("CombineBox5")
    
    
    shinyjs::hide("NextButtonCombine")
    
    shinyjs::hide("NextButtonD")
    shinyjs::disable("startbuttonD")
    
    shinyjs::hide("NextButtonQC")
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = T,
                                                              badgeColor = "green", badgeLabel = "optional"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T) 
  
  ####TAB 2.5 - QC (2nd dataset)#####
  
  observeEvent(input$tabs, {
    if (input$tabs == "QC2" & !is.null(seuratreactive$VFintegrated) | input$tabs == "QC2" & !is.null(seuratreactive$Removed2)) {
      shinyalert("Warning!", "If you adjust quality control parameters, all the following tabs will disappear and you will have to 
                 re-do the analysis at each following step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (is.na(input$nFeature12) |
        is.na(input$nFeature22) |
        is.na(input$nCount12) |
        is.na(input$nCount22) |
        is.na(input$percent.mt12) |
        is.na(input$percent.mt22) |
        is.na(input$percent.ribo12) |
        is.na(input$percent.ribo22))
    {
      shinyjs::disable("SQC2")
    }
    else {
      shinyjs::enable("SQC2")
    }
  })
  
  observeEvent(input$startbuttonQC2 |input$rewindQC2, {
    tryCatch({
      seuratreactive$obj_original2[["percent.mt"]] <- PercentageFeatureSet(seuratreactive$obj_original2, pattern = "^MT-|^mt-|^Mt-")
      seuratreactive$obj_original2[["percent.ribo"]] <- PercentageFeatureSet(seuratreactive$obj_original2, pattern = "^RP[SL][[:digit:]]|^RPLP[[:digit:]]|^RPSA|
                                                                              ^Rp[sl][[:digit:]]|^Rplp[[:digit:]]|^Rpsa|
                                                                              ^rp[sl][[:digit:]]|^rplp[[:digit:]]|^rpsa")
      seuratreactive$objQC2 <- seuratreactive$obj_original2
      
      if(sum(seuratreactive$obj_original2[["percent.mt"]]) == 0) {
        shinyalert("WARNING!", "No Mitochondrial genes detected in dataset (no genes that begin with MT detected), please check your gene names.", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      if(sum(seuratreactive$obj_original2[["percent.ribo"]]) == 0){
        shinyalert("WARNING!", "No Ribosomal genes detected in dataset (no genes that begin with RPS, RPL, RPLP, RPSA detected), please check your gene names.", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      seuratreactive$nFeature12 <- "0"
      seuratreactive$nFeature22 <- "100000"
      seuratreactive$nCount12 <- "0"
      seuratreactive$nCount22 <- "1000000"
      seuratreactive$percent.mt12 <- "0"
      seuratreactive$percent.mt22 <- "100"
      seuratreactive$percent.ribo12 <- "0"
      seuratreactive$percent.ribo22 <- "100"
      
      if(input$iscollapseboxQC2 == FALSE) {
        js$collapse("QCBoxIntro2")
      }
      shinyjs::show("QCBox12")
      shinyjs::show("QCBox22")
      shinyjs::show("QCBox32")
      shinyjs::show("QCBox42")
      
      
      shinyjs::hide("NextButtonD2")
      shinyjs::enable("startbuttonD2")
      
      shinyjs::disable("startbuttonQC2")
      shinyjs::show("NextButtonQC2")
      
      shinyjs::hide("NextButtonCombine")
      
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if you have the correct seurat object uploaded.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  }, ignoreInit = T)
  
  observeEvent(input$SQC2, {
    tryCatch({
      seuratreactive$objQC2 <- subset(seuratreactive$obj_original2, subset = nFeature_RNA >= input$nFeature12 & nFeature_RNA <= input$nFeature22
                                      & nCount_RNA >= input$nCount12 & nCount_RNA <= input$nCount22
                                      & percent.mt >= input$percent.mt12 & percent.mt <= input$percent.mt22
                                      & percent.ribo >= input$percent.ribo12 & percent.ribo <= input$percent.ribo22
      )
      
      seuratreactive$nFeature12 <- input$nFeature12
      seuratreactive$nFeature22 <- input$nFeature22
      seuratreactive$nCount12 <- input$nCount12
      seuratreactive$nCount22 <- input$nCount22
      seuratreactive$percent.mt12 <- input$percent.mt12
      seuratreactive$percent.mt22 <- input$percent.mt22
      seuratreactive$percent.ribo12 <- input$percent.ribo12
      seuratreactive$percent.ribo22 <- input$percent.ribo22
      
      
      shinyjs::hide("NextButtonD2")
      
      shinyjs::disable("startbuttonQC2")
      shinyjs::show("NextButtonQC2")
      
      shinyjs::hide("NextButtonCombine")
      
    },
    error = function(e) {
      shinyalert("ERROR!", "Selection parameters out of bounds, please re-select parameters.", type = "error", confirmButtonCol = "#337ab7")
    })  
  })
  
  output$plotCombineUI <- renderUI({
    plotOutput("plotCombine", height = (400 + input$QC2plotheight*40))
  })
  
  output$plotCombine <- renderPlot({
    validate(
      need(!is.null(seuratreactive$objQC2), "You must complete the quality control step"))
    VlnPlot(seuratreactive$objQC2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"), group.by = "orig.ident", pt.size = input$sizeQC2, ncol = 2)
  })
  
  output$plot2Combine <- renderPlot({
    validate(
      need(!is.null(seuratreactive$objQC2), "You must complete the quality control step"))
    FeatureScatter(seuratreactive$objQC2, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident", pt.size = 2) + NoLegend() })
  
  output$plot3Combine <- renderPlot({
    validate(
      need(!is.null(seuratreactive$objQC2), "You must complete the quality control step"))
    FeatureScatter(seuratreactive$objQC2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident", pt.size = 2) + NoLegend() })
  
  observeEvent(input$SQC2 | input$rewindQC2 | input$startbuttonQC2, {
    
    if (!is.null(seuratreactive$objQC2)) {
      if(input$iscollapseboxD2 == TRUE) {
        js$collapse("DBoxIntro2")
      }
      shinyjs::hide("DBox12")
      shinyjs::hide("DBox22")
      
      
      shinyjs::hide("NextButtonD2")
      shinyjs::enable("startbuttonD2")
      
      shinyjs::disable("startbuttonQC2")
      shinyjs::show("NextButtonQC2")
      
      shinyjs::hide("NextButtonCombine")
      
      updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      
      #D2
      seuratreactive$Remove2 <- NULL
      seuratreactive$plot.data.doublets2 <- NULL
      seuratreactive$Removed2 <- NULL
      seuratreactive$doublesim2 <- NULL
      
      #DR
      seuratreactive$obj <- NULL
      seuratreactive$SCTransformDR <- NULL
      seuratreactive$npcDR <- NULL
      seuratreactive$variablefeatures <- NULL
      seuratreactive$ALRAL <- NULL
      
      #DR2
      seuratreactive$obj <- NULL
      seuratreactive$SCTransformDR2 <- NULL
      seuratreactive$npcDR2 <- NULL
      seuratreactive$Combined.list <- NULL
      seuratreactive$VFintegrated <- NULL
      seuratreactive$integrationreduction <- NULL
      seuratreactive$kanchor <- NULL
      
      #C
      seuratreactive$integera <- NULL
      seuratreactive$integerb <- NULL
      seuratreactive$integerc <- NULL
      seuratreactive$integerd <- NULL
      seuratreactive$integere <- NULL
      seuratreactive$integerf <- NULL
      seuratreactive$integerg <- NULL
      seuratreactive$select1 <- NULL
      seuratreactive$tsne1 <- NULL
      seuratreactive$tsne2 <- NULL
      seuratreactive$tsne3 <- NULL
      seuratreactive$plot.data <- NULL
      seuratreactive$plot.data.original <- NULL
      
      seuratreactive$Identifiers <- NULL
      
      #CC
      seuratreactive$reductionCC1 <- NULL
      seuratreactive$reductionCC2 <- NULL
      seuratreactive$reductionCC3 <- NULL
      seuratreactive$GeneR <- NULL
      
      
      #L
      seuratreactive$reference_annotation <- NULL
      seuratreactive$select2 <- NULL
      seuratreactive$selecttissue <- NULL
      seuratreactive$new.cell.ids <- NULL
      seuratreactive$markers <- NULL
      
      #MEGENA
      seuratreactive$MEGENA_samples <- NULL
      seuratreactive$variablegenesMEGENA <- NULL
      seuratreactive$MEGENA_selectmethod <- NULL
      seuratreactive$MEGENApval <- NULL
      seuratreactive$MEGENAmin.size <- NULL
      seuratreactive$MEGENA_selectcluster <- NULL
      seuratreactive$n_MEGENA <- NULL
      seuratreactive$log2FCMEGENA <- NULL
      seuratreactive$padjustMEGENA <- NULL
      seuratreactive$testuseMEGENA <- NULL
      seuratreactive$minMEGENA <- NULL
      seuratreactive$pvalMEGENA <- NULL
      seuratreactive$summary.output <- NULL
      seuratreactive$markersMEGENA <- NULL
      seuratreactive$moduleGO_df <- NULL
      seuratreactive$VF <- NULL
      seuratreactive$g <- NULL
      
      #SCENIC
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$variablegenesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$n_SCENIC <- NULL
      seuratreactive$SCENICdb <- NULL
      seuratreactive$SCENIC_samples <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$SCENIC_selectcluster <- NULL
      seuratreactive$SCENIC_selectmethod <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
      
      #MAGMA
      seuratreactive$MAGMA_samples <- NULL
      seuratreactive$MAGMA_GWAS <- NULL
      seuratreactive$MAGMA_Final_Results <- NULL
      seuratreactive$meta <- NULL
      seuratreactive$MAGMA_selectcluster <- NULL
      seuratreactive$MAGMA_selectcluster2 <- NULL
      seuratreactive$MAGMA_samples2 <- NULL
      seuratreactive$textlabelMAGMA <- NULL
      seuratreactive$MAGMA_N <- NULL
      seuratreactive$GRCh <- NULL
      seuratreactive$MAGMA_Pop <- NULL
      seuratreactive$MAGMA_dirs <- NULL
      seuratreactive$genesOutPath2 <- NULL
      seuratreactive$n_MAGMA <- NULL
      seuratreactive$n_MAGMA2 <- NULL
      
      #CellChat
      seuratreactive$cellchat <- NULL
      seuratreactive$cellchatM <- NULL
      seuratreactive$object.list <- NULL
      seuratreactive$ChatChoose <- NULL
      seuratreactive$ChatSample <- NULL
      seuratreactive$Chatdb <- NULL
      seuratreactive$ChatMultiple1Sample <- NULL
      seuratreactive$ChatMultiple2Sample <- NULL
      seuratreactive$Chat_selectcluster <- NULL
      seuratreactive$Chat_number <- NULL
      seuratreactive$Chatpval <- NULL
      
      #DRUG
      seuratreactive$markersDRUG <- NULL
      seuratreactive$DRUGdf <- NULL
      
      #DRUG2
      seuratreactive$markersDRUG2 <- NULL
      seuratreactive$DRUGdf2 <- NULL
      
      #GE
      seuratreactive$titleGE <- NULL
      seuratreactive$GEInput <- NULL
      seuratreactive$GSEInput <- NULL
      seuratreactive$p <- NULL
      seuratreactive$q <- NULL
      seuratreactive$objGEyes <- NULL
      seuratreactive$Gene_vector_msig <- NULL
      
      #TA
      seuratreactive$TAselectedcells <- NULL
      seuratreactive$TA_pickercluster <- NULL
      seuratreactive$seurat_monocle <- NULL
      seuratreactive$Monocle3_genes <- NULL
      
      #MM
      seuratreactive$MM <- NULL
      seuratreactive$plot.dataM <- NULL
      seuratreactive$plot.dataMM <- NULL
      
      #DE
      seuratreactive$onlyposDE <- NULL
      seuratreactive$Subset_GENES <- NULL
      seuratreactive$DE_clusters_samples <- NULL
      seuratreactive$DECluster <- NULL
      seuratreactive$DEInput1_samples <- NULL
      seuratreactive$DEInput2_samples <- NULL
      seuratreactive$SelectComparison <- NULL
      seuratreactive$DE_clusters <- NULL
      seuratreactive$GOlabelsDE <- NULL
      seuratreactive$DEInput1 <- NULL
      seuratreactive$DEInput2 <- NULL
      seuratreactive$logDE <- NULL
      seuratreactive$minDE <- NULL
      seuratreactive$padjustDE1 <- NULL
      seuratreactive$testuseDE <- NULL
      seuratreactive$enrichInputDE <- NULL
      seuratreactive$pvalueDE <- NULL
      seuratreactive$pvalDE <- NULL
      seuratreactive$padjustDE <- NULL
      seuratreactive$volcanovalue1 <- NULL
      seuratreactive$volcanovalue2 <- NULL
      seuratreactive$pvaluevolcano <- NULL
      seuratreactive$markersDE <- NULL
      seuratreactive$YuDE <- NULL
      seuratreactive$markersDEvolcano <- NULL
      seuratreactive$PATHWAY <- NULL
      
      #SDE
      seuratreactive$Subset_GENESSDE <- NULL
      seuratreactive$selectedkey1 <- NULL
      seuratreactive$selectedkey2 <- NULL
      seuratreactive$onlyposSDE <- NULL
      seuratreactive$GOlabelsSDE <- NULL
      seuratreactive$logSDE <- NULL
      seuratreactive$minSDE <- NULL
      seuratreactive$padjustSDE1 <- NULL
      seuratreactive$testuseSDE <- NULL
      seuratreactive$enrichInputSDE <- NULL
      seuratreactive$pvalueSDE <- NULL
      seuratreactive$pvalSDE <- NULL
      seuratreactive$padjustSDE <- NULL
      seuratreactive$volcanoSDEvalue1 <- NULL
      seuratreactive$volcanoSDEvalue2 <- NULL
      seuratreactive$pvaluevolcanoSDE <- NULL
      seuratreactive$PATHWAYSDE <- NULL
      
      seuratreactive$markersSDE <- NULL
      seuratreactive$YuSDE <- NULL
      seuratreactive$markersSDEvolcano <- NULL
      
      #PA
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      #SPA
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = T),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
  }, ignoreInit = T)
  
  observeEvent(input$NextButtonQC2, {
    if(input$iscollapseboxD2 == TRUE) {
      js$collapse("DBoxIntro2")
    }
    shinyjs::hide("DBox12")
    shinyjs::hide("DBox22")
    
    
    shinyjs::hide("NextButtonD2")
    shinyjs::enable("startbuttonD2")
    
    shinyjs::disable("startbuttonQC2")
    shinyjs::hide("NextButtonQC2")
    
    shinyjs::hide("NextButtonCombine")
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                     menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                     menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = T,
                                                              badgeColor = "green", badgeLabel = "new"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T)
  
  ####TAB2.75 - Doublet Removal (2nd dataset) ####
  observeEvent(input$tabs, {
    if (input$tabs == "Doublet2" & !is.null(seuratreactive$VFintegrated)) {
      shinyalert("Warning!", "If you adjust doublet removal, all the following tabs will disappear and you will have to 
                 re-do the analysis at each step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (is.na(input$doublesim2)) {
      shinyjs::disable("doublesimGO2")
    }
  })
  
  observeEvent(input$startbuttonD2, {
    library(PCAtools)
    library(DoubletFinder)
    
    if (!is.null(seuratreactive$obj_original) & length(unique(seuratreactive$obj_original@meta.data$orig.ident)) > 1) {
      shinyalert("Warning!", "File contains multiple samples, doublet detection on aggregated scRNA-seq data representing
                    multiple distinct samples (i.e., control vs treatment, samples sequenced on different lanes) may be inaccurate due to
                   technical differences. Proceed with caution.", type = "warning", confirmButtonCol = "#337ab7")      
    }
    
    if(input$iscollapseboxD2 == FALSE) {
      js$collapse("DBoxIntro2")
    }
    shinyjs::show("DBox12")
    shinyjs::show("DBox22")
    
    shinyjs::enable("startbuttonDR2")
    shinyjs::hide("NextButtonDR2")
    
    shinyjs::disable("startbuttonD2")
    shinyjs::show("NextButtonD2")
    
    shinyjs::hide("NextButtonQC2")
    
  }, ignoreInit = T)
  
  observe(
    if (is.null(seuratreactive$Remove2)) {
      shinyjs::disable("doubleremove2")
    }
    else {
      shinyjs::enable("doubleremove2")
    }
  )
  
  observeEvent(input$doublesimGO2, {
    tryCatch({
      withProgress(message = 'Detecting doublets. Please wait...',
                   value = 0, {
                     incProgress(1/20)
                     
                     seuratreactive$objQC2 <- subset(seuratreactive$obj_original2, subset = nFeature_RNA >= as.numeric(seuratreactive$nFeature12) & nFeature_RNA <= as.numeric(seuratreactive$nFeature22)
                                                     & nCount_RNA >= as.numeric(seuratreactive$nCount12) & nCount_RNA <= as.numeric(seuratreactive$nCount22)
                                                     & percent.mt >= as.numeric(seuratreactive$percent.mt12) & percent.mt <= as.numeric(seuratreactive$percent.mt22)
                                                     & percent.ribo >= as.numeric(seuratreactive$percent.ribo12) & percent.ribo <= as.numeric(seuratreactive$percent.ribo22))
                     
                     #Identification of variable features
                     seuratreactive$objQC2 <- FindVariableFeatures(seuratreactive$objQC2, selection.method = "vst", nfeatures = 2000)
                     
                     #Normalization
                     seuratreactive$objQC2 <- NormalizeData(seuratreactive$objQC2)
                     
                     #Data scaling
                     seuratreactive$objQC2 <- ScaleData(seuratreactive$objQC2)
                     
                     #Dimensionality reduction
                     seuratreactive$objQC2 <- RunPCA(seuratreactive$objQC2, npcs = 50)
                     incProgress(7/20)
                     
                     elbow <- ElbowPlot(seuratreactive$objQC2, ndims = 50)
                     
                     Elbowpoint <- min(elbow$data[elbow$data$stdev <= findElbowPoint(elbow$data$stdev),]$dims) + 5
                     
                     seuratreactive$objQC2 <- FindNeighbors(seuratreactive$objQC2, dims = 1:Elbowpoint)
                     seuratreactive$objQC2 <- FindClusters(seuratreactive$objQC2)
                     
                     #2D
                     #RunUMAP
                     seuratreactive$objQC2 <- RunUMAP(seuratreactive$objQC2, dims = 1:Elbowpoint, n.components = 2L)
                     incProgress(9/20)
                     #Run tsne
                     seuratreactive$objQC2 <- RunTSNE(seuratreactive$objQC2, dims = 1:Elbowpoint, dim.embed = 2)
                     
                     #3D
                     seuratreactive$objQC2 <- RunUMAP(seuratreactive$objQC2, dims = 1:Elbowpoint, n.components = 3L)
                     incProgress(11/20)
                     #Run tsne
                     seuratreactive$objQC2 <- RunTSNE(seuratreactive$objQC2, dims = 1:Elbowpoint, dim.embed = 3)
                     incProgress(13/20)
                     
                     # pK Identification (no ground-truth)
                     sweep.res.list <- paramSweep_v3(seuratreactive$objQC2, PCs = 1:Elbowpoint, sct = FALSE)
                     sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
                     bcmvn <- find.pK(sweep.stats)
                     
                     incProgress(15/20)
                     
                     # Homotypic Doublet Proportion Estimate
                     homotypic.prop <- modelHomotypic(seuratreactive$objQC2@active.ident)           
                     nExp_poi <- round(input$doublesim2/100*nrow(seuratreactive$objQC2@meta.data))  ## Assuming 4% doublet formation rate - tailor for your dataset
                     nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
                     
                     seuratreactive$doublesim2 <- input$doublesim2
                     
                     seuratreactive$objQC2 <- doubletFinder_v3(seuratreactive$objQC2, PCs = 1:Elbowpoint, pN = 0.25, pK = as.numeric(paste0(bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK)), nExp = nExp_poi, reuse.pANN = F, sct = FALSE)
                     seuratreactive$objQC2 <- doubletFinder_v3(seuratreactive$objQC2, PCs = 1:Elbowpoint, pN = 0.25, pK = as.numeric(paste0(bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK)), nExp = nExp_poi.adj, 
                                                               reuse.pANN = paste0("pANN_", "0.25_", bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK, "_", nExp_poi), sct = FALSE)
                     
                     incProgress(19/20)
                     
                     seuratreactive$Remove2 <- cbind.data.frame(rownames(seuratreactive$objQC2@meta.data), seuratreactive$objQC2@meta.data[[paste0("DF.classifications_", "0.25_", bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK, "_", nExp_poi.adj)]])
                     colnames(seuratreactive$Remove2) <- c("Cell", "Doublet")
                     
                     seuratreactive$Remove2 <- seuratreactive$Remove2[grep('Doublet', seuratreactive$Remove2$Doublet),]
                     seuratreactive$Remove2 <- as.list(seuratreactive$Remove2$Cell)
                     
                     #Plot
                     plot.data2D <- FetchData(object = seuratreactive$objQC2, 
                                              vars = c("UMAP_1", "UMAP_2", "tSNE_1", "tSNE_2", "orig.ident", 
                                                       paste0("DF.classifications_", "0.25_", bcmvn[which(grepl(max(bcmvn$BCmetric), bcmvn$BCmetric)),]$pK, "_", nExp_poi.adj)))
                     
                     colnames(plot.data2D) <- c("UMAP_1_2D", "UMAP_2_2D", "tSNE_1_2D", "tSNE_2_2D", "Sample", "Detected Doublets")
                     
                     
                     plot.data3D <- FetchData(object = seuratreactive$objQC2, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "tSNE_1", "tSNE_2", "tSNE_3"))
                     
                     colnames(plot.data3D) <- c("UMAP_1_3D", "UMAP_2_3D", "UMAP_3_3D", "tSNE_1_3D", "tSNE_2_3D", "tSNE_3_3D")
                     
                     seuratreactive$plot.data.doublets2 <- merge(plot.data2D, plot.data3D, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data.doublets2) <- seuratreactive$plot.data.doublets2$Row.names
                     
                     seuratreactive$plot.data.doublets2 <- seuratreactive$plot.data.doublets2[,-1]
                     
                     shinyjs::enable("startbuttonDR2")
                     shinyjs::hide("NextButtonDR2")
                     
                     shinyjs::disable("startbuttonD2")
                     shinyjs::show("NextButtonD2")
                     
                     shinyjs::hide("NextButtonQC2")
                     incProgress(20/20)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Selection parameters out of bounds, please re-select parameters.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$doubleremove2, {
    tryCatch({
      withProgress(message = 'Performing Dimensionality Reduction. Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     if (is.null(seuratreactive$Remove2)) {
                       shinyalert("ERROR!", "Doublet removal failed. You need to detect your doublets first.", type = "error", confirmButtonCol = "#337ab7")
                     }
                     else {
                       incProgress(5/10)
                       
                       seuratreactive$objQC2 <- subset(seuratreactive$obj_original2, subset = nFeature_RNA >= as.numeric(seuratreactive$nFeature12) & nFeature_RNA <= as.numeric(seuratreactive$nFeature22)
                                                       & nCount_RNA >= as.numeric(seuratreactive$nCount12) & nCount_RNA <= as.numeric(seuratreactive$nCount22)
                                                       & percent.mt >= as.numeric(seuratreactive$percent.mt12) & percent.mt <= as.numeric(seuratreactive$percent.mt22)
                                                       & percent.ribo >= as.numeric(seuratreactive$percent.ribo12) & percent.ribo <= as.numeric(seuratreactive$percent.ribo22))
                       incProgress(7/10)
                       
                       seuratreactive$objQC2 <- seuratreactive$objQC2[,!colnames(seuratreactive$objQC2) %in% seuratreactive$Remove2]
                     }
                     
                     incProgress(10/10)
                     shinyjs::enable("startbuttonDR2")
                     shinyjs::hide("NextButtonDR2")
                     
                     shinyjs::disable("startbuttonD2")
                     shinyjs::show("NextButtonD2")
                     
                     shinyjs::hide("NextButtonQC2")
                     
                     seuratreactive$Removed2 <- "yes"
                     
                     shinyalert("SUCCESS!", "Doublets Removed!", type = "success", confirmButtonCol = "#337ab7")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "You have already removed your doublets!", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$plotdoublesim2UI <- renderUI({
    plotlyOutput("plotdoublesim2", height = (400 + input$D2plotheight*40))
  })
  
  output$plotdoublesim2 <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data.doublets2), "Please select your estimated percentage of doublets and click SIMULATE DOUBLETS to begin the detection process"))
    if(input$radioD2 == "UMAP") {
      plot_ly(data = seuratreactive$plot.data.doublets2, 
              x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
              color = seuratreactive$plot.data.doublets2[, input$plotclustersD2],
              colors = col_vector,
              type = "scatter", 
              mode = "markers", 
              marker = list(size = input$sizeD2),
              opacity = input$opacityD2,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets2),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets2$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD2))
    }
    else if (input$radioD2 == "3D UMAP") {
      plot_ly(data = seuratreactive$plot.data.doublets2, 
              x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D, 
              color = seuratreactive$plot.data.doublets2[, input$plotclustersD2],
              colors = col_vector,
              type = "scatter3d", 
              mode = "markers", 
              marker = list(size = input$sizeD2),
              opacity = input$opacityD2,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets2),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets2$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD2))
    }
    else if (input$radioD2 == "t-SNE") {
      plot_ly(data = seuratreactive$plot.data.doublets2, 
              x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
              color = seuratreactive$plot.data.doublets2[, input$plotclustersD2],
              colors = col_vector,
              type = "scatter", 
              mode = "markers", 
              marker = list(size = input$sizeD2), 
              opacity = input$opacityD2,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets2),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets2$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD2))
    }
    else if (input$radioD2 == "3D t-SNE") {
      plot_ly(data = seuratreactive$plot.data.doublets2, 
              x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
              color = seuratreactive$plot.data.doublets2[, input$plotclustersD2], 
              colors = col_vector,
              type = "scatter3d", 
              mode = "markers", 
              marker = list(size = input$sizeD2),
              opacity = input$opacityD2,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data.doublets2),
                                 "<br>Sample: ", seuratreactive$plot.data.doublets2$`Sample`)
      ) %>% layout(font = list(size = input$labelsizeD2))
    }
  })
  
  observeEvent(input$doublesimGO2 | input$startbuttonD2 | input$doubleremove2, {
    
    if(input$iscollapseboxDR2 == TRUE) {
      js$collapse("DRBoxIntro2")
    }
    shinyjs::hide("DRBox1DR2")
    shinyjs::hide("DRBox2DR2")
    shinyjs::hide("DRBox3DR2")
    shinyjs::hide("DRBox4DR2")
    shinyjs::hide("downloadMerged")
    
    shinyjs::enable("startbuttonDR2")
    shinyjs::hide("NextButtonDR2")
    
    shinyjs::disable("startbuttonD2")
    shinyjs::show("NextButtonD2")
    
    shinyjs::hide("NextButtonQC2")
    
    updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
    
    #DR
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR <- NULL
    seuratreactive$npcDR <- NULL
    seuratreactive$variablefeatures <- NULL
    seuratreactive$ALRAL <- NULL
    
    #DR2
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR2 <- NULL
    seuratreactive$npcDR2 <- NULL
    seuratreactive$Combined.list <- NULL
    seuratreactive$VFintegrated <- NULL
    seuratreactive$integrationreduction <- NULL
    seuratreactive$kanchor <- NULL
    
    #C
    seuratreactive$integera <- NULL
    seuratreactive$integerb <- NULL
    seuratreactive$integerc <- NULL
    seuratreactive$integerd <- NULL
    seuratreactive$integere <- NULL
    seuratreactive$integerf <- NULL
    seuratreactive$integerg <- NULL
    seuratreactive$select1 <- NULL
    seuratreactive$tsne1 <- NULL
    seuratreactive$tsne2 <- NULL
    seuratreactive$tsne3 <- NULL
    seuratreactive$plot.data <- NULL
    seuratreactive$plot.data.original <- NULL
    
    seuratreactive$Identifiers <- NULL
    
    #CC
    seuratreactive$reductionCC1 <- NULL
    seuratreactive$reductionCC2 <- NULL
    seuratreactive$reductionCC3 <- NULL
    seuratreactive$GeneR <- NULL
    
    
    #L
    seuratreactive$reference_annotation <- NULL
    seuratreactive$select2 <- NULL
    seuratreactive$selecttissue <- NULL
    seuratreactive$new.cell.ids <- NULL
    seuratreactive$markers <- NULL
    
    #MEGENA
    seuratreactive$MEGENA_samples <- NULL
    seuratreactive$variablegenesMEGENA <- NULL
    seuratreactive$MEGENA_selectmethod <- NULL
    seuratreactive$MEGENApval <- NULL
    seuratreactive$MEGENAmin.size <- NULL
    seuratreactive$MEGENA_selectcluster <- NULL
    seuratreactive$n_MEGENA <- NULL
    seuratreactive$log2FCMEGENA <- NULL
    seuratreactive$padjustMEGENA <- NULL
    seuratreactive$testuseMEGENA <- NULL
    seuratreactive$minMEGENA <- NULL
    seuratreactive$pvalMEGENA <- NULL
    seuratreactive$summary.output <- NULL
    seuratreactive$markersMEGENA <- NULL
    seuratreactive$moduleGO_df <- NULL
    seuratreactive$VF <- NULL
    seuratreactive$g <- NULL
    
    #SCENIC
    seuratreactive$variablefeaturesSCENIC <- NULL
    seuratreactive$variablegenesSCENIC <- NULL
    seuratreactive$regulonAUC <- NULL
    seuratreactive$regulons <- NULL
    seuratreactive$cellInfo <- NULL
    seuratreactive$n_SCENIC <- NULL
    seuratreactive$SCENICdb <- NULL
    seuratreactive$SCENIC_samples <- NULL
    seuratreactive$regulonnames <- NULL
    seuratreactive$titleSCENIC <- NULL
    seuratreactive$SCENICInput <- NULL
    seuratreactive$SCENIC_selectcluster <- NULL
    seuratreactive$SCENIC_selectmethod <- NULL
    seuratreactive$regulonActivity_byCellType_Scaled <- NULL
    seuratreactive$rss <- NULL
    
    #MAGMA
    seuratreactive$MAGMA_samples <- NULL
    seuratreactive$MAGMA_GWAS <- NULL
    seuratreactive$MAGMA_Final_Results <- NULL
    seuratreactive$meta <- NULL
    seuratreactive$MAGMA_selectcluster <- NULL
    seuratreactive$MAGMA_selectcluster2 <- NULL
    seuratreactive$MAGMA_samples2 <- NULL
    seuratreactive$textlabelMAGMA <- NULL
    seuratreactive$MAGMA_N <- NULL
    seuratreactive$GRCh <- NULL
    seuratreactive$MAGMA_Pop <- NULL
    seuratreactive$MAGMA_dirs <- NULL
    seuratreactive$genesOutPath2 <- NULL
    seuratreactive$n_MAGMA <- NULL
    seuratreactive$n_MAGMA2 <- NULL
    
    #CellChat
    seuratreactive$cellchat <- NULL
    seuratreactive$cellchatM <- NULL
    seuratreactive$object.list <- NULL
    seuratreactive$ChatChoose <- NULL
    seuratreactive$ChatSample <- NULL
    seuratreactive$Chatdb <- NULL
    seuratreactive$ChatMultiple1Sample <- NULL
    seuratreactive$ChatMultiple2Sample <- NULL
    seuratreactive$Chat_selectcluster <- NULL
    seuratreactive$Chat_number <- NULL
    seuratreactive$Chatpval <- NULL
    
    #DRUG
    seuratreactive$markersDRUG <- NULL
    seuratreactive$DRUGdf <- NULL
    
    #DRUG2
    seuratreactive$markersDRUG2 <- NULL
    seuratreactive$DRUGdf2 <- NULL
    
    #GE
    seuratreactive$titleGE <- NULL
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    seuratreactive$objGEyes <- NULL
    seuratreactive$Gene_vector_msig <- NULL
    
    #TA
    seuratreactive$TAselectedcells <- NULL
    seuratreactive$TA_pickercluster <- NULL
    seuratreactive$seurat_monocle <- NULL
    seuratreactive$Monocle3_genes <- NULL
    
    #MM
    seuratreactive$MM <- NULL
    seuratreactive$plot.dataM <- NULL
    seuratreactive$plot.dataMM <- NULL
    
    #DE
    seuratreactive$onlyposDE <- NULL
    seuratreactive$Subset_GENES <- NULL
    seuratreactive$DE_clusters_samples <- NULL
    seuratreactive$DECluster <- NULL
    seuratreactive$DEInput1_samples <- NULL
    seuratreactive$DEInput2_samples <- NULL
    seuratreactive$SelectComparison <- NULL
    seuratreactive$DE_clusters <- NULL
    seuratreactive$GOlabelsDE <- NULL
    seuratreactive$DEInput1 <- NULL
    seuratreactive$DEInput2 <- NULL
    seuratreactive$logDE <- NULL
    seuratreactive$minDE <- NULL
    seuratreactive$padjustDE1 <- NULL
    seuratreactive$testuseDE <- NULL
    seuratreactive$enrichInputDE <- NULL
    seuratreactive$pvalueDE <- NULL
    seuratreactive$pvalDE <- NULL
    seuratreactive$padjustDE <- NULL
    seuratreactive$volcanovalue1 <- NULL
    seuratreactive$volcanovalue2 <- NULL
    seuratreactive$pvaluevolcano <- NULL
    seuratreactive$markersDE <- NULL
    seuratreactive$YuDE <- NULL
    seuratreactive$markersDEvolcano <- NULL
    seuratreactive$PATHWAY <- NULL
    
    #SDE
    seuratreactive$Subset_GENESSDE <- NULL
    seuratreactive$selectedkey1 <- NULL
    seuratreactive$selectedkey2 <- NULL
    seuratreactive$onlyposSDE <- NULL
    seuratreactive$GOlabelsSDE <- NULL
    seuratreactive$logSDE <- NULL
    seuratreactive$minSDE <- NULL
    seuratreactive$padjustSDE1 <- NULL
    seuratreactive$testuseSDE <- NULL
    seuratreactive$enrichInputSDE <- NULL
    seuratreactive$pvalueSDE <- NULL
    seuratreactive$pvalSDE <- NULL
    seuratreactive$padjustSDE <- NULL
    seuratreactive$volcanoSDEvalue1 <- NULL
    seuratreactive$volcanoSDEvalue2 <- NULL
    seuratreactive$pvaluevolcanoSDE <- NULL
    seuratreactive$PATHWAYSDE <- NULL
    
    seuratreactive$markersSDE <- NULL
    seuratreactive$YuSDE <- NULL
    seuratreactive$markersSDEvolcano <- NULL
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                     menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                     menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = T),
                                                     menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F,
                                                              badgeColor = "green", badgeLabel = "new"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T)
  
  observeEvent(input$skipbuttonD2, {
    updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
    
    #D2
    seuratreactive$Remove2 <- NULL
    seuratreactive$plot.data.doublets2 <- NULL
    seuratreactive$Removed2 <- NULL
    seuratreactive$doublesim2 <- NULL
    
    #DR
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR <- NULL
    seuratreactive$npcDR <- NULL
    seuratreactive$variablefeatures <- NULL
    seuratreactive$ALRAL <- NULL
    
    #DR2
    seuratreactive$obj <- NULL
    seuratreactive$SCTransformDR2 <- NULL
    seuratreactive$npcDR2 <- NULL
    seuratreactive$Combined.list <- NULL
    seuratreactive$VFintegrated <- NULL
    seuratreactive$integrationreduction <- NULL
    seuratreactive$kanchor <- NULL
    
    #C
    seuratreactive$integera <- NULL
    seuratreactive$integerb <- NULL
    seuratreactive$integerc <- NULL
    seuratreactive$integerd <- NULL
    seuratreactive$integere <- NULL
    seuratreactive$integerf <- NULL
    seuratreactive$integerg <- NULL
    seuratreactive$select1 <- NULL
    seuratreactive$tsne1 <- NULL
    seuratreactive$tsne2 <- NULL
    seuratreactive$tsne3 <- NULL
    seuratreactive$plot.data <- NULL
    seuratreactive$plot.data.original <- NULL
    
    seuratreactive$Identifiers <- NULL
    
    #CC
    seuratreactive$reductionCC1 <- NULL
    seuratreactive$reductionCC2 <- NULL
    seuratreactive$reductionCC3 <- NULL
    seuratreactive$GeneR <- NULL
    
    
    #L
    seuratreactive$reference_annotation <- NULL
    seuratreactive$select2 <- NULL
    seuratreactive$selecttissue <- NULL
    seuratreactive$new.cell.ids <- NULL
    seuratreactive$markers <- NULL
    
    #MEGENA
    seuratreactive$MEGENA_samples <- NULL
    seuratreactive$variablegenesMEGENA <- NULL
    seuratreactive$MEGENA_selectmethod <- NULL
    seuratreactive$MEGENApval <- NULL
    seuratreactive$MEGENAmin.size <- NULL
    seuratreactive$MEGENA_selectcluster <- NULL
    seuratreactive$n_MEGENA <- NULL
    seuratreactive$log2FCMEGENA <- NULL
    seuratreactive$padjustMEGENA <- NULL
    seuratreactive$testuseMEGENA <- NULL
    seuratreactive$minMEGENA <- NULL
    seuratreactive$pvalMEGENA <- NULL
    seuratreactive$summary.output <- NULL
    seuratreactive$markersMEGENA <- NULL
    seuratreactive$moduleGO_df <- NULL
    seuratreactive$VF <- NULL
    seuratreactive$g <- NULL
    
    #SCENIC
    seuratreactive$variablefeaturesSCENIC <- NULL
    seuratreactive$variablegenesSCENIC <- NULL
    seuratreactive$regulonAUC <- NULL
    seuratreactive$regulons <- NULL
    seuratreactive$cellInfo <- NULL
    seuratreactive$n_SCENIC <- NULL
    seuratreactive$SCENICdb <- NULL
    seuratreactive$SCENIC_samples <- NULL
    seuratreactive$regulonnames <- NULL
    seuratreactive$titleSCENIC <- NULL
    seuratreactive$SCENICInput <- NULL
    seuratreactive$SCENIC_selectcluster <- NULL
    seuratreactive$SCENIC_selectmethod <- NULL
    seuratreactive$regulonActivity_byCellType_Scaled <- NULL
    seuratreactive$rss <- NULL
    
    #MAGMA
    seuratreactive$MAGMA_samples <- NULL
    seuratreactive$MAGMA_GWAS <- NULL
    seuratreactive$MAGMA_Final_Results <- NULL
    seuratreactive$meta <- NULL
    seuratreactive$MAGMA_selectcluster <- NULL
    seuratreactive$MAGMA_selectcluster2 <- NULL
    seuratreactive$MAGMA_samples2 <- NULL
    seuratreactive$textlabelMAGMA <- NULL
    seuratreactive$MAGMA_N <- NULL
    seuratreactive$GRCh <- NULL
    seuratreactive$MAGMA_Pop <- NULL
    seuratreactive$MAGMA_dirs <- NULL
    seuratreactive$genesOutPath2 <- NULL
    seuratreactive$n_MAGMA <- NULL
    seuratreactive$n_MAGMA2 <- NULL
    
    #CellChat
    seuratreactive$cellchat <- NULL
    seuratreactive$cellchatM <- NULL
    seuratreactive$object.list <- NULL
    seuratreactive$ChatChoose <- NULL
    seuratreactive$ChatSample <- NULL
    seuratreactive$Chatdb <- NULL
    seuratreactive$ChatMultiple1Sample <- NULL
    seuratreactive$ChatMultiple2Sample <- NULL
    seuratreactive$Chat_selectcluster <- NULL
    seuratreactive$Chat_number <- NULL
    seuratreactive$Chatpval <- NULL
    
    #DRUG
    seuratreactive$markersDRUG <- NULL
    seuratreactive$DRUGdf <- NULL
    
    #DRUG2
    seuratreactive$markersDRUG2 <- NULL
    seuratreactive$DRUGdf2 <- NULL
    
    #GE
    seuratreactive$titleGE <- NULL
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    seuratreactive$objGEyes <- NULL
    seuratreactive$Gene_vector_msig <- NULL
    
    #TA
    seuratreactive$TAselectedcells <- NULL
    seuratreactive$TA_pickercluster <- NULL
    seuratreactive$seurat_monocle <- NULL
    seuratreactive$Monocle3_genes <- NULL
    
    #MM
    seuratreactive$MM <- NULL
    seuratreactive$plot.dataM <- NULL
    seuratreactive$plot.dataMM <- NULL
    
    #DE
    seuratreactive$onlyposDE <- NULL
    seuratreactive$Subset_GENES <- NULL
    seuratreactive$DE_clusters_samples <- NULL
    seuratreactive$DECluster <- NULL
    seuratreactive$DEInput1_samples <- NULL
    seuratreactive$DEInput2_samples <- NULL
    seuratreactive$SelectComparison <- NULL
    seuratreactive$DE_clusters <- NULL
    seuratreactive$GOlabelsDE <- NULL
    seuratreactive$DEInput1 <- NULL
    seuratreactive$DEInput2 <- NULL
    seuratreactive$logDE <- NULL
    seuratreactive$minDE <- NULL
    seuratreactive$padjustDE1 <- NULL
    seuratreactive$testuseDE <- NULL
    seuratreactive$enrichInputDE <- NULL
    seuratreactive$pvalueDE <- NULL
    seuratreactive$pvalDE <- NULL
    seuratreactive$padjustDE <- NULL
    seuratreactive$volcanovalue1 <- NULL
    seuratreactive$volcanovalue2 <- NULL
    seuratreactive$pvaluevolcano <- NULL
    seuratreactive$markersDE <- NULL
    seuratreactive$YuDE <- NULL
    seuratreactive$markersDEvolcano <- NULL
    seuratreactive$PATHWAY <- NULL
    
    #SDE
    seuratreactive$Subset_GENESSDE <- NULL
    seuratreactive$selectedkey1 <- NULL
    seuratreactive$selectedkey2 <- NULL
    seuratreactive$onlyposSDE <- NULL
    seuratreactive$GOlabelsSDE <- NULL
    seuratreactive$logSDE <- NULL
    seuratreactive$minSDE <- NULL
    seuratreactive$padjustSDE1 <- NULL
    seuratreactive$testuseSDE <- NULL
    seuratreactive$enrichInputSDE <- NULL
    seuratreactive$pvalueSDE <- NULL
    seuratreactive$pvalSDE <- NULL
    seuratreactive$padjustSDE <- NULL
    seuratreactive$volcanoSDEvalue1 <- NULL
    seuratreactive$volcanoSDEvalue2 <- NULL
    seuratreactive$pvaluevolcanoSDE <- NULL
    seuratreactive$PATHWAYSDE <- NULL
    
    seuratreactive$markersSDE <- NULL
    seuratreactive$YuSDE <- NULL
    seuratreactive$markersSDEvolcano <- NULL
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    shinyjs::hide("DBox12")
    shinyjs::hide("DBox22")
    
    if(input$iscollapseboxDR2 == TRUE) {
      js$collapse("DRBoxIntro2")
    }
    shinyjs::hide("DRBox1DR2")
    shinyjs::hide("DRBox2DR2")
    shinyjs::hide("DRBox3DR2")
    shinyjs::hide("DRBox4DR2")
    shinyjs::hide("downloadMerged")
    
    shinyjs::enable("startbuttonDR2")
    shinyjs::hide("NextButtonDR2")
    
    
    shinyjs::hide("NextButtonD2")
    
    shinyjs::hide("NextButtonQC2")
    
    shinyjs::enable("startbuttonD2")
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                     menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                     menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                     menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T,
                                                              badgeColor = "green", badgeLabel = "new"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T)
  
  observeEvent(input$NextButtonD2, {
    
    if(input$iscollapseboxDR2 == TRUE) {
      js$collapse("DRBoxIntro2")
    }
    shinyjs::hide("DRBox1DR2")
    shinyjs::hide("DRBox2DR2")
    shinyjs::hide("DRBox3DR2")
    shinyjs::hide("DRBox4DR2")
    shinyjs::hide("downloadMerged")
    
    shinyjs::enable("startbuttonDR2")
    shinyjs::hide("NextButtonDR2")
    
    shinyjs::disable("startbuttonD2")
    shinyjs::hide("NextButtonD2")
    
    shinyjs::hide("NextButtonQC2")
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                     menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                     menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                     menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T,
                                                              badgeColor = "green", badgeLabel = "new"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  }, ignoreInit = T)
  
  ##TAB3 - Dimensionality Reduction####
  
  observeEvent(input$tabs, {
    if (input$tabs == "DR" & !is.null(seuratreactive$integera)) {
      shinyalert("Warning!", "If you adjust dimensionality reduction parameters, all the following tabs will disappear and you will have to 
                 re-do the analysis at each following step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observeEvent(input$startDR, {
    showModal(modalDialog(
      title = strong("Dimensionality Reduction Settings"),
      pickerInput("SCTransformDR", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Normalization Method')), 
                  choices = c("SCTransform", "LogNormalize"),
                  selected = "LogNormalize",
                  width = "100%"),
      numericInput("variablefeatures", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of Variable Features')), 
                   value = 2000, step = 100,
                   width = "100%"),
      numericInput("npcDR", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of Dimensions')), 
                   value = 50, step = 1,
                   width = "100%"),
      materialSwitch("ALRAL", label = h4(HTML('<h4 style = "text-align:justify;color:#000000">Impute dropouts (<a href="https://www.nature.com/articles/s41467-021-27729-z" target="_blank">ALRA method</a>) <br> *NOTE this will increase computational time')),
                     value = FALSE, status = "primary", width = "100%"),
      p(style="text-align: center;", actionBttn("startbuttonDR", "APPLY", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startDR, {
    updateVirtualSelect("variablefeatures", selected = seuratreactive$variablefeatures)
    updatePickerInput(session, "SCTransformDR", selected = seuratreactive$SCTransformDR)
    updateNumericInput(session, "npcDR", value = seuratreactive$npcDR)
    
    if (!is.null(input$SCTransformDR)) {
      if (input$SCTransformDR == "SCTransform") {
        updateNumericInput(session, "integera", value = 25)
        updateNumericInput(session, "integere", value = 25)
        updateNumericInput(session, "tsne1", value = 25)
      }
      else if (input$SCTransformDR == "LogNormalize") {
        updateNumericInput(session, "integera", value = 15)
        updateNumericInput(session, "integere", value = 15)
        updateNumericInput(session, "tsne1", value = 15)
      }
    }
    
    if (!is.null(input$variablefeatures)) {
      if (is.na(input$variablefeatures)) {
        shinyjs::disable("startbuttonDR")
      }
      else {
        shinyjs::enable("startbuttonDR")
      }
    }
  })
  
  observe({
    if (!is.null(input$variablefeatures)) {
      if (is.na(input$variablefeatures)) {
        shinyjs::disable("startbuttonDR")
      }
      else {
        shinyjs::enable("startbuttonDR")
      }
    }
  })
  
  observeEvent(input$startbuttonDR, {
    tryCatch({
      withProgress(message = 'Performing Dimensionality Reduction. Please wait...',
                   value = 0, {
                     incProgress(1/20)
                     
                     removeModal()
                     
                     if (input$SCTransformDR == "LogNormalize") {
                       seuratreactive$obj <- FindVariableFeatures(seuratreactive$objQC, selection.method = "vst", nfeatures = input$variablefeatures)
                       
                       incProgress(3/20)
                       
                       #Normalization
                       seuratreactive$obj <- NormalizeData(seuratreactive$obj)
                       incProgress(6/20)
                       #Data scaling
                       seuratreactive$obj <- ScaleData(seuratreactive$obj)
                       
                       if (input$ALRAL == TRUE) {
                         seuratreactive$ALRAL <- "yes"
                         seuratreactive$obj <- RunALRA(seuratreactive$obj)
                         
                         seuratreactive$obj@assays$RNA@data <- seuratreactive$obj@assays$alra@data
                         
                         DefaultAssay(seuratreactive$obj) <- "RNA"
                         
                         seuratreactive$obj[['alra']] <- NULL
                       }
                       else {
                         seuratreactive$ALRAL <- "no"
                       }
                       
                       incProgress(9/20)
                     }
                     else {
                       seuratreactive$obj <- SCTransform(seuratreactive$objQC, variable.features.n = input$variablefeatures, vst.flavor = "v2")
                       
                       if (input$ALRAL == TRUE) {
                         seuratreactive$ALRAL <- "yes"
                         seuratreactive$obj <- RunALRA(seuratreactive$obj)
                         
                         seuratreactive$obj@assays$SCT@data <- seuratreactive$obj@assays$alra@data
                         
                         DefaultAssay(seuratreactive$obj) <- "SCT"
                         
                         seuratreactive$obj[['alra']] <- NULL
                       }
                       else {
                         seuratreactive$ALRAL <- "no"
                       }
                     }
                     #Dimensionality reduction
                     seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = input$npcDR)
                     incProgress(18/20)
                     incProgress(20/20)
                     
                     seuratreactive$variablefeatures <- input$variablefeatures
                     seuratreactive$SCTransformDR <- input$SCTransformDR
                     seuratreactive$npcDR <- input$npcDR
                     
                     if(input$iscollapsebox1 == FALSE) {
                       js$collapse("DRBoxIntro")
                     }
                     shinyjs::show("DRBox1")
                     shinyjs::show("DRBox2")
                     shinyjs::show("DRBox3")
                     shinyjs::show("DRBox4")
                     
                     shinyjs::enable("startbuttonC")
                     shinyjs::hide("NextButtonC")
                     
                     if (!is.null(seuratreactive$obj)) {
                       shinyjs::show("NextButtonDR")
                     }
                     
                     shinyjs::hide("NextButtonCombine")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check the number of variable features that are incorporated. Try reducing the number of dimensions that are computed.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  output$plot4 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$variablefeatures), "You must complete the dimensionality reduction step"))
    LabelPoints(VariableFeaturePlot(seuratreactive$obj, pt.size = 2), points = head(VariableFeatures(seuratreactive$obj), 10), repel = TRUE)
  })
  
  output$plot5 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$variablefeatures), "You must complete the dimensionality reduction step"))
    DimHeatmap(seuratreactive$obj, dims = input$integer, nfeatures = input$PCnfeatures, cells = input$PCncells, balanced = TRUE, ncol = 1)
  })
  
  output$plot6 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$variablefeatures), "You must complete the dimensionality reduction step"))
    ElbowPlot(seuratreactive$obj, ndims = seuratreactive$npcDR)
  })
  
  output$plot7 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$variablefeatures), "You must complete the dimensionality reduction step"))
    VizDimLoadings(seuratreactive$obj, dims = input$integer2, nfeatures = input$Loadingsnfeatures)
  })
  
  observeEvent(input$startbuttonDR, {
    if (input$iscollapsebox2 == TRUE) {
      js$collapse("CBoxIntro")
    }
    shinyjs::hide("CBox1")
    shinyjs::hide("CBox2")
    shinyjs::hide("CBox3")
    
    shinyjs::enable("startbuttonC")
    shinyjs::hide("NextButtonC")
    
    if (!is.null(seuratreactive$obj)) {
      shinyjs::show("NextButtonDR")
    }
    
    shinyjs::hide("NextButtonCombine")
    
    updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
    
    #DR2
    seuratreactive$Combined.list <- NULL
    seuratreactive$VFintegrated <- NULL
    seuratreactive$integrationreduction <- NULL
    seuratreactive$kanchor <- NULL
    seuratreactive$SCTransformDR2 <- NULL
    seuratreactive$npcDR2 <- NULL
    
    #C
    seuratreactive$integera <- NULL
    seuratreactive$integerb <- NULL
    seuratreactive$integerc <- NULL
    seuratreactive$integerd <- NULL
    seuratreactive$integere <- NULL
    seuratreactive$integerf <- NULL
    seuratreactive$integerg <- NULL
    seuratreactive$select1 <- NULL
    seuratreactive$tsne1 <- NULL
    seuratreactive$tsne2 <- NULL
    seuratreactive$tsne3 <- NULL
    seuratreactive$plot.data <- NULL
    seuratreactive$plot.data.original <- NULL
    
    seuratreactive$Identifiers <- NULL
    
    #CC
    seuratreactive$reductionCC1 <- NULL
    seuratreactive$reductionCC2 <- NULL
    seuratreactive$reductionCC3 <- NULL
    seuratreactive$GeneR <- NULL
    
    
    #L
    seuratreactive$reference_annotation <- NULL
    seuratreactive$select2 <- NULL
    seuratreactive$selecttissue <- NULL
    seuratreactive$new.cell.ids <- NULL
    seuratreactive$markers <- NULL
    
    #MEGENA
    seuratreactive$MEGENA_samples <- NULL
    seuratreactive$variablegenesMEGENA <- NULL
    seuratreactive$MEGENA_selectmethod <- NULL
    seuratreactive$MEGENApval <- NULL
    seuratreactive$MEGENAmin.size <- NULL
    seuratreactive$MEGENA_selectcluster <- NULL
    seuratreactive$n_MEGENA <- NULL
    seuratreactive$log2FCMEGENA <- NULL
    seuratreactive$padjustMEGENA <- NULL
    seuratreactive$testuseMEGENA <- NULL
    seuratreactive$minMEGENA <- NULL
    seuratreactive$pvalMEGENA <- NULL
    seuratreactive$summary.output <- NULL
    seuratreactive$markersMEGENA <- NULL
    seuratreactive$moduleGO_df <- NULL
    seuratreactive$VF <- NULL
    seuratreactive$g <- NULL
    
    #SCENIC
    seuratreactive$variablefeaturesSCENIC <- NULL
    seuratreactive$variablegenesSCENIC <- NULL
    seuratreactive$regulonAUC <- NULL
    seuratreactive$regulons <- NULL
    seuratreactive$cellInfo <- NULL
    seuratreactive$n_SCENIC <- NULL
    seuratreactive$SCENICdb <- NULL
    seuratreactive$SCENIC_samples <- NULL
    seuratreactive$regulonnames <- NULL
    seuratreactive$titleSCENIC <- NULL
    seuratreactive$SCENICInput <- NULL
    seuratreactive$SCENIC_selectcluster <- NULL
    seuratreactive$SCENIC_selectmethod <- NULL
    seuratreactive$regulonActivity_byCellType_Scaled <- NULL
    seuratreactive$rss <- NULL
    
    #MAGMA
    seuratreactive$MAGMA_samples <- NULL
    seuratreactive$MAGMA_GWAS <- NULL
    seuratreactive$MAGMA_Final_Results <- NULL
    seuratreactive$meta <- NULL
    seuratreactive$MAGMA_selectcluster <- NULL
    seuratreactive$MAGMA_selectcluster2 <- NULL
    seuratreactive$MAGMA_samples2 <- NULL
    seuratreactive$textlabelMAGMA <- NULL
    seuratreactive$MAGMA_N <- NULL
    seuratreactive$GRCh <- NULL
    seuratreactive$MAGMA_Pop <- NULL
    seuratreactive$MAGMA_dirs <- NULL
    seuratreactive$genesOutPath2 <- NULL
    seuratreactive$n_MAGMA <- NULL
    seuratreactive$n_MAGMA2 <- NULL
    
    #CellChat
    seuratreactive$cellchat <- NULL
    seuratreactive$cellchatM <- NULL
    seuratreactive$object.list <- NULL
    seuratreactive$ChatChoose <- NULL
    seuratreactive$ChatSample <- NULL
    seuratreactive$Chatdb <- NULL
    seuratreactive$ChatMultiple1Sample <- NULL
    seuratreactive$ChatMultiple2Sample <- NULL
    seuratreactive$Chat_selectcluster <- NULL
    seuratreactive$Chat_number <- NULL
    seuratreactive$Chatpval <- NULL
    
    #DRUG
    seuratreactive$markersDRUG <- NULL
    seuratreactive$DRUGdf <- NULL
    
    #DRUG2
    seuratreactive$markersDRUG2 <- NULL
    seuratreactive$DRUGdf2 <- NULL
    
    #GE
    seuratreactive$titleGE <- NULL
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    seuratreactive$objGEyes <- NULL
    seuratreactive$Gene_vector_msig <- NULL
    
    #TA
    seuratreactive$TAselectedcells <- NULL
    seuratreactive$TA_pickercluster <- NULL
    seuratreactive$seurat_monocle <- NULL
    seuratreactive$Monocle3_genes <- NULL
    
    #MM
    seuratreactive$MM <- NULL
    seuratreactive$plot.dataM <- NULL
    seuratreactive$plot.dataMM <- NULL
    
    #DE
    seuratreactive$onlyposDE <- NULL
    seuratreactive$Subset_GENES <- NULL
    seuratreactive$DE_clusters_samples <- NULL
    seuratreactive$DECluster <- NULL
    seuratreactive$DEInput1_samples <- NULL
    seuratreactive$DEInput2_samples <- NULL
    seuratreactive$SelectComparison <- NULL
    seuratreactive$DE_clusters <- NULL
    seuratreactive$GOlabelsDE <- NULL
    seuratreactive$DEInput1 <- NULL
    seuratreactive$DEInput2 <- NULL
    seuratreactive$logDE <- NULL
    seuratreactive$minDE <- NULL
    seuratreactive$padjustDE1 <- NULL
    seuratreactive$testuseDE <- NULL
    seuratreactive$enrichInputDE <- NULL
    seuratreactive$pvalueDE <- NULL
    seuratreactive$pvalDE <- NULL
    seuratreactive$padjustDE <- NULL
    seuratreactive$volcanovalue1 <- NULL
    seuratreactive$volcanovalue2 <- NULL
    seuratreactive$pvaluevolcano <- NULL
    seuratreactive$markersDE <- NULL
    seuratreactive$YuDE <- NULL
    seuratreactive$markersDEvolcano <- NULL
    seuratreactive$PATHWAY <- NULL
    
    #SDE
    seuratreactive$Subset_GENESSDE <- NULL
    seuratreactive$selectedkey1 <- NULL
    seuratreactive$selectedkey2 <- NULL
    seuratreactive$onlyposSDE <- NULL
    seuratreactive$GOlabelsSDE <- NULL
    seuratreactive$logSDE <- NULL
    seuratreactive$minSDE <- NULL
    seuratreactive$padjustSDE1 <- NULL
    seuratreactive$testuseSDE <- NULL
    seuratreactive$enrichInputSDE <- NULL
    seuratreactive$pvalueSDE <- NULL
    seuratreactive$pvalSDE <- NULL
    seuratreactive$padjustSDE <- NULL
    seuratreactive$volcanoSDEvalue1 <- NULL
    seuratreactive$volcanoSDEvalue2 <- NULL
    seuratreactive$pvaluevolcanoSDE <- NULL
    seuratreactive$PATHWAYSDE <- NULL
    
    seuratreactive$markersSDE <- NULL
    seuratreactive$YuSDE <- NULL
    seuratreactive$markersSDEvolcano <- NULL
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),                                                           
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                     menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = T),
                                                     menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F,
                                                              badgeColor = "green", badgeLabel = "new"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  },ignoreInit = T)
  
  observeEvent(input$NextButtonDR, {    
    
    if (input$iscollapsebox2 == TRUE) {
      js$collapse("CBoxIntro")
    }
    shinyjs::hide("CBox1")
    shinyjs::hide("CBox2")
    shinyjs::hide("CBox3")
    
    shinyjs::enable("startbuttonC")
    shinyjs::hide("NextButtonC")
    
    shinyjs::hide("NextButtonDR")
    
    shinyjs::hide("NextButtonCombine")
    
    output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                     menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),                                                           
                                                     menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                     menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                     menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                     menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                     menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T,
                                                              badgeColor = "green", badgeLabel = "new"),
                                                     menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
    ))
  },ignoreInit = T)
  
  ####TAB 3.5 - Dimensionality reduction (2nd dataset) ######
  
  observeEvent(input$tabs, {
    if (input$tabs == "DR2" & !is.null(seuratreactive$integera)) {
      shinyalert("Warning!", "If you adjust dimensionality reduction parameters, all the following tabs will disappear and you will have to 
                 re-do the analysis at each following step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (!is.null(input$tabs)) {
      if (!is.null(seuratreactive$obj)) {
        
        if (input$tabs == "DR2" | input$tabs == "QC" | input$tabs == "QC2" | input$tabs == "C"| input$tabs == "CC" | input$tabs == "DR" | input$tabs == "MEGENA" | input$tabs == "SCENIC") {
          if (!is.null(seuratreactive$VFintegrated)) {
            if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
              DefaultAssay(seuratreactive$obj) <- "integrated"
            }
            else if (!is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
              DefaultAssay(seuratreactive$obj) <- "SCT"
            }
            else {
              DefaultAssay(seuratreactive$obj) <- "RNA"
            }
          }
          else if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform") {
            DefaultAssay(seuratreactive$obj) <- "SCT"
          }
          else {
            DefaultAssay(seuratreactive$obj) <- "RNA"
          }
        }
        else if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                 !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
          DefaultAssay(seuratreactive$obj) <- "SCT"
        }
        else {
          DefaultAssay(seuratreactive$obj) <- "RNA"
        }
      }
    }
  })
  
  observeEvent(input$downloadMerged, {
    showModal(modalDialog(
      title = strong("Download Integrated Gene Count Matrix"),
      h4(HTML('<h4 style = "text-align:justify">  
              Download the combined integrated gene count matrix (tab-delimited file), this file can be re-loaded into ICARUS using the count matrix (.txt) method of data input.       
                          ')),
      
      downloadBttn("downloadMergedFile", "Download Integrated Gene Count Matrix", size = "lg"),
      size = "l",
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$downloadMergedFile <- downloadHandler(
    filename = function(){
      paste0(paste0(levels(as.factor(seuratreactive$obj@meta.data$orig.ident)), collapse = "_"), "_CountMatrix.txt", sep = "")
    }, 
    
    content = function(file) {
      if (any(!grepl("_", colnames(seuratreactive$obj_original)))) {
        dataset1 <- RenameCells(seuratreactive$obj_original, add.cell.id = gsub('_', ".", unique(seuratreactive$obj_original@meta.data$orig.ident)))
      }
      else {
        dataset1 <- seuratreactive$obj_original
      }
      if (any(!grepl("_", colnames(seuratreactive$obj_original2)))) {
        dataset2 <- RenameCells(seuratreactive$obj_original2, add.cell.id = gsub('_', ".", unique(seuratreactive$obj_original2@meta.data$orig.ident)))
      }
      else {
        dataset2 <- seuratreactive$obj_original2
      }
      
      mergeddata <- merge(dataset1, dataset2)
      
      write.table(as.data.frame(mergeddata@assays$RNA@counts), file = file, quote = F, col.names = T, row.names = T, sep = "\t")
    }
  )
  
  observeEvent(input$startDR2, {
    showModal(modalDialog(
      title = strong("Dimensionality Reduction Settings"),
      numericInput("VFintegrated", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of Variable Features')), 
                   value = 2000, step = 100,
                   width = "100%"),
      pickerInput("SCTransformDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Normalization Method')), 
                  choices = c("SCTransform", "LogNormalize"),
                  selected = "LogNormalize",
                  width = "100%"),
      pickerInput("integrationreduction", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Reduction Method')), choices = c("cca", "rpca", "harmony"),
                  selected = "rpca",
                  width = "100%"),
      numericInput("npcDR2", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Number of Dimensions')), 
                   value = 50, step = 1,
                   width = "100%"),
      numericInput("kanchor", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select k anchors (strength of integration)')), 
                   value = 5,
                   min = 1, max = 50, step = 1,
                   width = "100%"),
      p(style="text-align: center;", actionBttn("startbuttonDR2", "INTEGRATE", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startDR2, {
    updatePickerInput(session, "integrationreduction", selected = seuratreactive$integrationreduction)
    updateNumericInput(session, "kanchor", value = seuratreactive$kanchor)
    updateNumericInput(session, "npcDR2", value = seuratreactive$npcDR2)
    updatePickerInput(session, "SCTransformDR2", selected = seuratreactive$SCTransformDR2)
    
    if (!is.null(input$SCTransformDR2)) {
      if (input$SCTransformDR2 == "SCTransform") {
        updateNumericInput(session, "integera", value = 25)
        updateNumericInput(session, "integere", value = 25)
        updateNumericInput(session, "tsne1", value = 25)
      }
      else if (input$SCTransformDR2 == "LogNormalize") {
        updateNumericInput(session, "integera", value = 15)
        updateNumericInput(session, "integere", value = 15)
        updateNumericInput(session, "tsne1", value = 15)
      }
    }
    
    if (!is.null(input$VFintegrated) && !is.null(input$kanchor)) {
      if (is.na(input$VFintegrated) |
          is.na(input$kanchor)) {
        shinyjs::disable("startbuttonDR2")
      }
      else {
        shinyjs::enable("startbuttonDR2")
      }
    }
    
    if (!is.null(input$integrationreduction)) {
      if (input$integrationreduction == "harmony") {
        shinyjs::hide("kanchor")
      }
      else if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
        shinyjs::show("kanchor")
      }
    }
  })
  
  observe({
    if (!is.null(input$SCTransformDR2)) {
      if (input$SCTransformDR2 == "SCTransform") {
        updateNumericInput(session, "integera", value = 25)
        updateNumericInput(session, "integere", value = 25)
        updateNumericInput(session, "tsne1", value = 25)
      }
      else if (input$SCTransformDR2 == "LogNormalize") {
        updateNumericInput(session, "integera", value = 15)
        updateNumericInput(session, "integere", value = 15)
        updateNumericInput(session, "tsne1", value = 15)
      }
    }
    
    if (!is.null(input$VFintegrated) && !is.null(input$kanchor)) {
      if (is.na(input$VFintegrated) |
          is.na(input$kanchor)) {
        shinyjs::disable("startbuttonDR2")
      }
      else {
        shinyjs::enable("startbuttonDR2")
      }
    }
    
    if (!is.null(input$integrationreduction)) {
      if (input$integrationreduction == "harmony") {
        shinyjs::hide("kanchor")
      }
      else if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
        shinyjs::show("kanchor")
      }
    }
  })
  
  observeEvent(input$startbuttonDR2, {
    tryCatch({
      withProgress(message = 'Performing Dimensionality Reduction. Please wait...',
                   value = 0, {
                     incProgress(1/20)
                     
                     removeModal()
                     
                     if (!is.null(seuratreactive$objQC2)) {
                       Combined <- merge(seuratreactive$objQC, seuratreactive$objQC2)
                     }
                     else {
                       Combined <- seuratreactive$objQC
                     }
                     
                     incProgress(2/20)
                     
                     # split the dataset into a list of two seurat objects 
                     seuratreactive$Combined.list <- SplitObject(Combined, split.by = "orig.ident")
                     
                     incProgress(5/20)
                     
                     # normalize and identify variable features for each dataset independently
                     seuratreactive$Combined.list <- lapply(seuratreactive$Combined.list, FUN = function(x) {
                       x <- NormalizeData(x)
                       x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = input$VFintegrated)
                     })
                     incProgress(7/20)
                     
                     if (input$SCTransformDR2 == "LogNormalize") {
                       
                       if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
                         
                         # select features that are repeatedly variable across datasets for integration
                         features <- SelectIntegrationFeatures(object.list = seuratreactive$Combined.list)

                         seuratreactive$Combined.list <- lapply(seuratreactive$Combined.list, FUN = function(x) {
                           x <- ScaleData(x, features = features)
                           x <- RunPCA(x, features = features, npcs = input$npcDR2)
                         })
                         
                         incProgress(10/20)
                         
                         #Perform integration
                         Combined.anchors <- FindIntegrationAnchors(object.list = seuratreactive$Combined.list, anchor.features = features,
                                                                    dims = 1:input$npcDR2,
                                                                    reduction = input$integrationreduction, 
                                                                    k.anchor = input$kanchor)
                         incProgress(12/20)
                         
                         # integrate data and keep full geneset
                         combined.dataset <- IntegrateData(anchorset = Combined.anchors, dims = 1:input$npcDR2)
                         incProgress(14/20)
                         
                         seuratreactive$obj <- combined.dataset
                         
                         #Data scaling
                         seuratreactive$obj <- ScaleData(seuratreactive$obj)
                         incProgress(16/20)
                         
                         #Dimensionality reduction
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = input$npcDR2)
                         
                         seuratreactive$kanchor <- input$kanchor
                       }
                       
                       else if (input$integrationreduction == "harmony") {
                         
                         Combined <- NormalizeData(Combined)
                         
                         Combined <- FindVariableFeatures(Combined, selection.method = "vst", nfeatures = input$VFintegrated)
                         
                         Combined <- ScaleData(Combined)
                         
                         Combined <- RunPCA(Combined, npcs = input$npcDR2)
                         
                         incProgress(12/20)
                         
                         Combined <- RunHarmony(Combined, "orig.ident", max.iter.harmony = 20)
                         
                         incProgress(17/20)
                         
                         seuratreactive$obj <- Combined
                         
                         seuratreactive$kanchor <- "Not Applicable"
                       }
                     }
                     else if (input$SCTransformDR2 == "SCTransform") {
                       
                       if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {

                         seuratreactive$Combined.list <- lapply(seuratreactive$Combined.list, FUN = function(x) {
                           x <- SCTransform(x, variable.features.n = input$VFintegrated, vst.flavor = "v2")
                           x <- RunPCA(x, npcs = input$npcDR2)
                         })
                         
                         incProgress(7/20)
                         
                         # select features that are repeatedly variable across datasets for integration
                         features <- SelectIntegrationFeatures(object.list = seuratreactive$Combined.list, nfeatures = input$VFintegrated)
                         
                         seuratreactive$Combined.list <- PrepSCTIntegration(object.list = seuratreactive$Combined.list, anchor.features = features)
                         
                         incProgress(10/20)
                         
                         #Perform integration
                         Combined.anchors <- FindIntegrationAnchors(object.list = seuratreactive$Combined.list, anchor.features = features,
                                                                    normalization.method = "SCT",
                                                                    dims = 1:input$npcDR2,
                                                                    reduction = input$integrationreduction, 
                                                                    k.anchor = input$kanchor)
                         
                         incProgress(12/20)
                         
                         # integrate data and keep full geneset
                         combined.dataset <- IntegrateData(anchorset = Combined.anchors, normalization.method = "SCT", dims = 1:input$npcDR2)
                         
                         incProgress(14/20)
                         
                         seuratreactive$obj <- combined.dataset
                         
                         incProgress(16/20)
                         
                         #Dimensionality reduction
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = input$npcDR2)
                         
                         seuratreactive$kanchor <- input$kanchor
                       }
                       
                       else if (input$integrationreduction == "harmony") {
                         
                         Combined <- SCTransform(Combined, variable.features.n = input$VFintegrated, vst.flavor = "v2")
                         
                         Combined <- RunPCA(Combined, npcs = input$npcDR2)
                         
                         incProgress(12/20)
                         
                         Combined <- RunHarmony(Combined, "orig.ident", max.iter.harmony = 20, assay.use = "SCT")
                         
                         incProgress(17/20)
                         
                         seuratreactive$obj <- Combined
                         
                         seuratreactive$kanchor <- "Not Applicable"
                       }
                     }
                     
                     incProgress(19/20)
                     
                     seuratreactive$VFintegrated <- input$VFintegrated
                     
                     seuratreactive$integrationreduction <- input$integrationreduction
                     
                     seuratreactive$SCTransformDR2 <- input$SCTransformDR2
                     
                     seuratreactive$npcDR2 <- input$npcDR2
                     
                     incProgress(20/20)
                     
                     js$collapse("DRBoxIntro2")
                     shinyjs::show("DRBox1DR2")
                     shinyjs::show("DRBox2DR2")
                     shinyjs::show("DRBox3DR2")
                     shinyjs::show("DRBox4DR2")
                     shinyjs::show("downloadMerged")
                     
                     shinyjs::enable("startbuttonC")
                     shinyjs::hide("NextButtonC")
                     
                     if (!is.null(seuratreactive$obj)) {
                       shinyjs::show("NextButtonDR2")
                     }
                     
                     shinyjs::hide("NextButtonCombine")
                     shinyjs::hide("NextButtonD2")
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "You may not have enough similarity between the two datasets to perform the integration. Try reduce the number of dimensions computed. Please also check the number of variable features that are incorporated.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  output$DR2VF1 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$VFintegrated), "You must complete the dimensionality reduction step"))
    LabelPoints(VariableFeaturePlot(seuratreactive$Combined.list[[input$integerVFPlot]], pt.size = 2), head(VariableFeatures(seuratreactive$Combined.list[[input$integerVFPlot]]), 10), repel = TRUE)
  })
  
  output$plot5DR2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$VFintegrated), "You must complete the dimensionality reduction step"))
    
    if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
      DimHeatmap(seuratreactive$obj, dims = input$integerDR2, nfeatures = input$PCnfeaturesDR2, cells = input$PCncellsDR2, balanced = TRUE, ncol = 1)
    }
    else if (seuratreactive$integrationreduction == "harmony") {
      DimHeatmap(seuratreactive$obj, dims = input$integerDR2, reduction = "harmony", nfeatures = input$PCnfeaturesDR2, cells = input$PCncellsDR2, balanced = TRUE, ncol = 1)
    }
  })
  
  output$plot6DR2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$VFintegrated), "You must complete the dimensionality reduction step"))
    if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
      ElbowPlot(seuratreactive$obj, ndims = seuratreactive$npcDR2)
    }
    else if (seuratreactive$integrationreduction == "harmony") {
      ElbowPlot(seuratreactive$obj, reduction = "harmony", ndims = seuratreactive$npcDR2)
    }
  })
  
  output$plot7DR2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$VFintegrated), "You must complete the dimensionality reduction step"))
    if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
      VizDimLoadings(seuratreactive$obj, dims = input$integer2DR2, nfeatures = input$LoadingsnfeaturesDR2)
    }
    else if (seuratreactive$integrationreduction == "harmony") {
      VizDimLoadings(seuratreactive$obj, dims = input$integer2DR2, reduction = "harmony", nfeatures = input$LoadingsnfeaturesDR2)
    }
  })
  
  observeEvent(input$startbuttonDR2, {
    
    if (!is.null(seuratreactive$VFintegrated)) {
      
      updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      #DR
      seuratreactive$variablefeatures <- NULL
      seuratreactive$SCTransformDR <- NULL
      seuratreactive$ALRAL <- NULL
      
      #C
      seuratreactive$integera <- NULL
      seuratreactive$integerb <- NULL
      seuratreactive$integerc <- NULL
      seuratreactive$integerd <- NULL
      seuratreactive$integere <- NULL
      seuratreactive$integerf <- NULL
      seuratreactive$integerg <- NULL
      seuratreactive$select1 <- NULL
      seuratreactive$tsne1 <- NULL
      seuratreactive$tsne2 <- NULL
      seuratreactive$tsne3 <- NULL
      seuratreactive$plot.data <- NULL
      seuratreactive$plot.data.original <- NULL
      
      seuratreactive$Identifiers <- NULL
      
      #CC
      seuratreactive$reductionCC1 <- NULL
      seuratreactive$reductionCC2 <- NULL
      seuratreactive$reductionCC3 <- NULL
      seuratreactive$GeneR <- NULL
      
      
      #L
      seuratreactive$reference_annotation <- NULL
      seuratreactive$select2 <- NULL
      seuratreactive$selecttissue <- NULL
      seuratreactive$new.cell.ids <- NULL
      seuratreactive$markers <- NULL
      
      #MEGENA
      seuratreactive$MEGENA_samples <- NULL
      seuratreactive$variablegenesMEGENA <- NULL
      seuratreactive$MEGENA_selectmethod <- NULL
      seuratreactive$MEGENApval <- NULL
      seuratreactive$MEGENAmin.size <- NULL
      seuratreactive$MEGENA_selectcluster <- NULL
      seuratreactive$n_MEGENA <- NULL
      seuratreactive$log2FCMEGENA <- NULL
      seuratreactive$padjustMEGENA <- NULL
      seuratreactive$testuseMEGENA <- NULL
      seuratreactive$minMEGENA <- NULL
      seuratreactive$pvalMEGENA <- NULL
      seuratreactive$summary.output <- NULL
      seuratreactive$markersMEGENA <- NULL
      seuratreactive$moduleGO_df <- NULL
      seuratreactive$VF <- NULL
      seuratreactive$g <- NULL
      
      #SCENIC
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$variablegenesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$n_SCENIC <- NULL
      seuratreactive$SCENICdb <- NULL
      seuratreactive$SCENIC_samples <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$SCENIC_selectcluster <- NULL
      seuratreactive$SCENIC_selectmethod <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
      
      #MAGMA
      seuratreactive$MAGMA_samples <- NULL
      seuratreactive$MAGMA_GWAS <- NULL
      seuratreactive$MAGMA_Final_Results <- NULL
      seuratreactive$meta <- NULL
      seuratreactive$MAGMA_selectcluster <- NULL
      seuratreactive$MAGMA_selectcluster2 <- NULL
      seuratreactive$MAGMA_samples2 <- NULL
      seuratreactive$textlabelMAGMA <- NULL
      seuratreactive$MAGMA_N <- NULL
      seuratreactive$GRCh <- NULL
      seuratreactive$MAGMA_Pop <- NULL
      seuratreactive$MAGMA_dirs <- NULL
      seuratreactive$genesOutPath2 <- NULL
      seuratreactive$n_MAGMA <- NULL
      seuratreactive$n_MAGMA2 <- NULL
      
      #CellChat
      seuratreactive$cellchat <- NULL
      seuratreactive$cellchatM <- NULL
      seuratreactive$object.list <- NULL
      seuratreactive$ChatChoose <- NULL
      seuratreactive$ChatSample <- NULL
      seuratreactive$Chatdb <- NULL
      seuratreactive$ChatMultiple1Sample <- NULL
      seuratreactive$ChatMultiple2Sample <- NULL
      seuratreactive$Chat_selectcluster <- NULL
      seuratreactive$Chat_number <- NULL
      seuratreactive$Chatpval <- NULL
      
      #DRUG
      seuratreactive$markersDRUG <- NULL
      seuratreactive$DRUGdf <- NULL
      
      #DRUG2
      seuratreactive$markersDRUG2 <- NULL
      seuratreactive$DRUGdf2 <- NULL
      
      #GE
      seuratreactive$titleGE <- NULL
      seuratreactive$GEInput <- NULL
      seuratreactive$GSEInput <- NULL
      seuratreactive$p <- NULL
      seuratreactive$q <- NULL
      seuratreactive$objGEyes <- NULL
      seuratreactive$Gene_vector_msig <- NULL
      
      #TA
      seuratreactive$TAselectedcells <- NULL
      seuratreactive$TA_pickercluster <- NULL
      seuratreactive$seurat_monocle <- NULL
      seuratreactive$Monocle3_genes <- NULL
      
      #MM
      seuratreactive$MM <- NULL
      seuratreactive$plot.dataM <- NULL
      seuratreactive$plot.dataMM <- NULL
      
      #DE
      seuratreactive$onlyposDE <- NULL
      seuratreactive$Subset_GENES <- NULL
      seuratreactive$DE_clusters_samples <- NULL
      seuratreactive$DECluster <- NULL
      seuratreactive$DEInput1_samples <- NULL
      seuratreactive$DEInput2_samples <- NULL
      seuratreactive$SelectComparison <- NULL
      seuratreactive$DE_clusters <- NULL
      seuratreactive$GOlabelsDE <- NULL
      seuratreactive$DEInput1 <- NULL
      seuratreactive$DEInput2 <- NULL
      seuratreactive$logDE <- NULL
      seuratreactive$minDE <- NULL
      seuratreactive$padjustDE1 <- NULL
      seuratreactive$testuseDE <- NULL
      seuratreactive$enrichInputDE <- NULL
      seuratreactive$pvalueDE <- NULL
      seuratreactive$pvalDE <- NULL
      seuratreactive$padjustDE <- NULL
      seuratreactive$volcanovalue1 <- NULL
      seuratreactive$volcanovalue2 <- NULL
      seuratreactive$pvaluevolcano <- NULL
      seuratreactive$markersDE <- NULL
      seuratreactive$YuDE <- NULL
      seuratreactive$markersDEvolcano <- NULL
      seuratreactive$PATHWAY <- NULL
      
      #SDE
      seuratreactive$Subset_GENESSDE <- NULL
      seuratreactive$selectedkey1 <- NULL
      seuratreactive$selectedkey2 <- NULL
      seuratreactive$onlyposSDE <- NULL
      seuratreactive$GOlabelsSDE <- NULL
      seuratreactive$logSDE <- NULL
      seuratreactive$minSDE <- NULL
      seuratreactive$padjustSDE1 <- NULL
      seuratreactive$testuseSDE <- NULL
      seuratreactive$enrichInputSDE <- NULL
      seuratreactive$pvalueSDE <- NULL
      seuratreactive$pvalSDE <- NULL
      seuratreactive$padjustSDE <- NULL
      seuratreactive$volcanoSDEvalue1 <- NULL
      seuratreactive$volcanoSDEvalue2 <- NULL
      seuratreactive$pvaluevolcanoSDE <- NULL
      seuratreactive$PATHWAYSDE <- NULL
      
      seuratreactive$markersSDE <- NULL
      seuratreactive$YuSDE <- NULL
      seuratreactive$markersSDEvolcano <- NULL
      
      #PA
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      #SPA
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      if (input$iscollapsebox2 == TRUE) {
        js$collapse("CBoxIntro")
      }
      shinyjs::hide("CBox1")
      shinyjs::hide("CBox2")
      shinyjs::hide("CBox3")
      
      shinyjs::enable("startbuttonC")
      shinyjs::hide("NextButtonC")
      
      if (!is.null(seuratreactive$obj)) {
        shinyjs::show("NextButtonDR2")
      }
      
      shinyjs::hide("NextButtonCombine")
      shinyjs::hide("NextButtonD2")
      
      if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),                                                           
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
        ))
      }
      else {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),                                                           
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = T),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
        ))
      }
    }
  },ignoreInit = T)
  
  observeEvent(input$NextButtonDR2, {
    if (input$iscollapsebox2 == TRUE) {
      js$collapse("CBoxIntro")
    }
    shinyjs::hide("CBox1")
    shinyjs::hide("CBox2")
    shinyjs::hide("CBox3")
    
    shinyjs::enable("startbuttonC")
    shinyjs::hide("NextButtonC")
    
    shinyjs::hide("NextButtonDR2")
    
    shinyjs::hide("NextButtonCombine")
    shinyjs::hide("NextButtonD2")
    
    if (!is.null(seuratreactive$objQC2)) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),                                                           
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
    else {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),                                                           
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
  },ignoreInit = T)
  
  
  ####TAB4 - Clustering#####
  observeEvent(input$tabs, {
    if (input$tabs == "Cluster" & !is.null(seuratreactive$reductionCC1)) {
      shinyalert("Warning!", "If you adjust clustering parameters, all the following tabs will disappear and you will have to 
                 re-do the analysis at each following step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observeEvent(input$startbuttonC, {
    tryCatch({
      if (input$iscollapsebox2 == FALSE) {
        js$collapse("CBoxIntro")
      }
      shinyjs::show("CBox1")
      shinyjs::show("CBox2")
      shinyjs::show("CBox3")
      
      shinyjs::enable("startbuttonCC")
      shinyjs::hide("NextButtonCC")
      
      shinyjs::disable("startbuttonC")
      
      shinyjs::hide("NextButtonDR2")
      shinyjs::hide("NextButtonDR")
      
      updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      
      #CC
      seuratreactive$reductionCC1 <- NULL
      seuratreactive$reductionCC2 <- NULL
      seuratreactive$reductionCC3 <- NULL
      seuratreactive$GeneR <- NULL
      
      
      #L
      seuratreactive$reference_annotation <- NULL
      seuratreactive$select2 <- NULL
      seuratreactive$selecttissue <- NULL
      seuratreactive$new.cell.ids <- NULL
      seuratreactive$markers <- NULL
      
      #MEGENA
      seuratreactive$MEGENA_samples <- NULL
      seuratreactive$variablegenesMEGENA <- NULL
      seuratreactive$MEGENA_selectmethod <- NULL
      seuratreactive$MEGENApval <- NULL
      seuratreactive$MEGENAmin.size <- NULL
      seuratreactive$MEGENA_selectcluster <- NULL
      seuratreactive$n_MEGENA <- NULL
      seuratreactive$log2FCMEGENA <- NULL
      seuratreactive$padjustMEGENA <- NULL
      seuratreactive$testuseMEGENA <- NULL
      seuratreactive$minMEGENA <- NULL
      seuratreactive$pvalMEGENA <- NULL
      seuratreactive$summary.output <- NULL
      seuratreactive$markersMEGENA <- NULL
      seuratreactive$moduleGO_df <- NULL
      seuratreactive$VF <- NULL
      seuratreactive$g <- NULL
      
      #SCENIC
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$variablegenesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$n_SCENIC <- NULL
      seuratreactive$SCENICdb <- NULL
      seuratreactive$SCENIC_samples <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$SCENIC_selectcluster <- NULL
      seuratreactive$SCENIC_selectmethod <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
      
      #MAGMA
      seuratreactive$MAGMA_samples <- NULL
      seuratreactive$MAGMA_GWAS <- NULL
      seuratreactive$MAGMA_Final_Results <- NULL
      seuratreactive$meta <- NULL
      seuratreactive$MAGMA_selectcluster <- NULL
      seuratreactive$MAGMA_selectcluster2 <- NULL
      seuratreactive$MAGMA_samples2 <- NULL
      seuratreactive$textlabelMAGMA <- NULL
      seuratreactive$MAGMA_N <- NULL
      seuratreactive$GRCh <- NULL
      seuratreactive$MAGMA_Pop <- NULL
      seuratreactive$MAGMA_dirs <- NULL
      seuratreactive$genesOutPath2 <- NULL
      seuratreactive$n_MAGMA <- NULL
      seuratreactive$n_MAGMA2 <- NULL
      
      #CellChat
      seuratreactive$cellchat <- NULL
      seuratreactive$cellchatM <- NULL
      seuratreactive$object.list <- NULL
      seuratreactive$ChatChoose <- NULL
      seuratreactive$ChatSample <- NULL
      seuratreactive$Chatdb <- NULL
      seuratreactive$ChatMultiple1Sample <- NULL
      seuratreactive$ChatMultiple2Sample <- NULL
      seuratreactive$Chat_selectcluster <- NULL
      seuratreactive$Chat_number <- NULL
      seuratreactive$Chatpval <- NULL
      
      #DRUG
      seuratreactive$markersDRUG <- NULL
      seuratreactive$DRUGdf <- NULL
      
      #DRUG2
      seuratreactive$markersDRUG2 <- NULL
      seuratreactive$DRUGdf2 <- NULL
      
      #GE
      seuratreactive$titleGE <- NULL
      seuratreactive$GEInput <- NULL
      seuratreactive$GSEInput <- NULL
      seuratreactive$p <- NULL
      seuratreactive$q <- NULL
      seuratreactive$objGEyes <- NULL
      seuratreactive$Gene_vector_msig <- NULL
      
      #TA
      seuratreactive$TAselectedcells <- NULL
      seuratreactive$TA_pickercluster <- NULL
      seuratreactive$seurat_monocle <- NULL
      seuratreactive$Monocle3_genes <- NULL
      
      #MM
      seuratreactive$MM <- NULL
      seuratreactive$plot.dataM <- NULL
      seuratreactive$plot.dataMM <- NULL
      
      #DE
      seuratreactive$onlyposDE <- NULL
      seuratreactive$Subset_GENES <- NULL
      seuratreactive$DE_clusters_samples <- NULL
      seuratreactive$DECluster <- NULL
      seuratreactive$DEInput1_samples <- NULL
      seuratreactive$DEInput2_samples <- NULL
      seuratreactive$SelectComparison <- NULL
      seuratreactive$DE_clusters <- NULL
      seuratreactive$GOlabelsDE <- NULL
      seuratreactive$DEInput1 <- NULL
      seuratreactive$DEInput2 <- NULL
      seuratreactive$logDE <- NULL
      seuratreactive$minDE <- NULL
      seuratreactive$padjustDE1 <- NULL
      seuratreactive$testuseDE <- NULL
      seuratreactive$enrichInputDE <- NULL
      seuratreactive$pvalueDE <- NULL
      seuratreactive$pvalDE <- NULL
      seuratreactive$padjustDE <- NULL
      seuratreactive$volcanovalue1 <- NULL
      seuratreactive$volcanovalue2 <- NULL
      seuratreactive$pvaluevolcano <- NULL
      seuratreactive$markersDE <- NULL
      seuratreactive$YuDE <- NULL
      seuratreactive$markersDEvolcano <- NULL
      seuratreactive$PATHWAY <- NULL
      
      #SDE
      seuratreactive$Subset_GENESSDE <- NULL
      seuratreactive$selectedkey1 <- NULL
      seuratreactive$selectedkey2 <- NULL
      seuratreactive$onlyposSDE <- NULL
      seuratreactive$GOlabelsSDE <- NULL
      seuratreactive$logSDE <- NULL
      seuratreactive$minSDE <- NULL
      seuratreactive$padjustSDE1 <- NULL
      seuratreactive$testuseSDE <- NULL
      seuratreactive$enrichInputSDE <- NULL
      seuratreactive$pvalueSDE <- NULL
      seuratreactive$pvalSDE <- NULL
      seuratreactive$padjustSDE <- NULL
      seuratreactive$volcanoSDEvalue1 <- NULL
      seuratreactive$volcanoSDEvalue2 <- NULL
      seuratreactive$pvaluevolcanoSDE <- NULL
      seuratreactive$PATHWAYSDE <- NULL
      
      seuratreactive$markersSDE <- NULL
      seuratreactive$YuSDE <- NULL
      seuratreactive$markersSDEvolcano <- NULL
      
      #PA
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      #SPA
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      
    },
    error = function(e) {
      shinyalert("ERROR!", "You have uploaded two identical datasets, clustering will not work.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observe({
    if (is.na(input$integera) |
        is.na(input$integerb) |
        is.na(input$integerc) |
        is.na(input$integerd) |
        is.na(input$integere) |
        is.na(input$integerf) |
        is.na(input$integerg) |
        is.na(input$select1) |
        is.na(input$tsne1) |
        is.na(input$tsne2) |
        is.na(input$tsne3)) {
      shinyjs::disable("button5")
    }
    else {
      shinyjs::enable("button5")
    }
  })
  
  
  observeEvent(input$button5, {
    tryCatch({
      withProgress(message = 'Please wait, this may take a while...',
                   value = 0, {
                     incProgress(1/10)
                     
                     if (is.null(seuratreactive$integrationreduction)) {
                       
                       #K-nearest neighbours
                       seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                       
                       #FindClusters
                       
                       
                       if (input$select1 == "Louvain") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                       }
                       else if (input$select1 == "SLM") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                       }
                       else if (input$select1 == "Leiden") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                       }
                       
                       incProgress(5/10)
                       
                       #2D
                       #RunUMAP
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                       incProgress(6/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                       
                       #3D
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                       incProgress(8/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                       incProgress(9/10)
                     }
                     
                     else if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
                       
                       DefaultAssay(seuratreactive$obj) <- "integrated"
                       
                       #K-nearest neighbours
                       seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                       
                       #FindClusters
                       if (input$select1 == "Louvain") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                       }
                       else if (input$select1 == "SLM") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                       }
                       else if (input$select1 == "Leiden") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                       }
                       
                       incProgress(5/10)
                       
                       #2D
                       #RunUMAP
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                       incProgress(6/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                       
                       #3D
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                       incProgress(8/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                       incProgress(9/10)
                     }
                     
                     else if (seuratreactive$integrationreduction == "harmony") {
                       #K-nearest neighbours
                       seuratreactive$obj <- FindNeighbors(seuratreactive$obj, reduction = "harmony", dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                       
                       #FindClusters
                       if (input$select1 == "Louvain") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                       }
                       else if (input$select1 == "SLM") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                       }
                       else if (input$select1 == "Leiden") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                       }
                       
                       incProgress(5/10)
                       
                       #2D
                       #RunUMAP
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                       incProgress(6/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)  
                       
                       #3D
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                       incProgress(8/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                       incProgress(9/10)
                     }
                     
                     plot.data2D <- FetchData(object = seuratreactive$obj, vars = c("UMAP_1", "UMAP_2", "tSNE_1", "tSNE_2"))
                     
                     colnames(plot.data2D) <- c("UMAP_1_2D", "UMAP_2_2D", "tSNE_1_2D", "tSNE_2_2D")
                     
                     
                     plot.data3D <- FetchData(object = seuratreactive$obj, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "tSNE_1", "tSNE_2", "tSNE_3"))
                     
                     colnames(plot.data3D) <- c("UMAP_1_3D", "UMAP_2_3D", "UMAP_3_3D", "tSNE_1_3D", "tSNE_2_3D", "tSNE_3_3D")
                     
                     seuratreactive$plot.data.original <- merge(plot.data2D, plot.data3D, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data.original) <- seuratreactive$plot.data.original$Row.names
                     
                     seuratreactive$plot.data.original <- seuratreactive$plot.data.original[,-1]
                     
                     #Merge identifiers
                     seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
                     
                     colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, seuratreactive$Identifiers, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     
                     seuratreactive$integera <- input$integera
                     seuratreactive$integerb <- input$integerb
                     seuratreactive$integerc <- input$integerc
                     seuratreactive$integerd <- input$integerd
                     seuratreactive$integere <- input$integere
                     seuratreactive$integerf <- input$integerf
                     seuratreactive$integerg <- input$integerg
                     seuratreactive$select1 <- input$select1
                     seuratreactive$tsne1 <- input$tsne1
                     seuratreactive$tsne2 <- input$tsne2
                     seuratreactive$tsne3 <- input$tsne3
                     
                     shinyjs::enable("startbuttonCC")
                     shinyjs::hide("NextButtonCC")
                     
                     shinyjs::disable("startbuttonC")
                     shinyjs::show("NextButtonC")
                     
                     shinyjs::hide("NextButtonDR2")
                     shinyjs::hide("NextButtonDR")
                     
                     incProgress(10/10)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check clustering parameters. Please make sure the number of dimensions does not exceed the number of dimensions computed during the dimensionality reduction step (refer to dimensionality reduction tab).", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$plot8UI <- renderUI({
    plotlyOutput("plot8", height = (400 + input$Cplotheight*40))
  })
  
  output$plot8 <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "Please select your clustering parameters and click APPLY to begin the clustering process"))
    if(input$radio == "UMAP") {
      plot_ly(data = seuratreactive$plot.data, 
              x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
              color = seuratreactive$plot.data[, input$plotclustersC],
              colors = col_vector,
              type = "scatter", 
              mode = "markers", 
              marker = list(size = input$size),
              opacity = input$opacity,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                 "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                 "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                 "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`)
      ) %>% layout(font = list(size = input$labelsize))
    }
    else if (input$radio == "3D UMAP") {
      plot_ly(data = seuratreactive$plot.data, 
              x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D, 
              color = seuratreactive$plot.data[, input$plotclustersC],
              colors = col_vector,
              type = "scatter3d", 
              mode = "markers", 
              marker = list(size = input$size),
              opacity = input$opacity,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                 "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                 "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                 "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`)
      ) %>% layout(font = list(size = input$labelsize))
    }
    else if (input$radio == "t-SNE") {
      plot_ly(data = seuratreactive$plot.data, 
              x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
              color = seuratreactive$plot.data[, input$plotclustersC],
              colors = col_vector,
              type = "scatter", 
              mode = "markers", 
              marker = list(size = input$size), 
              opacity = input$opacity,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                 "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                 "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                 "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`)
      ) %>% layout(font = list(size = input$labelsize))
    }
    else if (input$radio == "3D t-SNE") {
      plot_ly(data = seuratreactive$plot.data, 
              x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
              color = seuratreactive$plot.data[, input$plotclustersC], 
              colors = col_vector,
              type = "scatter3d", 
              mode = "markers", 
              marker = list(size = input$size),
              opacity = input$opacity,
              hoverinfo = "text",
              hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                 "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                 "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                 "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`)
      ) %>% layout(font = list(size = input$labelsize))
    }
  })
  
  output$sampledistributionUI <- renderUI({
    plotlyOutput("sampledistribution", height = (300 + input$HeightC*30))
  })
  
  output$sampledistribution <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "Please select your clustering parameters and click APPLY to begin the clustering process"))
    
    if(input$Sample_selectionC == "Seurat Clusters") {
      seuratreactive$Graph <- seuratreactive$Identifiers %>% group_by(Sample, `Seurat Clusters`) %>%  
        summarise(Value = sum(n())) %>% 
        group_by(Sample) %>% 
        mutate(total = sum(Value)) %>% 
        mutate(Percentage = round(Value/total, 3))
      
      if (input$switchPercent == FALSE) {
        ggplot(seuratreactive$Graph, aes(fill=`Seurat Clusters`, y=`Value`, x=`Sample`)) + 
          geom_bar(position="stack", stat="identity") + 
          ylab("") + 
          theme_bw() +
          theme(text = element_text(size=input$labelsizeC),
                axis.text.x= element_text(size = input$labelsizeC),
                axis.text.y= element_text(size = input$labelsizeC))
      }
      else {
        ggplot(seuratreactive$Graph, aes(x=Sample, y=Percentage, fill = `Seurat Clusters`)) +
          geom_bar(position = "fill", stat = "identity") +
          scale_y_continuous(labels = scales::percent) +
          ylab("") +
          theme_bw() +
          theme(text = element_text(size=input$labelsizeC),
                axis.text.x= element_text(size = input$labelsizeC),
                axis.text.y= element_text(size = input$labelsizeC))
      }
    }
    else if (input$Sample_selectionC == "Labelled Cells") {
      seuratreactive$Graph <- seuratreactive$Identifiers %>% group_by(Sample, `Labelled Cells`) %>%  
        summarise(Value = sum(n())) %>% 
        group_by(Sample) %>% 
        mutate(total = sum(Value)) %>% 
        mutate(Percentage = round(Value/total, 3))
      
      if (input$switchPercent == FALSE) {
        ggplot(seuratreactive$Graph, aes(fill=`Labelled Cells`, y=`Value`, x=`Sample`)) + 
          geom_bar(position="stack", stat="identity") + 
          ylab("") + 
          theme_bw() +
          theme(text = element_text(size=input$labelsizeC),
                axis.text.x= element_text(size = input$labelsizeC),
                axis.text.y= element_text(size = input$labelsizeC))
      }
      else {
        ggplot(seuratreactive$Graph, aes(x=Sample, y=Percentage, fill = `Labelled Cells`)) +
          geom_bar(position = "fill", stat = "identity") +
          scale_y_continuous(labels = scales::percent) +
          ylab("") +
          theme_bw() +
          theme(text = element_text(size=input$labelsizeC),
                axis.text.x= element_text(size = input$labelsizeC),
                axis.text.y= element_text(size = input$labelsizeC))
      }
    }
  })
  
  observeEvent(input$button5, {
    
    if (!is.null(seuratreactive$integera)) {
      
      if (input$iscollapsebox3 == TRUE) {
        js$collapse("CCBoxIntro")
      }
      shinyjs::hide("CCBox1")
      shinyjs::hide("CCBox1a")
      shinyjs::hide("CCBox1b")
      shinyjs::hide("CCBox1c")
      shinyjs::hide("CCBox2")
      
      shinyjs::enable("startbuttonCC")
      shinyjs::hide("NextButtonCC")
      
      shinyjs::disable("startbuttonC")
      shinyjs::show("NextButtonC")
      
      shinyjs::hide("NextButtonDR2")
      shinyjs::hide("NextButtonDR")
      
      updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      
      #CC
      seuratreactive$reductionCC1 <- NULL
      seuratreactive$reductionCC2 <- NULL
      seuratreactive$reductionCC3 <- NULL
      seuratreactive$GeneR <- NULL
      
      
      #L
      seuratreactive$reference_annotation <- NULL
      seuratreactive$select2 <- NULL
      seuratreactive$selecttissue <- NULL
      seuratreactive$new.cell.ids <- NULL
      seuratreactive$markers <- NULL
      
      #MEGENA
      seuratreactive$MEGENA_samples <- NULL
      seuratreactive$variablegenesMEGENA <- NULL
      seuratreactive$MEGENA_selectmethod <- NULL
      seuratreactive$MEGENApval <- NULL
      seuratreactive$MEGENAmin.size <- NULL
      seuratreactive$MEGENA_selectcluster <- NULL
      seuratreactive$n_MEGENA <- NULL
      seuratreactive$log2FCMEGENA <- NULL
      seuratreactive$padjustMEGENA <- NULL
      seuratreactive$testuseMEGENA <- NULL
      seuratreactive$minMEGENA <- NULL
      seuratreactive$pvalMEGENA <- NULL
      seuratreactive$summary.output <- NULL
      seuratreactive$markersMEGENA <- NULL
      seuratreactive$moduleGO_df <- NULL
      seuratreactive$VF <- NULL
      seuratreactive$g <- NULL
      
      #SCENIC
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$variablegenesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$n_SCENIC <- NULL
      seuratreactive$SCENICdb <- NULL
      seuratreactive$SCENIC_samples <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$SCENIC_selectcluster <- NULL
      seuratreactive$SCENIC_selectmethod <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
      
      #MAGMA
      seuratreactive$MAGMA_samples <- NULL
      seuratreactive$MAGMA_GWAS <- NULL
      seuratreactive$MAGMA_Final_Results <- NULL
      seuratreactive$meta <- NULL
      seuratreactive$MAGMA_selectcluster <- NULL
      seuratreactive$MAGMA_selectcluster2 <- NULL
      seuratreactive$MAGMA_samples2 <- NULL
      seuratreactive$textlabelMAGMA <- NULL
      seuratreactive$MAGMA_N <- NULL
      seuratreactive$GRCh <- NULL
      seuratreactive$MAGMA_Pop <- NULL
      seuratreactive$MAGMA_dirs <- NULL
      seuratreactive$genesOutPath2 <- NULL
      seuratreactive$n_MAGMA <- NULL
      seuratreactive$n_MAGMA2 <- NULL
      
      #CellChat
      seuratreactive$cellchat <- NULL
      seuratreactive$cellchatM <- NULL
      seuratreactive$object.list <- NULL
      seuratreactive$ChatChoose <- NULL
      seuratreactive$ChatSample <- NULL
      seuratreactive$Chatdb <- NULL
      seuratreactive$ChatMultiple1Sample <- NULL
      seuratreactive$ChatMultiple2Sample <- NULL
      seuratreactive$Chat_selectcluster <- NULL
      seuratreactive$Chat_number <- NULL
      seuratreactive$Chatpval <- NULL
      
      #DRUG
      seuratreactive$markersDRUG <- NULL
      seuratreactive$DRUGdf <- NULL
      
      #DRUG2
      seuratreactive$markersDRUG2 <- NULL
      seuratreactive$DRUGdf2 <- NULL
      
      #GE
      seuratreactive$titleGE <- NULL
      seuratreactive$GEInput <- NULL
      seuratreactive$GSEInput <- NULL
      seuratreactive$p <- NULL
      seuratreactive$q <- NULL
      seuratreactive$objGEyes <- NULL
      seuratreactive$Gene_vector_msig <- NULL
      
      #TA
      seuratreactive$TAselectedcells <- NULL
      seuratreactive$TA_pickercluster <- NULL
      seuratreactive$seurat_monocle <- NULL
      seuratreactive$Monocle3_genes <- NULL
      
      #MM
      seuratreactive$MM <- NULL
      seuratreactive$plot.dataM <- NULL
      seuratreactive$plot.dataMM <- NULL
      
      #DE
      seuratreactive$onlyposDE <- NULL
      seuratreactive$Subset_GENES <- NULL
      seuratreactive$DE_clusters_samples <- NULL
      seuratreactive$DECluster <- NULL
      seuratreactive$DEInput1_samples <- NULL
      seuratreactive$DEInput2_samples <- NULL
      seuratreactive$SelectComparison <- NULL
      seuratreactive$DE_clusters <- NULL
      seuratreactive$GOlabelsDE <- NULL
      seuratreactive$DEInput1 <- NULL
      seuratreactive$DEInput2 <- NULL
      seuratreactive$logDE <- NULL
      seuratreactive$minDE <- NULL
      seuratreactive$padjustDE1 <- NULL
      seuratreactive$testuseDE <- NULL
      seuratreactive$enrichInputDE <- NULL
      seuratreactive$pvalueDE <- NULL
      seuratreactive$pvalDE <- NULL
      seuratreactive$padjustDE <- NULL
      seuratreactive$volcanovalue1 <- NULL
      seuratreactive$volcanovalue2 <- NULL
      seuratreactive$pvaluevolcano <- NULL
      seuratreactive$markersDE <- NULL
      seuratreactive$YuDE <- NULL
      seuratreactive$markersDEvolcano <- NULL
      seuratreactive$PATHWAY <- NULL
      
      #SDE
      seuratreactive$Subset_GENESSDE <- NULL
      seuratreactive$selectedkey1 <- NULL
      seuratreactive$selectedkey2 <- NULL
      seuratreactive$onlyposSDE <- NULL
      seuratreactive$GOlabelsSDE <- NULL
      seuratreactive$logSDE <- NULL
      seuratreactive$minSDE <- NULL
      seuratreactive$padjustSDE1 <- NULL
      seuratreactive$testuseSDE <- NULL
      seuratreactive$enrichInputSDE <- NULL
      seuratreactive$pvalueSDE <- NULL
      seuratreactive$pvalSDE <- NULL
      seuratreactive$padjustSDE <- NULL
      seuratreactive$volcanoSDEvalue1 <- NULL
      seuratreactive$volcanoSDEvalue2 <- NULL
      seuratreactive$pvaluevolcanoSDE <- NULL
      seuratreactive$PATHWAYSDE <- NULL
      
      seuratreactive$markersSDE <- NULL
      seuratreactive$YuSDE <- NULL
      seuratreactive$markersSDEvolcano <- NULL
      
      #PA
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      #SPA
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      
      
      if (is.null(seuratreactive$objQC2)){
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
          )
          )
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
          )
          ) 
        }
      }
      else{
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = T),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
        ))
      }
    }
  }, ignoreInit = T)
  
  observeEvent(input$NextButtonC, {
    if (input$iscollapsebox3 == TRUE) {
      js$collapse("CCBoxIntro")
    }
    shinyjs::hide("CCBox1")
    shinyjs::hide("CCBox1a")
    shinyjs::hide("CCBox1b")
    shinyjs::hide("CCBox1c")
    shinyjs::hide("CCBox2")
    
    shinyjs::enable("startbuttonCC")
    shinyjs::hide("NextButtonCC")
    
    shinyjs::disable("startbuttonC")
    shinyjs::hide("NextButtonC")
    
    shinyjs::hide("NextButtonDR2")
    shinyjs::hide("NextButtonDR")
    
    if (is.null(seuratreactive$objQC2)){
      if (is.null(seuratreactive$VFintegrated)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
        )
        )
      }
      else {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
        )
        ) 
      }
    }
    else{
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                       menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)
      ))
    }
  }, ignoreInit = T)
  
  ####TAB 5 - Cell Cycle####
  
  observe({
    
    if (is.null(input$CC1) | !is.null(input$CC1) && input$CC1 == "")
    {
      shinyjs::disable("regress1")     
    }
    else{
      shinyjs::enable("regress1")
    }
  })
  
  observe({
    if (is.null(input$CC2) | !is.null(input$CC2) && input$CC2 == "")
    {
      shinyjs::disable("regress2")     
    }
    else{
      shinyjs::enable("regress2")
    }
  })
  
  observe({
    if (is.null(input$CC3) | !is.null(input$CC3) && input$CC3 == "")
    {
      shinyjs::disable("regress3")     
    }
    else{
      shinyjs::enable("regress3")
    }
  })
  
  observeEvent(input$tabs, {
    if (input$tabs == "CC" & !is.null(seuratreactive$new.cell.ids) |
        input$tabs == "CC" & !is.null(seuratreactive$summary.output) | 
        input$tabs == "CC" & !is.null(seuratreactive$regulons) |
        input$tabs == "CC" & !is.null(seuratreactive$objGEyes) | 
        input$tabs == "CC" & !is.null(seuratreactive$TA_pickercluster) |
        input$tabs == "CC" & !is.null(seuratreactive$MM) |
        input$tabs == "CC" & !is.null(seuratreactive$markersDE) |
        input$tabs == "CC" & !is.null(Yu$result) |
        input$tabs == "CC" & !is.null(seuratreactive$markersSDE) |
        input$tabs == "CC" & !is.null(YuSPA$result)) {
      shinyalert("Warning!", "If you adjust regression parameters, all the following tabs will disappear and you will have to 
                 re-do the analysis at each following step.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (input$CC1a == "") {
      shinyjs::hide("CCBox1a")
      shinyjs::hide("CCBox1b")
      shinyjs::hide("CCBox1c")
    }
    if (input$CC1a == "Confounding Effects") {
      shinyjs::show("CCBox1a")
      shinyjs::hide("CCBox1b")
      shinyjs::hide("CCBox1c")
    }
    if (input$CC1a == "Gene(s) of interest") {
      shinyjs::hide("CCBox1a")
      shinyjs::show("CCBox1b")
      shinyjs::hide("CCBox1c")
    }
    if (input$CC1a == "Gene pathway") {
      shinyjs::hide("CCBox1a")
      shinyjs::hide("CCBox1b")
      shinyjs::show("CCBox1c")
    }
  })
  
  observeEvent(input$startbuttonCC, {
    tryCatch({
      
      seuratreactive$obj <- CellCycleScoring(seuratreactive$obj, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes)
      
      IdentifiersCC <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
      
      IdentifiersCC[c("percent.ribo", "percent.mt")] <- signif(IdentifiersCC[c("percent.ribo", "percent.mt")], digits = 4)
      
      colnames(IdentifiersCC) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)", "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
      
      seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersCC, by = "row.names", all = T)
      
      rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
      
      seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
      
      seuratreactive$reductionCC1 <- "none"
      seuratreactive$reductionCC2 <- "none"
      seuratreactive$reductionCC3 <- "none"
      seuratreactive$GeneR <- "none"
      
      if (input$iscollapsebox3 == FALSE) {
        js$collapse("CCBoxIntro")
      }
      shinyjs::show("CCBox1")
      shinyjs::show("CCBox2")
      
      shinyjs::disable("startbuttonCC")
      shinyjs::show("NextButtonCC")
      
      shinyjs::hide("NextButtonC")
      
      updatePickerInput(session, 'CC1a', selected = "")
    },
    error = function(e) {
      shinyalert("WARNING!", "Low cell numbers are detected or cell cycle genes not present in dataset, click start again to continue anyway.", type = "warning", confirmButtonCol = "#337ab7")
      seuratreactive$obj <- NormalizeData(seuratreactive$obj)
    })
  })
  
  observeEvent(input$rewindCC |input$rewindCC2 | input$rewindCC3, {
    tryCatch({
      withProgress(message = "Please wait...",
                   value = 0, {
                     incProgress(1/10)
                     
                     if (is.null(seuratreactive$integrationreduction)) {
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize") {
                         #Data scaling
                         seuratreactive$obj <- ScaleData(seuratreactive$obj)
                       }
                       else {
                         seuratreactive$obj <- SCTransform(seuratreactive$obj, variable.features.n = seuratreactive$variablefeatures, vst.flavor = "v2")
                         
                         if (input$ALRAL == TRUE) {
                           seuratreactive$ALRAL <- "yes"
                           seuratreactive$obj <- RunALRA(seuratreactive$obj)
                           
                           seuratreactive$obj@assays$SCT@data <- seuratreactive$obj@assays$alra@data
                           
                           DefaultAssay(seuratreactive$obj) <- "SCT"
                           
                           seuratreactive$obj[['alra']] <- NULL
                         }
                         else {
                           seuratreactive$ALRAL <- "no"
                         }
                       }
                       
                       #Dimensionality reduction
                       seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR)
                       
                       #K-nearest neighbours
                       seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                       
                       #FindClusters
                       if (input$select1 == "Louvain") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                       }
                       else if (input$select1 == "SLM") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                       }
                       else if (input$select1 == "Leiden") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                       }
                       
                       incProgress(5/10)
                       
                       #2D
                       #RunUMAP
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                       incProgress(6/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                       incProgress(7/10)
                       
                       #3D
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                       incProgress(8/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                       incProgress(9/10)
                     }
                     
                     else if (!is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       
                       if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
                         
                         #Data scaling
                         seuratreactive$obj <- ScaleData(seuratreactive$obj)
                         
                         #Dimensionality reduction
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                         
                       }
                       else if (input$integrationreduction == "harmony") {
                         
                         seuratreactive$obj <- ScaleData(seuratreactive$obj)
                         
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         seuratreactive$obj <- RunHarmony(seuratreactive$obj, "orig.ident", max.iter.harmony = 20)
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, reduction = "harmony", dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                       }
                     }
                     
                     else if (!is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                       
                       if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
                         
                         seuratreactive$obj <- SCTransform(seuratreactive$obj, variable.features.n = seuratreactive$VFintegrated, vst.flavor = "v2")
                         
                         #Dimensionality reduction
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                       }
                       
                       else if (input$integrationreduction == "harmony") {
                         
                         seuratreactive$obj <- SCTransform(seuratreactive$obj, variable.features.n = seuratreactive$VFintegrated, vst.flavor = "v2")
                         
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         seuratreactive$obj <- RunHarmony(seuratreactive$obj, "orig.ident", max.iter.harmony = 20, assay.use = "SCT")
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, reduction = "harmony", dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                       }
                     }
                     
                     seuratreactive$obj <- CellCycleScoring(seuratreactive$obj, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes)
                     
                     plot.data2D <- FetchData(object = seuratreactive$obj, vars = c("UMAP_1", "UMAP_2", "tSNE_1", "tSNE_2"))
                     
                     colnames(plot.data2D) <- c("UMAP_1_2D", "UMAP_2_2D", "tSNE_1_2D", "tSNE_2_2D")
                     
                     
                     plot.data3D <- FetchData(object = seuratreactive$obj, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "tSNE_1", "tSNE_2", "tSNE_3"))
                     
                     colnames(plot.data3D) <- c("UMAP_1_3D", "UMAP_2_3D", "UMAP_3_3D", "tSNE_1_3D", "tSNE_2_3D", "tSNE_3_3D")
                     
                     seuratreactive$plot.data.original <- merge(plot.data2D, plot.data3D, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data.original) <- seuratreactive$plot.data.original$Row.names
                     
                     seuratreactive$plot.data.original <- seuratreactive$plot.data.original[,-1]
                     
                     IdentifiersCC <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersCC[c("percent.ribo", "percent.mt")] <- signif(IdentifiersCC[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersCC) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)", "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersCC, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$reductionCC1 <- "none"
                     seuratreactive$reductionCC2 <- "none"
                     seuratreactive$reductionCC3 <- "none"
                     seuratreactive$GeneR <- "none"
                     
                     shinyjs::disable("startbuttonCC")
                     shinyjs::show("NextButtonCC")
                     
                     shinyjs::hide("NextButtonC")
                     
                     incProgress(10/10)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if the previous step (clustering) was completed successfully.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  }, ignoreInit = T)
  
  observe({
    updateVirtualSelect('CC1', 
                        choices = c("Cell Cycle Phase", "nFeature_RNA", "nCount_RNA",
                                    "percent.ribo", "percent.mt"),
                        selected = seuratreactive$reductionCC1)
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) {
      if (!is.null(seuratreactive$VFintegrated)) {
        if (seuratreactive$integrationreduction == "rpca" | seuratreactive$integrationreduction == "cca") {
          updateVirtualSelect('CC2', choices = c(seuratreactive$obj[["integrated"]]@var.features), 
                              selected = seuratreactive$reductionCC2)
        }
        else if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                 !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
          updateVirtualSelect('CC2', choices = c(seuratreactive$obj[["RNA"]]@var.features), 
                              selected = seuratreactive$reductionCC2)
        }
        else {
          updateVirtualSelect('CC2', choices = c(seuratreactive$obj[["SCT"]]@var.features), 
                              selected = seuratreactive$reductionCC2)
        }
      }
      else if (!is.null(seuratreactive$variablefeatures)) {
        if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
            !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
          updateVirtualSelect('CC2', choices = c(seuratreactive$obj[["RNA"]]@var.features), 
                              selected = seuratreactive$reductionCC2)
        }
        else {
          updateVirtualSelect('CC2', choices = c(seuratreactive$obj[["SCT"]]@var.features), 
                              selected = seuratreactive$reductionCC2)
        }
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) {
      updateVirtualSelect('CC3', choices = c(sample(seuratreactive$msig$gs_name)), 
                          selected = seuratreactive$reductionCC3)
    }
  })
  
  output$GenepathwaytextCC <- renderUI({
    if (seuratreactive$reductionCC3 != "none") {
      HTML(paste("The following were regressed out of the dataset: ", paste(seuratreactive$reductionCC1, collapse = ", "), "<br>",
                 "The following genes were regressed out of the dataset: ", paste(seuratreactive$reductionCC2, collapse = ", "), "<br>",
                 "The following gene pathway were regressed out of the dataset: ", paste(seuratreactive$reductionCC3, collapse = ", "), "<br>",
                 "Genes in gene pathway: ", paste(strwrap(unique(seuratreactive$GeneR), 10), collapse = ", ")))
    }
    else {
      HTML(paste("The following were regressed out of the dataset: ", paste(seuratreactive$reductionCC1, collapse = ", "), "<br>",
                 "The following genes were regressed out of the dataset: ", paste(seuratreactive$reductionCC2, collapse = ", "), "<br>",
                 "The following gene pathway were regressed out of the dataset: ", paste(seuratreactive$reductionCC3, collapse = ", "), "<br>",
                 "Genes in gene pathway: none"))
    }
  })
  
  observeEvent(input$regress1 | input$regress2 | input$regress3, {
    tryCatch({
      withProgress(message = 'Performing Regression. Please wait, this may take a while...',
                   value = 0, {
                     incProgress(1/10)
                     
                     if (!is.null(input$CC1)) {
                       seuratreactive$reductionCC1 <- paste(input$CC1, collapse = ",")
                     }
                     else {
                       seuratreactive$reductionCC1 <- "none"
                     }
                     
                     if (!is.null(input$CC2)) {
                       percent.genes <- (Matrix::colSums(seuratreactive$obj[input$CC2])/Matrix::colSums(seuratreactive$obj))*100
                       seuratreactive$obj <- AddMetaData(object=seuratreactive$obj, metadata = percent.genes, col.name = "percent.custom.genes")
                       seuratreactive$reductionCC2 <- paste(input$CC2, collapse = ",")
                     }
                     else {
                       seuratreactive$obj@meta.data[["percent.custom.genes"]] <- NULL
                       seuratreactive$reductionCC2 <- "none"
                     }
                     
                     if (!is.null(input$CC3)) {
                       GeneR <- seuratreactive$msig[match(input$CC3, seuratreactive$msig$gs_name),]
                       
                       seuratreactive$GeneR <- GeneR[,-1] %>% as.character() %>% na.omit() %>% unique()
                       
                       
                       percent.pathway.genes <- (Matrix::colSums(seuratreactive$obj[seuratreactive$GeneR])/Matrix::colSums(seuratreactive$obj))*100
                       seuratreactive$obj <- AddMetaData(object=seuratreactive$obj, metadata = percent.pathway.genes, col.name = "percent.pathway.genes")
                       seuratreactive$reductionCC3 <- paste(input$CC3, collapse = ",")
                     }
                     else {
                       seuratreactive$obj@meta.data[["percent.pathway.genes"]] <- NULL
                       seuratreactive$reductionCC3 <- "none"
                       seuratreactive$GeneR <- "none"
                     }
                     
                     vars <- c()
                     
                     if (!is.null(input$CC1) && any(str_detect("Cell Cycle Phase", input$CC1))){
                       vars <- append(vars, c("S.Score", "G2M.Score"))
                     }
                     if (!is.null(input$CC1) && any(str_detect("nFeature_RNA", input$CC1))){
                       vars <- append(vars, "nFeature_RNA")
                     }
                     if (!is.null(input$CC1) && any(str_detect("nCount_RNA", input$CC1))){
                       vars <- append(vars, "nCount_RNA")
                     }
                     if (!is.null(input$CC1) && any(str_detect("percent.ribo", input$CC1))){
                       vars <- append(vars, "percent.ribo")
                     }
                     if (!is.null(input$CC1) && any(str_detect("percent.mt", input$CC1))){
                       vars <- append(vars, "percent.mt")
                     }
                     if (!is.null(seuratreactive$obj@meta.data$percent.custom.genes)) {
                       vars <- append(vars, "percent.custom.genes")
                     }
                     if (!is.null(seuratreactive$obj@meta.data$percent.pathway.genes)) {
                       vars <- append(vars, "percent.pathway.genes")
                     }
                     
                     
                     if (is.null(seuratreactive$integrationreduction)) {
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize") {
                         #Data scaling
                         seuratreactive$obj <- ScaleData(seuratreactive$obj, vars.to.regress = vars)
                       }
                       else {
                         seuratreactive$obj <- SCTransform(seuratreactive$obj, variable.features.n = seuratreactive$variablefeatures, vst.flavor = "v2", vars.to.regress = vars)
                         if (input$ALRAL == TRUE) {
                           seuratreactive$ALRAL <- "yes"
                           seuratreactive$obj <- RunALRA(seuratreactive$obj)
                           
                           seuratreactive$obj@assays$SCT@data <- seuratreactive$obj@assays$alra@data
                           
                           DefaultAssay(seuratreactive$obj) <- "SCT"
                           
                           seuratreactive$obj[['alra']] <- NULL
                         }
                         else {
                           seuratreactive$ALRAL <- "no"
                         }
                       }
                       
                       #Dimensionality reduction
                       seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR)
                       
                       #K-nearest neighbours
                       seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                       
                       #FindClusters
                       if (input$select1 == "Louvain") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                       }
                       else if (input$select1 == "SLM") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                       }
                       else if (input$select1 == "Leiden") {
                         seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                       }
                       
                       incProgress(5/10)
                       
                       #2D
                       #RunUMAP
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                       incProgress(6/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                       incProgress(7/10)
                       
                       #3D
                       seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                       incProgress(8/10)
                       #Run tsne
                       seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                       incProgress(9/10)
                     }
                     
                     else if (!is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       
                       if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
                         #Data scaling
                         seuratreactive$obj <- ScaleData(seuratreactive$obj, vars.to.regress = vars)
                         
                         #Dimensionality reduction
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                         
                       }
                       else if (input$integrationreduction == "harmony") {
                         
                         seuratreactive$obj <- ScaleData(seuratreactive$obj, vars.to.regress = vars)
                         
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         seuratreactive$obj <- RunHarmony(seuratreactive$obj, "orig.ident", max.iter.harmony = 20)
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, reduction = "harmony", dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                       }
                     }
                     
                     else if (!is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                       
                       if (input$integrationreduction == "rpca" | input$integrationreduction == "cca") {
                         
                         seuratreactive$obj <- SCTransform(seuratreactive$obj, variable.features.n = seuratreactive$VFintegrated, vst.flavor = "v2", vars.to.regress = vars)
                         
                         #Dimensionality reduction
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                       }
                       
                       else if (input$integrationreduction == "harmony") {
                         
                         seuratreactive$obj <- SCTransform(seuratreactive$obj, variable.features.n = seuratreactive$VFintegrated, vst.flavor = "v2", vars.to.regress = vars)
                         
                         seuratreactive$obj <- RunPCA(seuratreactive$obj, npcs = seuratreactive$npcDR2)
                         
                         seuratreactive$obj <- RunHarmony(seuratreactive$obj, "orig.ident", max.iter.harmony = 20, assay.use = "SCT")
                         
                         #K-nearest neighbours
                         seuratreactive$obj <- FindNeighbors(seuratreactive$obj, reduction = "harmony", dims = 1:input$integera, k.param = input$integerb, n.trees = input$integerc)
                         
                         #FindClusters
                         if (input$select1 == "Louvain") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 1)
                         }
                         else if (input$select1 == "SLM") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 3)
                         }
                         else if (input$select1 == "Leiden") {
                           seuratreactive$obj <- FindClusters(seuratreactive$obj, resolution = input$integerd, algorithm = 4)
                         }
                         
                         incProgress(5/10)
                         
                         #2D
                         #RunUMAP
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 2L)
                         incProgress(6/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 2)
                         incProgress(7/10)
                         
                         #3D
                         seuratreactive$obj <- RunUMAP(seuratreactive$obj, reduction = "harmony", dims = 1:input$integere, n.neighbors = input$integerf, min.dist = input$integerg, n.components = 3L)
                         incProgress(8/10)
                         #Run tsne
                         seuratreactive$obj <- RunTSNE(seuratreactive$obj, reduction = "harmony", dims = 1:input$tsne1, perplexity = input$tsne2, max_iter = input$tsne3, dim.embed = 3)
                         incProgress(9/10)
                       }
                     }
                     
                     plot.data2D <- FetchData(object = seuratreactive$obj, vars = c("UMAP_1", "UMAP_2", "tSNE_1", "tSNE_2"))
                     
                     colnames(plot.data2D) <- c("UMAP_1_2D", "UMAP_2_2D", "tSNE_1_2D", "tSNE_2_2D")
                     
                     
                     plot.data3D <- FetchData(object = seuratreactive$obj, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "tSNE_1", "tSNE_2", "tSNE_3"))
                     
                     colnames(plot.data3D) <- c("UMAP_1_3D", "UMAP_2_3D", "UMAP_3_3D", "tSNE_1_3D", "tSNE_2_3D", "tSNE_3_3D")
                     
                     seuratreactive$plot.data.original <- merge(plot.data2D, plot.data3D, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data.original) <- seuratreactive$plot.data.original$Row.names
                     
                     seuratreactive$plot.data.original <- seuratreactive$plot.data.original[,-1]
                     
                     
                     #Merge identifiers
                     IdentifiersCC <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersCC[c("percent.ribo", "percent.mt")] <- signif(IdentifiersCC[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersCC) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersCC, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     shinyjs::disable("startbuttonCC")
                     shinyjs::show("NextButtonCC")
                     
                     shinyjs::hide("NextButtonC")
                     
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if your starting species is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  output$plotCCUI <- renderUI({
    plotlyOutput("plotCC", height = (400 + input$CCplotheight*40))
  })
  
  output$plotCC <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    if(input$plotclustersCC == "Seurat Clusters" | input$plotclustersCC == "Cell Cycle Phase" | input$plotclustersCC == "Sample") {
      if(input$radioCC == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
      else if (input$radioCC == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
      else if (input$radioCC == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
      else if (input$radioCC == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
    }
    else {
      if(input$radioCC == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
      else if (input$radioCC == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
      else if (input$radioCC == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
      else if (input$radioCC == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersCC],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeCC),
                opacity = input$opacityCC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`,
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeCC))
      }
    }
  })
  
  observeEvent(input$regress1 | input$regress2 | input$regress3 | input$rewindCC | input$rewindCC2 | input$rewindCC3, {
    if (!is.null(seuratreactive$reductionCC1)) {
      
      updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      #L
      seuratreactive$reference_annotation <- NULL
      seuratreactive$select2 <- NULL
      seuratreactive$selecttissue <- NULL
      seuratreactive$new.cell.ids <- NULL
      seuratreactive$markers <- NULL
      
      #MEGENA
      seuratreactive$MEGENA_samples <- NULL
      seuratreactive$variablegenesMEGENA <- NULL
      seuratreactive$MEGENA_selectmethod <- NULL
      seuratreactive$MEGENApval <- NULL
      seuratreactive$MEGENAmin.size <- NULL
      seuratreactive$MEGENA_selectcluster <- NULL
      seuratreactive$n_MEGENA <- NULL
      seuratreactive$log2FCMEGENA <- NULL
      seuratreactive$padjustMEGENA <- NULL
      seuratreactive$testuseMEGENA <- NULL
      seuratreactive$minMEGENA <- NULL
      seuratreactive$pvalMEGENA <- NULL
      seuratreactive$summary.output <- NULL
      seuratreactive$markersMEGENA <- NULL
      seuratreactive$moduleGO_df <- NULL
      seuratreactive$VF <- NULL
      seuratreactive$g <- NULL
      
      #SCENIC
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$variablegenesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$n_SCENIC <- NULL
      seuratreactive$SCENICdb <- NULL
      seuratreactive$SCENIC_samples <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$SCENIC_selectcluster <- NULL
      seuratreactive$SCENIC_selectmethod <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
      
      #MAGMA
      seuratreactive$MAGMA_samples <- NULL
      seuratreactive$MAGMA_GWAS <- NULL
      seuratreactive$MAGMA_Final_Results <- NULL
      seuratreactive$meta <- NULL
      seuratreactive$MAGMA_selectcluster <- NULL
      seuratreactive$MAGMA_selectcluster2 <- NULL
      seuratreactive$MAGMA_samples2 <- NULL
      seuratreactive$textlabelMAGMA <- NULL
      seuratreactive$MAGMA_N <- NULL
      seuratreactive$GRCh <- NULL
      seuratreactive$MAGMA_Pop <- NULL
      seuratreactive$MAGMA_dirs <- NULL
      seuratreactive$genesOutPath2 <- NULL
      seuratreactive$n_MAGMA <- NULL
      seuratreactive$n_MAGMA2 <- NULL
      
      #CellChat
      seuratreactive$cellchat <- NULL
      seuratreactive$cellchatM <- NULL
      seuratreactive$object.list <- NULL
      seuratreactive$ChatChoose <- NULL
      seuratreactive$ChatSample <- NULL
      seuratreactive$Chatdb <- NULL
      seuratreactive$ChatMultiple1Sample <- NULL
      seuratreactive$ChatMultiple2Sample <- NULL
      seuratreactive$Chat_selectcluster <- NULL
      seuratreactive$Chat_number <- NULL
      seuratreactive$Chatpval <- NULL
      
      #DRUG
      seuratreactive$markersDRUG <- NULL
      seuratreactive$DRUGdf <- NULL
      
      #DRUG2
      seuratreactive$markersDRUG2 <- NULL
      seuratreactive$DRUGdf2 <- NULL
      
      #GE
      seuratreactive$titleGE <- NULL
      seuratreactive$GEInput <- NULL
      seuratreactive$GSEInput <- NULL
      seuratreactive$p <- NULL
      seuratreactive$q <- NULL
      seuratreactive$objGEyes <- NULL
      seuratreactive$Gene_vector_msig <- NULL
      
      #TA
      seuratreactive$TAselectedcells <- NULL
      seuratreactive$TA_pickercluster <- NULL
      seuratreactive$seurat_monocle <- NULL
      seuratreactive$Monocle3_genes <- NULL
      
      #MM
      seuratreactive$MM <- NULL
      seuratreactive$plot.dataM <- NULL
      seuratreactive$plot.dataMM <- NULL
      
      #DE
      seuratreactive$onlyposDE <- NULL
      seuratreactive$Subset_GENES <- NULL
      seuratreactive$DE_clusters_samples <- NULL
      seuratreactive$DECluster <- NULL
      seuratreactive$DEInput1_samples <- NULL
      seuratreactive$DEInput2_samples <- NULL
      seuratreactive$SelectComparison <- NULL
      seuratreactive$DE_clusters <- NULL
      seuratreactive$GOlabelsDE <- NULL
      seuratreactive$DEInput1 <- NULL
      seuratreactive$DEInput2 <- NULL
      seuratreactive$logDE <- NULL
      seuratreactive$minDE <- NULL
      seuratreactive$padjustDE1 <- NULL
      seuratreactive$testuseDE <- NULL
      seuratreactive$enrichInputDE <- NULL
      seuratreactive$pvalueDE <- NULL
      seuratreactive$pvalDE <- NULL
      seuratreactive$padjustDE <- NULL
      seuratreactive$volcanovalue1 <- NULL
      seuratreactive$volcanovalue2 <- NULL
      seuratreactive$pvaluevolcano <- NULL
      seuratreactive$markersDE <- NULL
      seuratreactive$YuDE <- NULL
      seuratreactive$markersDEvolcano <- NULL
      seuratreactive$PATHWAY <- NULL
      
      #SDE
      seuratreactive$Subset_GENESSDE <- NULL
      seuratreactive$selectedkey1 <- NULL
      seuratreactive$selectedkey2 <- NULL
      seuratreactive$onlyposSDE <- NULL
      seuratreactive$GOlabelsSDE <- NULL
      seuratreactive$logSDE <- NULL
      seuratreactive$minSDE <- NULL
      seuratreactive$padjustSDE1 <- NULL
      seuratreactive$testuseSDE <- NULL
      seuratreactive$enrichInputSDE <- NULL
      seuratreactive$pvalueSDE <- NULL
      seuratreactive$pvalSDE <- NULL
      seuratreactive$padjustSDE <- NULL
      seuratreactive$volcanoSDEvalue1 <- NULL
      seuratreactive$volcanoSDEvalue2 <- NULL
      seuratreactive$pvaluevolcanoSDE <- NULL
      seuratreactive$PATHWAYSDE <- NULL
      
      seuratreactive$markersSDE <- NULL
      seuratreactive$YuSDE <- NULL
      seuratreactive$markersSDEvolcano <- NULL
      
      #PA
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      #SPA
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      
      shinyjs::reset('fileinputMMload1')
      shinyjs::reset('fileinputMMload2')
      shinyjs::reset('fileinputMMload3')
      
      if (input$iscollapsebox4 == TRUE) {
        js$collapse("LBoxIntro")
      }
      shinyjs::hide("LBox1")
      shinyjs::hide("LBox2")
      shinyjs::hide("LBox3")
      shinyjs::hide("LBox4")
      shinyjs::hide("LBox5")
      shinyjs::hide("LBox6")
      shinyjs::hide("LBox7")
      shinyjs::hide("LBox8")
      
      shinyjs::disable("startbuttonCC")
      shinyjs::show("NextButtonCC")
      
      shinyjs::hide("NextButtonC")
      
      #DE
      if (input$iscollapsebox6 == TRUE) {
        js$collapse("DEBoxIntro")
      }
      shinyjs::hide("DEBox1")
      shinyjs::hide("DEBox1a")
      shinyjs::hide("DEBox1b")
      shinyjs::hide("DEBox2")
      shinyjs::hide("DEBox3")
      shinyjs::hide("DEBox4")
      shinyjs::hide("DEBox5")
      shinyjs::hide("DEBox6")
      shinyjs::enable("startbuttonDE")
      
      #SDE
      if (input$iscollapseboxSDE == TRUE) {
        js$collapse("SDEBoxIntro")
      }
      shinyjs::hide("SDEBox1")
      shinyjs::hide("SDEBox2")
      shinyjs::hide("SDEBox3")
      shinyjs::hide("SDEBox4")
      shinyjs::hide("SDEBox5")
      shinyjs::hide("SDEBox6")
      shinyjs::hide("SDEBox7")
      shinyjs::enable("startbuttonSDE")
      
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = T),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  }, ignoreInit = T)
  
  observeEvent(input$NextButtonCC, {
    if (is.null(seuratreactive$new.cell.ids) && is.null(seuratreactive$summary.output) && is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && is.null(seuratreactive$objGEyes) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      shinyalert("CONGRATS!", "You have completed data preprocessing! Analysis from this point on is in your hands, please explore the NEW tabs that have appeared in the left sidebar. Goodluck!", type = "success", confirmButtonCol = "#337ab7")
    }
  }, ignoreInit = T)
  
  observeEvent(input$NextButtonCC, {
    if (!is.null(seuratreactive$reductionCC1)) {
      shinyjs::reset('fileinputMMload1')
      shinyjs::reset('fileinputMMload2')
      shinyjs::reset('fileinputMMload3')
      
      if (input$iscollapsebox4 == TRUE) {
        js$collapse("LBoxIntro")
      }
      shinyjs::hide("LBox1")
      shinyjs::hide("LBox2")
      shinyjs::hide("LBox3")
      shinyjs::hide("LBox4")
      shinyjs::hide("LBox5")
      shinyjs::hide("LBox6")
      shinyjs::hide("LBox7")
      shinyjs::hide("LBox8")
      
      shinyjs::disable("startbuttonCC")
      shinyjs::hide("NextButtonCC")
      
      shinyjs::hide("NextButtonC")
      
      #DE
      if (input$iscollapsebox6 == TRUE) {
        js$collapse("DEBoxIntro")
      }
      shinyjs::hide("DEBox1")
      shinyjs::hide("DEBox1a")
      shinyjs::hide("DEBox1b")
      shinyjs::hide("DEBox2")
      shinyjs::hide("DEBox3")
      shinyjs::hide("DEBox4")
      shinyjs::hide("DEBox5")
      shinyjs::hide("DEBox6")
      shinyjs::enable("startbuttonDE")
      
      #SDE
      if (input$iscollapseboxSDE == TRUE) {
        js$collapse("SDEBoxIntro")
      }
      shinyjs::hide("SDEBox1")
      shinyjs::hide("SDEBox2")
      shinyjs::hide("SDEBox3")
      shinyjs::hide("SDEBox4")
      shinyjs::hide("SDEBox5")
      shinyjs::hide("SDEBox6")
      shinyjs::hide("SDEBox7")
      shinyjs::enable("startbuttonSDE")
      
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  }, ignoreInit = T)
  
  
  ####TAB 6 - Labelling Clusters####
  
  observe({
    if (!is.null(seuratreactive$new.cell.ids) && is.null(seuratreactive$summary.output) && is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && is.null(seuratreactive$objGEyes) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = T),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observe({
    if (input$selectLabelInput == "") {
      shinyjs::show("LBox3")
      shinyjs::hide("LBox4")
      shinyjs::hide("LBox5")
      shinyjs::hide("LBox6")
    }
    else if (input$selectLabelInput == "SingleR reference cell marker dataset") {
      shinyjs::show("LBox3")
      shinyjs::show("LBox4")
      shinyjs::hide("LBox5")
      shinyjs::hide("LBox6")
    }
    else if (input$selectLabelInput == "sctype reference cell marker dataset") {
      shinyjs::show("LBox3")
      shinyjs::hide("LBox4")
      shinyjs::hide("LBox5")
      shinyjs::show("LBox6")
    }
    else if (input$selectLabelInput == "Own cell labels") {
      shinyjs::show("LBox3")
      shinyjs::hide("LBox4")
      shinyjs::show("LBox5")
      shinyjs::hide("LBox6")
    }
  })
  
  observe({
    if (input$selecttissueSingleR == "") {
      shinyjs::hide("select2")
      shinyjs::hide("selectBrain")
      shinyjs::hide("selectPancreas")
      shinyjs::hide("selectMultiple")
      shinyjs::hide("selectHe")
      shinyjs::hide("ReferenceButtonBrain")
      shinyjs::hide("ReferenceButtonPancreas")
      shinyjs::hide("ReferenceButtonMultiple")
      shinyjs::hide("ReferenceButtonImmune")
      updateVirtualSelect("selectMultiple", selected = "Human Primary Cell Atlas")
    }
    else if (input$selecttissueSingleR == "Brain") {
      shinyjs::hide("select2")
      shinyjs::show("selectBrain")
      shinyjs::hide("selectPancreas")
      shinyjs::hide("selectMultiple")
      shinyjs::hide("selectHe")
      shinyjs::show("ReferenceButtonBrain")
      shinyjs::hide("ReferenceButtonPancreas")
      shinyjs::hide("ReferenceButtonMultiple")
      shinyjs::hide("ReferenceButtonImmune")
      updateVirtualSelect("selectMultiple", selected = "Human Primary Cell Atlas")
    }
    else if (input$selecttissueSingleR == "Pancreas") {
      shinyjs::hide("select2")
      shinyjs::hide("selectBrain")
      shinyjs::show("selectPancreas")
      shinyjs::hide("selectMultiple")
      shinyjs::hide("selectHe")
      shinyjs::hide("ReferenceButtonBrain")
      shinyjs::show("ReferenceButtonPancreas")
      shinyjs::hide("ReferenceButtonMultiple")
      shinyjs::hide("ReferenceButtonImmune")
      updateVirtualSelect("selectMultiple", selected = "Human Primary Cell Atlas")
    }
    else if (input$selecttissueSingleR == "Multiple") {
      shinyjs::hide("select2")
      shinyjs::hide("selectBrain")
      shinyjs::hide("selectPancreas")
      shinyjs::show("selectMultiple")
      shinyjs::hide("ReferenceButtonBrain")
      shinyjs::hide("ReferenceButtonPancreas")
      shinyjs::show("ReferenceButtonMultiple")
      shinyjs::hide("ReferenceButtonImmune")
    }
    else if (input$selecttissueSingleR == "Immune system") {
      shinyjs::show("select2")
      shinyjs::hide("selectBrain")
      shinyjs::hide("selectPancreas")
      shinyjs::hide("selectMultiple")
      shinyjs::hide("selectHe")
      shinyjs::hide("ReferenceButtonBrain")
      shinyjs::hide("ReferenceButtonPancreas")
      shinyjs::hide("ReferenceButtonMultiple")
      shinyjs::show("ReferenceButtonImmune")
      
      updateVirtualSelect("selectMultiple", selected = "Human Primary Cell Atlas")
    }
  })
  
  observe({
    if (input$selectMultiple == "He Organ Atlas") {
      shinyjs::show("selectHe")
    }
    else {
      shinyjs::hide("selectHe")
    }
  })
  
  observeEvent(input$startbuttonL,{
    library(scRNAseq)
    library(SingleR)
    library(celldex)
    library(scran)
    tryCatch({
      
      seuratreactive$obj <- SetIdent(seuratreactive$obj, value = "seurat_clusters")
      
      #Merge identifiers
      IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
      
      IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
      
      colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
      
      seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
      
      rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
      
      seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
      
      seuratreactive$reference_annotation <- "none"
      
      if (input$iscollapsebox4 == FALSE) {
        js$collapse("LBoxIntro")
      }
      shinyjs::show("LBox1")
      shinyjs::show("LBox2")
      shinyjs::show("LBox3")
      shinyjs::show("LBox7")
      shinyjs::show("LBox8")
      
      shinyjs::hide("NextButtonCC")
      
      updatePickerInput(session, 'selectLabelInput', selected = "")  
      
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if the previous step (data correction) was completed successfully.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  })
  
  output$LabelClustersTextInput <- renderUI({
    numclusters <- as.integer(length(levels(seuratreactive$obj$seurat_clusters)))
    lapply(1:numclusters, function(i) {
      textInput(paste("seuratlabel", levels(seuratreactive$obj$seurat_clusters)[i], sep = ""),
                paste("Cluster", levels(seuratreactive$obj$seurat_clusters)[i], "Label"))
    })
  })
  
  observe({
    if(!is.null(seuratreactive$integera)) {
      if (identical(seuratreactive$obj@active.ident, seuratreactive$obj$seurat_clusters)) {
        
        updatePickerInput(session, 'plotclustersC', choices = list("Seurat Clusters", "Sample"),
                          selected = "Seurat Clusters")
        
        updatePickerInput(session, 'plotclustersCC', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                          selected = "Cell Cycle Phase")
        
        updatePickerInput(session, 'Sample_selectionC', choices = list("Seurat Clusters"))  
      }
      else{
        
        updatePickerInput(session, 'plotclustersC', choices = list("Labelled Cells", "Seurat Clusters", "Sample"))
        
        updatePickerInput(session, 'plotclustersCC', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                          selected = "Cell Cycle Phase")
        
        updatePickerInput(session, 'Sample_selectionC', choices = list("Labelled Cells", "Seurat Clusters"))  
      }
    }
  })
  
  observe({
    if(!is.null(seuratreactive$reductionCC1)) {
      if (identical(seuratreactive$obj@active.ident, seuratreactive$obj$seurat_clusters)) {
        
        updatePickerInput(session, 'VlnInput', 
                          choices = list("Please select from the following" = "", "Seurat Clusters", "Cell Cycle Phase"),
                          selected = "") 
        
        updatePickerInput(session, 'GEDotInput', 
                          choices = list("Please select from the following" = "", "Seurat Clusters", "Cell Cycle Phase"),
                          selected = "") 
        
        updatePickerInput(session, 'TA_picker', choices = list("Seurat Clusters"))
        
        updatePickerInput(session, 'VlnInputMM', 
                          choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase"),
                          selected = "Seurat Clusters") 
        
        updatePickerInput(session, 'LC_clusters', choices = list("Seurat Clusters"))
        
        updatePickerInput(session, 'plotclustersL', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                   "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                          selected = "Seurat Clusters")
        
        updatePickerInput(session, 'plotclustersGE', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Gene Expression"),
                          selected = "Gene Expression")
        
        updatePickerInput(session, 'plotclustersSCENIC', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                        "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                        "Regulon"),
                          selected = "Regulon")
        
        if (!is.null(seuratreactive$TAselectedcells)) {
          updatePickerInput(session, 'plotclustersTA', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                      "Pseudotime"),
                            selected = "Pseudotime")
        }
        else {
          updatePickerInput(session, 'plotclustersTA', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                            selected = "Seurat Clusters")
        }
        
        if (is.null(seuratreactive$plot.data$Pseudotime)) {
          updatePickerInput(session, 'plotclustersTA', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                            selected = "Seurat Clusters")
        }
        
        updatePickerInput(session, 'plotclustersMM', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Gene expression"),
                          selected = "Gene expression")
        
        updatePickerInput(session, 'plotclustersDE', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Differential Expression"),
                          selected = "Sample")
        
        updatePickerInput(session, 'plotclustersSDE', choices = list("Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                     "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                          selected = "Seurat Clusters")
        
        if (!is.null(seuratreactive$DE_clusters)) {
          updatePickerInput(session, 'DE_clusters', choices = list("Seurat Clusters"), selected = seuratreactive$DE_clusters)
        }
        else {
          updatePickerInput(session, 'DE_clusters', choices = list("Seurat Clusters"))
        }
        
        if (!is.null(seuratreactive$DE_clusters_samples)) {
          updatePickerInput(session, 'DE_clusters_samples', choices = list("Seurat Clusters"), selected = seuratreactive$DE_clusters_samples)
        }
        else {
          updatePickerInput(session, 'DE_clusters_samples', choices = list("Seurat Clusters"))
        }
        
        if (!is.null(seuratreactive$MAGMA_selectcluster)) {
          updatePickerInput(session, 'MAGMA_selectcluster', choices = list("Seurat Clusters"), selected = seuratreactive$MAGMA_selectcluster)
        }
        else {
          updatePickerInput(session, 'MAGMA_selectcluster', choices = list("Seurat Clusters"))
        }
        
        if (!is.null(seuratreactive$MAGMA_selectcluster2)) {
          updatePickerInput(session, 'MAGMA_selectcluster2', choices = list("Seurat Clusters"), selected = seuratreactive$MAGMA_selectcluster2)
        }
        else {
          updatePickerInput(session, 'MAGMA_selectcluster2', choices = list("Seurat Clusters"))
        }
        
      }
      else{
        updatePickerInput(session, 'VlnInput', 
                          choices = list("Please select from the following" = "", "Labelled Cells", "Seurat Clusters", "Cell Cycle Phase"),
                          selected = "")
        
        updatePickerInput(session, 'GEDotInput', 
                          choices = list("Please select from the following" = "", "Labelled Cells", "Seurat Clusters", "Cell Cycle Phase"),
                          selected = "")
        
        updatePickerInput(session, 'TA_picker', choices = list("Seurat Clusters", "Labelled Cells"), selected = "Labelled Cells")
        
        updatePickerInput(session, 'VlnInputMM', 
                          choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase"),
                          selected = "Labelled Cells")
        
        updatePickerInput(session, 'LC_clusters', choices = list("Labelled Cells", "Seurat Clusters")) 
        
        updatePickerInput(session, 'plotclustersL', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                   "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                          selected = "Labelled Cells")
        
        updatePickerInput(session, 'plotclustersGE', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Gene Expression"),
                          selected = "Gene Expression")
        
        updatePickerInput(session, 'plotclustersSCENIC', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                        "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                        "Regulon"),
                          selected = "Regulon")
        
        if (!is.null(seuratreactive$TAselectedcells)) {
          updatePickerInput(session, 'plotclustersTA', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                      "Pseudotime"),
                            selected = "Pseudotime")
        }
        else {
          updatePickerInput(session, 'plotclustersTA', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                            selected = "Labelled Cells")
        }
        
        if (is.null(seuratreactive$plot.data$Pseudotime)) {
          updatePickerInput(session, 'plotclustersTA', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                            selected = "Labelled Cells")
        }
        
        updatePickerInput(session, 'plotclustersMM', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                                    "Gene expression"),
                          selected = "Gene expression")
        
        updatePickerInput(session, 'plotclustersDE', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Differential Expression"),
                          selected = "Sample")
        
        updatePickerInput(session, 'plotclustersSDE', choices = list("Labelled Cells", "Seurat Clusters", "Sample", "Cell Cycle Phase", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                     "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)"),
                          selected = "Labelled Cells")
        
        if (!is.null(seuratreactive$DE_clusters)) {
          updatePickerInput(session, 'DE_clusters', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$DE_clusters)
        }
        else {
          updatePickerInput(session, 'DE_clusters', choices = list("Seurat Clusters", "Labelled Cells"))
        }
        
        if (!is.null(seuratreactive$DE_clusters_samples)) {
          updatePickerInput(session, 'DE_clusters_samples', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$DE_clusters_samples)
        }
        else {
          updatePickerInput(session, 'DE_clusters_samples', choices = list("Seurat Clusters", "Labelled Cells"))
        }
        
        if (!is.null(seuratreactive$MAGMA_selectcluster)) {
          updatePickerInput(session, 'MAGMA_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$MAGMA_selectcluster)
        }
        else {
          updatePickerInput(session, 'MAGMA_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"))
        }
        
        if (!is.null(seuratreactive$MAGMA_selectcluster2)) {
          updatePickerInput(session, 'MAGMA_selectcluster2', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$MAGMA_selectcluster2)
        }
        else {
          updatePickerInput(session, 'MAGMA_selectcluster2', choices = list("Seurat Clusters", "Labelled Cells"))
        }
      }
    }
  })
  
  
  observeEvent(input$LabelClustersTextButton, {
    tryCatch({
      seuratreactive$obj <- SetIdent(seuratreactive$obj, value = "seurat_clusters")
      
      seuratreactive$new.cell.ids <- character(length(levels(seuratreactive$obj$seurat_clusters)))
      for (i in 1:length(levels(seuratreactive$obj$seurat_clusters))) {
        seuratreactive$new.cell.ids[i] <- eval(parse(text = paste("input$seuratlabel", levels(seuratreactive$obj$seurat_clusters)[i], sep = "")))
        updateTextInput(session, paste("seuratlabel", levels(seuratreactive$obj$seurat_clusters)[i], sep = ""),
                        value = seuratreactive$new.cell.ids[i])
      }
      
      names(seuratreactive$new.cell.ids) <- levels(seuratreactive$obj)
      seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
      
      #Merge identifiers
      IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
      
      IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
      
      colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
      
      seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
      
      rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
      
      seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
      
      seuratreactive$reference_annotation <- "custom"
      
      seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
      colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
      
      shinyjs::hide("NextButtonCC")
      
    },
    error = function(e) {
      shinyalert("ERROR!", "Manual labelling of clusters failed, please if check your labels are inputted correctly.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  })
  
  
  observeEvent(input$ReferenceButtonImmune, {
    tryCatch({
      withProgress(message = "Labelling clusters. Please wait...", 
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$obj <- SetIdent(seuratreactive$obj, value = "seurat_clusters")
                     
                     if (input$select2 == "Blueprint Encode"){
                       ref.se <- BlueprintEncodeData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     else if (input$select2 == "DICE Immune Cell Expression Data") {
                       ref.se <- DatabaseImmuneCellExpressionData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     else if (input$select2 == "Immunologic Genome Project") {
                       ref.se <- ImmGenData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     else if (input$select2 == "Monaco Immune Data") {
                       ref.se <- MonacoImmuneData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     else if (input$select2 == "Novershtern Hematopoietic Data") {
                       ref.se <- NovershternHematopoieticData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     seuratreactive$new.cell.ids <- annots$labels
                     names(seuratreactive$new.cell.ids) <- levels(seuratreactive$obj)
                     seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                     
                     #Merge identifiers
                     IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                 "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$reference_annotation <- "singleR database"
                     seuratreactive$select2 <- input$select2
                     
                     seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
                     colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
                     
                     shinyjs::hide("NextButtonCC")
                     
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "SingleR Labelling of clusters failed. Please check if the starting tissue is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$ReferenceButtonBrain, {
    tryCatch({
      withProgress(message = "Labelling clusters. Please wait...", 
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$obj <- SetIdent(seuratreactive$obj, value = "seurat_clusters")
                     
                     if (input$selectBrain == "Darmanis Brain Data") {
                       ref.se <- DarmanisBrainData()
                       
                       ref.se <- ref.se[,!is.na(ref.se$cell.type)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$cell.type, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectBrain == "Zhong Prefrontal Cortex Data") {
                       ref.se <- ZhongPrefrontalData()
                       
                       ref.se <- ref.se[,!is.na(ref.se$cell_types)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$cell_types, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     seuratreactive$new.cell.ids <- annots$labels
                     names(seuratreactive$new.cell.ids) <- levels(seuratreactive$obj)
                     seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                     
                     #Merge identifiers
                     IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                 "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$reference_annotation <- "singleR database"
                     seuratreactive$select2 <- input$selectBrain
                     
                     seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
                     colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
                     
                     shinyjs::hide("NextButtonCC")
                     
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "SingleR Labelling of clusters failed. Please check if the starting tissue is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$ReferenceButtonPancreas, {
    tryCatch({
      withProgress(message = "Labelling clusters. Please wait...", 
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$obj <- SetIdent(seuratreactive$obj, value = "seurat_clusters")
                     
                     if (input$selectPancreas == "Baron Mouse Pancreas Data") {
                       ref.se <- BaronPancreasData("mouse")
                       
                       ref.se <- ref.se[,!is.na(ref.se$label)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectPancreas == "Baron Pancreas Data") {
                       ref.se <- BaronPancreasData()
                       
                       ref.se <- ref.se[,!is.na(ref.se$label)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectPancreas == "Lawlor Pancreas Data") {
                       ref.se <- LawlorPancreasData()
                       
                       ref.se <- ref.se[,!is.na(ref.se$`cell type`)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$`cell type`, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectPancreas == "Muraro Pancreas Data") {
                       ref.se <- MuraroPancreasData()
                       
                       ref.se <- ref.se[,!is.na(ref.se$label)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectPancreas == "Segerstolpe Pancreas Data") {
                       ref.se <- SegerstolpePancreasData()
                       
                       ref.se <- ref.se[,!is.na(ref.se$`cell type`)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$`cell type`, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     seuratreactive$new.cell.ids <- annots$labels
                     names(seuratreactive$new.cell.ids) <- levels(seuratreactive$obj)
                     seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                     
                     #Merge identifiers
                     IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                 "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$reference_annotation <- "singleR database"
                     seuratreactive$select2 <- input$selectPancreas
                     
                     seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
                     colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
                     
                     shinyjs::hide("NextButtonCC")
                     
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "SingleR Labelling of clusters failed. Please check if the starting tissue is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$ReferenceButtonMultiple, {
    tryCatch({
      withProgress(message = "Labelling clusters. Please wait...", 
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$obj <- SetIdent(seuratreactive$obj, value = "seurat_clusters")
                     
                     if (input$selectMultiple == "Human Primary Cell Atlas") {
                       ref.se <- HumanPrimaryCellAtlasData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectMultiple == "Mouse RNA-seq") {
                       ref.se <- MouseRNAseqData()
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$label.main, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     else if (input$selectMultiple == "He Organ Atlas") {
                       ref.se <- HeOrganAtlasData(tissue = c(input$selectHe))
                       
                       ref.se <- ref.se[,!is.na(ref.se$Cell_type_in_each_tissue)]
                       
                       ref.se <- logNormCounts(ref.se)
                       
                       incProgress(5/10)
                       
                       annots <- SingleR(GetAssayData(seuratreactive$obj), ref = ref.se,
                                         clusters = Idents(seuratreactive$obj),
                                         colData(ref.se)$Cell_type_in_each_tissue, de.method = "wilcox")  
                       incProgress(8/10)
                     }
                     
                     seuratreactive$new.cell.ids <- annots$labels
                     names(seuratreactive$new.cell.ids) <- levels(seuratreactive$obj)
                     seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                     
                     #Merge identifiers
                     IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                 "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$reference_annotation <- "singleR database"
                     seuratreactive$select2 <- input$selectMultiple
                     
                     seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
                     colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
                     
                     shinyjs::hide("NextButtonCC")
                     
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "SingleR Labelling of clusters failed. Please check if the starting tissue is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$ReferenceButtonsctype, {
    tryCatch({
      withProgress(message = "Labelling clusters. Please wait...", 
                   value = 0, {
                     incProgress(1/10)
                     
                     # load gene set preparation function
                     source("gene_sets_prepare.R")
                     # load cell type annotation function
                     source("sctype_score_.R")
                     
                     seuratreactive$selecttissue <- input$selecttissue
                     
                     # prepare gene sets
                     gs_list = gene_sets_prepare(seuratreactive$selecttissue)
                     
                     
                     # get cell-type by cell matrix
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       es.max = sctype_score(scRNAseqData = seuratreactive$obj[["RNA"]]@scale.data, scaled = TRUE, 
                                             gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
                     }
                     else {
                       es.max = sctype_score(scRNAseqData = seuratreactive$obj[["SCT"]]@scale.data, scaled = TRUE, 
                                             gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
                     }
                     
                     incProgress(5/10)
                     
                     # merge by cluster
                     cL_results = do.call("rbind", lapply(unique(seuratreactive$obj@meta.data$seurat_clusters), function(cl){
                       es.max.cl = sort(rowSums(es.max[ ,rownames(seuratreactive$obj@meta.data[seuratreactive$obj@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
                       head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(seuratreactive$obj@meta.data$seurat_clusters==cl)), 10)
                     }))
                     sctype_scores = cL_results %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  
                     
                     incProgress(8/10)
                     
                     if (input$switchUnknown == TRUE) {
                       # set low-confident (low ScType score) clusters to "unknown"
                       sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = "Unknown"
                     }
                     
                     Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                     
                     seuratreactive$new.cell.ids <- sctype_scores$type
                     names(seuratreactive$new.cell.ids) <- sctype_scores$cluster
                     seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                     
                     
                     #Merge identifiers
                     IdentifiersL <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersL[c("percent.ribo", "percent.mt")] <- signif(IdentifiersL[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     colnames(IdentifiersL) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                 "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)", "Cell Cycle Phase", "Labelled Cells")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersL, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$reference_annotation <- "sctype database"
                     
                     seuratreactive$Identifiers <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "ident"))
                     colnames(seuratreactive$Identifiers) <- c("Seurat Clusters", "Sample", "Labelled Cells")
                     
                     shinyjs::hide("NextButtonCC")
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "sctype Labelling of clusters failed.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$plot9UI <- renderUI({
    plotlyOutput("plot9", height = (400 + input$Lplotheight*40))
  })
  
  output$plot9 <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    if(input$plotclustersL == "Seurat Clusters" | input$plotclustersL == "Cell Cycle Phase" | input$plotclustersL == "Labelled Cells" |
       input$plotclustersL == "Sample") {
      if(input$radioL == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
                
        ) %>% layout(font = list(size = input$labelsizeL))
      }
      else if (input$radioL == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
      else if (input$radioL == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
      else if (input$radioL == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
    }
    else {
      if(input$radioL == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
      else if (input$radioL == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
      else if (input$radioL == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
      else if (input$radioL == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersL],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeL),
                opacity = input$opacityL,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeL))
      }
    }
  })
  
  current_selectionL <- reactiveVal(NULL)
  
  
  observe({
    if (!is.null(seuratreactive$reductionCC1)) {
      if (input$LC_clusters == "Seurat Clusters"){
        updateVirtualSelect('refInput', choices = c(unique(levels(seuratreactive$obj$seurat_clusters))), selected = unique(levels(seuratreactive$obj$seurat_clusters))[1])
      }
      else {
        updateVirtualSelect('refInput', choices = c(unique(levels(seuratreactive$obj))), selected = unique(levels(seuratreactive$obj))[1])
      }
    }
  })
  
  observe({
    if (is.null(input$refInput) |
        is.na(input$minL) |
        is.na(input$log2FCL)) {
      shinyjs::disable("CMG")
    }
    else {
      shinyjs::enable("CMG")
    }
  })
  
  observeEvent(input$CMG, {
    library(igraph)
    library(ggraph)
    tryCatch({
      withProgress(message = 'Finding cell markers...',
                   value = 0, {
                     incProgress(1/10)
                     if (input$LC_clusters == "Seurat Clusters") {
                       
                       Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       markers <- FindAllMarkers(DEtestobj, group.by = "seurat_clusters",
                                                 min.pct = input$minL, logfc.threshold = input$log2FCL,
                                                 test.use = input$testuseL)
                       
                       markers$p_val_adj = p.adjust(markers$p_val, method=input$padjustL)
                       
                       incProgress(5/10)
                       
                       markers[,c(1,2,3,5)] <- signif(markers[,c(1,2,3,5)], digits = 4)
                       seuratreactive$markers <- markers[,c(7,2,3,1,5,6)]
                       colnames(seuratreactive$markers) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "p-value", "Adjusted p-value", "Cluster")
                       
                       seuratreactive$markers <- seuratreactive$markers[seuratreactive$markers$`Adjusted p-value` <= input$pvalL,]
                       
                       if (!is.null(seuratreactive$new.cell.ids)) {
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                       }
                       
                       incProgress(10/10)
                     }
                     
                     else if (input$LC_clusters == "Labelled Cells") {
                       
                       Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                       
                       seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       markers <- FindAllMarkers(DEtestobj,
                                                 min.pct = input$minL, logfc.threshold = input$log2FCL,
                                                 test.use = input$testuseL)
                       
                       markers$p_val_adj = p.adjust(markers$p_val, method=input$padjustL)
                       
                       
                       incProgress(5/10)
                       
                       markers[,c(1,2,3,5)] <- signif(markers[,c(1,2,3,5)], digits = 4)
                       seuratreactive$markers <- markers[,c(7,2,3,1,5,6)]
                       colnames(seuratreactive$markers) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "p-value", "Adjusted p-value", "Cluster")
                       
                       seuratreactive$markers <- seuratreactive$markers[seuratreactive$markers$`Adjusted p-value` <= input$pvalL,]
                       
                       incProgress(10/10)
                       
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if you've inputted all parameters. Please check if you have selected the correct test. Please also check your starting species is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  observe({
    if (!is.null(seuratreactive$markers)) {
      shinyjs::show("refInput")
    }
    else {
      shinyjs::hide("refInput")
    }
  })
  
  
  output$markers <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$markers), 'Please select your parameters and press APPLY to begin'))
    
    seuratreactive$markers[seuratreactive$markers$Cluster == input$refInput,-6]
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$UporDownUI <- renderUI({
    plotlyOutput("UporDown", height = (300 + input$HeightUporDown*30))
  })
  
  output$UporDown <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$markers), 'Please compute your marker genes'))
    
    summary <- seuratreactive$markers %>%
      group_by(Cluster) %>%
      summarise(sum = sum(`Log2 fold change` > 0), n = n()) %>% 
      mutate(Ratio = signif((sum/n) - 0.5, 4)) %>% 
      mutate(`Expression Change`= ifelse(Ratio>0,"Upregulated","Downregulated"))
    
    ggplot(summary, aes(fill=`Expression Change`, y=Ratio, x=Cluster)) + 
      geom_bar(position="stack", stat="identity") + 
      ylab("Upregulated markers/total markers") + 
      scale_y_continuous(limits = c(-0.5, 0.5), labels = function(y) y + 0.5) +
      scale_fill_manual(values=c("blue","red"), name="") +
      theme_bw() + 
      theme(axis.text.x = element_text(angle = 90, vjust=0.5),
            text = element_text(size=input$labelsizeUporDown))
  })
  
  output$MarkersMapUI <- renderUI({
    plotOutput("MarkersMap", height = (300 + input$Heightmarkers*30))
  })
  
  output$MarkersMap <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$markers), 'Please compute your marker genes'))
    
    markersmap <- seuratreactive$markers %>% 
      group_by(`Cluster`) %>% 
      top_n(n = input$MGNumber, wt = `Log2 fold change`)
    
    edges <- markersmap[,c("Cluster", "Gene Name", "Log2 fold change")]
    
    nodes1 <- markersmap %>%
      group_by(`Gene Name`) %>%
      summarise(node_color = "gene", n = 2) %>% 
      drop_na()
    
    nodes2 <- markersmap %>%
      group_by(Cluster) %>%
      summarise(node_color = "cluster", n = 2.1) %>% 
      drop_na()
    
    colnames(nodes2)[1] <- c("Gene Name")
    
    nodes <- full_join(nodes1, nodes2)
    
    graph = graph_from_data_frame(edges, directed = FALSE, nodes)
    
    if (input$checkboxgene_markers == TRUE && input$checkboxcluster_markers == TRUE) {
      ggraph(graph, layout = "fr") + 
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodemarkers, input$nodemarkers/2)) + 
        geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontmarkers, repel=TRUE) +
        geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontmarkers, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() + 
        guides(edge_width = FALSE, color = FALSE)
    }
    
    else if (input$checkboxgene_markers == FALSE && input$checkboxcluster_markers == TRUE) {
      ggraph(graph, layout = "fr") + 
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodemarkers, input$nodemarkers/2)) + 
        geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontmarkers, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() + 
        guides(edge_width = FALSE, color = FALSE)
    }
    
    else if (input$checkboxgene_markers == TRUE && input$checkboxcluster_markers == FALSE) {
      ggraph(graph, layout = "fr") + 
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodemarkers, input$nodemarkers/2)) + 
        geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontmarkers, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() + 
        guides(edge_width = FALSE, color = FALSE)
    }
  })
  
  
  ####TAB 7 - Gene Expression ####
  
  observe({
    if (is.null(seuratreactive$summary.output) && is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && !is.null(seuratreactive$objGEyes) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = T),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = T),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = T),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observeEvent(input$startbuttonGE, {
    if (input$iscollapsebox5 == FALSE) {
      js$collapse("GEBoxIntro")
    }
    shinyjs::show("GEBox1")
    shinyjs::show("GEBox2")
    shinyjs::show("GEBox3")
    shinyjs::show("GEBox4")
    
    seuratreactive$GEInput <- NULL
    seuratreactive$GSEInput <- NULL
    seuratreactive$p <- NULL
    seuratreactive$q <- NULL
    
    seuratreactive$plot.data$`Gene Expression` <- NULL
    
    seuratreactive$objGEyes <- "yes"
    
    updateVirtualSelect('GEInput', choices = c(rownames(seuratreactive$obj)))
    
    updateVirtualSelect('GSEInput', choices = c(sample(seuratreactive$msig$gs_name)))
    
  })
  
  observe({
    if (is.null(input$GEInput)) {
      shinyjs::disable("gene1")
    }
    else {
      shinyjs::enable("gene1")
    }
  })
  
  observe({
    if (is.null(input$GSEInput) | !is.null(input$GSEInput) && input$GSEInput == "") {
      shinyjs::disable("gene2")
    }
    else {
      shinyjs::enable("gene2")
    }
  })
  
  observe({
    if (input$GEDotInput == "" | input$GEDotInput == "Cell Cycle Phase") {
      shinyjs::hide("refineGEDOT")
    }
    else {
      shinyjs::show("refineGEDOT")
    }
  })
  
  observe({
    if (input$VlnInput == "" | input$VlnInput == "Cell Cycle Phase") {
      shinyjs::hide("refineGEVLN")
    }
    else {
      shinyjs::show("refineGEVLN")
    }
  })
  
  observe({
    if (input$checkboxGEDot == FALSE) {
      shinyjs::hide("refineGEDOT_SAMPLES")
    }
    else {
      shinyjs::show("refineGEDOT_SAMPLES")
    }
  })
  
  observe({
    if (input$checkboxvln == FALSE) {
      shinyjs::hide("refineGEVLN_SAMPLES")
    }
    else {
      shinyjs::show("refineGEVLN_SAMPLES")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) { 
      updateVirtualSelect('refineGEDOT_SAMPLES', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) { 
      updateVirtualSelect('refineGEVLN_SAMPLES', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))))
    }
  })
  
  observeEvent(input$gene1, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     mean <- FetchData(object = seuratreactive$obj, vars = input$GEInput) 
                     mean <- mutate(mean, Average = rowMeans(mean))
                     incProgress(3/10)
                     
                     seuratreactive$obj <- AddMetaData(object=seuratreactive$obj, metadata = mean$Average, col.name = "Gene_Expression")
                     
                     IdentifiersGE <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                      "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident", "Gene_Expression"))
                     
                     IdentifiersGE[c("percent.ribo", "percent.mt", "Gene_Expression")] <- signif(IdentifiersGE[c("percent.ribo", "percent.mt", "Gene_Expression")], digits = 4)
                     
                     colnames(IdentifiersGE) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                  "Cell Cycle Phase", "Labelled Cells", "Gene Expression")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersGE, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     incProgress(8/10)
                     
                     seuratreactive$titleGE <- paste(input$GEInput, collapse = " ")
                     
                     seuratreactive$GEInput <- input$GEInput
                     
                     seuratreactive$objVLN <- NULL
                     
                     seuratreactive$objGEDOT <- NULL
                     
                     incProgress(10/10)
                     
                     seuratreactive$Gene_vector_msig <- NULL
                     seuratreactive$GSEInput <- NULL
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please select a gene (or genes) to view its expression.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  })
  
  
  observeEvent(input$gene2, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     Gene_vector_msig <- seuratreactive$msig[match(input$GSEInput, seuratreactive$msig$gs_name),]
                     
                     incProgress(3/10)
                     
                     seuratreactive$Gene_vector_msig <- Gene_vector_msig[,-1] %>% as.character() %>% na.omit() %>% unique()
                     
                     mean <- FetchData(object = seuratreactive$obj, vars = unique(seuratreactive$Gene_vector_msig))
                     
                     mean <- mutate(mean, Average = rowMeans(mean))
                     incProgress(5/10)
                     
                     seuratreactive$obj <- AddMetaData(object=seuratreactive$obj, metadata = mean$Average, col.name = "Gene_Expression")
                     
                     IdentifiersGE <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                      "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident",
                                                                                      "Gene_Expression"))
                     
                     IdentifiersGE[c("percent.ribo", "percent.mt", "Gene_Expression")] <- signif(IdentifiersGE[c("percent.ribo", "percent.mt", "Gene_Expression")], digits = 4)
                     
                     incProgress(8/10)
                     
                     colnames(IdentifiersGE) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                  "Cell Cycle Phase", "Labelled Cells", "Gene Expression")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersGE, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     seuratreactive$titleGE <- paste(input$GSEInput, collapse = " ")
                     
                     seuratreactive$GSEInput <- input$GSEInput
                     
                     seuratreactive$objVLN <- NULL
                     
                     seuratreactive$objGEDOT <- NULL
                     
                     seuratreactive$GEInput <- NULL
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Gene in gene pathway not found in your dataset. Please also check if your starting species is correct.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  })
  
  output$Genepathwaytext <- renderUI({
    vector_msig <- ifelse (is.null(seuratreactive$GSEInput),
                           paste(""),
                           paste("<b>Gene pathway: </b>", paste(seuratreactive$GSEInput), "<br>",
                                 "Genes:", paste(strwrap(unique(seuratreactive$Gene_vector_msig), 10), collapse = ", "))
    )
    
    HTML(vector_msig)
  })
  
  output$plotGEUI <- renderUI({
    plotlyOutput("plotGE", height = (400 + input$GEplotheight*40))
  })
  
  output$plotGE <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data$`Gene Expression`), 'You must select a gene first to view its expression'))
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    if(input$plotclustersGE == "Gene Expression"){
      if(input$radioGE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(title = seuratreactive$titleGE, font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(title = seuratreactive$titleGE, font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
                
        ) %>% layout(title = seuratreactive$titleGE, font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(title = seuratreactive$titleGE, font = list(size = input$labelsizeGE))
      }
    }
    else if(input$plotclustersGE == "Seurat Clusters" | input$plotclustersGE == "Cell Cycle Phase" | input$plotclustersGE == "Labelled Cells" |
            input$plotclustersGE == "Sample") {  
      if(input$radioGE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
      }
      else if (input$radioGE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
        
      }
    }
    else {
      if(input$radioGE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
        
      }
      else if (input$radioGE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.data[, input$plotclustersGE],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeGE),
                opacity = input$opacityGE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.data$`Gene Expression`)
        ) %>% layout(font = list(size = input$labelsizeGE))
      }
    }
  })
  
  observeEvent(input$GEDOTGO2, {
    if (!is.null(seuratreactive$obj) && input$VlnInput == "Seurat Clusters" & input$checkboxvln == TRUE) {
      seuratreactive$objVLN <- seuratreactive$obj
      
      Idents(seuratreactive$objVLN) <- seuratreactive$objVLN@meta.data$seurat_clusters
      
      if (!is.null(input$refineGEVLN)) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, idents = input$refineGEVLN)
      }
      
      if (!is.null(input$refineGEVLN_SAMPLES) && input$checkboxvln == TRUE) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, subset = orig.ident == input$refineGEVLN_SAMPLES)
      }
      
      Idents(seuratreactive$objVLN) <- paste(Idents(seuratreactive$objVLN), seuratreactive$objVLN@meta.data$orig.ident, sep = "_")
    }
    
    else if (!is.null(seuratreactive$obj) && input$VlnInput == "Labelled Cells" & input$checkboxvln == TRUE) {
      seuratreactive$objVLN <- seuratreactive$obj
      
      if (!is.null(input$refineGEVLN)) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, idents = input$refineGEVLN)
      }
      
      if (!is.null(input$refineGEVLN_SAMPLES) && input$checkboxvln == TRUE) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, subset = orig.ident == input$refineGEVLN_SAMPLES)
      }
      
      Idents(seuratreactive$objVLN) <- paste(Idents(seuratreactive$objVLN), seuratreactive$objVLN@meta.data$orig.ident, sep = "_")
    }
    
    else if (!is.null(seuratreactive$obj) && input$VlnInput == "Cell Cycle Phase" & input$checkboxvln == TRUE) {
      seuratreactive$objVLN <- seuratreactive$obj
      
      Idents(seuratreactive$objVLN) <- seuratreactive$objVLN@meta.data$Phase
      
      if (!is.null(input$refineGEVLN)) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, idents = input$refineGEVLN)
      }
      
      if (!is.null(input$refineGEVLN_SAMPLES) && input$checkboxvln == TRUE) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, subset = orig.ident == input$refineGEVLN_SAMPLES)
      }
      
      Idents(seuratreactive$objVLN) <- paste(Idents(seuratreactive$objVLN), seuratreactive$objVLN@meta.data$orig.ident, sep = "_")
    }
    
    else if (!is.null(seuratreactive$obj) && input$VlnInput == "Seurat Clusters" & input$checkboxvln == FALSE) {
      seuratreactive$objVLN <- seuratreactive$obj
      
      Idents(seuratreactive$objVLN) <- seuratreactive$objVLN@meta.data$seurat_clusters
      
      if (!is.null(input$refineGEVLN)) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, idents = input$refineGEVLN)
      }
      
      if (!is.null(input$refineGEVLN_SAMPLES) && input$checkboxvln == TRUE) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, subset = orig.ident == input$refineGEVLN_SAMPLES)
      }
    }
    
    else if (!is.null(seuratreactive$obj) && input$VlnInput == "Labelled Cells" & input$checkboxvln == FALSE) {
      seuratreactive$objVLN <- seuratreactive$obj
      
      if (!is.null(input$refineGEVLN)) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, idents = input$refineGEVLN)
      }
      
      if (!is.null(input$refineGEVLN_SAMPLES) && input$checkboxvln == TRUE) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, subset = orig.ident == input$refineGEVLN_SAMPLES)
      }
    }
    
    else if (!is.null(seuratreactive$obj) && input$VlnInput == "Cell Cycle Phase" & input$checkboxvln == FALSE) {
      seuratreactive$objVLN <- seuratreactive$obj
      
      Idents(seuratreactive$objVLN) <- seuratreactive$objVLN@meta.data$Phase
      
      if (!is.null(input$refineGEVLN)) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, idents = input$refineGEVLN)
      }
      
      if (!is.null(input$refineGEVLN_SAMPLES) && input$checkboxvln == TRUE) {
        seuratreactive$objVLN <- subset(seuratreactive$objVLN, subset = orig.ident == input$refineGEVLN_SAMPLES)
      }
    }
  })
  
  output$VlnUI <- renderUI({
    plotOutput("Vln", height = (300 + input$HeightGEVln*30))
  })
  
  output$Vln <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data$`Gene Expression`), 'You must select a gene first to view its expression'))
    validate(
      need(!is.null(seuratreactive$objVLN), 'Please select your parameters and press APPLY to begin'))
    
    if (input$mergedGE == FALSE) {
      if (!is.null(seuratreactive$GEInput)) {
        p <- VlnPlot(seuratreactive$objVLN, features = seuratreactive$GEInput, pt.size = input$Vlnpt, ncol = 1, combine = F) 
        
        p1 <- list() 
        for (i in seq_along(p)){
          p1[[i]] = p[[i]] + theme(text = element_text(size=input$VlnsizeGE),
                                   axis.text.x = element_text(size = input$VlnsizeGE), 
                                   axis.text.y = element_text(size = input$VlnsizeGE)) 
        }
        
        print(plot_grid(plotlist = p1, ncol = 1))

        seuratreactive$p <- plot_grid(plotlist = p1, ncol = 1)
      }
      
      else if (!is.null(seuratreactive$GSEInput)) {
        p <- VlnPlot(seuratreactive$objVLN, features = intersect(unique(seuratreactive$Gene_vector_msig), rownames(seuratreactive$objVLN)), pt.size = input$Vlnpt, ncol = 1, combine = F) 
        
        p1 <- list() 
        for (i in seq_along(p)){
          p1[[i]] = p[[i]] + theme(text = element_text(size=input$VlnsizeGE),
                                   axis.text.x = element_text(size = input$VlnsizeGE), 
                                   axis.text.y = element_text(size = input$VlnsizeGE)) 
        }
        
        print(plot_grid(plotlist = p1, ncol = 1))
        
        seuratreactive$p <- plot_grid(plotlist = p1, ncol = 1)
      }
    }
    else {
      if (!is.null(seuratreactive$GEInput) | !is.null(seuratreactive$GSEInput)) {
        p <- VlnPlot(seuratreactive$objVLN, features = "Gene_Expression", pt.size = input$Vlnpt) + 
          theme(text = element_text(size=input$VlnsizeGE),
                                   axis.text.x = element_text(size = input$VlnsizeGE), 
                                   axis.text.y = element_text(size = input$VlnsizeGE)) 
         p[[1]][["labels"]][["title"]] <- seuratreactive$titleGE
        
        seuratreactive$p <- p
        
        print(p)
      }
    }
  }) 
  
  observeEvent(input$GEDOTGO, {
    if (!is.null(seuratreactive$obj) && input$GEDotInput == "Seurat Clusters" & input$checkboxGEDot == TRUE) {
      seuratreactive$objGEDOT <- seuratreactive$obj
      
      Idents(seuratreactive$objGEDOT) <- seuratreactive$objGEDOT@meta.data$seurat_clusters
      
      if (!is.null(input$refineGEDOT)) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, idents = input$refineGEDOT)
      }
      
      if (!is.null(input$refineGEDOT_SAMPLES) && input$checkboxGEDot == TRUE) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, subset = orig.ident == input$refineGEDOT_SAMPLES)
      }
      
      Idents(seuratreactive$objGEDOT) <- paste(Idents(seuratreactive$objGEDOT), seuratreactive$objGEDOT@meta.data$orig.ident, sep = "_")
    }
    
    else if (!is.null(seuratreactive$obj) && input$GEDotInput == "Labelled Cells" & input$checkboxGEDot == TRUE) {
      seuratreactive$objGEDOT <- seuratreactive$obj
      
      if (!is.null(input$refineGEDOT)) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, idents = input$refineGEDOT)
      }
      
      if (!is.null(input$refineGEDOT_SAMPLES) && input$checkboxGEDot == TRUE) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, subset = orig.ident == input$refineGEDOT_SAMPLES)
      }
      
      Idents(seuratreactive$objGEDOT) <- paste(Idents(seuratreactive$objGEDOT), seuratreactive$objGEDOT@meta.data$orig.ident, sep = "_")
    }
    
    else if (!is.null(seuratreactive$obj) && input$GEDotInput == "Cell Cycle Phase" & input$checkboxGEDot == TRUE) {
      seuratreactive$objGEDOT <- seuratreactive$obj
      
      Idents(seuratreactive$objGEDOT) <- seuratreactive$objGEDOT@meta.data$Phase
      
      if (!is.null(input$refineGEDOT)) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, idents = input$refineGEDOT)
      }
      
      if (!is.null(input$refineGEDOT_SAMPLES) && input$checkboxGEDot == TRUE) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, subset = orig.ident == input$refineGEDOT_SAMPLES)
      }
      
      Idents(seuratreactive$objGEDOT) <- paste(Idents(seuratreactive$objGEDOT), seuratreactive$objGEDOT@meta.data$orig.ident, sep = "_")
    }
    
    else if (!is.null(seuratreactive$obj) && input$GEDotInput == "Seurat Clusters" & input$checkboxGEDot == FALSE) {
      seuratreactive$objGEDOT <- seuratreactive$obj
      
      Idents(seuratreactive$objGEDOT) <- seuratreactive$objGEDOT@meta.data$seurat_clusters
      
      if (!is.null(input$refineGEDOT)) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, idents = c(input$refineGEDOT))
      }
      
      if (!is.null(input$refineGEDOT_SAMPLES) && input$checkboxGEDot == TRUE) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, subset = orig.ident == input$refineGEDOT_SAMPLES)
      }
    }
    
    else if (!is.null(seuratreactive$obj) && input$GEDotInput == "Labelled Cells" & input$checkboxGEDot == FALSE) {
      seuratreactive$objGEDOT <- seuratreactive$obj
      
      if (!is.null(input$refineGEDOT)) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, idents = input$refineGEDOT)
      }
      
      if (!is.null(input$refineGEDOT_SAMPLES) && input$checkboxGEDot == TRUE) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, subset = orig.ident == input$refineGEDOT_SAMPLES)
      }
    }
    
    else if (!is.null(seuratreactive$obj) && input$GEDotInput == "Cell Cycle Phase" & input$checkboxGEDot == FALSE) {
      seuratreactive$objGEDOT <- seuratreactive$obj
      
      Idents(seuratreactive$objGEDOT) <- seuratreactive$objGEDOT@meta.data$Phase
      
      if (!is.null(input$refineGEDOT)) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, idents = input$refineGEDOT)
      }
      
      if (!is.null(input$refineGEDOT_SAMPLES) && input$checkboxGEDot == TRUE) {
        seuratreactive$objGEDOT <- subset(seuratreactive$objGEDOT, subset = orig.ident == input$refineGEDOT_SAMPLES)
      }
    }
  })
  
  
  output$GEDotUI <- renderUI({
    plotOutput("GEDot", height = (300 + input$HeightGEDot*30))
  })
  
  output$GEDot <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data$`Gene Expression`), 'You must select a gene first to view its expression'))
    validate(
      need(!is.null(seuratreactive$objGEDOT), 'Please select your parameters and press APPLY to begin'))
    
    if (!is.null(seuratreactive$GEInput)) {
      p <- DotPlot(seuratreactive$objGEDOT, features = seuratreactive$GEInput, dot.scale = input$dotscale) +
        RotatedAxis() +
        scale_colour_gradient(low = "blue", high = "red") +
        guides(color = guide_colorbar(title = 'Average Expression')) +
        theme(text = element_text(size=input$GEDotSize),
              axis.text.x= element_text(size = input$GEDotSize),
              axis.text.y= element_text(size = input$GEDotSize))
      seuratreactive$q <- p
      p
    }
    
    else if (!is.null(seuratreactive$GSEInput)) {
      p <- DotPlot(seuratreactive$objGEDOT, features = unique(seuratreactive$Gene_vector_msig), dot.scale = input$dotscale) +
        RotatedAxis() +
        scale_colour_gradient(low = "blue", high = "red") +
        guides(color = guide_colorbar(title = 'Average Expression')) +
        theme(text = element_text(size=input$GEDotSize),
              axis.text.x= element_text(size = input$GEDotSize),
              axis.text.y= element_text(size = input$GEDotSize))
      p[["labels"]][["title"]] <- seuratreactive$titleGE
      seuratreactive$q <- p
      p 
    } 
  }) 
  
  ####TAB 8 - MEGENA####
  
  observe({
    if (!is.null(seuratreactive$summary.output) && is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$regulons) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = T),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = T),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = T),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observeEvent(input$startMEGENA, {
    showModal(modalDialog(
      title = strong("MEGENA Settings"),
      pickerInput("MEGENA_selectmethod", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select input method')),
                  choices = list("Variable Genes", "Differentially Expressed Genes", "Custom Differentially Expressed Genes"),
                  width = "100%"),
      numericInput("variablegenesMEGENA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Pick number of variable genes:')),
                   min = 1,
                   value = 500, step = 50, 
                   width = "100%"),
      pickerInput("MEGENA_selectcluster", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by')),
                  choices = list("Seurat Clusters", "Labelled Cells"),
                  width = "100%"),
      virtualSelectInput("MEGENA_samples", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sample(s)')), 
                         choices = NULL, 
                         multiple = T,
                         hideClearButton = F,
                         search = T,
                         disableSelectAll = F,
                         width = "100%"),
      h4(HTML('<h4 style = "text-align:justify;color:#000000;"> <b>Co-expression Module Identification Settings </b><br>')),
      numericInput("MEGENApval", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                   min = 0, max = 1,
                   value = 0.05,
                   width = "100%"),
      numericInput("MEGENAmin.size", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Minimum module size')),
                   min = 2, max = 100,
                   value = 10, 
                   width = "100%"),
      numericInput("n_MEGENA", h5(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Downsample MEGENA module construction to X cells per cluster/cell type (randomly selected)')),
                   min = 1,
                   value = 500, step = 50, 
                   width = "100%"),
      p(style="text-align: center;", actionBttn("startbuttonMEGENA", "APPLY", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startbuttonMEGENA,{
    library(MEGENA)
    library(DGCA)
    library(ggplot2)
    library(ggraph)
    library(GOstats)
    library(HGNChelper)
    library(reshape)
    source("MEGENA.ModuleSummary.R")
    
    tryCatch({
      withProgress(message = "Calculating Co-Expression modules. Please wait...",
                   value = 0, {
                     incProgress(1/20)
                     
                     removeModal()
                     
                     if (input$MEGENA_selectmethod == "Variable Genes") {
                       #Find variable features
                       
                       seuratreactive$VF <- FindVariableFeatures(seuratreactive$obj, nfeatures = input$variablegenesMEGENA)
                       
                       seuratreactive$VF <- VariableFeatures(seuratreactive$VF)
                     }
                     else if (input$MEGENA_selectmethod == "Differentially Expressed Genes") {
                       seuratreactive$VF <- seuratreactive$markersDE$`Gene Name`
                     }
                     else if (input$MEGENA_selectmethod == "Custom Differentially Expressed Genes") {
                       seuratreactive$VF <- seuratreactive$markersSDE$`Gene Name`
                     }
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       exprMat <- as.matrix(seuratreactive$obj@assays$RNA@data)
                     }
                     else {
                       exprMat <- as.matrix(seuratreactive$obj@assays$SCT@data)
                     }
                     
                     exprMat <- exprMat[rownames(exprMat) %in% seuratreactive$VF, ]
                     
                     incProgress(3/20)
                     
                     #Limit to cell cluster and sample
                     cellInfo <- FetchData(object = seuratreactive$obj, vars = c("orig.ident", "ident", "seurat_clusters"))
                     cellInfo <- cellInfo[cellInfo$orig.ident %in% input$MEGENA_samples, ]
                     
                     cellInfo <- tibble::rownames_to_column(cellInfo, "rownames")
                     
                     if (input$MEGENA_selectcluster == "Seurat Clusters") {
                       cellInfo <- cellInfo %>% mutate(sample(100000, size = nrow(cellInfo), replace = TRUE)) %>% 
                         group_by(seurat_clusters) %>% 
                         top_n(input$n_MEGENA)
                     }
                     else {
                       cellInfo <- cellInfo %>% mutate(sample(100000, size = nrow(cellInfo), replace = TRUE)) %>% 
                         group_by(ident) %>% 
                         top_n(input$n_MEGENA)
                     }
                     
                     cellInfo <- cellInfo %>% remove_rownames %>% column_to_rownames(var="rownames")
                     
                     incProgress(4/20)
                     
                     #subset out exprMat
                     exprMat <- exprMat[, colnames(exprMat) %in% rownames(cellInfo)]
                     exprMat <- exprMat[rowSums(exprMat)>0,]
                     
                     rho.out = calculate.rho.signed(exprMat,n.perm = 10,FDR.cutoff = input$MEGENApval,estimator = "pearson",
                                                    use.obs = "na.or.complete",
                                                    direction = "absolute",
                                                    rho.thresh = NULL,sort.el = TRUE)
                     
                     incProgress(8/20)
                     
                     if (length(rho.out[["signif.ijw"]][["row"]]) == 0) {
                       shinyalert("ERROR!", "Not enough cells to generate co-expression modules. Please include more cell clusters in your analysis. ", type = "error", confirmButtonCol = "#337ab7")
                       
                     }
                     else {
                       
                       n.cores <- 10; # number of cores/threads to call for PCP
                       doPar <-TRUE; # do we want to parallelize?
                       
                       run.par = doPar & (getDoParWorkers() == 1) 
                       if (run.par)
                       {
                         cl <- parallel::makeCluster(n.cores)
                         registerDoParallel(cl)
                         # check how many workers are there
                         cat(paste("number of cores to use:",getDoParWorkers(),"\n",sep = ""))
                       }
                       
                       el <- calculate.PFN(rho.out$signif.ijw,doPar = doPar, num.cores = n.cores,keep.track = FALSE)
                       
                       incProgress(11/20)
                       
                       seuratreactive$g <- graph.data.frame(el,directed = FALSE)
                       
                       #perform MCA clustering
                       MEGENA.output <- do.MEGENA(seuratreactive$g,
                                                  mod.pval = input$MEGENApval,hub.pval = input$MEGENApval, remove.unsig = TRUE,
                                                  min.size = input$MEGENAmin.size, max.size = vcount(seuratreactive$g)/2,
                                                  doPar = doPar ,num.cores = n.cores, n.perm = 100,
                                                  save.output = FALSE)
                       
                       incProgress(13/20)
                       
                       seuratreactive$summary.output <- MEGENA.ModuleSummary(MEGENA.output,
                                                                             mod.pvalue = 0.05,hub.pvalue = 0.05,
                                                                             min.size = 10,max.size = vcount(seuratreactive$g)/2,
                                                                             annot.table = NULL,id.col = 1,symbol.col = 2,
                                                                             output.sig = TRUE)
                       
                       if (getDoParWorkers() > 1)
                       {
                         env <- foreach:::.foreachGlobals
                         rm(list=ls(name=env), pos=env)
                       }
                       
                       #FindMarkers
                       if (input$MEGENA_selectcluster == "Seurat Clusters") {
                         
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                             !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                           DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                         }
                         else {
                           DEtestobj <- seuratreactive$obj
                         }
                         
                         seuratreactive$markersMEGENA <- FindAllMarkers(DEtestobj, logfc.threshold = input$log2FCMEGENA, min.pct = input$minMEGENA,
                                                                        test.use = input$testuseMEGENA,
                                                                        group.by = "seurat_clusters", only.pos = T)
                         
                         seuratreactive$markersMEGENA$p_val_adj = p.adjust(seuratreactive$markersMEGENA$p_val, method=input$padjustMEGENA)
                         
                         incProgress(15/20)
                         
                         #Load MEGENA matrix
                         seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[,c(7,2,3,5,6)]
                         colnames(seuratreactive$markersMEGENA) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "Adjusted p-value", "Cluster")
                         
                         #select only significant genes
                         seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$`Adjusted p-value` <= input$pvalMEGENA,]
                         
                         if (!is.null(seuratreactive$new.cell.ids)) {
                           Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                           
                           seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                         }
                         
                       }
                       else if (input$MEGENA_selectcluster == "Labelled Cells") {
                         
                         if (!is.null(seuratreactive$new.cell.ids)) {
                           Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                           
                           seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                         }
                         
                         if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                             !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                           DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                         }
                         else {
                           DEtestobj <- seuratreactive$obj
                         }
                         
                         seuratreactive$markersMEGENA <- FindAllMarkers(DEtestobj, logfc.threshold = input$log2FCMEGENA, min.pct = input$minMEGENA,
                                                                        test.use = input$testuseMEGENA,
                                                                        only.pos = T)
                         
                         seuratreactive$markersMEGENA$p_val_adj = p.adjust(seuratreactive$markersMEGENA$p_val, method=input$padjustMEGENA)
                         
                         
                         incProgress(15/20)
                         #Load MEGENA matrix
                         seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[,c(7,2,3,5,6)]
                         colnames(seuratreactive$markersMEGENA) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "Adjusted p-value", "Cluster")
                         
                         #select only significant genes
                         seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$`Adjusted p-value` <= input$pvalMEGENA,]
                       }
                       
                       seuratreactive$MEGENA_samples <- input$MEGENA_samples
                       seuratreactive$variablegenesMEGENA <- input$variablegenesMEGENA
                       seuratreactive$MEGENApval <- input$MEGENApval
                       seuratreactive$MEGENAmin.size <- input$MEGENAmin.size
                       seuratreactive$MEGENA_selectcluster <- input$MEGENA_selectcluster
                       seuratreactive$n_MEGENA <- input$n_MEGENA
                       seuratreactive$log2FCMEGENA <- input$log2FCMEGENA
                       seuratreactive$padjustMEGENA <- input$padjustMEGENA
                       seuratreactive$testuseMEGENA <- input$testuseMEGENA
                       seuratreactive$minMEGENA <- input$minMEGENA
                       seuratreactive$pvalMEGENA <- input$pvalMEGENA
                       
                       seuratreactive$MEGENA_selectmethod <- input$MEGENA_selectmethod
                       
                       if (input$iscollapseboxMEGENA == FALSE) {
                         js$collapse("MEGENABoxIntro")
                       }
                       shinyjs::show("MEGENABox1")
                       shinyjs::show("MEGENABox2")
                       shinyjs::show("MEGENABox3")
                       shinyjs::show("MEGENABox4")
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "MEGENA Co-Expression network analysis failed. Try including more genes in the analysis.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$SunMEGENA, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     if (!is.null(input$MEGENA_selectcluster) && input$MEGENA_selectcluster == "Seurat Clusters") {
                       
                       Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       seuratreactive$markersMEGENA <- FindAllMarkers(DEtestobj, logfc.threshold = input$log2FCMEGENA, min.pct = input$minMEGENA,
                                                                      test.use = input$testuseMEGENA,
                                                                      group.by = "seurat_clusters", only.pos = T)
                       
                       seuratreactive$markersMEGENA$p_val_adj = p.adjust(seuratreactive$markersMEGENA$p_val, method=input$padjustMEGENA)
                       
                       incProgress(5/10)
                       
                       #Load MEGENA matrix
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[,c(7,2,3,5,6)]
                       colnames(seuratreactive$markersMEGENA) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "Adjusted p-value", "Cluster")
                       
                       #select only significant genes
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$`Adjusted p-value` <= input$pvalMEGENA,]
                       
                       if (!is.null(seuratreactive$new.cell.ids)) {
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                       }
                     }
                     else if (!is.null(input$MEGENA_selectcluster) && input$MEGENA_selectcluster == "Labelled Cells") {
                       
                       if (!is.null(seuratreactive$new.cell.ids)) {
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                       }
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       seuratreactive$markersMEGENA <- FindAllMarkers(DEtestobj, logfc.threshold = input$log2FCMEGENA, min.pct = input$minMEGENA,
                                                                      test.use = input$testuseMEGENA,
                                                                      only.pos = T)
                       
                       seuratreactive$markersMEGENA$p_val_adj = p.adjust(seuratreactive$markersMEGENA$p_val, method=input$padjustMEGENA)
                       
                       incProgress(5/10)
                       #Load MEGENA matrix
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[,c(7,2,3,5,6)]
                       colnames(seuratreactive$markersMEGENA) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "Adjusted p-value", "Cluster")
                       
                       #select only significant genes
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$`Adjusted p-value` <= input$pvalMEGENA,]
                     }
                     else if (!is.null(seuratreactive$MEGENA_selectcluster) && seuratreactive$MEGENA_selectcluster == "Seurat Clusters") {
                       
                       Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       seuratreactive$markersMEGENA <- FindAllMarkers(DEtestobj, logfc.threshold = input$log2FCMEGENA, min.pct = input$minMEGENA,
                                                                      test.use = input$testuseMEGENA,
                                                                      group.by = "seurat_clusters", only.pos = T)
                       
                       seuratreactive$markersMEGENA$p_val_adj = p.adjust(seuratreactive$markersMEGENA$p_val, method=input$padjustMEGENA)
                       
                       incProgress(5/10)
                       
                       #Load MEGENA matrix
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[,c(7,2,3,5,6)]
                       colnames(seuratreactive$markersMEGENA) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "Adjusted p-value", "Cluster")
                       
                       #select only significant genes
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$`Adjusted p-value` <= input$pvalMEGENA,]
                       
                       if (!is.null(seuratreactive$new.cell.ids)) {
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                       }
                     }
                     else if (!is.null(seuratreactive$MEGENA_selectcluster) && seuratreactive$MEGENA_selectcluster == "Labelled Cells") {
                       
                       if (!is.null(seuratreactive$new.cell.ids)) {
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                       }
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       seuratreactive$markersMEGENA <- FindAllMarkers(DEtestobj, logfc.threshold = input$log2FCMEGENA, min.pct = input$minMEGENA,
                                                                      test.use = input$testuseMEGENA,
                                                                      only.pos = T)
                       
                       seuratreactive$markersMEGENA$p_val_adj = p.adjust(seuratreactive$markersMEGENA$p_val, method=input$padjustMEGENA)
                       
                       incProgress(5/10)
                       #Load MEGENA matrix
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[,c(7,2,3,5,6)]
                       colnames(seuratreactive$markersMEGENA) <- c("Gene Name", "Log2 fold change", "Percentage of cells expressed", "Adjusted p-value", "Cluster")
                       
                       #select only significant genes
                       seuratreactive$markersMEGENA <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$`Adjusted p-value` <= input$pvalMEGENA,]
                     }
                     
                     seuratreactive$log2FCMEGENA <- input$log2FCMEGENA
                     seuratreactive$minMEGENA <- input$minMEGENA
                     seuratreactive$pvalMEGENA <- input$pvalMEGENA
                     seuratreactive$padjustMEGENA <- input$padjustMEGENA
                     seuratreactive$testuseMEGENA <- input$testuseMEGENA
                     
                     incProgress(10/10)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if you've inputted all parameters. Please check if the correct test has been selected. Please also check your starting species is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$GO_MODGO, {
    showModal(modalDialog(
      title = strong("Coexpression Module GO Heatmap Computation"),
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete. <br><br><br>'),
         p(style="text-align: center;", actionBttn("GO_MODGOGO", "OK, I UNDERSTAND", size = "lg"))),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$GO_MODGOGO, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     removeModal()
                     
                     modules <- melt(seuratreactive$summary.output$modules)
                     modules$L1 <- sub("_", ".", modules$L1)
                     
                     incProgress(5/10)
                     if (seuratreactive$orgDb == "org.Hs.eg.db") {
                       moduleGO_res = moduleGO(genes = modules$value, labels = modules$L1, HGNC_clean = T,
                                               gene_ontology = "all",
                                               universe = seuratreactive$VF, pval_GO_cutoff = 1,
                                               annotation = seuratreactive$orgDb)
                     }
                     else {
                       moduleGO_res = moduleGO(genes = modules$value, labels = modules$L1, HGNC_clean = F,
                                               gene_ontology = "all",
                                               universe = seuratreactive$VF, pval_GO_cutoff = 1,
                                               annotation = seuratreactive$orgDb)
                     }
                     seuratreactive$moduleGO_df = extractModuleGO(moduleGO_res, labels = unique(modules$L1))
                     
                     incProgress(10/10)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Construction of Co-Expression Module Gene Ontology Heatmap Failed.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$startMEGENA, {
    updateVirtualSelect('MEGENA_samples', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                        selected = seuratreactive$MEGENA_samples)
    
    updateNumericInput(session, 'variablegenesMEGENA', value = seuratreactive$variablegenesMEGENA)
    updatePickerInput(session, 'MEGENA_selectmethod', selected = seuratreactive$MEGENA_selectmethod)
    updateNumericInput(session, 'MEGENAmin.size', value = seuratreactive$MEGENAmin.size)
    updateNumericInput(session, 'MEGENApval', value = seuratreactive$MEGENApval)
    updateNumericInput(session, 'n_MEGENA', value = seuratreactive$n_MEGENA)
    updatePickerInput(session, 'MEGENA_selectcluster', selected = seuratreactive$MEGENA_selectcluster)
    
    if (!is.null(input$variablegenesMEGENA) && !is.null(input$MEGENAmin.size) && !is.null(input$MEGENApval)) {
      if (is.na(input$variablegenesMEGENA) | is.null(input$MEGENA_samples) | is.null(input$MEGENA_selectcluster) | is.null(input$MEGENA_selectmethod) | is.na(input$MEGENAmin.size) | is.na(input$MEGENApval) | is.na(input$n_MEGENA)) {
        shinyjs::disable("startbuttonMEGENA")
      }
      else {
        shinyjs::enable("startbuttonMEGENA")
      }
    }
    
    if(!is.null(seuratreactive$reductionCC1)) {
      if (identical(seuratreactive$obj@active.ident, seuratreactive$obj$seurat_clusters)) {
        if (!is.null(seuratreactive$MEGENA_selectcluster)) {
          updatePickerInput(session, 'MEGENA_selectcluster', choices = list("Seurat Clusters"), selected = seuratreactive$MEGENA_selectcluster)
        }
        else {
          updatePickerInput(session, 'MEGENA_selectcluster', choices = list("Seurat Clusters"))
        }
      }
      else {
        if (!is.null(seuratreactive$MEGENA_selectcluster)) {
          updatePickerInput(session, 'MEGENA_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$MEGENA_selectcluster)
        }
        else {
          updatePickerInput(session, 'MEGENA_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"))
        }
      }
    }
  })
  
  observe({
    if (!is.null(input$MEGENA_selectmethod)) {
      if (input$MEGENA_selectmethod == "Differentially Expressed Genes" && is.null(seuratreactive$markersDE)) {
        shinyalert("ERROR!", "No differentially expressed genes detected! You must compute your differentially expressed genes first! Refer to the differentially expressed genes tab.", type = "error", confirmButtonCol = "#337ab7")
        updatePickerInput(session, 'MEGENA_selectmethod', selected = "Variable Genes")
        updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      }
      
      if (input$MEGENA_selectmethod == "Differentially Expressed Genes" && !is.null(seuratreactive$markersDE) && nrow(seuratreactive$markersDE) > 1000) {
        shinyalert("WARNING!", "You have more than 1,000 differentially expressed genes. Selecting a higher number of genes for co-expression module construction will drastically
                increase computing time (Analysis with 2,000 genes can take >1 hour).", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      if (input$MEGENA_selectmethod == "Custom Differentially Expressed Genes" && is.null(seuratreactive$markersSDE)) {
        shinyalert("ERROR!", "No custom differentially expressed genes detected! You must compute your custom differentially expressed genes first! Refer to the custom differentially expressed genes tab.", type = "error", confirmButtonCol = "#337ab7")
        updatePickerInput(session, 'MEGENA_selectmethod', selected = "Variable Genes")
        updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      }
      
      if (input$MEGENA_selectmethod == "Custom Differentially Expressed Genes" && !is.null(seuratreactive$markersSDE) && nrow(seuratreactive$markersSDE) > 1000) {
        shinyalert("WARNING!", "You have more than 1,000 differentially expressed genes. Selecting a higher number of genes for co-expression module construction will drastically
                increase computing time (Analysis with 2,000 genes can take >1 hour).", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      if (input$MEGENA_selectmethod == "Differentially Expressed Genes" | input$MEGENA_selectmethod == "Custom Differentially Expressed Genes") {
        shinyjs::hide("variablegenesMEGENA")
        updateNumericInput(session, 'variablegenesMEGENA', value = 500)
      }
      else if (input$MEGENA_selectmethod == "Variable Genes") {
        shinyjs::show("variablegenesMEGENA")
      }
    }
  })
  
  observe({
    if (is.null(seuratreactive$markersDE)) {
      if (input$radioMEGENAselect == "Differentially Expressed Genes (Number of DEGs)" | input$radioMEGENAselect == "Differentially Expressed Genes (Percentage of DEGs)") {
        shinyalert("ERROR!", "No differentially expressed genes detected! You must compute your custom differentially expressed genes first! Refer to the differentially expressed genes tab.", type = "error", confirmButtonCol = "#337ab7")
        updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      }
    }
    
    if (is.null(seuratreactive$markersSDE)) {
      if (input$radioMEGENAselect == "Custom Differentially Expressed Genes (Number of DEGs)" | input$radioMEGENAselect == "Custom Differentially Expressed Genes (Percentage of DEGs)") {
        shinyalert("ERROR!", "No custom differentially expressed genes detected! You must compute your custom differentially expressed genes first! Refer to the custom differentially expressed genes tab.", type = "error", confirmButtonCol = "#337ab7")
        updatePrettyRadioButtons(session, "radioMEGENAselect", selected = "Cell Type Module Activity")
      }
    }
  })
  
  output$textMEGENA <- renderUI({
    if (input$radioMEGENAselect == "Cell Type Module Activity") {
      HTML(paste("Table showing the percentage of cluster marker gene overlap with co-expression modules (higher values indicate higher co-expression module activity)"))
    }
    else if (input$radioMEGENAselect == "Differentially Expressed Genes (Number of DEGs)") {
      HTML(paste("Table showing the number of differential expressed gene overlap with co-expression modules"))
    }
    else if (input$radioMEGENAselect == "Differentially Expressed Genes (Percentage of DEGs)") {
      HTML(paste("Table showing the percentage of differential expressed gene overlap with co-expression modules"))
    }
    else if (input$radioMEGENAselect == "Custom Differentially Expressed Genes (Number of DEGs)") {
      HTML(paste("Table showing the number of custom differential expressed gene overlap with co-expression modules"))
    }
    else if (input$radioMEGENAselect == "Custom Differentially Expressed Genes (Percentage of DEGs)") {
      HTML(paste("Table showing the percentage of custom differential expressed gene overlap with co-expression modules"))
    }
  })
  
  observe({
    if (!is.null(input$variablegenesMEGENA) && !is.null(input$MEGENAmin.size) && !is.null(input$MEGENApval)) {
      if (is.na(input$variablegenesMEGENA) | is.null(input$MEGENA_samples) | is.null(input$MEGENA_selectcluster) | is.null(input$MEGENA_selectmethod) | is.na(input$MEGENAmin.size) | is.na(input$MEGENApval) | is.na(input$n_MEGENA)) {
        shinyjs::disable("startbuttonMEGENA")
      }
      else {
        shinyjs::enable("startbuttonMEGENA")
      }
    }
    
    if (!is.null(input$variablegenesMEGENA)) {
      if (!is.na(input$variablegenesMEGENA) && input$variablegenesMEGENA >= 2001) {
        shinyalert("WARNING!", "Selecting a higher number of variable genes for co-expression module construction will drastically
                increase computing time (Analysis with 2,000 variable genes can take >1 hour).", type = "warning", confirmButtonCol = "#337ab7")
      }
    }
    
    if (!is.null(input$n_MEGENA)) {
      if (!is.na(input$n_MEGENA) && input$n_MEGENA > 1000) {
        shinyalert("WARNING!", "Increasing the number of cells analysed (>1,000) per cluster/cell type will drastically
                increase computing time (Analysis with 5,000 cells can take upwards of 3-4 hours).", type = "warning", confirmButtonCol = "#337ab7")
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$summary.output)) {
      updateVirtualSelect("MEGENA_modulesInput", choices = c(names(seuratreactive$summary.output$modules)), selected = names(seuratreactive$summary.output$modules)[1])
    }
  })
  
  observe({
    if (!is.null(input$log2FCMEGENA) && !is.null(input$minMEGENA) && !is.null(input$pvalMEGENA) && !is.null(input$testuseMEGENA) && !is.null(input$padjustMEGENA)) {
      if (is.na(input$log2FCMEGENA) | is.na(input$minMEGENA) | is.na(input$pvalMEGENA)) {
        shinyjs::disable("SunMEGENA")
      }
      else {
        shinyjs::enable("SunMEGENA")
      }
    }
  })
  
  output$CoExMEGENA <- renderPlot({
    validate(
      need(!is.null(seuratreactive$summary.output), 'Please compute your MEGENA co-expression modules'))
    
    if (input$radioMEGENA == "View All Modules") {
      module.table <- seuratreactive$summary.output$module.table
      colnames(module.table)[1] <- "id" # first column of module table must be labelled as "id".
      
      plot <- plot_module_hierarchy(module.table = module.table,label.scaleFactor = input$genefontMEGENA,
                                    arrow.size = 0.015, node.label.color = "blue", edge.color = "red")
    }
    else {
      plot <- plot_module(output.summary = seuratreactive$summary.output,PFN = seuratreactive$g,subset.module = input$MEGENA_modulesInput,
                          layout = "kamada.kawai",label.hubs.only = input$checkboxMEGENA,
                          gene.set = NULL,color.code =  "grey",
                          output.plot = FALSE, out.dir = "modulePlot",col.names = palette(rainbow(6)),label.scaleFactor = input$genefontMEGENA,
                          hubLabel.col = "black", label.alpha = input$TranspraentMEGENA, show.topn.hubs = Inf,show.legend = TRUE)
    }
    plot[[1]]
  })
  
  output$MEGENA_table <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$summary.output), 'Please compute your MEGENA co-expression modules'))
    
    seuratreactive$summary.output$module.table[,c(1,2,3,4,6)]
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$Sunburst_MEGENA <- renderPlot({
    validate(
      need(!is.null(seuratreactive$summary.output), 'Please compute your MEGENA co-expression modules'))
    
    if (input$radioMEGENAselect == "Cell Type Module Activity") {
      simplelist <- vector(mode="list")
      
      for (i in unique(seuratreactive$markersMEGENA$Cluster)) {
        simplelist[[i]] <- seuratreactive$markersMEGENA[seuratreactive$markersMEGENA$Cluster == i,]$`Gene Name`
      }
      
      anotherlist <- vector(mode="list")
      yetanotherlist <- vector(mode="list")
      
      for (i in names(seuratreactive$summary.output$modules)) {
        anotherlist[[i]] <- mapply(function(x, y) {signif(length(intersect(x,y))*100/length(y), 4)}, simplelist, seuratreactive$summary.output$modules[i])
        yetanotherlist[[i]] <- ifelse(all(anotherlist[[i]] == 0), "*Uncertain", names(which.max(anotherlist[[i]])))
      }
      
      names <- as.data.frame(t(data.frame(yetanotherlist)))
      
      names <- names %>%
        rownames_to_column(var="module.id")
      
      seuratreactive$anotherlist <- anotherlist
      
      # get some coloring (with log transform option)
      mdf= seuratreactive$summary.output$module.table
      
      mdf <- full_join(mdf, names)
      
      mdf$V1 <- as.factor(mdf$V1)
      
      colnames(mdf)[7] <- "Cluster"
      
      seuratreactive$sunburst <- draw_sunburst_wt_fill(module.df = mdf,feat.col = "Cluster",
                                                       fill.type = "discrete",
                                                       id.col = "module.id",parent.col = "module.parent")
    }
    else if (input$radioMEGENAselect == "Differentially Expressed Genes (Number of DEGs)") {
      
      simplelist <- vector(mode="list")
      
      simplelist[[1]] <- seuratreactive$markersDE$`Gene Name`
      
      anotherlist <- vector(mode="list")
      
      for (i in names(seuratreactive$summary.output$modules)) {
        anotherlist[[i]] <- mapply(function(x, y) {signif(length(intersect(x,y)))}, simplelist, seuratreactive$summary.output$modules[i])
      }
      
      names <- as.data.frame(t(data.frame(anotherlist)))
      
      names <- names %>%
        rownames_to_column(var="module.id")
      
      seuratreactive$anotherlist <- anotherlist
      
      # get some coloring (with log transform option)
      mdf= seuratreactive$summary.output$module.table
      
      mdf <- full_join(mdf, names)
      
      mdf$V1 <- as.factor(mdf$V1)
      
      colnames(mdf)[7] <- "Number of DEGs overlap in module"
      mdf$`Number of DEGs overlap in module` <- as.numeric(mdf$`Number of DEGs overlap in module`)
      
      seuratreactive$sunburst <- draw_sunburst_wt_fill(module.df = mdf,feat.col = "Number of DEGs overlap in module",
                                                       fill.type = "continuous", log.transform = F,
                                                       fill.scale = scale_fill_gradient2(low = "white",mid = "white",high = "red"),
                                                       id.col = "module.id",parent.col = "module.parent")
    }
    else if (input$radioMEGENAselect == "Differentially Expressed Genes (Percentage of DEGs)") {
      simplelist <- vector(mode="list")
      
      simplelist[[1]] <- seuratreactive$markersDE$`Gene Name`
      
      anotherlist <- vector(mode="list")
      
      for (i in names(seuratreactive$summary.output$modules)) {
        anotherlist[[i]] <- mapply(function(x, y) {signif(length(intersect(x,y))*100/length(y))}, simplelist, seuratreactive$summary.output$modules[i])
      }
      
      names <- as.data.frame(t(data.frame(anotherlist)))
      
      names <- names %>%
        rownames_to_column(var="module.id")
      
      seuratreactive$anotherlist <- anotherlist
      
      # get some coloring (with log transform option)
      mdf= seuratreactive$summary.output$module.table
      
      mdf <- full_join(mdf, names)
      
      mdf$V1 <- as.factor(mdf$V1)
      
      colnames(mdf)[7] <- "Percenatage of DEGs overlap in module"
      mdf$`Percenatage of DEGs overlap in module` <- as.numeric(mdf$`Percenatage of DEGs overlap in module`)
      
      seuratreactive$sunburst <- draw_sunburst_wt_fill(module.df = mdf,feat.col = "Percenatage of DEGs overlap in module",
                                                       fill.type = "continuous", log.transform = F,
                                                       fill.scale = scale_fill_gradient2(low = "white",mid = "white",high = "red"),
                                                       id.col = "module.id",parent.col = "module.parent")
    }
    else if (input$radioMEGENAselect == "Custom Differentially Expressed Genes (Number of DEGs)") {
      
      simplelist <- vector(mode="list")
      
      simplelist[[1]] <- seuratreactive$markersSDE$`Gene Name`
      
      anotherlist <- vector(mode="list")
      
      for (i in names(seuratreactive$summary.output$modules)) {
        anotherlist[[i]] <- mapply(function(x, y) {signif(length(intersect(x,y)))}, simplelist, seuratreactive$summary.output$modules[i])
      }
      
      names <- as.data.frame(t(data.frame(anotherlist)))
      
      names <- names %>%
        rownames_to_column(var="module.id")
      
      seuratreactive$anotherlist <- anotherlist
      
      # get some coloring (with log transform option)
      mdf= seuratreactive$summary.output$module.table
      
      mdf <- full_join(mdf, names)
      
      mdf$V1 <- as.factor(mdf$V1)
      
      colnames(mdf)[7] <- "Number of DEGs overlap in module"
      mdf$`Number of DEGs overlap in module` <- as.numeric(mdf$`Number of DEGs overlap in module`)
      
      seuratreactive$sunburst <- draw_sunburst_wt_fill(module.df = mdf,feat.col = "Number of DEGs overlap in module",
                                                       fill.type = "continuous", log.transform = F,
                                                       fill.scale = scale_fill_gradient2(low = "white",mid = "white",high = "red"),
                                                       id.col = "module.id",parent.col = "module.parent")
    }
    else if (input$radioMEGENAselect == "Custom Differentially Expressed Genes (Percentage of DEGs)") {
      simplelist <- vector(mode="list")
      
      simplelist[[1]] <- seuratreactive$markersSDE$`Gene Name`
      
      anotherlist <- vector(mode="list")
      
      for (i in names(seuratreactive$summary.output$modules)) {
        anotherlist[[i]] <- mapply(function(x, y) {signif(length(intersect(x,y))*100/length(y))}, simplelist, seuratreactive$summary.output$modules[i])
      }
      
      names <- as.data.frame(t(data.frame(anotherlist)))
      
      names <- names %>%
        rownames_to_column(var="module.id")
      
      seuratreactive$anotherlist <- anotherlist
      
      # get some coloring (with log transform option)
      mdf= seuratreactive$summary.output$module.table
      
      mdf <- full_join(mdf, names)
      
      mdf$V1 <- as.factor(mdf$V1)
      
      colnames(mdf)[7] <- "Percenatage of DEGs overlap in module"
      mdf$`Percenatage of DEGs overlap in module` <- as.numeric(mdf$`Percenatage of DEGs overlap in module`)
      
      seuratreactive$sunburst <- draw_sunburst_wt_fill(module.df = mdf,feat.col = "Percenatage of DEGs overlap in module",
                                                       fill.type = "continuous", log.transform = F,
                                                       fill.scale = scale_fill_gradient2(low = "white",mid = "white",high = "red"),
                                                       id.col = "module.id",parent.col = "module.parent")
    }
    
    seuratreactive$sunburst
  })
  
  output$MEGENA_markertable <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$summary.output), 'Please compute your MEGENA co-expression modules'))
    
    if (input$radioMEGENAselect == "Cell Type Module Activity") {
      df <- as.data.frame(seuratreactive$anotherlist)
      
      df <- df %>% rownames_to_column(var="Cluster")
    }
    else {
      df <- as.data.frame(seuratreactive$anotherlist)
    }
    
    df
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$GOHeatmap_MEGENAUI <- renderUI({
    plotOutput("GOHeatmap_MEGENA", height = (300 + input$GOheightMEGENA*30))
  })
  
  output$GOHeatmap_MEGENA <- renderPlot({
    validate(
      need(!is.null(seuratreactive$moduleGO_df), 'Please press APPLY on the sidebar menu to begin computation.'))
    plotModuleGO(seuratreactive$moduleGO_df, nTerms = input$GOtermsMEGENA, text_size = input$GOtextMEGENA, coord_flip = TRUE, axis_x_text_angle = 90) +
      theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust = 1))
  })
  
  ####TAB 9 - SCENIC####
  
  observe({
    if (is.null(seuratreactive$MAGMA_Final_Results) && !is.null(seuratreactive$regulons) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = T),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = T),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = T),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observeEvent(input$startSCENIC, {
    showModal(modalDialog(
      title = strong("SCENIC Settings"),
      pickerInput("SCENIC_selectmethod", h5(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select input method')),
                  choices = list("Variable Genes", "Differentially Expressed Genes", "Custom Differentially Expressed Genes"),
                  width = "100%"),
      numericInput("variablegenesSCENIC", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Pick number of variable genes:')),
                   min = 1,
                   value = 500, step = 50, 
                   width = "100%"),
      virtualSelectInput("SCENIC_samples", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select sample(s)')), 
                         choices = NULL, 
                         multiple = T,
                         search = T,
                         disableSelectAll = F,
                         hideClearButton = F,
                         width = "100%"),
      pickerInput("SCENIC_selectcluster", h5(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select cells by')),
                  choices = list("Seurat Clusters", "Labelled Cells"),
                  width = "100%"),
      
      h4(HTML('<h4 style = "text-align:justify;color:#000000;"> <b>Gene regluatory network (regulon identification) settings</b><br>')),
      
      pickerInput("SCENICdb", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Select Transcription Factor Binding Motif Database')),
                  choices = list("Human", "Mouse", "Fly"),
                  selected = "Human",
                  width = "100%"),
      numericInput("n_SCENIC", h5(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">Downsample regulon construction to X cells per cluster/cell type (randomly selected)')),
                   min = 1,
                   value = 100, step = 50, 
                   width = "100%"),
      p(style="text-align: center;", actionBttn("startbuttonSCENIC", "APPLY", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of hours to days to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startSCENIC, {
    
    if (!is.null(input$variablegenesSCENIC) && !is.null(input$n_SCENIC)) {
      if (is.na(input$variablegenesSCENIC) |
          is.null(input$SCENIC_samples) |
          is.na(input$n_SCENIC) |
          is.null(input$SCENIC_selectcluster) |
          is.null(input$SCENIC_selectmethod)) {
        shinyjs::disable("startbuttonSCENIC")
      }
      else {
        shinyjs::enable("startbuttonSCENIC")
      }
    }
    
    updateVirtualSelect('SCENIC_samples', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                        selected = seuratreactive$SCENIC_samples)
    updateNumericInput(session, 'variablegenesSCENIC', value = seuratreactive$variablegenesSCENIC)
    updateNumericInput(session, 'n_SCENIC', value = seuratreactive$n_SCENIC)
    updatePickerInput(session, 'SCENICdb', selected = seuratreactive$SCENICdb)
    updatePickerInput(session, 'SCENIC_selectcluster', selected = seuratreactive$SCENIC_selectcluster)
    updatePickerInput(session, 'SCENIC_selectmethod', selected = seuratreactive$SCENIC_selectmethod)
    
    if(!is.null(seuratreactive$reductionCC1)) {
      if (identical(seuratreactive$obj@active.ident, seuratreactive$obj$seurat_clusters)) {
        if (!is.null(seuratreactive$SCENIC_selectcluster)) {
          updatePickerInput(session, 'SCENIC_selectcluster', choices = list("Seurat Clusters"), selected = seuratreactive$SCENIC_selectcluster)
        }
        else {
          updatePickerInput(session, 'SCENIC_selectcluster', choices = list("Seurat Clusters"))
        }
      }
      else {
        if (!is.null(seuratreactive$SCENIC_selectcluster)) {
          updatePickerInput(session, 'SCENIC_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$SCENIC_selectcluster)
        }
        else {
          updatePickerInput(session, 'SCENIC_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"))
        }
      }
    }
  })
  
  observe({
    if (!is.null(input$variablegenesSCENIC) && !is.null(input$n_SCENIC)) {
      if (is.na(input$variablegenesSCENIC) |
          is.null(input$SCENIC_selectmethod) |
          is.null(input$SCENIC_samples) |
          is.na(input$n_SCENIC) |
          is.null(input$SCENIC_selectcluster) |
          is.null(input$SCENIC_selectmethod)) {
        shinyjs::disable("startbuttonSCENIC")
      }
      else {
        shinyjs::enable("startbuttonSCENIC")
      }
    }
    
    if (!is.null(input$n_SCENIC)) {
      if (!is.na(input$n_SCENIC) && input$n_SCENIC >= 1000) {
        shinyalert("WARNING!", "Increasing the number of cells analysed (>1,000) per cluster/cell type will drastically
                increase computing time (Analysis with 5,000 cells can take upwards of 3-4 hours).", type = "warning", confirmButtonCol = "#337ab7")
      }
    }
    
    if (!is.null(input$variablegenesSCENIC)) {
      if (!is.na(input$variablegenesSCENIC) && input$variablegenesSCENIC >= 2001) {
        shinyalert("WARNING!", "Selecting a higher number of variable genes to include for analysis will drastically increase
     computational time for regulon construction (Analysis with 2,000 variable genes can take upwards of 3-4 hours).", type = "warning", confirmButtonCol = "#337ab7")
      }
    }
  })
  
  observe({
    if (!is.null(input$SCENIC_selectmethod)) {
      if (input$SCENIC_selectmethod == "Differentially Expressed Genes" && is.null(seuratreactive$markersDE)) {
        shinyalert("ERROR!", "No differentially expressed genes detected! You must compute your differentially expressed genes first! Refer to the differentially expressed genes tab.", type = "error", confirmButtonCol = "#337ab7")
        updatePickerInput(session, 'SCENIC_selectmethod', selected = "Variable Genes")
      }
      
      if (input$SCENIC_selectmethod == "Differentially Expressed Genes" && !is.null(seuratreactive$markersDE) && nrow(seuratreactive$markersDE) > 1000) {
        shinyalert("WARNING!", "You have more than 1,000 differentially expressed genes. Selecting a higher number of genes to include for analysis will drastically increase
     computational time for regulon construction (Analysis with 2,000 genes can take upwards of 3-4 hours).", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      if (input$SCENIC_selectmethod == "Custom Differentially Expressed Genes" && is.null(seuratreactive$markersSDE)) {
        shinyalert("ERROR!", "No custom differentially expressed genes detected! You must compute your custom differentially expressed genes first! Refer to the custom differentially expressed genes tab.", type = "error", confirmButtonCol = "#337ab7")
        updatePickerInput(session, 'SCENIC_selectmethod', selected = "Variable Genes")
      }
      
      if (input$SCENIC_selectmethod == "Custom Differentially Expressed Genes" && !is.null(seuratreactive$markersSDE) && nrow(seuratreactive$markersSDE) > 1000) {
        shinyalert("WARNING!", "You have more than 1,000 differentially expressed genes. Selecting a higher number of genes to include for analysis will drastically increase
     computational time for regulon construction (Analysis with 2,000 genes can take upwards of 3-4 hours).", type = "warning", confirmButtonCol = "#337ab7")
      }
      
      if (input$SCENIC_selectmethod == "Differentially Expressed Genes" | input$SCENIC_selectmethod == "Custom Differentially Expressed Genes") {
        shinyjs::hide("variablegenesSCENIC")
        updateNumericInput(session, 'variablegenesSCENIC', value = 500)
      }
      else if (input$SCENIC_selectmethod == "Variable Genes") {
        shinyjs::show("variablegenesSCENIC")
      }
    }
  })
  
  observeEvent(input$startbuttonSCENIC,{
    library(SCENIC)
    library(AUCell)
    library(RcisTarget)
    library(ggplot2)
    library(ggraph)
    library(igraph)
    library(ComplexHeatmap)
    
    tryCatch({
      withProgress(message = "Performing gene regulatory network analysis. Please wait...",
                   value = 0, {
                     incProgress(1/10)
                     
                     removeModal()
                     
                     seuratreactive$variablefeaturesSCENIC <- NULL
                     seuratreactive$variablegenesSCENIC <- NULL
                     seuratreactive$regulonAUC <- NULL
                     seuratreactive$regulons <- NULL
                     seuratreactive$cellInfo <- NULL
                     seuratreactive$n_SCENIC <- NULL
                     seuratreactive$SCENICdb <- NULL
                     seuratreactive$SCENIC_samples <- NULL
                     seuratreactive$regulonnames <- NULL
                     seuratreactive$titleSCENIC <- NULL
                     seuratreactive$SCENICInput <- NULL
                     seuratreactive$SCENIC_selectcluster <- NULL
                     seuratreactive$SCENIC_selectmethod <- NULL
                     seuratreactive$regulonActivity_byCellType_Scaled <- NULL
                     seuratreactive$rss <- NULL
                     
                     source("runSCENIC_2_createRegulons.R")
                     source("class_ScenicOptions.R")
                     source("runGenie3.R")
                     
                     unlink("int", recursive = TRUE)
                     unlink("output", recursive = TRUE)
                     
                     if (input$SCENIC_selectmethod == "Variable Genes") {
                       seuratreactive$variablefeaturesSCENIC <- FindVariableFeatures(seuratreactive$obj, selection.method = "vst", nfeatures = input$variablegenesSCENIC)
                       
                       seuratreactive$variablefeaturesSCENIC <- VariableFeatures(seuratreactive$variablefeaturesSCENIC)
                     }
                     else if (input$SCENIC_selectmethod == "Differentially Expressed Genes") {
                       seuratreactive$variablefeaturesSCENIC <- seuratreactive$markersDE$`Gene Name`
                     }
                     else if (input$SCENIC_selectmethod == "Custom Differentially Expressed Genes") {
                       seuratreactive$variablefeaturesSCENIC <- seuratreactive$markersSDE$`Gene Name`
                     }
                     
                     incProgress(2/10)
                     
                     #Generating expression matrix of DEGs
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       exprMat <- as.matrix(seuratreactive$obj@assays$RNA@data)
                     }
                     else {
                       exprMat <- as.matrix(seuratreactive$obj@assays$SCT@data)
                     }
                     
                     
                     #Getting cell info
                     seuratreactive$cellInfo <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident", "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     seuratreactive$cellInfo <- seuratreactive$cellInfo[seuratreactive$cellInfo$orig.ident %in% input$SCENIC_samples, ]
                     
                     seuratreactive$cellInfo <- tibble::rownames_to_column(seuratreactive$cellInfo, "rownames")
                     
                     if (input$SCENIC_selectcluster == "Seurat Clusters") {
                       seuratreactive$cellInfo <- seuratreactive$cellInfo %>% mutate(sample(100000, size = nrow(seuratreactive$cellInfo), replace = TRUE)) %>% 
                         group_by(seurat_clusters) %>% 
                         top_n(input$n_SCENIC)
                     }
                     else {
                       seuratreactive$cellInfo <- seuratreactive$cellInfo %>% mutate(sample(100000, size = nrow(seuratreactive$cellInfo), replace = TRUE)) %>% 
                         group_by(ident) %>% 
                         top_n(input$n_SCENIC)
                     }
                     
                     seuratreactive$cellInfo <- seuratreactive$cellInfo %>% remove_rownames %>% column_to_rownames(var="rownames")
                     
                     exprMat <- exprMat[rownames(exprMat) %in% seuratreactive$variablefeaturesSCENIC, colnames(exprMat) %in% rownames(seuratreactive$cellInfo)]
                     
                     #Initialize SCENIC
                     if (input$SCENICdb == "Mouse") {
                       org <- "mgi"
                     }
                     else if (input$SCENICdb == "Fly") {
                       org <- "dmel"
                     }
                     else {
                       org <- "hgnc"
                     }
                     
                     dbDir <- "cisTarget_databases"
                     myDatasetTitle <- "SCENIC"
                     data(defaultDbNames)
                     dbs <- defaultDbNames[[org]]
                     scenicOptions <- initializeScenic(org=org, dbDir=dbDir, dbs=dbs, datasetTitle=myDatasetTitle, nCores=1)
                     scenicOptions@inputDatasetInfo$cellInfo <- seuratreactive$cellInfo
                     scenicOptions@settings$seed <- 123
                     
                     ### Co-Expression network
                     genesKept <- geneFiltering(exprMat, scenicOptions)
                     exprMat_filtered <- exprMat[genesKept, ]
                     runCorrelation(exprMat_filtered, scenicOptions)
                     exprMat_filtered_log <- log2(exprMat_filtered+1) 
                     runGenie3(exprMat_filtered_log, scenicOptions)
                     
                     incProgress(5/10)
                     
                     scenicOptions <- runSCENIC_1_coexNetwork2modules(scenicOptions)
                     
                     scenicOptions <- runSCENIC_2_createRegulons(scenicOptions, coexMethod=NULL)
                     
                     incProgress(8/10)
                     
                     library(doParallel)
                     library(doRNG)
                     
                     scenicOptions <- runSCENIC_3_scoreCells(scenicOptions, exprMat_filtered_log)
                     
                     #Get information for plots
                     seuratreactive$regulonAUC <- loadInt(scenicOptions, "aucell_regulonAUC")
                     seuratreactive$regulonnames <- signif(seuratreactive$regulonAUC@assays@data@listData[["AUC"]],4)
                     
                     seuratreactive$regulons <- loadInt(scenicOptions, "regulons")
                     seuratreactive$regulons <- plyr::ldply(seuratreactive$regulons, rbind)
                     colnames(seuratreactive$regulons)[1] <- "Regulon"
                     seuratreactive$regulons <- seuratreactive$regulons[seuratreactive$regulons$Regulon %in% sub(" \\(.+", "", rownames(seuratreactive$regulonnames)),]
                     
                     seuratreactive$n_SCENIC <- input$n_SCENIC
                     seuratreactive$SCENICdb <- input$SCENICdb
                     seuratreactive$SCENIC_samples <- input$SCENIC_samples
                     seuratreactive$SCENIC_selectcluster <- input$SCENIC_selectcluster
                     seuratreactive$SCENIC_selectmethod <- input$SCENIC_selectmethod
                     seuratreactive$scenicOptions <- scenicOptions
                     seuratreactive$variablegenesSCENIC <- input$variablegenesSCENIC
                     
                     if (seuratreactive$SCENIC_selectcluster == "Seurat Clusters") {
                       regulonActivity_byCellType <- sapply(split(rownames(seuratreactive$cellInfo), seuratreactive$cellInfo$seurat_clusters),
                                                            function(cells) rowMeans(data.frame(getAUC(seuratreactive$regulonAUC)[,cells])))
                     }
                     else {
                       regulonActivity_byCellType <- sapply(split(rownames(seuratreactive$cellInfo), seuratreactive$cellInfo$ident),
                                                            function(cells) rowMeans(data.frame(getAUC(seuratreactive$regulonAUC)[,cells])))
                     }
                     
                     regulonActivity_byCellType_Scaled <- t(scale(t(regulonActivity_byCellType), center = T, scale=T))
                     
                     seuratreactive$regulonActivity_byCellType_Scaled <- regulonActivity_byCellType_Scaled[,colSums(is.na(regulonActivity_byCellType_Scaled))<nrow(regulonActivity_byCellType_Scaled)]
                     
                     if (seuratreactive$SCENIC_selectcluster == "Seurat Clusters") {
                       seuratreactive$rss <- calcRSS(AUC=getAUC(seuratreactive$regulonAUC), cellAnnotation=seuratreactive$cellInfo[colnames(seuratreactive$regulonAUC), "seurat_clusters"], )
                     }
                     else {
                       seuratreactive$rss <- calcRSS(AUC=getAUC(seuratreactive$regulonAUC), cellAnnotation=seuratreactive$cellInfo[colnames(seuratreactive$regulonAUC), "ident"], )
                     }
                     
                     
                     incProgress(10/10)
                     
                     updateVirtualSelect('SCENICInput', choices = c("Please select from the following" = "", rownames(seuratreactive$regulonnames)))
                     
                     if (input$iscollapseboxSCENIC == FALSE) {
                       js$collapse("SCENICBoxIntro")
                     }
                     shinyjs::show("SCENICBox1a")
                     shinyjs::show("SCENICBox1b")
                     shinyjs::show("SCENICBox1")
                     shinyjs::show("SCENICBox2")
                     shinyjs::show("SCENICBox3")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "SCENIC Regulatory Gene Network Failed. Check to see if the species database is correct. Try including more cells in the analysis.", type = "error", confirmButtonCol = "#337ab7")
      seuratreactive$variablefeaturesSCENIC <- NULL
      seuratreactive$regulonAUC <- NULL
      seuratreactive$regulons <- NULL
      seuratreactive$cellInfo <- NULL
      seuratreactive$regulonnames <- NULL
      seuratreactive$titleSCENIC <- NULL
      seuratreactive$SCENICInput <- NULL
      seuratreactive$regulonActivity_byCellType_Scaled <- NULL
      seuratreactive$rss <- NULL
    })
  })
  
  observeEvent(input$regulon1, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     regulon <- as.data.frame(t(seuratreactive$regulonnames))
                     
                     df <- regulon[, colnames(regulon) == input$SCENICInput]
                     
                     df <- as.data.frame(df)
                     
                     rownames(df) <- rownames(regulon)
                     colnames(df) <- "Regulon"
                     
                     incProgress(3/10)
                     
                     IdentifiersSCENIC <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                          "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident"))
                     
                     IdentifiersSCENIC[c("percent.ribo", "percent.mt")] <- signif(IdentifiersSCENIC[c("percent.ribo", "percent.mt")], digits = 4)
                     
                     
                     colnames(IdentifiersSCENIC) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                      "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                      "Cell Cycle Phase", "Labelled Cells")
                     
                     IdentifiersSCENIC <-  merge(IdentifiersSCENIC, df, by=0, all=TRUE)
                     
                     rownames(IdentifiersSCENIC) <- IdentifiersSCENIC$Row.names
                     
                     IdentifiersSCENIC <- IdentifiersSCENIC[,-1]
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersSCENIC, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     incProgress(8/10)
                     
                     seuratreactive$titleSCENIC <- paste(input$SCENICInput, collapse = " ")
                     
                     seuratreactive$SCENICInput <- input$SCENICInput
                     
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please select a regulon to view its expression.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observe({
    if (!is.null(seuratreactive$regulons)) {
      if (seuratreactive$SCENIC_selectmethod == "Seurat Clusters") {
        updateVirtualSelect('Regulon_Select2', choices = c(unique(levels(seuratreactive$obj$seurat_clusters))), selected = c(unique(levels(seuratreactive$obj$seurat_clusters)))[1])
      }
      else {
        updateVirtualSelect('Regulon_Select2', choices = c(unique(levels(seuratreactive$obj))), selected = c(unique(levels(seuratreactive$obj)))[1])
      }
    }
  })
  
  observe({
    if (input$radioSCENICrss == "All") {
      shinyjs::hide("Regulon_Select2")
    }
    if (input$radioSCENICrss == "Individual") {
      shinyjs::show("Regulon_Select2")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$regulonnames)) {
      updateVirtualSelect('SCENICInput', choices = c("Please select from the following" = "", rownames(seuratreactive$regulonnames)))
    }
  })
  
  output$plotSCENICUI <- renderUI({
    plotlyOutput("plotSCENIC", height = (400 + input$SCENICplotheight*40))
  })
  
  output$plotSCENIC <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data$`Regulon`), 'You must select a regulon first to view its expression'))
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    if(input$plotclustersSCENIC == "Regulon"){
      if(input$radioSCENIC == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(title = seuratreactive$titleSCENIC, font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(title = seuratreactive$titleSCENIC, font = list(size = input$labelsizeSCENIC)) 
        
      }
      else if (input$radioSCENIC == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
                
        ) %>% layout(title = seuratreactive$titleSCENIC, font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(title = seuratreactive$titleSCENIC, font = list(size = input$labelsizeSCENIC))
      }
    }
    else if(input$plotclustersSCENIC == "Seurat Clusters" | input$plotclustersSCENIC == "Cell Cycle Phase" | input$plotclustersSCENIC == "Labelled Cells" |
            input$plotclustersSCENIC == "Sample") {  
      if(input$radioSCENIC == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
      }
      else if (input$radioSCENIC == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
        
      }
    }
    else {
      if(input$radioSCENIC == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
        
      }
      else if (input$radioSCENIC == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.data[, input$plotclustersSCENIC],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeSCENIC),
                opacity = input$opacitySCENIC,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Regulon Activity (AUC): ", seuratreactive$plot.data$`Regulon`)
        ) %>% layout(font = list(size = input$labelsizeSCENIC))
      }
    }
  })
  
  output$Heatmap_SCENICUI <- renderUI({
    plotOutput("Heatmap_SCENIC", height = (300 + input$HeightSCENIC*30))
  })
  
  output$Heatmap_SCENIC <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$regulons), 'Please compute your marker genes'))
    
    Heatmap(seuratreactive$regulonActivity_byCellType_Scaled, name="Regulon activity", 
            column_names_gp = grid::gpar(fontsize = input$fontSCENIC),
            row_names_gp = grid::gpar(fontsize = input$fontSCENIC))
  })
  
  output$Dotplot_SCENICUI <- renderUI({
    plotlyOutput("Dotplot_SCENIC", height = (300 + input$HeightSCENIC2*30))
  })
  
  output$Dotplot_SCENIC <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$regulons), 'Please compute your marker genes'))
    
    if (input$radioSCENICrss == "All") {
      rssPlot <- plotRSS(seuratreactive$rss)
      p <- rssPlot$plot + theme(text = element_text(size=input$fontSCENIC2))
      
      ggplotly(p)
    }
    else {
      plotRSS_oneSet(seuratreactive$rss, setName = input$Regulon_Select2, n = length(rownames(seuratreactive$rss))-1) + 
        theme(text = element_text(size=input$fontSCENIC2))
    }
  })
  
  output$TF_Regulon_Map <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$regulons), 'Please compute your marker genes'))
    
    #Extra R function
    onlyDuplicatedExtended <- function(regulonNames)
    {
      regulonNames <- unname(regulonNames)
      tfs <- getTF(regulonNames)
      tfs <- gsub("_extended", "", tfs)
      splitRegulons <- split(regulonNames, tfs)[unique(tfs)]
      
      ret <- sapply(splitRegulons, function(x) 
      {
        split(x, !grepl("_extended", x))[[1]] # False (direct) will be first
      })
      
      return(ret)
    }
    
    if (input$checkboxshowextended == TRUE) {
      regulons <- seuratreactive$regulons[seuratreactive$regulons$Regulon %in% onlyDuplicatedExtended(seuratreactive$regulons$Regulon),]
    }
    else if (input$checkboxshowextended == FALSE) {
      regulons <- seuratreactive$regulons[-grep("_extended", seuratreactive$regulons$Regulon), ]
    }
    
    regulons <- regulons[,c(1:input$SCENICNumber)]
    
    regulons <- reshape2::melt(regulons, id.vars = "Regulon") %>% na.omit()
    
    regulons$Regulon <- paste0(regulons$Regulon, "_Regulon")
    
    
    edges <- regulons[,c("Regulon", "value")]
    
    nodes1 <- regulons %>%
      group_by(`Regulon`) %>%
      summarise(node_color = "Regulon") %>%
      drop_na()
    
    nodes2 <- regulons %>%
      group_by(`value`) %>%
      summarise(node_color = "TFs") %>%
      drop_na()
    
    colnames(nodes2)[1] <- "Regulon"
    
    nodes <- full_join(nodes1, nodes2)
    
    graph = graph_from_data_frame(edges, directed = FALSE, nodes)
    
    if (input$checkboxgene_SCENIC == TRUE && input$checkboxregulon_SCENIC == TRUE) {
      ggraph(graph, layout = "fr") +
        geom_edge_link(aes(width = 1), color = "light blue") +
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "Regulon", input$nodeSCENIC, input$nodeSCENIC/2)) +
        geom_node_text(aes(label = ifelse(node_color == "Regulon", name, NA_character_)), size = input$categoryfontSCENIC, repel=TRUE) +
        geom_node_text(aes(label = ifelse(node_color == "TFs", name, NA_character_)), size = input$genefontSCENIC, repel=TRUE) +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() +
        guides(color = FALSE, edge_width = FALSE)
    }
    
    else if (input$checkboxgene_SCENIC == FALSE && input$checkboxregulon_SCENIC == TRUE) {
      ggraph(graph, layout = "fr") +
        geom_edge_link(aes(width = 1), color = "light blue") +
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "Regulon", input$nodeSCENIC, input$nodeSCENIC/2)) +
        geom_node_text(aes(label = ifelse(node_color == "Regulon", name, NA_character_)), size = input$categoryfontSCENIC, repel=TRUE) +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() +
        guides(color = FALSE, edge_width = FALSE)
    }
    
    else if (input$checkboxgene_SCENIC == TRUE && input$checkboxregulon_SCENIC == FALSE) {
      ggraph(graph, layout = "fr") +
        geom_edge_link(aes(width = 1), color = "light blue") +
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "Regulon", input$nodeSCENIC, input$nodeSCENIC/2)) +
        geom_node_text(aes(label = ifelse(node_color == "TFs", name, NA_character_)), size = input$genefontSCENIC, repel=TRUE) +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() +
        guides(color = FALSE, edge_width = FALSE)
    }
  })
  
  ####TAB 10 - MAGMA ####
  
  observe({
    if (!is.null(seuratreactive$MAGMA_Final_Results) && is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = T),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = T),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = T),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observeEvent(input$tabs, {
    if (input$tabs == "MAGMA") {
      updatePickerInput(session, 'selectMAGMAInput', selected = "")  
    }
  })
  
  observe({
    if (input$selectMAGMAInput == "") {
      shinyjs::show("MAGMABox1")
      shinyjs::hide("MAGMABox1a")
      shinyjs::hide("MAGMABox1b")
      shinyjs::hide("MAGMABox2a")
      shinyjs::show("MAGMABox2b")
    }
    else if (input$selectMAGMAInput == "Own GWAS Data (GWAS Summary Statistics)") {
      shinyjs::show("MAGMABox1")
      shinyjs::show("MAGMABox1a")
      shinyjs::hide("MAGMABox1b")
      shinyjs::show("MAGMABox2a")
      shinyjs::hide("MAGMABox2b")
    }
    else if (input$selectMAGMAInput == "Public GWAS Datasets") {
      shinyjs::show("MAGMABox1")
      shinyjs::hide("MAGMABox1a")
      shinyjs::show("MAGMABox1b")
      shinyjs::hide("MAGMABox2a")
      shinyjs::show("MAGMABox2b")
    }
  })
  
  observeEvent(input$startbuttonMAGMA, {
    library(MAGMA.Celltyping)
    library(EWCE) 
    
    seuratreactive$meta <- data.table::fread( paste("https://raw.githubusercontent.com/neurogenomics/MAGMA_Files_Public/0a69b011f0c97276ad2623946fbfda41ea0d5860/metadata.csv") ) 
    seuratreactive$meta <- seuratreactive$meta[,-1]
    
    filter <- read.table("UNIQUE_MAGMA.txt")
    
    seuratreactive$meta <- seuratreactive$meta[seuratreactive$meta$id %in% filter$V1]
    
    seuratreactive$meta <- seuratreactive$meta[,c(1:4,8,10,11,14,16,17,18,19,20)]
    
    seuratreactive$meta <- mutate(seuratreactive$meta, genes_raw_url = paste0("https://github.com/neurogenomics/MAGMA_Files_Public/raw/0a69b011f0c97276ad2623946fbfda41ea0d5860/data/MAGMA_Files/",
                                                                              seuratreactive$meta$id, ".tsv.gz.35UP.10DOWN/", seuratreactive$meta$id, ".tsv.gz.35UP.10DOWN.genes.raw"))
    
    seuratreactive$meta <- mutate(seuratreactive$meta, genes_out_url = paste0("https://github.com/neurogenomics/MAGMA_Files_Public/raw/0a69b011f0c97276ad2623946fbfda41ea0d5860/data/MAGMA_Files/",
                                                                              seuratreactive$meta$id, ".tsv.gz.35UP.10DOWN/", seuratreactive$meta$id, ".tsv.gz.35UP.10DOWN.genes.out"))
    
    if (input$iscollapseboxMAGMA == FALSE) {
      js$collapse("MAGMABoxIntro")
    }
    shinyjs::show("MAGMABox1")
    shinyjs::hide("MAGMABox1a")
    shinyjs::hide("MAGMABox1b")
    shinyjs::hide("MAGMABox2a")
    shinyjs::show("MAGMABox2b")
    shinyjs::show("MAGMABox3")
    shinyjs::show("MAGMABox4")
  })
  
  observeEvent(input$fileinputMAGMA, {
    tryCatch({
      withProgress(message = 'Loading GWAS Summary Statistics...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file = input$fileinputMAGMA
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seuratreactive$UploadedGWAS <- read.table(file$datapath, sep = "\t", header = T, check.names = F)
                     
                     if (all(grepl(pattern = "BP|CHR|SNP", colnames(seuratreactive$UploadedGWAS)[1:3]))) {
                       if (any(grepl("Z|OR|BETA|LOG_ODDS|SIGNED_SUMSTAT", colnames(seuratreactive$UploadedGWAS)))) {
                         
                         seuratreactive$path_formatted <- paste(input$fileinputMAGMA$datapath)
                       }
                       else {
                         shinyalert("ERROR!", "File contains does not contain either of the following columns: Z, OR, BETA, LOG_ODDS, SIGNED_SUMSTAT. Please check your uploaded data.", type = "error", confirmButtonCol = "#337ab7")
                         shinyjs::disable("MAGMAGO2")
                       }
                     }
                     else {
                       shinyalert("ERROR!", "File contains does not contain either of the following columns: BP, CHR, SNP. Please check your uploaded data.", type = "error", confirmButtonCol = "#337ab7")
                       shinyjs::disable("MAGMAGO2")
                     }
                     incProgress(0.8)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "File is not in correct format.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observe({
    if (is.null(input$MAGMA_samples) |
        is.null(input$MAGMA_GWAS) |
        is.null(input$MAGMA_selectcluster) |
        !is.null(input$n_MAGMA2) && is.na(input$n_MAGMA2)) {
      shinyjs::disable("MAGMAGO")
    }
    else {
      shinyjs::enable("MAGMAGO")
    }
  })
  
  observe({
    if (is.null(input$MAGMA_samples2) | is.null(input$textlabelMAGMA) |
        is.na(input$MAGMA_N) |
        is.null(input$GRCh) |
        is.null(input$fileinputMAGMA) |
        is.null(input$MAGMA_selectcluster2) |
        !is.null(input$n_MAGMA) && is.na(input$n_MAGMA)) {
      shinyjs::disable("MAGMAGO2")
    }
    else {
      shinyjs::enable("MAGMAGO2")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) { 
      updateVirtualSelect('MAGMA_samples', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                          selected = seuratreactive$MAGMA_samples)
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) { 
      updateVirtualSelect('MAGMA_samples2', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                          selected = seuratreactive$MAGMA_samples2)
    }
  })
  
  observe({
    if (!is.null(seuratreactive$meta)) { 
      updateVirtualSelect('MAGMA_GWAS', choices = c(seuratreactive$meta$id),
                          selected = seuratreactive$MAGMA_GWAS)
    }
  })
  
  observeEvent(input$MAGMAGOGOGO, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     removeModal()
                     
                     source("MAGMA_fixpath.R")
                     source("MAGMA_getos.R")
                     source("MAGMA_messageparallel.R")
                     source("MAGMA_Gitdownload.R")
                     source("MAGMA_messager.R")
                     source("MAGMA_importmeta.R")
                     source("MAGMA_importmagmafiles.R")
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       exp <- seuratreactive$obj@assays$RNA@data
                     }
                     else {
                       exp <- seuratreactive$obj@assays$SCT@data
                     }
                     
                     cellInfo <- FetchData(object = seuratreactive$obj, vars = c("orig.ident", "ident", "seurat_clusters"))
                     
                     cellInfo <- cellInfo[cellInfo$orig.ident %in% input$MAGMA_samples, ]
                     
                     cellInfo <- tibble::rownames_to_column(cellInfo, "rownames")
                     
                     if (input$MAGMA_selectcluster == "Seurat Clusters") {
                       cellInfo <- cellInfo %>% mutate(sample(100000, size = nrow(cellInfo), replace = TRUE)) %>% 
                         group_by(seurat_clusters) %>% 
                         top_n(input$n_MAGMA2)
                     }
                     else {
                       cellInfo <- cellInfo %>% mutate(sample(100000, size = nrow(cellInfo), replace = TRUE)) %>% 
                         group_by(ident) %>% 
                         top_n(input$n_MAGMA2)
                     }
                     
                     cellInfo <- cellInfo %>% remove_rownames %>% column_to_rownames(var="rownames")
                     
                     exp <- exp[, colnames(exp) %in% rownames(cellInfo)]
                     
                     #Set up CTD
                     if (input$MAGMA_selectcluster == "Seurat Clusters") {
                       annotLevels <- list(as.character(cellInfo$seurat_clusters))
                     }
                     else {
                       annotLevels <- list(as.character(cellInfo$ident))
                     }
                     
                     if (seuratreactive$PAREACTOME == "sscrofa") {
                       CTD <- EWCE::generate_celltype_data(
                         exp = exp,
                         annotLevels = annotLevels,
                         groupName = "Seurat",
                         input_species = "hsapiens",
                         return_ctd=TRUE) 
                     }
                     else {
                       CTD <- EWCE::generate_celltype_data(
                         exp = exp,
                         annotLevels = annotLevels,
                         groupName = "Seurat",
                         input_species = seuratreactive$PAREACTOME,
                         return_ctd=TRUE) 
                     }
                     
                     table <- seuratreactive$meta[,c(1:2)]
                     colnames(table) <- c("GWAS", "Trait")
                     
                     seuratreactive$magma_dirs <- import_magma_files(ids = input$MAGMA_GWAS)
                     
                     if (!is.null(seuratreactive$genesOutPath2)) {
                       combineddir <- c(seuratreactive$genesOutPath2, seuratreactive$magma_dirs)
                     }
                     else {
                       combineddir <- seuratreactive$magma_dirs
                     }
                     
                     if (seuratreactive$PAREACTOME == "sscrofa") {
                       MAGMA_results <- celltype_associations_pipeline(
                         magma_dirs = combineddir,
                         ctd = CTD$ctd,
                         ctd_species = "hsapiens", 
                         ctd_name = "scRNA-seq_Dataset", 
                         run_linear = TRUE, 
                         run_top10 = TRUE,
                         force_new = TRUE,
                         save_dir = NULL,
                         standardise = T)
                     }
                     else {
                       MAGMA_results <- celltype_associations_pipeline(
                         magma_dirs = combineddir,
                         ctd = CTD$ctd,
                         ctd_species = seuratreactive$PAREACTOME, 
                         ctd_name = "scRNA-seq_Dataset", 
                         run_linear = TRUE, 
                         run_top10 = TRUE,
                         force_new = TRUE,
                         save_dir = NULL,
                         standardise = T)
                     }
                     
                     if (input$Linearmode2 == TRUE) {
                       merged_results <- MAGMA.Celltyping::merge_results(
                         MAGMA_results = MAGMA_results,
                         filetype = "ctAssocMerged")
                     }
                     else {
                       merged_results <- MAGMA.Celltyping::merge_results(
                         MAGMA_results = MAGMA_results,
                         filetype = "ctAssocsLinear")
                     }
                     
                     merged_results$GWAS <- sub("\\..+", "", merged_results$GWAS)
                     
                     merged_results$GWAS <- sub("^0$", input$textlabelMAGMA, merged_results$GWAS)
                     
                     seuratreactive$MAGMA_Final_Results <- full_join(merged_results, table)
                     
                     seuratreactive$MAGMA_Final_Results$Trait[is.na(seuratreactive$MAGMA_Final_Results$Trait)] <- seuratreactive$MAGMA_Final_Results$GWAS[which(is.na(seuratreactive$MAGMA_Final_Results$Trait))]
                     
                     seuratreactive$MAGMA_Final_Results <- na.omit(seuratreactive$MAGMA_Final_Results)
                     
                     incProgress(10/10)
                     
                     if (!is.null(seuratreactive$genesOutPath2)) {
                       seuratreactive$MAGMA_samples <- input$MAGMA_samples
                       seuratreactive$MAGMA_GWAS <- input$MAGMA_GWAS
                       seuratreactive$MAGMA_selectcluster <- input$MAGMA_selectcluster
                       seuratreactive$n_MAGMA2 <- input$n_MAGMA2
                       
                       seuratreactive$MAGMA_samples2 <- input$MAGMA_samples
                       seuratreactive$MAGMA_selectcluster2 <- input$MAGMA_selectcluster
                       seuratreactive$textlabelMAGMA <- input$textlabelMAGMA
                       seuratreactive$MAGMA_N <- input$MAGMA_N
                       seuratreactive$GRCh <- input$GRCh
                       seuratreactive$MAGMA_Pop <- input$MAGMA_Pop
                       seuratreactive$n_MAGMA <- input$n_MAGMA2
                     }
                     else {
                       seuratreactive$MAGMA_samples <- input$MAGMA_samples
                       seuratreactive$MAGMA_GWAS <- input$MAGMA_GWAS
                       seuratreactive$MAGMA_selectcluster <- input$MAGMA_selectcluster
                       seuratreactive$n_MAGMA2 <- input$n_MAGMA2
                       
                       seuratreactive$MAGMA_samples2 <- NULL
                       seuratreactive$MAGMA_selectcluster2 <- NULL
                       seuratreactive$textlabelMAGMA <- NULL
                       seuratreactive$MAGMA_N <- NULL
                       seuratreactive$GRCh <- NULL
                       seuratreactive$MAGMA_Pop <- NULL
                       seuratreactive$n_MAGMA <- NULL
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "MAGMA Computation Failed. One or more of your selected GWAS datasets were not formatted correctly, please try another GWAS dataset. Please also check if your dataset species is supported. Try linear enrichment only.", type = "error", confirmButtonCol = "#337ab7")
      seuratreactive$magma_dirs <- NULL
    })
  })
  
  observeEvent(input$MAGMAGO2GOGO, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     removeModal()
                     
                     source("MAGMA_fixpath.R")
                     source("MAGMA_getos.R")
                     source("MAGMA_messageparallel.R")
                     source("MAGMA_Gitdownload.R")
                     source("MAGMA_messager.R")
                     source("MAGMA_importmeta.R")
                     source("MAGMA_importmagmafiles.R")
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       exp <- seuratreactive$obj@assays$RNA@data
                     }
                     else {
                       exp <- seuratreactive$obj@assays$SCT@data
                     }
                     
                     cellInfo <- FetchData(object = seuratreactive$obj, vars = c("orig.ident", "ident", "seurat_clusters"))
                     
                     cellInfo <- cellInfo[cellInfo$orig.ident %in% input$MAGMA_samples2, ]
                     
                     cellInfo <- tibble::rownames_to_column(cellInfo, "rownames")
                     
                     if (input$MAGMA_selectcluster2 == "Seurat Clusters") {
                       cellInfo <- cellInfo %>% mutate(sample(100000, size = nrow(cellInfo), replace = TRUE)) %>% 
                         group_by(seurat_clusters) %>% 
                         top_n(input$n_MAGMA)
                     }
                     else {
                       cellInfo <- cellInfo %>% mutate(sample(100000, size = nrow(cellInfo), replace = TRUE)) %>% 
                         group_by(ident) %>% 
                         top_n(input$n_MAGMA)
                     }
                     
                     exp <- exp[, colnames(exp) %in% cellInfo$rownames]
                     
                     #Set up CTD
                     if (input$MAGMA_selectcluster2 == "Seurat Clusters") {
                       annotLevels <- list(as.character(cellInfo$seurat_clusters))
                     }
                     else {
                       annotLevels <- list(as.character(cellInfo$ident))
                     }
                     
                     if (seuratreactive$PAREACTOME == "sscrofa") {
                       CTD <- EWCE::generate_celltype_data(
                         exp = exp,
                         annotLevels = annotLevels,
                         groupName = "Seurat",
                         input_species = "hsapiens",
                         return_ctd=TRUE) 
                     }
                     else {
                       CTD <- EWCE::generate_celltype_data(
                         exp = exp,
                         annotLevels = annotLevels,
                         groupName = "Seurat",
                         input_species = seuratreactive$PAREACTOME,
                         return_ctd=TRUE) 
                     }
                     
                     if (input$MAGMA_Pop == "European") {
                       genesOutPath <- MAGMA.Celltyping::map_snps_to_genes(
                         path_formatted = seuratreactive$path_formatted,
                         population = "eur",
                         genome_build = input$GRCh,
                         N = input$MAGMA_N)
                     }
                     else if (input$MAGMA_Pop == "African") {
                       genesOutPath <- MAGMA.Celltyping::map_snps_to_genes(
                         path_formatted = seuratreactive$path_formatted,
                         population = "afr",
                         genome_build = input$GRCh,
                         N = input$MAGMA_N)
                     }
                     else if (input$MAGMA_Pop == "Ad Mixed American") {
                       genesOutPath <- MAGMA.Celltyping::map_snps_to_genes(
                         path_formatted = seuratreactive$path_formatted,
                         population = "amr",
                         genome_build = input$GRCh,
                         N = input$MAGMA_N)
                     }
                     else if (input$MAGMA_Pop == "East Asian") {
                       genesOutPath <- MAGMA.Celltyping::map_snps_to_genes(
                         path_formatted = seuratreactive$path_formatted,
                         population = "eas",
                         genome_build = input$GRCh,
                         N = input$MAGMA_N)
                     }
                     else if (input$MAGMA_Pop == "South Asian") {
                       genesOutPath <- MAGMA.Celltyping::map_snps_to_genes(
                         path_formatted = seuratreactive$path_formatted,
                         population = "sas",
                         genome_build = input$GRCh,
                         N = input$MAGMA_N)
                     }
                     
                     file.rename(paste0(genesOutPath, ".txt"), genesOutPath)
                     
                     seuratreactive$genesOutPath2 <- sub("[a-zA-Z0-9_]+.\\w+.35UP.10DOWN.genes.out$", "", genesOutPath)
                     
                     print(seuratreactive$genesOutPath2)
                     
                     table <- seuratreactive$meta[,c(1:2)]
                     colnames(table) <- c("GWAS", "Trait")
                     
                     if (!is.null(seuratreactive$magma_dirs)) {
                       combineddir <- c(seuratreactive$genesOutPath2, seuratreactive$magma_dirs)
                     }
                     else {
                       combineddir <- seuratreactive$genesOutPath2
                     }
                     
                     if (seuratreactive$PAREACTOME == "sscrofa") {
                       MAGMA_results <- MAGMA.Celltyping::celltype_associations_pipeline(
                         magma_dirs = combineddir,
                         ctd = CTD$ctd,
                         ctd_species = "hsapiens", 
                         ctd_name = "scRNA-seq_Dataset", 
                         run_linear = TRUE, 
                         run_top10 = TRUE,
                         force_new = TRUE,
                         save_dir = F,
                         standardise = T)
                     }
                     else {
                       MAGMA_results <- MAGMA.Celltyping::celltype_associations_pipeline(
                         magma_dirs = combineddir,
                         ctd = CTD$ctd,
                         ctd_species = seuratreactive$PAREACTOME, 
                         ctd_name = "scRNA-seq_Dataset", 
                         run_linear = TRUE, 
                         run_top10 = TRUE,
                         force_new = TRUE,
                         save_dir = NULL,
                         standardise = T)
                     }
                     
                     if (input$Linearmode1 == TRUE) {
                       merged_results <- MAGMA.Celltyping::merge_results(
                         MAGMA_results = MAGMA_results,
                         filetype = "ctAssocMerged")
                     }
                     else {
                       merged_results <- MAGMA.Celltyping::merge_results(
                         MAGMA_results = MAGMA_results,
                         filetype = "ctAssocsLinear")
                     }
                     
                     merged_results$GWAS <- sub("\\..+", "", merged_results$GWAS)
                     
                     merged_results$GWAS <- sub("^0$", input$textlabelMAGMA, merged_results$GWAS)
                     
                     seuratreactive$MAGMA_Final_Results <- full_join(merged_results, table)
                     
                     seuratreactive$MAGMA_Final_Results$Trait[is.na(seuratreactive$MAGMA_Final_Results$Trait)] <- seuratreactive$MAGMA_Final_Results$GWAS[which(is.na(seuratreactive$MAGMA_Final_Results$Trait))]
                     
                     seuratreactive$MAGMA_Final_Results <- na.omit(seuratreactive$MAGMA_Final_Results)
                     
                     incProgress(10/10)
                     
                     if (!is.null(seuratreactive$magma_dirs)) {
                       seuratreactive$MAGMA_samples2 <- input$MAGMA_samples2
                       seuratreactive$MAGMA_selectcluster2 <- input$MAGMA_selectcluster2
                       seuratreactive$textlabelMAGMA <- input$textlabelMAGMA
                       seuratreactive$MAGMA_N <- input$MAGMA_N
                       seuratreactive$GRCh <- input$GRCh
                       seuratreactive$MAGMA_Pop <- input$MAGMA_Pop
                       seuratreactive$n_MAGMA <- input$n_MAGMA
                       
                       seuratreactive$MAGMA_samples <- input$MAGMA_samples2
                       seuratreactive$MAGMA_GWAS <- input$MAGMA_GWAS
                       seuratreactive$MAGMA_selectcluster <- input$MAGMA_selectcluster2
                       seuratreactive$n_MAGMA2 <- input$n_MAGMA
                     }
                     else {
                       seuratreactive$MAGMA_samples2 <- input$MAGMA_samples2
                       seuratreactive$MAGMA_selectcluster2 <- input$MAGMA_selectcluster2
                       seuratreactive$textlabelMAGMA <- input$textlabelMAGMA
                       seuratreactive$MAGMA_N <- input$MAGMA_N
                       seuratreactive$GRCh <- input$GRCh
                       seuratreactive$MAGMA_Pop <- input$MAGMA_Pop
                       seuratreactive$n_MAGMA <- input$n_MAGMA
                       
                       seuratreactive$MAGMA_samples <- NULL
                       seuratreactive$MAGMA_GWAS <- NULL
                       seuratreactive$MAGMA_selectcluster <- NULL
                       seuratreactive$n_MAGMA2 <- NULL
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "MAGMA Computation Failed. One or more of your selected GWAS datasets were not formatted correctly, please try another GWAS dataset. Please also check if your dataset species is supported. Try Linear enrichment only.", type = "error", confirmButtonCol = "#337ab7")
      seuratreactive$genesOutPath2 <- NULL
    })
  })
  
  observeEvent(input$MAGMAGO2, {
    showModal(modalDialog(
      title = strong("MAGMA Own Data Input"),
      h3(HTML('<h3 style = "text-align:justify"> Please note that only HUMAN GWAS summary statistics are accepted. Other species are currently not supported. <br>'),
         p(style="text-align: center;", actionBttn("MAGMAGO2GOGO", "OK, I UNDERSTAND", size = "lg"))),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$MAGMAGO, {
    showModal(modalDialog(
      title = strong("MAGMA"),
      h3(HTML('<h3 style = "text-align:justify"> Please note that analysis may fail for non-human datasets. <br>'),
         p(style="text-align: center;", actionBttn("MAGMAGOGOGO", "OK, I UNDERSTAND", size = "lg"))),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  output$MAGMA_GWAS_table_Uploaded <- DT::renderDataTable({
    
    validate(
      need(!is.null(seuratreactive$UploadedGWAS), 'You must load your data first...'))
    
    seuratreactive$UploadedGWAS
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$MAGMA_GWAS_table <- DT::renderDataTable({
    seuratreactive$meta
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$MAGMA_HeatUI <- renderUI({
    plotlyOutput("MAGMA_Heat", height = (300 + input$HeightMAGMA*30))
  })
  
  output$MAGMA_Heat <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$MAGMA_Final_Results), 'Please compute your MAGMA GWAS Associations'))
    
    MAGMA.Celltyping::results_heatmap(
      merged_results = seuratreactive$MAGMA_Final_Results, 
      fdr_thresh = 1,
      y_var = "Trait",
      x_var = "Celltype_id",
      fill_var = "-log10(FDR)") + theme(text = element_text(size = input$fontMAGMA),
                                        axis.text.x = element_text(size = input$fontMAGMA, angle=90, vjust=0.5)) 
  })
  
  output$MAGMA_table <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$MAGMA_Final_Results), 'Please compute your MAGMA GWAS Associations'))
    
    seuratreactive$MAGMA_Final_Results[,c("Celltype_id", "GWAS", "Trait", "EnrichmentMode", "P", "log10p", "FDR")]
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  ####TAB 11 - Trajectory Analysis##### 
  
  observe({
    if (!is.null(seuratreactive$TA_pickercluster) && is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = T),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = T),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = T),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observeEvent(input$startbuttonTA, {
    library(monocle3)
    library(SeuratWrappers)
    
    seuratreactive$plot.data$Pseudotime <- NULL
    
    if (input$iscollapseboxTA == FALSE) {
      js$collapse("TABoxIntro")
    }
    shinyjs::show("TABox1")
    shinyjs::show("TABox4")
    shinyjs::show("TABox5")
    shinyjs::show("TABox6")
    shinyjs::show("TABox7")
    shinyjs::show("TABox8")
    shinyjs::show("TABox9")
    shinyjs::show("TABox10")
    shinyjs::hide("TABox2")
    shinyjs::hide("TABox3")
    
    updatePickerInput(session, "TAselect", selected = "")
    
    updateVirtualSelect('MonocleInput', choices = c(rownames(seuratreactive$obj)))
    
  }, ignoreInit = T)
  
  observe({
    if (input$TAselect == "Cell cluster/Cell type") {
      shinyjs::show("TABox2")
      shinyjs::hide("TABox3")
    }
    else if (input$TAselect == "") {
      shinyjs::hide("TABox2")
      shinyjs::hide("TABox3")
    }
    else if (input$TAselect == "Lasso select") {
      shinyjs::hide("TABox2")
      shinyjs::show("TABox3")
    }
  })
  
  observe({
    if (is.null(input$TA_pickercluster)) {
      shinyjs::disable("TAGO")
    }
    else{
      shinyjs::enable("TAGO")
    }
  })
  
  observe({
    if (is.null(selectedTA()$key)) {
      shinyjs::disable("TAGO2")
    }
    else{
      shinyjs::enable("TAGO2")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$reductionCC1)) {
      if (input$TA_picker == "Seurat Clusters"){
        updateVirtualSelect('TA_pickercluster', choices = c(unique(levels(seuratreactive$obj$seurat_clusters)))
                            , selected = seuratreactive$TA_pickercluster)
      }
      else {
        updateVirtualSelect('TA_pickercluster', choices = c(unique(levels(seuratreactive$obj)))
                            , selected = seuratreactive$TA_pickercluster)
      }
    }
  })
  
  output$plotTAUI <- renderUI({
    plotlyOutput("plotTA", height = (400 + input$TAplotheight*40))
  })
  
  output$plotTA <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    if(input$plotclustersTA == "Seurat Clusters" | input$plotclustersTA == "Cell Cycle Phase" | input$plotclustersTA == "Labelled Cells" |
       input$plotclustersTA == "Sample") {
      if(input$radioTA == "UMAP") {
        plot_ly(data = seuratreactive$plot.data, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected") %>% layout(font = list(size = input$labelsizeTA))
      }
      else if (input$radioTA == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
        
      }
      else if (input$radioTA == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
        
      }
      else if (input$radioTA == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
        
      }
    }
    else {
      if(input$radioTA == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
        
      }
      else if (input$radioTA == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
        
      }
      else if (input$radioTA == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
      }
      else if (input$radioTA == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersTA],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeTA),
                opacity = input$opacityTA,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`,
                                   "<br>Pseudotime: ", seuratreactive$plot.data$Pseudotime),
                source = "plotTAselected"
        ) %>% layout(font = list(size = input$labelsizeTA))
      }
    }
  })
  
  selectedTA <- reactive({
    req(seuratreactive$plot.data)
    event_data("plotly_selected", source = "plotTAselected")
  })
  
  output$selectedplotTA <- DT::renderDataTable({
    seuratreactive$lassoTA <- subset(seuratreactive$plot.data.table, seuratreactive$plot.data.table$Cell %in% selectedTA()$key)
    seuratreactive$lassoTA    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  observeEvent(input$TAGO2, {
    tryCatch({
      withProgress(message = 'Performing trajectory analysis...',
                   value = 0, {
                     
                     if (!is.null(selectedTA()$key) |
                         !is.null(seuratreactive$TAselectedcells)) {
                       
                       seuratreactive$TAselectedcells <- selectedTA()$key
                       
                       incProgress(1/10)
                       
                       #Calculating trajectories with monocle3
                       seuratreactive$seurat_monocle <- as.cell_data_set(seuratreactive$obj)
                       
                       seuratreactive$seurat_monocle <- cluster_cells(cds = seuratreactive$seurat_monocle, reduction_method = "UMAP")
                       
                       seuratreactive$seurat_monocle <- learn_graph(seuratreactive$seurat_monocle)
                       incProgress(3/10)
                       
                       # order cells
                       seuratreactive$seurat_monocle <- order_cells(seuratreactive$seurat_monocle, reduction_method = "UMAP", root_cells = seuratreactive$TAselectedcells)
                       
                       seuratreactive$TA_pickercluster <- "lasso select"
                       
                       # plot trajectories colored by pseudotime
                       plot_cells(
                         cds = seuratreactive$seurat_monocle,
                         color_cells_by = "pseudotime"
                       )
                       
                       incProgress(5/10)
                       
                       seuratreactive$obj <- AddMetaData(
                         object = seuratreactive$obj,
                         metadata = seuratreactive$seurat_monocle@principal_graph_aux@listData$UMAP$pseudotime,
                         col.name = "Pseudotime"
                       )
                       
                       IdentifiersTA <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                        "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident",
                                                                                        "Pseudotime"))
                       
                       is.na(IdentifiersTA) <- do.call(cbind,lapply(IdentifiersTA, is.infinite))
                       
                       IdentifiersTA[c("percent.ribo", "percent.mt", "Pseudotime")] <- signif(IdentifiersTA[c("percent.ribo", "percent.mt", "Pseudotime")], digits = 4)
                       
                       incProgress(8/10)
                       
                       colnames(IdentifiersTA) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                    "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                    "Cell Cycle Phase", "Labelled Cells", "Pseudotime")
                       
                       seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersTA, by = "row.names", all = T)
                       
                       rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                       
                       seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                       
                       incProgress(9/10)
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Trajectory Analysis failed, try select a different root.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  
  observeEvent(input$TAGO, {
    tryCatch({
      withProgress(message = 'Performing trajectory analysis...',
                   value = 0, {
                     incProgress(1/10)
                     
                     #Calculating trajectories with monocle3
                     seuratreactive$seurat_monocle <- as.cell_data_set(seuratreactive$obj)
                     
                     seuratreactive$seurat_monocle <- cluster_cells(cds = seuratreactive$seurat_monocle, reduction_method = "UMAP")
                     
                     seuratreactive$seurat_monocle <- learn_graph(seuratreactive$seurat_monocle)
                     
                     incProgress(3/10)
                     
                     # order cells
                     if (input$TA_picker == "Seurat Clusters") {
                       seuratreactive$TAselectedcells <- colnames(seuratreactive$obj[,seuratreactive$obj$seurat_clusters %in% input$TA_pickercluster])
                     }
                     else {
                       seuratreactive$TAselectedcells <- colnames(seuratreactive$obj[,seuratreactive$obj@active.ident %in% input$TA_pickercluster])
                     }
                     
                     seuratreactive$seurat_monocle <- order_cells(seuratreactive$seurat_monocle, reduction_method = "UMAP", root_cells = seuratreactive$TAselectedcells)
                     
                     seuratreactive$TA_pickercluster <- input$TA_pickercluster
                     
                     # plot trajectories colored by pseudotime
                     plot_cells(
                       cds = seuratreactive$seurat_monocle,
                       color_cells_by = "pseudotime"
                     )
                     
                     incProgress(5/10)
                     
                     seuratreactive$obj <- AddMetaData(
                       object = seuratreactive$obj,
                       metadata = seuratreactive$seurat_monocle@principal_graph_aux@listData$UMAP$pseudotime,
                       col.name = "Pseudotime"
                     )
                     
                     IdentifiersTA <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                      "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident",
                                                                                      "Pseudotime"))
                     
                     is.na(IdentifiersTA) <- do.call(cbind,lapply(IdentifiersTA, is.infinite))
                     
                     IdentifiersTA[c("percent.ribo", "percent.mt", "Pseudotime")] <- signif(IdentifiersTA[c("percent.ribo", "percent.mt", "Pseudotime")], digits = 4)
                     
                     incProgress(8/10)
                     
                     colnames(IdentifiersTA) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                  "Cell Cycle Phase", "Labelled Cells", "Pseudotime")
                     
                     seuratreactive$plot.data <- merge(seuratreactive$plot.data.original, IdentifiersTA, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.data) <- seuratreactive$plot.data$Row.names
                     
                     seuratreactive$plot.data <- seuratreactive$plot.data[,-1]
                     
                     incProgress(9/10)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Trajectory Analysis failed, try select a different root.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  observeEvent(input$GO_TA, {
    showModal(modalDialog(
      title = strong("Gene changes across pseudotime"),
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete depending on the size of the dataset(s). <br><br><br>'),
         p(style="text-align: center;", actionBttn("GOGO_TA", "OK, I UNDERSTAND", size = "lg"))),
      easyClose = TRUE,
      footer = NULL
    ))
  })
  
  observeEvent(input$GOGO_TA, {
    tryCatch({
      withProgress(message = 'Calculating ...',
                   value = 0, {
                     incProgress(1/10)
                     
                     removeModal()
                     
                     seuratreactive$Monocle3_genes <- graph_test(seuratreactive$seurat_monocle, neighbor_graph="principal_graph", cores=10)
                     
                     incProgress(9/10)
                     seuratreactive$Monocle3_genes <- seuratreactive$Monocle3_genes[,-1]
                     
                     incProgress(10/10)
                     seuratreactive$Monocle3_genes <- tibble::rownames_to_column(seuratreactive$Monocle3_genes, "Gene Name")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Trajectory Analysis failed, try select a different root.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$plotmonocleTAUI <- renderUI({
    plotlyOutput("plotmonocleTA", height = (400 + input$TA2plotheight*40))
  })
  
  output$plotmonocleTA <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$seurat_monocle), "Please select your root cells and compute trajectories"))
    
    if (input$reductionmethodTA == "2D") {
      plot_cells(
        cds = seuratreactive$seurat_monocle,
        color_cells_by = "pseudotime",
        reduction_method = "UMAP",
        cell_size = input$sizemonocleTA/4,
        graph_label_size = input$sizemonocleTA,
        alpha = input$opacitymonocleTA
      ) + theme(text = element_text(size=input$labelsizemonocleTA),
                axis.text.x= element_text(size = input$labelsizemonocleTA),
                axis.text.y= element_text(size = input$labelsizemonocleTA))
    }
    else if (input$reductionmethodTA == "3D") {
      plot_cells_3d(
        cds = seuratreactive$seurat_monocle,
        color_cells_by = "pseudotime",
        reduction_method = "UMAP",
        cell_size = input$sizemonocleTA*20,
        alpha = input$opacitymonocleTA
      ) + theme(text = element_text(size=input$labelsizemonocleTA),
                axis.text.x= element_text(size = input$labelsizemonocleTA),
                axis.text.y= element_text(size = input$labelsizemonocleTA))
    }
  })
  
  output$monocle3DEG <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$seurat_monocle), "Please select your root cells and compute trajectories"))
    validate(
      need(!is.null(seuratreactive$Monocle3_genes), "Please compute your gene changes across pseudotime, start analysis by pressing the START button in the sidebar"))
    
    seuratreactive$Monocle3_genes
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$monocle3DEGvizUI <- renderUI({
    plotOutput("monocle3DEGviz", height = (300 + input$HeightDEGviz*30))
  })
  
  output$monocle3DEGviz <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$seurat_monocle), "Please select your root cells and compute trajectories"))
    validate(
      need(!is.null(input$MonocleInput), "Please select your gene of interest"))
    
    fData(seuratreactive$seurat_monocle)$gene_short_name <- rownames(fData(seuratreactive$seurat_monocle))
    
    my_genes <- row.names(subset(fData(seuratreactive$seurat_monocle), gene_short_name %in% input$MonocleInput)) 
    cds_subset <- seuratreactive$seurat_monocle[my_genes,]
    plot_genes_in_pseudotime(cds_subset, color_cells_by = "monocle3_pseudotime") + theme_bw(base_size = input$fontsizemonocle3select) + guides(colour = FALSE)
  })
  
  output$monocle3vizUI <- renderUI({
    plotOutput("monocle3viz", height = (300 + input$Heightpseudotimeviz*30))
  })
  
  output$monocle3viz <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$seurat_monocle), "Please select your root cells and compute trajectories"))
    
    if (input$monocle3pseudo == "Seurat Clusters") {
      seuratreactive$seurat_monocle$monocle3_pseudotime <- pseudotime(seuratreactive$seurat_monocle)
      data.pseudo <- as.data.frame(colData(seuratreactive$seurat_monocle))
      
      ggplot(data.pseudo, aes(monocle3_pseudotime, reorder(seurat_clusters, monocle3_pseudotime), fill = seurat_clusters)) + 
        geom_boxplot() + 
        ylab("Seurat Clusters") +
        xlab("Monocle3 Pseudotime") +
        theme_bw(base_size = input$fontsizepseudotimeviz)
    }
    else {
      seuratreactive$seurat_monocle$monocle3_pseudotime <- pseudotime(seuratreactive$seurat_monocle)
      data.pseudo <- as.data.frame(colData(seuratreactive$seurat_monocle))
      
      ggplot(data.pseudo, aes(monocle3_pseudotime, reorder(ident, monocle3_pseudotime), fill = ident)) + 
        geom_boxplot() + 
        ylab("Labelled Cells") + 
        xlab("Monocle3 Pseudotime") +
        theme_bw(base_size = input$fontsizepseudotimeviz)
    }
  })
  
  
  ####TAB 12 - CellChat ####
  
  observe({
    if (!is.null(seuratreactive$Chatdb) && is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = T),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = T),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = T),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  observeEvent(input$startChat, {
    showModal(modalDialog(
      title = strong("CellChat Settings"),
      pickerInput("ChatChoose", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Please select from the following')), 
                  choices = list("Please select from the following" = "", "Single Dataset Analysis", "Comparison Between Two Datasets"),
                  selected = "",
                  width = "100%"),
      virtualSelectInput("ChatSample", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Select Sample (Single Dataset Analysis)')),
                         choices = NULL,
                         multiple = T,
                         search = T,
                         hideClearButton = F,
                         disableSelectAll = F,
                         width = "100%"),
      virtualSelectInput("ChatMultiple1Sample", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Select Sample 1 (Comparison Between Two Datasets)')),
                         choices = NULL,
                         multiple = T,
                         search = T,
                         hideClearButton = F,
                         disableSelectAll = F,
                         width = "100%"),
      virtualSelectInput("ChatMultiple2Sample", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Select Sample 2 (Comparison Between Two Datasets)')),
                         choices = NULL,
                         multiple = T,
                         search = T,
                         hideClearButton = F,
                         disableSelectAll = F,
                         width = "100%"),
      pickerInput("Chat_selectcluster", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top: 25px;">Select cells by')),
                  choices = list("Seurat Clusters", "Labelled Cells"),
                  width = "100%"),
      pickerInput("Chatdb", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Select Database')),
                  choices = list("Human", "Mouse"),
                  width = "100%"),
      numericInput("Chatpval", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-50px;">p-value threshold')),
                   min = 0, max = 1,
                   value = 0.05,
                   width = "100%"),
      numericInput("Chat_number", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-40px;">Filter out the cell-cell communication in groups with less than X cells')), 
                   value = 10, min = 1, step = 1,
                   width = "100%"),
      
      p(style="text-align: center;", actionBttn("startbuttonChat1", "APPLY", size = "lg")),
      p(style="text-align: center;", actionBttn("startbuttonChat2", "APPLY", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of hours to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startChat, {
    
    updatePickerInput(session, "ChatChoose", selected = seuratreactive$ChatChoose)
    updateVirtualSelect('ChatSample', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                        selected = seuratreactive$ChatSample)
    updatePickerInput(session, "Chatdb", selected = seuratreactive$Chatdb)
    updateVirtualSelect('ChatMultiple1Sample', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                        selected = seuratreactive$ChatMultiple1Sample)
    updateVirtualSelect('ChatMultiple2Sample', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                        selected = seuratreactive$ChatMultiple2Sample)
    
    
    if (!is.null(input$ChatChoose)) {
      if (input$ChatChoose == "") {
        shinyjs::disable("startbuttonChat1")
        shinyjs::disable("startbuttonChat2")
        shinyjs::hide("ChatSample")
        shinyjs::hide("ChatMultiple1Sample")
        shinyjs::hide("ChatMultiple2Sample")
        shinyjs::hide("startbuttonChat2")
        shinyjs::show("startbuttonChat1")
      }
      else if (input$ChatChoose == "Single Dataset Analysis") {
        shinyjs::show("ChatSample")
        shinyjs::hide("ChatMultiple1Sample")
        shinyjs::hide("ChatMultiple2Sample")
        shinyjs::show("startbuttonChat1")
        shinyjs::hide("startbuttonChat2")
        
        if (is.null(input$Chat_selectcluster) | is.null(input$ChatSample) | is.na(input$Chatpval) | is.na(input$Chatpval)) {
          shinyjs::disable("startbuttonChat2")
          shinyjs::disable("startbuttonChat1")
        }
        else {
          shinyjs::disable("startbuttonChat2")
          shinyjs::enable("startbuttonChat1")
        }
        
      }
      else if (input$ChatChoose == "Comparison Between Two Datasets") {
        shinyjs::hide("ChatSample")
        shinyjs::show("ChatMultiple1Sample")
        shinyjs::show("ChatMultiple2Sample")
        shinyjs::hide("startbuttonChat1")
        shinyjs::show("startbuttonChat2")
        if (!is.null(input$ChatMultiple1Sample)) {
          if (paste(input$ChatMultiple1Sample, collapse = "") == paste(input$ChatMultiple2Sample, collapse = "") | is.null(input$ChatMultiple1Sample) | is.null(input$ChatMultiple2Sample) |is.null(input$Chat_selectcluster) |
              is.na(input$Chatpval) | is.na(input$Chatpval)) {
            shinyjs::disable("startbuttonChat2")
            shinyjs::disable("startbuttonChat1")
          }
          else {
            shinyjs::enable("startbuttonChat2")
            shinyjs::disable("startbuttonChat1")
          }
        }
      }
    }
    
    if(!is.null(seuratreactive$reductionCC1)) {
      if (identical(seuratreactive$obj@active.ident, seuratreactive$obj$seurat_clusters)) {
        if (!is.null(seuratreactive$Chat_selectcluster)) {
          updatePickerInput(session, 'Chat_selectcluster', choices = list("Seurat Clusters"), selected = seuratreactive$Chat_selectcluster)
        }
        else {
          updatePickerInput(session, 'Chat_selectcluster', choices = list("Seurat Clusters"))
        }
      }
      else {
        if (!is.null(seuratreactive$Chat_selectcluster)) {
          updatePickerInput(session, 'Chat_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"), selected = seuratreactive$Chat_selectcluster)
        }
        else {
          updatePickerInput(session, 'Chat_selectcluster', choices = list("Seurat Clusters", "Labelled Cells"))
        }
      }
    }
    
  })
  
  observe({
    if (!is.null(input$ChatChoose)) {
      if (input$ChatChoose == "") {
        shinyjs::disable("startbuttonChat1")
        shinyjs::disable("startbuttonChat2")
        shinyjs::hide("ChatSample")
        shinyjs::hide("ChatMultiple1Sample")
        shinyjs::hide("ChatMultiple2Sample")
        shinyjs::hide("startbuttonChat2")
        shinyjs::show("startbuttonChat1")
      }
      else if (input$ChatChoose == "Single Dataset Analysis") {
        shinyjs::show("ChatSample")
        shinyjs::hide("ChatMultiple1Sample")
        shinyjs::hide("ChatMultiple2Sample")
        shinyjs::show("startbuttonChat1")
        shinyjs::hide("startbuttonChat2")
        
        if (is.null(input$Chat_selectcluster) | is.null(input$ChatSample) | is.na(input$Chatpval) | is.na(input$Chatpval)) {
          shinyjs::disable("startbuttonChat2")
          shinyjs::disable("startbuttonChat1")
        }
        else {
          shinyjs::disable("startbuttonChat2")
          shinyjs::enable("startbuttonChat1")
        }
        
      }
      else if (input$ChatChoose == "Comparison Between Two Datasets") {
        shinyjs::hide("ChatSample")
        shinyjs::show("ChatMultiple1Sample")
        shinyjs::show("ChatMultiple2Sample")
        shinyjs::hide("startbuttonChat1")
        shinyjs::show("startbuttonChat2")
        if (!is.null(input$ChatMultiple1Sample)) {
          if (paste(input$ChatMultiple1Sample, collapse = "") == paste(input$ChatMultiple2Sample, collapse = "") | is.null(input$ChatMultiple1Sample) | is.null(input$ChatMultiple2Sample) |is.null(input$Chat_selectcluster) |
              is.na(input$Chatpval) | is.na(input$Chatpval)) {
            shinyjs::disable("startbuttonChat2")
            shinyjs::disable("startbuttonChat1")
          }
          else {
            shinyjs::enable("startbuttonChat2")
            shinyjs::disable("startbuttonChat1")
          }
        }
      }
      
    }
  })
  
  observe({
    if (!is.null(input$Dataset_SelectMM) && !is.null(input$Dataset_SelectMM2)) {
      if (input$Dataset_SelectMM == input$Dataset_SelectMM2) {
        shinyjs::disable("GOCHAT")
      }
      else {
        shinyjs::enable("GOCHAT")
      }
    }
  })
  
  observe({
    if (input$radioRolesChatM == "Heatmap") {
      shinyjs::show("signallingroles_MM")
    }
    else {
      shinyjs::hide("signallingroles_MM")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) { 
      updateVirtualSelect('ChatSample', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                          selected = seuratreactive$ChatSample)
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj)) { 
      updateVirtualSelect('ChatMultiple1Sample', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                          selected = seuratreactive$ChatMultiple1Sample)
      
      updateVirtualSelect('ChatMultiple2Sample', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                          selected = seuratreactive$ChatMultiple2Sample)
    }
  })
  
  observe({
    if (!is.null(seuratreactive$obj) && !is.null(seuratreactive$cellchatM)) { 
      updatePickerInput(session, 'Dataset_SelectMM', choices = c(names(seuratreactive$cellchatM@idents)[1:2]),
                        selected = seuratreactive$Dataset_SelectMM)
      
      updatePickerInput(session, 'Dataset_SelectMM2', choices = c(names(seuratreactive$cellchatM@idents)[1:2]),
                        selected = seuratreactive$Dataset_SelectMM2)
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchat)) {
      updatePickerInput(session, 'Communication_Select', choices = c(rownames(seuratreactive$cellchat@net$weight)))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchat)) {
      updateVirtualSelect('source_select', choices = c(rownames(seuratreactive$cellchat@net$weight)))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchat)) {
      updateVirtualSelect('target_select', choices = c(rownames(seuratreactive$cellchat@net$weight)))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updatePickerInput(session, 'Communication_SelectM1', choices = c(rownames(seuratreactive$cellchatM@net[[1]]$weight)))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updatePickerInput(session, 'Communication_SelectM2', choices = c(rownames(seuratreactive$cellchatM@net[[2]]$weight)))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updateVirtualSelect('source_selectM', choices = unique(c(rownames(seuratreactive$cellchatM@net[[1]]$weight), rownames(seuratreactive$cellchatM@net[[2]]$weight))))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updateVirtualSelect('source_selectMM', choices = unique(c(rownames(seuratreactive$cellchatM@net[[1]]$weight), rownames(seuratreactive$cellchatM@net[[2]]$weight))))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updateVirtualSelect('target_selectM', choices = unique(c(rownames(seuratreactive$cellchatM@net[[1]]$weight), rownames(seuratreactive$cellchatM@net[[2]]$weight))))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updateVirtualSelect('target_selectMM', choices = unique(c(rownames(seuratreactive$cellchatM@net[[1]]$weight), rownames(seuratreactive$cellchatM@net[[2]]$weight))))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchat)) {
      updateVirtualSelect('Signalling_Select', choices = c(seuratreactive$cellchat@netP$pathways))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      if (input$SignallingDataset_SelectM == "Dataset 1") {
        updateVirtualSelect('Signalling_SelectM', choices = c(seuratreactive$cellchatM@netP[[1]]$pathways))
      }
      else if (input$SignallingDataset_SelectM == "Dataset 2") {
        updateVirtualSelect('Signalling_SelectM', choices = c(seuratreactive$cellchatM@netP[[2]]$pathways))
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchat)) {
      updateVirtualSelect('Ligand_Select', choices = c(seuratreactive$cellchat@netP$pathways))
    }
  })
  
  observe({
    if (!is.null(seuratreactive$cellchatM)) {
      updateVirtualSelect('Ligand_SelectM', choices = c(seuratreactive$cellchatM@netP[[1]]$pathways))
    }
  })
  
  output$Interactions_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (input$checkboxCommunication == TRUE) {
      par(mfrow = c(1,2), xpd=TRUE)
      netVisual_circle(seuratreactive$cellchat@net$count, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Number of interactions")
      netVisual_circle(seuratreactive$cellchat@net$weight, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
    }
    else {
      par(mfrow = c(1,2), xpd=TRUE)
      
      mat1 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$count), ncol = ncol(seuratreactive$cellchat@net$count), dimnames = dimnames(seuratreactive$cellchat@net$count))
      mat1[input$Communication_Select, ] <- seuratreactive$cellchat@net$count[input$Communication_Select, ]
      netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$count), title.name = paste(input$Communication_Select, "Number of interactions"))
      
      mat2 <- matrix(0, nrow = nrow(seuratreactive$cellchat@net$weight), ncol = ncol(seuratreactive$cellchat@net$weight), dimnames = dimnames(seuratreactive$cellchat@net$weight))
      mat2[input$Communication_Select, ] <- seuratreactive$cellchat@net$weight[input$Communication_Select, ]
      netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$cellchat@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$cellchat@net$weight), title.name = paste(input$Communication_Select, "Interaction weights/strength"))
    }
  })
  
  output$Signalling_CellChatUI <- renderUI({
    plotOutput("Signalling_CellChat", height = (400 + input$HeightSignalling_Chat*30))
  })
  
  output$Signalling_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioChat == "Hierarchy Plot") {
      par(mfrow=c(1,1))
      netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select,  vertex.receiver = seq(1,length(unique(seuratreactive$cellchat@idents))/2), layout = "hierarchy", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10)
    }
    else if (input$radioChat == "Circle Plot") {
      par(mfrow=c(1,1))
      netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "circle", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10)
    }
    else if (input$radioChat == "Chord Plot") {
      par(mfrow=c(1,1))
      netVisual_aggregate(seuratreactive$cellchat, signaling = input$Signalling_Select, layout = "chord", pt.title = input$sizeSignalling, vertex.label.cex = input$sizeSignalling/10)
    }
    else if (input$radioChat == "Heatmap") {
      par(mfrow=c(1,1))
      netVisual_heatmap(seuratreactive$cellchat, signaling = input$Signalling_Select, color.heatmap = "Reds", font.size = input$sizeSignalling, font.size.title = input$sizeSignalling + 2)
    }
    else if (input$radioChat == "Signalling Roles") {
      par(mfrow=c(1,1))
      netAnalysis_signalingRole_network(seuratreactive$cellchat, signaling = input$Signalling_Select, width = input$WidthSignalling_Chat, height = input$HeightSignalling_Chat, font.size = input$sizeSignalling)
    }
  })
  
  output$Ligand_CellChatUI <- renderUI({
    plotOutput("Ligand_CellChat", height = (400 + input$HeightLigand*30))
  })
  
  output$Ligand_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (!is.null(input$Ligand_Select)) {
      if (input$radioLigandChat == "Bar Graph") {
        par(mfrow=c(1,1))
        netAnalysis_contribution(seuratreactive$cellchat, signaling = input$Ligand_Select, font.size = input$sizeLigand, font.size.title = input$sizeLigand)
      }
      else if (input$radioLigandChat == "Dot Plot") {
        par(mfrow=c(1,1))
        netVisual_bubble(seuratreactive$cellchat, sources.use = input$source_select, targets.use = input$target_select, signaling = input$Ligand_Select, remove.isolate = FALSE, font.size = input$sizeLigand, font.size.title = input$sizeLigand, angle.x = 90)
      }
      else if (input$radioLigandChat == "Chord Plot") {
        par(mfrow=c(1,1))
        netVisual_chord_gene(seuratreactive$cellchat, signaling = input$Ligand_Select, lab.cex = input$sizeLigand/10)
      }
      else if (input$radioLigandChat == "Gene Expression") {
        par(mfrow=c(1,1))
        plotGeneExpression(seuratreactive$cellchat, signaling = input$Ligand_Select)
      }
    }
  })
  
  output$Roles_CellChatUI <- renderUI({
    plotOutput("Roles_CellChat", height = (400 + input$HeightRoles*30))
  })
  
  output$Roles_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioRolesChat == "Heatmap") {
      ht1 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
      ht2 <- netAnalysis_signalingRole_heatmap(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizeRoles, font.size.title = input$sizeRoles + 2, width = input$WidthRoles, height = input$HeightRoles)
      ht1 + ht2
    }
    else if (input$radioRolesChat == "Scatter Plot") {
      netAnalysis_signalingRole_scatter(seuratreactive$cellchat, font.size = input$sizeRoles, font.size.title = input$sizeRoles, label.size	= input$sizeRoles - 8)
    }
  })
  
  output$Patterns_CellChatUI <- renderUI({
    plotOutput("Patterns_CellChat", height = (400 + input$HeightPatterns*30))
  })
  
  output$Patterns_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioPatternsChat == "River Plot") {
      netAnalysis_river(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2)
    }
    else if (input$radioPatternsChat == "Dot Plot") {
      netAnalysis_dot(seuratreactive$cellchat, pattern = "incoming", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2)
    }
  })
  
  output$Patterns_CellChatUI2 <- renderUI({
    plotOutput("Patterns_CellChat2", height = (400 + input$HeightPatterns*30))
  })
  
  output$Patterns_CellChat2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioPatternsChat == "River Plot") {
      netAnalysis_river(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns/8 + 2, font.size.title = input$sizePatterns + 2)
    }
    else if (input$radioPatternsChat == "Dot Plot") {
      netAnalysis_dot(seuratreactive$cellchat, pattern = "outgoing", font.size = input$sizePatterns, font.size.title = input$sizePatterns + 2)
    }
  })
  
  output$Manifold_CellChatUI <- renderUI({
    plotOutput("Manifold_CellChat", height = (400 + input$HeightManifold*30))
  })
  
  output$Manifold_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchat), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioManifoldChat == "Functional Classification") {
      netVisual_embedding(seuratreactive$cellchat, type = "functional", label.size = input$sizeManifold)
    }
    else if (input$radioManifoldChat == "Structural Classification") {
      netVisual_embedding(seuratreactive$cellchat, type = "structural", label.size = input$sizeManifold)
    }
  })
  
  observeEvent(input$GOPATTERNS, {
    tryCatch({
      withProgress(message = 'Identifying incoming and outgoing communication patterns...',
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$cellchat <- identifyCommunicationPatterns(seuratreactive$cellchat, pattern = "incoming", k = input$Patterns_Select, heatmap.show = F)
                     incProgress(5/10)
                     seuratreactive$cellchat <- identifyCommunicationPatterns(seuratreactive$cellchat, pattern = "outgoing", k = input$Patterns_Select, heatmap.show = F)
                     incProgress(10/10)
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "CellChat incoming and outgoing communcation pattern identification failed. Please increase/decrease the number of input patterns.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  observeEvent(input$startbuttonChat1, {
    library(CellChat)
    library(patchwork)
    library(NMF)
    library(ggalluvial)
    library(ComplexHeatmap)
    
    tryCatch({
      withProgress(message = 'Performing cell-cell communication network analysis...',
                   value = 0, {
                     incProgress(1/20)
                     
                     removeModal()
                     
                     split_seurat <- SplitObject(seuratreactive$obj, split.by = "orig.ident")
                     
                     split_seurat <- split_seurat[input$ChatSample]
                     
                     split_seurat <- Reduce(merge, split_seurat)
                     
                     if (input$Chat_selectcluster == "Seurat Clusters") {
                       
                       split_seurat <- SetIdent(split_seurat, value = "seurat_clusters")
                       
                       new_ids = paste0("Seurat_Cluster_", levels(split_seurat))
                       names(new_ids) <- levels(split_seurat)
                       split_seurat <- RenameIdents(split_seurat, new_ids)
                     }
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       seuratreactive$cellchat <- createCellChat(split_seurat, assay = "RNA", group.by = "ident")
                     }
                     else {
                       seuratreactive$cellchat <- createCellChat(split_seurat, assay = "SCT", group.by = "ident")
                     }
                     
                     seuratreactive$cellchat <- setIdent(seuratreactive$cellchat, ident.use = "ident")
                     
                     if (input$Chatdb == "Human") {
                       seuratreactive$cellchat@DB <- CellChatDB.human
                     }
                     else if (input$Chatdb == "Mouse") {
                       seuratreactive$cellchat@DB <- CellChatDB.mouse
                     }
                     
                     incProgress(2/20)
                     
                     seuratreactive$cellchat <- subsetData(seuratreactive$cellchat) # This step is necessary even if using the whole database
                     
                     seuratreactive$cellchat <- identifyOverExpressedGenes(seuratreactive$cellchat, thresh.p = input$Chatpval)
                     incProgress(7/20)
                     seuratreactive$cellchat <- identifyOverExpressedInteractions(seuratreactive$cellchat)
                     incProgress(8/20)
                     # project gene expression data onto Protein-Protein Interaction network
                     if (input$Chatdb == "Human") {
                       seuratreactive$cellchat <- projectData(seuratreactive$cellchat, PPI.human)
                     }
                     else if (input$Chatdb == "Mouse") {
                       seuratreactive$cellchat <- projectData(seuratreactive$cellchat, PPI.mouse)
                     }
                     incProgress(12/20)
                     
                     seuratreactive$cellchat <- computeCommunProb(seuratreactive$cellchat)
                     # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
                     seuratreactive$cellchat <- filterCommunication(seuratreactive$cellchat, min.cells = input$Chat_number)
                     incProgress(16/20)
                     
                     #Interactions
                     seuratreactive$cellchat <- computeCommunProbPathway(seuratreactive$cellchat, thresh = input$Chatpval)
                     seuratreactive$cellchat <- aggregateNet(seuratreactive$cellchat, thresh = input$Chatpval)
                     
                     seuratreactive$cellchat <- netAnalysis_computeCentrality(seuratreactive$cellchat, slot.name = "netP", thresh = input$Chatpval)
                     
                     seuratreactive$cellchat <- computeNetSimilarity(seuratreactive$cellchat, type = "functional")
                     seuratreactive$cellchat <- netEmbedding(seuratreactive$cellchat, type = "functional", umap.method = "uwot")
                     seuratreactive$cellchat <- netClustering(seuratreactive$cellchat, type = "functional", do.plot = F, nCores = 10)
                     
                     incProgress(18/20)
                     seuratreactive$cellchat <- computeNetSimilarity(seuratreactive$cellchat, type = "structural")
                     seuratreactive$cellchat <- netEmbedding(seuratreactive$cellchat, type = "structural", umap.method = "uwot")
                     seuratreactive$cellchat <- netClustering(seuratreactive$cellchat, type = "structural", do.plot = F, nCores = 10)
                     
                     seuratreactive$cellchat <- identifyCommunicationPatterns(seuratreactive$cellchat, pattern = "incoming", k = input$Patterns_Select, heatmap.show = F)
                     seuratreactive$cellchat <- identifyCommunicationPatterns(seuratreactive$cellchat, pattern = "outgoing", k = input$Patterns_Select, heatmap.show = F)
                     
                     incProgress(20/20)
                     
                     seuratreactive$cellchatM <- NULL
                     seuratreactive$object.list <- NULL
                     seuratreactive$ChatChoose <- input$ChatChoose
                     seuratreactive$ChatSample <- input$ChatSample
                     seuratreactive$Chatdb <- input$Chatdb
                     seuratreactive$ChatMultiple1Sample <- NULL
                     seuratreactive$ChatMultiple2Sample <- NULL
                     seuratreactive$Chat_selectcluster <- input$Chat_selectcluster
                     seuratreactive$Chat_number <- input$Chat_number
                     seuratreactive$Chatpval <- input$Chatpval
                     
                     js$collapse("ChatBoxIntro")
                     
                     shinyjs::show("ChatBox1")
                     shinyjs::show("ChatBox2")
                     shinyjs::show("ChatBox3")
                     shinyjs::show("ChatBox4")
                     shinyjs::show("ChatBox5")
                     shinyjs::show("ChatBox6")
                     shinyjs::hide("ChatBox7")
                     shinyjs::hide("ChatBox8")
                     shinyjs::hide("ChatBox9")
                     shinyjs::hide("ChatBox10")
                     shinyjs::hide("ChatBox11")
                     shinyjs::hide("ChatBox12")
                     shinyjs::hide("ChatBox13")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "CellChat Cell-Cell Communication Analysis Failed. Please check if you have selected the correct species.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$Interactions_CellChatM <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (input$checkboxCommunicationM == TRUE) {
      par(mfrow = c(1,2), xpd=TRUE)
      for (i in 1:length(seuratreactive$object.list)) {
        netVisual_circle(seuratreactive$object.list[[i]]@net$count, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$count), edge.width.max = 12, title.name = paste0("Number of interactions - ", names(seuratreactive$object.list)[i]))
      }
    }
    else {
      par(mfrow = c(1,2), xpd=TRUE)
      
      mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$count), ncol = ncol(seuratreactive$object.list[[1]]@net$count), dimnames = dimnames(seuratreactive$object.list[[1]]@net$count))
      mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$count[input$Communication_SelectM1, ]
      netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$count), title.name = paste(input$Communication_SelectM1, "Number of interactions - ", names(seuratreactive$object.list)[1]))
      
      mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$count), ncol = ncol(seuratreactive$object.list[[2]]@net$count), dimnames = dimnames(seuratreactive$object.list[[2]]@net$count))
      mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$count[input$Communication_SelectM2, ]
      netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$count), title.name = paste(input$Communication_SelectM2, "Number of interactions - ", names(seuratreactive$object.list)[2]))
      
    }
  })
  
  output$Interactions_CellChatM2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (input$checkboxCommunicationM == TRUE) {
      par(mfrow = c(1,2), xpd=TRUE)
      for (i in 1:length(seuratreactive$object.list)) {
        netVisual_circle(seuratreactive$object.list[[i]]@net$weight, weight.scale = T, label.edge= F, edge.weight.max = max(seuratreactive$object.list[[i]]@net$weight), edge.width.max = 12, title.name = paste0("Interaction weights/strength - ", names(seuratreactive$object.list)[i]))
      }
    }
    else {
      par(mfrow = c(1,2), xpd=TRUE)
      mat1 <- matrix(0, nrow = nrow(seuratreactive$object.list[[1]]@net$weight), ncol = ncol(seuratreactive$object.list[[1]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[1]]@net$weight))
      mat1[input$Communication_SelectM1, ] <- seuratreactive$object.list[[1]]@net$weight[input$Communication_SelectM1, ]
      netVisual_circle(mat1, vertex.weight = as.numeric(table(seuratreactive$object.list[[1]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[1]]@net$weight), title.name = paste(input$Communication_SelectM1, "Interaction weights/strength - ", names(seuratreactive$object.list)[1]))
      
      mat2 <- matrix(0, nrow = nrow(seuratreactive$object.list[[2]]@net$weight), ncol = ncol(seuratreactive$object.list[[2]]@net$weight), dimnames = dimnames(seuratreactive$object.list[[2]]@net$weight))
      mat2[input$Communication_SelectM2, ] <- seuratreactive$object.list[[2]]@net$weight[input$Communication_SelectM2, ]
      netVisual_circle(mat2, vertex.weight = as.numeric(table(seuratreactive$object.list[[2]]@idents)), weight.scale = T, edge.weight.max = max(seuratreactive$object.list[[2]]@net$weight), title.name = paste(input$Communication_SelectM2, "Interaction weights/strength - ", names(seuratreactive$object.list)[2]))
    }
  })
  
  output$CellCell_CellChatUI <- renderUI({
    plotOutput("CellCell_CellChat", height = (400 + input$HeightCellCell*30))
  })
  
  output$CellCell_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioCellCell == "Communication Network") {
      par(mfrow = c(1,2), xpd=TRUE)
      netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T)
      netVisual_diffInteraction(seuratreactive$cellchatM, weight.scale = T, measure = "weight")
      
    }
    else if (input$radioCellCell == "Bar Graph") {
      gg1 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), size.text = input$sizeCellCell)
      gg2 <- compareInteractions(seuratreactive$cellchatM, show.legend = F, group = c(1,2), measure = "weight", size.text = input$sizeCellCell)
      gg1 + gg2
    }
    else if (input$radioCellCell == "Heatmap") {
      gg1 <- netVisual_heatmap(seuratreactive$cellchatM, font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
      gg2 <- netVisual_heatmap(seuratreactive$cellchatM, measure = "weight", font.size = input$sizeCellCell, font.size.title = input$sizeCellCell +2)
      gg1 + gg2
    }
  })
  
  output$SignallingM_CellChatUI <- renderUI({
    plotOutput("SignallingM_CellChat", height = (400 + input$HeightSignalling_ChatM*30))
  })
  
  output$SignallingM_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (!is.null(input$Signalling_SelectM)) {
      
      if (input$SignallingDataset_SelectM == "Dataset 1") {
        if (input$radioChatM == "Hierarchy Plot") {
          par(mfrow=c(1,1))
          netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[1]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10)
        }
        else if (input$radioChatM == "Circle Plot") {
          par(mfrow=c(1,1))
          netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10)
        }
        else if (input$radioChatM == "Chord Plot") {
          par(mfrow=c(1,1))
          netVisual_aggregate(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10)
        }
        else if (input$radioChatM == "Heatmap") {
          par(mfrow=c(1,1))
          netVisual_heatmap(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2)
        }
        else if (input$radioChatM == "Signalling Roles") {
          par(mfrow=c(1,1))
          netAnalysis_signalingRole_network(seuratreactive$object.list[[1]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM)
        }
      }
      else {
        if (input$radioChatM == "Hierarchy Plot") {
          par(mfrow=c(1,1))
          netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM,  vertex.receiver = seq(1,length(unique(seuratreactive$object.list[[2]]@idents))/2), layout = "hierarchy", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10)
        }
        else if (input$radioChatM == "Circle Plot") {
          par(mfrow=c(1,1))
          netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "circle", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10)
        }
        else if (input$radioChatM == "Chord Plot") {
          par(mfrow=c(1,1))
          netVisual_aggregate(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, layout = "chord", pt.title = input$sizeSignallingM, vertex.label.cex = input$sizeSignallingM/10)
        }
        else if (input$radioChatM == "Heatmap") {
          par(mfrow=c(1,1))
          netVisual_heatmap(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, color.heatmap = "Reds", font.size = input$sizeSignallingM, font.size.title = input$sizeSignallingM + 2)
        }
        else if (input$radioChatM == "Signalling Roles") {
          par(mfrow=c(1,1))
          netAnalysis_signalingRole_network(seuratreactive$object.list[[2]], signaling = input$Signalling_SelectM, width = input$WidthSignalling_ChatM, height = input$HeightSignalling_ChatM, font.size = input$sizeSignallingM)
        }
      }
    }
  })
  
  output$SignallingM_CellChattext <- renderUI({
    html <- ifelse (input$SignallingDataset_SelectM == "Dataset 1",
                    paste("Dataset 1 - ", names(seuratreactive$object.list)[1]),
                    paste("Dataset 2 - ", names(seuratreactive$object.list)[2])
    )
    
    HTML(html)
  })
  
  output$LigandM_CellChatUI <- renderUI({
    plotOutput("LigandM_CellChat", height = (400 + input$HeightLigandM*30))
  })
  
  output$LigandM_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (!is.null(input$Ligand_SelectM)) {
      
      if (input$radioLigandChatM == "Dot Plot") {
        par(mfrow=c(1,1))
        netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectM, targets.use = input$target_selectM, comparison = c(1, 2), signaling = input$Ligand_SelectM, remove.isolate = F, font.size = input$sizeLigandM, font.size.title = input$sizeLigandM, angle.x = 90)
      }
      else if (input$radioLigandChatM == "Gene Expression") {
        seuratreactive$cellchatM@meta$datasets = factor(seuratreactive$cellchatM@meta$datasets, levels = c(names(seuratreactive$object.list)[1], names(seuratreactive$object.list)[2])) # set factor level
        plotGeneExpression(seuratreactive$cellchatM, signaling = input$Ligand_SelectM, split.by = "datasets", colors.ggplot = T)
      }
    }
  })
  
  observeEvent(input$GOCHAT, {
    tryCatch({
      withProgress(message = 'Identifying upregulated and downregulated signalling ligand receptor pairs...',
                   value = 0, {
                     incProgress(1/10)
                     
                     # perform differential expression analysis
                     seuratreactive$cellchatM <- identifyOverExpressedGenes(seuratreactive$cellchatM, group.dataset = "datasets", pos.dataset = input$Dataset_SelectMM, features.name = input$Dataset_SelectMM, only.pos = FALSE, thresh.pc = 0.1, thresh.fc = 0.1, thresh.p = 1)
                     incProgress(2/10)
                     # map the results of differential expression analysis onto the inferred cell-cell communications to easily manage/subset the ligand-receptor pairs of interest
                     net <- netMappingDEG(seuratreactive$cellchatM, features.name = input$Dataset_SelectMM)
                     incProgress(3/10)
                     # extract the ligand-receptor pairs with upregulated ligands in LS
                     seuratreactive$net.up <- subsetCommunication(seuratreactive$cellchatM, net = net, datasets = input$Dataset_SelectMM, ligand.logFC = 0.2, receptor.logFC = NULL)
                     incProgress(5/10)
                     
                     # extract the ligand-receptor pairs with upregulated ligands and upregulated receptors in NL, i.e.,downregulated in LS
                     seuratreactive$net.down <- subsetCommunication(seuratreactive$cellchatM, net = net, datasets = input$Dataset_SelectMM2, ligand.logFC = -0.1, receptor.logFC = -0.1)
                     incProgress(10/10)
                     
                     seuratreactive$Dataset_SelectMM <- input$Dataset_SelectMM
                     seuratreactive$Dataset_SelectMM2 <- input$Dataset_SelectMM2
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "CellChat identification of upregulated/downregulated signalling ligand-receptor pairs failed. No significant signaling interactions were detected.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  output$LigandMM1_CellChatUI <- renderUI({
    plotOutput("LigandMM1_CellChat", height = (400 + input$HeightLigandMM*30))
  })
  
  output$LigandMM1_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$net.up), 'Please compute your cell-cell communication analysis'))
    
    netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.up[, "interaction_name", drop = F], comparison = c(1, 2), angle.x = 90, remove.isolate = F,title.name = paste0("Up-regulated signaling in ", input$Dataset_SelectMM), font.size = input$sizeLigandMM, font.size.title = input$sizeLigandMM)
  })
  
  output$LigandMM2_CellChatUI <- renderUI({
    plotOutput("LigandMM2_CellChat", height = (400 + input$HeightLigandMM*30))
  })
  
  output$LigandMM2_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$net.down), 'Please compute your cell-cell communication analysis'))
    
    netVisual_bubble(seuratreactive$cellchatM, sources.use = input$source_selectMM, targets.use = input$target_selectMM, pairLR.use = seuratreactive$net.down[, "interaction_name", drop = F], comparison = c(1, 2),  angle.x = 90, remove.isolate = F,title.name = paste0("Down-regulated signaling in ", input$Dataset_SelectMM), font.size = input$sizeLigandMM, font.size.title = input$sizeLigandMM)
  })
  
  output$RolesM_CellChatUI <- renderUI({
    plotOutput("RolesM_CellChat", height = (400 + input$HeightRolesM*30))
  })
  
  output$RolesM_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioRolesChatM == "Scatter Plot") {
      num.link <- sapply(seuratreactive$object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
      gg <- list()
      for (i in 1:length(seuratreactive$object.list)) {
        gg[[i]] <- netAnalysis_signalingRole_scatter(seuratreactive$object.list[[i]], title = names(seuratreactive$object.list)[i], weight.MinMax = c(min(num.link[[i]]), max(num.link[[i]])),  font.size = input$sizeRolesM, font.size.title = input$sizeRolesM, label.size	= input$sizeRolesM - 8)
      }
      
      patchwork::wrap_plots(plots = gg)
    }
    else if (input$radioRolesChatM == "Bar Graph") {
      gg1 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = T, do.stat = TRUE, font.size = input$sizeRolesM)
      gg2 <- rankNet(seuratreactive$cellchatM, mode = "comparison", stacked = F, do.stat = TRUE, font.size = input$sizeRolesM)
      gg1 + gg2
    }
    else if (input$radioRolesChatM == "Heatmap") {
      if (input$signallingroles_MM == "Outgoing") {
        pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
        ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
        ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "outgoing", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
        ht1 + ht2
      }
      else if (input$signallingroles_MM == "Incoming") {
        pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
        ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
        ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "incoming", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
        ht1 + ht2
      }
      else if (input$signallingroles_MM == "All") {
        pathway.union <- union(seuratreactive$object.list[[1]]@netP$pathways, seuratreactive$object.list[[2]]@netP$pathways)
        ht1 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[1]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[1], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
        ht2 = netAnalysis_signalingRole_heatmap(seuratreactive$object.list[[2]], pattern = "all", signaling = pathway.union, title = names(seuratreactive$object.list)[2], font.size = input$sizeRolesM, font.size.title = input$sizeRolesM + 2, width = input$WidthRolesM, height = input$HeightRolesM)
        ht1 + ht2
      }
    }
  })
  
  output$ManifoldM_CellChatUI <- renderUI({
    plotOutput("ManifoldM_CellChat", height = (400 + input$HeightManifoldM*30))
  })
  
  output$ManifoldM_CellChat <- renderPlot({
    validate(
      need(!is.null(seuratreactive$cellchatM), 'Please compute your cell-cell communication analysis'))
    
    if (input$radioManifoldChatM == "Functional Classification") {
      netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "functional", label.size = input$sizeManifoldM)
    }
    else if (input$radioManifoldChatM == "Structural Classification") {
      netVisual_embeddingPairwise(seuratreactive$cellchatM, type = "structural", label.size = input$sizeManifoldM)
    }
  })
  
  observeEvent(input$startbuttonChat2, {
    library(CellChat)
    library(patchwork)
    library(NMF)
    library(ggalluvial)
    library(ComplexHeatmap)
    
    tryCatch({
      withProgress(message = 'Performing cell-cell communication network analysis...',
                   value = 0, {
                     incProgress(1/20)
                     
                     removeModal()
                     
                     #Dataset 1
                     split_seurat1 <- SplitObject(seuratreactive$obj, split.by = "orig.ident")
                     
                     split_seurat1 <- split_seurat1[input$ChatMultiple1Sample]
                     
                     split_seurat1 <- Reduce(merge, split_seurat1)
                     
                     if (input$Chat_selectcluster == "Seurat Clusters") {
                       
                       split_seurat1 <- SetIdent(split_seurat1, value = "seurat_clusters")
                       
                       new_ids = paste0("Seurat_Cluster_", levels(split_seurat1))
                       names(new_ids) <- levels(split_seurat1)
                       split_seurat1 <- RenameIdents(split_seurat1, new_ids)
                     }
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       cellchat1 <- createCellChat(split_seurat1, assay = "RNA", group.by = "ident")
                     }
                     else {
                       cellchat1 <- createCellChat(split_seurat1, assay = "SCT", group.by = "ident")
                     }
                     
                     cellchat1 <- setIdent(cellchat1, ident.use = "ident") # set "labels" as default cell identity
                     
                     if (input$Chatdb == "Human") {
                       cellchat1@DB <- CellChatDB.human
                     }
                     else if (input$Chatdb == "Mouse") {
                       cellchat1@DB <- CellChatDB.mouse
                     }
                     
                     incProgress(2/40)
                     
                     # subset the expression data of signaling genes for saving computation cost
                     cellchat1 <- subsetData(cellchat1) # This step is necessary even if using the whole database
                     
                     cellchat1 <- identifyOverExpressedGenes(cellchat1, thresh.p = input$Chatpval)
                     
                     incProgress(7/40)
                     
                     cellchat1 <- identifyOverExpressedInteractions(cellchat1)
                     
                     incProgress(8/40)
                     # project gene expression data onto PPI network (optional)
                     if (input$Chatdb == "Human") {
                       cellchat1 <- projectData(cellchat1, PPI.human)
                     }
                     else if (input$Chatdb == "Mouse") {
                       cellchat1 <- projectData(cellchat1, PPI.mouse)
                     }
                     
                     incProgress(12/40)
                     
                     cellchat1 <- computeCommunProb(cellchat1)
                     # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
                     cellchat1 <- filterCommunication(cellchat1, min.cells = input$Chat_number)
                     
                     incProgress(16/40)
                     
                     #Interactions
                     cellchat1 <- computeCommunProbPathway(cellchat1, thresh = input$Chatpval)
                     cellchat1 <- aggregateNet(cellchat1, thresh = input$Chatpval)
                     
                     cellchat1 <- netAnalysis_computeCentrality(cellchat1, slot.name = "netP", thresh = input$Chatpval)
                     
                     incProgress(20/40)
                     
                     #Dataset 2
                     split_seurat2 <- SplitObject(seuratreactive$obj, split.by = "orig.ident")
                     
                     split_seurat2 <- split_seurat2[input$ChatMultiple2Sample]
                     
                     split_seurat2 <- Reduce(merge, split_seurat2)
                     
                     if (input$Chat_selectcluster == "Seurat Clusters") {
                       
                       split_seurat2 <- SetIdent(split_seurat2, value = "seurat_clusters")
                       
                       new_ids = paste0("Seurat_Cluster_", levels(split_seurat2))
                       names(new_ids) <- levels(split_seurat2)
                       split_seurat2 <- RenameIdents(split_seurat2, new_ids)
                     }
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       cellchat2 <- createCellChat(split_seurat2, assay = "RNA", group.by = "ident")
                     }
                     else {
                       cellchat2 <- createCellChat(split_seurat2, assay = "SCT", group.by = "ident")
                     }
                     
                     cellchat2 <- setIdent(cellchat2, ident.use = "ident") # set "labels" as default cell identity
                     
                     if (input$Chatdb == "Human") {
                       cellchat2@DB <- CellChatDB.human
                     }
                     else if (input$Chatdb == "Mouse") {
                       cellchat2@DB <- CellChatDB.mouse
                     }
                     
                     incProgress(22/40)
                     
                     # subset the expression data of signaling genes for saving computation cost
                     cellchat2 <- subsetData(cellchat2) # This step is necessary even if using the whole database
                     
                     cellchat2 <- identifyOverExpressedGenes(cellchat2, thresh.p = input$Chatpval)
                     
                     incProgress(27/40)
                     
                     cellchat2 <- identifyOverExpressedInteractions(cellchat2)
                     
                     incProgress(28/40)
                     # project gene expression data onto PPI network (optional)
                     if (input$Chatdb == "Human") {
                       cellchat2 <- projectData(cellchat2, PPI.human)
                     }
                     else if (input$Chatdb == "Mouse") {
                       cellchat2 <- projectData(cellchat2, PPI.mouse)
                     }
                     
                     incProgress(32/40)
                     
                     cellchat2 <- computeCommunProb(cellchat2)
                     # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
                     cellchat2 <- filterCommunication(cellchat2, min.cells = input$Chat_number)
                     
                     incProgress(36/40)
                     
                     #Interactions
                     cellchat2 <- computeCommunProbPathway(cellchat2, thresh = input$Chatpval)
                     cellchat2 <- aggregateNet(cellchat2, thresh = input$Chatpval)
                     
                     cellchat2 <- netAnalysis_computeCentrality(cellchat2, slot.name = "netP", thresh = input$Chatpval)
                     
                     #Merge the two dataframes
                     seuratreactive$object.list <- list(cellchat1, cellchat2)
                     names(seuratreactive$object.list) <- c(paste0(unique(cellchat1@meta[["orig.ident"]]), collapse = " "), paste0(unique(cellchat2@meta[["orig.ident"]]), collapse = " "))
                     
                     if (identical(levels(unique(cellchat1@idents)), levels(unique(cellchat2@idents)))) {
                       seuratreactive$cellchatM <- mergeCellChat(seuratreactive$object.list, add.names = names(seuratreactive$object.list), cell.prefix = F)
                     }
                     else { 
                       seuratreactive$cellchatM <- mergeCellChat(seuratreactive$object.list, add.names = names(seuratreactive$object.list), cell.prefix = TRUE)
                       seuratreactive$cellchatM <- liftCellChat(seuratreactive$cellchatM, group.new = union(levels(cellchat1@idents), levels(cellchat2@idents)))
                     }
                     
                     incProgress(38/40)
                     
                     seuratreactive$cellchatM <- computeNetSimilarityPairwise(seuratreactive$cellchatM, type = "functional")
                     seuratreactive$cellchatM <- netEmbedding(seuratreactive$cellchatM, type = "functional", umap.method = "uwot")
                     seuratreactive$cellchatM <- netClustering(seuratreactive$cellchatM, type = "functional", do.plot = F, nCores = 10)
                     
                     seuratreactive$cellchatM <- computeNetSimilarityPairwise(seuratreactive$cellchatM, type = "structural")
                     seuratreactive$cellchatM <- netEmbedding(seuratreactive$cellchatM, type = "structural", umap.method = "uwot")
                     seuratreactive$cellchatM <- netClustering(seuratreactive$cellchatM, type = "structural", do.plot = F, nCores = 10)
                     
                     incProgress(40/40)
                     
                     seuratreactive$cellchat <- NULL
                     seuratreactive$ChatChoose <- input$ChatChoose
                     seuratreactive$ChatSample <- NULL
                     seuratreactive$Chatdb <- input$Chatdb
                     seuratreactive$ChatMultiple1Sample <- input$ChatMultiple1Sample
                     seuratreactive$ChatMultiple2Sample <- input$ChatMultiple2Sample
                     seuratreactive$Chat_selectcluster <- input$Chat_selectcluster
                     seuratreactive$Chat_number <- input$Chat_number
                     seuratreactive$Chatpval <- input$Chatpval
                     
                     js$collapse("ChatBoxIntro")
                     shinyjs::hide("ChatBox1")
                     shinyjs::hide("ChatBox2")
                     shinyjs::hide("ChatBox3")
                     shinyjs::hide("ChatBox4")
                     shinyjs::hide("ChatBox5")
                     shinyjs::hide("ChatBox6")
                     shinyjs::show("ChatBox7")
                     shinyjs::show("ChatBox8")
                     shinyjs::show("ChatBox9")
                     shinyjs::show("ChatBox10")
                     shinyjs::show("ChatBox11")
                     shinyjs::show("ChatBox12")
                     shinyjs::show("ChatBox13")
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "CellChat Cell-Cell Communication Analysis Failed. Please check if you have selected the correct species.", type = "error", confirmButtonCol = "#337ab7")
    })
  })
  
  ####TAB 13 - Multimodal ####
  
  observe({
    if (!is.null(seuratreactive$MM) && is.null(seuratreactive$markersDE) && is.null(Yu$result) && is.null(seuratreactive$markersDRUG) && is.null(seuratreactive$markersSDE) && is.null(YuSPA$result) && is.null(seuratreactive$markersDRUG2)) {
      if (is.null(seuratreactive$objQC2)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = T),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = T),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = T),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  })
  
  shinyjs::disable("buttonfileinputMMload1")
  shinyjs::disable("buttonfileinputMMload2")
  shinyjs::disable("buttonfileinputMMload3")
  
  observeEvent(input$startbuttonMM, {
    if (input$iscollapseboxMM == FALSE) {
      js$collapse("MMBoxIntro")
    }
    if (input$iscollapseboxMM2 == FALSE) {
      js$collapse("MMBox4")
    }
    if (input$iscollapseboxMM3 == FALSE) {
      js$collapse("MMBox5")
    }
    shinyjs::hide("MMBox1")
    shinyjs::hide("MMBox2")
    shinyjs::hide("MMBox3")
    shinyjs::show("MMBox4")
    shinyjs::show("MMBox5")
    
    if (is.null(seuratreactive$markersDE)) {
      if (input$iscollapsebox6 == TRUE) {
        js$collapse("DEBoxIntro")
      }
      shinyjs::hide("DEBox1")
      shinyjs::hide("DEBox1a")
      shinyjs::hide("DEBox1b")
      shinyjs::hide("DEBox2")
      shinyjs::hide("DEBox3")
      shinyjs::hide("DEBox4")
      shinyjs::hide("DEBox5")
      shinyjs::hide("DEBox6")
      shinyjs::enable("startbuttonDE")
      shinyjs::hide("NextButtonDE")
    }
    
    if (is.null(seuratreactive$markersSDE)) {
      if (input$iscollapseboxSDE == TRUE) {
        js$collapse("SDEBoxIntro")
      }
      shinyjs::hide("SDEBox1")
      shinyjs::hide("SDEBox2")
      shinyjs::hide("SDEBox3")
      shinyjs::hide("SDEBox4")
      shinyjs::hide("SDEBox5")
      shinyjs::hide("SDEBox6")
      shinyjs::hide("SDEBox7")
      shinyjs::enable("startbuttonSDE")
      shinyjs::hide("NextButtonSDE")
    }
    
  })
  
  observe({
    if (input$radioLoadMM == "10X data") {
      shinyjs::show("fileinputMMload1")
      shinyjs::show("buttonfileinputMMload1")
      shinyjs::hide("fileinputMMload2")
      shinyjs::hide("buttonfileinputMMload2")
      shinyjs::hide("fileinputMMload3")
      shinyjs::hide("buttonfileinputMMload3")
      shinyjs::hide("existingdataMM")
      shinyjs::hide("buttonexistingdataMM")
    }
    
    else if (input$radioLoadMM == "Count Matrix (txt)") {
      shinyjs::hide("fileinputMMload1")
      shinyjs::hide("buttonfileinputMMload1")
      shinyjs::show("fileinputMMload2")
      shinyjs::show("buttonfileinputMMload2")
      shinyjs::hide("fileinputMMload3")
      shinyjs::hide("buttonfileinputMMload3")
      shinyjs::hide("existingdataMM")
      shinyjs::hide("buttonexistingdataMM")
    }
    
    else if (input$radioLoadMM == "Seurat object (RDS file)") {
      shinyjs::hide("fileinputMMload1")
      shinyjs::hide("buttonfileinputMMload1")
      shinyjs::hide("fileinputMMload2")
      shinyjs::hide("buttonfileinputMMload2")
      shinyjs::show("fileinputMMload3")
      shinyjs::show("buttonfileinputMMload3")
      shinyjs::hide("existingdataMM")
      shinyjs::hide("buttonexistingdataMM")
    }
    
    else if (input$radioLoadMM == "Sample datasets") {
      shinyjs::hide("fileinputMMload1")
      shinyjs::hide("buttonfileinputMMload1")
      shinyjs::hide("fileinputMMload2")
      shinyjs::hide("buttonfileinputMMload2")
      shinyjs::hide("fileinputMMload3")
      shinyjs::hide("buttonfileinputMMload3")
      shinyjs::show("existingdataMM")
      shinyjs::show("buttonexistingdataMM")
    }
  })
  
  observe({
    if(is.null(input$existingdataMM) | !is.null(input$existingdataMM) && input$existingdataMM == "") {
      shinyjs::disable("buttonexistingdataMM")
    }
    else{
      shinyjs::enable("buttonexistingdataMM")
    }
  })
  
  observeEvent(input$fileinputMMload1, {
    
    tryCatch({
      withProgress(message = 'Loading 10X files...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     file1 <- input$fileinputMMload1[[1, 'datapath']]
                     file2 <- input$fileinputMMload1[[2, 'datapath']]
                     file3 <- input$fileinputMMload1[[3, 'datapath']]
                     
                     filesdir = dirname(file1)
                     
                     file.rename(file1, paste0(filesdir,'/',input$fileinputMMload1$name[1]))
                     file.rename(file2, paste0(filesdir,'/',input$fileinputMMload1$name[2]))
                     file.rename(file3, paste0(filesdir,'/',input$fileinputMMload1$name[3]))
                     
                     seurat <- Read10X(data.dir = filesdir)
                     
                     if (any(names(seurat) == "Antibody Capture")) {
                       rownames(x = seurat[["Antibody Capture"]]) <- gsub(pattern = "_.+", replacement = "",
                                                                          x = rownames(x = seurat[["Antibody Capture"]]))
                       
                       seuratreactive$MM <- as.data.frame(seurat[["Antibody Capture"]])
                       
                       seuratreactive$MM <- seuratreactive$MM[,colnames(seuratreactive$MM) %in% colnames(seuratreactive$obj)]
                       
                       if(length(colnames(seuratreactive$MM)) > 0) {
                         
                         seuratreactive$obj[["ADT"]] <- CreateAssayObject(counts = seuratreactive$MM)
                         
                         shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                         
                         shinyjs::enable("buttonfileinputMMload1")
                         
                         incProgress(0.8)
                       }
                       else{
                         shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
                         shinyjs::disable("buttonfileinputMMload1")
                         seuratreactive$MM <- NULL
                       }
                     }
                     else {
                       shinyalert("ERROR!", "Multimodal data was not detected in your 10X files, please check to see if you have uploaded the correct files.", type = "error", confirmButtonCol = "#337ab7")
                       shinyjs::disable("buttonfileinputMMload1")
                       seuratreactive$MM <- NULL
                     }
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check if you have loaded 3 files (barcodes.tsv, features.tsv, matrix.mtx).", type = "error", confirmButtonCol = "#337ab7")
      shinyjs::disable("buttonfileinputMMload1")
      seuratreactive$MM <- NULL
    })
  })
  
  observeEvent(input$fileinputMMload2, {
    
    tryCatch({
      withProgress(message = 'Loading multimodal data...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     
                     file = input$fileinputMMload2
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seuratreactive$MM <- read.table(file$datapath, sep = "\t", header = T, row.names = 1, check.names = F)
                     
                     seuratreactive$MM <- seuratreactive$MM[,colnames(seuratreactive$MM) %in% colnames(seuratreactive$obj)]
                     
                     if(length(colnames(seuratreactive$MM)) > 0) {
                       adt_assay <- CreateAssayObject(counts = seuratreactive$MM)
                       
                       seuratreactive$obj[["ADT"]] <- adt_assay
                       
                       shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                       
                       shinyjs::enable("buttonfileinputMMload2")
                       
                       incProgress(0.8)
                     }
                     else{
                       shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
                       shinyjs::disable("buttonfileinputMMload2")
                       seuratreactive$MM <- NULL
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
      shinyjs::disable("buttonfileinputMMload2")
      seuratreactive$MM <- NULL
    })
  })
  
  observeEvent(input$fileinputMMload3, {
    
    tryCatch({
      withProgress(message = 'Loading multimodal data...', 
                   value = 0, {
                     incProgress(0.1)
                     
                     
                     file = input$fileinputMMload3
                     if (is.null(file)) {
                       return(NULL)
                     }
                     
                     seuratreactive$MM <- readRDS(file$datapath)
                     
                     if (any(names(seuratreactive$MM) == "ADT")) {
                       
                       seuratreactive$MM <- as.data.frame(seuratreactive$MM@assays$ADT@counts)   
                       
                       seuratreactive$MM <- seuratreactive$MM[,colnames(seuratreactive$MM) %in% colnames(seuratreactive$obj)]
                       
                       if(length(colnames(seuratreactive$MM)) > 0) {
                         adt_assay <- CreateAssayObject(counts = seuratreactive$MM)
                         
                         seuratreactive$obj[["ADT"]] <- adt_assay
                         
                         shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                         
                         shinyjs::enable("buttonfileinputMMload3")
                         
                         incProgress(0.8)
                       }
                       
                       else{
                         shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
                         shinyjs::disable("buttonfileinputMMload3")
                         seuratreactive$MM <- NULL
                       }
                     }
                     else {
                       shinyalert("ERROR!", "No ADT assay found in Seurat object, please make sure your multimodal data is present in your Seurat object under the ADT named assay.", type = "error", confirmButtonCol = "#337ab7")
                       shinyjs::disable("buttonfileinputMMload3")
                       seuratreactive$MM <- NULL
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
      shinyjs::disable("buttonfileinputMMload3")
      seuratreactive$MM <- NULL
    })
  })
  
  observeEvent(input$buttonfileinputMMload2 | input$buttonfileinputMMload1 | input$buttonfileinputMMload3, {
    
    shinyjs::show("MMBox1")
    shinyjs::show("MMBox2")
    shinyjs::show("MMBox3")
    if (input$iscollapseboxMM2 == FALSE) {
      js$collapse("MMBox4")
    }
    if (input$iscollapseboxMM3 == FALSE) {
      js$collapse("MMBox5")
    }
    
  }, ignoreInit = T)
  
  
  observeEvent(input$buttonexistingdataMM, {
    tryCatch({    
      if (input$existingdataMM == "8k CBMCs (Seurat CITE-seq multimodal tutorial)") {
        withProgress(message = 'Loading 8k CBMCs multimodal data...', 
                     value = 0, {
                       incProgress(0.1)
                       
                       seuratreactive$MM <- read.table("cbmc.adt.txt", sep = "\t", header = T, row.names = 1, check.names = F)
                       
                       seuratreactive$MM <- seuratreactive$MM[,colnames(seuratreactive$MM) %in% colnames(seuratreactive$obj)]
                       
                       if(length(colnames(seuratreactive$MM)) > 0) {
                         adt_assay <- CreateAssayObject(counts = seuratreactive$MM)
                         
                         seuratreactive$obj[["ADT"]] <- adt_assay
                         
                         shinyalert("SUCCESS!", "File uploaded successfully.", type = "success", confirmButtonCol = "#337ab7")
                         
                         incProgress(0.8)
                         
                         shinyjs::show("MMBox1")
                         shinyjs::show("MMBox2")
                         shinyjs::show("MMBox3")
                         if (input$iscollapseboxMM2 == FALSE) {
                           js$collapse("MMBox4")
                         }
                         if (input$iscollapseboxMM3 == FALSE) {
                           js$collapse("MMBox5")
                         }
                         
                         
                       }
                       else{
                         shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
                         seuratreactive$MM <- NULL
                       }
                     })
      }
    },
    error = function(e) {
      shinyalert("ERROR!", "No overlapping cells were detected between multimodal data and loaded single cell RNA-seq dataset.", type = "error", confirmButtonCol = "#337ab7")
      seuratreactive$MM <- NULL
    })
  })
  
  current_selectionMM <- reactiveVal(NULL)
  
  observeEvent(input$geneMM, {
    current_selectionMM(input$MMInput)
  })
  
  observe({
    if (!is.null(seuratreactive$reductionCC1)) {
      updateVirtualSelect('MMInput', choices = c(rownames(seuratreactive$obj@assays$ADT)), selected = current_selectionMM())
    }
  })
  
  observe({
    if (is.null(input$MMInput) | !is.null(input$MMInput) && input$MMInput == "") {
      shinyjs::disable("geneMM")
    }
    else {
      shinyjs::enable("geneMM")
    }
  })
  
  output$MatrixTableMM <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$MM), 'You must load your data first...'))
    
    head(as.data.frame(seuratreactive$MM),c(500L, 500L)) %>% na.omit()
    
  }, options = list(scrollX = TRUE))
  
  
  observeEvent(input$geneMM, {
    tryCatch({
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       DefaultAssay(seuratreactive$obj) <- "RNA"
                     }
                     else {
                       DefaultAssay(seuratreactive$obj) <- "SCT"
                     }
                     
                     
                     mean1 <- FetchData(object = seuratreactive$obj, vars = input$MMInput) 
                     mean1 <- mutate(mean1, Average = rowMeans(mean1))
                     incProgress(3/10)
                     
                     seuratreactive$obj <- AddMetaData(object=seuratreactive$obj, metadata = mean1$Average, col.name = "Gene_ExpressionMM")
                     
                     IdentifiersMM <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                      "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident",
                                                                                      "Gene_ExpressionMM"))
                     
                     IdentifiersMM[c("percent.ribo", "percent.mt", "Gene_ExpressionMM")] <- signif(IdentifiersMM[c("percent.ribo", "percent.mt", "Gene_ExpressionMM")], digits = 4)
                     
                     incProgress(8/10)
                     
                     colnames(IdentifiersMM) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                  "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                  "Cell Cycle Phase", "Labelled Cells", "Gene expression")
                     
                     seuratreactive$plot.dataM <- merge(seuratreactive$plot.data.original, IdentifiersMM, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.dataM) <- seuratreactive$plot.dataM$Row.names
                     
                     seuratreactive$plot.dataM <- seuratreactive$plot.dataM[,-1]
                     
                     seuratreactive$titleMM <- paste(input$MMInput, collapse = " ")
                     
                     seuratreactive$MMInput <- input$MMInput
                     
                     incProgress(10/10)
                     
                   })
      
      withProgress(message = 'Please wait...',
                   value = 0, {
                     incProgress(1/10)
                     
                     DefaultAssay(seuratreactive$obj) <- "ADT"
                     
                     seuratreactive$obj <- NormalizeData(seuratreactive$obj, normalization.method = "CLR", margin = 2)
                     
                     mean2 <- FetchData(object = seuratreactive$obj, vars = input$MMInput) 
                     mean2 <- mutate(mean2, Average = rowMeans(mean2))
                     incProgress(3/10)
                     
                     seuratreactive$obj <- AddMetaData(object=seuratreactive$obj, metadata = mean2$Average, col.name = "Gene_ExpressionMM2")
                     
                     IdentifiersMM2 <- FetchData(object = seuratreactive$obj, vars = c("seurat_clusters", "orig.ident",
                                                                                       "nFeature_RNA", "nCount_RNA", "percent.ribo", "percent.mt", "Phase", "ident",
                                                                                       "Gene_ExpressionMM2"))
                     
                     IdentifiersMM2[c("percent.ribo", "percent.mt", "Gene_ExpressionMM2")] <- signif(IdentifiersMM2[c("percent.ribo", "percent.mt", "Gene_ExpressionMM2")], digits = 4)
                     
                     incProgress(8/10)
                     
                     colnames(IdentifiersMM2) <- c("Seurat Clusters", "Sample", "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                   "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)",
                                                   "Cell Cycle Phase", "Labelled Cells", "Gene expression")
                     
                     seuratreactive$plot.dataMM <- merge(seuratreactive$plot.data.original, IdentifiersMM2, by = "row.names", all = T)
                     
                     rownames(seuratreactive$plot.dataMM) <- seuratreactive$plot.dataMM$Row.names
                     
                     seuratreactive$plot.dataMM <- seuratreactive$plot.dataMM[,-1]
                     
                     seuratreactive$querygeneMM <- input$MMInput
                     
                     if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "LogNormalize" | 
                         !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "LogNormalize") {
                       DefaultAssay(seuratreactive$obj) <- "RNA"
                     }
                     else {
                       DefaultAssay(seuratreactive$obj) <- "SCT"
                     }
                     incProgress(10/10)
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please select a gene (or genes) to view its expression.", type = "error", confirmButtonCol = "#337ab7")
    }) 
  })
  
  output$plotMM1UI <- renderUI({
    plotlyOutput("plotMM1", height = (400 + input$MMplotheight*40))
  })
  
  output$plotMM1 <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.dataM$`Gene expression`), 'You must select a gene first to view its expression'))
    validate(
      need(!is.null(seuratreactive$plot.dataM), "You must complete the clustering step"))
    validate(
      need(all(seuratreactive$querygeneMM %in% rownames(seuratreactive$obj)), "Query gene(s) not present in dataset"))
    
    if(input$plotclustersMM == "Gene expression"){
      if(input$radioMM == "UMAP") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM)) 
        
      }
      else if (input$radioMM == "t-SNE") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
                
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
      }
    }
    else if(input$plotclustersMM == "Seurat Clusters" | input$plotclustersMM == "Cell Cycle Phase" | input$plotclustersMM == "Labelled Cells" |
            input$plotclustersMM == "Sample") {  
      if(input$radioMM == "UMAP") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
      }
      else if (input$radioMM == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "t-SNE") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
    }
    else {
      if(input$radioMM == "UMAP") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "t-SNE") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.dataM, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.dataM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataM),
                                   "<br>Sample: ", seuratreactive$plot.dataM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
      }
    }
  })
  
  output$plotMM2UI <- renderUI({
    plotlyOutput("plotMM2", height = (400 + input$MMplotheight*40))
  })
  
  output$plotMM2 <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.dataMM$`Gene expression`), 'You must select a gene first to view its expression'))
    
    if(input$plotclustersMM == "Gene expression"){
      if(input$radioMM == "UMAP") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "t-SNE") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
                
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(title = seuratreactive$titleMM, font = list(size = input$labelsizeMM))
      }
    }
    else if(input$plotclustersMM == "Seurat Clusters" | input$plotclustersMM == "Cell Cycle Phase" | input$plotclustersMM == "Labelled Cells" |
            input$plotclustersMM == "Sample") {  
      if(input$radioMM == "UMAP") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
      }
      else if (input$radioMM == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "t-SNE") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = col_vector,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
    }
    else {
      if(input$radioMM == "UMAP") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "t-SNE") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
        
      }
      else if (input$radioMM == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.dataMM, 
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D, 
                color = seuratreactive$plot.dataMM[, input$plotclustersMM],
                colors = Spectrals,
                type = "scatter3d", 
                mode = "markers", 
                marker = list(size = input$sizeMM),
                opacity = input$opacityMM,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.dataMM),
                                   "<br>Sample: ", seuratreactive$plot.dataMM$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.dataMM$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.dataMM$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.dataMM$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.dataMM$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.dataMM$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.dataMM$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.dataMM$`Ribosomal Percentage`,
                                   "<br>Gene Expression: ", seuratreactive$plot.dataMM$`Gene expression`)
        ) %>% layout(font = list(size = input$labelsizeMM))
      }
    }
  })
  
  output$VlnMMUI <- renderUI({
    plotOutput("VlnMM", height = (300 + input$HeightMMVln*30))
  })
  
  output$VlnMM2UI <- renderUI({
    plotOutput("VlnMM2", height = (300 + input$HeightMMVln*30))
  })
  
  output$VlnMM <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.dataM$`Gene expression`), 'You must select a gene first to view its expression'))
    validate(
      need(all(seuratreactive$querygeneMM %in% rownames(seuratreactive$obj)), "Query gene(s) not present in dataset"))
    
    if (input$checkboxvlnMM == TRUE) {
      if (input$VlnInputMM == "Seurat Clusters") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", group.by = "seurat_clusters", split.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
      else if (input$VlnInputMM == "Labelled Cells") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", pt.size = input$VlnptMM, split.by = "orig.ident") + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
      else if (input$VlnInputMM == "Sample") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", group.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
      else if (input$VlnInputMM == "Cell Cycle Phase") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", group.by = "Phase", pt.size = input$VlnptMM, split.by = "orig.ident") + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
    }
    else {
      if (input$VlnInputMM == "Seurat Clusters") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", group.by = "seurat_clusters", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
      else if (input$VlnInputMM == "Labelled Cells") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
      else if (input$VlnInputMM == "Sample") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", group.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
      else if (input$VlnInputMM == "Cell Cycle Phase") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM", group.by = "Phase", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM <- p
        p
      }
    }
  })
  
  output$VlnMM2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.dataMM$`Gene expression`), 'You must select a gene first to view its expression'))    
    
    if (input$checkboxvlnMM == TRUE) {
      if (input$VlnInputMM == "Seurat Clusters") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", group.by = "seurat_clusters", split.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
      else if (input$VlnInputMM == "Labelled Cells") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", split.by = "orig.ident", pt.size = input$VlnptMM) +
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
      else if (input$VlnInputMM == "Sample") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", group.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
      else if (input$VlnInputMM == "Cell Cycle Phase") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", group.by = "Phase", split.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
    }
    
    else {
      if (input$VlnInputMM == "Seurat Clusters") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", group.by = "seurat_clusters", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
      else if (input$VlnInputMM == "Labelled Cells") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
      else if (input$VlnInputMM == "Sample") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", group.by = "orig.ident", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
      else if (input$VlnInputMM == "Cell Cycle Phase") {
        p <- VlnPlot(seuratreactive$obj, features = "Gene_ExpressionMM2", group.by = "Phase", pt.size = input$VlnptMM) + 
          theme(text = element_text(size=input$VlnsizeMM),
                axis.text.x= element_text(size = input$VlnsizeMM),
                axis.text.y= element_text(size = input$VlnsizeMM))
        p[[1]][["labels"]][["title"]] <- seuratreactive$titleMM
        seuratreactive$pMM2 <- p
        p
      }
    }
  })
  
  
  ####TAB 14 - Differential Gene Expression##### 
  shinyjs::disable("DE1") 
  shinyjs::disable("DE2") 
  shinyjs::disable("buttonDEGDE") 
  shinyjs::disable("buttonEnrichDE")
  shinyjs::disable("buttonvolcano")
  shinyjs::disable("GP_actions")
  
  observeEvent(input$startbuttonDE, {
    if (input$iscollapsebox6 == FALSE) {
      js$collapse("DEBoxIntro")
    }
    shinyjs::show("DEBox1")
    shinyjs::show("DEBox2")
    shinyjs::show("DEBox3")
    shinyjs::show("DEBox4")
    shinyjs::show("DEBox5")
    shinyjs::show("DEBox6")
    
    shinyjs::disable("startbuttonDE")
    
    seuratreactive$title <- NULL
    
    updatePickerInput(session, 'SelectComparison', selected = "")  
    
  }, ignoreInit = T)
  
  observe({
    if (!is.null(input$DEInput1)) {
      if (paste(input$DEInput1, collapse = "") == paste(input$DEInput2, collapse = "") | is.null(input$DEInput1) | is.null(input$DEInput2))
      {
        shinyjs::disable("DE1")
        shinyjs::disable("buttonDEGDE")
        shinyjs::disable("buttonEnrichDE")
      }
      else{
        shinyjs::enable("DE1")
        shinyjs::enable("buttonDEGDE")
        shinyjs::enable("buttonEnrichDE")
      }
    }
  })
  
  observe({
    if (!is.null(input$DEInput1_samples)) {
      if (paste(input$DEInput1_samples, collapse = "") == paste(input$DEInput2_samples, collapse = "") | is.null(input$DECluster) | is.null(input$DEInput1_samples) | is.null(input$DEInput2_samples))
      {
        shinyjs::disable("DE2")
        shinyjs::disable("buttonDEGDE")
        shinyjs::disable("buttonEnrichDE")
      }
      else{
        shinyjs::enable("DE2")
        shinyjs::enable("buttonDEGDE")
        shinyjs::enable("buttonEnrichDE")
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$markersDEvolcano))
    {
      shinyjs::enable("buttonvolcano") 
    }
    else{
      shinyjs::disable("buttonvolcano")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$Subset_GENES))
    {
      shinyjs::enable("GP_actions") 
    }
    else{
      shinyjs::disable("GP_actions")
    }
  })
  
  observe({
    if (is.null(input$SelectComparison) | !is.null(input$SelectComparison) && input$SelectComparison == "") {
      shinyjs::disable("buttonDEGDE") 
      shinyjs::disable("buttonEnrichDE") 
    }
  })
  
  observe({
    if (is.null(input$SelectComparison) | !is.null(input$SelectComparison) && input$SelectComparison == "") {
      shinyjs::hide("DEBox1a")
      shinyjs::hide("DEBox1b")
      updatePickerInput(session, "plotclustersDE", selected = "Sample")
    }
    if (input$SelectComparison == "Between Clusters") {
      shinyjs::show("DEBox1a")
      shinyjs::hide("DEBox1b")
      updatePickerInput(session, "plotclustersDE", selected = "Differential Expression")
    }
    if (input$SelectComparison == "Between Samples") {
      shinyjs::hide("DEBox1a")
      shinyjs::show("DEBox1b")
      updatePickerInput(session, "plotclustersDE", selected = "Differential Expression")
    }
  })
  
  observeEvent(input$tabs, {
    if (input$tabs == "DE" & !is.null(Yu$result)) {
      shinyalert("Warning!", "If you adjust differential expression comparisons, the pathway analysis and drug gene interactions tabs will disappear and you will have to 
                 re-do the analysis.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  observe({
    if (!is.null(seuratreactive$reductionCC1)) {
      if (input$GEDotInput == "Seurat Clusters") {
        updateVirtualSelect('refineGEDOT', choices = c(unique(levels(seuratreactive$obj$seurat_clusters))))
      }
      else {
        updateVirtualSelect('refineGEDOT', choices = c(unique(levels(seuratreactive$obj))))
      }
      
      if (input$VlnInput == "Seurat Clusters") {
        updateVirtualSelect('refineGEVLN', choices = c(unique(levels(seuratreactive$obj$seurat_clusters))))
      }
      else {
        updateVirtualSelect('refineGEVLN', choices = c(unique(levels(seuratreactive$obj))))
      }
      
      if (input$DE_clusters == "Seurat Clusters"){
        updateVirtualSelect('DEInput1', choices = c(unique(levels(seuratreactive$obj$seurat_clusters)))
                            , selected = seuratreactive$DEInput1)
      }
      else {
        updateVirtualSelect('DEInput1', choices = c(unique(levels(seuratreactive$obj)))
                            , selected = seuratreactive$DEInput1)
      }
      
      if (input$DE_clusters == "Seurat Clusters"){
        updateVirtualSelect('DEInput2', choices = c(unique(levels(seuratreactive$obj$seurat_clusters)))
                            , selected = seuratreactive$DEInput2)
      }
      else {
        updateVirtualSelect('DEInput2', choices = c(unique(levels(seuratreactive$obj)))
                            , selected = seuratreactive$DEInput2)
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$reductionCC1)) {
      if (input$DE_clusters_samples == "Seurat Clusters"){
        updateVirtualSelect('DECluster', choices = c(unique(levels(seuratreactive$obj$seurat_clusters)))
                            , selected = seuratreactive$DECluster)
      }
      else {
        updateVirtualSelect('DECluster', choices = c(unique(levels(seuratreactive$obj)))
                            , selected = seuratreactive$DECluster)
      }
      if (!is.null(seuratreactive$reductionCC1)) {
        updateVirtualSelect('DEInput1_samples', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                            selected = seuratreactive$DEInput1_samples)
        updateVirtualSelect('DEInput2_samples', choices = c(levels(as.factor(seuratreactive$obj@meta.data$orig.ident))),
                            selected = seuratreactive$DEInput2_samples)
      }
    }
  })
  
  output$plotDE <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(input$SelectComparison), "Please choose either Between Clusters or Between Sample comparisons"))
    
    if(input$plotclustersDE == "Differential Expression") {
      
      if (input$SelectComparison == "Between Clusters") {
        
        if(input$radioDE == "UMAP") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~UMAP_1_2D, y = ~UMAP_2_2D,
                  color = ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput1), paste(input$DEInput1, collapse = ","),
                                 ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput2), paste(input$DEInput2, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
        }
        else if (input$radioDE == "3D UMAP") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                  color = ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput1), paste(input$DEInput1, collapse = ","),
                                 ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput2), paste(input$DEInput2, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter3d",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
          
        }
        else if (input$radioDE == "t-SNE") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                  color = ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput1), paste(input$DEInput1, collapse = ","),
                                 ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput2), paste(input$DEInput2, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
          
        }
        else if (input$radioDE == "3D t-SNE") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                  color = ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput1), paste(input$DEInput1, collapse = ","),
                                 ifelse(seuratreactive$plot.data[, input$DE_clusters] %in% c(input$DEInput2), paste(input$DEInput2, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter3d",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
          
        }
      }
      
      else if (input$SelectComparison == "Between Samples") {
        
        if(input$radioDE == "UMAP") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~UMAP_1_2D, y = ~UMAP_2_2D,
                  color = ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput1_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput1_samples, collapse = ","), 
                                 ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput2_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput2_samples, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
        }
        else if (input$radioDE == "3D UMAP") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                  color = ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput1_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput1_samples, collapse = ","), 
                                 ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput2_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput2_samples, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter3d",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
          
        }
        else if (input$radioDE == "t-SNE") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                  color = ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput1_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput1_samples, collapse = ","), 
                                 ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput2_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput2_samples, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
          
        }
        else if (input$radioDE == "3D t-SNE") {
          plot_ly(data = seuratreactive$plot.data,
                  x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                  color = ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput1_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput1_samples, collapse = ","), 
                                 ifelse(seuratreactive$plot.data[, "Sample"] %in% c(input$DEInput2_samples) & seuratreactive$plot.data[, input$DE_clusters_samples] %in% c(input$DECluster), paste(input$DEInput2_samples, collapse = ","),
                                        "*Non-selected")),
                  colors = c("grey", "red", "green"),
                  type = "scatter3d",
                  mode = "markers",
                  marker = list(size = input$sizeDE),
                  opacity = input$opacityDE,
                  hoverinfo = "text",
                  hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                     "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                     "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                     "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                     "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                     "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                     "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                     "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                     "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
          ) %>% layout(font = list(size = input$labelsizeDE))
          
        }
      }
    }
    
    else if(input$plotclustersDE == "Seurat Clusters" | input$plotclustersDE == "Cell Cycle Phase" | input$plotclustersDE == "Labelled Cells" |
            input$plotclustersDE == "Sample") {
      if(input$radioDE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
      }
      else if (input$radioDE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
        
      }
      else if (input$radioDE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
        
      }
      else if (input$radioDE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
        
      }
    }
    else {
      if(input$radioDE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
        
      }
      else if (input$radioDE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
        
      }
      else if (input$radioDE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
      }
      else if (input$radioDE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersDE],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeDE),
                opacity = input$opacityDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`)
        ) %>% layout(font = list(size = input$labelsizeDE))
      }
    }
  })
  
  observe({
    if (is.na(input$logDE) |
        is.na(input$minDE)) {
      shinyjs::disable("buttonDEGDE")
    }
    else {
      if (!is.null(seuratreactive$markersDE)) {
        shinyjs::enable("buttonDEGDE")
      }
    }
  })
  
  observe({
    if (is.na(input$pvalueDE) | is.na(input$pvalDE)) {
      shinyjs::disable("buttonEnrichDE")
    }
    else {
      if (!is.null(seuratreactive$markersDE)) {
        shinyjs::enable("buttonEnrichDE")
      }
    }
  })
  
  observe({
    if (is.na(input$pvolcano)) {
      shinyjs::disable("buttonvolcano")
    }
    else {
      if (!is.null(seuratreactive$markersDEvolcano)) {
        shinyjs::enable("buttonvolcano")
      }
    }
  })
  
  observe({
    if (is.null(input$GPInput) | !is.null(input$GPInput) && input$GPInput == "") {
      shinyjs::disable("GP_actions")
    }
    else {
      if (!is.null(seuratreactive$markersDEvolcano)) {
        shinyjs::enable("GP_actions")
      }
    }
  })
  
  observeEvent(input$DE1 | input$buttonDEGDE | input$buttonEnrichDE | input$DE2, {
    tryCatch({
      withProgress(message = 'Finding differentially expressed genes, this may take a while...',
                   value = 0, {
                     incProgress(1/10)
                     
                     if (input$SelectComparison == "Between Clusters") {
                       
                       if (input$DE_clusters == "Seurat Clusters") {
                         
                         if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                             !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                           DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                         }
                         else {
                           DEtestobj <- seuratreactive$obj
                         }
                         
                         markersDE <- FindMarkers(DEtestobj, ident.1 = input$DEInput1, ident.2 = input$DEInput2, group.by = "seurat_clusters", only.pos = input$onlyposDE,
                                                  logfc.threshold = input$logDE, min.pct = input$minDE,
                                                  test.use = input$testuseDE)
                         
                         markersDE$p_val_adj = p.adjust(markersDE$p_val, method=input$padjustDE1)
                         
                         markersDE <- signif(markersDE, digits = 4)
                         setDT(markersDE, keep.rownames = TRUE)[]
                         seuratreactive$markersDE <- markersDE
                         
                         colnames(seuratreactive$markersDE) <- c("Gene Name", "p-value", "Log2 fold change", "pct.1",
                                                                 "pct.2", "Adjusted p-value")
                         
                         seuratreactive$markersDE <- seuratreactive$markersDE[seuratreactive$markersDE$`Adjusted p-value` <= input$pvalDE,]
                         
                         incProgress(3/10)
                         
                         GENES_ENTREZ <- bitr(seuratreactive$markersDE$`Gene Name`, fromType="SYMBOL", toType="ENTREZID", OrgDb = seuratreactive$orgDb)
                         colnames(GENES_ENTREZ)[1] <- "Gene Name"
                         
                         incProgress(5/10)
                         
                         seuratreactive$PATHWAY <- full_join(seuratreactive$markersDE, GENES_ENTREZ) %>% na.omit() %>% arrange(desc(`Log2 fold change`))
                         
                         geneList = seuratreactive$PATHWAY$`Log2 fold change`
                         
                         names(geneList) = seuratreactive$PATHWAY$ENTREZID
                         
                         geneList = sort(geneList, decreasing = TRUE)
                         
                         if (input$enrichInputDE == "Gene Ontology") {
                           
                           Yu <- gseGO(geneList      = geneList,
                                       OrgDb         = seuratreactive$orgDb,
                                       ont           = input$GOlabelsDE,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- input$GOlabelsDE
                           
                           Yu <- clusterProfiler::simplify(Yu)
                           
                           seuratreactive$YuDE <- Yu@result
                           
                         }
                         else if (input$enrichInputDE == "KEGG Pathways") {
                           Yu <- gseKEGG(geneList         = geneList,
                                         organism     = seuratreactive$EKEGG,
                                         pAdjustMethod = input$padjustDE,
                                         pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "WikiPathways") {
                           Yu <- gseWP(geneList = geneList,
                                       organism = input$species1,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "REACTOME") {
                           Yu <- gsePathway(geneList = geneList,
                                            organism = seuratreactive$REACT,
                                            pAdjustMethod = input$padjustDE,
                                            pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         incProgress(8/10)
                         
                         rownames(seuratreactive$YuDE) <- NULL
                         seuratreactive$YuDE <- seuratreactive$YuDE[,c(1,2,6)]
                         colnames(seuratreactive$YuDE) <- c("ID", "Description", "Adjusted p-value")
                         seuratreactive$YuDE$`Adjusted p-value` <- signif(seuratreactive$YuDE$`Adjusted p-value`, digits = 4)
                         
                         incProgress(10/10)
                         
                         seuratreactive$onlyposDE <- input$onlyposDE
                         seuratreactive$DEInput1_samples <- input$DEInput1_samples
                         seuratreactive$DEInput2_samples <- input$DEInput2_samples
                         seuratreactive$DECluster <- input$DECluster
                         seuratreactive$DE_clusters_samples <- input$DE_clusters_samples
                         seuratreactive$SelectComparison <- input$SelectComparison
                         seuratreactive$DE_clusters <- input$DE_clusters
                         seuratreactive$DEInput1 <- input$DEInput1
                         seuratreactive$DEInput2 <- input$DEInput2
                         seuratreactive$logDE <- input$logDE
                         seuratreactive$minDE <- input$minDE
                         seuratreactive$padjustDE1 <- input$padjustDE1
                         seuratreactive$testuseDE <- input$testuseDE
                         seuratreactive$enrichInputDE <- input$enrichInputDE
                         seuratreactive$pvalueDE <- input$pvalueDE
                         seuratreactive$pvalDE <- input$pvalDE
                         seuratreactive$padjustDE <- input$padjustDE
                       }
                       
                       else if (input$DE_clusters == "Labelled Cells") {
                         
                         if (!is.null(seuratreactive$new.cell.ids)) {
                           Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                           
                           seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                         }
                         
                         if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                             !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                           DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                         }
                         else {
                           DEtestobj <- seuratreactive$obj
                         }
                         
                         markersDE <- FindMarkers(DEtestobj, ident.1 = input$DEInput1, ident.2 = input$DEInput2, only.pos = input$onlyposDE,
                                                  logfc.threshold = input$logDE, min.pct = input$minDE,
                                                  test.use = input$testuseDE)
                         
                         markersDE$p_val_adj = p.adjust(markersDE$p_val, method=input$padjustDE1)
                         
                         markersDE <- signif(markersDE, digits = 4)
                         setDT(markersDE, keep.rownames = TRUE)[]
                         
                         seuratreactive$markersDE <- markersDE
                         
                         colnames(seuratreactive$markersDE) <- c("Gene Name", "p-value", "Log2 fold change", "pct.1",
                                                                 "pct.2", "Adjusted p-value")
                         
                         seuratreactive$markersDE <- seuratreactive$markersDE[seuratreactive$markersDE$`Adjusted p-value` <= input$pvalDE,]
                         
                         
                         incProgress(3/10)
                         
                         GENES_ENTREZ <- bitr(seuratreactive$markersDE$`Gene Name`, fromType="SYMBOL", toType="ENTREZID", OrgDb = seuratreactive$orgDb)
                         colnames(GENES_ENTREZ)[1] <- "Gene Name"
                         
                         incProgress(5/10)
                         
                         seuratreactive$PATHWAY <- full_join(seuratreactive$markersDE, GENES_ENTREZ) %>% na.omit() %>% arrange(desc(`Log2 fold change`))
                         
                         geneList = seuratreactive$PATHWAY$`Log2 fold change`
                         
                         names(geneList) = seuratreactive$PATHWAY$ENTREZID
                         
                         geneList = sort(geneList, decreasing = TRUE)
                         
                         if (input$enrichInputDE == "Gene Ontology") {
                           
                           Yu <- gseGO(geneList      = geneList,
                                       OrgDb         = seuratreactive$orgDb,
                                       ont           = input$GOlabelsDE,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff  = input$pvalueDE)
                           
                           Yu <- clusterProfiler::simplify(Yu)
                           
                           seuratreactive$YuDE <- Yu@result
                           
                           seuratreactive$GOlabelsDE <- input$GOlabelsDE
                         }
                         else if (input$enrichInputDE == "KEGG Pathways") {
                           Yu <- gseKEGG(geneList         = geneList,
                                         organism     = seuratreactive$EKEGG,
                                         pAdjustMethod = input$padjustDE,
                                         pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                         }
                         else if (input$enrichInputDE == "WikiPathways") {
                           Yu <- gseWP(geneList = geneList,
                                       organism = input$species1,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff = input$pvalueDE)
                           
                           seuratreactive$YuDE <- Yu@result
                           seuratreactive$GOlabelsDE <- "NA"
                         }
                         else if (input$enrichInputDE == "REACTOME") {
                           Yu <- gsePathway(geneList = geneList,
                                            organism = seuratreactive$REACT,
                                            pAdjustMethod = input$padjustDE,
                                            pvalueCutoff  = input$pvalueDE)
                           
                           seuratreactive$YuDE <- Yu@result
                           seuratreactive$GOlabelsDE <- "NA"
                         }
                         incProgress(8/10)
                         
                         rownames(seuratreactive$YuDE) <- NULL
                         seuratreactive$YuDE <- seuratreactive$YuDE[,c(1,2,6)]
                         colnames(seuratreactive$YuDE) <- c("ID", "Description", "Adjusted p-value")
                         seuratreactive$YuDE$`Adjusted p-value` <- signif(seuratreactive$YuDE$`Adjusted p-value`, digits = 4)
                         
                         
                         incProgress(10/10)
                         
                         seuratreactive$onlyposDE <- input$onlyposDE
                         seuratreactive$DEInput1_samples <- input$DEInput1_samples
                         seuratreactive$DEInput2_samples <- input$DEInput2_samples
                         seuratreactive$DECluster <- input$DECluster
                         seuratreactive$DE_clusters_samples <- input$DE_clusters_samples
                         seuratreactive$SelectComparison <- input$SelectComparison
                         seuratreactive$DE_clusters <- input$DE_clusters
                         seuratreactive$DEInput1 <- input$DEInput1
                         seuratreactive$DEInput2 <- input$DEInput2
                         seuratreactive$logDE <- input$logDE
                         seuratreactive$minDE <- input$minDE
                         seuratreactive$padjustDE1 <- input$padjustDE1
                         seuratreactive$testuseDE <- input$testuseDE
                         seuratreactive$enrichInputDE <- input$enrichInputDE
                         seuratreactive$pvalueDE <- input$pvalueDE
                         seuratreactive$pvalDE <- input$pvalDE
                         seuratreactive$padjustDE <- input$padjustDE
                         
                       }
                     }
                     
                     if (input$SelectComparison == "Between Samples") {
                       
                       if (input$DE_clusters_samples == "Seurat Clusters") {
                         
                         Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                         
                         if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                             !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                           DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                         }
                         else {
                           DEtestobj <- seuratreactive$obj
                         }
                         
                         markersDE <- FindMarkers(DEtestobj, ident.1 = input$DEInput1_samples, ident.2 = input$DEInput2_samples, group.by = "orig.ident", only.pos = input$onlyposDE,
                                                  subset.ident = input$DECluster,
                                                  logfc.threshold = input$logDE, min.pct = input$minDE,
                                                  test.use = input$testuseDE)
                         
                         markersDE$p_val_adj = p.adjust(markersDE$p_val, method=input$padjustDE1)
                         
                         
                         markersDE <- signif(markersDE, digits = 4)
                         setDT(markersDE, keep.rownames = TRUE)[]
                         seuratreactive$markersDE <- markersDE
                         
                         colnames(seuratreactive$markersDE) <- c("Gene Name", "p-value", "Log2 fold change", "pct.1",
                                                                 "pct.2", "Adjusted p-value")
                         
                         seuratreactive$markersDE <- seuratreactive$markersDE[seuratreactive$markersDE$`Adjusted p-value` <= input$pvalDE,]
                         
                         incProgress(3/10)
                         
                         GENES_ENTREZ <- bitr(seuratreactive$markersDE$`Gene Name`, fromType="SYMBOL", toType="ENTREZID", OrgDb = seuratreactive$orgDb)
                         colnames(GENES_ENTREZ)[1] <- "Gene Name"
                         
                         incProgress(5/10)
                         
                         seuratreactive$PATHWAY <- full_join(seuratreactive$markersDE, GENES_ENTREZ) %>% na.omit() %>% arrange(desc(`Log2 fold change`))
                         
                         geneList = seuratreactive$PATHWAY$`Log2 fold change`
                         
                         names(geneList) = seuratreactive$PATHWAY$ENTREZID
                         
                         geneList = sort(geneList, decreasing = TRUE)
                         
                         if (input$enrichInputDE == "Gene Ontology") {
                           
                           Yu <- gseGO(geneList      = geneList,
                                       OrgDb         = seuratreactive$orgDb,
                                       ont           = input$GOlabelsDE,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- input$GOlabelsDE
                           
                           Yu <- clusterProfiler::simplify(Yu)
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "KEGG Pathways") {
                           Yu <- gseKEGG(geneList         = geneList,
                                         organism     = seuratreactive$EKEGG,
                                         pAdjustMethod = input$padjustDE,
                                         pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "WikiPathways") {
                           Yu <- gseWP(geneList = geneList,
                                       organism = input$species1,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "REACTOME") {
                           Yu <- gsePathway(geneList = geneList,
                                            organism = seuratreactive$REACT,
                                            pAdjustMethod = input$padjustDE,
                                            pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         incProgress(8/10)
                         
                         rownames(seuratreactive$YuDE) <- NULL
                         seuratreactive$YuDE <- seuratreactive$YuDE[,c(1,2,6)]
                         colnames(seuratreactive$YuDE) <- c("ID", "Description", "Adjusted p-value")
                         seuratreactive$YuDE$`Adjusted p-value` <- signif(seuratreactive$YuDE$`Adjusted p-value`, digits = 4)
                         
                         incProgress(10/10)
                         
                         seuratreactive$onlyposDE <- input$onlyposDE
                         seuratreactive$DEInput1_samples <- input$DEInput1_samples
                         seuratreactive$DEInput2_samples <- input$DEInput2_samples
                         seuratreactive$DECluster <- input$DECluster
                         seuratreactive$DE_clusters_samples <- input$DE_clusters_samples
                         seuratreactive$SelectComparison <- input$SelectComparison
                         seuratreactive$DE_clusters <- input$DE_clusters
                         seuratreactive$DEInput1 <- input$DEInput1
                         seuratreactive$DEInput2 <- input$DEInput2
                         seuratreactive$logDE <- input$logDE
                         seuratreactive$minDE <- input$minDE
                         seuratreactive$padjustDE1 <- input$padjustDE1
                         seuratreactive$testuseDE <- input$testuseDE
                         seuratreactive$enrichInputDE <- input$enrichInputDE
                         seuratreactive$pvalueDE <- input$pvalueDE
                         seuratreactive$pvalDE <- input$pvalDE
                         seuratreactive$padjustDE <- input$padjustDE
                         
                         if (!is.null(seuratreactive$new.cell.ids)) {
                           Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                           
                           seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                         }
                       }
                       
                       else if (input$DE_clusters_samples == "Labelled Cells") {
                         
                         if (!is.null(seuratreactive$new.cell.ids)) {
                           Idents(seuratreactive$obj) <- seuratreactive$obj@meta.data$seurat_clusters
                           
                           seuratreactive$obj <- RenameIdents(seuratreactive$obj, seuratreactive$new.cell.ids)
                         }
                         
                         if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                             !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                           DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                         }
                         else {
                           DEtestobj <- seuratreactive$obj
                         }
                         
                         markersDE <- FindMarkers(DEtestobj, ident.1 = input$DEInput1_samples, ident.2 = input$DEInput2_samples, group.by = "orig.ident", only.pos = input$onlyposDE,
                                                  subset.ident = input$DECluster,
                                                  logfc.threshold = input$logDE, min.pct = input$minDE,
                                                  test.use = input$testuseDE)
                         
                         markersDE$p_val_adj = p.adjust(markersDE$p_val, method=input$padjustDE1)
                         
                         markersDE <- signif(markersDE, digits = 4)
                         setDT(markersDE, keep.rownames = TRUE)[]
                         
                         seuratreactive$markersDE <- markersDE
                         
                         colnames(seuratreactive$markersDE) <- c("Gene Name", "p-value", "Log2 fold change", "pct.1",
                                                                 "pct.2", "Adjusted p-value")
                         
                         seuratreactive$markersDE <- seuratreactive$markersDE[seuratreactive$markersDE$`Adjusted p-value` <= input$pvalDE,]
                         
                         
                         incProgress(3/10)
                         
                         GENES_ENTREZ <- bitr(seuratreactive$markersDE$`Gene Name`, fromType="SYMBOL", toType="ENTREZID", OrgDb = seuratreactive$orgDb)
                         colnames(GENES_ENTREZ)[1] <- "Gene Name"
                         
                         incProgress(5/10)
                         
                         seuratreactive$PATHWAY <- full_join(seuratreactive$markersDE, GENES_ENTREZ) %>% na.omit() %>% arrange(desc(`Log2 fold change`))
                         
                         geneList = seuratreactive$PATHWAY$`Log2 fold change`
                         
                         names(geneList) = seuratreactive$PATHWAY$ENTREZID
                         
                         geneList = sort(geneList, decreasing = TRUE)
                         
                         if (input$enrichInputDE == "Gene Ontology") {
                           
                           Yu <- gseGO(geneList      = geneList,
                                       OrgDb         = seuratreactive$orgDb,
                                       ont           = input$GOlabelsDE,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- input$GOlabelsDE
                           
                           Yu <- clusterProfiler::simplify(Yu)
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "KEGG Pathways") {
                           Yu <- gseKEGG(geneList         = geneList,
                                         organism     = seuratreactive$EKEGG,
                                         pAdjustMethod = input$padjustDE,
                                         pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "WikiPathways") {
                           Yu <- gseWP(geneList = geneList,
                                       organism = input$species1,
                                       pAdjustMethod = input$padjustDE,
                                       pvalueCutoff = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         else if (input$enrichInputDE == "REACTOME") {
                           Yu <- gsePathway(geneList = geneList,
                                            organism = seuratreactive$REACT,
                                            pAdjustMethod = input$padjustDE,
                                            pvalueCutoff  = input$pvalueDE)
                           seuratreactive$GOlabelsDE <- "NA"
                           
                           seuratreactive$YuDE <- Yu@result
                         }
                         incProgress(8/10)
                         
                         rownames(seuratreactive$YuDE) <- NULL
                         seuratreactive$YuDE <- seuratreactive$YuDE[,c(1,2,6)]
                         colnames(seuratreactive$YuDE) <- c("ID", "Description", "Adjusted p-value")
                         seuratreactive$YuDE$`Adjusted p-value` <- signif(seuratreactive$YuDE$`Adjusted p-value`, digits = 4)
                         
                         
                         incProgress(10/10)
                         
                         seuratreactive$onlyposDE <- input$onlyposDE
                         seuratreactive$DEInput1_samples <- input$DEInput1_samples
                         seuratreactive$DEInput2_samples <- input$DEInput2_samples
                         seuratreactive$DECluster <- input$DECluster
                         seuratreactive$DE_clusters_samples <- input$DE_clusters_samples
                         seuratreactive$SelectComparison <- input$SelectComparison
                         seuratreactive$DE_clusters <- input$DE_clusters
                         seuratreactive$DEInput1 <- input$DEInput1
                         seuratreactive$DEInput2 <- input$DEInput2
                         seuratreactive$logDE <- input$logDE
                         seuratreactive$minDE <- input$minDE
                         seuratreactive$padjustDE1 <- input$padjustDE1
                         seuratreactive$testuseDE <- input$testuseDE
                         seuratreactive$enrichInputDE <- input$enrichInputDE
                         seuratreactive$pvalueDE <- input$pvalueDE
                         seuratreactive$pvalDE <- input$pvalDE
                         seuratreactive$padjustDE <- input$padjustDE
                         
                       }
                     }
                     
                   })
    },
    error = function(e) {
      shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
      seuratreactive$YuDE <- NULL
    })
  }, ignoreInit = T)
  
  observe({
    if(input$enrichInputDE == "Gene Ontology") {
      shinyjs::show("GOlabelsDE")
    }
    else{
      shinyjs::hide("GOlabelsDE")
    }
  })
  
  
  output$markersDE <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$markersDE), 'You have not made any comparisons yet... ')
    )
    
    seuratreactive$markersDE
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$GO_DE <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$markersDE), 'You have not made any comparisons yet... ')
    )
    validate(
      need(!is.null(seuratreactive$YuDE), 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.')
    )
    validate(
      need(nrow(seuratreactive$YuDE) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.')
    )
    
    seuratreactive$YuDE
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  
  observeEvent(input$DE1 | input$buttonDEGDE | input$buttonEnrichDE | input$buttonvolcano | input$DE2, {
    tryCatch({
      
      seuratreactive$markersDEvolcano <- seuratreactive$markersDE
      
      seuratreactive$markersDEvolcano$Expression <- "None"
      
      seuratreactive$markersDEvolcano$Expression[seuratreactive$markersDEvolcano$`Log2 fold change` > input$logvolcano[2] & seuratreactive$markersDEvolcano$`Adjusted p-value` <= input$pvolcano] <- "UPREGULATED"
      
      seuratreactive$markersDEvolcano$Expression[seuratreactive$markersDEvolcano$`Log2 fold change` < input$logvolcano[1] & seuratreactive$markersDEvolcano$`Adjusted p-value` <= input$pvolcano] <- "DOWNREGULATED"
      
      seuratreactive$volcanovalue1 <- input$logvolcano[1]
      
      seuratreactive$volcanovalue2 <- input$logvolcano[2]
      
      seuratreactive$pvaluevolcano <- input$pvolcano
      
      seuratreactive$Subset_GENES <- NULL
      
    },
    error = function(e) {
      shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
      seuratreactive$YuDE <- NULL
    }) 
    
  }, ignoreInit = T)
  
  output$volcanoplot <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$markersDEvolcano), 'You have not made any comparisons yet... '))
    
    p <- ggplot(data=seuratreactive$markersDEvolcano, aes(x= `Log2 fold change`, y=-log10(`Adjusted p-value`), col = `Expression`,
                                                          text = paste0("Gene Name:", `Gene Name`, "\n",
                                                                        "Log2 fold change:", `Log2 fold change`, "\n",
                                                                        "Adjusted p-value:", `Adjusted p-value`, "\n",
                                                                        "Expression:", `Expression`)
                                                          
    )) +
      geom_point(size = input$sizevolcanoDE, alpha = input$opacityvolcanoDE) + 
      theme_minimal() +
      theme(legend.position="none", text = element_text(size=input$labelsizevolcanoDE)) +
      ylab("-log10(p-value)") +
      xlab("Log2 fold change") +
      scale_color_manual(values=c("blue", "black", "red")) +
      geom_vline(xintercept=c(seuratreactive$volcanovalue1, seuratreactive$volcanovalue2), col="red", linetype = "dotted", size = 1.5) +
      geom_hline(yintercept=-log10(seuratreactive$pvaluevolcano), col="blue", linetype = "dashed", size = 1.5)
    
    ggplotly(p, tooltip = "text")
  })        
  
  observe({
    if (!is.null(seuratreactive$msig)) {
      updateVirtualSelect('GPInput', choices = sample(seuratreactive$msig$gs_name))
    }
  })
  
  observeEvent(input$GP_actions, {
    tryCatch({
      Gene_vector_msig <- seuratreactive$msig[match(input$GPInput, seuratreactive$msig$gs_name),]
      
      Gene_vector_msig <- Gene_vector_msig[,-1] %>% as.character() %>% na.omit() %>% unique()
      
      Subset_GENES <- seuratreactive$markersDE[seuratreactive$markersDE$`Gene Name` %in% Gene_vector_msig,]
      
      seuratreactive$Subset_GENES <- arrange(Subset_GENES, Subset_GENES$`Log2 fold change`)
      
      seuratreactive$title <- input$GPInput
    },
    error = function(e) {
      shinyalert("ERROR!", "Please select a pathway to look at.", type = "error", confirmButtonCol = "#337ab7")
      
    }) 
  })
  
  
  output$GP_plot <- renderPlotly({
    validate(
      need(input$GP_actions, 'Please select a pathway to look at...'))
    validate(
      need(length(rownames(seuratreactive$Subset_GENES)) > 0, 'No expression of genes in selected pathway. Consider easing p-value thresholds and/or logfc thresholds'))
    tryCatch({
      
      q <- ggplot(seuratreactive$Subset_GENES, aes(x=reorder(seuratreactive$Subset_GENES$`Gene Name`, seuratreactive$Subset_GENES$`Log2 fold change`), y=seuratreactive$Subset_GENES$`Log2 fold change`,
                                                   text = paste0("Gene Name: ", `Gene Name`, "\n",
                                                                 "Log2 fold change: ", `Log2 fold change`, "\n",
                                                                 "Adjusted p-value: ", `Adjusted p-value`))) +
        geom_point(size = input$sizeGPDE, alpha = input$opacityGPDE, col = "red") + 
        theme_minimal() +
        theme(legend.position="none", text = element_text(size=input$labelsizeGPDE)) +
        ylab("Log2 fold change") +
        xlab("") +
        ggtitle(seuratreactive$title) +
        theme(axis.text.x = element_blank(),
              plot.title = element_text(vjust=0.5)) +
        geom_hline(yintercept=-log10(1), size = 1)
      
      ggplotly(q, tooltip = "text")
    },
    error = function(e) {
    }) 
  })
  
  observeEvent(input$DE1 | input$buttonDEGDE | input$buttonEnrichDE | input$buttonvolcano | input$DE2, {
    if(!is.null(seuratreactive$markersDE)) {
      if(length(rownames(seuratreactive$YuDE)) < 1) {
        shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7") 
      }
    }
    
    
    if (!is.null(seuratreactive$DE_clusters)) {
      seuratreactive$PAchoose <- NULL
      seuratreactive$padjustPA <- NULL
      seuratreactive$pvaluePA <- NULL
      seuratreactive$GOlabelsPA <- NULL
      Yu$result <- NULL
      Yu$ema <- NULL
      seuratreactive$geneListPA <- NULL
      
      if (input$iscollapsebox7 == TRUE) {
        js$collapse("PABoxIntro")
      }
      shinyjs::hide("PABox1")
      shinyjs::hide("PABox2")
      shinyjs::hide("PABox3")
      shinyjs::hide("PABox4")
      shinyjs::hide("PABox5")
      
      if (input$iscollapseboxDRUG == TRUE) {
        js$collapse("DRUGBoxIntro")
      }
      shinyjs::hide("DRUGBox1")
      shinyjs::hide("DRUGBox2")
      
      shinyjs::disable("startbuttonDE")
      shinyjs::show("NextButtonDE")
      
      if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      else if (is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      
      else if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      else if (is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F))) 
        }
      }
      
      else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  }, ignoreInit = T)
  
  
  observeEvent(input$NextButtonDE, {
    
    #PA
    seuratreactive$PAchoose <- NULL
    seuratreactive$padjustPA <- NULL
    seuratreactive$pvaluePA <- NULL
    seuratreactive$GOlabelsPA <- NULL
    Yu$result <- NULL
    Yu$ema <- NULL
    seuratreactive$geneListPA <- NULL
    
    if (input$iscollapsebox7 == TRUE) {
      js$collapse("PABoxIntro")
    }
    shinyjs::hide("PABox1")
    shinyjs::hide("PABox2")
    shinyjs::hide("PABox3")
    shinyjs::hide("PABox4")
    shinyjs::hide("PABox5")
    
    shinyjs::disable("startbuttonDE")
    shinyjs::hide("NextButtonDE")
    
    if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
      if (is.null(seuratreactive$VFintegrated)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
    
    else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                       menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                       menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                       menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                       menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                       menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                       menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                       menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                       menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                       menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                       menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                       menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                       menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                       menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
    }
    else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                       menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                       menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                       menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                       menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                       menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                       menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                       menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                       menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                       menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                       menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = T),
                                                       menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                       menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F),
                                                       menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
    }
    
    else if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
      if (is.null(seuratreactive$VFintegrated)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
    
    else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                       menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                       menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                       menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                       menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                       menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                       menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                       menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                       menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                       menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                       menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                       menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
    }
  }, ignoreInit = T)
  
  
  
  ####TAB 15 - Custom Differential expression#####
  shinyjs::disable("buttonSDEGDE") 
  shinyjs::disable("SelectedCellsGO") 
  shinyjs::disable("buttonEnrichSDE") 
  shinyjs::disable("buttonvolcanoSDE")
  shinyjs::disable("GP_actionsSDE")
  
  observeEvent(input$startbuttonSDE, {
    if (input$iscollapseboxSDE == FALSE) {
      js$collapse("SDEBoxIntro")
    }
    shinyjs::show("SDEBox1")
    shinyjs::show("SDEBox2")
    shinyjs::show("SDEBox3")
    shinyjs::show("SDEBox4")
    shinyjs::show("SDEBox5")
    shinyjs::show("SDEBox6")
    shinyjs::show("SDEBox7")
    
    shinyjs::disable("startbuttonSDE")
    
    seuratreactive$titleSDE <- NULL
  }, ignoreInit = T)
  
  observe({
    if (!is.null(seuratreactive$markersSDEvolcano))
    {
      shinyjs::enable("buttonvolcanoSDE") 
    }
    else{
      shinyjs::disable("buttonvolcanoSDE")
    }
  })
  
  observe({
    if (!is.null(seuratreactive$Subset_GENESSDE))
    {
      shinyjs::enable("GP_actionsSDE") 
    }
    else{
      shinyjs::disable("GP_actionsSDE")
    }
  })
  
  observe({
    if (is.null(selected()$key) | is.null(selected2()$key)) {
      shinyjs::disable("buttonSDEGDE") 
      shinyjs::disable("SelectedCellsGO") 
      shinyjs::disable("buttonEnrichSDE") 
      shinyjs::disable("buttonvolcanoSDE")
      shinyjs::disable("GP_actionsSDE")
    }
    
    else if (!is.null(selected()$key) | !is.null(selected2()$key)) {
      if (identical(selected()$key, selected2()$key) |
          length(selected()$key) <= 5 | length(selected2()$key) <= 5)
      {
        shinyjs::disable("buttonSDEGDE") 
        shinyjs::disable("SelectedCellsGO") 
        shinyjs::disable("buttonEnrichSDE") 
        shinyjs::disable("buttonvolcanoSDE")
        shinyjs::disable("GP_actionsSDE")
      }
      else{
        shinyjs::enable("buttonSDEGDE") 
        shinyjs::enable("buttonEnrichSDE") 
        shinyjs::enable("SelectedCellsGO") 
      }
    }
  })
  
  observeEvent(input$tabs, {
    if (input$tabs == "SDE" & !is.null(YuSPA$result)) {
      shinyalert("WARNING!", "If you adjust differential expression comparisons, the custom pathway analysis and custom drug gene interactions tabs will disappear and you will have to 
                 re-do the analysis.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  output$plotSDE1UI <- renderUI({
    plotlyOutput("plotSDE1", height = (300 + input$customplotheight*40))
  })
  
  output$plotSDE2UI <- renderUI({
    plotlyOutput("plotSDE2", height = (300 + input$customplotheight*40))
  })
  
  output$plotSDE1 <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    if(input$plotclustersSDE == "Seurat Clusters" | input$plotclustersSDE == "Cell Cycle Phase" | input$plotclustersSDE == "Labelled Cells" |
       input$plotclustersSDE == "Sample") {
      if(input$radioSDE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
      }
      else if (input$radioSDE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
    }
    else {
      if(input$radioSDE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
      }
      else if (input$radioSDE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot1") %>% layout(font = list(size = input$labelsizeSDE))
      }
    }
  })
  
  output$plotSDE2 <- renderPlotly({
    
    if(input$plotclustersSDE == "Seurat Clusters" | input$plotclustersSDE == "Cell Cycle Phase" | input$plotclustersSDE == "Labelled Cells" |
       input$plotclustersSDE == "Sample") {
      if(input$radioSDE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
      }
      else if (input$radioSDE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = col_vector,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
    }
    else {
      if(input$radioSDE == "UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_2D, y = ~UMAP_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "3D UMAP") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~UMAP_1_3D, y = ~UMAP_2_3D, z = ~UMAP_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
        
      }
      else if (input$radioSDE == "t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_2D, y = ~tSNE_2_2D, key = rownames(seuratreactive$plot.data),
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
      }
      else if (input$radioSDE == "3D t-SNE") {
        plot_ly(data = seuratreactive$plot.data,
                x = ~tSNE_1_3D, y = ~tSNE_2_3D, z = ~tSNE_3_3D,
                color = seuratreactive$plot.data[, input$plotclustersSDE],
                colors = Spectrals,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = input$sizeSDE),
                opacity = input$opacitySDE,
                hoverinfo = "text",
                hovertext = paste0("Cell: ", rownames(seuratreactive$plot.data),
                                   "<br>Sample: ", seuratreactive$plot.data$`Sample`,
                                   "<br>Seurat Cluster: ", seuratreactive$plot.data$`Seurat Clusters`,
                                   "<br>Cell Label: ", seuratreactive$plot.data$`Labelled Cells`, 
                                   "<br>Cell Cycle Phase: ", seuratreactive$plot.data$`Cell Cycle Phase`,
                                   "<br>Unique Genes: ", seuratreactive$plot.data$`Number of unique genes`,
                                   "<br>UMIs: ", seuratreactive$plot.data$`Number of UMIs per cell`,
                                   "<br>Mitochondrial percentage: ", seuratreactive$plot.data$`Mitochondrial Percentage`,
                                   "<br>Ribosomal percentage: ", seuratreactive$plot.data$`Ribosomal Percentage`),
                source = "plot2") %>% layout(font = list(size = input$labelsizeSDE))
      }
    }
  })
  
  observe({
    if (is.na(input$logSDE) |
        is.na(input$minSDE)) {
      shinyjs::disable("buttonSDEGDE")
    }
    else {
      if (!is.null(seuratreactive$markersSDE)) {
        shinyjs::enable("buttonSDEGDE")
      }
    }
  })
  
  observe({
    if (is.na(input$pvalueSDE) | is.na(input$pvalSDE)) {
      shinyjs::disable("buttonEnrichSDE")
    }
    else {
      if (!is.null(seuratreactive$YuSDE)) {
        shinyjs::enable("buttonEnrichSDE")
      }
    }
  })
  
  observe({
    if (is.na(input$pvolcanoSDE)) {
      shinyjs::disable("buttonvolcanoSDE")
    }
    else {
      if (!is.null(seuratreactive$markersSDEvolcano)) {
        shinyjs::enable("buttonvolcanoSDE")
      }
    }
  })
  
  observe({
    if (is.null(input$GPInputSDE) | !is.null(input$GPInputSDE) && input$GPInputSDE == "") {
      shinyjs::disable("GP_actionsSDE")
    }
    else {
      if (!is.null(seuratreactive$markersSDEvolcano)) {
        shinyjs::enable("GP_actionsSDE")
      }
    }
  })
  
  
  observeEvent(input$tabs, {
    tryCatch({
      if (input$tabs == "SDE" | input$tabs == "TA") {
        if (identical(seuratreactive$plot.data$`Seurat Clusters`, seuratreactive$plot.data$`Labelled Cells`)) {
          seuratreactive$plot.data.table  <- seuratreactive$plot.data[c("Seurat Clusters", "Sample", "Cell Cycle Phase",
                                                                        "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                        "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)")]
          seuratreactive$plot.data.table <- tibble::rownames_to_column(seuratreactive$plot.data.table, "Cell")
        }
        else{
          seuratreactive$plot.data.table  <- seuratreactive$plot.data[c("Seurat Clusters", "Labelled Cells", "Sample", "Cell Cycle Phase",
                                                                        "Number of unique genes (nFeature_RNA)", "Number of UMIs per cell (nCount_RNA)",
                                                                        "Ribosomal Percentage (percent.ribo)", "Mitochondrial Percentage (percent.mt)")]
          seuratreactive$plot.data.table <- tibble::rownames_to_column(seuratreactive$plot.data.table, "Cell")
        }
      }
    },
    error = function(e) {
      shinyalert("ERROR!", "Unexpected error. Panic!", type = "error", confirmButtonCol = "#337ab7")
      
    }) 
  }, ignoreInit = T)
  
  selected <- reactive({
    req(seuratreactive$plot.data)
    event_data("plotly_selected", source = "plot1")
  })
  
  selected2 <- reactive({
    req(seuratreactive$plot.data)
    event_data("plotly_selected", source = "plot2")
  })
  
  
  output$selectedplot1 <- DT::renderDataTable({
    seuratreactive$lasso1 <- subset(seuratreactive$plot.data.table, seuratreactive$plot.data.table$Cell %in% selected()$key)
    seuratreactive$lasso1    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$selectedplot2 <- DT::renderDataTable({
    seuratreactive$lasso2 <- subset(seuratreactive$plot.data.table, seuratreactive$plot.data.table$Cell %in% selected2()$key)
    seuratreactive$lasso2
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  observeEvent(input$SelectedCellsGO | input$buttonSDEGDE | input$buttonEnrichSDE, {
    tryCatch({
      withProgress(message = 'Finding differentially expressed genes, this may take a while...',
                   value = 0, {
                     incProgress(1/10)
                     if (!is.null(selected()$key) && !is.null(selected2()$key) |
                         !is.null(seuratreactive$selectedkey1) && !is.null(seuratreactive$selectedkey2)) {
                       
                       if (!is.null(selected()$key) && !is.null(selected2()$key)) {
                         seuratreactive$selectedkey1 <- selected()$key
                         seuratreactive$selectedkey2 <- selected2()$key
                       }
                       
                       if (!is.null(seuratreactive$SCTransformDR) && seuratreactive$SCTransformDR == "SCTransform" | 
                           !is.null(seuratreactive$SCTransformDR2) && seuratreactive$SCTransformDR2 == "SCTransform") {
                         DEtestobj <- PrepSCTFindMarkers(seuratreactive$obj)
                       }
                       else {
                         DEtestobj <- seuratreactive$obj
                       }
                       
                       markersSDE <- FindMarkers(DEtestobj, ident.1 = seuratreactive$selectedkey1, ident.2 = seuratreactive$selectedkey2, only.pos = input$onlyposSDE,
                                                 logfc.threshold = input$logSDE, min.pct = input$minSDE,
                                                 test.use = input$testuseSDE)
                       
                       markersSDE$p_val_adj = p.adjust(markersSDE$p_val, method=input$padjustSDE1)
                       
                       markersSDE <- signif(markersSDE, digits = 4)
                       setDT(markersSDE, keep.rownames = TRUE)[]
                       seuratreactive$markersSDE <- markersSDE
                       
                       colnames(seuratreactive$markersSDE) <- c("Gene Name", "p-value", "Log2 fold change", "pct.1",
                                                                "pct.2", "Adjusted p-value")
                       
                       seuratreactive$markersSDE <- seuratreactive$markersSDE[seuratreactive$markersSDE$`Adjusted p-value` <= input$pvalSDE,]
                       
                       incProgress(3/10)
                       
                       GENES_ENTREZ <- bitr(seuratreactive$markersSDE$`Gene Name`, fromType="SYMBOL", toType="ENTREZID", OrgDb = seuratreactive$orgDb)
                       colnames(GENES_ENTREZ)[1] <- "Gene Name"
                       
                       incProgress(5/10)
                       
                       seuratreactive$PATHWAYSDE <- full_join(seuratreactive$markersSDE, GENES_ENTREZ) %>% na.omit() %>% arrange(desc(`Log2 fold change`))
                       
                       geneList = seuratreactive$PATHWAYSDE$`Log2 fold change`
                       
                       names(geneList) = seuratreactive$PATHWAYSDE$ENTREZID
                       
                       geneList = sort(geneList, decreasing = TRUE)
                       
                       if (input$enrichInputSDE == "Gene Ontology") {
                         
                         Yu <- gseGO(geneList      = geneList,
                                     OrgDb         = seuratreactive$orgDb,
                                     ont           = input$GOlabelsSDE,
                                     pAdjustMethod = input$padjustSDE,
                                     pvalueCutoff  = input$pvalueSDE)
                         
                         Yu <- clusterProfiler::simplify(Yu)
                         
                         seuratreactive$YuSDE <- Yu@result
                         
                         seuratreactive$GOlabelsSDE <- input$GOlabelsSDE
                       }
                       else if (input$enrichInputSDE == "KEGG Pathways") {
                         Yu <- gseKEGG(geneList         = geneList,
                                       organism     = seuratreactive$EKEGG,
                                       pAdjustMethod = input$padjustSDE,
                                       pvalueCutoff  = input$pvalueSDE)
                         seuratreactive$GOlabelsSDE <- "NA"
                         
                         seuratreactive$YuSDE <- Yu@result
                       }
                       else if (input$enrichInputSDE == "WikiPathways") {
                         Yu <- gseWP(geneList = geneList,
                                     organism = input$species1,
                                     pAdjustMethod = input$padjustSDE,
                                     pvalueCutoff = input$pvalueSDE)
                         seuratreactive$GOlabelsSDE <- "NA"
                         
                         seuratreactive$YuSDE <- Yu@result
                       }
                       else if (input$enrichInputSDE == "REACTOME") {
                         Yu <- gsePathway(geneList = geneList,
                                          organism = seuratreactive$REACT,
                                          pAdjustMethod = input$padjustSDE,
                                          pvalueCutoff  = input$pvalueSDE)
                         seuratreactive$GOlabelsSDE <- "NA"
                         
                         seuratreactive$YuSDE <- Yu@result
                       }
                       
                       incProgress(8/10)
                       
                       rownames(seuratreactive$YuSDE) <- NULL
                       seuratreactive$YuSDE <- seuratreactive$YuSDE[,c(1,2,6)]
                       colnames(seuratreactive$YuSDE) <- c("ID", "Description", "Adjusted p-value")
                       seuratreactive$YuSDE$`Adjusted p-value` <- signif(seuratreactive$YuSDE$`Adjusted p-value`, digits = 4)
                       
                       seuratreactive$onlyposSDE <- input$onlyposSDE
                       seuratreactive$logSDE <- input$logSDE
                       seuratreactive$minSDE <- input$minSDE
                       seuratreactive$padjustSDE1 <- input$padjustSDE1
                       seuratreactive$testuseSDE <- input$testuseSDE
                       seuratreactive$enrichInputSDE <- input$enrichInputSDE
                       seuratreactive$pvalueSDE <- input$pvalueSDE
                       seuratreactive$pvalSDE <- input$pvaSDE
                       seuratreactive$padjustSDE <- input$padjustSDE
                       
                       incProgress(10/10)
                     }
                   })
    },
    error = function(e) {
      shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
      seuratreactive$YuSDE <- NULL
    })
  }, ignoreInit = T)
  
  observe({
    if(input$enrichInputSDE == "Gene Ontology") {
      shinyjs::show("GOlabelsSDE")
    }
    else{
      shinyjs::hide("GOlabelsSDE")
    }
  })
  
  output$markersSDEs <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$markersSDE), 'You have not made any comparisons yet... ')
    )
    
    seuratreactive$markersSDE
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$GO_SDEs <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$markersSDE), 'You have not made any comparisons yet... ')
    )
    validate(
      need(!is.null(seuratreactive$YuSDE), 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.')
    )
    validate(
      need(nrow(seuratreactive$YuSDE) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.')
    )
    
    seuratreactive$YuSDE
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  
  observeEvent(input$SelectedCellsGO| input$buttonSDEGDE | input$buttonEnrichSDE | input$buttonvolcanoSDE, {
    tryCatch({
      
      seuratreactive$markersSDEvolcano <- seuratreactive$markersSDE
      
      seuratreactive$markersSDEvolcano$Expression <- "None"
      
      seuratreactive$markersSDEvolcano$Expression[seuratreactive$markersSDEvolcano$`Log2 fold change` > input$logvolcanoSDE[2] & seuratreactive$markersSDEvolcano$`Adjusted p-value` <= input$pvolcanoSDE] <- "UPREGULATED"
      
      seuratreactive$markersSDEvolcano$Expression[seuratreactive$markersSDEvolcano$`Log2 fold change` < input$logvolcanoSDE[1] & seuratreactive$markersSDEvolcano$`Adjusted p-value` <= input$pvolcanoSDE] <- "DOWNREGULATED"
      
      seuratreactive$volcanoSDEvalue1 <- input$logvolcanoSDE[1]
      
      seuratreactive$volcanoSDEvalue2 <- input$logvolcanoSDE[2]
      
      seuratreactive$pvaluevolcanoSDE <- input$pvolcanoSDE
      
      seuratreactive$Subset_GENESSDE <- NULL
      
    },
    error = function(e) {
      shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
      seuratreactive$YuSDE <- NULL
    }) 
    
  }, ignoreInit = T)
  
  output$volcanoplotSDE <- renderPlotly({
    validate(
      need(!is.null(seuratreactive$markersSDEvolcano), 'You have not made any comparisons yet... '))
    
    p <- ggplot(data=seuratreactive$markersSDEvolcano, aes(x= `Log2 fold change`, y=-log10(`Adjusted p-value`), col = `Expression`,
                                                           text = paste0("Gene Name:", `Gene Name`, "\n",
                                                                         "Log2 fold change:", `Log2 fold change`, "\n",
                                                                         "Adjusted p-value:", `Adjusted p-value`, "\n",
                                                                         "Expression:", `Expression`)
                                                           
    )) +
      geom_point(size = input$sizevolcanoSDE, alpha = input$opacityvolcanoSDE) +
      theme_minimal() +
      theme(legend.position="none", text = element_text(size=input$labelsizevolcanoSDE)) +
      ylab("-log10(p-value)") +
      xlab("Log2 fold change") +
      scale_color_manual(values=c("blue", "black", "red")) +
      geom_vline(xintercept=c(seuratreactive$volcanoSDEvalue1, seuratreactive$volcanoSDEvalue2), col="red", linetype = "dotted", size = 1.5) +
      geom_hline(yintercept=-log10(seuratreactive$pvaluevolcanoSDE), col="blue", linetype = "dashed", size = 1.5)
    
    ggplotly(p, tooltip = "text")
  })
  
  observe({
    if (!is.null(seuratreactive$msig$gs_name)) {
      updateVirtualSelect('GPInputSDE', choices = sample(seuratreactive$msig$gs_name))
    }
  })
  
  observeEvent(input$GP_actionsSDE, {
    tryCatch({
      Gene_vector_msig <- seuratreactive$msig[match(input$GPInputSDE, seuratreactive$msig$gs_name),]
      
      Gene_vector_msig <- Gene_vector_msig[,-1] %>% as.character() %>% na.omit() %>% unique()
      
      Subset_GENES <- seuratreactive$markersSDE[seuratreactive$markersSDE$`Gene Name` %in% Gene_vector_msig,]
      
      seuratreactive$Subset_GENESSDE <- arrange(Subset_GENES, Subset_GENES$`Log2 fold change`)
      
      seuratreactive$titleSDE <- input$GPInputSDE
    },
    error = function(e) {
      shinyalert("ERROR!", "Please select a pathway to look at.", type = "error", confirmButtonCol = "#337ab7")
      
    }) 
    
  })
  
  
  output$GP_plotSDE <- renderPlotly({
    validate(
      need(input$GP_actionsSDE, 'Please select a pathway to look at...'))
    validate(
      need(length(rownames(seuratreactive$Subset_GENESSDE)) > 0, 'No expression of genes in selected pathway. Consider easing p-value thresholds and/or logfc thresholds'))
    tryCatch({
      
      q <- ggplot(seuratreactive$Subset_GENESSDE, aes(x=reorder(seuratreactive$Subset_GENESSDE$`Gene Name`, seuratreactive$Subset_GENESSDE$`Log2 fold change`), y=seuratreactive$Subset_GENESSDE$`Log2 fold change`,
                                                      text = paste0("Gene Name: ", `Gene Name`, "\n",
                                                                    "Log2 fold change: ", `Log2 fold change`, "\n",
                                                                    "Adjusted p-value: ", `Adjusted p-value`))) +
        geom_point(size = input$sizeGPSDE, alpha = input$opacityGPSDE, col = "red") +
        theme_minimal() +
        theme(legend.position="none", text = element_text(size=input$labelsizeGPSDE)) +
        ylab("Log2 fold change") +
        xlab("") +
        ggtitle(seuratreactive$titleSDE) +
        theme(axis.text.x = element_blank(),
              plot.title = element_text(vjust=0.5)) +
        geom_hline(yintercept=-log10(1), size = 1)
      
      ggplotly(q, tooltip = "text")
    },
    error = function(e) {
    }) 
  })
  
  observeEvent(input$SelectedCellsGO| input$buttonSDEGDE | input$buttonEnrichSDE | input$buttonvolcanoSDE, {
    if(!is.null(seuratreactive$markersSDE)) {
      if(length(rownames(seuratreactive$markersSDE)) < 1) {
        shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
        seuratreactive$YuSDE <- NULL
      }
    }
    
    if (!is.null(seuratreactive$selectedkey1)) {
      seuratreactive$SPAchoose <- NULL
      seuratreactive$padjustSPA <- NULL
      seuratreactive$pvalueSPA <- NULL
      seuratreactive$GOlabelsSPA <- NULL
      YuSPA$result <- NULL
      YuSPA$ema <- NULL
      seuratreactive$geneListSPA <- NULL
      
      if (input$iscollapseboxSPA == TRUE) {
        js$collapse("SPABoxIntro")
      }
      shinyjs::hide("SPABox1")
      shinyjs::hide("SPABox2")
      shinyjs::hide("SPABox3")
      shinyjs::hide("SPABox4")
      shinyjs::hide("SPABox5")
      
      if (input$iscollapseboxDRUG2 == TRUE) {
        js$collapse("DRUGBoxIntro2")
      }
      shinyjs::hide("DRUGBox12")
      shinyjs::hide("DRUGBox22")
      
      shinyjs::disable("startbuttonSDE")
      shinyjs::show("NextButtonSDE")
      
      if (is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      else if (is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE)& !is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE)& is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      
      else if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                    badgeColor = "green", badgeLabel = "new"),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      else if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        if (is.null(seuratreactive$VFintegrated)) {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
        else {
          output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                           menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                           menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                           menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                           menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                           menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                           menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                           menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                           menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                           menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                           menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                           menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                           menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                           menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                           menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                           menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                           menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                           menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                           menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                           menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                           menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
        }
      }
      
      else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & is.null(seuratreactive$markersSDE)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                         menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = T),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
  }, ignoreInit = T)
  
  
  
  observeEvent(input$NextButtonSDE, {
    
    #SPA
    seuratreactive$SPAchoose <- NULL
    seuratreactive$padjustSPA <- NULL
    seuratreactive$pvalueSPA <- NULL
    seuratreactive$GOlabelsSPA <- NULL
    YuSPA$result <- NULL
    YuSPA$ema <- NULL
    seuratreactive$geneListSPA <- NULL
    
    if (input$iscollapseboxSPA == TRUE) {
      js$collapse("SPABoxIntro")
    }
    shinyjs::hide("SPABox1")
    shinyjs::hide("SPABox2")
    shinyjs::hide("SPABox3")
    shinyjs::hide("SPABox4")
    shinyjs::hide("SPABox5")
    
    shinyjs::disable("startbuttonSDE")
    shinyjs::hide("NextButtonSDE")
    
    if (is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
      if (is.null(seuratreactive$VFintegrated)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
    
    else if (!is.null(seuratreactive$objQC2) & is.null(seuratreactive$markersDE)& !is.null(seuratreactive$markersSDE)) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                       menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                       menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                       menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                       menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                       menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                       menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                       menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                       menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                       menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                       menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                       menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                       menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
    }
    
    else if (is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
      if (is.null(seuratreactive$VFintegrated)) {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
      else {
        output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                         menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                         menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                         menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                         menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                         menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                         menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                         menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                         menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                         menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                         menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                         menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                         menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                         menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                         menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                         menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                         menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                         menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                         menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                         menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                         menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                  badgeColor = "green", badgeLabel = "new"),
                                                         menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
      }
    }
    
    else if (!is.null(seuratreactive$objQC2) & !is.null(seuratreactive$markersDE) & !is.null(seuratreactive$markersSDE)) {
      output$dynamic_content <- renderMenu(sidebarMenu(id = "tabs",
                                                       menuItem("Load your data", icon = icon("spinner"), tabName = "Load", selected = F),
                                                       menuItem("Quality Control", icon = icon("bell"), tabName = "QC", selected = F),
                                                       menuItem("Doublet Removal", icon = icon("check-double"), tabName = "Doublet", selected = F),
                                                       menuItem("Load 2nd Dataset", icon = icon("handshake"), tabName = "Combine", selected = F),
                                                       menuItem("Quality Control (2nd Dataset)", icon = icon("bell"), tabName = "QC2", selected = F),
                                                       menuItem("Doublet Removal (2nd Dataset)", icon = icon("check-double"), tabName = "Doublet2", selected = F),
                                                       menuItem("Dimensionality Reduction", icon = icon("dashboard"), tabName = "DR2", selected = F),
                                                       menuItem("Clustering", icon = icon("puzzle-piece"), tabName = "Cluster", selected = F),
                                                       menuItem("Data Correction", icon = icon("recycle"), tabName = "CC", selected = F),
                                                       menuItem("Labelling Clusters", icon = icon("tags"), tabName = "Label", selected = F),
                                                       menuItem("Gene Expression", icon = icon("chart-area"), tabName = "GE", selected = F),
                                                       menuItem("Gene Co-Expression", icon = icon("code-branch"), tabName = "MEGENA", selected = F),
                                                       menuItem("Gene Regulatory Networks", icon = icon("circle-nodes"), tabName = "SCENIC", selected = F),
                                                       menuItem("GWAS Associations", icon = icon("city"), tabName = "MAGMA", selected = F),
                                                       menuItem("Trajectory Analysis", icon = icon("chart-line"), tabName = "TA", selected = F),
                                                       menuItem("Cell Communication", icon = icon("comments"), tabName = "Chat", selected = F),
                                                       menuItem("Multimodal Analysis", icon = icon("database"), tabName = "MM", selected = F),
                                                       menuItem("Differential Expression", icon = icon("balance-scale-right"), tabName = "DE", selected = F),
                                                       menuItem("Pathway Analysis", icon = icon("project-diagram"), tabName = "PA", selected = F),
                                                       menuItem("Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG", selected = F),
                                                       menuItem("Custom Differential Expression", icon = icon("balance-scale-left"), tabName = "SDE", selected = F),
                                                       menuItem("Custom Pathway Analysis", icon = icon("project-diagram"), tabName = "SPA", selected = T,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Custom Drug Gene Interactions", icon = icon("capsules"), tabName = "DRUG2", selected = F,
                                                                badgeColor = "green", badgeLabel = "new"),
                                                       menuItem("Logs", icon = icon("newspaper"), tabName = "Logs", selected = F)))
    }
  }, ignoreInit = T)
  
  
  ####TAB 14.5 - PATHWAY ANALYSIS (Differential Expression Analysis)#### 
  
  observeEvent(input$startPA, {
    showModal(modalDialog(
      title = strong("Pathway Analysis Settings"),
      pickerInput("PAchoose", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Choose Enrichment Analysis')), 
                  choices = list("Gene Ontology", "WikiPathways",
                                 "KEGG Pathways", "REACTOME"),
                  selected = "Gene Ontology",
                  width = "100%"),
      pickerInput("GOlabelsPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Select sub-ontology')), 
                  choices = list("BP", "MF",
                                 "CC"),
                  selected = "BP",
                  width = "100%"),
      pickerInput("padjustPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">p-value adjustment method')),
                  choices = list("none", "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                  selected = "fdr",
                  width = "100%"),
      numericInput("pvaluePA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">p-value threshold')), 
                   value = 0.05, min = 0, max = 1, step = 0.001, width = "100%"),
      p(style="text-align: center;", actionBttn("startbuttonPA", "APPLY", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startPA, {
    
    updatePickerInput(session, 'PAchoose', selected = seuratreactive$PAchoose)
    updatePickerInput(session, 'GOlabelsPA', selected = seuratreactive$GOlabelsPA)
    updateNumericInput(session, 'pvaluePA', value = seuratreactive$pvaluePA)
    updatePickerInput(session, 'padjustPA', selected = seuratreactive$padjustPA)
    
    if (!is.null(input$PAchoose)) {
      if(input$PAchoose != "Gene Ontology") {
        updatePickerInput(session, 'GOlabelsPA', choices = "NA")
      }
      else{
        updatePickerInput(session, "GOlabelsPA", choices = list("BP", "MF","CC"))
      }
    }
    
    if (!is.null(input$PAchoose) | !is.null(input$GOlabelsPA) | !is.null(input$pvaluePA) | !is.null(input$padjustPA)) {
      if (is.null(input$PAchoose) | is.null(input$GOlabelsPA) | is.na(input$pvaluePA) | is.null(input$padjustPA)) {
        shinyjs::disable("startbuttonPA")
      }
      else {
        shinyjs::enable("startbuttonPA")
      }
    }
  })
  
  observeEvent(input$startbuttonPA, {
    tryCatch({
      withProgress(message = 'Calculating pathways...',
                   value = 0, {
                     
                     if (!is.null(seuratreactive$PATHWAY)) {
                       incProgress(1/10)
                       
                       removeModal()
                       
                       seuratreactive$geneListPA = seuratreactive$PATHWAY$`Log2 fold change`
                       
                       names(seuratreactive$geneListPA) = seuratreactive$PATHWAY$ENTREZID
                       
                       seuratreactive$geneListPA = sort(seuratreactive$geneListPA, decreasing = TRUE)
                       
                       incProgress(5/10)
                       
                       if (input$PAchoose == "Gene Ontology") {
                         Yu$result <- gseGO(gene          = seuratreactive$geneListPA,
                                            OrgDb         = seuratreactive$orgDb,
                                            ont           = input$GOlabelsPA,
                                            pAdjustMethod = input$padjustPA,
                                            pvalueCutoff  = input$pvaluePA)
                         seuratreactive$GOlabelsPA <- input$GOlabelsPA
                         
                         Yu$result <- clusterProfiler::simplify(Yu$result)
                       }
                       else if (input$PAchoose == "KEGG Pathways") {
                         Yu$result <- gseKEGG(gene         = seuratreactive$geneListPA,
                                              organism     = seuratreactive$EKEGG,
                                              pAdjustMethod = input$padjustPA,
                                              pvalueCutoff  = input$pvaluePA)
                         seuratreactive$GOlabelsPA <- "NA"
                       }
                       else if (input$PAchoose == "WikiPathways") {
                         Yu$result <- gseWP(seuratreactive$geneListPA, 
                                            organism = input$species1, 
                                            pAdjustMethod = input$padjustPA,
                                            pvalueCutoff  = input$pvaluePA)
                         seuratreactive$GOlabelsPA <- "NA"
                       }
                       else if (input$PAchoose == "REACTOME") {
                         Yu$result <- gsePathway(gene = seuratreactive$geneListPA,
                                                 organism = seuratreactive$REACT,
                                                 pAdjustMethod = input$padjustPA,
                                                 pvalueCutoff  = input$pvaluePA)
                         seuratreactive$GOlabelsPA <- "NA"
                       }
                       seuratreactive$PAchoose <- input$PAchoose
                       seuratreactive$padjustPA <- input$padjustPA
                       seuratreactive$pvaluePA <- input$pvaluePA
                       
                       Yu$result <- setReadable(Yu$result, seuratreactive$orgDb, 'ENTREZID')
                       
                       incProgress(10/10)
                       
                       if (!is.null(seuratreactive$PATHWAY)) {
                         if (input$iscollapsebox7 == FALSE) {
                           js$collapse("PABoxIntro")
                         }
                         shinyjs::show("PABox1")
                         shinyjs::show("PABox2")
                         shinyjs::show("PABox3")
                         shinyjs::show("PABox4")
                         shinyjs::show("PABox5")
                         seuratreactive$PAPA <- "none"
                       }
                       
                       shinyjs::hide("NextButtonDE")
                     }
                   })
    },
    error = function(e) {
      shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
    }) 
  }, ignoreInit = T)
  
  Yu <- reactiveValues()
  
  observe({
    if (!is.null(input$PAchoose) | !is.null(input$GOlabelsPA) | !is.null(input$pvaluePA) | !is.null(input$padjustPA)) {
      if (is.null(input$PAchoose) | is.null(input$GOlabelsPA) | is.na(input$pvaluePA) | is.null(input$padjustPA)) {
        shinyjs::disable("startbuttonPA")
      }
      else {
        shinyjs::enable("startbuttonPA")
      }
    }
  })
  
  observe({
    if (is.null(input$PAPA) | !is.null(input$PAPA) && input$PAPA == "") {
      shinyjs::disable("PAPAInput")
    }
    else {
      shinyjs::enable("PAPAInput")
    }
  })
  
  observe({
    if (!is.null(input$PAchoose)) {
      if(input$PAchoose != "Gene Ontology") {
        updatePickerInput(session, 'GOlabelsPA', choices = "NA")
      }
      else{
        updatePickerInput(session, "GOlabelsPA", choices = list("BP", "MF","CC"))
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$PAREACTOME)) {
      updateVirtualSelect('PAPA', choices = names(pathways(seuratreactive$PAREACTOME, "reactome")))
    }
  })
  
  output$dotplotUI <- renderUI({
    plotOutput("dotplot", height = (300 + input$HeightPA_Dot*40))
  })
  
  output$dotplot <- renderPlot({
    validate(
      need(nrow(Yu$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    if (input$radioPA == "Dot Plot") {
      seuratreactive$dotplotPA <- dotplot(Yu$result, showCategory = input$DotPlotNumber, font.size = input$fontPA_Dot)
    }
    else if (input$radioPA == "Ridge Plot") {
      seuratreactive$dotplotPA <- ridgeplot(Yu$result, showCategory = input$DotPlotNumber) +
        theme(text = element_text(size=input$fontPA_Dot),
              axis.text.x= element_text(size = input$fontPA_Dot),
              axis.text.y= element_text(size = input$fontPA_Dot))
    }
    else if (input$radioPA == "Heatmap") {
      seuratreactive$dotplotPA <- heatplot(Yu$result, foldChange=seuratreactive$geneListPA, showCategory=input$DotPlotNumber) +
        theme(axis.text.x = element_text(angle = 90, vjust=0.5),
              text = element_text(size=input$fontPA_Dot))
    }
    seuratreactive$dotplotPA
  })
  
  output$GCplotUI <- renderUI({
    plotOutput("GCplot", height = (300 + input$HeightPA_GC*30))
  })
  
  output$GCplot <- renderPlot({
    validate(
      need(nrow(Yu$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    if (input$checkboxgene_GC == TRUE && input$checkboxcategory_GC == TRUE) {
      seuratreactive$GCplotPA <- cnetplot(Yu$result, node_label="all", showCategory = input$GCNumber,
                                          foldChange = seuratreactive$geneListPA,
                                          cex_label_category = input$categoryfontPA_GC, cex_label_gene = input$genefontPA_GC, cex_category = input$nodePA_GC)
      seuratreactive$GCplotPA
    }
    else if (input$checkboxgene_GC == FALSE && input$checkboxcategory_GC == TRUE) {
      seuratreactive$GCplotPA <- cnetplot(Yu$result, node_label="category", showCategory = input$GCNumber,
                                          foldChange = seuratreactive$geneListPA,
                                          cex_label_category = input$categoryfontPA_GC, cex_label_gene = input$genefontPA_GC, cex_category = input$nodePA_GC)
      seuratreactive$GCplotPA
    }
    else if (input$checkboxgene_GC == TRUE && input$checkboxcategory_GC == FALSE) {
      seuratreactive$GCplotPA <- cnetplot(Yu$result, node_label="gene", showCategory = input$GCNumber,
                                          foldChange = seuratreactive$geneListPA,
                                          cex_label_category = input$categoryfontPA_GC, cex_label_gene = input$genefontPA_GC, cex_category = input$nodePA_GC)
      seuratreactive$GCplotPA
    }
    else if (input$checkboxgene_GC == FALSE && input$checkboxcategory_GC == FALSE) {
      seuratreactive$GCplotPA <- cnetplot(Yu$result, showCategory = input$GCNumber,
                                          foldChange = seuratreactive$geneListPA,
                                          node_label="none")
      seuratreactive$GCplotPA
    }
  })
  
  output$emapplotUI <- renderUI({
    plotOutput("emapplot", height = (300 + input$HeightPA_EM*30))
  })
  
  output$emapplot <- renderPlot({
    validate(
      need(nrow(Yu$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    Yu$ema <- pairwise_termsim(Yu$result)
    
    seuratreactive$emapplotPA <- emapplot(Yu$ema, showCategory = input$EMNumber, cex_label_category = input$fontPA_EM, cex_line = input$linePA_EM, cex_category = input$nodePA_EM)
    
    seuratreactive$emapplotPA
  })
  
  output$treeplotUI <- renderUI({
    plotOutput("treeplot", height = (300 + input$HeightPA_T*30))
  })
  
  output$treeplot <- renderPlot({
    validate(
      need(nrow(Yu$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    Yu$ema <- pairwise_termsim(Yu$result)
    
    seuratreactive$treeplotPA <- treeplot(Yu$ema, fontsize = input$fontPA_T, showCategory = input$TreeNumber, nCluster = input$ClustersPA_T)
    
    seuratreactive$treeplotPA
  })
  
  observeEvent(input$PAPAInput, {
    tryCatch({
      Yu$PAPA <- viewPathway(input$PAPA, 
                             readable = TRUE, 
                             organism = seuratreactive$REACT,
                             foldChange = seuratreactive$geneListPA)
      seuratreactive$PAPA <- input$PAPA
    },
    error = function(e) {
      shinyalert("WARNING!", "Pathway is not expressed, please select another pathway to view.", type = "warning", confirmButtonCol = "#337ab7")
    })
  })
  
  output$pathwayplotUI <- renderUI({
    plotOutput("pathwayplot", height = (300 + input$HeightPA_PP*30))
  })
  
  output$pathwayplot <- renderPlot({
    validate(
      need(!is.null(Yu$PAPA), 'Please choose a pathway to look at...'))
    Yu$PAPA
  })
  
  ####TAB 14.75 - Drug Gene Interactions (Differential Expression Analysis)####
  
  observeEvent(input$startbuttonDRUG, {
    library(igraph)
    library(ggraph)
    
    tryCatch({
      withProgress(message = 'Finding significant Drug Gene interactions...',
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$markersDRUG <- seuratreactive$markersDE[, c("Gene Name", "Log2 fold change", "Adjusted p-value")]
                     
                     GeneDrugs <- read.table("interactions.tsv", fill = T, header = T, sep = "\t")
                     colnames(GeneDrugs)[1] <- "Gene Name"
                     
                     seuratreactive$DRUGdf <- left_join(GeneDrugs, seuratreactive$markersDRUG) %>% na.omit()
                     
                     seuratreactive$DRUGdf <- seuratreactive$DRUGdf[,c("Gene Name", "Log2 fold change", "Adjusted p-value", "drug_claim_name", 
                                                                       "drug_concept_id", "interaction_claim_source", "interaction_types", "interaction_group_score", "PMIDs")]
                     
                     colnames(seuratreactive$DRUGdf) <- c("Gene Name", "Log2 fold change", "Adjusted p-value", "Drug Claim Name", 
                                                          "Drug Concept ID", "Interaction Claim Source", "Interaction types", "Interaction Group Score", "PMIDs")
                     
                     seuratreactive$DRUGdf <- arrange(seuratreactive$DRUGdf, `Adjusted p-value`)
                     
                     incProgress(10/10)
                     
                     js$collapse("DRUGBoxIntro")
                     shinyjs::show("DRUGBox1")
                     shinyjs::show("DRUGBox2")
                     
                     shinyjs::hide("NextButtonDE")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check your starting species is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  observeEvent(input$tabs, {
    if (input$tabs == "DRUG" && seuratreactive$REACT != "human") {
      shinyalert("WARNING!", "Drug-Gene interaction analysis only works for human datasets.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  output$markersDRUG <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    seuratreactive$DRUGdf
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  
  output$DRUGMap <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$DRUGdf), "You must compute your Gene-Drug interactions"))
    
    DRUGmap <- arrange(seuratreactive$DRUGdf, `Adjusted p-value`, `Log2 fold change`)
    
    DRUGmap <- DRUGmap[1:input$DRUGNumber,]
    
    edges <- DRUGmap[,c("Gene Name", "Drug Claim Name", "Log2 fold change")]
    
    nodes1 <- DRUGmap %>%
      group_by(`Gene Name`) %>%
      summarise(node_color = "gene", n = 2) %>%
      drop_na()
    
    nodes2 <- DRUGmap %>%
      group_by(`Drug Claim Name`) %>%
      summarise(node_color = "cluster", n = 2.1) %>%
      drop_na()
    
    colnames(nodes2)[1] <- c("Gene Name")
    
    nodes <- full_join(nodes1, nodes2)
    
    graph = graph_from_data_frame(edges, directed = FALSE, nodes)
    
    if (input$checkboxgene_DRUG == TRUE && input$checkboxcluster_DRUG == TRUE) {
      ggraph(graph, layout = "fr") +
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) +
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG, input$nodeDRUG/2)) +
        geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG, repel=TRUE) +
        geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() +
        guides(edge_width = FALSE, color = FALSE)
    }
    
    else if (input$checkboxgene_DRUG == FALSE && input$checkboxcluster_DRUG == TRUE) {
      ggraph(graph, layout = "fr") +
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) +
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG, input$nodeDRUG/2)) +
        geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() +
        guides(edge_width = FALSE, color = FALSE)
    }
    
    else if (input$checkboxgene_DRUG == TRUE && input$checkboxcluster_DRUG == FALSE) {
      ggraph(graph, layout = "fr") +
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) +
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG, input$nodeDRUG/2)) +
        geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() +
        guides(edge_width = FALSE, color = FALSE)
    }
  })
  
  
  ####TAB 15.5 - PATHWAY ANAlYSIS (Custom Differential Expression Analysis)#### 
  
  observeEvent(input$startSPA, {
    showModal(modalDialog(
      title = strong("Pathway Analysis Settings"),
      pickerInput("SPAchoose", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Choose Enrichment Analysis')), 
                  choices = list("Gene Ontology", "WikiPathways",
                                 "KEGG Pathways", "REACTOME"),
                  selected = "Gene Ontology",
                  width = "100%"),
      pickerInput("GOlabelsSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">Select sub-ontology')), 
                  choices = list("BP", "MF",
                                 "CC"),
                  selected = "BP",
                  width = "100%"),
      pickerInput("padjustSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">p-value adjustment method')),
                  choices = list("none", "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"),
                  selected = "fdr",
                  width = "100%"),
      numericInput("pvalueSPA", h4(HTML('<h4 style = "text-align:justify;color:#000000; margin-top:-60px;">p-value threshold')), 
                   value = 0.05, min = 0, max = 1, step = 0.001, width = "100%"),
      p(style="text-align: center;", actionBttn("startbuttonSPA", "APPLY", size = "lg")),
      
      h4(HTML('<h4 style = "text-align:justify"> Please note that this step is computationally intensive and could take upwards of 30 minutes to complete depending on the size of the dataset(s).')),
      easyClose = TRUE,
      size = "m",
      footer = NULL
    ))
  })
  
  observeEvent(input$startSPA, {
    
    updatePickerInput(session, 'SPAchoose', selected = seuratreactive$SPAchoose)
    updatePickerInput(session, 'GOlabelsSPA', selected = seuratreactive$GOlabelsSPA)
    updateNumericInput(session, 'pvalueSPA', value = seuratreactive$pvalueSPA)
    updatePickerInput(session, 'padjustSPA', selected = seuratreactive$padjustSPA)
    
    if (!is.null(input$SPAchoose)) {
      if(input$SPAchoose != "Gene Ontology") {
        updatePickerInput(session, 'GOlabelsSPA', choices = "NA")
      }
      else{
        updatePickerInput(session, "GOlabelsSPA", choices = list("BP", "MF","CC"))
      }
    }
    
    if (!is.null(input$pvalueSPA)) {
      if (is.null(input$SPAchoose) | is.null(input$GOlabelsSPA) | is.na(input$pvalueSPA) | is.null(input$padjustSPA)) {
        shinyjs::disable("startbuttonSPA")
      }
      else {
        shinyjs::enable("startbuttonSPA")
      }
    }
  })
  
  observeEvent(input$startbuttonSPA, {
    tryCatch({
      withProgress(message = 'Calculating pathways...',
                   value = 0, {
                     if (!is.null(seuratreactive$PATHWAYSDE)) {
                       incProgress(1/10)
                       
                       removeModal()
                       
                       seuratreactive$geneListSPA = seuratreactive$PATHWAYSDE$`Log2 fold change`
                       
                       names(seuratreactive$geneListSPA) = seuratreactive$PATHWAYSDE$ENTREZID
                       
                       seuratreactive$geneListSPA = sort(seuratreactive$geneListSPA, decreasing = TRUE)
                       
                       incProgress(5/10)
                       
                       if (input$SPAchoose == "Gene Ontology") {
                         YuSPA$result <- gseGO(gene          = seuratreactive$geneListSPA,
                                               OrgDb         = seuratreactive$orgDb,
                                               ont           = input$GOlabelsSPA,
                                               pAdjustMethod = input$padjustSPA,
                                               pvalueCutoff  = input$pvalueSPA)
                         
                         YuSPA$result <- clusterProfiler::simplify(YuSPA$result)
                         
                         seuratreactive$GOlabelsSPA <- input$GOlabelsSPA
                       }
                       else if (input$SPAchoose == "KEGG Pathways") {
                         YuSPA$result <- gseKEGG(gene         = seuratreactive$geneListSPA,
                                                 organism     = seuratreactive$EKEGG,
                                                 pAdjustMethod = input$padjustSPA,
                                                 pvalueCutoff  = input$pvalueSPA)
                         seuratreactive$GOlabelsSPA <- "NA"
                       }
                       else if (input$SPAchoose == "WikiPathways") {
                         YuSPA$result <- gseWP(seuratreactive$geneListSPA,
                                               organism = input$species1,
                                               pAdjustMethod = input$padjustSPA,
                                               pvalueCutoff  = input$pvalueSPA)
                         seuratreactive$GOlabelsSPA <- "NA"
                       }
                       else if (input$SPAchoose == "REACTOME") {
                         YuSPA$result <- gsePathway(gene = seuratreactive$geneListSPA,
                                                    organism = seuratreactive$REACT,
                                                    pAdjustMethod = input$padjustSPA,
                                                    pvalueCutoff  = input$pvalueSPA)
                         seuratreactive$GOlabelsSPA <- "NA"
                       }
                       
                       YuSPA$result <- setReadable(YuSPA$result, seuratreactive$orgDb, 'ENTREZID')
                       
                       seuratreactive$SPAchoose <- input$SPAchoose
                       seuratreactive$padjustSPA <- input$padjustSPA
                       seuratreactive$pvalueSPA <- input$pvalueSPA
                       
                       incProgress(10/10)
                       
                       if (!is.null(seuratreactive$PATHWAYSDE)) {
                         js$collapse("SPABoxIntro")
                         shinyjs::show("SPABox1")
                         shinyjs::show("SPABox2")
                         shinyjs::show("SPABox3")
                         shinyjs::show("SPABox4")
                         shinyjs::show("SPABox5")
                         seuratreactive$PAPASPA <- "none"
                       }
                       
                       shinyjs::hide("NextButtonSDE")
                     }
                   })
    },
    error = function(e) {
      shinyalert("WARNING!", "No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.", type = "warning", confirmButtonCol = "#337ab7")
    }) 
  }, ignoreInit = T)
  
  YuSPA <- reactiveValues()
  
  observe({
    if (!is.null(input$pvalueSPA)) {
      if (is.null(input$SPAchoose) | is.null(input$GOlabelsSPA) | is.na(input$pvalueSPA) | is.null(input$padjustSPA)) {
        shinyjs::disable("startbuttonSPA")
      }
      else {
        shinyjs::enable("startbuttonSPA")
      }
    }
  })
  
  observe({
    if (is.null(input$PAPASPA) | !is.null(input$PAPASPA) && input$PAPASPA == "") {
      shinyjs::disable("PAPAInputSPA")
    }
    else {
      shinyjs::enable("PAPAInputSPA")
    }
  })
  
  observe({
    if (!is.null(input$SPAchoose)) {
      if(input$SPAchoose != "Gene Ontology") {
        updatePickerInput(session, 'GOlabelsSPA', choices = "NA")
      }
      else{
        updatePickerInput(session, "GOlabelsSPA", choices = list("BP", "MF","CC"))
      }
    }
  })
  
  observe({
    if (!is.null(seuratreactive$PAREACTOME)) {
      updateVirtualSelect('PAPASPA', choices = names(pathways(seuratreactive$PAREACTOME, "reactome")))
    }
  })
  
  output$dotplotUISPA <- renderUI({
    plotOutput("dotplotSPA", height = (300 + input$HeightSPA_Dot*40))
  })
  
  output$dotplotSPA <- renderPlot({
    validate(
      need(nrow(YuSPA$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    if (input$radioSPA == "Dot Plot") {
      seuratreactive$dotplotSPA <- dotplot(YuSPA$result, showCategory = input$DotPlotNumberSPA, font.size = input$fontSPA_Dot)
    }
    else if (input$radioSPA == "Ridge Plot") {
      seuratreactive$dotplotSPA <- ridgeplot(YuSPA$result, showCategory = input$DotPlotNumberSPA) +
        theme(text = element_text(size=input$fontSPA_Dot),
              axis.text.x= element_text(size = input$fontSPA_Dot),
              axis.text.y= element_text(size = input$fontSPA_Dot))
    }
    else if (input$radioSPA == "Heatmap") {
      seuratreactive$dotplotSPA <- heatplot(YuSPA$result, foldChange=seuratreactive$geneListSPA, showCategory=input$DotPlotNumberSPA) +
        theme(text = element_text(size=input$fontSPA_Dot),
              axis.text.x= element_text(size = input$fontSPA_Dot),
              axis.text.y= element_text(size = input$fontSPA_Dot))
    }
    seuratreactive$dotplotSPA
  })
  
  output$treeplotSPAUI <- renderUI({
    plotOutput("treeplotSPA", height = (300 + input$HeightSPA_T*30))
  })
  
  output$treeplotSPA <- renderPlot({
    validate(
      need(nrow(YuSPA$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    YuSPA$ema <- pairwise_termsim(YuSPA$result)
    
    seuratreactive$treeplotSPA <- treeplot(YuSPA$ema, fontsize = input$fontSPA_T, showCategory = input$TreeNumberSPA, nCluster = input$ClustersSPA_T)
    
    seuratreactive$treeplotSPA
  })
  
  output$GCplotUISPA <- renderUI({
    plotOutput("GCplotSPA", height = (300 + input$HeightSPA_GC*30))
  })
  
  output$GCplotSPA <- renderPlot({
    validate(
      need(nrow(YuSPA$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    if (input$checkboxgene_GCSPA == TRUE && input$checkboxcategory_GCSPA == TRUE) {
      seuratreactive$GCplotSPA <- cnetplot(YuSPA$result, node_label="all", showCategory = input$GCNumberSPA,
                                           foldChange = seuratreactive$geneListSPA,
                                           cex_label_category = input$categoryfontSPA_GC, cex_label_gene = input$genefontSPA_GC, cex_category = input$nodeSPA_GC)
      seuratreactive$GCplotSPA
    }
    else if (input$checkboxgene_GCSPA == FALSE && input$checkboxcategory_GCSPA == TRUE) {
      seuratreactive$GCplotSPA <- cnetplot(YuSPA$result, node_label="category", showCategory = input$GCNumberSPA,
                                           foldChange = seuratreactive$geneListSPA,
                                           cex_label_category = input$categoryfontSPA_GC, cex_label_gene = input$genefontSPA_GC, cex_category = input$nodeSPA_GC)
      seuratreactive$GCplotSPA
    }
    else if (input$checkboxgene_GCSPA == TRUE && input$checkboxcategory_GCSPA == FALSE) {
      seuratreactive$GCplotSPA <- cnetplot(YuSPA$result, node_label="gene", showCategory = input$GCNumberSPA,
                                           foldChange = seuratreactive$geneListSPA,
                                           cex_label_category = input$categoryfontSPA_GC, cex_label_gene = input$genefontSPA_GC, cex_category = input$nodeSPA_GC)
      seuratreactive$GCplotSPA
    }
    else if (input$checkboxgene_GCSPA == FALSE && input$checkboxcategory_GCSPA == FALSE) {
      seuratreactive$GCplotSPA <- cnetplot(YuSPA$result, showCategory = input$GCNumberSPA,
                                           foldChange = seuratreactive$geneListSPA,
                                           node_label="none")
      seuratreactive$GCplotSPA
    }
  })
  
  output$emapplotUISPA <- renderUI({
    plotOutput("emapplotSPA", height = (300 + input$HeightSPA_EM*30))
  })
  
  output$emapplotSPA <- renderPlot({
    validate(
      need(nrow(YuSPA$result) > 0, 'No enriched terms found. Please make sure you have enough variability between the two groups of cells being compared. Try increasing the number of cells compared. Please also check if your starting species is correct.'))
    
    YuSPA$ema <- pairwise_termsim(YuSPA$result)
    seuratreactive$emapplotSPA <- emapplot(YuSPA$ema, showCategory = input$EMNumberSPA, cex_label_category = input$fontSPA_EM, cex_line = input$lineSPA_EM, cex_category = input$nodeSPA_EM)
    seuratreactive$emapplotSPA
  })
  
  observeEvent(input$PAPAInputSPA, {
    tryCatch({
      YuSPA$PAPA <- viewPathway(input$PAPASPA, 
                                readable = TRUE, 
                                organism = seuratreactive$REACT,
                                foldChange = seuratreactive$geneListSPA)
      seuratreactive$PAPASPA <- input$PAPASPA
    },
    error = function(e) {
      shinyalert("WARNING!", "Pathway not expressed, please select another pathway to view.", type = "warning", confirmButtonCol = "#337ab7")
    })
  })
  
  output$pathwayplotSPAUI <- renderUI({
    plotOutput("pathwayplotSPA", height = (300 + input$HeightSPA_PP*30))
  })
  
  output$pathwayplotSPA <- renderPlot({
    validate(
      need(!is.null(YuSPA$PAPA), 'Please choose a pathway to look at...'))
    YuSPA$PAPA
  }) 
  
  
  ####TAB 15.75 - Drug Gene Interactions (Custom Differential Expression Analysis)####
  
  observeEvent(input$startbuttonDRUG2, {
    library(igraph)
    library(ggraph)
    
    tryCatch({
      withProgress(message = 'Finding significant Drug Gene interactions...',
                   value = 0, {
                     incProgress(1/10)
                     
                     seuratreactive$markersDRUG2 <- seuratreactive$markersSDE[, c("Gene Name", "Log2 fold change", "Adjusted p-value")]
                     
                     GeneDrugs <- read.table("interactions.tsv", fill = T, header = T, sep = "\t")
                     colnames(GeneDrugs)[1] <- "Gene Name"
                     
                     seuratreactive$DRUGdf2 <- left_join(GeneDrugs, seuratreactive$markersDRUG2) %>% na.omit()
                     
                     seuratreactive$DRUGdf2 <- seuratreactive$DRUGdf2[,c("Gene Name", "Log2 fold change", "Adjusted p-value", "drug_claim_name", 
                                                                         "drug_concept_id", "interaction_claim_source", "interaction_types", "interaction_group_score", "PMIDs")]
                     
                     colnames(seuratreactive$DRUGdf2) <- c("Gene Name", "Log2 fold change", "Adjusted p-value", "Drug Claim Name", 
                                                           "Drug Concept ID", "Interaction Claim Source", "Interaction types", "Interaction Group Score", "PMIDs")
                     
                     seuratreactive$DRUGdf2 <- arrange(seuratreactive$DRUGdf2, `Adjusted p-value`)
                     
                     incProgress(10/10)
                     
                     js$collapse("DRUGBoxIntro2")
                     shinyjs::show("DRUGBox12")
                     shinyjs::show("DRUGBox22")
                     
                     shinyjs::hide("NextButtonSDE")
                     
                   })
    },
    error = function(e) {
      shinyalert("ERROR!", "Please check your starting species is correct.", type = "error", confirmButtonCol = "#337ab7")
    })
  }, ignoreInit = T)
  
  observeEvent(input$tabs, {
    if (input$tabs == "DRUG2" && seuratreactive$REACT != "human") {
      shinyalert("WARNING!", "Drug-Gene interaction analysis only works for human datasets.", type = "warning", confirmButtonCol = "#337ab7")      
    }
  })
  
  output$markersDRUG2 <- DT::renderDataTable({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    
    seuratreactive$DRUGdf2
    
  }, options = list(scrollX = TRUE), rownames= FALSE)
  
  output$DRUGMap2 <- renderPlot({
    validate(
      need(!is.null(seuratreactive$plot.data), "You must complete the clustering step"))
    validate(
      need(!is.null(seuratreactive$DRUGdf2), 'Please compute your Gene-Drug interactions'))
    
    DRUGmap <- arrange(seuratreactive$DRUGdf2, `Adjusted p-value`, `Log2 fold change`)
    
    DRUGmap <- DRUGmap[1:input$DRUGNumber2,]
    
    edges <- DRUGmap[,c("Gene Name", "Drug Claim Name", "Log2 fold change")]
    
    nodes1 <- DRUGmap %>%
      group_by(`Gene Name`) %>%
      summarise(node_color = "gene", n = 2) %>% 
      drop_na()
    
    nodes2 <- DRUGmap %>%
      group_by(`Drug Claim Name`) %>%
      summarise(node_color = "cluster", n = 2.1) %>% 
      drop_na()
    
    colnames(nodes2)[1] <- c("Gene Name")
    
    nodes <- full_join(nodes1, nodes2)
    
    graph = graph_from_data_frame(edges, directed = FALSE, nodes)
    
    if (input$checkboxgene_DRUG2 == TRUE && input$checkboxcluster_DRUG2 == TRUE) {
      ggraph(graph, layout = "fr") + 
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG2, input$nodeDRUG2/2)) + 
        geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG2, repel=TRUE) +
        geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG2, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() + 
        guides(edge_width = FALSE, color = FALSE)
    }
    
    else if (input$checkboxgene_DRUG2 == FALSE && input$checkboxcluster_DRUG2 == TRUE) {
      ggraph(graph, layout = "fr") + 
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG2, input$nodeDRUG2/2)) + 
        geom_node_text(aes(label = ifelse(node_color == "cluster", name, NA_character_)), size = input$categoryfontDRUG2, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() + 
        guides(edge_width = FALSE, color = FALSE)
    }
    
    else if (input$checkboxgene_DRUG2 == TRUE && input$checkboxcluster_DRUG2 == FALSE) {
      ggraph(graph, layout = "fr") + 
        geom_edge_link(aes(width = 1, color = `Log2 fold change`)) + 
        scale_edge_width(range = c(1, 2))+
        geom_node_point(aes(color = node_color), size = ifelse(nodes$node_color == "cluster", input$nodeDRUG2, input$nodeDRUG2/2)) + 
        geom_node_text(aes(label = ifelse(node_color == "gene", name, NA_character_)), size = input$genefontDRUG2, repel=TRUE) +
        scale_edge_color_gradient2(low="blue", high="red", mid = "white") +
        scale_color_manual(values = c("grey", "green")) +
        theme_graph() + 
        guides(edge_width = FALSE, color = FALSE)
    }
  })
  
}

shinyApp(ui, server)